{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "L56fjA27DmhA"
   },
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "t0BTCsz7Y48y",
    "outputId": "f6510678-f977-4ec8-ce30-1f2acd4fe666"
   },
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import os\n",
    "import dill\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "import sys\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Download and describe data\n",
    "\n",
    "This notebook collects data from the [Han et al. 2018](https://www.sciencedirect.com/science/article/pii/S0092867418301168#sec4) mouse cell atlas.\n",
    "\n",
    "## Dataset description\n",
    "\n",
    "- Data can be found at the [GEO Accession GSE108097](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE108097).\n",
    "- This notebook previously collects data from one sample, [\"Brain1\"](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM2906405), for the purposes of data analysis and exploration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  815M  100  815M    0     0  37.4M      0  0:00:21  0:00:21 --:--:-- 39.2M\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'version' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_49093/1618536247.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     38\u001b[0m     \u001b[0msample_dict\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mspecies_sample_dict\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     39\u001b[0m     \u001b[0mfileclass\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgenome_fasta_file\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 40\u001b[0;31m     \u001b[0mversion\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgenome_version\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     41\u001b[0m )\n\u001b[1;32m     42\u001b[0m annot = url_download_biofile(\n",
      "\u001b[0;32m~/glial-origins/utils/biofile_handling.py\u001b[0m in \u001b[0;36murl_download_biofile\u001b[0;34m(url, protocol, sample_dict, fileclass, **kwargs)\u001b[0m\n\u001b[1;32m    155\u001b[0m             \u001b[0msubprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'gunzip'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutput_loc\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    156\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 157\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mfileclass\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0munzip_filename\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_dict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    158\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    159\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mprotocol\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'curl'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/glial-origins/utils/biofile_handling.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, filename, sample_dict, **kwargs)\u001b[0m\n\u001b[1;32m     46\u001b[0m         \u001b[0mbiofile\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__init__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfilename\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_dict\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     47\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0ms3uri\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m's3://arcadia-reference-datasets/organisms/'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspecies\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'/genomics_reference/genome/'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfilename\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 48\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mversion\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mversion\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     49\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     50\u001b[0m \u001b[0;32mclass\u001b[0m \u001b[0mtransdecoder_cdna_file\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbiofile\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'version' is not defined"
     ]
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify the name of the species folder in Amazon S3\n",
    "species = 'Mus_musculus'\n",
    "\n",
    "# Specify any particular identifying conditions, eg tissue type:\n",
    "conditions = 'adultbrain'\n",
    "\n",
    "# Specify url and other variables\n",
    "genome_fasta_url = 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/635/GCF_000001635.23_GRCm38.p3/GCF_000001635.23_GRCm38.p3_genomic.fna.gz'\n",
    "genome_version = 'GRCm38.p3'\n",
    "\n",
    "annot_url = 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/635/GCF_000001635.23_GRCm38.p3/GCF_000001635.23_GRCm38.p3_genomic.gff.gz'\n",
    "gxc_url = 'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM2906nnn/GSM2906405/suppl/GSM2906405_Brain1_dge.txt.gz'\n",
    "\n",
    "###########\n",
    "# runtime #\n",
    "###########\n",
    "\n",
    "protocol = 'curl'\n",
    "\n",
    "species_prefix = prefixify(species)\n",
    "\n",
    "# Specify folder as destination for file downloads\n",
    "output_folder = '../../output/' + species_prefix + '_' + conditions + '/'\n",
    "\n",
    "if not os.path.exists(output_folder):\n",
    "    os.mkdir(output_folder)\n",
    "    \n",
    "species_sample_dict = sample_dict(species, conditions, output_folder)\n",
    "\n",
    "\n",
    "genome_fasta = url_download_biofile(\n",
    "    url = genome_fasta_url,\n",
    "    protocol = protocol, \n",
    "    sample_dict = species_sample_dict,\n",
    "    fileclass = genome_fasta_file,\n",
    "    version = genome_version\n",
    ")\n",
    "annot = url_download_biofile(\n",
    "    url = annot_url, \n",
    "    protocol = protocol,\n",
    "    sample_dict = species_sample_dict,\n",
    "    fileclass = genome_gff_file,\n",
    "    genome_fasta_file = genome_fasta\n",
    ")\n",
    "gxc = url_download_biofile(\n",
    "    url = gxc_url, \n",
    "    protocol = protocol,\n",
    "    sample_dict = species_sample_dict,\n",
    "    fileclass = gxc_file,\n",
    "    genome_fasta_file = genome_fasta\n",
    ")\n",
    "\n",
    "sample_docket = docket(species_sample_dict)\n",
    "keyfiles = {\n",
    "    'annot': annot,\n",
    "    'genome_fasta': genome_fasta,\n",
    "    'gxc': gxc\n",
    "}\n",
    "sample_docket.add_keyfiles(keyfiles)\n",
    "\n",
    "display(vars(sample_docket))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Load in the gxc matrix and get gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Brain_1.CCGCTAAATAAATAAGGG</th>\n",
       "      <th>Brain_1.AACGCCGATCTTGCCCTC</th>\n",
       "      <th>Brain_1.ACCTGAAGTTTATCGTAA</th>\n",
       "      <th>Brain_1.CTCGCACTGAAACCGCTA</th>\n",
       "      <th>Brain_1.ATCAACATCTCTTCGGGT</th>\n",
       "      <th>Brain_1.GCGAATAGGGTCTATGTA</th>\n",
       "      <th>Brain_1.CGAGTAAGGGTCTAGTCG</th>\n",
       "      <th>Brain_1.ATCTCTTCGTAAGTTGCC</th>\n",
       "      <th>Brain_1.AACGCCTAAGGGCTCGCA</th>\n",
       "      <th>Brain_1.AACGCCTCACTTATACAG</th>\n",
       "      <th>...</th>\n",
       "      <th>Brain_1.TTGGACGCCTAGGAGATC</th>\n",
       "      <th>Brain_1.TTAACTAAAGTTTATGTA</th>\n",
       "      <th>Brain_1.GTCCCGGGACATAGGACT</th>\n",
       "      <th>Brain_1.CCGCTAGGGTTTGCTCAA</th>\n",
       "      <th>Brain_1.TGATCAGCTGTGTCAAAG</th>\n",
       "      <th>Brain_1.TCACTTGAATTATGAAGC</th>\n",
       "      <th>Brain_1.AAGTACGCTGTGTATGTA</th>\n",
       "      <th>Brain_1.CCTAGATAGAGAATTTGC</th>\n",
       "      <th>Brain_1.CATCCCATTTGCGGCTGC</th>\n",
       "      <th>Brain_1.CAAAGTGGGTTTAGCGAG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0610005C13Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610007P14Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009E02Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n-R5s28</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n-R5s37</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n-R5s52</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n-R5s88</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>n-R5s98</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20165 rows × 3679 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Brain_1.CCGCTAAATAAATAAGGG  Brain_1.AACGCCGATCTTGCCCTC  \\\n",
       "0610005C13Rik                           0                           0   \n",
       "0610007P14Rik                           0                           3   \n",
       "0610009B22Rik                           0                           3   \n",
       "0610009E02Rik                           0                           0   \n",
       "0610009L18Rik                           0                           0   \n",
       "...                                   ...                         ...   \n",
       "n-R5s28                                 0                           0   \n",
       "n-R5s37                                 0                           0   \n",
       "n-R5s52                                 0                           0   \n",
       "n-R5s88                                 0                           0   \n",
       "n-R5s98                                 0                           0   \n",
       "\n",
       "               Brain_1.ACCTGAAGTTTATCGTAA  Brain_1.CTCGCACTGAAACCGCTA  \\\n",
       "0610005C13Rik                           0                           0   \n",
       "0610007P14Rik                           1                           2   \n",
       "0610009B22Rik                           0                           2   \n",
       "0610009E02Rik                           0                           0   \n",
       "0610009L18Rik                           0                           0   \n",
       "...                                   ...                         ...   \n",
       "n-R5s28                                 0                           0   \n",
       "n-R5s37                                 0                           0   \n",
       "n-R5s52                                 0                           0   \n",
       "n-R5s88                                 0                           0   \n",
       "n-R5s98                                 0                           0   \n",
       "\n",
       "               Brain_1.ATCAACATCTCTTCGGGT  Brain_1.GCGAATAGGGTCTATGTA  \\\n",
       "0610005C13Rik                           0                           0   \n",
       "0610007P14Rik                           0                           0   \n",
       "0610009B22Rik                           1                           0   \n",
       "0610009E02Rik                           0                           0   \n",
       "0610009L18Rik                           2                           0   \n",
       "...                                   ...                         ...   \n",
       "n-R5s28                                 0                           0   \n",
       "n-R5s37                                 0                           0   \n",
       "n-R5s52                                 0                           0   \n",
       "n-R5s88                                 0                           0   \n",
       "n-R5s98                                 0                           0   \n",
       "\n",
       "               Brain_1.CGAGTAAGGGTCTAGTCG  Brain_1.ATCTCTTCGTAAGTTGCC  \\\n",
       "0610005C13Rik                           0                           0   \n",
       "0610007P14Rik                           1                           2   \n",
       "0610009B22Rik                           0                           0   \n",
       "0610009E02Rik                           0                           0   \n",
       "0610009L18Rik                           0                           0   \n",
       "...                                   ...                         ...   \n",
       "n-R5s28                                 0                           0   \n",
       "n-R5s37                                 0                           0   \n",
       "n-R5s52                                 0                           0   \n",
       "n-R5s88                                 0                           0   \n",
       "n-R5s98                                 0                           0   \n",
       "\n",
       "               Brain_1.AACGCCTAAGGGCTCGCA  Brain_1.AACGCCTCACTTATACAG  ...  \\\n",
       "0610005C13Rik                           0                           0  ...   \n",
       "0610007P14Rik                           0                           0  ...   \n",
       "0610009B22Rik                           0                           0  ...   \n",
       "0610009E02Rik                           0                           0  ...   \n",
       "0610009L18Rik                           0                           0  ...   \n",
       "...                                   ...                         ...  ...   \n",
       "n-R5s28                                 0                           0  ...   \n",
       "n-R5s37                                 0                           0  ...   \n",
       "n-R5s52                                 0                           0  ...   \n",
       "n-R5s88                                 0                           0  ...   \n",
       "n-R5s98                                 0                           0  ...   \n",
       "\n",
       "               Brain_1.TTGGACGCCTAGGAGATC  Brain_1.TTAACTAAAGTTTATGTA  \\\n",
       "0610005C13Rik                           0                           0   \n",
       "0610007P14Rik                           0                           0   \n",
       "0610009B22Rik                           0                           0   \n",
       "0610009E02Rik                           0                           0   \n",
       "0610009L18Rik                           0                           0   \n",
       "...                                   ...                         ...   \n",
       "n-R5s28                                 0                           0   \n",
       "n-R5s37                                 0                           0   \n",
       "n-R5s52                                 0                           0   \n",
       "n-R5s88                                 0                           0   \n",
       "n-R5s98                                 0                           0   \n",
       "\n",
       "               Brain_1.GTCCCGGGACATAGGACT  Brain_1.CCGCTAGGGTTTGCTCAA  \\\n",
       "0610005C13Rik                           0                           0   \n",
       "0610007P14Rik                           0                           0   \n",
       "0610009B22Rik                           0                           0   \n",
       "0610009E02Rik                           0                           0   \n",
       "0610009L18Rik                           0                           0   \n",
       "...                                   ...                         ...   \n",
       "n-R5s28                                 0                           0   \n",
       "n-R5s37                                 0                           0   \n",
       "n-R5s52                                 0                           0   \n",
       "n-R5s88                                 0                           0   \n",
       "n-R5s98                                 0                           0   \n",
       "\n",
       "               Brain_1.TGATCAGCTGTGTCAAAG  Brain_1.TCACTTGAATTATGAAGC  \\\n",
       "0610005C13Rik                           0                           0   \n",
       "0610007P14Rik                           0                           0   \n",
       "0610009B22Rik                           0                           0   \n",
       "0610009E02Rik                           0                           0   \n",
       "0610009L18Rik                           0                           0   \n",
       "...                                   ...                         ...   \n",
       "n-R5s28                                 0                           0   \n",
       "n-R5s37                                 0                           0   \n",
       "n-R5s52                                 0                           0   \n",
       "n-R5s88                                 0                           0   \n",
       "n-R5s98                                 0                           0   \n",
       "\n",
       "               Brain_1.AAGTACGCTGTGTATGTA  Brain_1.CCTAGATAGAGAATTTGC  \\\n",
       "0610005C13Rik                           0                           0   \n",
       "0610007P14Rik                           0                           0   \n",
       "0610009B22Rik                           1                           0   \n",
       "0610009E02Rik                           0                           0   \n",
       "0610009L18Rik                           0                           1   \n",
       "...                                   ...                         ...   \n",
       "n-R5s28                                 0                           0   \n",
       "n-R5s37                                 0                           0   \n",
       "n-R5s52                                 0                           0   \n",
       "n-R5s88                                 0                           0   \n",
       "n-R5s98                                 0                           0   \n",
       "\n",
       "               Brain_1.CATCCCATTTGCGGCTGC  Brain_1.CAAAGTGGGTTTAGCGAG  \n",
       "0610005C13Rik                           0                           0  \n",
       "0610007P14Rik                           0                           0  \n",
       "0610009B22Rik                           0                           0  \n",
       "0610009E02Rik                           0                           0  \n",
       "0610009L18Rik                           0                           0  \n",
       "...                                   ...                         ...  \n",
       "n-R5s28                                 0                           0  \n",
       "n-R5s37                                 0                           0  \n",
       "n-R5s52                                 0                           0  \n",
       "n-R5s88                                 0                           0  \n",
       "n-R5s98                                 0                           0  \n",
       "\n",
       "[20165 rows x 3679 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0610005C13Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0610007P14Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0610009B22Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0610009E02Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0610009L18Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20160</th>\n",
       "      <td>n-R5s28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20161</th>\n",
       "      <td>n-R5s37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20162</th>\n",
       "      <td>n-R5s52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20163</th>\n",
       "      <td>n-R5s88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20164</th>\n",
       "      <td>n-R5s98</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20165 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_name\n",
       "0      0610005C13Rik\n",
       "1      0610007P14Rik\n",
       "2      0610009B22Rik\n",
       "3      0610009E02Rik\n",
       "4      0610009L18Rik\n",
       "...              ...\n",
       "20160        n-R5s28\n",
       "20161        n-R5s37\n",
       "20162        n-R5s52\n",
       "20163        n-R5s88\n",
       "20164        n-R5s98\n",
       "\n",
       "[20165 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes_matrix = pd.read_csv(sample_docket.gxc.path, sep = ' ')\n",
    "display(genes_matrix)\n",
    "\n",
    "genes_list = pd.DataFrame({'gene_name':genes_matrix.index})\n",
    "display(genes_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jvfcFNT2FuaZ"
   },
   "source": [
    "# 4. Convert GFF genome file to GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1.0</td>\n",
       "      <td>195471971.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id0;Dbxref=taxon:10090;Name=1;chromosome=1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>BestRefSeq%2CGnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>3199731.0</td>\n",
       "      <td>3671742.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Dbxref=GeneID:497097,MGI:MGI:3528744;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3199731.0</td>\n",
       "      <td>3671742.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Dbxref=GeneID:497097,Genb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552.0</td>\n",
       "      <td>3671742.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Dbxref=GeneID:497097,Genban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702.0</td>\n",
       "      <td>3421901.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Dbxref=GeneID:497097,Genban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284598</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>15515.0</td>\n",
       "      <td>15558.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169987;Note=ETAS2%3B extended terminatio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284599</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>16035.0</td>\n",
       "      <td>16058.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169988;Note=CSB1%3B conserved sequencing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284600</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>16089.0</td>\n",
       "      <td>16104.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169989;Note=CSB2%3B conserved sequencing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284601</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>16114.0</td>\n",
       "      <td>16131.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169990;Note=CSB3%3B conserved sequencing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284602</th>\n",
       "      <td>###</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2284603 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0                    1       2          3            4  \\\n",
       "0        NC_000067.6               RefSeq  region        1.0  195471971.0   \n",
       "1        NC_000067.6  BestRefSeq%2CGnomon    gene  3199731.0    3671742.0   \n",
       "2        NC_000067.6               Gnomon    mRNA  3199731.0    3671742.0   \n",
       "3        NC_000067.6               Gnomon    exon  3670552.0    3671742.0   \n",
       "4        NC_000067.6               Gnomon    exon  3421702.0    3421901.0   \n",
       "...              ...                  ...     ...        ...          ...   \n",
       "2284598  NC_005089.1               RefSeq  region    15515.0      15558.0   \n",
       "2284599  NC_005089.1               RefSeq  region    16035.0      16058.0   \n",
       "2284600  NC_005089.1               RefSeq  region    16089.0      16104.0   \n",
       "2284601  NC_005089.1               RefSeq  region    16114.0      16131.0   \n",
       "2284602          ###                  NaN     NaN        NaN          NaN   \n",
       "\n",
       "           5    6    7                                                  8  \n",
       "0          .    +    .  ID=id0;Dbxref=taxon:10090;Name=1;chromosome=1;...  \n",
       "1          .    -    .  ID=gene0;Dbxref=GeneID:497097,MGI:MGI:3528744;...  \n",
       "2          .    -    .  ID=rna0;Parent=gene0;Dbxref=GeneID:497097,Genb...  \n",
       "3          .    -    .  ID=id1;Parent=rna0;Dbxref=GeneID:497097,Genban...  \n",
       "4          .    -    .  ID=id2;Parent=rna0;Dbxref=GeneID:497097,Genban...  \n",
       "...      ...  ...  ...                                                ...  \n",
       "2284598    .    +    .  ID=id1169987;Note=ETAS2%3B extended terminatio...  \n",
       "2284599    .    +    .  ID=id1169988;Note=CSB1%3B conserved sequencing...  \n",
       "2284600    .    +    .  ID=id1169989;Note=CSB2%3B conserved sequencing...  \n",
       "2284601    .    +    .  ID=id1169990;Note=CSB3%3B conserved sequencing...  \n",
       "2284602  NaN  NaN  NaN                                                NaN  \n",
       "\n",
       "[2284603 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'ID=id1;Parent=rna0;Dbxref=GeneID:497097,Genbank:XM_006495550.2,MGI:MGI:3528744;gbkey=mRNA;gene=Xkr4;product=X Kell blood group precursor related family member 4%2C transcript variant X1;transcript_id=XM_006495550.2'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load in the original GFF-based annotation\n",
    "models = pd.read_csv(sample_docket.annot.path, skiprows = 7, header = None, sep = '\\t')\n",
    "display(models)\n",
    "\n",
    "attributes_column = 8\n",
    "\n",
    "# Check the structure of fields in the Ensembl additional fields section\n",
    "display(models[attributes_column][3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning: could not find the corresponding gene segment for CDS chain 40891296-40891885 of gene gene14227\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41047556-41047995 of gene gene14234\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41048394-41048824 of gene gene14235\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41059443-41059886 of gene gene14236\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41062359-41062803 of gene gene14237\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41080253-41080654 of gene gene14239\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41084669-41085107 of gene gene14240\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41106805-41107303 of gene gene14243\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41113567-41114067 of gene gene14244\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41116036-41116468 of gene gene14245\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41118864-41119364 of gene gene14246\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41121396-41121832 of gene gene14247\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41127592-41128076 of gene gene14248\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41130147-41130585 of gene gene14249\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41135167-41135616 of gene gene14250\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41141211-41141658 of gene gene14251\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41151791-41152230 of gene gene14252\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41163092-41163556 of gene gene14253\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41175137-41175608 of gene gene14254\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41178589-41179048 of gene gene14255\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41188273-41188977 of gene gene14256\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41202587-41203044 of gene gene14257\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41212075-41212534 of gene gene14258\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41216073-41216526 of gene gene14259\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41218091-41218557 of gene gene14260\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41221674-41222210 of gene gene14261\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41227525-41227913 of gene gene14262\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41258121-41258579 of gene gene14264\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41266697-41267150 of gene gene14265\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41271403-41271881 of gene gene14266\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41281376-41281990 of gene gene14267\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41538219-41539665 of gene gene14287\n",
      "Warning: could not find the corresponding gene segment for CDS chain 41546730-41548181 of gene gene14296\n",
      "Warning: could not find the corresponding gene segment for CDS chain 67555636-67556216 of gene gene14818\n",
      "Warning: could not find the corresponding gene segment for CDS chain 67593876-67594308 of gene gene14820\n",
      "Warning: could not find the corresponding gene segment for CDS chain 67609745-67610508 of gene gene14821\n",
      "Warning: could not find the corresponding gene segment for CDS chain 67711174-67711628 of gene gene14824\n",
      "Warning: could not find the corresponding gene segment for CDS chain 67720729-67721225 of gene gene14826\n",
      "Warning: could not find the corresponding gene segment for CDS chain 67724914-67725661 of gene gene14827\n",
      "Warning: could not find the corresponding gene segment for CDS chain 67759700-67760413 of gene gene14830\n",
      "Warning: could not find the corresponding gene segment for CDS chain 67766036-67766772 of gene gene14831\n",
      "Warning: could not find the corresponding gene segment for CDS chain 67839793-67840266 of gene gene14834\n",
      "Warning: could not find the corresponding gene segment for CDS chain 67847409-67847874 of gene gene14835\n",
      "Warning: could not find the corresponding gene segment for CDS chain 67861159-67861659 of gene gene14836\n",
      "Warning: could not find the corresponding gene segment for CDS chain 67876011-67876477 of gene gene14837\n",
      "Warning: could not find the corresponding gene segment for CDS chain 67896177-67896642 of gene gene14839\n",
      "Warning: could not find the corresponding gene segment for CDS chain 67913573-67914052 of gene gene14840\n",
      "Warning: could not find the corresponding gene segment for CDS chain 67942076-67942540 of gene gene14841\n",
      "Warning: could not find the corresponding gene segment for CDS chain 67954230-67954690 of gene gene14842\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68016742-68017488 of gene gene14843\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68036818-68037318 of gene gene14844\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68049983-68050454 of gene gene14845\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68056345-68056810 of gene gene14846\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68066369-68066828 of gene gene14847\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68082547-68083145 of gene gene14848\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68105414-68105880 of gene gene14850\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68121091-68121825 of gene gene14851\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68151885-68152617 of gene gene14852\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68160797-68161716 of gene gene14853\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68164459-68164923 of gene gene14854\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68178801-68179564 of gene gene14855\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68219981-68220705 of gene gene14857\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68256404-68256869 of gene gene14858\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68270527-68271265 of gene gene14859\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68302439-68303158 of gene gene14860\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68326098-68326752 of gene gene14862\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68339527-68339995 of gene gene14863\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68348661-68349441 of gene gene14864\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68425605-68426072 of gene gene14866\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68437468-68437925 of gene gene14867\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68466126-68466590 of gene gene14868\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68479082-68479573 of gene gene14870\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68519012-68519477 of gene gene14872\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68541658-68542422 of gene gene14873\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68570763-68571235 of gene gene14874\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68591380-68591832 of gene gene14875\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68631965-68632430 of gene gene14876\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68647999-68648531 of gene gene14877\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68670752-68671335 of gene gene14878\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68680379-68680848 of gene gene14879\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68704508-68704978 of gene gene14880\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68712925-68713508 of gene gene14881\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68736297-68736764 of gene gene14883\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68755038-68755593 of gene gene14884\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68768555-68769103 of gene gene14885\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68807180-68807708 of gene gene14886\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68834846-68835307 of gene gene14888\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68862265-68863031 of gene gene14889\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68902801-68903266 of gene gene14890\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68910411-68910938 of gene gene14891\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68930269-68930734 of gene gene14892\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68939602-68940067 of gene gene14893\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68961378-68962019 of gene gene14894\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68979198-68979663 of gene gene14895\n",
      "Warning: could not find the corresponding gene segment for CDS chain 68990758-68991294 of gene gene14896\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69009143-69009513 of gene gene14897\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69016558-69017080 of gene gene14898\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69042972-69043505 of gene gene14899\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69051444-69051794 of gene gene14900\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69059690-69060224 of gene gene14901\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69110904-69111432 of gene gene14902\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69137866-69138332 of gene gene14903\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69156118-69156659 of gene gene14904\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69177416-69177787 of gene gene14905\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69184826-69185361 of gene gene14906\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69197583-69198116 of gene gene14908\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69226854-69227383 of gene gene14909\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69243160-69243688 of gene gene14911\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69267888-69268412 of gene gene14912\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69283794-69284319 of gene gene14913\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69304834-69305363 of gene gene14914\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69324177-69324646 of gene gene14915\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69334604-69335076 of gene gene14917\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69342472-69343004 of gene gene14918\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69362722-69363187 of gene gene14919\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69377944-69378472 of gene gene14920\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69392439-69392815 of gene gene14921\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69399812-69400344 of gene gene14922\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69409499-69409875 of gene gene14923\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69416893-69417425 of gene gene14924\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69438218-69438741 of gene gene14925\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69463288-69463725 of gene gene14926\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69500254-69500758 of gene gene14927\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69544368-69544892 of gene gene14929\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69569737-69570109 of gene gene14930\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69575975-69576500 of gene gene14931\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69587295-69587810 of gene gene14933\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69599886-69600262 of gene gene14934\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69607275-69607801 of gene gene14935\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69617048-69617423 of gene gene14936\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69631648-69632178 of gene gene14937\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69648824-69649354 of gene gene14938\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69681406-69681939 of gene gene14939\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69700767-69701287 of gene gene14940\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69716498-69716956 of gene gene14941\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69726573-69727125 of gene gene14942\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69750854-69751310 of gene gene14943\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69764523-69764992 of gene gene14944\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69775750-69776306 of gene gene14945\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69814631-69815100 of gene gene14946\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69823355-69823911 of gene gene14947\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69834792-69835245 of gene gene14948\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69858420-69858884 of gene gene14949\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69868595-69869193 of gene gene14950\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69879350-69879819 of gene gene14951\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69900424-69900977 of gene gene14952\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69943186-69943648 of gene gene14953\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69963312-69963861 of gene gene14954\n",
      "Warning: could not find the corresponding gene segment for CDS chain 69992465-69992977 of gene gene14955\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70010949-70011673 of gene gene14956\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70044112-70044678 of gene gene14957\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70058632-70059199 of gene gene14958\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70074024-70074584 of gene gene14959\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70117061-70117617 of gene gene14961\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70138462-70139082 of gene gene14963\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70171809-70172270 of gene gene14965\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70193228-70193797 of gene gene14966\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70215433-70215957 of gene gene14968\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70216858-70217421 of gene gene14969\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70260409-70260933 of gene gene14971\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70297554-70298119 of gene gene14973\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70314895-70315452 of gene gene14974\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70335841-70336479 of gene gene14976\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70340876-70341448 of gene gene14977\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70355877-70356445 of gene gene14978\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70371469-70371993 of gene gene14980\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70386672-70387238 of gene gene14981\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70406469-70406992 of gene gene14982\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70434952-70435476 of gene gene14985\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70457503-70458036 of gene gene14987\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70518250-70518849 of gene gene14990\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70553274-70554440 of gene gene14991\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70572633-70573230 of gene gene14992\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70588189-70588777 of gene gene14993\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70588871-70589452 of gene gene14994\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70607437-70608036 of gene gene14995\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70642405-70643565 of gene gene14996\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70663298-70663895 of gene gene14997\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70671789-70672377 of gene gene14998\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70686946-70687534 of gene gene14999\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70698468-70699067 of gene gene15000\n",
      "Warning: could not find the corresponding gene segment for CDS chain 70703578-70704177 of gene gene15001\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113256204-113260236 of gene gene31382\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113269263-113273248 of gene gene31384\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113287389-113288932 of gene gene31385\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113304314-113307933 of gene gene31386\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113326544-113330523 of gene gene31387\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113357442-113361232 of gene gene31388\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113407535-113416324 of gene gene31392\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113418826-113422730 of gene gene31393\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113578504-113578990 of gene gene31422\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113588267-113588700 of gene gene31423\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113603460-113603928 of gene gene31426\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113612965-113613442 of gene gene31428\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113661771-113662228 of gene gene31434\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113807314-113807747 of gene gene31446\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113875772-113876239 of gene gene31451\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113884761-113885342 of gene gene31454\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113896408-113896946 of gene gene31455\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113912025-113912483 of gene gene31456\n",
      "Warning: could not find the corresponding gene segment for CDS chain 113994469-113994898 of gene gene31461\n",
      "Warning: could not find the corresponding gene segment for CDS chain 114048241-114048704 of gene gene31464\n",
      "Warning: could not find the corresponding gene segment for CDS chain 114059845-114060273 of gene gene31465\n",
      "Warning: could not find the corresponding gene segment for CDS chain 114109001-114109430 of gene gene31470\n",
      "Warning: could not find the corresponding gene segment for CDS chain 114127533-114127984 of gene gene31471\n",
      "Warning: could not find the corresponding gene segment for CDS chain 114140692-114141122 of gene gene31472\n",
      "Warning: could not find the corresponding gene segment for CDS chain 114153180-114153644 of gene gene31473\n",
      "Warning: could not find the corresponding gene segment for CDS chain 114176438-114176867 of gene gene31475\n",
      "Warning: could not find the corresponding gene segment for CDS chain 114196439-114196875 of gene gene31476\n",
      "Warning: could not find the corresponding gene segment for CDS chain 114222788-114223254 of gene gene31477\n",
      "Warning: could not find the corresponding gene segment for CDS chain 114479007-114479451 of gene gene31500\n",
      "Warning: could not find the corresponding gene segment for CDS chain 114523441-114523887 of gene gene31506\n",
      "Warning: could not find the corresponding gene segment for CDS chain 114843693-114844083 of gene gene31541\n",
      "Warning: could not find the corresponding gene segment for CDS chain 114937154-114937544 of gene gene31550\n",
      "Warning: could not find the corresponding gene segment for CDS chain 115158445-115158835 of gene gene31564\n",
      "Warning: could not find the corresponding gene segment for CDS chain 115208137-115208527 of gene gene31568\n",
      "Warning: could not find the corresponding gene segment for CDS chain 115312166-115312597 of gene gene31576\n",
      "Warning: could not find the corresponding gene segment for CDS chain 115460999-115461431 of gene gene31584\n",
      "Warning: could not find the corresponding gene segment for CDS chain 115495625-115496058 of gene gene31587\n",
      "Warning: could not find the corresponding gene segment for CDS chain 115507587-115507977 of gene gene31588\n",
      "Warning: could not find the corresponding gene segment for CDS chain 115623203-115623595 of gene gene31600\n",
      "Warning: could not find the corresponding gene segment for CDS chain 115802690-115803080 of gene gene31611\n",
      "Warning: could not find the corresponding gene segment for CDS chain 115980702-115981134 of gene gene31629\n",
      "Warning: could not find the corresponding gene segment for CDS chain 19178042-19178492 of gene gene31859\n",
      "Warning: could not find the corresponding gene segment for CDS chain 19184981-19185506 of gene gene31860\n",
      "Warning: could not find the corresponding gene segment for CDS chain 19190368-19190929 of gene gene31861\n",
      "Warning: could not find the corresponding gene segment for CDS chain 19192278-19192725 of gene gene31862\n",
      "Warning: could not find the corresponding gene segment for CDS chain 19214103-19216618 of gene gene31864\n",
      "Warning: could not find the corresponding gene segment for CDS chain 19242845-19243301 of gene gene31865\n",
      "Warning: could not find the corresponding gene segment for CDS chain 19260885-19263397 of gene gene31867\n",
      "Warning: could not find the corresponding gene segment for CDS chain 19305060-19307551 of gene gene31868\n",
      "Warning: could not find the corresponding gene segment for CDS chain 19340016-19340454 of gene gene31871\n",
      "Warning: could not find the corresponding gene segment for CDS chain 19344558-19352343 of gene gene31873\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52427967-52428876 of gene gene34790\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52567293-52568053 of gene gene34798\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52580762-52581213 of gene gene34799\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52607307-52607804 of gene gene34800\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52622566-52623082 of gene gene34801\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52638515-52638962 of gene gene34802\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52654817-52655329 of gene gene34803\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52667411-52667864 of gene gene34804\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52683964-52684472 of gene gene34805\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52710917-52711402 of gene gene34806\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52725308-52726919 of gene gene34807\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52744314-52744838 of gene gene34808\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52753398-52753846 of gene gene34809\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52769863-52770392 of gene gene34810\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52778510-52778995 of gene gene34812\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52792316-52792776 of gene gene34813\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52795165-52795633 of gene gene34814\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52810997-52811496 of gene gene34815\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52821922-52822359 of gene gene34816\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52824357-52824913 of gene gene34817\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52835861-52836387 of gene gene34818\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52844560-52845041 of gene gene34819\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52851355-52851866 of gene gene34820\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52855665-52856166 of gene gene34821\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52863365-52863933 of gene gene34823\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52887587-52888080 of gene gene34824\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52892186-52892618 of gene gene34825\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52899972-52900470 of gene gene34826\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52915915-52916426 of gene gene34829\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52920678-52921110 of gene gene34830\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52929789-52930282 of gene gene34832\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52942764-52943289 of gene gene34833\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52948668-52949169 of gene gene34834\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52956828-52957400 of gene gene34835\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52972532-52973038 of gene gene34837\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52983456-52983770 of gene gene34838\n",
      "Warning: could not find the corresponding gene segment for CDS chain 52990889-52991389 of gene gene34839\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53001713-53002231 of gene gene34840\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53007632-53008098 of gene gene34841\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53020459-53020991 of gene gene34842\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53032902-53033419 of gene gene34844\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53042428-53042891 of gene gene34845\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53047287-53047821 of gene gene34846\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53072763-53073275 of gene gene34847\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53078547-53079045 of gene gene34848\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53082063-53082522 of gene gene34849\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53091346-53091875 of gene gene34850\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53101801-53102265 of gene gene34851\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53104871-53105344 of gene gene34852\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53122114-53122611 of gene gene34853\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53132700-53133137 of gene gene34854\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53135122-53135678 of gene gene34855\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53146922-53147448 of gene gene34856\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53155607-53156088 of gene gene34857\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53162399-53162910 of gene gene34858\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53166677-53167178 of gene gene34859\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53174325-53174896 of gene gene34861\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53198525-53199018 of gene gene34862\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53203122-53203544 of gene gene34863\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53210906-53211407 of gene gene34864\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53226884-53227400 of gene gene34867\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53231668-53232102 of gene gene34868\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53240751-53241244 of gene gene34870\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53253792-53254317 of gene gene34871\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53259678-53260179 of gene gene34872\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53267963-53268535 of gene gene34873\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53283669-53284175 of gene gene34875\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53294596-53295030 of gene gene34876\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53302059-53302559 of gene gene34877\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53312904-53313422 of gene gene34878\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53318818-53319284 of gene gene34879\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53324676-53325208 of gene gene34880\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53337122-53337639 of gene gene34882\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53345961-53346424 of gene gene34883\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53351090-53351621 of gene gene34884\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53362368-53364106 of gene gene34885\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53370077-53370575 of gene gene34886\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53374821-53375278 of gene gene34887\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53390633-53391136 of gene gene34888\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53418388-53418873 of gene gene34889\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53428761-53430377 of gene gene34890\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53443314-53443839 of gene gene34891\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53454327-53454784 of gene gene34892\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53461209-53461738 of gene gene34893\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53469756-53470231 of gene gene34895\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53488106-53488567 of gene gene34896\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53491152-53491622 of gene gene34897\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53505790-53506286 of gene gene34898\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53516929-53517366 of gene gene34899\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53519303-53519859 of gene gene34900\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53530786-53531313 of gene gene34901\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53538266-53538738 of gene gene34902\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53545014-53545525 of gene gene34903\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53554057-53554558 of gene gene34904\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53559676-53560247 of gene gene34905\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53586471-53586964 of gene gene34906\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53591080-53591504 of gene gene34907\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53598911-53599410 of gene gene34908\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53616397-53616914 of gene gene34910\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53621752-53622245 of gene gene34911\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53634888-53635399 of gene gene34912\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53649425-53649994 of gene gene34914\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53666005-53666506 of gene gene34916\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53676196-53676628 of gene gene34917\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53683677-53684177 of gene gene34918\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53704007-53704514 of gene gene34919\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53709943-53710419 of gene gene34920\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53716760-53717291 of gene gene34921\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53729558-53730074 of gene gene34922\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53738375-53738838 of gene gene34923\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53743174-53743705 of gene gene34924\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53757410-53757921 of gene gene34925\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53763196-53763694 of gene gene34926\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53766735-53767179 of gene gene34927\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53777071-53777576 of gene gene34928\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53795455-53795963 of gene gene34930\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53806639-53807115 of gene gene34931\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53831105-53831827 of gene gene34932\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53845322-53845827 of gene gene34933\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53863110-53863661 of gene gene34934\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53876016-53876752 of gene gene34935\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53881612-53882217 of gene gene34936\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53926978-53927595 of gene gene34938\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53946073-53946660 of gene gene34940\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53960993-53961602 of gene gene34942\n",
      "Warning: could not find the corresponding gene segment for CDS chain 53977153-53977678 of gene gene34943\n",
      "Warning: could not find the corresponding gene segment for CDS chain 54075004-54075515 of gene gene34946\n",
      "Warning: could not find the corresponding gene segment for CDS chain 54142851-54146108 of gene gene34951\n",
      "Warning: could not find the corresponding gene segment for CDS chain 54148662-54149201 of gene gene34952\n",
      "Warning: could not find the corresponding gene segment for CDS chain 54220521-54224198 of gene gene35013\n",
      "Warning: could not find the corresponding gene segment for CDS chain 19085017-19085460 of gene gene37959\n",
      "Warning: could not find the corresponding gene segment for CDS chain 19241208-19241679 of gene gene37967\n",
      "Warning: could not find the corresponding gene segment for CDS chain 19260403-19260844 of gene gene37968\n",
      "Warning: could not find the corresponding gene segment for CDS chain 61232397-61232935 of gene gene43615\n",
      "Warning: could not find the corresponding gene segment for CDS chain 138385-142276 of gene gene46566\n",
      "Warning: could not find the corresponding gene segment for CDS chain 154428-1266230 of gene gene46567\n",
      "Warning: could not find the corresponding gene segment for CDS chain 170882-1058542 of gene gene46568\n",
      "Warning: could not find the corresponding gene segment for CDS chain 191755-791762 of gene gene46569\n",
      "Warning: could not find the corresponding gene segment for CDS chain 228395-231303 of gene gene46570\n",
      "Warning: could not find the corresponding gene segment for CDS chain 278805-287519 of gene gene46572\n",
      "Warning: could not find the corresponding gene segment for CDS chain 290081-1174475 of gene gene46573\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1307139-1307593 of gene gene46576\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1345173-1345626 of gene gene46577\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1369629-1370082 of gene gene46578\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1397346-1399594 of gene gene46579\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1441056-1441509 of gene gene46580\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1453087-1479273 of gene gene46581\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1500168-1502939 of gene gene46582\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1518594-1519146 of gene gene46583\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1527699-1528169 of gene gene46584\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1532723-1533194 of gene gene46586\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1549249-1549767 of gene gene46587\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1580632-1583295 of gene gene46589\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1612110-1614773 of gene gene46591\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1652261-1653173 of gene gene46592\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1668521-1669012 of gene gene46593\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1695479-1695959 of gene gene46594\n",
      "Warning: could not find the corresponding gene segment for CDS chain 72669-73602 of gene gene46607\n",
      "Warning: could not find the corresponding gene segment for CDS chain 234790-235550 of gene gene46614\n",
      "Warning: could not find the corresponding gene segment for CDS chain 246939-247390 of gene gene46615\n",
      "Warning: could not find the corresponding gene segment for CDS chain 267059-267557 of gene gene46616\n",
      "Warning: could not find the corresponding gene segment for CDS chain 282342-282858 of gene gene46617\n",
      "Warning: could not find the corresponding gene segment for CDS chain 298270-298714 of gene gene46618\n",
      "Warning: could not find the corresponding gene segment for CDS chain 313424-313941 of gene gene46619\n",
      "Warning: could not find the corresponding gene segment for CDS chain 321439-321892 of gene gene46620\n",
      "Warning: could not find the corresponding gene segment for CDS chain 337968-338476 of gene gene46621\n",
      "Warning: could not find the corresponding gene segment for CDS chain 365129-365614 of gene gene46622\n",
      "Warning: could not find the corresponding gene segment for CDS chain 380537-382152 of gene gene46623\n",
      "Warning: could not find the corresponding gene segment for CDS chain 399727-400251 of gene gene46624\n",
      "Warning: could not find the corresponding gene segment for CDS chain 409424-409872 of gene gene46625\n",
      "Warning: could not find the corresponding gene segment for CDS chain 418813-419342 of gene gene46626\n",
      "Warning: could not find the corresponding gene segment for CDS chain 427451-427936 of gene gene46628\n",
      "Warning: could not find the corresponding gene segment for CDS chain 441206-441666 of gene gene46629\n",
      "Warning: could not find the corresponding gene segment for CDS chain 444246-444716 of gene gene46630\n",
      "Warning: could not find the corresponding gene segment for CDS chain 458650-459146 of gene gene46631\n",
      "Warning: could not find the corresponding gene segment for CDS chain 469073-469510 of gene gene46632\n",
      "Warning: could not find the corresponding gene segment for CDS chain 471472-472024 of gene gene46633\n",
      "Warning: could not find the corresponding gene segment for CDS chain 746382-746890 of gene gene46634\n",
      "Warning: could not find the corresponding gene segment for CDS chain 912496-913023 of gene gene46635\n",
      "Warning: could not find the corresponding gene segment for CDS chain 929766-930238 of gene gene46636\n",
      "Warning: could not find the corresponding gene segment for CDS chain 936548-937059 of gene gene46637\n",
      "Warning: could not find the corresponding gene segment for CDS chain 941045-941546 of gene gene46638\n",
      "Warning: could not find the corresponding gene segment for CDS chain 946658-947226 of gene gene46639\n",
      "Warning: could not find the corresponding gene segment for CDS chain 968042-968538 of gene gene46640\n",
      "Warning: could not find the corresponding gene segment for CDS chain 972643-973077 of gene gene46641\n",
      "Warning: could not find the corresponding gene segment for CDS chain 980450-980949 of gene gene46642\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1009451-1009978 of gene gene46644\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1014262-1014694 of gene gene46645\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1023341-1023834 of gene gene46647\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1036320-1036831 of gene gene46648\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1042291-1042792 of gene gene46649\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1050316-1050885 of gene gene46650\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1066061-1066562 of gene gene46652\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1076069-1076383 of gene gene46653\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1083543-1084046 of gene gene46654\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1094682-1095189 of gene gene46655\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1100574-1101040 of gene gene46656\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1106877-1107409 of gene gene46657\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1119275-1119789 of gene gene46659\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1128287-1128750 of gene gene46660\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1133144-1133675 of gene gene46661\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1158653-1159165 of gene gene46662\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1164437-1164935 of gene gene46663\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1167954-1168413 of gene gene46664\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1185353-1185854 of gene gene46665\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1207164-1207672 of gene gene46667\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1222997-1223473 of gene gene46668\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1256240-1256965 of gene gene46669\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1278997-1279502 of gene gene46670\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1297109-1297660 of gene gene46671\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1311342-1312080 of gene gene46672\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1316937-1317548 of gene gene46673\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1369211-1369829 of gene gene46674\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1373771-1374358 of gene gene46676\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1384602-1385202 of gene gene46678\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1400172-1400696 of gene gene46679\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1519291-1519802 of gene gene46682\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1580287-1583552 of gene gene46687\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1586041-1586580 of gene gene46688\n",
      "Warning: could not find the corresponding gene segment for CDS chain 1657818-1661492 of gene gene46749\n",
      "Warning: could not find the corresponding gene segment for CDS chain 12209-16119 of gene gene47297\n",
      "Warning: could not find the corresponding gene segment for CDS chain 33640-37532 of gene gene47298\n",
      "Warning: could not find the corresponding gene segment for CDS chain 44362-53897 of gene gene47299\n"
     ]
    }
   ],
   "source": [
    "# convert the GFF file to GTF using gffread\n",
    "models_asgtf = gffread_gff_to_gtf(sample_docket.annot, species_sample_dict, gffread_loc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3199731</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3199731</td>\n",
       "      <td>3207317</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3213439</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224794</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>14145</td>\n",
       "      <td>15288</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"gene48832\"; gene_name \"CYTB\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224795</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111579\"; gene_id \"gene48833\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224796</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111579\"; gene_id \"gene48833\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224797</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111580\"; gene_id \"gene48834\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224798</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111580\"; gene_id \"gene48834\"...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2224799 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0       1           2        3        4  5  6  7  \\\n",
       "0        NC_000067.6  Gnomon  transcript  3199731  3671742  .  -  .   \n",
       "1        NC_000067.6  Gnomon        exon  3199731  3207317  .  -  .   \n",
       "2        NC_000067.6  Gnomon        exon  3213439  3216968  .  -  .   \n",
       "3        NC_000067.6  Gnomon        exon  3421702  3421901  .  -  .   \n",
       "4        NC_000067.6  Gnomon        exon  3670552  3671742  .  -  .   \n",
       "...              ...     ...         ...      ...      ... .. .. ..   \n",
       "2224794  NC_005089.1  RefSeq         CDS    14145    15288  .  +  0   \n",
       "2224795  NC_005089.1  RefSeq  transcript    15289    15355  .  +  .   \n",
       "2224796  NC_005089.1  RefSeq        exon    15289    15355  .  +  .   \n",
       "2224797  NC_005089.1  RefSeq  transcript    15356    15422  .  -  .   \n",
       "2224798  NC_005089.1  RefSeq        exon    15356    15422  .  -  .   \n",
       "\n",
       "                                                         8  \n",
       "0        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "1        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "2        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "3        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "4        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "...                                                    ...  \n",
       "2224794       transcript_id \"gene48832\"; gene_name \"CYTB\";  \n",
       "2224795  transcript_id \"rna111579\"; gene_id \"gene48833\"...  \n",
       "2224796  transcript_id \"rna111579\"; gene_id \"gene48833\"...  \n",
       "2224797  transcript_id \"rna111580\"; gene_id \"gene48834\"...  \n",
       "2224798  transcript_id \"rna111580\"; gene_id \"gene48834\"...  \n",
       "\n",
       "[2224799 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'transcript_id \"rna0\"; gene_id \"gene0\"; gene_name \"Xkr4\";'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load the newly-generated GTF file as a dataframe\n",
    "models_asgtf_df = pd.read_csv(models_asgtf.path, skiprows = 0, header = None, sep = '\\t')\n",
    "\n",
    "display(models_asgtf_df)\n",
    "display(models_asgtf_df[attributes_column][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3199731</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3199731</td>\n",
       "      <td>3207317</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3213439</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224793</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14145</td>\n",
       "      <td>15288</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"gene48832\"; gene_id \"gene48832\"...</td>\n",
       "      <td>{'transcript_id': 'gene48832', 'gene_id': 'gen...</td>\n",
       "      <td>CYTB</td>\n",
       "      <td>gene48832</td>\n",
       "      <td>gene48832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224795</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111579\"; gene_id \"gene48833\"...</td>\n",
       "      <td>{'transcript_id': 'rna111579', 'gene_id': 'gen...</td>\n",
       "      <td>TrnT</td>\n",
       "      <td>gene48833</td>\n",
       "      <td>rna111579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224796</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111579\"; gene_id \"gene48833\"...</td>\n",
       "      <td>{'transcript_id': 'rna111579', 'gene_id': 'gen...</td>\n",
       "      <td>TrnT</td>\n",
       "      <td>gene48833</td>\n",
       "      <td>rna111579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224797</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111580\"; gene_id \"gene48834\"...</td>\n",
       "      <td>{'transcript_id': 'rna111580', 'gene_id': 'gen...</td>\n",
       "      <td>TrnP</td>\n",
       "      <td>gene48834</td>\n",
       "      <td>rna111580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224798</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111580\"; gene_id \"gene48834\"...</td>\n",
       "      <td>{'transcript_id': 'rna111580', 'gene_id': 'gen...</td>\n",
       "      <td>TrnP</td>\n",
       "      <td>gene48834</td>\n",
       "      <td>rna111580</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1287511 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0       1           2        3        4  5  6  7  \\\n",
       "0        NC_000067.6  Gnomon  transcript  3199731  3671742  .  -  .   \n",
       "1        NC_000067.6  Gnomon        exon  3199731  3207317  .  -  .   \n",
       "2        NC_000067.6  Gnomon        exon  3213439  3216968  .  -  .   \n",
       "3        NC_000067.6  Gnomon        exon  3421702  3421901  .  -  .   \n",
       "4        NC_000067.6  Gnomon        exon  3670552  3671742  .  -  .   \n",
       "...              ...     ...         ...      ...      ... .. .. ..   \n",
       "2224793  NC_005089.1  RefSeq  transcript    14145    15288  .  +  .   \n",
       "2224795  NC_005089.1  RefSeq  transcript    15289    15355  .  +  .   \n",
       "2224796  NC_005089.1  RefSeq        exon    15289    15355  .  +  .   \n",
       "2224797  NC_005089.1  RefSeq  transcript    15356    15422  .  -  .   \n",
       "2224798  NC_005089.1  RefSeq        exon    15356    15422  .  -  .   \n",
       "\n",
       "                                                         8  \\\n",
       "0        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "1        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "2        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "3        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "4        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "...                                                    ...   \n",
       "2224793  transcript_id \"gene48832\"; gene_id \"gene48832\"...   \n",
       "2224795  transcript_id \"rna111579\"; gene_id \"gene48833\"...   \n",
       "2224796  transcript_id \"rna111579\"; gene_id \"gene48833\"...   \n",
       "2224797  transcript_id \"rna111580\"; gene_id \"gene48834\"...   \n",
       "2224798  transcript_id \"rna111580\"; gene_id \"gene48834\"...   \n",
       "\n",
       "                                          field_dictionary gene_name  \\\n",
       "0        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...      Xkr4   \n",
       "1        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...      Xkr4   \n",
       "2        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...      Xkr4   \n",
       "3        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...      Xkr4   \n",
       "4        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...      Xkr4   \n",
       "...                                                    ...       ...   \n",
       "2224793  {'transcript_id': 'gene48832', 'gene_id': 'gen...      CYTB   \n",
       "2224795  {'transcript_id': 'rna111579', 'gene_id': 'gen...      TrnT   \n",
       "2224796  {'transcript_id': 'rna111579', 'gene_id': 'gen...      TrnT   \n",
       "2224797  {'transcript_id': 'rna111580', 'gene_id': 'gen...      TrnP   \n",
       "2224798  {'transcript_id': 'rna111580', 'gene_id': 'gen...      TrnP   \n",
       "\n",
       "           gene_id transcript_id  \n",
       "0            gene0          rna0  \n",
       "1            gene0          rna0  \n",
       "2            gene0          rna0  \n",
       "3            gene0          rna0  \n",
       "4            gene0          rna0  \n",
       "...            ...           ...  \n",
       "2224793  gene48832     gene48832  \n",
       "2224795  gene48833     rna111579  \n",
       "2224796  gene48833     rna111579  \n",
       "2224797  gene48834     rna111580  \n",
       "2224798  gene48834     rna111580  \n",
       "\n",
       "[1287511 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Use a custom function to extract useful fields from the additional fields section (column 8)\n",
    "# Pull from that dict to fill in additional useful columns\n",
    "models_asgtf_df['field_dictionary'] = models_asgtf_df[8].apply(convert_fields_to_dict_gtf)\n",
    "models_asgtf_df['gene_name'] = [d.get('gene_name') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['gene_id'] = [d.get('gene_id') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['transcript_id'] = [d.get('transcript_id') for d in models_asgtf_df['field_dictionary']]\n",
    "\n",
    "# Remove CDS annotations because they interfere with TransDecoder cDNA generation\n",
    "models_asgtf_df = models_asgtf_df[models_asgtf_df[2] != 'CDS']\n",
    "display(models_asgtf_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Generate gtf-idmm\n",
    "This file maps the gene_name to `gene_id` and `transcript_id` fields generated by the conversion from GFF to GTF, which will be needed for downstream processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>LOC105243853</td>\n",
       "      <td>gene2</td>\n",
       "      <td>rna4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224789</th>\n",
       "      <td>ND6</td>\n",
       "      <td>gene48830</td>\n",
       "      <td>gene48830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224791</th>\n",
       "      <td>TrnE</td>\n",
       "      <td>gene48831</td>\n",
       "      <td>rna111578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224793</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>gene48832</td>\n",
       "      <td>gene48832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224795</th>\n",
       "      <td>TrnT</td>\n",
       "      <td>gene48833</td>\n",
       "      <td>rna111579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224797</th>\n",
       "      <td>TrnP</td>\n",
       "      <td>gene48834</td>\n",
       "      <td>rna111580</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>116982 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_name    gene_id transcript_id\n",
       "0                Xkr4      gene0          rna0\n",
       "8                Xkr4      gene0          rna1\n",
       "15               Xkr4      gene0          rna2\n",
       "22               Xkr4      gene0          rna3\n",
       "29       LOC105243853      gene2          rna4\n",
       "...               ...        ...           ...\n",
       "2224789           ND6  gene48830     gene48830\n",
       "2224791          TrnE  gene48831     rna111578\n",
       "2224793          CYTB  gene48832     gene48832\n",
       "2224795          TrnT  gene48833     rna111579\n",
       "2224797          TrnP  gene48834     rna111580\n",
       "\n",
       "[116982 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract gene_name, gene_id, and transcript_id fields to generate an ID mapping matrix (idmm)\n",
    "idmm_df = models_asgtf_df[['gene_name', 'gene_id', 'transcript_id']].drop_duplicates()\n",
    "idmm_df.dropna(inplace = True)\n",
    "display(idmm_df)\n",
    "\n",
    "# generate a filename and file for the idmm\n",
    "idmm_filename = '_'.join([species_prefix, conditions, 'gtf-idmm.tsv'])\n",
    "idmm = idmm_file(idmm_filename, species_sample_dict, idmm_df.columns)\n",
    "\n",
    "# save to file and add to the docket\n",
    "idmm_df.to_csv(idmm.path, sep = '\\t')\n",
    "sample_docket.add_keyfile(idmm, 'gtf-idmm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Generate updated gtf\n",
    "Generated an updated GTF file using transcript_id as the key. For some datasets, transcripts do not consistently get gene names and gene IDs added, which causes Transdecoder to throw errors. This resolves that problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "models_asgtf_updated_df = models_asgtf_df.merge(idmm_df, on = 'transcript_id')\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_name': x['gene_name_y']}), axis = 1)\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_id': x['gene_id_y']}), axis = 1)\n",
    "models_asgtf_updated_df['new_fields'] = models_asgtf_updated_df['field_dictionary'].apply(convert_dict_to_fields_gtf)\n",
    "models_asgtf_updated_df = models_asgtf_updated_df[[0, 1, 2, 3, 4, 5, 6, 7, 'new_fields']]\n",
    "\n",
    "models_asgtf_updated_filename = models_asgtf.filename.replace('.gtf', '_updated.gtf')\n",
    "models_asgtf_updated = genome_gtf_file(models_asgtf_updated_filename, species_sample_dict)\n",
    "\n",
    "models_asgtf_updated_df.to_csv(models_asgtf_updated.path, header = None, index = None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-parsing cufflinks output: ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_transcripts.gtf\n",
      "-parsing genome fasta: ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic.fna\n",
      "^C\n"
     ]
    }
   ],
   "source": [
    "version = 'GRCm38p3'\n",
    "format_details = 'transdecoder_cDNA'\n",
    "txome_filename = '_'.join([species_prefix, conditions, version, format_details + '.fa'] )\n",
    "txome_filepath = downloads_folder + txome_filename\n",
    "\n",
    "!{transdecoder_loc + 'util/gtf_genome_to_cdna_fasta.pl'} {gtftrans_filepath} {genome_unzip_filepath} > {txome_filepath}\n",
    "\n",
    "file_set.add(txome_filepath)\n",
    "file_dict[txome_filepath] = 'organisms/' + species_folder + '/genomics_reference/transcriptome/' + txome_filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-- Skipping CMD: /home/ec2-user/TransDecoder-TransDecoder-v5.5.0/util/compute_base_probs.pl ../../output/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa 0 > /home/ec2-user/2022_arcadia_glialorigins_working/notebooks/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/base_freqs.dat, checkpoint [/home/ec2-user/2022_arcadia_glialorigins_working/notebooks/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir.__checkpoints_longorfs/base_freqs_file.ok] exists.\n",
      "-skipping long orf extraction, already completed earlier as per checkpoint: /home/ec2-user/2022_arcadia_glialorigins_working/notebooks/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir.__checkpoints_longorfs/TD.longorfs.ok\n",
      "-- Skipping CMD: /home/ec2-user/TransDecoder-TransDecoder-v5.5.0/util/get_top_longest_fasta_entries.pl Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/longest_orfs.cds 5000 5000 > Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/longest_orfs.cds.top_longest_5000, checkpoint [/home/ec2-user/2022_arcadia_glialorigins_working/notebooks/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir.__checkpoints/get_longest_orfs.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/TransDecoder-TransDecoder-v5.5.0/util/exclude_similar_proteins.pl Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/longest_orfs.cds.top_longest_5000 > Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/longest_orfs.cds.top_longest_5000.nr, checkpoint [/home/ec2-user/2022_arcadia_glialorigins_working/notebooks/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir.__checkpoints/nr.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/TransDecoder-TransDecoder-v5.5.0/util/get_top_longest_fasta_entries.pl Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/longest_orfs.cds.top_longest_5000.nr 500 > Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/longest_orfs.cds.top_500_longest, checkpoint [/home/ec2-user/2022_arcadia_glialorigins_working/notebooks/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir.__checkpoints/top_train_select.ok] exists.\n",
      "PCT_GC: 46.2\n",
      "-- Skipping CMD: /home/ec2-user/TransDecoder-TransDecoder-v5.5.0/util/seq_n_baseprobs_to_loglikelihood_vals.pl Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/longest_orfs.cds.top_500_longest Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/base_freqs.dat > Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/hexamer.scores, checkpoint [/home/ec2-user/2022_arcadia_glialorigins_working/notebooks/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir.__checkpoints/hexamer_scores.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/TransDecoder-TransDecoder-v5.5.0/util/score_CDS_likelihood_all_6_frames.pl Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/longest_orfs.cds Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/hexamer.scores > Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/longest_orfs.cds.scores, checkpoint [/home/ec2-user/2022_arcadia_glialorigins_working/notebooks/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir.__checkpoints/score_cdss.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/TransDecoder-TransDecoder-v5.5.0/util/select_best_ORFs_per_transcript.pl --gff3_file Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/longest_orfs.gff3 --cds_scores Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/longest_orfs.cds.scores  --min_length_auto_accept 749  > Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/longest_orfs.cds.best_candidates.gff3 , checkpoint [/home/ec2-user/2022_arcadia_glialorigins_working/notebooks/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir.__checkpoints/select_best_orfs.5ff0b26f8fe5cefb2210bfd598afe17b.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/TransDecoder-TransDecoder-v5.5.0/util/train_start_PWM.pl --transcripts ../../output/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa --selected_orfs Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/longest_orfs.cds.top_500_longest --out_prefix Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/start_refinement, checkpoint [/home/ec2-user/2022_arcadia_glialorigins_working/notebooks/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir.__checkpoints/train_start_PWM.5ff0b26f8fe5cefb2210bfd598afe17b.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/TransDecoder-TransDecoder-v5.5.0/util/start_codon_refinement.pl --transcripts ../../output/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa --gff3_file Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/longest_orfs.cds.best_candidates.gff3 --workdir Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir > Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/longest_orfs.cds.best_candidates.gff3.revised_starts.gff3, checkpoint [/home/ec2-user/2022_arcadia_glialorigins_working/notebooks/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir.__checkpoints/revise_starts.5ff0b26f8fe5cefb2210bfd598afe17b.ok] exists.\n",
      "-- Skipping CMD: cp Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir/longest_orfs.cds.best_candidates.gff3.revised_starts.gff3 Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.gff3, checkpoint [/home/ec2-user/2022_arcadia_glialorigins_working/notebooks/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir.__checkpoints/copy_to_final_outputfile.5ff0b26f8fe5cefb2210bfd598afe17b.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/TransDecoder-TransDecoder-v5.5.0/util/gff3_file_to_bed.pl Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.gff3 > Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.bed, checkpoint [/home/ec2-user/2022_arcadia_glialorigins_working/notebooks/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir.__checkpoints/make_final_bed.5ff0b26f8fe5cefb2210bfd598afe17b.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/TransDecoder-TransDecoder-v5.5.0/util/gff3_file_to_proteins.pl --gff3 Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.gff3 --fasta ../../output/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa  --genetic_code Universal > Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.pep, checkpoint [/home/ec2-user/2022_arcadia_glialorigins_working/notebooks/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir.__checkpoints/make_final_pep.5ff0b26f8fe5cefb2210bfd598afe17b.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/TransDecoder-TransDecoder-v5.5.0/util/gff3_file_to_proteins.pl --gff3 Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.gff3 --fasta ../../output/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa --seqType CDS  --genetic_code Universal > Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.cds, checkpoint [/home/ec2-user/2022_arcadia_glialorigins_working/notebooks/Drer_adultbrain/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder_dir.__checkpoints/make_final_cds.5ff0b26f8fe5cefb2210bfd598afe17b.ok] exists.\n",
      "transdecoder is finished.  See output files Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.*\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!{tdlongorf_loc} -t {txome_filepath}\n",
    "!{tdpredict_loc} -t {txome_filepath}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "iEug6TQ6Cc9D"
   },
   "source": [
    "# 13. Push files to AWS S3\n",
    "\n",
    "Iteratively moves through the file_set and file_dict variables and populates files into the right place in AWS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "sVRRQITMeC2f",
    "outputId": "8aa0c336-100e-43a2-a5b7-f4dc1b20f14a"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upload: ../../output/Mmus_adultbrain/GSM2906406_Brain2_dge.txt to s3://arcadia-reference-datasets/organisms/Mus_musculus/functional_sequencing/scRNA-Seq/GSM2906406_Brain2_dge.txt\n",
      "upload: ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic.gff to s3://arcadia-reference-datasets/organisms/Mus_musculus/genomics_reference/annotation/GCF_000001635.23_GRCm38.p3_genomic.gff\n",
      "upload: ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic.fna to s3://arcadia-reference-datasets/organisms/Mus_musculus/genomics_reference/genome/GCF_000001635.23_GRCm38.p3_genomic.fna\n"
     ]
    }
   ],
   "source": [
    "for filename in file_set:\n",
    "    input = filename\n",
    "    path = 's3://arcadia-reference-datasets/' + file_dict[filename]\n",
    "    !aws s3 cp $input $path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "id": "U92TyitFeDh7"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.bed',\n",
       " 'Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.cds',\n",
       " 'Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.gff3',\n",
       " 'Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.pep']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upload: ./Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.bed to s3://arcadia-reference-datasets/organisms/Danio_rerio/genomics_reference/proteome/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.bed\n",
      "upload: ./Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.cds to s3://arcadia-reference-datasets/organisms/Danio_rerio/genomics_reference/proteome/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.cds\n",
      "upload: ./Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.gff3 to s3://arcadia-reference-datasets/organisms/Danio_rerio/genomics_reference/proteome/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.gff3\n",
      "upload: ./Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.pep to s3://arcadia-reference-datasets/organisms/Danio_rerio/genomics_reference/proteome/Drer_adultbrain_GRCz10_transdecoder_cDNA.fa.transdecoder.pep\n"
     ]
    }
   ],
   "source": [
    "suffixes = ['.bed', '.cds', '.gff3', '.pep']\n",
    "filenames_forexport = [txome_filename + '.transdecoder' + i for i in suffixes]\n",
    "display(filenames_forexport)\n",
    "\n",
    "for filepath in filenames_forexport:\n",
    "    s3address = 's3://arcadia-reference-datasets/organisms/' + species_folder + '/genomics_reference/proteome/' + filepath\n",
    "    !aws s3 cp {'./' + filepath} {s3address}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "Xtro_bro.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
