{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "L56fjA27DmhA"
   },
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "t0BTCsz7Y48y",
    "outputId": "f6510678-f977-4ec8-ce30-1f2acd4fe666"
   },
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import os\n",
    "import dill\n",
    "import copy\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "import sys\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Download and describe data\n",
    "\n",
    "This notebook collects data from the [Han et al. 2018](https://www.sciencedirect.com/science/article/pii/S0092867418301168#sec4) mouse cell atlas.\n",
    "\n",
    "## Dataset description\n",
    "\n",
    "- Data can be found at the [GEO Accession GSE108097](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE108097).\n",
    "- This notebook previously collects data from one sample, [\"Brain1\"](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM2906405), for the purposes of data analysis and exploration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'species': 'Mus_musculus',\n",
       " 'conditions': 'adultbrain',\n",
       " 'directory': '../../output/Mmus_adultbrain/',\n",
       " 'files': {},\n",
       " 'annot': <biofile_handling.GenomeGffFile at 0x7f8f926fd480>,\n",
       " 'genome_fasta': <biofile_handling.GenomeFastaFile at 0x7f8f926fd6c0>,\n",
       " 'gxc': <biofile_handling.GxcFile at 0x7f8f926fffa0>}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify the name of the species folder in Amazon S3\n",
    "species = 'Mus_musculus'\n",
    "\n",
    "# Specify any particular identifying conditions, eg tissue type:\n",
    "conditions = 'adultbrain'\n",
    "\n",
    "# Specify url and other variables\n",
    "genome_fasta_url = 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/635/GCF_000001635.23_GRCm38.p3/GCF_000001635.23_GRCm38.p3_genomic.fna.gz'\n",
    "genome_version = 'GRCm38.p3'\n",
    "\n",
    "annot_url = 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/635/GCF_000001635.23_GRCm38.p3/GCF_000001635.23_GRCm38.p3_genomic.gff.gz'\n",
    "gxc_url = 'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM2906nnn/GSM2906405/suppl/GSM2906405_Brain1_dge.txt.gz'\n",
    "\n",
    "###########\n",
    "# runtime #\n",
    "###########\n",
    "\n",
    "protocol = 'curl'\n",
    "\n",
    "species_prefix = prefixify(species)\n",
    "\n",
    "# Specify folder as destination for file downloads\n",
    "output_folder = '../../output/' + prefixify(species) + '_' + conditions + '/'\n",
    "\n",
    "if not os.path.exists(output_folder):\n",
    "    os.mkdir(output_folder)\n",
    "    \n",
    "species_SampleDict = SampleDict(species, conditions, output_folder)\n",
    "\n",
    "genome_fasta = GenomeFastaFile(\n",
    "    filename = '',\n",
    "    sampledict = species_SampleDict,\n",
    "    version = genome_version,\n",
    "    url = genome_fasta_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "annot = GenomeGffFile(\n",
    "    filename = '',\n",
    "    sampledict = species_SampleDict,\n",
    "    GenomeFastaFile = genome_fasta,\n",
    "    url = annot_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "gxc = GxcFile(\n",
    "    filename = '',\n",
    "    sampledict = species_SampleDict,\n",
    "    GenomeFastaFile = genome_fasta,\n",
    "    GenomeAnnotFile = annot,\n",
    "    url = gxc_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "\n",
    "sample_Docket = Docket(species_SampleDict)\n",
    "keyfiles = {\n",
    "    'annot': annot,\n",
    "    'genome_fasta': genome_fasta,\n",
    "    'gxc': gxc\n",
    "}\n",
    "sample_Docket.add_keyfiles(keyfiles)\n",
    "\n",
    "display(vars(sample_Docket))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Load in the gxc matrix and get gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Brain_1.CCGCTAAATAAATAAGGG</th>\n",
       "      <th>Brain_1.AACGCCGATCTTGCCCTC</th>\n",
       "      <th>Brain_1.ACCTGAAGTTTATCGTAA</th>\n",
       "      <th>Brain_1.CTCGCACTGAAACCGCTA</th>\n",
       "      <th>Brain_1.ATCAACATCTCTTCGGGT</th>\n",
       "      <th>Brain_1.GCGAATAGGGTCTATGTA</th>\n",
       "      <th>Brain_1.CGAGTAAGGGTCTAGTCG</th>\n",
       "      <th>Brain_1.ATCTCTTCGTAAGTTGCC</th>\n",
       "      <th>Brain_1.AACGCCTAAGGGCTCGCA</th>\n",
       "      <th>...</th>\n",
       "      <th>Brain_1.TTGGACGCCTAGGAGATC</th>\n",
       "      <th>Brain_1.TTAACTAAAGTTTATGTA</th>\n",
       "      <th>Brain_1.GTCCCGGGACATAGGACT</th>\n",
       "      <th>Brain_1.CCGCTAGGGTTTGCTCAA</th>\n",
       "      <th>Brain_1.TGATCAGCTGTGTCAAAG</th>\n",
       "      <th>Brain_1.TCACTTGAATTATGAAGC</th>\n",
       "      <th>Brain_1.AAGTACGCTGTGTATGTA</th>\n",
       "      <th>Brain_1.CCTAGATAGAGAATTTGC</th>\n",
       "      <th>Brain_1.CATCCCATTTGCGGCTGC</th>\n",
       "      <th>Brain_1.CAAAGTGGGTTTAGCGAG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0610005C13Rik</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0610007P14Rik</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0610009B22Rik</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0610009E02Rik</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0610009L18Rik</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20160</th>\n",
       "      <td>n-R5s28</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20161</th>\n",
       "      <td>n-R5s37</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20162</th>\n",
       "      <td>n-R5s52</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20163</th>\n",
       "      <td>n-R5s88</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20164</th>\n",
       "      <td>n-R5s98</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20165 rows × 3680 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_name  Brain_1.CCGCTAAATAAATAAGGG  Brain_1.AACGCCGATCTTGCCCTC  \\\n",
       "0      0610005C13Rik                           0                           0   \n",
       "1      0610007P14Rik                           0                           3   \n",
       "2      0610009B22Rik                           0                           3   \n",
       "3      0610009E02Rik                           0                           0   \n",
       "4      0610009L18Rik                           0                           0   \n",
       "...              ...                         ...                         ...   \n",
       "20160        n-R5s28                           0                           0   \n",
       "20161        n-R5s37                           0                           0   \n",
       "20162        n-R5s52                           0                           0   \n",
       "20163        n-R5s88                           0                           0   \n",
       "20164        n-R5s98                           0                           0   \n",
       "\n",
       "       Brain_1.ACCTGAAGTTTATCGTAA  Brain_1.CTCGCACTGAAACCGCTA  \\\n",
       "0                               0                           0   \n",
       "1                               1                           2   \n",
       "2                               0                           2   \n",
       "3                               0                           0   \n",
       "4                               0                           0   \n",
       "...                           ...                         ...   \n",
       "20160                           0                           0   \n",
       "20161                           0                           0   \n",
       "20162                           0                           0   \n",
       "20163                           0                           0   \n",
       "20164                           0                           0   \n",
       "\n",
       "       Brain_1.ATCAACATCTCTTCGGGT  Brain_1.GCGAATAGGGTCTATGTA  \\\n",
       "0                               0                           0   \n",
       "1                               0                           0   \n",
       "2                               1                           0   \n",
       "3                               0                           0   \n",
       "4                               2                           0   \n",
       "...                           ...                         ...   \n",
       "20160                           0                           0   \n",
       "20161                           0                           0   \n",
       "20162                           0                           0   \n",
       "20163                           0                           0   \n",
       "20164                           0                           0   \n",
       "\n",
       "       Brain_1.CGAGTAAGGGTCTAGTCG  Brain_1.ATCTCTTCGTAAGTTGCC  \\\n",
       "0                               0                           0   \n",
       "1                               1                           2   \n",
       "2                               0                           0   \n",
       "3                               0                           0   \n",
       "4                               0                           0   \n",
       "...                           ...                         ...   \n",
       "20160                           0                           0   \n",
       "20161                           0                           0   \n",
       "20162                           0                           0   \n",
       "20163                           0                           0   \n",
       "20164                           0                           0   \n",
       "\n",
       "       Brain_1.AACGCCTAAGGGCTCGCA  ...  Brain_1.TTGGACGCCTAGGAGATC  \\\n",
       "0                               0  ...                           0   \n",
       "1                               0  ...                           0   \n",
       "2                               0  ...                           0   \n",
       "3                               0  ...                           0   \n",
       "4                               0  ...                           0   \n",
       "...                           ...  ...                         ...   \n",
       "20160                           0  ...                           0   \n",
       "20161                           0  ...                           0   \n",
       "20162                           0  ...                           0   \n",
       "20163                           0  ...                           0   \n",
       "20164                           0  ...                           0   \n",
       "\n",
       "       Brain_1.TTAACTAAAGTTTATGTA  Brain_1.GTCCCGGGACATAGGACT  \\\n",
       "0                               0                           0   \n",
       "1                               0                           0   \n",
       "2                               0                           0   \n",
       "3                               0                           0   \n",
       "4                               0                           0   \n",
       "...                           ...                         ...   \n",
       "20160                           0                           0   \n",
       "20161                           0                           0   \n",
       "20162                           0                           0   \n",
       "20163                           0                           0   \n",
       "20164                           0                           0   \n",
       "\n",
       "       Brain_1.CCGCTAGGGTTTGCTCAA  Brain_1.TGATCAGCTGTGTCAAAG  \\\n",
       "0                               0                           0   \n",
       "1                               0                           0   \n",
       "2                               0                           0   \n",
       "3                               0                           0   \n",
       "4                               0                           0   \n",
       "...                           ...                         ...   \n",
       "20160                           0                           0   \n",
       "20161                           0                           0   \n",
       "20162                           0                           0   \n",
       "20163                           0                           0   \n",
       "20164                           0                           0   \n",
       "\n",
       "       Brain_1.TCACTTGAATTATGAAGC  Brain_1.AAGTACGCTGTGTATGTA  \\\n",
       "0                               0                           0   \n",
       "1                               0                           0   \n",
       "2                               0                           1   \n",
       "3                               0                           0   \n",
       "4                               0                           0   \n",
       "...                           ...                         ...   \n",
       "20160                           0                           0   \n",
       "20161                           0                           0   \n",
       "20162                           0                           0   \n",
       "20163                           0                           0   \n",
       "20164                           0                           0   \n",
       "\n",
       "       Brain_1.CCTAGATAGAGAATTTGC  Brain_1.CATCCCATTTGCGGCTGC  \\\n",
       "0                               0                           0   \n",
       "1                               0                           0   \n",
       "2                               0                           0   \n",
       "3                               0                           0   \n",
       "4                               1                           0   \n",
       "...                           ...                         ...   \n",
       "20160                           0                           0   \n",
       "20161                           0                           0   \n",
       "20162                           0                           0   \n",
       "20163                           0                           0   \n",
       "20164                           0                           0   \n",
       "\n",
       "       Brain_1.CAAAGTGGGTTTAGCGAG  \n",
       "0                               0  \n",
       "1                               0  \n",
       "2                               0  \n",
       "3                               0  \n",
       "4                               0  \n",
       "...                           ...  \n",
       "20160                           0  \n",
       "20161                           0  \n",
       "20162                           0  \n",
       "20163                           0  \n",
       "20164                           0  \n",
       "\n",
       "[20165 rows x 3680 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0610005C13Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0610007P14Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0610009B22Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0610009E02Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0610009L18Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20160</th>\n",
       "      <td>n-R5s28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20161</th>\n",
       "      <td>n-R5s37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20162</th>\n",
       "      <td>n-R5s52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20163</th>\n",
       "      <td>n-R5s88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20164</th>\n",
       "      <td>n-R5s98</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20165 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_name\n",
       "0      0610005C13Rik\n",
       "1      0610007P14Rik\n",
       "2      0610009B22Rik\n",
       "3      0610009E02Rik\n",
       "4      0610009L18Rik\n",
       "...              ...\n",
       "20160        n-R5s28\n",
       "20161        n-R5s37\n",
       "20162        n-R5s52\n",
       "20163        n-R5s88\n",
       "20164        n-R5s98\n",
       "\n",
       "[20165 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Exception: Must coerce data gxc matrix into common format ###\n",
    "\n",
    "if 'coerced' not in sample_Docket.gxc.path:\n",
    "    genes_matrix = pd.read_csv(sample_Docket.gxc.path, sep = ' ')\n",
    "    \n",
    "    genes_matrix.insert(0, 'gene_name', genes_matrix.index)\n",
    "    genes_matrix.reset_index(inplace = True, drop = True)\n",
    "    display(genes_matrix)\n",
    "\n",
    "    new_matrix = copy.deepcopy(sample_Docket.gxc)\n",
    "    new_matrix.filename = sample_Docket.gxc.filename.replace('.txt', '_coerced.txt')\n",
    "    new_matrix.path = new_matrix.directory + new_matrix.filename\n",
    "\n",
    "    genes_matrix.to_csv(new_matrix.path, index = None, sep = '\\t')\n",
    "    sample_Docket.gxc = new_matrix\n",
    "else:\n",
    "    genes_matrix = pd.read_csv(sample_Docket.gxc.path, sep = '\\t')\n",
    "    display(genes_matrix)\n",
    "    \n",
    "#################################################################\n",
    "\n",
    "gxc_genes_list = pd.DataFrame({'gene_name':genes_matrix['gene_name']})\n",
    "display(gxc_genes_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jvfcFNT2FuaZ"
   },
   "source": [
    "# 3. Get mapping identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1.0</td>\n",
       "      <td>195471971.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id0;Dbxref=taxon:10090;Name=1;chromosome=1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>BestRefSeq%2CGnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>3199731.0</td>\n",
       "      <td>3671742.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Dbxref=GeneID:497097,MGI:MGI:3528744;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3199731.0</td>\n",
       "      <td>3671742.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Dbxref=GeneID:497097,Genb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552.0</td>\n",
       "      <td>3671742.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Dbxref=GeneID:497097,Genban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702.0</td>\n",
       "      <td>3421901.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Dbxref=GeneID:497097,Genban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284598</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>15515.0</td>\n",
       "      <td>15558.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169987;Note=ETAS2%3B extended terminatio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284599</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>16035.0</td>\n",
       "      <td>16058.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169988;Note=CSB1%3B conserved sequencing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284600</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>16089.0</td>\n",
       "      <td>16104.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169989;Note=CSB2%3B conserved sequencing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284601</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>16114.0</td>\n",
       "      <td>16131.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169990;Note=CSB3%3B conserved sequencing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284602</th>\n",
       "      <td>###</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2284603 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0                    1       2          3            4  \\\n",
       "0        NC_000067.6               RefSeq  region        1.0  195471971.0   \n",
       "1        NC_000067.6  BestRefSeq%2CGnomon    gene  3199731.0    3671742.0   \n",
       "2        NC_000067.6               Gnomon    mRNA  3199731.0    3671742.0   \n",
       "3        NC_000067.6               Gnomon    exon  3670552.0    3671742.0   \n",
       "4        NC_000067.6               Gnomon    exon  3421702.0    3421901.0   \n",
       "...              ...                  ...     ...        ...          ...   \n",
       "2284598  NC_005089.1               RefSeq  region    15515.0      15558.0   \n",
       "2284599  NC_005089.1               RefSeq  region    16035.0      16058.0   \n",
       "2284600  NC_005089.1               RefSeq  region    16089.0      16104.0   \n",
       "2284601  NC_005089.1               RefSeq  region    16114.0      16131.0   \n",
       "2284602          ###                  NaN     NaN        NaN          NaN   \n",
       "\n",
       "           5    6    7                                                  8  \n",
       "0          .    +    .  ID=id0;Dbxref=taxon:10090;Name=1;chromosome=1;...  \n",
       "1          .    -    .  ID=gene0;Dbxref=GeneID:497097,MGI:MGI:3528744;...  \n",
       "2          .    -    .  ID=rna0;Parent=gene0;Dbxref=GeneID:497097,Genb...  \n",
       "3          .    -    .  ID=id1;Parent=rna0;Dbxref=GeneID:497097,Genban...  \n",
       "4          .    -    .  ID=id2;Parent=rna0;Dbxref=GeneID:497097,Genban...  \n",
       "...      ...  ...  ...                                                ...  \n",
       "2284598    .    +    .  ID=id1169987;Note=ETAS2%3B extended terminatio...  \n",
       "2284599    .    +    .  ID=id1169988;Note=CSB1%3B conserved sequencing...  \n",
       "2284600    .    +    .  ID=id1169989;Note=CSB2%3B conserved sequencing...  \n",
       "2284601    .    +    .  ID=id1169990;Note=CSB3%3B conserved sequencing...  \n",
       "2284602  NaN  NaN  NaN                                                NaN  \n",
       "\n",
       "[2284603 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'ID=id1;Parent=rna0;Dbxref=GeneID:497097,Genbank:XM_006495550.2,MGI:MGI:3528744;gbkey=mRNA;gene=Xkr4;product=X Kell blood group precursor related family member 4%2C transcript variant X1;transcript_id=XM_006495550.2'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load in the original GFF-based annotation\n",
    "models = pd.read_csv(sample_Docket.annot.path, skiprows = 7, header = None, sep = '\\t')\n",
    "display(models)\n",
    "\n",
    "attributes_column = 8\n",
    "\n",
    "# Check the structure of fields in the GFF additional fields section\n",
    "display(models[attributes_column][3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Dbxref</th>\n",
       "      <th>dbxref_dict</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1.0</td>\n",
       "      <td>195471971.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id0;Dbxref=taxon:10090;Name=1;chromosome=1;...</td>\n",
       "      <td>{'ID': 'id0', 'Dbxref': 'taxon:10090', 'Name':...</td>\n",
       "      <td>None</td>\n",
       "      <td>taxon:10090</td>\n",
       "      <td>{'taxon': '10090'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>BestRefSeq%2CGnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>3199731.0</td>\n",
       "      <td>3671742.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Dbxref=GeneID:497097,MGI:MGI:3528744;...</td>\n",
       "      <td>{'ID': 'gene0', 'Dbxref': 'GeneID:497097,MGI:M...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>GeneID:497097,MGI:MGI:3528744</td>\n",
       "      <td>{'GeneID': '497097', 'MGI': '3528744'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3199731.0</td>\n",
       "      <td>3671742.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Dbxref=GeneID:497097,Genb...</td>\n",
       "      <td>{'ID': 'rna0', 'Parent': 'gene0', 'Dbxref': 'G...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>GeneID:497097,Genbank:XM_006495550.2,MGI:MGI:3...</td>\n",
       "      <td>{'GeneID': '497097', 'Genbank': 'XM_006495550....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552.0</td>\n",
       "      <td>3671742.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Dbxref=GeneID:497097,Genban...</td>\n",
       "      <td>{'ID': 'id1', 'Parent': 'rna0', 'Dbxref': 'Gen...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>GeneID:497097,Genbank:XM_006495550.2,MGI:MGI:3...</td>\n",
       "      <td>{'GeneID': '497097', 'Genbank': 'XM_006495550....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702.0</td>\n",
       "      <td>3421901.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Dbxref=GeneID:497097,Genban...</td>\n",
       "      <td>{'ID': 'id2', 'Parent': 'rna0', 'Dbxref': 'Gen...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>GeneID:497097,Genbank:XM_006495550.2,MGI:MGI:3...</td>\n",
       "      <td>{'GeneID': '497097', 'Genbank': 'XM_006495550....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284597</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>15451.0</td>\n",
       "      <td>15509.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169986;Note=ETAS1%3B extended terminatio...</td>\n",
       "      <td>{'ID': 'id1169986', 'Note': 'ETAS1%3B extended...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284598</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>15515.0</td>\n",
       "      <td>15558.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169987;Note=ETAS2%3B extended terminatio...</td>\n",
       "      <td>{'ID': 'id1169987', 'Note': 'ETAS2%3B extended...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284599</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>16035.0</td>\n",
       "      <td>16058.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169988;Note=CSB1%3B conserved sequencing...</td>\n",
       "      <td>{'ID': 'id1169988', 'Note': 'CSB1%3B conserved...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284600</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>16089.0</td>\n",
       "      <td>16104.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169989;Note=CSB2%3B conserved sequencing...</td>\n",
       "      <td>{'ID': 'id1169989', 'Note': 'CSB2%3B conserved...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284601</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>16114.0</td>\n",
       "      <td>16131.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169990;Note=CSB3%3B conserved sequencing...</td>\n",
       "      <td>{'ID': 'id1169990', 'Note': 'CSB3%3B conserved...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2284240 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0                    1       2          3            4  5  \\\n",
       "0        NC_000067.6               RefSeq  region        1.0  195471971.0  .   \n",
       "1        NC_000067.6  BestRefSeq%2CGnomon    gene  3199731.0    3671742.0  .   \n",
       "2        NC_000067.6               Gnomon    mRNA  3199731.0    3671742.0  .   \n",
       "3        NC_000067.6               Gnomon    exon  3670552.0    3671742.0  .   \n",
       "4        NC_000067.6               Gnomon    exon  3421702.0    3421901.0  .   \n",
       "...              ...                  ...     ...        ...          ... ..   \n",
       "2284597  NC_005089.1               RefSeq  region    15451.0      15509.0  .   \n",
       "2284598  NC_005089.1               RefSeq  region    15515.0      15558.0  .   \n",
       "2284599  NC_005089.1               RefSeq  region    16035.0      16058.0  .   \n",
       "2284600  NC_005089.1               RefSeq  region    16089.0      16104.0  .   \n",
       "2284601  NC_005089.1               RefSeq  region    16114.0      16131.0  .   \n",
       "\n",
       "         6  7                                                  8  \\\n",
       "0        +  .  ID=id0;Dbxref=taxon:10090;Name=1;chromosome=1;...   \n",
       "1        -  .  ID=gene0;Dbxref=GeneID:497097,MGI:MGI:3528744;...   \n",
       "2        -  .  ID=rna0;Parent=gene0;Dbxref=GeneID:497097,Genb...   \n",
       "3        -  .  ID=id1;Parent=rna0;Dbxref=GeneID:497097,Genban...   \n",
       "4        -  .  ID=id2;Parent=rna0;Dbxref=GeneID:497097,Genban...   \n",
       "...     .. ..                                                ...   \n",
       "2284597  +  .  ID=id1169986;Note=ETAS1%3B extended terminatio...   \n",
       "2284598  +  .  ID=id1169987;Note=ETAS2%3B extended terminatio...   \n",
       "2284599  +  .  ID=id1169988;Note=CSB1%3B conserved sequencing...   \n",
       "2284600  +  .  ID=id1169989;Note=CSB2%3B conserved sequencing...   \n",
       "2284601  +  .  ID=id1169990;Note=CSB3%3B conserved sequencing...   \n",
       "\n",
       "                                          field_dictionary gene_name  \\\n",
       "0        {'ID': 'id0', 'Dbxref': 'taxon:10090', 'Name':...      None   \n",
       "1        {'ID': 'gene0', 'Dbxref': 'GeneID:497097,MGI:M...      Xkr4   \n",
       "2        {'ID': 'rna0', 'Parent': 'gene0', 'Dbxref': 'G...      Xkr4   \n",
       "3        {'ID': 'id1', 'Parent': 'rna0', 'Dbxref': 'Gen...      Xkr4   \n",
       "4        {'ID': 'id2', 'Parent': 'rna0', 'Dbxref': 'Gen...      Xkr4   \n",
       "...                                                    ...       ...   \n",
       "2284597  {'ID': 'id1169986', 'Note': 'ETAS1%3B extended...      None   \n",
       "2284598  {'ID': 'id1169987', 'Note': 'ETAS2%3B extended...      None   \n",
       "2284599  {'ID': 'id1169988', 'Note': 'CSB1%3B conserved...      None   \n",
       "2284600  {'ID': 'id1169989', 'Note': 'CSB2%3B conserved...      None   \n",
       "2284601  {'ID': 'id1169990', 'Note': 'CSB3%3B conserved...      None   \n",
       "\n",
       "                                                    Dbxref  \\\n",
       "0                                              taxon:10090   \n",
       "1                            GeneID:497097,MGI:MGI:3528744   \n",
       "2        GeneID:497097,Genbank:XM_006495550.2,MGI:MGI:3...   \n",
       "3        GeneID:497097,Genbank:XM_006495550.2,MGI:MGI:3...   \n",
       "4        GeneID:497097,Genbank:XM_006495550.2,MGI:MGI:3...   \n",
       "...                                                    ...   \n",
       "2284597                                               None   \n",
       "2284598                                               None   \n",
       "2284599                                               None   \n",
       "2284600                                               None   \n",
       "2284601                                               None   \n",
       "\n",
       "                                               dbxref_dict  \n",
       "0                                       {'taxon': '10090'}  \n",
       "1                   {'GeneID': '497097', 'MGI': '3528744'}  \n",
       "2        {'GeneID': '497097', 'Genbank': 'XM_006495550....  \n",
       "3        {'GeneID': '497097', 'Genbank': 'XM_006495550....  \n",
       "4        {'GeneID': '497097', 'Genbank': 'XM_006495550....  \n",
       "...                                                    ...  \n",
       "2284597                                               None  \n",
       "2284598                                               None  \n",
       "2284599                                               None  \n",
       "2284600                                               None  \n",
       "2284601                                               None  \n",
       "\n",
       "[2284240 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Remove any rows with NaNs\n",
    "models.dropna(inplace = True)\n",
    "\n",
    "# Extract field and database cross-ref (dbxref) information into columns\n",
    "models['field_dictionary'] = models[attributes_column].apply(convert_fields_to_dict_gff)\n",
    "models['gene_name'] = [d.get('gene') for d in models['field_dictionary']]\n",
    "models['Dbxref'] = [d.get('Dbxref') for d in models['field_dictionary']]\n",
    "models['dbxref_dict'] = models['Dbxref'].apply(convert_dbxref_to_dict)\n",
    "\n",
    "display(models)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Extract gene IDs for mapping to UniprotKB\n",
    "Specify which set of identifiers will be use to query the [Uniprot ID Mapping Tool](https://www.uniprot.org/id-mapping) via API.\n",
    "\n",
    "If using an identifier from the `dbxref_dict`, specify the name via string in the `dbxref_datafield` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>MGI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>MGI:3528744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Gm18956</td>\n",
       "      <td>MGI:5011141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Gm7341</td>\n",
       "      <td>MGI:3648497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>Gm6101</td>\n",
       "      <td>MGI:3644986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>Rp1</td>\n",
       "      <td>MGI:1341105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284583</th>\n",
       "      <td>ND6</td>\n",
       "      <td>MGI:102495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284585</th>\n",
       "      <td>TrnE</td>\n",
       "      <td>MGI:102488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284588</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>MGI:102501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284590</th>\n",
       "      <td>TrnT</td>\n",
       "      <td>MGI:102473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284593</th>\n",
       "      <td>TrnP</td>\n",
       "      <td>MGI:102478</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>40680 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name          MGI\n",
       "1            Xkr4  MGI:3528744\n",
       "17        Gm18956  MGI:5011141\n",
       "36         Gm7341  MGI:3648497\n",
       "136        Gm6101  MGI:3644986\n",
       "137           Rp1  MGI:1341105\n",
       "...           ...          ...\n",
       "2284583       ND6   MGI:102495\n",
       "2284585      TrnE   MGI:102488\n",
       "2284588      CYTB   MGI:102501\n",
       "2284590      TrnT   MGI:102473\n",
       "2284593      TrnP   MGI:102478\n",
       "\n",
       "[40680 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dbxref_datafield = 'MGI'\n",
    "\n",
    "\n",
    "models.dropna(axis = 0, subset = ['dbxref_dict'], inplace = True)\n",
    "models[dbxref_datafield] = [d.get(dbxref_datafield) for d in models['dbxref_dict']]\n",
    "\n",
    "models_subset = models[['gene_name', dbxref_datafield]].dropna().drop_duplicates()\n",
    "\n",
    "### EXCEPTION ###\n",
    "# added extra info because the MGI prefix contains a colon\n",
    "models_subset['MGI'] = 'MGI:' + models_subset[dbxref_datafield]\n",
    "\n",
    "\n",
    "display(models_subset)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Generate gene list file to query Uniprot ID Mapping API\n",
    "Generate a text file ending in `_ids.txt` for sending to the ID mapping API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 40680 gene ids to ../../output/Mmus_adultbrain/Mmus_adultbrain_MGI_ids.txt\n"
     ]
    }
   ],
   "source": [
    "gene_list = models_subset[dbxref_datafield].unique()\n",
    "\n",
    "genelist_object = GeneListFile(\n",
    "    filename = '', # filename is automatically generated from sampledict\n",
    "    sampledict = species_SampleDict,\n",
    "    sources = [sample_Docket.annot],\n",
    "    genes = gene_list,\n",
    "    identifier = dbxref_datafield\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Query Uniprot ID Mapping API\n",
    "Specify the `from_type` variable based on the Uniprot name of the identifier.\n",
    "\n",
    "The table below lists some databases and the `from_type` string that the API accepts for that datatype.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "| datatype | `from_type` string | description |\n",
    "| ---: | :--- | :--- |\n",
    "| Mouse Genome Informatics | `MGI` | ID starts with 'MGI:' |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  470k    0    52  100  470k     20   188k  0:00:02  0:00:02 --:--:--  189k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100    24    0    24    0     0     34      0 --:--:-- --:--:-- --:--:--    34\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>From</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>Q5GH67</td>\n",
       "      <td>XKR4_MOUSE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>XK-related protein 4 [Cleaved into: XK-related...</td>\n",
       "      <td>Xkr4 Xrg4</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>Q5DTU2</td>\n",
       "      <td>Q5DTU2_MOUSE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>MKIAA1889 protein</td>\n",
       "      <td>Xkr4 AY534250 mKIAA1889</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>P56716</td>\n",
       "      <td>RP1_MOUSE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Oxygen-regulated protein 1 (Retinitis pigmento...</td>\n",
       "      <td>Rp1 Orp1 Rp1h</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>2095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>A0A140LHJ6</td>\n",
       "      <td>A0A140LHJ6_MOUSE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Oxygen-regulated protein 1</td>\n",
       "      <td>Rp1</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>1371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>Q4VA02</td>\n",
       "      <td>Q4VA02_MOUSE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Rp1 protein</td>\n",
       "      <td>Rp1 Rp1h</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>2096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78494</th>\n",
       "      <td>MGI:102501</td>\n",
       "      <td>P00158</td>\n",
       "      <td>CYB_MOUSE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Cytochrome b (Complex III subunit 3) (Complex ...</td>\n",
       "      <td>Mt-Cyb Cob Cytb mt-Cytb Mtcyb</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78495</th>\n",
       "      <td>MGI:102501</td>\n",
       "      <td>Q36290</td>\n",
       "      <td>Q36290_MOUSE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Cytochrome b</td>\n",
       "      <td>mt-Cytb</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78496</th>\n",
       "      <td>MGI:102501</td>\n",
       "      <td>Q3V2F9</td>\n",
       "      <td>Q3V2F9_MOUSE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Cytochrome b</td>\n",
       "      <td>mt-Cytb</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78497</th>\n",
       "      <td>MGI:102501</td>\n",
       "      <td>Q3V2T7</td>\n",
       "      <td>Q3V2T7_MOUSE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Cytochrome b (Complex III subunit 3) (Complex ...</td>\n",
       "      <td>mt-Cytb</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78498</th>\n",
       "      <td>MGI:102501</td>\n",
       "      <td>Q7JCZ3</td>\n",
       "      <td>Q7JCZ3_MOUSE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Cytochrome b</td>\n",
       "      <td>mt-Cytb CYTB Cytb cytb mt-CYTB</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>381</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78499 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              From       Entry        Entry Name    Reviewed  \\\n",
       "0      MGI:3528744      Q5GH67        XKR4_MOUSE    reviewed   \n",
       "1      MGI:3528744      Q5DTU2      Q5DTU2_MOUSE  unreviewed   \n",
       "2      MGI:1341105      P56716         RP1_MOUSE    reviewed   \n",
       "3      MGI:1341105  A0A140LHJ6  A0A140LHJ6_MOUSE  unreviewed   \n",
       "4      MGI:1341105      Q4VA02      Q4VA02_MOUSE  unreviewed   \n",
       "...            ...         ...               ...         ...   \n",
       "78494   MGI:102501      P00158         CYB_MOUSE    reviewed   \n",
       "78495   MGI:102501      Q36290      Q36290_MOUSE  unreviewed   \n",
       "78496   MGI:102501      Q3V2F9      Q3V2F9_MOUSE  unreviewed   \n",
       "78497   MGI:102501      Q3V2T7      Q3V2T7_MOUSE  unreviewed   \n",
       "78498   MGI:102501      Q7JCZ3      Q7JCZ3_MOUSE  unreviewed   \n",
       "\n",
       "                                           Protein names  \\\n",
       "0      XK-related protein 4 [Cleaved into: XK-related...   \n",
       "1                                      MKIAA1889 protein   \n",
       "2      Oxygen-regulated protein 1 (Retinitis pigmento...   \n",
       "3                             Oxygen-regulated protein 1   \n",
       "4                                            Rp1 protein   \n",
       "...                                                  ...   \n",
       "78494  Cytochrome b (Complex III subunit 3) (Complex ...   \n",
       "78495                                       Cytochrome b   \n",
       "78496                                       Cytochrome b   \n",
       "78497  Cytochrome b (Complex III subunit 3) (Complex ...   \n",
       "78498                                       Cytochrome b   \n",
       "\n",
       "                           Gene Names              Organism  Length  \n",
       "0                           Xkr4 Xrg4  Mus musculus (Mouse)     647  \n",
       "1             Xkr4 AY534250 mKIAA1889  Mus musculus (Mouse)     106  \n",
       "2                       Rp1 Orp1 Rp1h  Mus musculus (Mouse)    2095  \n",
       "3                                 Rp1  Mus musculus (Mouse)    1371  \n",
       "4                            Rp1 Rp1h  Mus musculus (Mouse)    2096  \n",
       "...                               ...                   ...     ...  \n",
       "78494   Mt-Cyb Cob Cytb mt-Cytb Mtcyb  Mus musculus (Mouse)     381  \n",
       "78495                         mt-Cytb  Mus musculus (Mouse)     366  \n",
       "78496                         mt-Cytb  Mus musculus (Mouse)     381  \n",
       "78497                         mt-Cytb  Mus musculus (Mouse)      94  \n",
       "78498  mt-Cytb CYTB Cytb cytb mt-CYTB  Mus musculus (Mouse)     381  \n",
       "\n",
       "[78499 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from_type = 'MGI'\n",
    "to_type = 'UniProtKB'\n",
    "\n",
    "uniprot_idmm_object = genelist_object.get_uniprot_ids(ID_MAPPER_LOC, from_type, to_type)\n",
    "sample_Docket.add_keyfiles({'uniprot_idmm': uniprot_idmm_object})\n",
    "\n",
    "uniprot_idmm = pd.read_csv(sample_Docket.uniprot_idmm.path, sep = '\\t')\n",
    "display(uniprot_idmm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Extract results and generate Uniprot IDMM\n",
    "Generates an idmm that links `gene_name`, the `dbxref_datafield` seleted above, and `uniprot_id` returned by API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_12872/3311337425.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  uniprot_idpairs.rename(columns = {'From': dbxref_datafield, 'Entry': 'uniprot_id'}, inplace = True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MGI</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>Q5GH67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>Q5DTU2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>P56716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>A0A140LHJ6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>Q4VA02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78494</th>\n",
       "      <td>MGI:102501</td>\n",
       "      <td>P00158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78495</th>\n",
       "      <td>MGI:102501</td>\n",
       "      <td>Q36290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78496</th>\n",
       "      <td>MGI:102501</td>\n",
       "      <td>Q3V2F9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78497</th>\n",
       "      <td>MGI:102501</td>\n",
       "      <td>Q3V2T7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78498</th>\n",
       "      <td>MGI:102501</td>\n",
       "      <td>Q7JCZ3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78499 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               MGI  uniprot_id\n",
       "0      MGI:3528744      Q5GH67\n",
       "1      MGI:3528744      Q5DTU2\n",
       "2      MGI:1341105      P56716\n",
       "3      MGI:1341105  A0A140LHJ6\n",
       "4      MGI:1341105      Q4VA02\n",
       "...            ...         ...\n",
       "78494   MGI:102501      P00158\n",
       "78495   MGI:102501      Q36290\n",
       "78496   MGI:102501      Q3V2F9\n",
       "78497   MGI:102501      Q3V2T7\n",
       "78498   MGI:102501      Q7JCZ3\n",
       "\n",
       "[78499 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>MGI</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>Q5GH67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>MGI:3528744</td>\n",
       "      <td>Q5DTU2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Rp1</td>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>P56716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rp1</td>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>A0A140LHJ6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rp1</td>\n",
       "      <td>MGI:1341105</td>\n",
       "      <td>Q4VA02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78494</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>MGI:102501</td>\n",
       "      <td>P00158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78495</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>MGI:102501</td>\n",
       "      <td>Q36290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78496</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>MGI:102501</td>\n",
       "      <td>Q3V2F9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78497</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>MGI:102501</td>\n",
       "      <td>Q3V2T7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78498</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>MGI:102501</td>\n",
       "      <td>Q7JCZ3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78499 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name          MGI  uniprot_id\n",
       "0          Xkr4  MGI:3528744      Q5GH67\n",
       "1          Xkr4  MGI:3528744      Q5DTU2\n",
       "2           Rp1  MGI:1341105      P56716\n",
       "3           Rp1  MGI:1341105  A0A140LHJ6\n",
       "4           Rp1  MGI:1341105      Q4VA02\n",
       "...         ...          ...         ...\n",
       "78494      CYTB   MGI:102501      P00158\n",
       "78495      CYTB   MGI:102501      Q36290\n",
       "78496      CYTB   MGI:102501      Q3V2F9\n",
       "78497      CYTB   MGI:102501      Q3V2T7\n",
       "78498      CYTB   MGI:102501      Q7JCZ3\n",
       "\n",
       "[78499 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "uniprot_idpairs = uniprot_idmm[['From', 'Entry']]\n",
    "uniprot_idpairs.rename(columns = {'From': dbxref_datafield, 'Entry': 'uniprot_id'}, inplace = True)\n",
    "display(uniprot_idpairs)\n",
    "\n",
    "uniprot_output_idmm = models_subset.merge(uniprot_idpairs, on = dbxref_datafield)\n",
    "display(uniprot_output_idmm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate a filename and file for the idmm\n",
    "uniprot_output_idmm_filename = '_'.join([species_prefix, conditions, 'uniprot-idmm.tsv'])\n",
    "uniprot_output_idmm_object = IdmmFile(uniprot_output_idmm_filename, species_SampleDict, kind = 'uniprot_idmm', sources = [sample_Docket.annot])\n",
    "\n",
    "# save to file and add to the Docket\n",
    "uniprot_output_idmm.to_csv(uniprot_output_idmm_object.path, sep = '\\t')\n",
    "sample_Docket.add_keyfile(uniprot_output_idmm_object, 'uniprot_idmm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8. Convert GFF to GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted file GCF_000001635.23_GRCm38.p3_genomic.gtf already exists at:\n",
      " ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic.gtf\n"
     ]
    }
   ],
   "source": [
    "# convert the GFF file to GTF using gffread\n",
    "models_asgtf = sample_Docket.annot.to_gtf(GFFREAD_LOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3199731</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3199731</td>\n",
       "      <td>3207317</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3213439</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224794</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>14145</td>\n",
       "      <td>15288</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"gene48832\"; gene_name \"CYTB\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224795</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111579\"; gene_id \"gene48833\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224796</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111579\"; gene_id \"gene48833\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224797</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111580\"; gene_id \"gene48834\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224798</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111580\"; gene_id \"gene48834\"...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2224799 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0       1           2        3        4  5  6  7  \\\n",
       "0        NC_000067.6  Gnomon  transcript  3199731  3671742  .  -  .   \n",
       "1        NC_000067.6  Gnomon        exon  3199731  3207317  .  -  .   \n",
       "2        NC_000067.6  Gnomon        exon  3213439  3216968  .  -  .   \n",
       "3        NC_000067.6  Gnomon        exon  3421702  3421901  .  -  .   \n",
       "4        NC_000067.6  Gnomon        exon  3670552  3671742  .  -  .   \n",
       "...              ...     ...         ...      ...      ... .. .. ..   \n",
       "2224794  NC_005089.1  RefSeq         CDS    14145    15288  .  +  0   \n",
       "2224795  NC_005089.1  RefSeq  transcript    15289    15355  .  +  .   \n",
       "2224796  NC_005089.1  RefSeq        exon    15289    15355  .  +  .   \n",
       "2224797  NC_005089.1  RefSeq  transcript    15356    15422  .  -  .   \n",
       "2224798  NC_005089.1  RefSeq        exon    15356    15422  .  -  .   \n",
       "\n",
       "                                                         8  \n",
       "0        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "1        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "2        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "3        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "4        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "...                                                    ...  \n",
       "2224794       transcript_id \"gene48832\"; gene_name \"CYTB\";  \n",
       "2224795  transcript_id \"rna111579\"; gene_id \"gene48833\"...  \n",
       "2224796  transcript_id \"rna111579\"; gene_id \"gene48833\"...  \n",
       "2224797  transcript_id \"rna111580\"; gene_id \"gene48834\"...  \n",
       "2224798  transcript_id \"rna111580\"; gene_id \"gene48834\"...  \n",
       "\n",
       "[2224799 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'transcript_id \"rna0\"; gene_id \"gene0\"; gene_name \"Xkr4\";'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load the newly-generated GTF file as a dataframe\n",
    "models_asgtf_df = pd.read_csv(models_asgtf.path, skiprows = 0, header = None, sep = '\\t')\n",
    "\n",
    "display(models_asgtf_df)\n",
    "display(models_asgtf_df[attributes_column][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3199731</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3199731</td>\n",
       "      <td>3207317</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3213439</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224793</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14145</td>\n",
       "      <td>15288</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"gene48832\"; gene_id \"gene48832\"...</td>\n",
       "      <td>{'transcript_id': 'gene48832', 'gene_id': 'gen...</td>\n",
       "      <td>CYTB</td>\n",
       "      <td>gene48832</td>\n",
       "      <td>gene48832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224795</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111579\"; gene_id \"gene48833\"...</td>\n",
       "      <td>{'transcript_id': 'rna111579', 'gene_id': 'gen...</td>\n",
       "      <td>TrnT</td>\n",
       "      <td>gene48833</td>\n",
       "      <td>rna111579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224796</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111579\"; gene_id \"gene48833\"...</td>\n",
       "      <td>{'transcript_id': 'rna111579', 'gene_id': 'gen...</td>\n",
       "      <td>TrnT</td>\n",
       "      <td>gene48833</td>\n",
       "      <td>rna111579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224797</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111580\"; gene_id \"gene48834\"...</td>\n",
       "      <td>{'transcript_id': 'rna111580', 'gene_id': 'gen...</td>\n",
       "      <td>TrnP</td>\n",
       "      <td>gene48834</td>\n",
       "      <td>rna111580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224798</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111580\"; gene_id \"gene48834\"...</td>\n",
       "      <td>{'transcript_id': 'rna111580', 'gene_id': 'gen...</td>\n",
       "      <td>TrnP</td>\n",
       "      <td>gene48834</td>\n",
       "      <td>rna111580</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1287511 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0       1           2        3        4  5  6  7  \\\n",
       "0        NC_000067.6  Gnomon  transcript  3199731  3671742  .  -  .   \n",
       "1        NC_000067.6  Gnomon        exon  3199731  3207317  .  -  .   \n",
       "2        NC_000067.6  Gnomon        exon  3213439  3216968  .  -  .   \n",
       "3        NC_000067.6  Gnomon        exon  3421702  3421901  .  -  .   \n",
       "4        NC_000067.6  Gnomon        exon  3670552  3671742  .  -  .   \n",
       "...              ...     ...         ...      ...      ... .. .. ..   \n",
       "2224793  NC_005089.1  RefSeq  transcript    14145    15288  .  +  .   \n",
       "2224795  NC_005089.1  RefSeq  transcript    15289    15355  .  +  .   \n",
       "2224796  NC_005089.1  RefSeq        exon    15289    15355  .  +  .   \n",
       "2224797  NC_005089.1  RefSeq  transcript    15356    15422  .  -  .   \n",
       "2224798  NC_005089.1  RefSeq        exon    15356    15422  .  -  .   \n",
       "\n",
       "                                                         8  \\\n",
       "0        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "1        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "2        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "3        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "4        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "...                                                    ...   \n",
       "2224793  transcript_id \"gene48832\"; gene_id \"gene48832\"...   \n",
       "2224795  transcript_id \"rna111579\"; gene_id \"gene48833\"...   \n",
       "2224796  transcript_id \"rna111579\"; gene_id \"gene48833\"...   \n",
       "2224797  transcript_id \"rna111580\"; gene_id \"gene48834\"...   \n",
       "2224798  transcript_id \"rna111580\"; gene_id \"gene48834\"...   \n",
       "\n",
       "                                          field_dictionary gene_name  \\\n",
       "0        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...      Xkr4   \n",
       "1        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...      Xkr4   \n",
       "2        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...      Xkr4   \n",
       "3        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...      Xkr4   \n",
       "4        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...      Xkr4   \n",
       "...                                                    ...       ...   \n",
       "2224793  {'transcript_id': 'gene48832', 'gene_id': 'gen...      CYTB   \n",
       "2224795  {'transcript_id': 'rna111579', 'gene_id': 'gen...      TrnT   \n",
       "2224796  {'transcript_id': 'rna111579', 'gene_id': 'gen...      TrnT   \n",
       "2224797  {'transcript_id': 'rna111580', 'gene_id': 'gen...      TrnP   \n",
       "2224798  {'transcript_id': 'rna111580', 'gene_id': 'gen...      TrnP   \n",
       "\n",
       "           gene_id transcript_id  \n",
       "0            gene0          rna0  \n",
       "1            gene0          rna0  \n",
       "2            gene0          rna0  \n",
       "3            gene0          rna0  \n",
       "4            gene0          rna0  \n",
       "...            ...           ...  \n",
       "2224793  gene48832     gene48832  \n",
       "2224795  gene48833     rna111579  \n",
       "2224796  gene48833     rna111579  \n",
       "2224797  gene48834     rna111580  \n",
       "2224798  gene48834     rna111580  \n",
       "\n",
       "[1287511 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Use a custom function to extract useful fields from the additional fields section (column 8)\n",
    "# Pull from that dict to fill in additional useful columns\n",
    "models_asgtf_df['field_dictionary'] = models_asgtf_df[attributes_column].apply(convert_fields_to_dict_gtf)\n",
    "models_asgtf_df['gene_name'] = [d.get('gene_name') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['gene_id'] = [d.get('gene_id') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['transcript_id'] = [d.get('transcript_id') for d in models_asgtf_df['field_dictionary']]\n",
    "\n",
    "# Remove CDS annotations because they interfere with TransDecoder cDNA generation\n",
    "models_asgtf_df = models_asgtf_df[models_asgtf_df[2] != 'CDS']\n",
    "display(models_asgtf_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9. Generate gtf-idmm\n",
    "This file maps the `gene_name` to `gene_id` and `transcript_id` fields generated by the conversion from GFF to GTF, which will be needed for downstream processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>LOC105243853</td>\n",
       "      <td>gene2</td>\n",
       "      <td>rna4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224789</th>\n",
       "      <td>ND6</td>\n",
       "      <td>gene48830</td>\n",
       "      <td>gene48830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224791</th>\n",
       "      <td>TrnE</td>\n",
       "      <td>gene48831</td>\n",
       "      <td>rna111578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224793</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>gene48832</td>\n",
       "      <td>gene48832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224795</th>\n",
       "      <td>TrnT</td>\n",
       "      <td>gene48833</td>\n",
       "      <td>rna111579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224797</th>\n",
       "      <td>TrnP</td>\n",
       "      <td>gene48834</td>\n",
       "      <td>rna111580</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>116982 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_name    gene_id transcript_id\n",
       "0                Xkr4      gene0          rna0\n",
       "8                Xkr4      gene0          rna1\n",
       "15               Xkr4      gene0          rna2\n",
       "22               Xkr4      gene0          rna3\n",
       "29       LOC105243853      gene2          rna4\n",
       "...               ...        ...           ...\n",
       "2224789           ND6  gene48830     gene48830\n",
       "2224791          TrnE  gene48831     rna111578\n",
       "2224793          CYTB  gene48832     gene48832\n",
       "2224795          TrnT  gene48833     rna111579\n",
       "2224797          TrnP  gene48834     rna111580\n",
       "\n",
       "[116982 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract gene_name, gene_id, and transcript_id fields to generate an ID mapping matrix (idmm)\n",
    "idmm_df = models_asgtf_df[['gene_name', 'gene_id', 'transcript_id']].drop_duplicates()\n",
    "idmm_df.dropna(inplace = True)\n",
    "display(idmm_df)\n",
    "\n",
    "# generate a filename and file for the idmm\n",
    "idmm_filename = '_'.join([species_prefix, conditions, 'gtf-idmm.tsv'])\n",
    "idmm = IdmmFile(idmm_filename, species_SampleDict, kind = 'gtf_idmm', sources = [sample_Docket.annot])\n",
    "\n",
    "# save to file and add to the Docket\n",
    "idmm_df.to_csv(idmm.path, sep = '\\t')\n",
    "sample_Docket.add_keyfile(idmm, 'gtf_idmm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 10. Generate updated gtf\n",
    "Generated an updated GTF file using transcript_id as the key. For some datasets, transcripts do not consistently get gene names and gene IDs added, which causes Transdecoder to throw errors. This resolves that problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "models_asgtf_updated_df = models_asgtf_df.merge(idmm_df, on = 'transcript_id')\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_name': x['gene_name_y']}), axis = 1)\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_id': x['gene_id_y']}), axis = 1)\n",
    "models_asgtf_updated_df['new_fields'] = models_asgtf_updated_df['field_dictionary'].apply(convert_dict_to_fields_gtf)\n",
    "models_asgtf_updated_df = models_asgtf_updated_df[[0, 1, 2, 3, 4, 5, 6, 7, 'new_fields']]\n",
    "\n",
    "models_asgtf_updated_filename = models_asgtf.filename.replace('.gtf', '_updated.gtf')\n",
    "models_asgtf_updated = GenomeGtfFile(models_asgtf_updated_filename, species_SampleDict, GenomeFastaFile = sample_Docket.genome_fasta)\n",
    "\n",
    "models_asgtf_updated_df.to_csv(models_asgtf_updated.path, header = None, index = None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 11. Generate cDNA and peptide files\n",
    "Using the updated gtf file and genome file, generate cDNA sequence.\n",
    "\n",
    "Then, using the cDNA sequence, generate peptide sequences using transdecoder.\n",
    "\n",
    "Expect this step to take some time, probably ~20-30min."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/compute_base_probs.pl ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna 0 > ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//base_freqs.dat, checkpoint [/home/ec2-user/glial-origins/notebooks/Mmus_adultbrain/../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir/.__checkpoints_longorfs/base_freqs_file.ok] exists.\n",
      "-skipping long orf extraction, already completed earlier as per checkpoint: ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir/.__checkpoints_longorfs/TD.longorfs.ok\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//longest_orfs.cds 5000 5000 > ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000, checkpoint [/home/ec2-user/glial-origins/notebooks/Mmus_adultbrain/../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir/.__checkpoints/get_longest_orfs.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/exclude_similar_proteins.pl ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000 > ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr, checkpoint [/home/ec2-user/glial-origins/notebooks/Mmus_adultbrain/../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir/.__checkpoints/nr.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr 500 > ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest, checkpoint [/home/ec2-user/glial-origins/notebooks/Mmus_adultbrain/../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir/.__checkpoints/top_train_select.ok] exists.\n",
      "PCT_GC: 49\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/seq_n_baseprobs_to_loglikelihood_vals.pl ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//base_freqs.dat > ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//hexamer.scores, checkpoint [/home/ec2-user/glial-origins/notebooks/Mmus_adultbrain/../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir/.__checkpoints/hexamer_scores.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/score_CDS_likelihood_all_6_frames.pl ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//longest_orfs.cds ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//hexamer.scores > ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//longest_orfs.cds.scores, checkpoint [/home/ec2-user/glial-origins/notebooks/Mmus_adultbrain/../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir/.__checkpoints/score_cdss.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/select_best_ORFs_per_transcript.pl --gff3_file ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//longest_orfs.gff3 --cds_scores ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//longest_orfs.cds.scores  --min_length_auto_accept 749  > ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 , checkpoint [/home/ec2-user/glial-origins/notebooks/Mmus_adultbrain/../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir/.__checkpoints/select_best_orfs.cba57cf5b92a2cb48c08ed58d3d876a7.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/train_start_PWM.pl --transcripts ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna --selected_orfs ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest --out_prefix ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//start_refinement, checkpoint [/home/ec2-user/glial-origins/notebooks/Mmus_adultbrain/../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir/.__checkpoints/train_start_PWM.cba57cf5b92a2cb48c08ed58d3d876a7.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/start_codon_refinement.pl --transcripts ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna --gff3_file ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 --workdir ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir/ > ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3, checkpoint [/home/ec2-user/glial-origins/notebooks/Mmus_adultbrain/../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir/.__checkpoints/revise_starts.cba57cf5b92a2cb48c08ed58d3d876a7.ok] exists.\n",
      "-- Skipping CMD: cp ../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3 GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3, checkpoint [/home/ec2-user/glial-origins/notebooks/Mmus_adultbrain/../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir/.__checkpoints/copy_to_final_outputfile.cba57cf5b92a2cb48c08ed58d3d876a7.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_bed.pl GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3 > GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.bed, checkpoint [/home/ec2-user/glial-origins/notebooks/Mmus_adultbrain/../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir/.__checkpoints/make_final_bed.cba57cf5b92a2cb48c08ed58d3d876a7.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3 --fasta ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna  --genetic_code Universal > GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep, checkpoint [/home/ec2-user/glial-origins/notebooks/Mmus_adultbrain/../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir/.__checkpoints/make_final_pep.cba57cf5b92a2cb48c08ed58d3d876a7.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3 --fasta ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna --seqType CDS  --genetic_code Universal > GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.cds, checkpoint [/home/ec2-user/glial-origins/notebooks/Mmus_adultbrain/../../output/Mmus_adultbrain/Mus_musculus_adultbrain.transdecoder_dir/.__checkpoints/make_final_cds.cba57cf5b92a2cb48c08ed58d3d876a7.ok] exists.\n",
      "transdecoder is finished.  See output files GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.*\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cdna = sample_Docket.genome_fasta.get_transdecoder_cdna_gtf(models_asgtf_updated, TRANSDECODER_LOC)\n",
    "sample_Docket.add_keyfile(cdna, 'cdna')\n",
    "\n",
    "transdecoder_files = sample_Docket.cdna.to_pep_files(TDLONGORF_LOC, TDPREDICT_LOC)\n",
    "sample_Docket.add_keyfiles(transdecoder_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "iEug6TQ6Cc9D"
   },
   "source": [
    "# 12. Push files to AWS S3\n",
    "\n",
    "Iteratively moves through the file_set and file_dict variables and populates files into the right place in AWS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 13. Pickle the `sample_Docket` variable for use by the next script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_Docket.pickle()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "Xtro_bro.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
