{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3df6c5ba-21b7-4f9a-aedb-9db92d7882cf",
   "metadata": {},
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aac310d7-4800-4b49-aa99-48d455aa03d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import os\n",
    "import dill\n",
    "import datetime\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "import sys\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84395a54-f8ad-4342-bbd1-6e959c6470da",
   "metadata": {},
   "source": [
    "# 1. Collect BioFileDockets for each starting dataset\n",
    "Collects BioFileDockets for each of the species datasets in the `species_dict`.\n",
    "\n",
    "This `dict` expects key:value pairs in the form of `species`:`conditions` and is passed to the MultiSpeciesBioFileDocket class, which creates a folder in `output/` for the analysis.\n",
    "\n",
    "The folder name takes the following format:\n",
    "\n",
    "- `species_concat` + `_` + `global_conditions` + `analysis_type`\n",
    "\n",
    "Where `species_concat` is the string concatenation of the alphabetize list of species prefixes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e1ea15e7-b9de-4247-a0c8-c641879a58c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "species_dict = {\n",
    "    'Danio_rerio': 'adultbrain',\n",
    "    'Mus_musculus': 'adultbrain',\n",
    "    'Xenopus_laevis': 'adultbrain'\n",
    "}\n",
    "\n",
    "global_conditions = 'adultbrain'\n",
    "\n",
    "analysis_type = 'OrthoFinder'\n",
    "\n",
    "################\n",
    "\n",
    "sample_MSD = MultiSpeciesBioFileDocket(\n",
    "    species_dict = species_dict,\n",
    "    global_conditions = global_conditions,\n",
    "    analysis_type = analysis_type\n",
    "    )\n",
    "\n",
    "sample_MSD.make_directory()\n",
    "sample_MSD.get_BioFileDockets()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e3bf493-da0f-42bc-904d-fc4c6dd6ca2b",
   "metadata": {},
   "source": [
    "# 2. Get the peptide file paths from each BioFileDocket"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8bb0f3ca-f82a-48a8-93e1-7c0b9f60178b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Drer': <biofile_handling.TransdecoderOutFile at 0x7f966b6ee5c0>,\n",
       " 'Mmus': <biofile_handling.TransdecoderOutFile at 0x7f966b6ef490>,\n",
       " 'Xlae': <biofile_handling.TransdecoderOutFile at 0x7f966b6efac0>}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep has already been moved\n",
      "file /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep has already been moved\n",
      "file /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.pep has already been moved\n"
     ]
    }
   ],
   "source": [
    "# Extract the peptide file info from the sample_MSD\n",
    "peptide_files = {pre: sample_MSD.species_BioFileDockets[pre].transdecoder_pep for pre in sample_MSD.species_BioFileDockets}\n",
    "display(peptide_files)\n",
    "\n",
    "# Move the files from the starting directory to the OrthoFinder run directory\n",
    "for pepfile in peptide_files.values():\n",
    "    start = os.path.abspath(pepfile.path)\n",
    "    end = sample_MSD.directory + pepfile.filename\n",
    "    if not os.path.exists(end):\n",
    "        print('copying', start, 'to', end)\n",
    "        subprocess.run(['cp', start, end])\n",
    "    else:\n",
    "        print('file', start, 'has already been moved')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4963be4b-109a-416b-bc59-4459f8d3dd3b",
   "metadata": {},
   "source": [
    "# 3. Run OrthoFinder on all collected peptide files\n",
    "\n",
    "Before running, will check if analysis has already been done - if so, continues."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "03e128ac-c038-49ed-bf45-2c19b59630ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../../output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep30\n",
      "OF folder already exists at ../../output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/\n",
      "Checking for results folder\n",
      "Results folder found at ../../output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder</th>\n",
       "      <th>GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder</th>\n",
       "      <th>XENLA_9.2_genome_cDNA.fa.transdecoder</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OG0000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene10395_t.p1, gene12148_t.p1, gene1307_t.p1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OG0000001</td>\n",
       "      <td>rna75711.p2</td>\n",
       "      <td>gene10036.p1, gene10201.p2, gene10535.p1, gene...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OG0000002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene2928_t.p3, gene38029_t.p2, gene41218_t.p1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OG0000003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene264_t.p1, rna1157.p2, rna12598.p1, rna1516...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OG0000004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene50425_t.p1, rna27145.p2, rna34455.p1, rna4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32348</th>\n",
       "      <td>OG0032348</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rna97220.p1, rna97221.p1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32349</th>\n",
       "      <td>OG0032349</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rna97260.p1, rna97260.p2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32350</th>\n",
       "      <td>OG0032350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rna97447.p1, rna97448.p1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32351</th>\n",
       "      <td>OG0032351</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rna97780.p1, rna97781.p2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32352</th>\n",
       "      <td>OG0032352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rna9904.p1, rna9906.p1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32353 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Orthogroup GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder  \\\n",
       "0      OG0000000                                                NaN         \n",
       "1      OG0000001                                        rna75711.p2         \n",
       "2      OG0000002                                                NaN         \n",
       "3      OG0000003                                                NaN         \n",
       "4      OG0000004                                                NaN         \n",
       "...          ...                                                ...         \n",
       "32348  OG0032348                                                NaN         \n",
       "32349  OG0032349                                                NaN         \n",
       "32350  OG0032350                                                NaN         \n",
       "32351  OG0032351                                                NaN         \n",
       "32352  OG0032352                                                NaN         \n",
       "\n",
       "      GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder  \\\n",
       "0                                                    NaN     \n",
       "1      gene10036.p1, gene10201.p2, gene10535.p1, gene...     \n",
       "2                                                    NaN     \n",
       "3                                                    NaN     \n",
       "4                                                    NaN     \n",
       "...                                                  ...     \n",
       "32348                                                NaN     \n",
       "32349                                                NaN     \n",
       "32350                                                NaN     \n",
       "32351                                                NaN     \n",
       "32352                                                NaN     \n",
       "\n",
       "                   XENLA_9.2_genome_cDNA.fa.transdecoder  \n",
       "0      gene10395_t.p1, gene12148_t.p1, gene1307_t.p1,...  \n",
       "1                                                    NaN  \n",
       "2      gene2928_t.p3, gene38029_t.p2, gene41218_t.p1,...  \n",
       "3      gene264_t.p1, rna1157.p2, rna12598.p1, rna1516...  \n",
       "4      gene50425_t.p1, rna27145.p2, rna34455.p1, rna4...  \n",
       "...                                                  ...  \n",
       "32348                           rna97220.p1, rna97221.p1  \n",
       "32349                           rna97260.p1, rna97260.p2  \n",
       "32350                           rna97447.p1, rna97448.p1  \n",
       "32351                           rna97780.p1, rna97781.p2  \n",
       "32352                             rna9904.p1, rna9906.p1  \n",
       "\n",
       "[32353 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Generate OrthoFinder run folder information based on today's date\n",
    "OF_folder = sample_MSD.directory + 'OrthoFinder/'\n",
    "today = datetime.date.today()\n",
    "datefolder = OF_folder + 'Results_' + today.strftime('%b%d')\n",
    "print(datefolder)\n",
    "\n",
    "# Check of an OrthoFinder folder already exists; if not, runs analysis\n",
    "if not os.path.exists(OF_folder):\n",
    "    subprocess.run([ORTHOFINDER_LOC, '-f', sample_MSD.directory])\n",
    "else:\n",
    "    print('OF folder already exists at', OF_folder)\n",
    "    print('Checking for results folder')\n",
    "    folders = os.listdir(OF_folder)\n",
    "    datefolder = OF_folder + [i for i in folders if 'Results' in i][0]\n",
    "    print('Results folder found at', datefolder)\n",
    "\n",
    "# Generates OrthoGroups file object\n",
    "orthogroups_file = OrthoFinderOutputFile(\n",
    "    filename = '/Orthogroups/Orthogroups.tsv', \n",
    "    multispeciesbiofiledocket = sample_MSD,\n",
    "    directory = datefolder)\n",
    "\n",
    "orthogroups_df = pd.read_csv(orthogroups_file.path, sep = '\\t')\n",
    "display(orthogroups_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61be3acd-f098-44bb-bbf4-9c9bea6a1150",
   "metadata": {},
   "source": [
    "# 4. Generate Orthofinder_exc file from gxc file of each starting dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dbac4c37-1aff-4fdc-9992-a2203b9ef111",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Orthofinder_exc file already exists at ../../output/Drer_adultbrain/GSM3768152_Brain_8_dge_asOrthogroup.txt\n",
      "Orthofinder_exc file already exists at ../../output/Mmus_adultbrain/GSM2906405_Brain1_dge_coerced_asOrthogroup.txt\n",
      "Orthofinder_exc file already exists at ../../output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup.txt\n"
     ]
    }
   ],
   "source": [
    "# Iterates through all of the species in the Species BioFileDocket\n",
    "for pre in sample_MSD.species_BioFileDockets.keys():\n",
    "    \n",
    "    # Generates filename automatically\n",
    "    exc_filename = sample_MSD.species_BioFileDockets[pre].gxc.filename.replace('.' + sample_MSD.species_BioFileDockets[pre].gxc.filetype, '_asOrthogroup.' + sample_MSD.species_BioFileDockets[pre].gxc.filetype)\n",
    "    \n",
    "    # Generates file object\n",
    "    exc = ExcFile(\n",
    "            filename = exc_filename,\n",
    "            sampledict = sample_MSD.SampleDicts[pre],\n",
    "            gxcfile = sample_MSD.species_BioFileDockets[pre].gxc,\n",
    "            embedding = 'Orthogroup'\n",
    "            )\n",
    "    \n",
    "    # Checks whether an Orthofinder_exc file already exists; avoid re-generating if it does\n",
    "    if os.path.exists(exc.path):\n",
    "        print('Orthofinder_exc file already exists at', exc.path)\n",
    "        sample_MSD.species_BioFileDockets[pre].add_keyfile(exc, 'Orthogroup_exc')\n",
    "        \n",
    "        continue\n",
    "    \n",
    "    # Loads the gtf id mapping matrix                                                                    \n",
    "    gtf_idmm_df = pd.read_csv(sample_MSD.species_BioFileDockets[pre].gtf_idmm.path, index_col = 0, sep = '\\t')\n",
    "    \n",
    "    # Automatically gets the expected column name of the OrthoFinder file\n",
    "    species_column = sample_MSD.species_BioFileDockets[pre].cdna.filename + '.transdecoder'\n",
    "    \n",
    "    # Copies orthogroups dataframe to do transformations without modifying original\n",
    "    orthogroups_df_copy = orthogroups_df.copy(deep = True)\n",
    "    \n",
    "    # Expands orthogroups column for species-specific dataset\n",
    "    orthogroups_df_copy['protein_id'] = orthogroups_df_copy[species_column].str.split(', ')\n",
    "    orthogroups_df_copy = orthogroups_df_copy.explode('protein_id')\n",
    "    orthogroups_df_copy['transcript_id'] = orthogroups_df_copy['protein_id'].str.split('.', expand = True)[0]\n",
    "    \n",
    "    # Gets id mapping between transcript, protein, and Orthogroup ids\n",
    "    og_keys = orthogroups_df_copy[['transcript_id', 'protein_id', 'Orthogroup']].drop_duplicates()\n",
    "    og_keys.dropna(inplace = True)\n",
    "    \n",
    "    # Merges original idmm with orthogroup info, generating a new idmm to be used in downstream analysis\n",
    "    og_idmm_df = gtf_idmm_df.merge(og_keys, on = 'transcript_id')  \n",
    "    og_idmm_filename = pre + '_' + sample_MSD.SampleDicts[pre].conditions + '_og-idmm.tsv'\n",
    "    \n",
    "    og_idmm = IdmmFile(\n",
    "        filename = og_idmm_filename,\n",
    "        sampledict = sample_MSD.SampleDicts[pre],\n",
    "        kind = 'og_idmm',\n",
    "        sources = [orthogroups_file, sample_MSD.species_BioFileDockets[pre].gtf_idmm]\n",
    "        )\n",
    "    og_idmm_df.to_csv(og_idmm.path, sep = '\\t', index = None)\n",
    "    \n",
    "    # Adds the new og_idmm to the BioFileDocket for the species\n",
    "    sample_MSD.species_BioFileDockets[pre].add_keyfile(og_idmm, 'og_idmm')\n",
    "    sample_MSD.species_BioFileDockets[pre].add_keyfile(orthogroups_file, sample_MSD.species_concat + '_OGfile')\n",
    "    \n",
    "    # Extracts gene_name to Orthogroup mapping keys\n",
    "    og_gene_keys = og_idmm_df[['gene_name', 'Orthogroup']].drop_duplicates()\n",
    "    \n",
    "    # Reads in original gxc matrix file\n",
    "    gxc_df = pd.read_csv(sample_MSD.species_BioFileDockets[pre].gxc.path, sep = '\\t')\n",
    "    # Automatically gets the first column name of file for later use\n",
    "    gxc_original_dataname = gxc_df.columns[0]\n",
    "    # Renames that column to 'gene_name' for easier merging\n",
    "    gxc_df.rename(columns = {gxc_original_dataname: 'gene_name'}, inplace = True)\n",
    "    # Merges gxc with orthogroup gene keys to generate exc dataframe\n",
    "    exc_df = og_gene_keys.merge(gxc_df, on = 'gene_name')\n",
    "    \n",
    "    # Removes 'gene_name' column\n",
    "    exc_df = exc_df.drop(columns = 'gene_name')\n",
    "    # Aggregates read counts per cell by Orthogroup ID\n",
    "    exc_df = exc_df.groupby('Orthogroup').agg({i: ('first' if i == 'Orthogroup' else 'sum') for i in exc_df.columns}).reset_index(drop = True)\n",
    "    \n",
    "    # Saves new exc matrix to file and puts it into the species BioFileDocket\n",
    "    exc_df.to_csv(exc.path, sep = '\\t', index = None)\n",
    "    sample_MSD.species_BioFileDockets[pre].add_keyfile(exc, 'Orthogroup_exc')\n",
    "    \n",
    "    display(exc_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c135c9ee-8f04-4bf5-9916-7b3eeb0f42b0",
   "metadata": {},
   "source": [
    "# 5. Generate merged exc file for all samples in dataset\n",
    "\n",
    "(TO BE WRITTEN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c30778eb-5480-4fda-84a2-462932b09b6a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "984a991a-f659-464f-aaf0-62637fb3bc88",
   "metadata": {},
   "source": [
    "# 6. Pickle the species_BioFileDocket files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f873483e-f652-4d15-bc4c-fe2ff9892456",
   "metadata": {},
   "outputs": [],
   "source": [
    "for pre in sample_MSD.species_BioFileDockets.keys():\n",
    "    sample_MSD.species_BioFileDockets[pre].pickle()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1e678ce-1ec4-4a8d-9ebf-c77ed8d236b4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb57f7f7-81b2-4369-b0f5-2bc28412a0d0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "090a25cf-13eb-440e-9150-6525caeef939",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c50f981e-021c-4e4a-8198-a1ffb333ff17",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2006722c-b684-403e-a1bf-1c22ec0fa79d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
