{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3df6c5ba-21b7-4f9a-aedb-9db92d7882cf",
   "metadata": {},
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "aac310d7-4800-4b49-aa99-48d455aa03d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess, os, dill, sys\n",
    "import datetime\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84395a54-f8ad-4342-bbd1-6e959c6470da",
   "metadata": {},
   "source": [
    "# 1. Collect BioFileDockets for each starting dataset\n",
    "Collects BioFileDockets for each of the species datasets in the `species_dict`.\n",
    "\n",
    "This `dict` expects key:value pairs in the form of `species`:`conditions` and is passed to the MultiSpeciesBioFileDocket class, which creates a folder in `output/` for the analysis.\n",
    "\n",
    "The folder name takes the following format:\n",
    "\n",
    "- `species_concat` + `_` + `global_conditions` + `analysis_type`\n",
    "\n",
    "Where `species_concat` is the string concatenation of the alphabetize list of species prefixes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e1ea15e7-b9de-4247-a0c8-c641879a58c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/ already exists\n",
      "file GCF_000002035.5_GRCz10_genomic.gff already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.gff\n",
      "file GCF_000002035.5_GRCz10_genomic.fna already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.fna\n",
      "file GSM3768152_Brain_8_dge.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GSM3768152_Brain_8_dge.txt\n",
      "file Drer_adultbrain_uniprot-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_uniprot-idmm.tsv\n",
      "file Drer_adultbrain_gtf-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_gtf-idmm.tsv\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep\n",
      "file GSM3768152_Brain_8_dge_asOrthogroup.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GSM3768152_Brain_8_dge_asOrthogroup.txt\n",
      "file DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv\n",
      "file GCF_000001635.23_GRCm38.p3_genomic.gff already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic.gff\n",
      "file GCF_000001635.23_GRCm38.p3_genomic.fna already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic.fna\n",
      "file GSM2906405_Brain1_dge_coerced.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GSM2906405_Brain1_dge_coerced.txt\n",
      "file Mmus_adultbrain_uniprot-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_adultbrain_uniprot-idmm.tsv\n",
      "file Mmus_adultbrain_gtf-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_adultbrain_gtf-idmm.tsv\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.bed already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.bed\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.cds already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.cds\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3 already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep\n",
      "file Mmus_adultbrain_Brain_1_cellannot.tsv already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_adultbrain_Brain_1_cellannot.tsv\n",
      "file GSM2906405_Brain1_dge_coerced_asOrthogroup.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GSM2906405_Brain1_dge_coerced_asOrthogroup.txt\n",
      "file DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv\n",
      "file XENLA_9.2_Xenbase.gtf already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_Xenbase.gtf\n",
      "file XENLA_9.2_GCA.gff already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_GCA.gff\n",
      "file XENLA_9.2_genome.fa already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome.fa\n",
      "file GSM6214268_Xenopus_brain_COL65_dge.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge.txt\n",
      "file Xlae_adultbrain_Xenbase_UniProtIDs.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_Xenbase_UniProtIDs.txt\n",
      "file Xlae_adultbrain_uniprot-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_uniprot-idmm.tsv\n",
      "file Xlae_adultbrain_gtf-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_gtf-idmm.tsv\n",
      "file XENLA_9.2_genome_cDNA.fa already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.bed already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.bed\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.cds already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.cds\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.gff3\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.pep already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.pep\n",
      "file GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup.txt\n",
      "file DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv\n"
     ]
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "species_dict = {\n",
    "    'Danio_rerio': 'adultbrain',\n",
    "    'Mus_musculus': 'adultbrain',\n",
    "    'Xenopus_laevis': 'adultbrain'\n",
    "}\n",
    "\n",
    "global_conditions = 'adultbrain'\n",
    "\n",
    "analysis_type = 'OrthoFinder'\n",
    "\n",
    "################\n",
    "\n",
    "sample_MSD = MultiSpeciesBioFileDocket(\n",
    "    species_dict = species_dict,\n",
    "    global_conditions = global_conditions,\n",
    "    analysis_type = analysis_type\n",
    "    )\n",
    "\n",
    "sample_MSD.get_BioFileDockets()\n",
    "sample_MSD.s3_to_local()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e3bf493-da0f-42bc-904d-fc4c6dd6ca2b",
   "metadata": {},
   "source": [
    "# 2. Get the peptide file paths from each BioFileDocket"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8bb0f3ca-f82a-48a8-93e1-7c0b9f60178b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Drer': <biofile_handling.TransdecoderOutFile at 0x7f2904c98790>,\n",
       " 'Mmus': <biofile_handling.TransdecoderOutFile at 0x7f2904c99000>,\n",
       " 'Xlae': <biofile_handling.TransdecoderOutFile at 0x7f2904c99990>}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep has already been moved\n",
      "file /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep has already been moved\n",
      "file /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.pep has already been moved\n"
     ]
    }
   ],
   "source": [
    "# Extract the peptide file info from the sample_MSD\n",
    "peptide_files = {pre: sample_MSD.species_BioFileDockets[pre].transdecoder_pep for pre in sample_MSD.species_BioFileDockets}\n",
    "display(peptide_files)\n",
    "\n",
    "# Move the files from the starting directory to the OrthoFinder run directory\n",
    "for pepfile in peptide_files.values():\n",
    "    start = os.path.abspath(pepfile.path)\n",
    "    end = sample_MSD.directory + pepfile.filename\n",
    "    if not os.path.exists(end):\n",
    "        print('copying', start, 'to', end)\n",
    "        subprocess.run(['cp', start, end])\n",
    "    else:\n",
    "        print('file', start, 'has already been moved')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4963be4b-109a-416b-bc59-4459f8d3dd3b",
   "metadata": {},
   "source": [
    "# 3. Run OrthoFinder on all collected peptide files\n",
    "\n",
    "Before running, will check if analysis has already been done - if so, continues."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "03e128ac-c038-49ed-bf45-2c19b59630ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Jan04\n",
      "OF folder already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/\n",
      "Checking for results folder\n",
      "Results folder found at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder</th>\n",
       "      <th>GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder</th>\n",
       "      <th>XENLA_9.2_genome_cDNA.fa.transdecoder</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OG0000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene10395_t.p1, gene12148_t.p1, gene1307_t.p1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OG0000001</td>\n",
       "      <td>rna75711.p2</td>\n",
       "      <td>gene10036.p1, gene10201.p2, gene10535.p1, gene...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OG0000002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene2928_t.p3, gene38029_t.p2, gene41218_t.p1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OG0000003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene264_t.p1, rna1157.p2, rna12598.p1, rna1516...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OG0000004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene50425_t.p1, rna27145.p2, rna34455.p1, rna4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32348</th>\n",
       "      <td>OG0032348</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rna97220.p1, rna97221.p1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32349</th>\n",
       "      <td>OG0032349</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rna97260.p1, rna97260.p2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32350</th>\n",
       "      <td>OG0032350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rna97447.p1, rna97448.p1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32351</th>\n",
       "      <td>OG0032351</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rna97780.p1, rna97781.p2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32352</th>\n",
       "      <td>OG0032352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rna9904.p1, rna9906.p1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32353 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Orthogroup GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder  \\\n",
       "0      OG0000000                                                NaN         \n",
       "1      OG0000001                                        rna75711.p2         \n",
       "2      OG0000002                                                NaN         \n",
       "3      OG0000003                                                NaN         \n",
       "4      OG0000004                                                NaN         \n",
       "...          ...                                                ...         \n",
       "32348  OG0032348                                                NaN         \n",
       "32349  OG0032349                                                NaN         \n",
       "32350  OG0032350                                                NaN         \n",
       "32351  OG0032351                                                NaN         \n",
       "32352  OG0032352                                                NaN         \n",
       "\n",
       "      GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder  \\\n",
       "0                                                    NaN     \n",
       "1      gene10036.p1, gene10201.p2, gene10535.p1, gene...     \n",
       "2                                                    NaN     \n",
       "3                                                    NaN     \n",
       "4                                                    NaN     \n",
       "...                                                  ...     \n",
       "32348                                                NaN     \n",
       "32349                                                NaN     \n",
       "32350                                                NaN     \n",
       "32351                                                NaN     \n",
       "32352                                                NaN     \n",
       "\n",
       "                   XENLA_9.2_genome_cDNA.fa.transdecoder  \n",
       "0      gene10395_t.p1, gene12148_t.p1, gene1307_t.p1,...  \n",
       "1                                                    NaN  \n",
       "2      gene2928_t.p3, gene38029_t.p2, gene41218_t.p1,...  \n",
       "3      gene264_t.p1, rna1157.p2, rna12598.p1, rna1516...  \n",
       "4      gene50425_t.p1, rna27145.p2, rna34455.p1, rna4...  \n",
       "...                                                  ...  \n",
       "32348                           rna97220.p1, rna97221.p1  \n",
       "32349                           rna97260.p1, rna97260.p2  \n",
       "32350                           rna97447.p1, rna97448.p1  \n",
       "32351                           rna97780.p1, rna97781.p2  \n",
       "32352                             rna9904.p1, rna9906.p1  \n",
       "\n",
       "[32353 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Generate OrthoFinder run folder information based on today's date\n",
    "OF_folder = sample_MSD.directory + 'OrthoFinder/'\n",
    "today = datetime.date.today()\n",
    "datefolder = OF_folder + 'Results_' + today.strftime('%b%d')\n",
    "print(datefolder)\n",
    "\n",
    "# Check of an OrthoFinder folder already exists; if not, runs analysis\n",
    "if not os.path.exists(OF_folder):\n",
    "    subprocess.run([ORTHOFINDER_LOC, '-f', sample_MSD.directory])\n",
    "else:\n",
    "    print('OF folder already exists at', OF_folder)\n",
    "    print('Checking for results folder')\n",
    "    folders = os.listdir(OF_folder)\n",
    "    datefolder = OF_folder + [i for i in folders if 'Results' in i][0]\n",
    "    print('Results folder found at', datefolder)\n",
    "\n",
    "# Generates OrthoGroups file object\n",
    "orthogroups_file = OrthoFinderOutputFile(\n",
    "    sampledict = sample_MSD.sampledict,\n",
    "    species_dict = sample_MSD.species_dict,\n",
    "    filename = '/'.join(datefolder.split('/')[-2:]) + '/Orthogroups/Orthogroups.tsv'\n",
    ")\n",
    "\n",
    "orthogroups_df = pd.read_csv(orthogroups_file.path, sep = '\\t')\n",
    "display(orthogroups_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61be3acd-f098-44bb-bbf4-9c9bea6a1150",
   "metadata": {},
   "source": [
    "# 4. Generate Orthofinder_exc file from gxc file of each starting dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dbac4c37-1aff-4fdc-9992-a2203b9ef111",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using Orthogroup embeddings as expected from OrthoFinder\n",
      "Expanding column GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder\n",
      "Extracting keys for Orthogroup\n",
      "Saving keys to Drer_adultbrain_og_idmm.tsv\n",
      "Done saving Drer_adultbrain_og_idmm.tsv\n",
      "key \"og_idmm\" already exists, ignoring\n",
      "key \"DrerMmusXlae_OGfile\" already exists, ignoring\n",
      "Generating exc file at /home/ec2-user/glial-origins/output/Drer_adultbrain/GSM3768152_Brain_8_dge_asOrthogroup.txt\n",
      "Preview of exc file: /home/ec2-user/glial-origins/output/Drer_adultbrain/GSM3768152_Brain_8_dge_asOrthogroup.txt\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>ACAATATATTGTACCTGA</th>\n",
       "      <th>ACGTTGATGGCGTAGAGA</th>\n",
       "      <th>AACCTAACCTGAATTTGC</th>\n",
       "      <th>CTCGCAGCCCTCTATGTA</th>\n",
       "      <th>ACGTTGCGTATTTAGTCG</th>\n",
       "      <th>AACCTATAGAGACCGACG</th>\n",
       "      <th>ACGAGCGCTGTGGCCTAG</th>\n",
       "      <th>GCGAATGGACATGGACAT</th>\n",
       "      <th>TCTACCGCTCAAGCTCAA</th>\n",
       "      <th>...</th>\n",
       "      <th>CGGCAGTCAAAGATCTCT</th>\n",
       "      <th>GACACTGCGAATCTGTGT</th>\n",
       "      <th>GCAGGAGGCTGCTAAGGG</th>\n",
       "      <th>TATGTATACTTCCGCACC</th>\n",
       "      <th>TGGATGTTCCGCACAATA</th>\n",
       "      <th>AACCTATGGATGGGGTTT</th>\n",
       "      <th>AAGCGGAGGACTCTCCAT</th>\n",
       "      <th>ACCTGACTCGCAAGCGAG</th>\n",
       "      <th>ACGTTGCAAAGTTTCATA</th>\n",
       "      <th>ATCAACTGCAATTTCCGC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OG0000001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OG0000005</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OG0000009</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OG0000013</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OG0000014</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13611</th>\n",
       "      <td>OG0030137</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13612</th>\n",
       "      <td>OG0030138</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13614</th>\n",
       "      <td>OG0030140</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13615</th>\n",
       "      <td>OG0030141</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13616</th>\n",
       "      <td>OG0030142</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12997 rows × 10001 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Orthogroup  ACAATATATTGTACCTGA  ACGTTGATGGCGTAGAGA  AACCTAACCTGAATTTGC  \\\n",
       "0      OG0000001                   0                   0                   0   \n",
       "1      OG0000005                   0                   1                   0   \n",
       "2      OG0000009                   0                   0                   0   \n",
       "3      OG0000013                   0                   1                   0   \n",
       "4      OG0000014                   1                   1                   0   \n",
       "...          ...                 ...                 ...                 ...   \n",
       "13611  OG0030137                   1                   0                   0   \n",
       "13612  OG0030138                   0                   0                   0   \n",
       "13614  OG0030140                   0                   0                   0   \n",
       "13615  OG0030141                   0                   0                   0   \n",
       "13616  OG0030142                   0                   0                   1   \n",
       "\n",
       "       CTCGCAGCCCTCTATGTA  ACGTTGCGTATTTAGTCG  AACCTATAGAGACCGACG  \\\n",
       "0                       0                   1                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "13611                   0                   0                   0   \n",
       "13612                   0                   0                   0   \n",
       "13614                   0                   0                   0   \n",
       "13615                   0                   0                   0   \n",
       "13616                   0                   0                   0   \n",
       "\n",
       "       ACGAGCGCTGTGGCCTAG  GCGAATGGACATGGACAT  TCTACCGCTCAAGCTCAA  ...  \\\n",
       "0                       1                   1                   0  ...   \n",
       "1                       1                   0                   0  ...   \n",
       "2                       0                   0                   0  ...   \n",
       "3                       0                   0                   2  ...   \n",
       "4                       0                   0                   0  ...   \n",
       "...                   ...                 ...                 ...  ...   \n",
       "13611                   0                   0                   0  ...   \n",
       "13612                   0                   0                   0  ...   \n",
       "13614                   0                   0                   0  ...   \n",
       "13615                   0                   0                   0  ...   \n",
       "13616                   0                   0                   0  ...   \n",
       "\n",
       "       CGGCAGTCAAAGATCTCT  GACACTGCGAATCTGTGT  GCAGGAGGCTGCTAAGGG  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   2                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "13611                   0                   0                   0   \n",
       "13612                   0                   0                   0   \n",
       "13614                   0                   0                   0   \n",
       "13615                   0                   0                   0   \n",
       "13616                   0                   0                   0   \n",
       "\n",
       "       TATGTATACTTCCGCACC  TGGATGTTCCGCACAATA  AACCTATGGATGGGGTTT  \\\n",
       "0                       0                   0                   1   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "13611                   0                   0                   0   \n",
       "13612                   0                   0                   0   \n",
       "13614                   0                   0                   0   \n",
       "13615                   0                   0                   0   \n",
       "13616                   0                   0                   0   \n",
       "\n",
       "       AAGCGGAGGACTCTCCAT  ACCTGACTCGCAAGCGAG  ACGTTGCAAAGTTTCATA  \\\n",
       "0                       0                   1                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "13611                   0                   0                   0   \n",
       "13612                   0                   0                   0   \n",
       "13614                   0                   0                   0   \n",
       "13615                   0                   0                   0   \n",
       "13616                   0                   0                   0   \n",
       "\n",
       "       ATCAACTGCAATTTCCGC  \n",
       "0                       0  \n",
       "1                       0  \n",
       "2                       0  \n",
       "3                       0  \n",
       "4                       0  \n",
       "...                   ...  \n",
       "13611                   0  \n",
       "13612                   0  \n",
       "13614                   0  \n",
       "13615                   0  \n",
       "13616                   0  \n",
       "\n",
       "[12997 rows x 10001 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exc file saved at /home/ec2-user/glial-origins/output/Drer_adultbrain/GSM3768152_Brain_8_dge_asOrthogroup.txt\n",
      "key \"Orthogroup_exc\" already exists, ignoring\n",
      "Expanding column GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder\n",
      "Extracting keys for Orthogroup\n",
      "Saving keys to Mmus_adultbrain_og_idmm.tsv\n",
      "Done saving Mmus_adultbrain_og_idmm.tsv\n",
      "key \"og_idmm\" already exists, ignoring\n",
      "key \"DrerMmusXlae_OGfile\" already exists, ignoring\n",
      "Generating exc file at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GSM2906405_Brain1_dge_coerced_asOrthogroup.txt\n",
      "Preview of exc file: /home/ec2-user/glial-origins/output/Mmus_adultbrain/GSM2906405_Brain1_dge_coerced_asOrthogroup.txt\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>Brain_1.CCGCTAAATAAATAAGGG</th>\n",
       "      <th>Brain_1.AACGCCGATCTTGCCCTC</th>\n",
       "      <th>Brain_1.ACCTGAAGTTTATCGTAA</th>\n",
       "      <th>Brain_1.CTCGCACTGAAACCGCTA</th>\n",
       "      <th>Brain_1.ATCAACATCTCTTCGGGT</th>\n",
       "      <th>Brain_1.GCGAATAGGGTCTATGTA</th>\n",
       "      <th>Brain_1.CGAGTAAGGGTCTAGTCG</th>\n",
       "      <th>Brain_1.ATCTCTTCGTAAGTTGCC</th>\n",
       "      <th>Brain_1.AACGCCTAAGGGCTCGCA</th>\n",
       "      <th>...</th>\n",
       "      <th>Brain_1.TTGGACGCCTAGGAGATC</th>\n",
       "      <th>Brain_1.TTAACTAAAGTTTATGTA</th>\n",
       "      <th>Brain_1.GTCCCGGGACATAGGACT</th>\n",
       "      <th>Brain_1.CCGCTAGGGTTTGCTCAA</th>\n",
       "      <th>Brain_1.TGATCAGCTGTGTCAAAG</th>\n",
       "      <th>Brain_1.TCACTTGAATTATGAAGC</th>\n",
       "      <th>Brain_1.AAGTACGCTGTGTATGTA</th>\n",
       "      <th>Brain_1.CCTAGATAGAGAATTTGC</th>\n",
       "      <th>Brain_1.CATCCCATTTGCGGCTGC</th>\n",
       "      <th>Brain_1.CAAAGTGGGTTTAGCGAG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OG0000001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OG0000005</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OG0000008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OG0000011</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OG0000017</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18406</th>\n",
       "      <td>OG0028363</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18408</th>\n",
       "      <td>OG0028365</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18409</th>\n",
       "      <td>OG0028366</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18413</th>\n",
       "      <td>OG0028373</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18417</th>\n",
       "      <td>OG0028379</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13426 rows × 3680 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Orthogroup  Brain_1.CCGCTAAATAAATAAGGG  Brain_1.AACGCCGATCTTGCCCTC  \\\n",
       "0      OG0000001                           0                           0   \n",
       "1      OG0000005                           0                           0   \n",
       "2      OG0000008                           0                           0   \n",
       "3      OG0000011                           3                           0   \n",
       "4      OG0000017                           0                           1   \n",
       "...          ...                         ...                         ...   \n",
       "18406  OG0028363                           1                           0   \n",
       "18408  OG0028365                           0                           0   \n",
       "18409  OG0028366                           0                           0   \n",
       "18413  OG0028373                           0                           1   \n",
       "18417  OG0028379                           0                           0   \n",
       "\n",
       "       Brain_1.ACCTGAAGTTTATCGTAA  Brain_1.CTCGCACTGAAACCGCTA  \\\n",
       "0                               0                           0   \n",
       "1                               0                           0   \n",
       "2                               0                           0   \n",
       "3                               0                           0   \n",
       "4                               2                           1   \n",
       "...                           ...                         ...   \n",
       "18406                           0                           0   \n",
       "18408                           0                           0   \n",
       "18409                           0                           0   \n",
       "18413                           0                           0   \n",
       "18417                           0                           0   \n",
       "\n",
       "       Brain_1.ATCAACATCTCTTCGGGT  Brain_1.GCGAATAGGGTCTATGTA  \\\n",
       "0                               0                           0   \n",
       "1                               0                           0   \n",
       "2                               0                           1   \n",
       "3                               0                          15   \n",
       "4                               0                           0   \n",
       "...                           ...                         ...   \n",
       "18406                           0                           0   \n",
       "18408                           0                           0   \n",
       "18409                           0                           0   \n",
       "18413                           0                           0   \n",
       "18417                           0                           0   \n",
       "\n",
       "       Brain_1.CGAGTAAGGGTCTAGTCG  Brain_1.ATCTCTTCGTAAGTTGCC  \\\n",
       "0                               0                           0   \n",
       "1                               0                           0   \n",
       "2                               0                           0   \n",
       "3                               0                           0   \n",
       "4                               0                           0   \n",
       "...                           ...                         ...   \n",
       "18406                           0                           0   \n",
       "18408                           0                           0   \n",
       "18409                           0                           0   \n",
       "18413                           0                           0   \n",
       "18417                           0                           0   \n",
       "\n",
       "       Brain_1.AACGCCTAAGGGCTCGCA  ...  Brain_1.TTGGACGCCTAGGAGATC  \\\n",
       "0                               0  ...                           0   \n",
       "1                               0  ...                           0   \n",
       "2                               0  ...                           0   \n",
       "3                               0  ...                           0   \n",
       "4                               0  ...                           0   \n",
       "...                           ...  ...                         ...   \n",
       "18406                           1  ...                           0   \n",
       "18408                           0  ...                           0   \n",
       "18409                           0  ...                           0   \n",
       "18413                           0  ...                           0   \n",
       "18417                           0  ...                           0   \n",
       "\n",
       "       Brain_1.TTAACTAAAGTTTATGTA  Brain_1.GTCCCGGGACATAGGACT  \\\n",
       "0                               0                           0   \n",
       "1                               0                           0   \n",
       "2                               0                           0   \n",
       "3                               0                           0   \n",
       "4                               0                           0   \n",
       "...                           ...                         ...   \n",
       "18406                           0                           0   \n",
       "18408                           0                           1   \n",
       "18409                           0                           0   \n",
       "18413                           0                           0   \n",
       "18417                           0                           0   \n",
       "\n",
       "       Brain_1.CCGCTAGGGTTTGCTCAA  Brain_1.TGATCAGCTGTGTCAAAG  \\\n",
       "0                               0                           0   \n",
       "1                               0                           0   \n",
       "2                               0                           0   \n",
       "3                               0                           0   \n",
       "4                               0                           0   \n",
       "...                           ...                         ...   \n",
       "18406                           0                           0   \n",
       "18408                           0                           0   \n",
       "18409                           0                           0   \n",
       "18413                           0                           0   \n",
       "18417                           0                           0   \n",
       "\n",
       "       Brain_1.TCACTTGAATTATGAAGC  Brain_1.AAGTACGCTGTGTATGTA  \\\n",
       "0                               0                           0   \n",
       "1                               0                           0   \n",
       "2                               0                           0   \n",
       "3                               0                           0   \n",
       "4                               0                           0   \n",
       "...                           ...                         ...   \n",
       "18406                           0                           0   \n",
       "18408                           0                           0   \n",
       "18409                           0                           0   \n",
       "18413                           0                           0   \n",
       "18417                           0                           0   \n",
       "\n",
       "       Brain_1.CCTAGATAGAGAATTTGC  Brain_1.CATCCCATTTGCGGCTGC  \\\n",
       "0                               0                           0   \n",
       "1                               0                           0   \n",
       "2                               0                           0   \n",
       "3                               0                           0   \n",
       "4                               2                           0   \n",
       "...                           ...                         ...   \n",
       "18406                           0                           0   \n",
       "18408                           0                           0   \n",
       "18409                           0                           0   \n",
       "18413                           0                           0   \n",
       "18417                           0                           0   \n",
       "\n",
       "       Brain_1.CAAAGTGGGTTTAGCGAG  \n",
       "0                               0  \n",
       "1                               0  \n",
       "2                               0  \n",
       "3                               0  \n",
       "4                               0  \n",
       "...                           ...  \n",
       "18406                           0  \n",
       "18408                           0  \n",
       "18409                           0  \n",
       "18413                           0  \n",
       "18417                           0  \n",
       "\n",
       "[13426 rows x 3680 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exc file saved at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GSM2906405_Brain1_dge_coerced_asOrthogroup.txt\n",
      "key \"Orthogroup_exc\" already exists, ignoring\n",
      "Expanding column XENLA_9.2_genome_cDNA.fa.transdecoder\n",
      "Extracting keys for Orthogroup\n",
      "Saving keys to Xlae_adultbrain_og_idmm.tsv\n",
      "Done saving Xlae_adultbrain_og_idmm.tsv\n",
      "key \"og_idmm\" already exists, ignoring\n",
      "key \"DrerMmusXlae_OGfile\" already exists, ignoring\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "\"['gene_name'] not in index\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [6], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mgxc_to_exc\u001b[49m\u001b[43m(\u001b[49m\u001b[43msample_MSD\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43msample_MSD\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      2\u001b[0m \u001b[43m           \u001b[49m\u001b[43membedding_df\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43morthogroups_df\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m      3\u001b[0m \u001b[43m           \u001b[49m\u001b[43mexc_file\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[43morthogroups_file\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[43m           \u001b[49m\u001b[43moverwrite\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/glial-origins/notebooks/DrerMmusXlae_adultbrain_OrthoFinder/../../utils/biofile_handling.py:1300\u001b[0m, in \u001b[0;36mgxc_to_exc\u001b[0;34m(sample_MSD, embedding_df, exc_file, overwrite)\u001b[0m\n\u001b[1;32m   1297\u001b[0m sample_MSD\u001b[38;5;241m.\u001b[39mspecies_BioFileDockets[pre]\u001b[38;5;241m.\u001b[39madd_keyfile(sample_MSD\u001b[38;5;241m.\u001b[39mspecies_concat \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m fileouttype, exc_file)\n\u001b[1;32m   1299\u001b[0m \u001b[38;5;66;03m# Extracts gene_name to embedding mapping keys\u001b[39;00m\n\u001b[0;32m-> 1300\u001b[0m gene_keys \u001b[38;5;241m=\u001b[39m \u001b[43mexc_idmm_df\u001b[49m\u001b[43m[\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mgene_name\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43membedding\u001b[49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mdrop_duplicates()\n\u001b[1;32m   1302\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGenerating exc file at\u001b[39m\u001b[38;5;124m'\u001b[39m, exc\u001b[38;5;241m.\u001b[39mpath)\n\u001b[1;32m   1303\u001b[0m \u001b[38;5;66;03m# Reads in original gxc matrix file\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/glial_origins/lib/python3.10/site-packages/pandas/core/frame.py:3511\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3509\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m is_iterator(key):\n\u001b[1;32m   3510\u001b[0m         key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(key)\n\u001b[0;32m-> 3511\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_indexer_strict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcolumns\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m   3513\u001b[0m \u001b[38;5;66;03m# take() does not accept boolean indexers\u001b[39;00m\n\u001b[1;32m   3514\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(indexer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mbool\u001b[39m:\n",
      "File \u001b[0;32m~/miniconda3/envs/glial_origins/lib/python3.10/site-packages/pandas/core/indexes/base.py:5796\u001b[0m, in \u001b[0;36mIndex._get_indexer_strict\u001b[0;34m(self, key, axis_name)\u001b[0m\n\u001b[1;32m   5793\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   5794\u001b[0m     keyarr, indexer, new_indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reindex_non_unique(keyarr)\n\u001b[0;32m-> 5796\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_raise_if_missing\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeyarr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis_name\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   5798\u001b[0m keyarr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtake(indexer)\n\u001b[1;32m   5799\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, Index):\n\u001b[1;32m   5800\u001b[0m     \u001b[38;5;66;03m# GH 42790 - Preserve name from an Index\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/glial_origins/lib/python3.10/site-packages/pandas/core/indexes/base.py:5859\u001b[0m, in \u001b[0;36mIndex._raise_if_missing\u001b[0;34m(self, key, indexer, axis_name)\u001b[0m\n\u001b[1;32m   5856\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNone of [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m] are in the [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00maxis_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   5858\u001b[0m not_found \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(ensure_index(key)[missing_mask\u001b[38;5;241m.\u001b[39mnonzero()[\u001b[38;5;241m0\u001b[39m]]\u001b[38;5;241m.\u001b[39munique())\n\u001b[0;32m-> 5859\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnot_found\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not in index\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['gene_name'] not in index\""
     ]
    }
   ],
   "source": [
    "gxc_to_exc(sample_MSD = sample_MSD,\n",
    "           embedding_df = orthogroups_df, \n",
    "           exc_file = orthogroups_file,\n",
    "           overwrite = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c135c9ee-8f04-4bf5-9916-7b3eeb0f42b0",
   "metadata": {},
   "source": [
    "# 5. Generate merged exc file for all samples in dataset\n",
    "\n",
    "(TO BE WRITTEN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c30778eb-5480-4fda-84a2-462932b09b6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "collector_df = pd.DataFrame()\n",
    "\n",
    "display(collector_df)\n",
    "export_filename = '_'.join([sample_MSD.species_concat, sample_MSD.global_conditions, 'asOrthogroup', 'combined_exc.tsv'])\n",
    "\n",
    "jointexc = JointExcFile(\n",
    "    species_dict = sample_MSD.species_dict,\n",
    "    sampledict = sample_MSD.sampledict,\n",
    "    filename = export_filename,\n",
    "    embedding = 'Orthogroup',\n",
    "    sources = [sample_MSD.species_BioFileDockets[pre].Orthogroup_exc for pre in sample_MSD.species_BioFileDockets]\n",
    ")\n",
    "\n",
    "if not os.path.exists(jointexc.path):\n",
    "\n",
    "    # Iterates through all of the species in the Species BioFileDocket\n",
    "    for i, pre in enumerate(sample_MSD.species_BioFileDockets.keys()):\n",
    "    \n",
    "        df = pd.read_csv(sample_MSD.species_BioFileDockets[pre].Orthogroup_exc.path, sep = '\\t')\n",
    "        df = df.add_prefix(pre + '_')\n",
    "        df.rename(columns = {pre + '_' + 'Orthogroup': 'Orthogroup'}, inplace = True)\n",
    "    \n",
    "        if i == 0:\n",
    "            collector_df = df\n",
    "        else:\n",
    "            collector_df = collector_df.merge(df, on = 'Orthogroup', how = 'inner')\n",
    "\n",
    "    collector_df.to_csv(jointexc.path, sep = '\\t', index = None)\n",
    "\n",
    "# Iterates through all of the species in the Species BioFileDocket\n",
    "for pre in sample_MSD.species_BioFileDockets.keys():\n",
    "    sample_MSD.species_BioFileDockets[pre].add_keyfile('Orthogroup_jointexc', jointexc)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "984a991a-f659-464f-aaf0-62637fb3bc88",
   "metadata": {},
   "source": [
    "# 6. Pickle the species_BioFileDocket files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f873483e-f652-4d15-bc4c-fe2ff9892456",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_000002035.5_GRCz10_genomic.gff already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM3768152_Brain_8_dge.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_uniprot-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_gtf-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM3768152_Brain_8_dge_asOrthogroup.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_BioFileDocket.pkl already exists in S3 bucket; overwriting.\n",
      "upload: ../../output/Drer_adultbrain/Drer_adultbrain_BioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/Drer_adultbrain_BioFileDocket.pkl\n",
      "GCF_000001635.23_GRCm38.p3_genomic.gff already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM2906405_Brain1_dge_coerced.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Mmus_adultbrain_uniprot-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Mmus_adultbrain_gtf-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.bed already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.cds already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3 already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Mmus_adultbrain_Brain_1_cellannot.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM2906405_Brain1_dge_coerced_asOrthogroup.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Mmus_adultbrain_BioFileDocket.pkl already exists in S3 bucket; overwriting.\n",
      "upload: ../../output/Mmus_adultbrain/Mmus_adultbrain_BioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/Mmus_adultbrain_BioFileDocket.pkl\n",
      "XENLA_9.2_Xenbase.gtf already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_GCA.gff already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome.fa already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM6214268_Xenopus_brain_COL65_dge.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_Xenbase_UniProtIDs.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_uniprot-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_gtf-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.bed already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.cds already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.pep already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_BioFileDocket.pkl already exists in S3 bucket; overwriting.\n",
      "upload: ../../output/Xlae_adultbrain/Xlae_adultbrain_BioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/Xlae_adultbrain_BioFileDocket.pkl\n"
     ]
    }
   ],
   "source": [
    "for pre in sample_MSD.species_BioFileDockets.keys():\n",
    "    sample_MSD.species_BioFileDockets[pre].local_to_s3()\n",
    "    sample_MSD.species_BioFileDockets[pre].pickle()\n",
    "    sample_MSD.species_BioFileDockets[pre].push_to_s3(overwrite = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1e678ce-1ec4-4a8d-9ebf-c77ed8d236b4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d38322d-c4a4-49b3-9f64-8d64594feec7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0874b4e2-e713-4c31-91fb-2c73c6fe45bb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be8b8325-e7ee-43e3-b9b5-91625e69becf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
