{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3df6c5ba-21b7-4f9a-aedb-9db92d7882cf",
   "metadata": {},
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "aac310d7-4800-4b49-aa99-48d455aa03d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import os\n",
    "import dill\n",
    "import datetime\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "import sys\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84395a54-f8ad-4342-bbd1-6e959c6470da",
   "metadata": {},
   "source": [
    "# 1. Collect BioFileDockets for each starting dataset\n",
    "Collects BioFileDockets for each of the species datasets in the `species_dict`.\n",
    "\n",
    "This `dict` expects key:value pairs in the form of `species`:`conditions` and is passed to the MultiSpeciesBioFileDocket class, which creates a folder in `output/` for the analysis.\n",
    "\n",
    "The folder name takes the following format:\n",
    "\n",
    "- `species_concat` + `_` + `global_conditions` + `analysis_type`\n",
    "\n",
    "Where `species_concat` is the string concatenation of the alphabetize list of species prefixes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e1ea15e7-b9de-4247-a0c8-c641879a58c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file GCF_000002035.5_GRCz10_genomic.gff already exists at ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.gff\n",
      "file GCF_000002035.5_GRCz10_genomic.fna already exists at ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.fna\n",
      "file GSM3768152_Brain_8_dge.txt already exists at ../../output/Drer_adultbrain/GSM3768152_Brain_8_dge.txt\n",
      "file Drer_adultbrain_uniprot-idmm.tsv already exists at ../../output/Drer_adultbrain/Drer_adultbrain_uniprot-idmm.tsv\n",
      "file Drer_adultbrain_gtf-idmm.tsv already exists at ../../output/Drer_adultbrain/Drer_adultbrain_gtf-idmm.tsv\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna already exists at ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed already exists at ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds already exists at ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 already exists at ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep already exists at ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep\n",
      "file GSM3768152_Brain_8_dge_asOrthogroup.txt already exists at ../../output/Drer_adultbrain/GSM3768152_Brain_8_dge_asOrthogroup.txt\n",
      "file Drer_adultbrain_og-idmm.tsv already exists at ../../output/Drer_adultbrain/Drer_adultbrain_og-idmm.tsv\n",
      "file DrerMmusXlae_adultbrain_Orthogroups.tsv already exists at ../../output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28/Orthogroups/DrerMmusXlae_adultbrain_Orthogroups.tsv\n",
      "file Drer_Exc_top_200DE_Orthogroup_ids.txt already exists at ../../output/Drer_adultbrain/Drer_Exc_top_200DE_Orthogroup_ids.txt\n",
      "file Drer_adultbrain_Brain_8_cellannot.tsv already exists at ../../output/Drer_adultbrain/Drer_adultbrain_Brain_8_cellannot.tsv\n",
      "file DrerMmusXlae_adultbrain_Orthogroups.tsv already exists at ../../output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28/Orthogroups/DrerMmusXlae_adultbrain_Orthogroups.tsv\n",
      "file DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists at ../../output/DrerMmusXlae_adultbrain_OrthoFinder/DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv\n",
      "file Drer_uniprot-taxid_7955_genes.tsv already exists at ../../output/Drer_adultbrain/Drer_uniprot-taxid_7955_genes.tsv\n",
      "file Drer_adultbrain_sc-idmm.tsv already exists at ../../output/Drer_adultbrain/Drer_adultbrain_sc-idmm.tsv\n",
      "file DrerMmusXlae_adultbrain_struclusters_file.tsv already exists at ../../output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_struclusters_file.tsv\n",
      "file GSM3768152_Brain_8_dge_asStrucluster.txt already exists at ../../output/Drer_adultbrain/GSM3768152_Brain_8_dge_asStrucluster.txt\n",
      "file DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv already exists at ../../output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv\n",
      "file Drer_Exc_top_250DE_StruCluster_ids.txt already exists at ../../output/Drer_adultbrain/Drer_Exc_top_250DE_StruCluster_ids.txt\n",
      "file GCF_000001635.23_GRCm38.p3_genomic.gff already exists at ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic.gff\n",
      "file GCF_000001635.23_GRCm38.p3_genomic.fna already exists at ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic.fna\n",
      "file GSM2906405_Brain1_dge_coerced.txt already exists at ../../output/Mmus_adultbrain/GSM2906405_Brain1_dge_coerced.txt\n",
      "file Mmus_adultbrain_uniprot-idmm.tsv already exists at ../../output/Mmus_adultbrain/Mmus_adultbrain_uniprot-idmm.tsv\n",
      "file Mmus_adultbrain_gtf-idmm.tsv already exists at ../../output/Mmus_adultbrain/Mmus_adultbrain_gtf-idmm.tsv\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna already exists at ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.bed already exists at ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.bed\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.cds already exists at ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.cds\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3 already exists at ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep already exists at ../../output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep\n",
      "file GSM2906405_Brain1_dge_coerced_asOrthogroup.txt already exists at ../../output/Mmus_adultbrain/GSM2906405_Brain1_dge_coerced_asOrthogroup.txt\n",
      "file Mmus_adultbrain_og-idmm.tsv already exists at ../../output/Mmus_adultbrain/Mmus_adultbrain_og-idmm.tsv\n",
      "file DrerMmusXlae_adultbrain_Orthogroups.tsv already exists at ../../output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28/Orthogroups/DrerMmusXlae_adultbrain_Orthogroups.tsv\n",
      "file Mmus_Exc_top_200DE_Orthogroup_ids.txt already exists at ../../output/Mmus_adultbrain/Mmus_Exc_top_200DE_Orthogroup_ids.txt\n",
      "file Mmus_adultbrain_Brain_1_cellannot.tsv already exists at ../../output/Mmus_adultbrain/Mmus_adultbrain_Brain_1_cellannot.tsv\n",
      "file DrerMmusXlae_adultbrain_Orthogroups.tsv already exists at ../../output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28/Orthogroups/DrerMmusXlae_adultbrain_Orthogroups.tsv\n",
      "file DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists at ../../output/DrerMmusXlae_adultbrain_OrthoFinder/DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv\n",
      "file Mmus_uniprot-taxid_10090_genes.tsv already exists at ../../output/Mmus_adultbrain/Mmus_uniprot-taxid_10090_genes.tsv\n",
      "file Mmus_adultbrain_sc-idmm.tsv already exists at ../../output/Mmus_adultbrain/Mmus_adultbrain_sc-idmm.tsv\n",
      "file DrerMmusXlae_adultbrain_struclusters_file.tsv already exists at ../../output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_struclusters_file.tsv\n",
      "file GSM2906405_Brain1_dge_coerced_asStrucluster.txt already exists at ../../output/Mmus_adultbrain/GSM2906405_Brain1_dge_coerced_asStrucluster.txt\n",
      "file Mmus_Exc_top_250DE_StruCluster_ids.txt already exists at ../../output/Mmus_adultbrain/Mmus_Exc_top_250DE_StruCluster_ids.txt\n",
      "file DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv already exists at ../../output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv\n",
      "file XENLA_9.2_Xenbase.gtf already exists at ../../output/Xlae_adultbrain/XENLA_9.2_Xenbase.gtf\n",
      "file XENLA_9.2_GCA.gff already exists at ../../output/Xlae_adultbrain/XENLA_9.2_GCA.gff\n",
      "file XENLA_9.2_genome.fa already exists at ../../output/Xlae_adultbrain/XENLA_9.2_genome.fa\n",
      "file GSM6214268_Xenopus_brain_COL65_dge.txt already exists at ../../output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge.txt\n",
      "file Xlae_adultbrain_uniprot-idmm.tsv already exists at ../../output/Xlae_adultbrain/Xlae_adultbrain_uniprot-idmm.tsv\n",
      "file Xlae_adultbrain_gtf-idmm.tsv already exists at ../../output/Xlae_adultbrain/Xlae_adultbrain_gtf-idmm.tsv\n",
      "file XENLA_9.2_genome_cDNA.fa already exists at ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.bed already exists at ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.bed\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.cds already exists at ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.cds\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 already exists at ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.gff3\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.pep already exists at ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.pep\n",
      "file DrerMmusXlae_adultbrain_Orthogroups.tsv already exists at ../../output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28/Orthogroups/DrerMmusXlae_adultbrain_Orthogroups.tsv\n",
      "file GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup.txt already exists at ../../output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup.txt\n",
      "file Xlae_adultbrain_og-idmm.tsv already exists at ../../output/Xlae_adultbrain/Xlae_adultbrain_og-idmm.tsv\n",
      "file DrerMmusXlae_adultbrain_Orthogroups.tsv already exists at ../../output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28/Orthogroups/DrerMmusXlae_adultbrain_Orthogroups.tsv\n",
      "file Xlae_adultbrain_Brain3_4_cellannot.tsv already exists at ../../output/Xlae_adultbrain/Xlae_adultbrain_Brain3_4_cellannot.tsv\n",
      "file DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists at ../../output/DrerMmusXlae_adultbrain_OrthoFinder/DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv\n",
      "file Xlae_uniprot-taxid_8355_genes.tsv already exists at ../../output/Xlae_adultbrain/Xlae_uniprot-taxid_8355_genes.tsv\n",
      "file Xlae_adultbrain_sc-idmm.tsv already exists at ../../output/Xlae_adultbrain/Xlae_adultbrain_sc-idmm.tsv\n",
      "file DrerMmusXlae_adultbrain_struclusters_file.tsv already exists at ../../output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_struclusters_file.tsv\n",
      "file GSM6214268_Xenopus_brain_COL65_dge_asStrucluster.txt already exists at ../../output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge_asStrucluster.txt\n",
      "file Xlae_Exc_top_250DE_StruCluster_ids.txt already exists at ../../output/Xlae_adultbrain/Xlae_Exc_top_250DE_StruCluster_ids.txt\n",
      "file Xlae_Exc_top_200DE_Orthogroup_ids.txt already exists at ../../output/Xlae_adultbrain/Xlae_Exc_top_200DE_Orthogroup_ids.txt\n",
      "file DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv already exists at ../../output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv\n"
     ]
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "species_dict = {\n",
    "    'Danio_rerio': 'adultbrain',\n",
    "    'Mus_musculus': 'adultbrain',\n",
    "    'Xenopus_laevis': 'adultbrain'\n",
    "}\n",
    "\n",
    "global_conditions = 'adultbrain'\n",
    "\n",
    "analysis_type = 'OrthoFinder'\n",
    "\n",
    "################\n",
    "\n",
    "sample_MSD = MultiSpeciesBioFileDocket(\n",
    "    species_dict = species_dict,\n",
    "    global_conditions = global_conditions,\n",
    "    analysis_type = analysis_type\n",
    "    )\n",
    "\n",
    "sample_MSD.make_directory()\n",
    "sample_MSD.get_BioFileDockets()\n",
    "sample_MSD.s3_to_local()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e3bf493-da0f-42bc-904d-fc4c6dd6ca2b",
   "metadata": {},
   "source": [
    "# 2. Get the peptide file paths from each BioFileDocket"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8bb0f3ca-f82a-48a8-93e1-7c0b9f60178b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Drer': <biofile_handling.TransdecoderOutFile at 0x7f444bb7ae30>,\n",
       " 'Mmus': <biofile_handling.TransdecoderOutFile at 0x7f444b67cdc0>,\n",
       " 'Xlae': <biofile_handling.TransdecoderOutFile at 0x7f444b3cad10>}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep has already been moved\n",
      "file /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep has already been moved\n",
      "file /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.pep has already been moved\n"
     ]
    }
   ],
   "source": [
    "# Extract the peptide file info from the sample_MSD\n",
    "peptide_files = {pre: sample_MSD.species_BioFileDockets[pre].transdecoder_pep for pre in sample_MSD.species_BioFileDockets}\n",
    "display(peptide_files)\n",
    "\n",
    "# Move the files from the starting directory to the OrthoFinder run directory\n",
    "for pepfile in peptide_files.values():\n",
    "    start = os.path.abspath(pepfile.path)\n",
    "    end = sample_MSD.directory + pepfile.filename\n",
    "    if not os.path.exists(end):\n",
    "        print('copying', start, 'to', end)\n",
    "        subprocess.run(['cp', start, end])\n",
    "    else:\n",
    "        print('file', start, 'has already been moved')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4963be4b-109a-416b-bc59-4459f8d3dd3b",
   "metadata": {},
   "source": [
    "# 3. Run OrthoFinder on all collected peptide files\n",
    "\n",
    "Before running, will check if analysis has already been done - if so, continues."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "03e128ac-c038-49ed-bf45-2c19b59630ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../../output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Dec01\n",
      "OF folder already exists at ../../output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/\n",
      "Checking for results folder\n",
      "Results folder found at ../../output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder</th>\n",
       "      <th>GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder</th>\n",
       "      <th>XENLA_9.2_genome_cDNA.fa.transdecoder</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OG0000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene10395_t.p1, gene12148_t.p1, gene1307_t.p1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OG0000001</td>\n",
       "      <td>rna75711.p2</td>\n",
       "      <td>gene10036.p1, gene10201.p2, gene10535.p1, gene...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OG0000002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene2928_t.p3, gene38029_t.p2, gene41218_t.p1,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OG0000003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene264_t.p1, rna1157.p2, rna12598.p1, rna1516...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OG0000004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene50425_t.p1, rna27145.p2, rna34455.p1, rna4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32348</th>\n",
       "      <td>OG0032348</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rna97220.p1, rna97221.p1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32349</th>\n",
       "      <td>OG0032349</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rna97260.p1, rna97260.p2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32350</th>\n",
       "      <td>OG0032350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rna97447.p1, rna97448.p1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32351</th>\n",
       "      <td>OG0032351</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rna97780.p1, rna97781.p2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32352</th>\n",
       "      <td>OG0032352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rna9904.p1, rna9906.p1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32353 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Orthogroup GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder  \\\n",
       "0      OG0000000                                                NaN         \n",
       "1      OG0000001                                        rna75711.p2         \n",
       "2      OG0000002                                                NaN         \n",
       "3      OG0000003                                                NaN         \n",
       "4      OG0000004                                                NaN         \n",
       "...          ...                                                ...         \n",
       "32348  OG0032348                                                NaN         \n",
       "32349  OG0032349                                                NaN         \n",
       "32350  OG0032350                                                NaN         \n",
       "32351  OG0032351                                                NaN         \n",
       "32352  OG0032352                                                NaN         \n",
       "\n",
       "      GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder  \\\n",
       "0                                                    NaN     \n",
       "1      gene10036.p1, gene10201.p2, gene10535.p1, gene...     \n",
       "2                                                    NaN     \n",
       "3                                                    NaN     \n",
       "4                                                    NaN     \n",
       "...                                                  ...     \n",
       "32348                                                NaN     \n",
       "32349                                                NaN     \n",
       "32350                                                NaN     \n",
       "32351                                                NaN     \n",
       "32352                                                NaN     \n",
       "\n",
       "                   XENLA_9.2_genome_cDNA.fa.transdecoder  \n",
       "0      gene10395_t.p1, gene12148_t.p1, gene1307_t.p1,...  \n",
       "1                                                    NaN  \n",
       "2      gene2928_t.p3, gene38029_t.p2, gene41218_t.p1,...  \n",
       "3      gene264_t.p1, rna1157.p2, rna12598.p1, rna1516...  \n",
       "4      gene50425_t.p1, rna27145.p2, rna34455.p1, rna4...  \n",
       "...                                                  ...  \n",
       "32348                           rna97220.p1, rna97221.p1  \n",
       "32349                           rna97260.p1, rna97260.p2  \n",
       "32350                           rna97447.p1, rna97448.p1  \n",
       "32351                           rna97780.p1, rna97781.p2  \n",
       "32352                             rna9904.p1, rna9906.p1  \n",
       "\n",
       "[32353 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Generate OrthoFinder run folder information based on today's date\n",
    "OF_folder = sample_MSD.directory + 'OrthoFinder/'\n",
    "today = datetime.date.today()\n",
    "datefolder = OF_folder + 'Results_' + today.strftime('%b%d')\n",
    "print(datefolder)\n",
    "\n",
    "# Check of an OrthoFinder folder already exists; if not, runs analysis\n",
    "if not os.path.exists(OF_folder):\n",
    "    subprocess.run([ORTHOFINDER_LOC, '-f', sample_MSD.directory])\n",
    "else:\n",
    "    print('OF folder already exists at', OF_folder)\n",
    "    print('Checking for results folder')\n",
    "    folders = os.listdir(OF_folder)\n",
    "    datefolder = OF_folder + [i for i in folders if 'Results' in i][0]\n",
    "    print('Results folder found at', datefolder)\n",
    "\n",
    "# Generates OrthoGroups file object\n",
    "orthogroups_file = OrthoFinderOutputFile(\n",
    "    filename = '/Orthogroups/Orthogroups.tsv', \n",
    "    multispeciesbiofiledocket = sample_MSD,\n",
    "    directory = datefolder)\n",
    "\n",
    "orthogroups_df = pd.read_csv(orthogroups_file.path, sep = '\\t')\n",
    "display(orthogroups_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61be3acd-f098-44bb-bbf4-9c9bea6a1150",
   "metadata": {},
   "source": [
    "# 4. Generate Orthofinder_exc file from gxc file of each starting dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dbac4c37-1aff-4fdc-9992-a2203b9ef111",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using Orthogroup embeddings as expected from OrthoFinder\n",
      "Orthogroup_excfile already exists at ../../output/Drer_adultbrain/GSM3768152_Brain_8_dge_asOrthogroup.txt skipping\n",
      "Orthogroup_excfile already exists at ../../output/Mmus_adultbrain/GSM2906405_Brain1_dge_coerced_asOrthogroup.txt skipping\n",
      "Orthogroup_excfile already exists at ../../output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup.txt skipping\n"
     ]
    }
   ],
   "source": [
    "gxc_to_exc(sample_MSD = sample_MSD,\n",
    "           embedding_df = orthogroups_df, \n",
    "           exc_file = orthogroups_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c135c9ee-8f04-4bf5-9916-7b3eeb0f42b0",
   "metadata": {},
   "source": [
    "# 5. Generate merged exc file for all samples in dataset\n",
    "\n",
    "(TO BE WRITTEN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c30778eb-5480-4fda-84a2-462932b09b6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "collector_df = pd.DataFrame()\n",
    "\n",
    "display(collector_df)\n",
    "export_filename = '_'.join([sample_MSD.species_concat, sample_MSD.global_conditions, 'asOrthogroup', 'combined_exc.tsv'])\n",
    "\n",
    "jointexc = JointExcFile(\n",
    "    multispeciesbiofiledocket = sample_MSD,\n",
    "    directory = sample_MSD.directory,\n",
    "    filename = export_filename,\n",
    "    embedding = 'Orthogroup',\n",
    "    sources = [sample_MSD.species_BioFileDockets[pre].Orthogroup_exc for pre in sample_MSD.species_BioFileDockets]\n",
    ")\n",
    "\n",
    "if not os.path.exists(jointexc.path):\n",
    "\n",
    "    # Iterates through all of the species in the Species BioFileDocket\n",
    "    for i, pre in enumerate(sample_MSD.species_BioFileDockets.keys()):\n",
    "    \n",
    "        df = pd.read_csv(sample_MSD.species_BioFileDockets[pre].Orthogroup_exc.path, sep = '\\t')\n",
    "        df = df.add_prefix(pre + '_')\n",
    "        df.rename(columns = {pre + '_' + 'Orthogroup': 'Orthogroup'}, inplace = True)\n",
    "    \n",
    "        if i == 0:\n",
    "            collector_df = df\n",
    "        else:\n",
    "            collector_df = collector_df.merge(df, on = 'Orthogroup', how = 'inner')\n",
    "\n",
    "    collector_df.to_csv(jointexc.path, sep = '\\t', index = None)\n",
    "\n",
    "# Iterates through all of the species in the Species BioFileDocket\n",
    "for pre in sample_MSD.species_BioFileDockets.keys():\n",
    "    sample_MSD.species_BioFileDockets[pre].add_keyfile(jointexc, 'Orthogroup_jointexc')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "984a991a-f659-464f-aaf0-62637fb3bc88",
   "metadata": {},
   "source": [
    "# 6. Pickle the species_BioFileDocket files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f873483e-f652-4d15-bc4c-fe2ff9892456",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_000002035.5_GRCz10_genomic.gff already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM3768152_Brain_8_dge.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_uniprot-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_gtf-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM3768152_Brain_8_dge_asOrthogroup.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_og-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_Orthogroups.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_Exc_top_200DE_Orthogroup_ids.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_Brain_8_cellannot.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_Orthogroups.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_uniprot-taxid_7955_genes.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_sc-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "upload: ../../output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_struclusters_file.tsv to s3://arcadia-reference-datasets/FoldSeek_output/DrerMmusXlae_adultbrain_struclusters_file.tsv\n",
      "GSM3768152_Brain_8_dge_asStrucluster.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_sample_BioFileDocket.pkl already exists in S3 bucket; overwriting.\n",
      "upload: ../../output/Drer_adultbrain/Drer_adultbrain_sample_BioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/Drer_adultbrain_sample_BioFileDocket.pkl\n",
      "GCF_000001635.23_GRCm38.p3_genomic.gff already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM2906405_Brain1_dge_coerced.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Mmus_adultbrain_uniprot-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Mmus_adultbrain_gtf-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.bed already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.cds already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3 already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM2906405_Brain1_dge_coerced_asOrthogroup.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Mmus_adultbrain_og-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_Orthogroups.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Mmus_Exc_top_200DE_Orthogroup_ids.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Mmus_adultbrain_Brain_1_cellannot.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_Orthogroups.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Mmus_uniprot-taxid_10090_genes.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Mmus_adultbrain_sc-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_struclusters_file.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM2906405_Brain1_dge_coerced_asStrucluster.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Mmus_adultbrain_sample_BioFileDocket.pkl already exists in S3 bucket; overwriting.\n",
      "upload: ../../output/Mmus_adultbrain/Mmus_adultbrain_sample_BioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/Mmus_adultbrain_sample_BioFileDocket.pkl\n",
      "XENLA_9.2_Xenbase.gtf already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_GCA.gff already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome.fa already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM6214268_Xenopus_brain_COL65_dge.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_uniprot-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_gtf-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.bed already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.cds already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.pep already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_Orthogroups.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_og-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_Orthogroups.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_Brain3_4_cellannot.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_uniprot-taxid_8355_genes.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_sc-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_struclusters_file.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM6214268_Xenopus_brain_COL65_dge_asStrucluster.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_sample_BioFileDocket.pkl already exists in S3 bucket; overwriting.\n",
      "upload: ../../output/Xlae_adultbrain/Xlae_adultbrain_sample_BioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/Xlae_adultbrain_sample_BioFileDocket.pkl\n"
     ]
    }
   ],
   "source": [
    "for pre in sample_MSD.species_BioFileDockets.keys():\n",
    "    sample_MSD.species_BioFileDockets[pre].local_to_s3()\n",
    "    sample_MSD.species_BioFileDockets[pre].pickle()\n",
    "    sample_MSD.species_BioFileDockets[pre].push_to_s3(overwrite = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1e678ce-1ec4-4a8d-9ebf-c77ed8d236b4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d38322d-c4a4-49b3-9f64-8d64594feec7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0874b4e2-e713-4c31-91fb-2c73c6fe45bb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be8b8325-e7ee-43e3-b9b5-91625e69becf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
