{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "L56fjA27DmhA"
   },
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "t0BTCsz7Y48y",
    "outputId": "f6510678-f977-4ec8-ce30-1f2acd4fe666"
   },
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import os\n",
    "import dill\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "import sys\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Download and describe data\n",
    "\n",
    "This notebook collects data from the [Liao et al. 2022](https://www.sciencedirect.com/science/article/pii/S0092867418301168#sec4) Xenopus laevis adult cell atlas.\n",
    "\n",
    "## Dataset description\n",
    "\n",
    "- Data can be found at the [GEO Accession GSE195790](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE195790).\n",
    "- This notebook previously collects data from one sample, [\"Xenopus_brain_COL65\"](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM6214268), for the purposes of data analysis and exploration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  730M  100  730M    0     0  19.3M      0  0:00:37  0:00:37 --:--:-- 16.0M\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  109M  100  109M    0     0  6503k      0  0:00:17  0:00:17 --:--:-- 5801k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  150M  100  150M    0     0  26.4M      0  0:00:05  0:00:05 --:--:-- 30.3M\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 18.7M  100 18.7M    0     0  8600k      0  0:00:02  0:00:02 --:--:-- 8605k\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'species': 'Xenopus_laevis',\n",
       " 'conditions': 'adultbrain',\n",
       " 'directory': '../../output/Xlae_adultbrain/',\n",
       " 'files': {},\n",
       " 'annot': <biofile_handling.GenomeGtfFile at 0x7f981479b8b0>,\n",
       " 'annot2': <biofile_handling.GenomeGffFile at 0x7f981479b730>,\n",
       " 'genome_fasta': <biofile_handling.GenomeFastaFile at 0x7f976386f2e0>,\n",
       " 'gxc': <biofile_handling.GxcFile at 0x7f981479b940>}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify the name of the species folder in Amazon S3\n",
    "species = 'Xenopus_laevis'\n",
    "\n",
    "# Specify any particular identifying conditions, eg tissue type:\n",
    "conditions = 'adultbrain'\n",
    "\n",
    "# Specify url and other variables\n",
    "genome_fasta_url = 'https://ftp.xenbase.org/pub/Genomics/JGI/Xenla9.2/XENLA_9.2_genome.fa.gz'\n",
    "genome_version = 'XENLA_9.2'\n",
    "\n",
    "annot_url = 'https://ftp.xenbase.org/pub/Genomics/JGI/Xenla9.2/XENLA_9.2_Xenbase.gtf'\n",
    "\n",
    "### EXCEPTION ###\n",
    "# Need to get a second annotation in order to pull usable ID mapping information\n",
    "annot2_url = 'https://ftp.xenbase.org/pub/Genomics/JGI/Xenla9.2/XENLA_9.2_GCA.gff3'\n",
    "#################\n",
    "\n",
    "gxc_url = 'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM6214nnn/GSM6214268/suppl/GSM6214268_Xenopus_brain_COL65_dge.txt.gz'\n",
    "\n",
    "###########\n",
    "# runtime #\n",
    "###########\n",
    "\n",
    "protocol = 'curl'\n",
    "\n",
    "species_prefix = prefixify(species)\n",
    "\n",
    "# Specify folder as destination for file downloads\n",
    "output_folder = '../../output/' + prefixify(species) + '_' + conditions + '/'\n",
    "\n",
    "if not os.path.exists(output_folder):\n",
    "    os.mkdir(output_folder)\n",
    "    \n",
    "species_SampleDict = SampleDict(species, conditions, output_folder)\n",
    "\n",
    "genome_fasta = GenomeFastaFile(\n",
    "    filename = '',\n",
    "    sampledict = species_SampleDict,\n",
    "    version = genome_version,\n",
    "    url = genome_fasta_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "annot = GenomeGtfFile(\n",
    "    filename = '',\n",
    "    sampledict = species_SampleDict,\n",
    "    GenomeFastaFile = genome_fasta,\n",
    "    url = annot_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "annot2 = GenomeGffFile(\n",
    "    filename = '',\n",
    "    sampledict = species_SampleDict,\n",
    "    GenomeFastaFile = genome_fasta,\n",
    "    url = annot2_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "gxc = GxcFile(\n",
    "    filename = '',\n",
    "    sampledict = species_SampleDict,\n",
    "    GenomeFastaFile = genome_fasta,\n",
    "    GenomeAnnotFile = annot,\n",
    "    url = gxc_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "\n",
    "sample_Docket = Docket(species_SampleDict)\n",
    "keyfiles = {\n",
    "    'annot': annot,\n",
    "    'annot2': annot2,\n",
    "    'genome_fasta': genome_fasta,\n",
    "    'gxc': gxc\n",
    "}\n",
    "sample_Docket.add_keyfiles(keyfiles)\n",
    "\n",
    "display(vars(sample_Docket))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Load in the gxc matrix and get gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>AACCTATTCATATAAGGG</th>\n",
       "      <th>CTCGCATCAAAGTTAACT</th>\n",
       "      <th>AACCTAGTATACTTCCGC</th>\n",
       "      <th>AACCTAAAAGTTCTGAAA</th>\n",
       "      <th>CTCGCACGCACCCTCCAT</th>\n",
       "      <th>ACGTTGTATTGTAGCGAG</th>\n",
       "      <th>ACGAGCATGCTTTAGTCG</th>\n",
       "      <th>AACCTAGTCCCGCCATCT</th>\n",
       "      <th>AACCTAGCGAATTAGAGA</th>\n",
       "      <th>...</th>\n",
       "      <th>TCACTTGTTGCCATGCTT</th>\n",
       "      <th>TCGGGTTGTCACACTTAT</th>\n",
       "      <th>TCGTAATCGTAAGTTGCC</th>\n",
       "      <th>TGTCACGAATTACACAAG</th>\n",
       "      <th>TGTGCGTACTTCTAGTCG</th>\n",
       "      <th>TTAACTATACAGTGGATG</th>\n",
       "      <th>TTGGACACTTATGATCTT</th>\n",
       "      <th>AAAGTTACTTATGCCCTC</th>\n",
       "      <th>AACCTACGCACCTGCGGA</th>\n",
       "      <th>AACCTATAGTCGCTGTGT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.S</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>42Sp43.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42Sp50.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AK6.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AK6.S</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25907</th>\n",
       "      <td>zyg11b.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25908</th>\n",
       "      <td>zyg11b.S</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25909</th>\n",
       "      <td>zzef1.S</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25910</th>\n",
       "      <td>zzz3.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25911</th>\n",
       "      <td>zzz3.S</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25912 rows × 30001 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GENE  AACCTATTCATATAAGGG  CTCGCATCAAAGTTAACT  AACCTAGTATACTTCCGC  \\\n",
       "0           3.S                   0                   0                   0   \n",
       "1      42Sp43.L                   0                   0                   0   \n",
       "2      42Sp50.L                   0                   0                   0   \n",
       "3         AK6.L                   0                   0                   0   \n",
       "4         AK6.S                   1                   0                   0   \n",
       "...         ...                 ...                 ...                 ...   \n",
       "25907  zyg11b.L                   0                   0                   0   \n",
       "25908  zyg11b.S                   0                   0                   0   \n",
       "25909   zzef1.S                   0                   0                   0   \n",
       "25910    zzz3.L                   0                   0                   0   \n",
       "25911    zzz3.S                   0                   0                   0   \n",
       "\n",
       "       AACCTAAAAGTTCTGAAA  CTCGCACGCACCCTCCAT  ACGTTGTATTGTAGCGAG  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "25907                   1                   0                   0   \n",
       "25908                   0                   0                   0   \n",
       "25909                   0                   1                   0   \n",
       "25910                   0                   0                   0   \n",
       "25911                   0                   0                   0   \n",
       "\n",
       "       ACGAGCATGCTTTAGTCG  AACCTAGTCCCGCCATCT  AACCTAGCGAATTAGAGA  ...  \\\n",
       "0                       0                   0                   0  ...   \n",
       "1                       0                   0                   0  ...   \n",
       "2                       0                   0                   0  ...   \n",
       "3                       1                   0                   0  ...   \n",
       "4                       0                   0                   0  ...   \n",
       "...                   ...                 ...                 ...  ...   \n",
       "25907                   0                   0                   0  ...   \n",
       "25908                   0                   0                   0  ...   \n",
       "25909                   0                   0                   0  ...   \n",
       "25910                   2                   0                   0  ...   \n",
       "25911                   0                   0                   0  ...   \n",
       "\n",
       "       TCACTTGTTGCCATGCTT  TCGGGTTGTCACACTTAT  TCGTAATCGTAAGTTGCC  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "25907                   0                   0                   0   \n",
       "25908                   0                   0                   0   \n",
       "25909                   0                   0                   0   \n",
       "25910                   0                   0                   0   \n",
       "25911                   0                   0                   0   \n",
       "\n",
       "       TGTCACGAATTACACAAG  TGTGCGTACTTCTAGTCG  TTAACTATACAGTGGATG  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "25907                   0                   0                   0   \n",
       "25908                   0                   0                   0   \n",
       "25909                   0                   0                   0   \n",
       "25910                   0                   0                   0   \n",
       "25911                   0                   0                   0   \n",
       "\n",
       "       TTGGACACTTATGATCTT  AAAGTTACTTATGCCCTC  AACCTACGCACCTGCGGA  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "25907                   0                   0                   0   \n",
       "25908                   0                   0                   0   \n",
       "25909                   0                   0                   0   \n",
       "25910                   0                   0                   0   \n",
       "25911                   0                   0                   0   \n",
       "\n",
       "       AACCTATAGTCGCTGTGT  \n",
       "0                       0  \n",
       "1                       0  \n",
       "2                       0  \n",
       "3                       0  \n",
       "4                       0  \n",
       "...                   ...  \n",
       "25907                   0  \n",
       "25908                   0  \n",
       "25909                   0  \n",
       "25910                   0  \n",
       "25911                   0  \n",
       "\n",
       "[25912 rows x 30001 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>42Sp43.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42Sp50.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AK6.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AK6.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25907</th>\n",
       "      <td>zyg11b.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25908</th>\n",
       "      <td>zyg11b.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25909</th>\n",
       "      <td>zzef1.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25910</th>\n",
       "      <td>zzz3.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25911</th>\n",
       "      <td>zzz3.S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25912 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name\n",
       "0           3.S\n",
       "1      42Sp43.L\n",
       "2      42Sp50.L\n",
       "3         AK6.L\n",
       "4         AK6.S\n",
       "...         ...\n",
       "25907  zyg11b.L\n",
       "25908  zyg11b.S\n",
       "25909   zzef1.S\n",
       "25910    zzz3.L\n",
       "25911    zzz3.S\n",
       "\n",
       "[25912 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes_matrix = pd.read_csv(sample_Docket.gxc.path, sep = '\\t')\n",
    "display(genes_matrix)\n",
    "\n",
    "gxc_genes_list = pd.DataFrame({'gene_name':genes_matrix['GENE']})\n",
    "display(gxc_genes_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jvfcFNT2FuaZ"
   },
   "source": [
    "# 3. Get mapping identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2136</td>\n",
       "      <td>2204</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2205</td>\n",
       "      <td>3023</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3024</td>\n",
       "      <td>3092</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3093</td>\n",
       "      <td>4723</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>4724</td>\n",
       "      <td>4798</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771995</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104527008</td>\n",
       "      <td>104527068</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771996</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771997</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771998</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771999</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>772000 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0        1     2          3          4  5  6  7  \\\n",
       "0             MT  Xenbase  exon       2136       2204  .  +  .   \n",
       "1             MT  Xenbase  exon       2205       3023  .  +  .   \n",
       "2             MT  Xenbase  exon       3024       3092  .  +  .   \n",
       "3             MT  Xenbase  exon       3093       4723  .  +  .   \n",
       "4             MT  Xenbase  exon       4724       4798  .  +  .   \n",
       "...          ...      ...   ...        ...        ... .. .. ..   \n",
       "771995  chr9_10S  Xenbase   CDS  104527008  104527068  .  -  0   \n",
       "771996  chr9_10S  Xenbase  exon  104530653  104530753  .  -  2   \n",
       "771997  chr9_10S  Xenbase   CDS  104530653  104530753  .  -  2   \n",
       "771998  chr9_10S  Xenbase  exon  104531027  104531054  .  -  0   \n",
       "771999  chr9_10S  Xenbase   CDS  104531027  104531054  .  -  0   \n",
       "\n",
       "                                                        8  \n",
       "0       gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...  \n",
       "1       gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...  \n",
       "2       gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...  \n",
       "3       gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...  \n",
       "4       gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...  \n",
       "...                                                   ...  \n",
       "771995  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771996  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771997  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771998  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771999  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "\n",
       "[772000 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"; transcript_id \"gene42065_t\"; transcript_name \"mt-trna-phe.L\";'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load in the original GFF-based annotation\n",
    "models = pd.read_csv(sample_Docket.annot.path, skiprows = 0, header = None, sep = '\\t', on_bad_lines = 'skip', comment='#')\n",
    "display(models)\n",
    "\n",
    "attributes_column = 8\n",
    "\n",
    "# Check the structure of fields in the GFF additional fields section\n",
    "display(models[attributes_column][0])\n",
    "\n",
    "# Remove any rows with NaNs\n",
    "models.dropna(inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2136</td>\n",
       "      <td>2204</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...</td>\n",
       "      <td>{'gene_id': 'gene42065', 'gene_name': 'mt-trna...</td>\n",
       "      <td>mt-trna-phe.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2205</td>\n",
       "      <td>3023</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...</td>\n",
       "      <td>{'gene_id': 'gene34778', 'gene_name': 'mt-rnr1...</td>\n",
       "      <td>mt-rnr1.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3024</td>\n",
       "      <td>3092</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...</td>\n",
       "      <td>{'gene_id': 'gene48202', 'gene_name': 'mt-trna...</td>\n",
       "      <td>mt-trna-val.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3093</td>\n",
       "      <td>4723</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...</td>\n",
       "      <td>{'gene_id': 'gene44770', 'gene_name': 'mt-rnr2...</td>\n",
       "      <td>mt-rnr2.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>4724</td>\n",
       "      <td>4798</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...</td>\n",
       "      <td>{'gene_id': 'gene43253', 'gene_name': 'mt-trna...</td>\n",
       "      <td>mt-trna-leu1.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771995</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104527008</td>\n",
       "      <td>104527068</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771996</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771997</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771998</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771999</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>772000 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0        1     2          3          4  5  6  7  \\\n",
       "0             MT  Xenbase  exon       2136       2204  .  +  .   \n",
       "1             MT  Xenbase  exon       2205       3023  .  +  .   \n",
       "2             MT  Xenbase  exon       3024       3092  .  +  .   \n",
       "3             MT  Xenbase  exon       3093       4723  .  +  .   \n",
       "4             MT  Xenbase  exon       4724       4798  .  +  .   \n",
       "...          ...      ...   ...        ...        ... .. .. ..   \n",
       "771995  chr9_10S  Xenbase   CDS  104527008  104527068  .  -  0   \n",
       "771996  chr9_10S  Xenbase  exon  104530653  104530753  .  -  2   \n",
       "771997  chr9_10S  Xenbase   CDS  104530653  104530753  .  -  2   \n",
       "771998  chr9_10S  Xenbase  exon  104531027  104531054  .  -  0   \n",
       "771999  chr9_10S  Xenbase   CDS  104531027  104531054  .  -  0   \n",
       "\n",
       "                                                        8  \\\n",
       "0       gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...   \n",
       "1       gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...   \n",
       "2       gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...   \n",
       "3       gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...   \n",
       "4       gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...   \n",
       "...                                                   ...   \n",
       "771995  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "771996  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "771997  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "771998  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "771999  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "\n",
       "                                         field_dictionary          gene_name  \n",
       "0       {'gene_id': 'gene42065', 'gene_name': 'mt-trna...      mt-trna-phe.L  \n",
       "1       {'gene_id': 'gene34778', 'gene_name': 'mt-rnr1...          mt-rnr1.L  \n",
       "2       {'gene_id': 'gene48202', 'gene_name': 'mt-trna...      mt-trna-val.L  \n",
       "3       {'gene_id': 'gene44770', 'gene_name': 'mt-rnr2...          mt-rnr2.L  \n",
       "4       {'gene_id': 'gene43253', 'gene_name': 'mt-trna...     mt-trna-leu1.L  \n",
       "...                                                   ...                ...  \n",
       "771995  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S  \n",
       "771996  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S  \n",
       "771997  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S  \n",
       "771998  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S  \n",
       "771999  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S  \n",
       "\n",
       "[772000 rows x 11 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "49282"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract field and database cross-ref (dbxref) information into columns\n",
    "models['field_dictionary'] = models[attributes_column].apply(convert_fields_to_dict_gtf)\n",
    "models['gene_name'] = [d.get('gene_name') for d in models['field_dictionary']]\n",
    "\n",
    "display(models)\n",
    "display(len(models['gene_name'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Name=Xelaev18004747m;end_range=18399,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Note=transcript XELAEV_18...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>17924</td>\n",
       "      <td>18243</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809901</th>\n",
       "      <td>Scaffold94051</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8</td>\n",
       "      <td>205</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809902</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene45941;Name=Xelaev18004691m;end_range=23...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809903</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna47731;Parent=gene45941;Note=transcript X...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809904</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id452950;Parent=rna47731;Note=transcript XE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809905</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>809906 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0        1     2      3      4  5  6  7  \\\n",
       "0               chr1L  Genbank  gene  17924  18399  .  -  .   \n",
       "1               chr1L  Genbank  mRNA  17924  18399  .  -  .   \n",
       "2               chr1L  Genbank  exon  18336  18399  .  -  .   \n",
       "3               chr1L  Genbank  exon  17924  18243  .  -  .   \n",
       "4               chr1L  Genbank   CDS  18336  18399  .  -  0   \n",
       "...               ...      ...   ...    ...    ... .. .. ..   \n",
       "809901  Scaffold94051  Genbank   CDS      8    205  .  +  0   \n",
       "809902  Scaffold95291  Genbank  gene      9    236  .  -  .   \n",
       "809903  Scaffold95291  Genbank  mRNA      9    236  .  -  .   \n",
       "809904  Scaffold95291  Genbank  exon      9    236  .  -  .   \n",
       "809905  Scaffold95291  Genbank   CDS      9    236  .  -  0   \n",
       "\n",
       "                                                        8  \n",
       "0       ID=gene0;Name=Xelaev18004747m;end_range=18399,...  \n",
       "1       ID=rna0;Parent=gene0;Note=transcript XELAEV_18...  \n",
       "2       ID=id1;Parent=rna0;Note=transcript XELAEV_1800...  \n",
       "3       ID=id2;Parent=rna0;Note=transcript XELAEV_1800...  \n",
       "4       ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...  \n",
       "...                                                   ...  \n",
       "809901  ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...  \n",
       "809902  ID=gene45941;Name=Xelaev18004691m;end_range=23...  \n",
       "809903  ID=rna47731;Parent=gene45941;Note=transcript X...  \n",
       "809904  ID=id452950;Parent=rna47731;Note=transcript XE...  \n",
       "809905  ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...  \n",
       "\n",
       "[809906 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "models2 = pd.read_csv(sample_Docket.annot2.path, skiprows = 0, header = None, sep = '\\t', on_bad_lines = 'skip', comment='#')\n",
    "display(models2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Xenbase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Name=Xelaev18004747m;end_range=18399,...</td>\n",
       "      <td>{'ID': 'gene0', 'Name': 'Xelaev18004747m', 'en...</td>\n",
       "      <td>Xelaev18004747m</td>\n",
       "      <td>XB-GENE-5942444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Note=transcript XELAEV_18...</td>\n",
       "      <td>{'ID': 'rna0', 'Parent': 'gene0', 'Note': 'tra...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "      <td>{'ID': 'id1', 'Parent': 'rna0', 'Note': 'trans...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>17924</td>\n",
       "      <td>18243</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "      <td>{'ID': 'id2', 'Parent': 'rna0', 'Note': 'trans...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...</td>\n",
       "      <td>{'ID': 'cds0', 'Parent': 'rna0', 'Dbxref': 'Ph...</td>\n",
       "      <td>OCT98948.1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809901</th>\n",
       "      <td>Scaffold94051</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8</td>\n",
       "      <td>205</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...</td>\n",
       "      <td>{'ID': 'cds47730', 'Parent': 'rna47730', 'Dbxr...</td>\n",
       "      <td>OCT55143.1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809902</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene45941;Name=Xelaev18004691m;end_range=23...</td>\n",
       "      <td>{'ID': 'gene45941', 'Name': 'Xelaev18004691m',...</td>\n",
       "      <td>Xelaev18004691m</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809903</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna47731;Parent=gene45941;Note=transcript X...</td>\n",
       "      <td>{'ID': 'rna47731', 'Parent': 'gene45941', 'Not...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809904</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id452950;Parent=rna47731;Note=transcript XE...</td>\n",
       "      <td>{'ID': 'id452950', 'Parent': 'rna47731', 'Note...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809905</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...</td>\n",
       "      <td>{'ID': 'cds47731', 'Parent': 'rna47731', 'Dbxr...</td>\n",
       "      <td>OCT55142.1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>809906 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0        1     2      3      4  5  6  7  \\\n",
       "0               chr1L  Genbank  gene  17924  18399  .  -  .   \n",
       "1               chr1L  Genbank  mRNA  17924  18399  .  -  .   \n",
       "2               chr1L  Genbank  exon  18336  18399  .  -  .   \n",
       "3               chr1L  Genbank  exon  17924  18243  .  -  .   \n",
       "4               chr1L  Genbank   CDS  18336  18399  .  -  0   \n",
       "...               ...      ...   ...    ...    ... .. .. ..   \n",
       "809901  Scaffold94051  Genbank   CDS      8    205  .  +  0   \n",
       "809902  Scaffold95291  Genbank  gene      9    236  .  -  .   \n",
       "809903  Scaffold95291  Genbank  mRNA      9    236  .  -  .   \n",
       "809904  Scaffold95291  Genbank  exon      9    236  .  -  .   \n",
       "809905  Scaffold95291  Genbank   CDS      9    236  .  -  0   \n",
       "\n",
       "                                                        8  \\\n",
       "0       ID=gene0;Name=Xelaev18004747m;end_range=18399,...   \n",
       "1       ID=rna0;Parent=gene0;Note=transcript XELAEV_18...   \n",
       "2       ID=id1;Parent=rna0;Note=transcript XELAEV_1800...   \n",
       "3       ID=id2;Parent=rna0;Note=transcript XELAEV_1800...   \n",
       "4       ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...   \n",
       "...                                                   ...   \n",
       "809901  ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...   \n",
       "809902  ID=gene45941;Name=Xelaev18004691m;end_range=23...   \n",
       "809903  ID=rna47731;Parent=gene45941;Note=transcript X...   \n",
       "809904  ID=id452950;Parent=rna47731;Note=transcript XE...   \n",
       "809905  ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...   \n",
       "\n",
       "                                         field_dictionary        gene_name  \\\n",
       "0       {'ID': 'gene0', 'Name': 'Xelaev18004747m', 'en...  Xelaev18004747m   \n",
       "1       {'ID': 'rna0', 'Parent': 'gene0', 'Note': 'tra...             None   \n",
       "2       {'ID': 'id1', 'Parent': 'rna0', 'Note': 'trans...             None   \n",
       "3       {'ID': 'id2', 'Parent': 'rna0', 'Note': 'trans...             None   \n",
       "4       {'ID': 'cds0', 'Parent': 'rna0', 'Dbxref': 'Ph...       OCT98948.1   \n",
       "...                                                   ...              ...   \n",
       "809901  {'ID': 'cds47730', 'Parent': 'rna47730', 'Dbxr...       OCT55143.1   \n",
       "809902  {'ID': 'gene45941', 'Name': 'Xelaev18004691m',...  Xelaev18004691m   \n",
       "809903  {'ID': 'rna47731', 'Parent': 'gene45941', 'Not...             None   \n",
       "809904  {'ID': 'id452950', 'Parent': 'rna47731', 'Note...             None   \n",
       "809905  {'ID': 'cds47731', 'Parent': 'rna47731', 'Dbxr...       OCT55142.1   \n",
       "\n",
       "                Xenbase  \n",
       "0       XB-GENE-5942444  \n",
       "1                  None  \n",
       "2                  None  \n",
       "3                  None  \n",
       "4                  None  \n",
       "...                 ...  \n",
       "809901             None  \n",
       "809902             None  \n",
       "809903             None  \n",
       "809904             None  \n",
       "809905             None  \n",
       "\n",
       "[809906 rows x 12 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Xenbase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>42sp43.L</td>\n",
       "      <td>XB-GENE-6252610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>42sp50.L</td>\n",
       "      <td>XB-GENE-5853356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LOC100036702.L</td>\n",
       "      <td>XB-GENE-18005853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LOC100037907.L</td>\n",
       "      <td>XB-GENE-18005786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LOC100037907.S</td>\n",
       "      <td>XB-GENE-18005873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21951</th>\n",
       "      <td>zyx.L</td>\n",
       "      <td>XB-GENE-6252950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21952</th>\n",
       "      <td>zyx.S</td>\n",
       "      <td>XB-GENE-6253946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21953</th>\n",
       "      <td>zzef1.S</td>\n",
       "      <td>XB-GENE-6488145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21954</th>\n",
       "      <td>zzz3.L</td>\n",
       "      <td>XB-GENE-6486634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21955</th>\n",
       "      <td>zzz3.S</td>\n",
       "      <td>XB-GENE-17341117</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21956 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_name           Xenbase\n",
       "0            42sp43.L   XB-GENE-6252610\n",
       "1            42sp50.L   XB-GENE-5853356\n",
       "2      LOC100036702.L  XB-GENE-18005853\n",
       "3      LOC100037907.L  XB-GENE-18005786\n",
       "4      LOC100037907.S  XB-GENE-18005873\n",
       "...               ...               ...\n",
       "21951           zyx.L   XB-GENE-6252950\n",
       "21952           zyx.S   XB-GENE-6253946\n",
       "21953         zzef1.S   XB-GENE-6488145\n",
       "21954          zzz3.L   XB-GENE-6486634\n",
       "21955          zzz3.S  XB-GENE-17341117\n",
       "\n",
       "[21956 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "models2['field_dictionary'] = models2[attributes_column].apply(convert_fields_to_dict_gff)\n",
    "models2['gene_name'] = [d.get('Name') for d in models2['field_dictionary']]\n",
    "models2['Xenbase'] = [d.get('Alias') for d in models2['field_dictionary']]\n",
    "display(models2)\n",
    "\n",
    "xenbase_keys = models2[['gene_name', 'Xenbase']].dropna().drop_duplicates()\n",
    "xenbase_keys = xenbase_keys.groupby('gene_name').agg({'gene_name': 'first', 'Xenbase': ','.join}).reset_index(drop = True)\n",
    "display(xenbase_keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19286"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>new_fields</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2136</td>\n",
       "      <td>2204</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2205</td>\n",
       "      <td>3023</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3024</td>\n",
       "      <td>3092</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3093</td>\n",
       "      <td>4723</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>4724</td>\n",
       "      <td>4798</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771995</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104527008</td>\n",
       "      <td>104527068</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771996</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771997</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771998</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771999</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>772000 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0        1     2          3          4  5  6  7  \\\n",
       "0             MT  Xenbase  exon       2136       2204  .  +  .   \n",
       "1             MT  Xenbase  exon       2205       3023  .  +  .   \n",
       "2             MT  Xenbase  exon       3024       3092  .  +  .   \n",
       "3             MT  Xenbase  exon       3093       4723  .  +  .   \n",
       "4             MT  Xenbase  exon       4724       4798  .  +  .   \n",
       "...          ...      ...   ...        ...        ... .. .. ..   \n",
       "771995  chr9_10S  Xenbase   CDS  104527008  104527068  .  -  0   \n",
       "771996  chr9_10S  Xenbase  exon  104530653  104530753  .  -  2   \n",
       "771997  chr9_10S  Xenbase   CDS  104530653  104530753  .  -  2   \n",
       "771998  chr9_10S  Xenbase  exon  104531027  104531054  .  -  0   \n",
       "771999  chr9_10S  Xenbase   CDS  104531027  104531054  .  -  0   \n",
       "\n",
       "                                               new_fields  \n",
       "0       gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...  \n",
       "1       gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...  \n",
       "2       gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...  \n",
       "3       gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...  \n",
       "4       gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...  \n",
       "...                                                   ...  \n",
       "771995  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771996  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771997  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771998  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771999  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "\n",
       "[772000 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "models = models.merge(xenbase_keys, on = 'gene_name', how = 'left')\n",
    "new_models = models.copy(deep = True)\n",
    "display(len(new_models[['gene_name', 'Xenbase']].dropna().drop_duplicates()))\n",
    "\n",
    "new_models['Xenbase'] = new_models['Xenbase'].astype(str)\n",
    "new_models.apply(lambda x: x['field_dictionary'].update({'Xenbase': x['Xenbase']}), axis = 1)\n",
    "new_models['new_fields'] = new_models['field_dictionary'].apply(convert_dict_to_fields_gtf)\n",
    "\n",
    "new_models = new_models[[0, 1, 2, 3, 4, 5, 6, 7, 'new_fields']]\n",
    "\n",
    "models_asgtf_updated_filename = sample_Docket.annot.filename.replace('.gtf', '_updated.gtf')\n",
    "models_asgtf_updated = GenomeGtfFile(models_asgtf_updated_filename, species_SampleDict, GenomeFastaFile = sample_Docket.genome_fasta)\n",
    "\n",
    "new_models.to_csv(models_asgtf_updated.path, sep = '\\t', index = None, header = None)\n",
    "\n",
    "display(new_models)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Extract gene IDs for mapping to UniprotKB\n",
    "Specify which set of identifiers will be use to query the [Uniprot ID Mapping Tool](https://www.uniprot.org/id-mapping) via API.\n",
    "\n",
    "If using an identifier from the `dbxref_dict`, specify the name via string in the `dbxref_datafield` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Xenbase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>vdac2.S</td>\n",
       "      <td>XB-GENE-954844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>samd8.S</td>\n",
       "      <td>XB-GENE-17343426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>dusp13.S</td>\n",
       "      <td>XB-GENE-17337763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>adk.S</td>\n",
       "      <td>XB-GENE-997231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>vcl.S</td>\n",
       "      <td>XB-GENE-5759162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771494</th>\n",
       "      <td>nif3l1.S</td>\n",
       "      <td>XB-GENE-6078750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771661</th>\n",
       "      <td>tpsg1.S</td>\n",
       "      <td>XB-GENE-5893017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771750</th>\n",
       "      <td>loc100491461.S</td>\n",
       "      <td>XB-GENE-6489054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771854</th>\n",
       "      <td>rnf112.1.S</td>\n",
       "      <td>XB-GENE-5960855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771885</th>\n",
       "      <td>rnf112.S</td>\n",
       "      <td>XB-GENE-17336033</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19286 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_name           Xenbase\n",
       "192            vdac2.S    XB-GENE-954844\n",
       "212            samd8.S  XB-GENE-17343426\n",
       "227           dusp13.S  XB-GENE-17337763\n",
       "258              adk.S    XB-GENE-997231\n",
       "281              vcl.S   XB-GENE-5759162\n",
       "...                ...               ...\n",
       "771494        nif3l1.S   XB-GENE-6078750\n",
       "771661         tpsg1.S   XB-GENE-5893017\n",
       "771750  loc100491461.S   XB-GENE-6489054\n",
       "771854      rnf112.1.S   XB-GENE-5960855\n",
       "771885        rnf112.S  XB-GENE-17336033\n",
       "\n",
       "[19286 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dbxref_datafield = ''\n",
    "datafield = 'Xenbase'\n",
    "\n",
    "if dbxref_datafield != '':\n",
    "    models.dropna(axis = 0, subset = ['dbxref_dict'], inplace = True)\n",
    "    models[dbxref_datafield] = [d.get(dbxref_datafield) for d in models['dbxref_dict']]\n",
    "\n",
    "    models_subset = models[['gene_name', dbxref_datafield]].dropna().drop_duplicates()\n",
    "    display(models_subset)\n",
    "\n",
    "elif datafield == 'gene_name':\n",
    "    models_subset = models[['gene_name']].dropna().drop_duplicates()\n",
    "    display(models_subset)\n",
    "\n",
    "elif datafield != '':\n",
    "    models_subset = models[['gene_name', datafield]].dropna().drop_duplicates()\n",
    "    display(models_subset)\n",
    "\n",
    "else:\n",
    "    raise Exception('You must provide a data field for ID mapping.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Generate gene list file to query Uniprot ID Mapping API\n",
    "Generate a text file ending in `_ids.txt` for sending to the ID mapping API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 19285 gene ids to ../../output/Xlae_adultbrain/Xlae_adultbrain_Xenbase_ids.txt\n"
     ]
    }
   ],
   "source": [
    "datafield = dbxref_datafield if dbxref_datafield != '' else datafield\n",
    "\n",
    "gene_list = models_subset[datafield].unique()\n",
    "\n",
    "genelist_object = GeneListFile(\n",
    "    filename = '', # filename is automatically generated from sampledict\n",
    "    sampledict = species_SampleDict,\n",
    "    sources = [sample_Docket.annot],\n",
    "    genes = gene_list,\n",
    "    identifier = datafield\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Query Uniprot ID Mapping API\n",
    "Specify the `from_type` variable based on the Uniprot name of the identifier.\n",
    "The table below lists some databases and the `from_type` string that the API accepts for that datatype.\n",
    "\n",
    "For some reason, the ID mapping API usually returns \"resource not found\" the first time you use it.\n",
    "Try re-running this cell if you receive that error.\n",
    "\n",
    "\n",
    "| datatype | `from_type` string | description |\n",
    "| ---: | :--- | :--- |\n",
    "| Mouse Genome Informatics | `MGI` | ID starts with `MGI:` |\n",
    "| Zebrafish Information Network | `ZFIN` | ID starts with `ZDB-GENE-` |\n",
    "| NCBI Gene ID | `GeneID` | varies, usually numeric string |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  305k    0    52  100  305k     28   168k  0:00:01  0:00:01 --:--:--  168k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100    24    0    24    0     0     38      0 --:--:-- --:--:-- --:--:--    38\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>From</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XB-GENE-954844</td>\n",
       "      <td>Q52KY1</td>\n",
       "      <td>Q52KY1_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Voltage-dependent anion-selective channel prot...</td>\n",
       "      <td>vdac2.S vdac2 vdac2.L</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XB-GENE-997231</td>\n",
       "      <td>Q6DJM1</td>\n",
       "      <td>Q6DJM1_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Adenosine kinase (AK) (EC 2.7.1.20) (Adenosine...</td>\n",
       "      <td>adk.S adk MGC82032</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XB-GENE-5910051</td>\n",
       "      <td>Q6GM53</td>\n",
       "      <td>IKKA_XENLA</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Inhibitor of nuclear factor kappa-B kinase sub...</td>\n",
       "      <td>chuk ikka</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XB-GENE-958583</td>\n",
       "      <td>Q641H0</td>\n",
       "      <td>Q641H0_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Annexin</td>\n",
       "      <td>anxa11.S anx11 anxa11 cap50 MGC81584</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XB-GENE-6251658</td>\n",
       "      <td>Q5FWP7</td>\n",
       "      <td>Q5FWP7_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>MGC84996 protein (nuclear receptor coactivator...</td>\n",
       "      <td>ncoa4.S ara70 ele1 MGC84996 ncoa4 ncoa4-b ptc3...</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17505</th>\n",
       "      <td>XB-GENE-6078750</td>\n",
       "      <td>A0A1L8ERE3</td>\n",
       "      <td>A0A1L8ERE3_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>NIF3-like protein 1</td>\n",
       "      <td>nif3l1.S nif3l1 nif3l1.L</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17506</th>\n",
       "      <td>XB-GENE-6078750</td>\n",
       "      <td>Q0IHC9</td>\n",
       "      <td>Q0IHC9_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>NIF3-like protein 1</td>\n",
       "      <td>nif3l1.S MGC154449 nif3l1 nif3l1.L</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17507</th>\n",
       "      <td>XB-GENE-5893017</td>\n",
       "      <td>A0A1L8ERF3</td>\n",
       "      <td>A0A1L8ERF3_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>serine protease 27</td>\n",
       "      <td>tpsg1.S</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17508</th>\n",
       "      <td>XB-GENE-5960855</td>\n",
       "      <td>A0A1L8ERI6</td>\n",
       "      <td>A0A1L8ERI6_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>RING finger protein 112</td>\n",
       "      <td>rnf112.1.S</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17509</th>\n",
       "      <td>XB-GENE-17336033</td>\n",
       "      <td>A0A1L8ERH8</td>\n",
       "      <td>A0A1L8ERH8_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>RING finger protein 112</td>\n",
       "      <td>rnf112.S</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>481</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17510 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   From       Entry        Entry Name    Reviewed  \\\n",
       "0        XB-GENE-954844      Q52KY1      Q52KY1_XENLA  unreviewed   \n",
       "1        XB-GENE-997231      Q6DJM1      Q6DJM1_XENLA  unreviewed   \n",
       "2       XB-GENE-5910051      Q6GM53        IKKA_XENLA    reviewed   \n",
       "3        XB-GENE-958583      Q641H0      Q641H0_XENLA  unreviewed   \n",
       "4       XB-GENE-6251658      Q5FWP7      Q5FWP7_XENLA  unreviewed   \n",
       "...                 ...         ...               ...         ...   \n",
       "17505   XB-GENE-6078750  A0A1L8ERE3  A0A1L8ERE3_XENLA  unreviewed   \n",
       "17506   XB-GENE-6078750      Q0IHC9      Q0IHC9_XENLA  unreviewed   \n",
       "17507   XB-GENE-5893017  A0A1L8ERF3  A0A1L8ERF3_XENLA  unreviewed   \n",
       "17508   XB-GENE-5960855  A0A1L8ERI6  A0A1L8ERI6_XENLA  unreviewed   \n",
       "17509  XB-GENE-17336033  A0A1L8ERH8  A0A1L8ERH8_XENLA  unreviewed   \n",
       "\n",
       "                                           Protein names  \\\n",
       "0      Voltage-dependent anion-selective channel prot...   \n",
       "1      Adenosine kinase (AK) (EC 2.7.1.20) (Adenosine...   \n",
       "2      Inhibitor of nuclear factor kappa-B kinase sub...   \n",
       "3                                                Annexin   \n",
       "4      MGC84996 protein (nuclear receptor coactivator...   \n",
       "...                                                  ...   \n",
       "17505                                NIF3-like protein 1   \n",
       "17506                                NIF3-like protein 1   \n",
       "17507                                 serine protease 27   \n",
       "17508                            RING finger protein 112   \n",
       "17509                            RING finger protein 112   \n",
       "\n",
       "                                              Gene Names  \\\n",
       "0                                  vdac2.S vdac2 vdac2.L   \n",
       "1                                     adk.S adk MGC82032   \n",
       "2                                              chuk ikka   \n",
       "3                   anxa11.S anx11 anxa11 cap50 MGC81584   \n",
       "4      ncoa4.S ara70 ele1 MGC84996 ncoa4 ncoa4-b ptc3...   \n",
       "...                                                  ...   \n",
       "17505                           nif3l1.S nif3l1 nif3l1.L   \n",
       "17506                 nif3l1.S MGC154449 nif3l1 nif3l1.L   \n",
       "17507                                            tpsg1.S   \n",
       "17508                                         rnf112.1.S   \n",
       "17509                                           rnf112.S   \n",
       "\n",
       "                                   Organism  Length  \n",
       "0      Xenopus laevis (African clawed frog)     283  \n",
       "1      Xenopus laevis (African clawed frog)     361  \n",
       "2      Xenopus laevis (African clawed frog)     743  \n",
       "3      Xenopus laevis (African clawed frog)     502  \n",
       "4      Xenopus laevis (African clawed frog)     625  \n",
       "...                                     ...     ...  \n",
       "17505  Xenopus laevis (African clawed frog)     369  \n",
       "17506  Xenopus laevis (African clawed frog)     344  \n",
       "17507  Xenopus laevis (African clawed frog)     280  \n",
       "17508  Xenopus laevis (African clawed frog)     625  \n",
       "17509  Xenopus laevis (African clawed frog)     481  \n",
       "\n",
       "[17510 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from_type = 'Xenbase'\n",
    "to_type = 'UniProtKB'\n",
    "\n",
    "uniprot_idmm_object = genelist_object.get_uniprot_ids(ID_MAPPER_LOC, from_type, to_type)\n",
    "sample_Docket.add_keyfiles({'uniprot_idmm': uniprot_idmm_object})\n",
    "\n",
    "uniprot_idmm = pd.read_csv(sample_Docket.uniprot_idmm.path, sep = '\\t')\n",
    "display(uniprot_idmm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Extract results and generate Uniprot IDMM\n",
    "Generates an idmm that links `gene_name`, the `dbxref_datafield` seleted above, and `uniprot_id` returned by API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_9203/899271637.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  uniprot_idpairs.rename(columns = {'From': datafield, 'Entry': 'uniprot_id'}, inplace = True)\n",
      "/tmp/ipykernel_9203/899271637.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  uniprot_idpairs[datafield] = uniprot_idpairs[datafield].astype(str)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Xenbase</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XB-GENE-954844</td>\n",
       "      <td>Q52KY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XB-GENE-997231</td>\n",
       "      <td>Q6DJM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XB-GENE-5910051</td>\n",
       "      <td>Q6GM53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XB-GENE-958583</td>\n",
       "      <td>Q641H0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XB-GENE-6251658</td>\n",
       "      <td>Q5FWP7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17505</th>\n",
       "      <td>XB-GENE-6078750</td>\n",
       "      <td>A0A1L8ERE3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17506</th>\n",
       "      <td>XB-GENE-6078750</td>\n",
       "      <td>Q0IHC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17507</th>\n",
       "      <td>XB-GENE-5893017</td>\n",
       "      <td>A0A1L8ERF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17508</th>\n",
       "      <td>XB-GENE-5960855</td>\n",
       "      <td>A0A1L8ERI6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17509</th>\n",
       "      <td>XB-GENE-17336033</td>\n",
       "      <td>A0A1L8ERH8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17510 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Xenbase  uniprot_id\n",
       "0        XB-GENE-954844      Q52KY1\n",
       "1        XB-GENE-997231      Q6DJM1\n",
       "2       XB-GENE-5910051      Q6GM53\n",
       "3        XB-GENE-958583      Q641H0\n",
       "4       XB-GENE-6251658      Q5FWP7\n",
       "...                 ...         ...\n",
       "17505   XB-GENE-6078750  A0A1L8ERE3\n",
       "17506   XB-GENE-6078750      Q0IHC9\n",
       "17507   XB-GENE-5893017  A0A1L8ERF3\n",
       "17508   XB-GENE-5960855  A0A1L8ERI6\n",
       "17509  XB-GENE-17336033  A0A1L8ERH8\n",
       "\n",
       "[17510 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Xenbase</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>vdac2.S</td>\n",
       "      <td>XB-GENE-954844</td>\n",
       "      <td>Q52KY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>adk.S</td>\n",
       "      <td>XB-GENE-997231</td>\n",
       "      <td>Q6DJM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chuk.S</td>\n",
       "      <td>XB-GENE-5910051</td>\n",
       "      <td>Q6GM53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>anxa11.S</td>\n",
       "      <td>XB-GENE-958583</td>\n",
       "      <td>Q641H0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ncoa4.S</td>\n",
       "      <td>XB-GENE-6251658</td>\n",
       "      <td>Q5FWP7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17474</th>\n",
       "      <td>nif3l1.S</td>\n",
       "      <td>XB-GENE-6078750</td>\n",
       "      <td>A0A1L8ERE3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17475</th>\n",
       "      <td>nif3l1.S</td>\n",
       "      <td>XB-GENE-6078750</td>\n",
       "      <td>Q0IHC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17476</th>\n",
       "      <td>tpsg1.S</td>\n",
       "      <td>XB-GENE-5893017</td>\n",
       "      <td>A0A1L8ERF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17477</th>\n",
       "      <td>rnf112.1.S</td>\n",
       "      <td>XB-GENE-5960855</td>\n",
       "      <td>A0A1L8ERI6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17478</th>\n",
       "      <td>rnf112.S</td>\n",
       "      <td>XB-GENE-17336033</td>\n",
       "      <td>A0A1L8ERH8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17479 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name           Xenbase  uniprot_id\n",
       "0         vdac2.S    XB-GENE-954844      Q52KY1\n",
       "1           adk.S    XB-GENE-997231      Q6DJM1\n",
       "2          chuk.S   XB-GENE-5910051      Q6GM53\n",
       "3        anxa11.S    XB-GENE-958583      Q641H0\n",
       "4         ncoa4.S   XB-GENE-6251658      Q5FWP7\n",
       "...           ...               ...         ...\n",
       "17474    nif3l1.S   XB-GENE-6078750  A0A1L8ERE3\n",
       "17475    nif3l1.S   XB-GENE-6078750      Q0IHC9\n",
       "17476     tpsg1.S   XB-GENE-5893017  A0A1L8ERF3\n",
       "17477  rnf112.1.S   XB-GENE-5960855  A0A1L8ERI6\n",
       "17478    rnf112.S  XB-GENE-17336033  A0A1L8ERH8\n",
       "\n",
       "[17479 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "uniprot_idpairs = uniprot_idmm[['From', 'Entry']]\n",
    "uniprot_idpairs.rename(columns = {'From': datafield, 'Entry': 'uniprot_id'}, inplace = True)\n",
    "uniprot_idpairs[datafield] = uniprot_idpairs[datafield].astype(str)\n",
    "display(uniprot_idpairs)\n",
    "\n",
    "uniprot_output_idmm = models_subset.merge(uniprot_idpairs, on = datafield)\n",
    "display(uniprot_output_idmm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate a filename and file for the idmm\n",
    "uniprot_output_idmm_filename = '_'.join([species_prefix, conditions, 'uniprot-idmm.tsv'])\n",
    "uniprot_output_idmm_object = IdmmFile(uniprot_output_idmm_filename, species_SampleDict, kind = 'uniprot_idmm', sources = [sample_Docket.annot])\n",
    "\n",
    "# save to file and add to the Docket\n",
    "uniprot_output_idmm.to_csv(uniprot_output_idmm_object.path, sep = '\\t')\n",
    "sample_Docket.add_keyfile(uniprot_output_idmm_object, 'uniprot_idmm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9. Generate gtf-idmm\n",
    "This file maps the `gene_name` to `gene_id` and `transcript_id` fields generated by the conversion from GFF to GTF, which will be needed for downstream processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2136</td>\n",
       "      <td>2204</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2205</td>\n",
       "      <td>3023</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3024</td>\n",
       "      <td>3092</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3093</td>\n",
       "      <td>4723</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>4724</td>\n",
       "      <td>4798</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771995</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104527008</td>\n",
       "      <td>104527068</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771996</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771997</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771998</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771999</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>772000 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0        1     2          3          4  5  6  7  \\\n",
       "0             MT  Xenbase  exon       2136       2204  .  +  .   \n",
       "1             MT  Xenbase  exon       2205       3023  .  +  .   \n",
       "2             MT  Xenbase  exon       3024       3092  .  +  .   \n",
       "3             MT  Xenbase  exon       3093       4723  .  +  .   \n",
       "4             MT  Xenbase  exon       4724       4798  .  +  .   \n",
       "...          ...      ...   ...        ...        ... .. .. ..   \n",
       "771995  chr9_10S  Xenbase   CDS  104527008  104527068  .  -  0   \n",
       "771996  chr9_10S  Xenbase  exon  104530653  104530753  .  -  2   \n",
       "771997  chr9_10S  Xenbase   CDS  104530653  104530753  .  -  2   \n",
       "771998  chr9_10S  Xenbase  exon  104531027  104531054  .  -  0   \n",
       "771999  chr9_10S  Xenbase   CDS  104531027  104531054  .  -  0   \n",
       "\n",
       "                                                        8  \n",
       "0       gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...  \n",
       "1       gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...  \n",
       "2       gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...  \n",
       "3       gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...  \n",
       "4       gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...  \n",
       "...                                                   ...  \n",
       "771995  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771996  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771997  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771998  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771999  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "\n",
       "[772000 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; transcript_id \"gene34778_t\"; transcript_name \"mt-rnr1.L\"; Xenbase \"nan\"'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "models_asgtf_df = pd.read_csv(models_asgtf_updated.path, sep = '\\t', header = None)\n",
    "\n",
    "display(models_asgtf_df)\n",
    "display(models_asgtf_df[attributes_column][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2136</td>\n",
       "      <td>2204</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...</td>\n",
       "      <td>{'gene_id': 'gene42065', 'gene_name': 'mt-trna...</td>\n",
       "      <td>mt-trna-phe.L</td>\n",
       "      <td>gene42065</td>\n",
       "      <td>gene42065_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2205</td>\n",
       "      <td>3023</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...</td>\n",
       "      <td>{'gene_id': 'gene34778', 'gene_name': 'mt-rnr1...</td>\n",
       "      <td>mt-rnr1.L</td>\n",
       "      <td>gene34778</td>\n",
       "      <td>gene34778_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3024</td>\n",
       "      <td>3092</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...</td>\n",
       "      <td>{'gene_id': 'gene48202', 'gene_name': 'mt-trna...</td>\n",
       "      <td>mt-trna-val.L</td>\n",
       "      <td>gene48202</td>\n",
       "      <td>gene48202_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3093</td>\n",
       "      <td>4723</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...</td>\n",
       "      <td>{'gene_id': 'gene44770', 'gene_name': 'mt-rnr2...</td>\n",
       "      <td>mt-rnr2.L</td>\n",
       "      <td>gene44770</td>\n",
       "      <td>gene44770_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>4724</td>\n",
       "      <td>4798</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...</td>\n",
       "      <td>{'gene_id': 'gene43253', 'gene_name': 'mt-trna...</td>\n",
       "      <td>mt-trna-leu1.L</td>\n",
       "      <td>gene43253</td>\n",
       "      <td>gene43253_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771990</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104524309</td>\n",
       "      <td>104524351</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "      <td>gene24857</td>\n",
       "      <td>rna64354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771992</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104525747</td>\n",
       "      <td>104525868</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "      <td>gene24857</td>\n",
       "      <td>rna64354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771994</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104527008</td>\n",
       "      <td>104527068</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "      <td>gene24857</td>\n",
       "      <td>rna64354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771996</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "      <td>gene24857</td>\n",
       "      <td>rna64354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771998</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "      <td>gene24857</td>\n",
       "      <td>rna64354</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>453943 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0        1     2          3          4  5  6  7  \\\n",
       "0             MT  Xenbase  exon       2136       2204  .  +  .   \n",
       "1             MT  Xenbase  exon       2205       3023  .  +  .   \n",
       "2             MT  Xenbase  exon       3024       3092  .  +  .   \n",
       "3             MT  Xenbase  exon       3093       4723  .  +  .   \n",
       "4             MT  Xenbase  exon       4724       4798  .  +  .   \n",
       "...          ...      ...   ...        ...        ... .. .. ..   \n",
       "771990  chr9_10S  Xenbase  exon  104524309  104524351  .  -  0   \n",
       "771992  chr9_10S  Xenbase  exon  104525747  104525868  .  -  2   \n",
       "771994  chr9_10S  Xenbase  exon  104527008  104527068  .  -  0   \n",
       "771996  chr9_10S  Xenbase  exon  104530653  104530753  .  -  2   \n",
       "771998  chr9_10S  Xenbase  exon  104531027  104531054  .  -  0   \n",
       "\n",
       "                                                        8  \\\n",
       "0       gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...   \n",
       "1       gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...   \n",
       "2       gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...   \n",
       "3       gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...   \n",
       "4       gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...   \n",
       "...                                                   ...   \n",
       "771990  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "771992  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "771994  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "771996  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "771998  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "\n",
       "                                         field_dictionary          gene_name  \\\n",
       "0       {'gene_id': 'gene42065', 'gene_name': 'mt-trna...      mt-trna-phe.L   \n",
       "1       {'gene_id': 'gene34778', 'gene_name': 'mt-rnr1...          mt-rnr1.L   \n",
       "2       {'gene_id': 'gene48202', 'gene_name': 'mt-trna...      mt-trna-val.L   \n",
       "3       {'gene_id': 'gene44770', 'gene_name': 'mt-rnr2...          mt-rnr2.L   \n",
       "4       {'gene_id': 'gene43253', 'gene_name': 'mt-trna...     mt-trna-leu1.L   \n",
       "...                                                   ...                ...   \n",
       "771990  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S   \n",
       "771992  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S   \n",
       "771994  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S   \n",
       "771996  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S   \n",
       "771998  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S   \n",
       "\n",
       "          gene_id transcript_id  \n",
       "0       gene42065   gene42065_t  \n",
       "1       gene34778   gene34778_t  \n",
       "2       gene48202   gene48202_t  \n",
       "3       gene44770   gene44770_t  \n",
       "4       gene43253   gene43253_t  \n",
       "...           ...           ...  \n",
       "771990  gene24857      rna64354  \n",
       "771992  gene24857      rna64354  \n",
       "771994  gene24857      rna64354  \n",
       "771996  gene24857      rna64354  \n",
       "771998  gene24857      rna64354  \n",
       "\n",
       "[453943 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Use a custom function to extract useful fields from the additional fields section (column 8)\n",
    "# Pull from that dict to fill in additional useful columns\n",
    "models_asgtf_df['field_dictionary'] = models_asgtf_df[attributes_column].apply(convert_fields_to_dict_gtf)\n",
    "models_asgtf_df['gene_name'] = [d.get('gene_name') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['gene_id'] = [d.get('gene_id') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['transcript_id'] = [d.get('transcript_id') for d in models_asgtf_df['field_dictionary']]\n",
    "\n",
    "# Remove CDS annotations because they interfere with TransDecoder cDNA generation\n",
    "models_asgtf_df = models_asgtf_df[models_asgtf_df[2] != 'CDS']\n",
    "display(models_asgtf_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mt-trna-phe.L</td>\n",
       "      <td>gene42065</td>\n",
       "      <td>gene42065_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mt-rnr1.L</td>\n",
       "      <td>gene34778</td>\n",
       "      <td>gene34778_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mt-trna-val.L</td>\n",
       "      <td>gene48202</td>\n",
       "      <td>gene48202_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mt-rnr2.L</td>\n",
       "      <td>gene44770</td>\n",
       "      <td>gene44770_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mt-trna-leu1.L</td>\n",
       "      <td>gene43253</td>\n",
       "      <td>gene43253_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771955</th>\n",
       "      <td>vkorc1.S</td>\n",
       "      <td>gene13974</td>\n",
       "      <td>rna40672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771956</th>\n",
       "      <td>vkorc1.S</td>\n",
       "      <td>gene13974</td>\n",
       "      <td>rna40673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771962</th>\n",
       "      <td>LOC108703288</td>\n",
       "      <td>gene37157</td>\n",
       "      <td>rna77797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771977</th>\n",
       "      <td>LOC108702356 [provisional:urgcp]</td>\n",
       "      <td>gene34762</td>\n",
       "      <td>rna74435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771984</th>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "      <td>gene24857</td>\n",
       "      <td>rna64354</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88005 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               gene_name    gene_id transcript_id\n",
       "0                          mt-trna-phe.L  gene42065   gene42065_t\n",
       "1                              mt-rnr1.L  gene34778   gene34778_t\n",
       "2                          mt-trna-val.L  gene48202   gene48202_t\n",
       "3                              mt-rnr2.L  gene44770   gene44770_t\n",
       "4                         mt-trna-leu1.L  gene43253   gene43253_t\n",
       "...                                  ...        ...           ...\n",
       "771955                          vkorc1.S  gene13974      rna40672\n",
       "771956                          vkorc1.S  gene13974      rna40673\n",
       "771962                      LOC108703288  gene37157      rna77797\n",
       "771977  LOC108702356 [provisional:urgcp]  gene34762      rna74435\n",
       "771984                 Xetrov90026942m.S  gene24857      rna64354\n",
       "\n",
       "[88005 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract gene_name, gene_id, and transcript_id fields to generate an ID mapping matrix (idmm)\n",
    "idmm_df = models_asgtf_df[['gene_name', 'gene_id', 'transcript_id']].drop_duplicates()\n",
    "idmm_df.dropna(inplace = True)\n",
    "display(idmm_df)\n",
    "\n",
    "# generate a filename and file for the idmm\n",
    "idmm_filename = '_'.join([species_prefix, conditions, 'gtf-idmm.tsv'])\n",
    "idmm = IdmmFile(idmm_filename, species_SampleDict, kind = 'gtf_idmm', sources = [sample_Docket.annot])\n",
    "\n",
    "# save to file and add to the Docket\n",
    "idmm_df.to_csv(idmm.path, sep = '\\t')\n",
    "sample_Docket.add_keyfile(idmm, 'gtf_idmm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 11. Generate cDNA and peptide files\n",
    "Using the updated gtf file and genome file, generate cDNA sequence.\n",
    "\n",
    "Then, using the cDNA sequence, generate peptide sequences using transdecoder.\n",
    "\n",
    "Expect this step to take some time, probably ~20-30min."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "-parsing cufflinks output: ../../output/Xlae_adultbrain/XENLA_9.2_Xenbase_updated.gtf\n",
      "-parsing genome fasta: ../../output/Xlae_adultbrain/XENLA_9.2_genome.fa\n",
      "-done parsing genome.\n",
      "// processing MT\n",
      "// processing Scaffold100\n",
      "// processing Scaffold1000\n",
      "// processing Scaffold100330\n",
      "// processing Scaffold10049\n",
      "// processing Scaffold10052\n",
      "// processing Scaffold100627\n",
      "// processing Scaffold10094\n",
      "// processing Scaffold101\n",
      "// processing Scaffold10127\n",
      "// processing Scaffold10158\n",
      "// processing Scaffold10166\n",
      "// processing Scaffold101757\n",
      "// processing Scaffold101947\n",
      "// processing Scaffold101966\n",
      "// processing Scaffold102\n",
      "// processing Scaffold10215\n",
      "// processing Scaffold10219\n",
      "// processing Scaffold1022\n",
      "// processing Scaffold1024\n",
      "// processing Scaffold10260\n",
      "// processing Scaffold10281\n",
      "// processing Scaffold102870\n",
      "// processing Scaffold103\n",
      "// processing Scaffold1032\n",
      "// processing Scaffold103277\n",
      "// processing Scaffold103338\n",
      "// processing Scaffold103460\n",
      "// processing Scaffold10353\n",
      "// processing Scaffold10361\n",
      "// processing Scaffold1037\n",
      "// processing Scaffold104\n",
      "// processing Scaffold10422\n",
      "// processing Scaffold10429\n",
      "// processing Scaffold1045\n",
      "// processing Scaffold105\n",
      "// processing Scaffold1050\n",
      "// processing Scaffold1052\n",
      "// processing Scaffold10527\n",
      "// processing Scaffold1053\n",
      "// processing Scaffold10545\n",
      "// processing Scaffold10554\n",
      "// processing Scaffold105971\n",
      "// processing Scaffold106\n",
      "// processing Scaffold10614\n",
      "// processing Scaffold10631\n",
      "// processing Scaffold10656\n",
      "// processing Scaffold10678\n",
      "// processing Scaffold1068\n",
      "// processing Scaffold107\n",
      "// processing Scaffold107097\n",
      "// processing Scaffold10719\n",
      "// processing Scaffold107334\n",
      "// processing Scaffold10742\n",
      "// processing Scaffold1076\n",
      "// processing Scaffold10766\n",
      "// processing Scaffold10776\n",
      "// processing Scaffold108\n",
      "// processing Scaffold10808\n",
      "// processing Scaffold10822\n",
      "// processing Scaffold108574\n",
      "// processing Scaffold10870\n",
      "// processing Scaffold109\n",
      "// processing Scaffold109172\n",
      "// processing Scaffold1094\n",
      "// processing Scaffold10969\n",
      "// processing Scaffold110\n",
      "// processing Scaffold1103\n",
      "// processing Scaffold110332\n",
      "// processing Scaffold11038\n",
      "// processing Scaffold11046\n",
      "// processing Scaffold110469\n",
      "// processing Scaffold11074\n",
      "// processing Scaffold1109\n",
      "// processing Scaffold11092\n",
      "// processing Scaffold111\n",
      "// processing Scaffold111047\n",
      "// processing Scaffold1112\n",
      "// processing Scaffold11121\n",
      "// processing Scaffold1114\n",
      "// processing Scaffold1117\n",
      "// processing Scaffold1119\n",
      "// processing Scaffold112\n",
      "// processing Scaffold112450\n",
      "// processing Scaffold113\n",
      "// processing Scaffold11305\n",
      "// processing Scaffold11341\n",
      "// processing Scaffold11369\n",
      "// processing Scaffold11372\n",
      "// processing Scaffold11373\n",
      "// processing Scaffold11376\n",
      "// processing Scaffold1139\n",
      "// processing Scaffold114\n",
      "// processing Scaffold114193\n",
      "// processing Scaffold11427\n",
      "// processing Scaffold11438\n",
      "// processing Scaffold11468\n",
      "// processing Scaffold11473\n",
      "// processing Scaffold115\n",
      "// processing Scaffold11522\n",
      "// processing Scaffold11564\n",
      "// processing Scaffold116\n",
      "// processing Scaffold11616\n",
      "// processing Scaffold11618\n",
      "// processing Scaffold11627\n",
      "// processing Scaffold11633\n",
      "// processing Scaffold11669\n",
      "// processing Scaffold117\n",
      "// processing Scaffold11701\n",
      "// processing Scaffold11739\n",
      "// processing Scaffold11772\n",
      "// processing Scaffold11776\n",
      "// processing Scaffold118\n",
      "// processing Scaffold11846\n",
      "// processing Scaffold1185\n",
      "// processing Scaffold11872\n",
      "// processing Scaffold11941\n",
      "// processing Scaffold11981\n",
      "// processing Scaffold120\n",
      "// processing Scaffold12031\n",
      "// processing Scaffold1204\n",
      "// processing Scaffold12066\n",
      "// processing Scaffold12085\n",
      "// processing Scaffold12099\n",
      "// processing Scaffold121\n",
      "// processing Scaffold12139\n",
      "// processing Scaffold12192\n",
      "// processing Scaffold122\n",
      "// processing Scaffold1223\n",
      "// processing Scaffold12266\n",
      "// processing Scaffold123\n",
      "// processing Scaffold12300\n",
      "// processing Scaffold12366\n",
      "// processing Scaffold1237\n",
      "// processing Scaffold124\n",
      "// processing Scaffold1240\n",
      "// processing Scaffold12436\n",
      "// processing Scaffold12441\n",
      "// processing Scaffold1249\n",
      "// processing Scaffold125\n",
      "// processing Scaffold1254\n",
      "// processing Scaffold12543\n",
      "// processing Scaffold1255\n",
      "// processing Scaffold126\n",
      "// processing Scaffold1260\n",
      "// processing Scaffold1261\n",
      "// processing Scaffold1265\n",
      "// processing Scaffold12656\n",
      "// processing Scaffold12673\n",
      "// processing Scaffold12695\n",
      "// processing Scaffold127\n",
      "// processing Scaffold1276\n",
      "// processing Scaffold12767\n",
      "// processing Scaffold12782\n",
      "// processing Scaffold128\n",
      "// processing Scaffold12811\n",
      "// processing Scaffold12843\n",
      "// processing Scaffold129\n",
      "// processing Scaffold12907\n",
      "// processing Scaffold12966\n",
      "// processing Scaffold12978\n",
      "// processing Scaffold1298\n",
      "// processing Scaffold12993\n",
      "// processing Scaffold130\n",
      "// processing Scaffold1303\n",
      "// processing Scaffold1305\n",
      "// processing Scaffold131\n",
      "// processing Scaffold13100\n",
      "// processing Scaffold13178\n",
      "// processing Scaffold132\n",
      "// processing Scaffold13220\n",
      "// processing Scaffold13230\n",
      "// processing Scaffold133\n",
      "// processing Scaffold13311\n",
      "// processing Scaffold13317\n",
      "// processing Scaffold13319\n",
      "// processing Scaffold1334\n",
      "// processing Scaffold13371\n",
      "// processing Scaffold13374\n",
      "// processing Scaffold134\n",
      "// processing Scaffold1341\n",
      "// processing Scaffold13416\n",
      "// processing Scaffold1346\n",
      "// processing Scaffold13493\n",
      "// processing Scaffold1352\n",
      "// processing Scaffold1353\n",
      "// processing Scaffold13553\n",
      "// processing Scaffold13631\n",
      "// processing Scaffold13679\n",
      "// processing Scaffold13727\n",
      "// processing Scaffold1376\n",
      "// processing Scaffold13775\n",
      "// processing Scaffold138\n",
      "// processing Scaffold13835\n",
      "// processing Scaffold13844\n",
      "// processing Scaffold13863\n",
      "// processing Scaffold139\n",
      "// processing Scaffold13957\n",
      "// processing Scaffold140\n",
      "// processing Scaffold14010\n",
      "// processing Scaffold1403\n",
      "// processing Scaffold14031\n",
      "// processing Scaffold14056\n",
      "// processing Scaffold1409\n",
      "// processing Scaffold14124\n",
      "// processing Scaffold1413\n",
      "// processing Scaffold14158\n",
      "// processing Scaffold1417\n",
      "// processing Scaffold142\n",
      "// processing Scaffold14205\n",
      "// processing Scaffold1421\n",
      "// processing Scaffold14242\n",
      "// processing Scaffold14272\n",
      "// processing Scaffold1429\n",
      "// processing Scaffold143\n",
      "// processing Scaffold1433\n",
      "// processing Scaffold14365\n",
      "// processing Scaffold14381\n",
      "// processing Scaffold14397\n",
      "// processing Scaffold144\n",
      "// processing Scaffold14458\n",
      "// processing Scaffold14478\n",
      "// processing Scaffold14498\n",
      "// processing Scaffold145\n",
      "// processing Scaffold14551\n",
      "// processing Scaffold146\n",
      "// processing Scaffold1462\n",
      "// processing Scaffold14633\n",
      "// processing Scaffold14684\n",
      "// processing Scaffold14695\n",
      "// processing Scaffold147\n",
      "// processing Scaffold14712\n",
      "// processing Scaffold148\n",
      "// processing Scaffold14802\n",
      "// processing Scaffold14809\n",
      "// processing Scaffold149\n",
      "// processing Scaffold1493\n",
      "// processing Scaffold1499\n",
      "// processing Scaffold150\n",
      "// processing Scaffold15029\n",
      "// processing Scaffold1503\n",
      "// processing Scaffold1506\n",
      "// processing Scaffold151\n",
      "// processing Scaffold1510\n",
      "// processing Scaffold15115\n",
      "// processing Scaffold15180\n",
      "// processing Scaffold1522\n",
      "// processing Scaffold1523\n",
      "// processing Scaffold153\n",
      "// processing Scaffold15301\n",
      "// processing Scaffold1532\n",
      "// processing Scaffold15327\n",
      "// processing Scaffold1533\n",
      "// processing Scaffold15350\n",
      "// processing Scaffold15387\n",
      "// processing Scaffold154\n",
      "// processing Scaffold15422\n",
      "// processing Scaffold15485\n",
      "// processing Scaffold155\n",
      "// processing Scaffold15509\n",
      "// processing Scaffold1551\n",
      "// processing Scaffold15526\n",
      "// processing Scaffold156\n",
      "// processing Scaffold15613\n",
      "// processing Scaffold15667\n",
      "// processing Scaffold157\n",
      "// processing Scaffold15729\n",
      "// processing Scaffold15795\n",
      "// processing Scaffold158\n",
      "// processing Scaffold1580\n",
      "// processing Scaffold15816\n",
      "// processing Scaffold15842\n",
      "// processing Scaffold1587\n",
      "// processing Scaffold15922\n",
      "// processing Scaffold15940\n",
      "// processing Scaffold15953\n",
      "// processing Scaffold15959\n",
      "// processing Scaffold15970\n",
      "// processing Scaffold15982\n",
      "// processing Scaffold16010\n",
      "// processing Scaffold1607\n",
      "// processing Scaffold1609\n",
      "// processing Scaffold161\n",
      "// processing Scaffold16170\n",
      "// processing Scaffold1619\n",
      "// processing Scaffold162\n",
      "// processing Scaffold16207\n",
      "// processing Scaffold16274\n",
      "// processing Scaffold1629\n",
      "// processing Scaffold16328\n",
      "// processing Scaffold16334\n",
      "// processing Scaffold16369\n",
      "// processing Scaffold16377\n",
      "// processing Scaffold1638\n",
      "// processing Scaffold16396\n",
      "// processing Scaffold164\n",
      "// processing Scaffold16431\n",
      "// processing Scaffold16461\n",
      "// processing Scaffold16467\n",
      "// processing Scaffold16491\n",
      "// processing Scaffold165\n",
      "// processing Scaffold16546\n",
      "// processing Scaffold1657\n",
      "// processing Scaffold16577\n",
      "// processing Scaffold16583\n",
      "// processing Scaffold16589\n",
      "// processing Scaffold166\n",
      "// processing Scaffold16636\n",
      "// processing Scaffold16644\n",
      "// processing Scaffold16665\n",
      "// processing Scaffold16714\n",
      "// processing Scaffold16715\n",
      "// processing Scaffold16721\n",
      "// processing Scaffold1675\n",
      "// processing Scaffold1677\n",
      "// processing Scaffold16781\n",
      "// processing Scaffold168\n",
      "// processing Scaffold1683\n",
      "// processing Scaffold1686\n",
      "// processing Scaffold169\n",
      "// processing Scaffold16915\n",
      "// processing Scaffold16918\n",
      "// processing Scaffold16923\n",
      "// processing Scaffold16965\n",
      "// processing Scaffold16973\n",
      "// processing Scaffold170\n",
      "// processing Scaffold17034\n",
      "// processing Scaffold1704\n",
      "// processing Scaffold1713\n",
      "// processing Scaffold17159\n",
      "// processing Scaffold1718\n",
      "// processing Scaffold172\n",
      "// processing Scaffold17200\n",
      "// processing Scaffold17231\n",
      "// processing Scaffold1724\n",
      "// processing Scaffold17266\n",
      "// processing Scaffold17279\n",
      "// processing Scaffold173\n",
      "// processing Scaffold17320\n",
      "// processing Scaffold1733\n",
      "// processing Scaffold17346\n",
      "// processing Scaffold17402\n",
      "// processing Scaffold17410\n",
      "// processing Scaffold17460\n",
      "// processing Scaffold175\n",
      "// processing Scaffold17547\n",
      "// processing Scaffold1755\n",
      "// processing Scaffold1759\n",
      "// processing Scaffold176\n",
      "// processing Scaffold17651\n",
      "// processing Scaffold177\n",
      "// processing Scaffold1770\n",
      "// processing Scaffold17797\n",
      "// processing Scaffold17798\n",
      "// processing Scaffold178\n",
      "// processing Scaffold1780\n",
      "// processing Scaffold17803\n",
      "// processing Scaffold17823\n",
      "// processing Scaffold179\n",
      "// processing Scaffold17907\n",
      "// processing Scaffold17914\n",
      "// processing Scaffold17931\n",
      "// processing Scaffold1795\n",
      "// processing Scaffold1798\n",
      "// processing Scaffold17986\n",
      "// processing Scaffold18008\n",
      "// processing Scaffold18033\n",
      "// processing Scaffold1804\n",
      "// processing Scaffold18042\n",
      "// processing Scaffold18077\n",
      "// processing Scaffold181\n",
      "// processing Scaffold1816\n",
      "// processing Scaffold1817\n",
      "// processing Scaffold1821\n",
      "// processing Scaffold18292\n",
      "// processing Scaffold18294\n",
      "// processing Scaffold183\n",
      "// processing Scaffold18309\n",
      "// processing Scaffold1831\n",
      "// processing Scaffold1832\n",
      "// processing Scaffold1834\n",
      "// processing Scaffold184\n",
      "// processing Scaffold185\n",
      "// processing Scaffold18507\n",
      "// processing Scaffold18527\n",
      "// processing Scaffold1858\n",
      "// processing Scaffold186\n",
      "// processing Scaffold1868\n",
      "// processing Scaffold18689\n",
      "// processing Scaffold18692\n",
      "// processing Scaffold187\n",
      "// processing Scaffold1871\n",
      "// processing Scaffold18726\n",
      "// processing Scaffold18788\n",
      "// processing Scaffold188\n",
      "// processing Scaffold18859\n",
      "// processing Scaffold1887\n",
      "// processing Scaffold1888\n",
      "// processing Scaffold189\n",
      "// processing Scaffold18990\n",
      "// processing Scaffold19\n",
      "// processing Scaffold190\n",
      "// processing Scaffold191\n",
      "// processing Scaffold192\n",
      "// processing Scaffold19210\n",
      "// processing Scaffold19234\n",
      "// processing Scaffold193\n",
      "// processing Scaffold19357\n",
      "// processing Scaffold19361\n",
      "// processing Scaffold19367\n",
      "// processing Scaffold194\n",
      "// processing Scaffold19414\n",
      "// processing Scaffold19438\n",
      "// processing Scaffold19464\n",
      "// processing Scaffold19487\n",
      "// processing Scaffold195\n",
      "// processing Scaffold19512\n",
      "// processing Scaffold1952\n",
      "// processing Scaffold19590\n",
      "// processing Scaffold196\n",
      "// processing Scaffold1961\n",
      "// processing Scaffold19639\n",
      "// processing Scaffold19665\n",
      "// processing Scaffold19676\n",
      "// processing Scaffold19697\n",
      "// processing Scaffold197\n",
      "// processing Scaffold19746\n",
      "// processing Scaffold19748\n",
      "// processing Scaffold19767\n",
      "// processing Scaffold19774\n",
      "// processing Scaffold19784\n",
      "// processing Scaffold198\n",
      "// processing Scaffold19825\n",
      "// processing Scaffold19837\n",
      "// processing Scaffold19870\n",
      "// processing Scaffold19885\n",
      "// processing Scaffold199\n",
      "// processing Scaffold19939\n",
      "// processing Scaffold19945\n",
      "// processing Scaffold19961\n",
      "// processing Scaffold20\n",
      "// processing Scaffold20072\n",
      "// processing Scaffold201\n",
      "// processing Scaffold2012\n",
      "// processing Scaffold2016\n",
      "// processing Scaffold20184\n",
      "// processing Scaffold20199\n",
      "// processing Scaffold20225\n",
      "// processing Scaffold2028\n",
      "// processing Scaffold203\n",
      "// processing Scaffold20323\n",
      "// processing Scaffold20387\n",
      "// processing Scaffold20413\n",
      "// processing Scaffold20422\n",
      "// processing Scaffold20467\n",
      "// processing Scaffold205\n",
      "// processing Scaffold20510\n",
      "// processing Scaffold20539\n",
      "// processing Scaffold20587\n",
      "// processing Scaffold206\n",
      "// processing Scaffold2063\n",
      "// processing Scaffold20650\n",
      "// processing Scaffold207\n",
      "// processing Scaffold2074\n",
      "// processing Scaffold20791\n",
      "// processing Scaffold20793\n",
      "// processing Scaffold208\n",
      "// processing Scaffold20836\n",
      "// processing Scaffold20886\n",
      "// processing Scaffold20890\n",
      "// processing Scaffold209\n",
      "// processing Scaffold20961\n",
      "// processing Scaffold20985\n",
      "// processing Scaffold21\n",
      "// processing Scaffold210\n",
      "// processing Scaffold21072\n",
      "// processing Scaffold211\n",
      "// processing Scaffold21129\n",
      "// processing Scaffold21194\n",
      "// processing Scaffold21202\n",
      "// processing Scaffold21221\n",
      "// processing Scaffold21250\n",
      "// processing Scaffold21267\n",
      "// processing Scaffold21279\n",
      "// processing Scaffold213\n",
      "// processing Scaffold21313\n",
      "// processing Scaffold21339\n",
      "// processing Scaffold214\n",
      "// processing Scaffold21499\n",
      "// processing Scaffold215\n",
      "// processing Scaffold2155\n",
      "// processing Scaffold21560\n",
      "// processing Scaffold21569\n",
      "// processing Scaffold216\n",
      "// processing Scaffold21630\n",
      "// processing Scaffold21635\n",
      "// processing Scaffold21651\n",
      "// processing Scaffold21653\n",
      "// processing Scaffold217\n",
      "// processing Scaffold2171\n",
      "// processing Scaffold21790\n",
      "// processing Scaffold21797\n",
      "// processing Scaffold2181\n",
      "// processing Scaffold2182\n",
      "// processing Scaffold2184\n",
      "// processing Scaffold21849\n",
      "// processing Scaffold21850\n",
      "// processing Scaffold21929\n",
      "// processing Scaffold21944\n",
      "// processing Scaffold21977\n",
      "// processing Scaffold21998\n",
      "// processing Scaffold22\n",
      "// processing Scaffold220\n",
      "// processing Scaffold22002\n",
      "// processing Scaffold2204\n",
      "// processing Scaffold22062\n",
      "// processing Scaffold221\n",
      "// processing Scaffold2213\n",
      "// processing Scaffold22188\n",
      "// processing Scaffold22198\n",
      "// processing Scaffold22217\n",
      "// processing Scaffold22288\n",
      "// processing Scaffold223\n",
      "// processing Scaffold22366\n",
      "// processing Scaffold22410\n",
      "// processing Scaffold22417\n",
      "// processing Scaffold22459\n",
      "// processing Scaffold225\n",
      "// processing Scaffold22536\n",
      "// processing Scaffold22597\n",
      "// processing Scaffold226\n",
      "// processing Scaffold22610\n",
      "// processing Scaffold22614\n",
      "// processing Scaffold22637\n",
      "// processing Scaffold2268\n",
      "// processing Scaffold22685\n",
      "// processing Scaffold22699\n",
      "// processing Scaffold227\n",
      "// processing Scaffold22723\n",
      "// processing Scaffold22737\n",
      "// processing Scaffold22755\n",
      "// processing Scaffold22760\n",
      "// processing Scaffold22794\n",
      "// processing Scaffold228\n",
      "// processing Scaffold22831\n",
      "// processing Scaffold22879\n",
      "// processing Scaffold229\n",
      "// processing Scaffold22900\n",
      "// processing Scaffold22934\n",
      "// processing Scaffold22992\n",
      "// processing Scaffold23\n",
      "// processing Scaffold23012\n",
      "// processing Scaffold232\n",
      "// processing Scaffold23229\n",
      "// processing Scaffold23249\n",
      "// processing Scaffold23254\n",
      "// processing Scaffold233\n",
      "// processing Scaffold23327\n",
      "// processing Scaffold2333\n",
      "// processing Scaffold23335\n",
      "// processing Scaffold234\n",
      "// processing Scaffold23466\n",
      "// processing Scaffold23469\n",
      "// processing Scaffold23502\n",
      "// processing Scaffold23583\n",
      "// processing Scaffold236\n",
      "// processing Scaffold23611\n",
      "// processing Scaffold23651\n",
      "// processing Scaffold237\n",
      "// processing Scaffold23701\n",
      "// processing Scaffold23727\n",
      "// processing Scaffold23728\n",
      "// processing Scaffold23732\n",
      "// processing Scaffold23741\n",
      "// processing Scaffold238\n",
      "// processing Scaffold23805\n",
      "// processing Scaffold2384\n",
      "// processing Scaffold23928\n",
      "// processing Scaffold23938\n",
      "// processing Scaffold2394\n",
      "// processing Scaffold23953\n",
      "// processing Scaffold23985\n",
      "// processing Scaffold23995\n",
      "// processing Scaffold24\n",
      "// processing Scaffold24009\n",
      "// processing Scaffold24017\n",
      "// processing Scaffold24020\n",
      "// processing Scaffold24051\n",
      "// processing Scaffold241\n",
      "// processing Scaffold2411\n",
      "// processing Scaffold24120\n",
      "// processing Scaffold24166\n",
      "// processing Scaffold2417\n",
      "// processing Scaffold24189\n",
      "// processing Scaffold243\n",
      "// processing Scaffold24320\n",
      "// processing Scaffold24331\n",
      "// processing Scaffold24337\n",
      "// processing Scaffold244\n",
      "// processing Scaffold24427\n",
      "// processing Scaffold2443\n",
      "// processing Scaffold24448\n",
      "// processing Scaffold24485\n",
      "// processing Scaffold245\n",
      "// processing Scaffold24514\n",
      "// processing Scaffold24523\n",
      "// processing Scaffold24554\n",
      "// processing Scaffold2457\n",
      "// processing Scaffold24580\n",
      "// processing Scaffold24625\n",
      "// processing Scaffold2463\n",
      "// processing Scaffold24639\n",
      "// processing Scaffold247\n",
      "// processing Scaffold24711\n",
      "// processing Scaffold24718\n",
      "// processing Scaffold248\n",
      "// processing Scaffold24819\n",
      "// processing Scaffold24843\n",
      "// processing Scaffold2485\n",
      "// processing Scaffold249\n",
      "// processing Scaffold24904\n",
      "// processing Scaffold24945\n",
      "// processing Scaffold24962\n",
      "// processing Scaffold25\n",
      "// processing Scaffold250\n",
      "// processing Scaffold25028\n",
      "// processing Scaffold2504\n",
      "// processing Scaffold25050\n",
      "// processing Scaffold25053\n",
      "// processing Scaffold251\n",
      "// processing Scaffold25114\n",
      "// processing Scaffold25129\n",
      "// processing Scaffold25157\n",
      "// processing Scaffold25161\n",
      "// processing Scaffold25165\n",
      "// processing Scaffold25208\n",
      "// processing Scaffold2523\n",
      "// processing Scaffold25251\n",
      "// processing Scaffold25301\n",
      "// processing Scaffold25353\n",
      "// processing Scaffold254\n",
      "// processing Scaffold25422\n",
      "// processing Scaffold25452\n",
      "// processing Scaffold2549\n",
      "// processing Scaffold255\n",
      "// processing Scaffold25509\n",
      "// processing Scaffold256\n",
      "// processing Scaffold25627\n",
      "// processing Scaffold25634\n",
      "// processing Scaffold25648\n",
      "// processing Scaffold25714\n",
      "// processing Scaffold25783\n",
      "// processing Scaffold258\n",
      "// processing Scaffold25834\n",
      "// processing Scaffold25835\n",
      "// processing Scaffold25854\n",
      "// processing Scaffold25897\n",
      "// processing Scaffold259\n",
      "// processing Scaffold25933\n",
      "// processing Scaffold25994\n",
      "// processing Scaffold26\n",
      "// processing Scaffold26023\n",
      "// processing Scaffold26051\n",
      "// processing Scaffold26092\n",
      "// processing Scaffold261\n",
      "// processing Scaffold2615\n",
      "// processing Scaffold26151\n",
      "// processing Scaffold26164\n",
      "// processing Scaffold26172\n",
      "// processing Scaffold2628\n",
      "// processing Scaffold263\n",
      "// processing Scaffold26341\n",
      "// processing Scaffold26344\n",
      "// processing Scaffold2646\n",
      "// processing Scaffold26471\n",
      "// processing Scaffold2656\n",
      "// processing Scaffold26565\n",
      "// processing Scaffold26588\n",
      "// processing Scaffold2659\n",
      "// processing Scaffold266\n",
      "// processing Scaffold26605\n",
      "// processing Scaffold26628\n",
      "// processing Scaffold26655\n",
      "// processing Scaffold26770\n",
      "// processing Scaffold26786\n",
      "// processing Scaffold26792\n",
      "// processing Scaffold268\n",
      "// processing Scaffold2681\n",
      "// processing Scaffold26824\n",
      "// processing Scaffold26893\n",
      "// processing Scaffold269\n",
      "// processing Scaffold26908\n",
      "// processing Scaffold27\n",
      "// processing Scaffold27041\n",
      "// processing Scaffold27045\n",
      "// processing Scaffold27052\n",
      "// processing Scaffold27093\n",
      "// processing Scaffold27139\n",
      "// processing Scaffold272\n",
      "// processing Scaffold27224\n",
      "// processing Scaffold273\n",
      "// processing Scaffold274\n",
      "// processing Scaffold2741\n",
      "// processing Scaffold27438\n",
      "// processing Scaffold27460\n",
      "// processing Scaffold27541\n",
      "// processing Scaffold276\n",
      "// processing Scaffold27603\n",
      "// processing Scaffold27609\n",
      "// processing Scaffold27772\n",
      "// processing Scaffold27825\n",
      "// processing Scaffold27875\n",
      "// processing Scaffold279\n",
      "// processing Scaffold27911\n",
      "// processing Scaffold27916\n",
      "// processing Scaffold27930\n",
      "// processing Scaffold28\n",
      "// processing Scaffold280\n",
      "// processing Scaffold2802\n",
      "// processing Scaffold28029\n",
      "// processing Scaffold28140\n",
      "// processing Scaffold282\n",
      "// processing Scaffold28210\n",
      "// processing Scaffold28219\n",
      "// processing Scaffold28229\n",
      "// processing Scaffold28253\n",
      "// processing Scaffold283\n",
      "// processing Scaffold28370\n",
      "// processing Scaffold284\n",
      "// processing Scaffold28425\n",
      "// processing Scaffold28428\n",
      "// processing Scaffold28429\n",
      "// processing Scaffold28495\n",
      "// processing Scaffold285\n",
      "// processing Scaffold2856\n",
      "// processing Scaffold28562\n",
      "// processing Scaffold286\n",
      "// processing Scaffold28656\n",
      "// processing Scaffold287\n",
      "// processing Scaffold28704\n",
      "// processing Scaffold28783\n",
      "// processing Scaffold288\n",
      "// processing Scaffold28836\n",
      "// processing Scaffold28840\n",
      "// processing Scaffold28841\n",
      "// processing Scaffold28858\n",
      "// processing Scaffold28928\n",
      "// processing Scaffold28944\n",
      "// processing Scaffold28949\n",
      "// processing Scaffold28975\n",
      "// processing Scaffold28977\n",
      "// processing Scaffold29\n",
      "// processing Scaffold290\n",
      "// processing Scaffold29026\n",
      "// processing Scaffold291\n",
      "// processing Scaffold29124\n",
      "// processing Scaffold29217\n",
      "// processing Scaffold29231\n",
      "// processing Scaffold29244\n",
      "// processing Scaffold29253\n",
      "// processing Scaffold29254\n",
      "// processing Scaffold29372\n",
      "// processing Scaffold29414\n",
      "// processing Scaffold2943\n",
      "// processing Scaffold29437\n",
      "// processing Scaffold295\n",
      "// processing Scaffold2961\n",
      "// processing Scaffold2962\n",
      "// processing Scaffold29746\n",
      "// processing Scaffold29752\n",
      "// processing Scaffold29832\n",
      "// processing Scaffold29894\n",
      "// processing Scaffold2996\n",
      "// processing Scaffold29976\n",
      "// processing Scaffold29994\n",
      "// processing Scaffold30\n",
      "// processing Scaffold300\n",
      "// processing Scaffold30013\n",
      "// processing Scaffold30034\n",
      "// processing Scaffold30128\n",
      "// processing Scaffold30135\n",
      "// processing Scaffold30139\n",
      "// processing Scaffold30160\n",
      "// processing Scaffold30196\n",
      "// processing Scaffold303\n",
      "// processing Scaffold30352\n",
      "// processing Scaffold3039\n",
      "// processing Scaffold304\n",
      "// processing Scaffold3043\n",
      "// processing Scaffold30431\n",
      "// processing Scaffold3045\n",
      "// processing Scaffold3047\n",
      "// processing Scaffold30566\n",
      "// processing Scaffold30588\n",
      "// processing Scaffold306\n",
      "// processing Scaffold30617\n",
      "// processing Scaffold3074\n",
      "// processing Scaffold30740\n",
      "// processing Scaffold30785\n",
      "// processing Scaffold3080\n",
      "// processing Scaffold3081\n",
      "// processing Scaffold30815\n",
      "// processing Scaffold3086\n",
      "// processing Scaffold309\n",
      "// processing Scaffold31\n",
      "// processing Scaffold3105\n",
      "// processing Scaffold31073\n",
      "// processing Scaffold311\n",
      "// processing Scaffold3113\n",
      "// processing Scaffold31147\n",
      "// processing Scaffold31153\n",
      "// processing Scaffold31157\n",
      "// processing Scaffold312\n",
      "// processing Scaffold3126\n",
      "// processing Scaffold31340\n",
      "// processing Scaffold3137\n",
      "// processing Scaffold314\n",
      "// processing Scaffold31413\n",
      "// processing Scaffold31549\n",
      "// processing Scaffold3156\n",
      "// processing Scaffold31617\n",
      "// processing Scaffold31635\n",
      "// processing Scaffold31679\n",
      "// processing Scaffold317\n",
      "// processing Scaffold31700\n",
      "// processing Scaffold31747\n",
      "// processing Scaffold318\n",
      "// processing Scaffold31814\n",
      "// processing Scaffold31883\n",
      "// processing Scaffold31915\n",
      "// processing Scaffold32\n",
      "// processing Scaffold32051\n",
      "// processing Scaffold3206\n",
      "// processing Scaffold32066\n",
      "// processing Scaffold32075\n",
      "// processing Scaffold321\n",
      "// processing Scaffold32141\n",
      "// processing Scaffold322\n",
      "// processing Scaffold32353\n",
      "// processing Scaffold32366\n",
      "// processing Scaffold324\n",
      "// processing Scaffold326\n",
      "// processing Scaffold3267\n",
      "// processing Scaffold32683\n",
      "// processing Scaffold32686\n",
      "// processing Scaffold32746\n",
      "// processing Scaffold32791\n",
      "// processing Scaffold3295\n",
      "// processing Scaffold32975\n",
      "// processing Scaffold32978\n",
      "// processing Scaffold32984\n",
      "// processing Scaffold33\n",
      "// processing Scaffold331\n",
      "// processing Scaffold33137\n",
      "// processing Scaffold332\n",
      "// processing Scaffold33203\n",
      "// processing Scaffold33267\n",
      "// processing Scaffold333\n",
      "// processing Scaffold3342\n",
      "// processing Scaffold33444\n",
      "// processing Scaffold33527\n",
      "// processing Scaffold33547\n",
      "// processing Scaffold33562\n",
      "// processing Scaffold33584\n",
      "// processing Scaffold33629\n",
      "// processing Scaffold3371\n",
      "// processing Scaffold33715\n",
      "// processing Scaffold33747\n",
      "// processing Scaffold3375\n",
      "// processing Scaffold33764\n",
      "// processing Scaffold33766\n",
      "// processing Scaffold33786\n",
      "// processing Scaffold338\n",
      "// processing Scaffold339\n",
      "// processing Scaffold33997\n",
      "// processing Scaffold34\n",
      "// processing Scaffold340\n",
      "// processing Scaffold3400\n",
      "// processing Scaffold34037\n",
      "// processing Scaffold342\n",
      "// processing Scaffold34210\n",
      "// processing Scaffold34230\n",
      "// processing Scaffold34234\n",
      "// processing Scaffold34238\n",
      "// processing Scaffold34268\n",
      "// processing Scaffold34303\n",
      "// processing Scaffold34422\n",
      "// processing Scaffold345\n",
      "// processing Scaffold34535\n",
      "// processing Scaffold34548\n",
      "// processing Scaffold346\n",
      "// processing Scaffold3462\n",
      "// processing Scaffold34638\n",
      "// processing Scaffold34646\n",
      "// processing Scaffold34679\n",
      "// processing Scaffold347\n",
      "// processing Scaffold34732\n",
      "// processing Scaffold3474\n",
      "// processing Scaffold34791\n",
      "// processing Scaffold348\n",
      "// processing Scaffold3483\n",
      "// processing Scaffold34879\n",
      "// processing Scaffold34887\n",
      "// processing Scaffold349\n",
      "// processing Scaffold35\n",
      "// processing Scaffold35071\n",
      "// processing Scaffold3512\n",
      "// processing Scaffold3514\n",
      "// processing Scaffold35165\n",
      "// processing Scaffold352\n",
      "// processing Scaffold3522\n",
      "// processing Scaffold353\n",
      "// processing Scaffold35321\n",
      "// processing Scaffold35343\n",
      "// processing Scaffold35416\n",
      "// processing Scaffold3549\n",
      "// processing Scaffold355\n",
      "// processing Scaffold35541\n",
      "// processing Scaffold35582\n",
      "// processing Scaffold35626\n",
      "// processing Scaffold35645\n",
      "// processing Scaffold35646\n",
      "// processing Scaffold35690\n",
      "// processing Scaffold35787\n",
      "// processing Scaffold35829\n",
      "// processing Scaffold3584\n",
      "// processing Scaffold35863\n",
      "// processing Scaffold3587\n",
      "// processing Scaffold35956\n",
      "// processing Scaffold35984\n",
      "// processing Scaffold35989\n",
      "// processing Scaffold3599\n",
      "// processing Scaffold36\n",
      "// processing Scaffold3601\n",
      "// processing Scaffold3603\n",
      "// processing Scaffold36074\n",
      "// processing Scaffold36080\n",
      "// processing Scaffold361\n",
      "// processing Scaffold362\n",
      "// processing Scaffold36222\n",
      "// processing Scaffold36266\n",
      "// processing Scaffold36273\n",
      "// processing Scaffold363\n",
      "// processing Scaffold36409\n",
      "// processing Scaffold36446\n",
      "// processing Scaffold3654\n",
      "// processing Scaffold36593\n",
      "// processing Scaffold3662\n",
      "// processing Scaffold3669\n",
      "// processing Scaffold36732\n",
      "// processing Scaffold36738\n",
      "// processing Scaffold36797\n",
      "// processing Scaffold368\n",
      "// processing Scaffold36810\n",
      "// processing Scaffold36889\n",
      "// processing Scaffold369\n",
      "// processing Scaffold36904\n",
      "// processing Scaffold36906\n",
      "// processing Scaffold36949\n",
      "// processing Scaffold3695\n",
      "// processing Scaffold3699\n",
      "// processing Scaffold37\n",
      "// processing Scaffold37012\n",
      "// processing Scaffold37014\n",
      "// processing Scaffold3705\n",
      "// processing Scaffold37090\n",
      "// processing Scaffold371\n",
      "// processing Scaffold37132\n",
      "// processing Scaffold37153\n",
      "// processing Scaffold37191\n",
      "// processing Scaffold372\n",
      "// processing Scaffold37235\n",
      "// processing Scaffold37260\n",
      "// processing Scaffold3728\n",
      "// processing Scaffold373\n",
      "// processing Scaffold3735\n",
      "// processing Scaffold3739\n",
      "// processing Scaffold374\n",
      "// processing Scaffold3746\n",
      "// processing Scaffold375\n",
      "// processing Scaffold3751\n",
      "// processing Scaffold3752\n",
      "// processing Scaffold37581\n",
      "// processing Scaffold376\n",
      "// processing Scaffold3768\n",
      "// processing Scaffold377\n",
      "// processing Scaffold37799\n",
      "// processing Scaffold3780\n",
      "// processing Scaffold37823\n",
      "// processing Scaffold37871\n",
      "// processing Scaffold3789\n",
      "// processing Scaffold3791\n",
      "// processing Scaffold38\n",
      "// processing Scaffold38032\n",
      "// processing Scaffold38064\n",
      "// processing Scaffold38099\n",
      "// processing Scaffold381\n",
      "// processing Scaffold38197\n",
      "// processing Scaffold382\n",
      "// processing Scaffold383\n",
      "// processing Scaffold38356\n",
      "// processing Scaffold38407\n",
      "// processing Scaffold38434\n",
      "// processing Scaffold38465\n",
      "// processing Scaffold38489\n",
      "// processing Scaffold3851\n",
      "// processing Scaffold38548\n",
      "// processing Scaffold38675\n",
      "// processing Scaffold38679\n",
      "// processing Scaffold38681\n",
      "// processing Scaffold38720\n",
      "// processing Scaffold38785\n",
      "// processing Scaffold38796\n",
      "// processing Scaffold38836\n",
      "// processing Scaffold38988\n",
      "// processing Scaffold39\n",
      "// processing Scaffold390\n",
      "// processing Scaffold39040\n",
      "// processing Scaffold391\n",
      "// processing Scaffold39122\n",
      "// processing Scaffold3915\n",
      "// processing Scaffold39178\n",
      "// processing Scaffold392\n",
      "// processing Scaffold39352\n",
      "// processing Scaffold394\n",
      "// processing Scaffold395\n",
      "// processing Scaffold39526\n",
      "// processing Scaffold39656\n",
      "// processing Scaffold39671\n",
      "// processing Scaffold39726\n",
      "// processing Scaffold3973\n",
      "// processing Scaffold39816\n",
      "// processing Scaffold3985\n",
      "// processing Scaffold39890\n",
      "// processing Scaffold40\n",
      "// processing Scaffold40024\n",
      "// processing Scaffold40070\n",
      "// processing Scaffold402\n",
      "// processing Scaffold40247\n",
      "// processing Scaffold40271\n",
      "// processing Scaffold403\n",
      "// processing Scaffold4032\n",
      "// processing Scaffold40326\n",
      "// processing Scaffold4037\n",
      "// processing Scaffold40467\n",
      "// processing Scaffold4048\n",
      "// processing Scaffold4055\n",
      "// processing Scaffold40612\n",
      "// processing Scaffold40645\n",
      "// processing Scaffold40661\n",
      "// processing Scaffold40698\n",
      "// processing Scaffold4072\n",
      "// processing Scaffold4074\n",
      "// processing Scaffold4083\n",
      "// processing Scaffold4092\n",
      "// processing Scaffold40934\n",
      "// processing Scaffold40946\n",
      "// processing Scaffold40965\n",
      "// processing Scaffold40979\n",
      "// processing Scaffold40981\n",
      "// processing Scaffold41\n",
      "// processing Scaffold41019\n",
      "// processing Scaffold411\n",
      "// processing Scaffold41187\n",
      "// processing Scaffold4122\n",
      "// processing Scaffold41224\n",
      "// processing Scaffold4132\n",
      "// processing Scaffold4138\n",
      "// processing Scaffold414\n",
      "// processing Scaffold41720\n",
      "// processing Scaffold41851\n",
      "// processing Scaffold41860\n",
      "// processing Scaffold41872\n",
      "// processing Scaffold4188\n",
      "// processing Scaffold419\n",
      "// processing Scaffold4198\n",
      "// processing Scaffold42\n",
      "// processing Scaffold42058\n",
      "// processing Scaffold42095\n",
      "// processing Scaffold4218\n",
      "// processing Scaffold42207\n",
      "// processing Scaffold42224\n",
      "// processing Scaffold4225\n",
      "// processing Scaffold42375\n",
      "// processing Scaffold424\n",
      "// processing Scaffold4248\n",
      "// processing Scaffold425\n",
      "// processing Scaffold4250\n",
      "// processing Scaffold42511\n",
      "// processing Scaffold42545\n",
      "// processing Scaffold42631\n",
      "// processing Scaffold4265\n",
      "// processing Scaffold427\n",
      "// processing Scaffold42721\n",
      "// processing Scaffold42744\n",
      "// processing Scaffold4278\n",
      "// processing Scaffold42848\n",
      "// processing Scaffold42859\n",
      "// processing Scaffold42890\n",
      "// processing Scaffold42933\n",
      "// processing Scaffold42959\n",
      "// processing Scaffold43\n",
      "// processing Scaffold430\n",
      "// processing Scaffold4302\n",
      "// processing Scaffold43096\n",
      "// processing Scaffold43099\n",
      "// processing Scaffold431\n",
      "// processing Scaffold43129\n",
      "// processing Scaffold43138\n",
      "// processing Scaffold43225\n",
      "// processing Scaffold43237\n",
      "// processing Scaffold43252\n",
      "// processing Scaffold4329\n",
      "// processing Scaffold43397\n",
      "// processing Scaffold43430\n",
      "// processing Scaffold4345\n",
      "// processing Scaffold43481\n",
      "// processing Scaffold4350\n",
      "// processing Scaffold43510\n",
      "// processing Scaffold43546\n",
      "// processing Scaffold436\n",
      "// processing Scaffold4362\n",
      "// processing Scaffold43625\n",
      "// processing Scaffold43660\n",
      "// processing Scaffold4381\n",
      "// processing Scaffold4389\n",
      "// processing Scaffold4394\n",
      "// processing Scaffold43966\n",
      "// processing Scaffold44\n",
      "// processing Scaffold44065\n",
      "// processing Scaffold441\n",
      "// processing Scaffold44106\n",
      "// processing Scaffold4411\n",
      "// processing Scaffold44118\n",
      "// processing Scaffold44224\n",
      "// processing Scaffold44398\n",
      "// processing Scaffold44469\n",
      "// processing Scaffold4452\n",
      "// processing Scaffold4453\n",
      "// processing Scaffold44546\n",
      "// processing Scaffold44549\n",
      "// processing Scaffold44669\n",
      "// processing Scaffold4467\n",
      "// processing Scaffold44700\n",
      "// processing Scaffold4478\n",
      "// processing Scaffold44812\n",
      "// processing Scaffold449\n",
      "// processing Scaffold44955\n",
      "// processing Scaffold45\n",
      "// processing Scaffold450\n",
      "// processing Scaffold4507\n",
      "// processing Scaffold451\n",
      "// processing Scaffold45163\n",
      "// processing Scaffold45180\n",
      "// processing Scaffold452\n",
      "// processing Scaffold453\n",
      "// processing Scaffold45382\n",
      "// processing Scaffold45383\n",
      "// processing Scaffold454\n",
      "// processing Scaffold45438\n",
      "// processing Scaffold455\n",
      "// processing Scaffold4558\n",
      "// processing Scaffold45714\n",
      "// processing Scaffold45721\n",
      "// processing Scaffold45759\n",
      "// processing Scaffold45783\n",
      "// processing Scaffold45814\n",
      "// processing Scaffold45827\n",
      "// processing Scaffold45906\n",
      "// processing Scaffold4595\n",
      "// processing Scaffold45974\n",
      "// processing Scaffold46\n",
      "// processing Scaffold4600\n",
      "// processing Scaffold46009\n",
      "// processing Scaffold4601\n",
      "// processing Scaffold46092\n",
      "// processing Scaffold46115\n",
      "// processing Scaffold463\n",
      "// processing Scaffold46400\n",
      "// processing Scaffold4641\n",
      "// processing Scaffold46504\n",
      "// processing Scaffold46540\n",
      "// processing Scaffold46560\n",
      "// processing Scaffold46595\n",
      "// processing Scaffold46647\n",
      "// processing Scaffold46721\n",
      "// processing Scaffold46813\n",
      "// processing Scaffold46859\n",
      "// processing Scaffold4689\n",
      "// processing Scaffold46896\n",
      "// processing Scaffold4690\n",
      "// processing Scaffold46927\n",
      "// processing Scaffold46931\n",
      "// processing Scaffold47\n",
      "// processing Scaffold471\n",
      "// processing Scaffold47115\n",
      "// processing Scaffold472\n",
      "// processing Scaffold47308\n",
      "// processing Scaffold47320\n",
      "// processing Scaffold47351\n",
      "// processing Scaffold474\n",
      "// processing Scaffold47526\n",
      "// processing Scaffold47594\n",
      "// processing Scaffold47730\n",
      "// processing Scaffold478\n",
      "// processing Scaffold47853\n",
      "// processing Scaffold47911\n",
      "// processing Scaffold48\n",
      "// processing Scaffold48073\n",
      "// processing Scaffold4808\n",
      "// processing Scaffold48145\n",
      "// processing Scaffold48196\n",
      "// processing Scaffold482\n",
      "// processing Scaffold4820\n",
      "// processing Scaffold48201\n",
      "// processing Scaffold4822\n",
      "// processing Scaffold4829\n",
      "// processing Scaffold48412\n",
      "// processing Scaffold48451\n",
      "// processing Scaffold485\n",
      "// processing Scaffold48544\n",
      "// processing Scaffold48672\n",
      "// processing Scaffold48715\n",
      "// processing Scaffold488\n",
      "// processing Scaffold4888\n",
      "// processing Scaffold48974\n",
      "// processing Scaffold49\n",
      "// processing Scaffold49060\n",
      "// processing Scaffold49115\n",
      "// processing Scaffold49173\n",
      "// processing Scaffold492\n",
      "// processing Scaffold494\n",
      "// processing Scaffold49413\n",
      "// processing Scaffold49424\n",
      "// processing Scaffold49488\n",
      "// processing Scaffold49668\n",
      "// processing Scaffold497\n",
      "// processing Scaffold49765\n",
      "// processing Scaffold4977\n",
      "// processing Scaffold49882\n",
      "// processing Scaffold499\n",
      "// processing Scaffold4991\n",
      "// processing Scaffold49921\n",
      "// processing Scaffold4994\n",
      "// processing Scaffold49961\n",
      "// processing Scaffold50\n",
      "// processing Scaffold5000\n",
      "// processing Scaffold501\n",
      "// processing Scaffold50106\n",
      "// processing Scaffold50141\n",
      "// processing Scaffold50365\n",
      "// processing Scaffold50373\n",
      "// processing Scaffold505\n",
      "// processing Scaffold5050\n",
      "// processing Scaffold50506\n",
      "// processing Scaffold50660\n",
      "// processing Scaffold507\n",
      "// processing Scaffold50754\n",
      "// processing Scaffold50809\n",
      "// processing Scaffold5085\n",
      "// processing Scaffold509\n",
      "// processing Scaffold5090\n",
      "// processing Scaffold5098\n",
      "// processing Scaffold51\n",
      "// processing Scaffold51035\n",
      "// processing Scaffold51052\n",
      "// processing Scaffold5106\n",
      "// processing Scaffold51072\n",
      "// processing Scaffold5108\n",
      "// processing Scaffold512\n",
      "// processing Scaffold51274\n",
      "// processing Scaffold51286\n",
      "// processing Scaffold5141\n",
      "// processing Scaffold51452\n",
      "// processing Scaffold5151\n",
      "// processing Scaffold5156\n",
      "// processing Scaffold51595\n",
      "// processing Scaffold517\n",
      "// processing Scaffold5180\n",
      "// processing Scaffold51959\n",
      "// processing Scaffold52\n",
      "// processing Scaffold52030\n",
      "// processing Scaffold522\n",
      "// processing Scaffold5224\n",
      "// processing Scaffold52264\n",
      "// processing Scaffold52276\n",
      "// processing Scaffold5228\n",
      "// processing Scaffold52476\n",
      "// processing Scaffold52477\n",
      "// processing Scaffold5260\n",
      "// processing Scaffold52695\n",
      "// processing Scaffold529\n",
      "// processing Scaffold52927\n",
      "// processing Scaffold5296\n",
      "// processing Scaffold52986\n",
      "// processing Scaffold53\n",
      "// processing Scaffold53002\n",
      "// processing Scaffold53056\n",
      "// processing Scaffold53146\n",
      "// processing Scaffold53173\n",
      "// processing Scaffold532\n",
      "// processing Scaffold5323\n",
      "// processing Scaffold53233\n",
      "// processing Scaffold533\n",
      "// processing Scaffold53362\n",
      "// processing Scaffold53455\n",
      "// processing Scaffold53539\n",
      "// processing Scaffold53564\n",
      "// processing Scaffold536\n",
      "// processing Scaffold53706\n",
      "// processing Scaffold5379\n",
      "// processing Scaffold53865\n",
      "// processing Scaffold53899\n",
      "// processing Scaffold539\n",
      "// processing Scaffold53957\n",
      "// processing Scaffold54\n",
      "// processing Scaffold5422\n",
      "// processing Scaffold54296\n",
      "// processing Scaffold54343\n",
      "// processing Scaffold54429\n",
      "// processing Scaffold54476\n",
      "// processing Scaffold545\n",
      "// processing Scaffold548\n",
      "// processing Scaffold54836\n",
      "// processing Scaffold55\n",
      "// processing Scaffold550\n",
      "// processing Scaffold5501\n",
      "// processing Scaffold55190\n",
      "// processing Scaffold55205\n",
      "// processing Scaffold5523\n",
      "// processing Scaffold553\n",
      "// processing Scaffold55435\n",
      "// processing Scaffold555\n",
      "// processing Scaffold55668\n",
      "// processing Scaffold55682\n",
      "// processing Scaffold557\n",
      "// processing Scaffold55773\n",
      "// processing Scaffold55793\n",
      "// processing Scaffold5582\n",
      "// processing Scaffold5584\n",
      "// processing Scaffold559\n",
      "// processing Scaffold55914\n",
      "// processing Scaffold5593\n",
      "// processing Scaffold56\n",
      "// processing Scaffold5603\n",
      "// processing Scaffold562\n",
      "// processing Scaffold5620\n",
      "// processing Scaffold56236\n",
      "// processing Scaffold5626\n",
      "// processing Scaffold5635\n",
      "// processing Scaffold5636\n",
      "// processing Scaffold5642\n",
      "// processing Scaffold565\n",
      "// processing Scaffold56649\n",
      "// processing Scaffold56700\n",
      "// processing Scaffold568\n",
      "// processing Scaffold56891\n",
      "// processing Scaffold56999\n",
      "// processing Scaffold57\n",
      "// processing Scaffold57012\n",
      "// processing Scaffold57034\n",
      "// processing Scaffold5710\n",
      "// processing Scaffold57112\n",
      "// processing Scaffold5714\n",
      "// processing Scaffold5721\n",
      "// processing Scaffold57267\n",
      "// processing Scaffold57429\n",
      "// processing Scaffold57449\n",
      "// processing Scaffold575\n",
      "// processing Scaffold57516\n",
      "// processing Scaffold57562\n",
      "// processing Scaffold57657\n",
      "// processing Scaffold577\n",
      "// processing Scaffold57772\n",
      "// processing Scaffold57813\n",
      "// processing Scaffold57823\n",
      "// processing Scaffold579\n",
      "// processing Scaffold57932\n",
      "// processing Scaffold58\n",
      "// processing Scaffold58077\n",
      "// processing Scaffold581\n",
      "// processing Scaffold58129\n",
      "// processing Scaffold58131\n",
      "// processing Scaffold58544\n",
      "// processing Scaffold5858\n",
      "// processing Scaffold58703\n",
      "// processing Scaffold58722\n",
      "// processing Scaffold58746\n",
      "// processing Scaffold58824\n",
      "// processing Scaffold5884\n",
      "// processing Scaffold5887\n",
      "// processing Scaffold58976\n",
      "// processing Scaffold59\n",
      "// processing Scaffold59021\n",
      "// processing Scaffold59028\n",
      "// processing Scaffold5907\n",
      "// processing Scaffold5916\n",
      "// processing Scaffold5920\n",
      "// processing Scaffold5923\n",
      "// processing Scaffold593\n",
      "// processing Scaffold59438\n",
      "// processing Scaffold59544\n",
      "// processing Scaffold59567\n",
      "// processing Scaffold596\n",
      "// processing Scaffold59730\n",
      "// processing Scaffold59738\n",
      "// processing Scaffold59808\n",
      "// processing Scaffold599\n",
      "// processing Scaffold59968\n",
      "// processing Scaffold60\n",
      "// processing Scaffold60026\n",
      "// processing Scaffold602\n",
      "// processing Scaffold60253\n",
      "// processing Scaffold603\n",
      "// processing Scaffold6031\n",
      "// processing Scaffold60321\n",
      "// processing Scaffold6040\n",
      "// processing Scaffold60404\n",
      "// processing Scaffold6041\n",
      "// processing Scaffold60418\n",
      "// processing Scaffold60467\n",
      "// processing Scaffold60515\n",
      "// processing Scaffold60526\n",
      "// processing Scaffold6053\n",
      "// processing Scaffold6058\n",
      "// processing Scaffold6061\n",
      "// processing Scaffold60684\n",
      "// processing Scaffold60685\n",
      "// processing Scaffold60886\n",
      "// processing Scaffold60916\n",
      "// processing Scaffold61\n",
      "// processing Scaffold61013\n",
      "// processing Scaffold61032\n",
      "// processing Scaffold61041\n",
      "// processing Scaffold61113\n",
      "// processing Scaffold61152\n",
      "// processing Scaffold61186\n",
      "// processing Scaffold61294\n",
      "// processing Scaffold61298\n",
      "// processing Scaffold61339\n",
      "// processing Scaffold617\n",
      "// processing Scaffold61829\n",
      "// processing Scaffold61840\n",
      "// processing Scaffold6187\n",
      "// processing Scaffold61988\n",
      "// processing Scaffold61989\n",
      "// processing Scaffold62\n",
      "// processing Scaffold6206\n",
      "// processing Scaffold62071\n",
      "// processing Scaffold62093\n",
      "// processing Scaffold62158\n",
      "// processing Scaffold622\n",
      "// processing Scaffold6223\n",
      "// processing Scaffold62383\n",
      "// processing Scaffold6241\n",
      "// processing Scaffold62428\n",
      "// processing Scaffold6247\n",
      "// processing Scaffold62495\n",
      "// processing Scaffold6256\n",
      "// processing Scaffold62611\n",
      "// processing Scaffold627\n",
      "// processing Scaffold6285\n",
      "// processing Scaffold62896\n",
      "// processing Scaffold62941\n",
      "// processing Scaffold62946\n",
      "// processing Scaffold62951\n",
      "// processing Scaffold63\n",
      "// processing Scaffold63262\n",
      "// processing Scaffold63302\n",
      "// processing Scaffold63420\n",
      "// processing Scaffold6347\n",
      "// processing Scaffold63521\n",
      "// processing Scaffold63588\n",
      "// processing Scaffold63686\n",
      "// processing Scaffold63783\n",
      "// processing Scaffold64\n",
      "// processing Scaffold640\n",
      "// processing Scaffold64020\n",
      "// processing Scaffold64024\n",
      "// processing Scaffold64067\n",
      "// processing Scaffold641\n",
      "// processing Scaffold6412\n",
      "// processing Scaffold64126\n",
      "// processing Scaffold64156\n",
      "// processing Scaffold64157\n",
      "// processing Scaffold64395\n",
      "// processing Scaffold64514\n",
      "// processing Scaffold64563\n",
      "// processing Scaffold64633\n",
      "// processing Scaffold6464\n",
      "// processing Scaffold64670\n",
      "// processing Scaffold6475\n",
      "// processing Scaffold64791\n",
      "// processing Scaffold6493\n",
      "// processing Scaffold65\n",
      "// processing Scaffold65113\n",
      "// processing Scaffold65126\n",
      "// processing Scaffold65227\n",
      "// processing Scaffold65302\n",
      "// processing Scaffold65428\n",
      "// processing Scaffold65454\n",
      "// processing Scaffold6548\n",
      "// processing Scaffold65537\n",
      "// processing Scaffold65918\n",
      "// processing Scaffold6594\n",
      "// processing Scaffold65985\n",
      "// processing Scaffold66\n",
      "// processing Scaffold660\n",
      "// processing Scaffold66062\n",
      "// processing Scaffold6615\n",
      "// processing Scaffold663\n",
      "// processing Scaffold66432\n",
      "// processing Scaffold66554\n",
      "// processing Scaffold667\n",
      "// processing Scaffold66752\n",
      "// processing Scaffold6681\n",
      "// processing Scaffold66811\n",
      "// processing Scaffold66902\n",
      "// processing Scaffold6698\n",
      "// processing Scaffold67\n",
      "// processing Scaffold6707\n",
      "// processing Scaffold67133\n",
      "// processing Scaffold6717\n",
      "// processing Scaffold67190\n",
      "// processing Scaffold67434\n",
      "// processing Scaffold6750\n",
      "// processing Scaffold67519\n",
      "// processing Scaffold67595\n",
      "// processing Scaffold6762\n",
      "// processing Scaffold6763\n",
      "// processing Scaffold6775\n",
      "// processing Scaffold67830\n",
      "// processing Scaffold67865\n",
      "// processing Scaffold679\n",
      "// processing Scaffold6792\n",
      "// processing Scaffold67955\n",
      "// processing Scaffold67958\n",
      "// processing Scaffold68\n",
      "// processing Scaffold680\n",
      "// processing Scaffold6844\n",
      "// processing Scaffold6861\n",
      "// processing Scaffold68754\n",
      "// processing Scaffold688\n",
      "// processing Scaffold68816\n",
      "// processing Scaffold68859\n",
      "// processing Scaffold6886\n",
      "// processing Scaffold68895\n",
      "// processing Scaffold68950\n",
      "// processing Scaffold69\n",
      "// processing Scaffold69001\n",
      "// processing Scaffold69006\n",
      "// processing Scaffold69049\n",
      "// processing Scaffold6924\n",
      "// processing Scaffold69301\n",
      "// processing Scaffold69347\n",
      "// processing Scaffold69353\n",
      "// processing Scaffold69397\n",
      "// processing Scaffold69448\n",
      "// processing Scaffold695\n",
      "// processing Scaffold6951\n",
      "// processing Scaffold69566\n",
      "// processing Scaffold69680\n",
      "// processing Scaffold69699\n",
      "// processing Scaffold697\n",
      "// processing Scaffold6974\n",
      "// processing Scaffold6978\n",
      "// processing Scaffold699\n",
      "// processing Scaffold70\n",
      "// processing Scaffold7004\n",
      "// processing Scaffold70270\n",
      "// processing Scaffold70531\n",
      "// processing Scaffold70566\n",
      "// processing Scaffold70606\n",
      "// processing Scaffold70631\n",
      "// processing Scaffold70718\n",
      "// processing Scaffold709\n",
      "// processing Scaffold70944\n",
      "// processing Scaffold70969\n",
      "// processing Scaffold71\n",
      "// processing Scaffold71102\n",
      "// processing Scaffold7115\n",
      "// processing Scaffold71350\n",
      "// processing Scaffold71831\n",
      "// processing Scaffold71929\n",
      "// processing Scaffold7197\n",
      "// processing Scaffold72\n",
      "// processing Scaffold72079\n",
      "// processing Scaffold72304\n",
      "// processing Scaffold72381\n",
      "// processing Scaffold7241\n",
      "// processing Scaffold7247\n",
      "// processing Scaffold72624\n",
      "// processing Scaffold72886\n",
      "// processing Scaffold729\n",
      "// processing Scaffold73\n",
      "// processing Scaffold73036\n",
      "// processing Scaffold73114\n",
      "// processing Scaffold7315\n",
      "// processing Scaffold73153\n",
      "// processing Scaffold73250\n",
      "// processing Scaffold73368\n",
      "// processing Scaffold73386\n",
      "// processing Scaffold73430\n",
      "// processing Scaffold73524\n",
      "// processing Scaffold73647\n",
      "// processing Scaffold737\n",
      "// processing Scaffold73714\n",
      "// processing Scaffold7387\n",
      "// processing Scaffold74\n",
      "// processing Scaffold74043\n",
      "// processing Scaffold7410\n",
      "// processing Scaffold74256\n",
      "// processing Scaffold7458\n",
      "// processing Scaffold74604\n",
      "// processing Scaffold7463\n",
      "// processing Scaffold74705\n",
      "// processing Scaffold7471\n",
      "// processing Scaffold75\n",
      "// processing Scaffold75197\n",
      "// processing Scaffold753\n",
      "// processing Scaffold75348\n",
      "// processing Scaffold7552\n",
      "// processing Scaffold75526\n",
      "// processing Scaffold75635\n",
      "// processing Scaffold75744\n",
      "// processing Scaffold75746\n",
      "// processing Scaffold7581\n",
      "// processing Scaffold75926\n",
      "// processing Scaffold76\n",
      "// processing Scaffold760\n",
      "// processing Scaffold7603\n",
      "// processing Scaffold76052\n",
      "// processing Scaffold76108\n",
      "// processing Scaffold7616\n",
      "// processing Scaffold76173\n",
      "// processing Scaffold7619\n",
      "// processing Scaffold7624\n",
      "// processing Scaffold76258\n",
      "// processing Scaffold76261\n",
      "// processing Scaffold76277\n",
      "// processing Scaffold763\n",
      "// processing Scaffold7635\n",
      "// processing Scaffold7636\n",
      "// processing Scaffold764\n",
      "// processing Scaffold76428\n",
      "// processing Scaffold7654\n",
      "// processing Scaffold76703\n",
      "// processing Scaffold7693\n",
      "// processing Scaffold77\n",
      "// processing Scaffold7707\n",
      "// processing Scaffold77347\n",
      "// processing Scaffold7736\n",
      "// processing Scaffold774\n",
      "// processing Scaffold77471\n",
      "// processing Scaffold775\n",
      "// processing Scaffold7754\n",
      "// processing Scaffold77555\n",
      "// processing Scaffold7760\n",
      "// processing Scaffold7763\n",
      "// processing Scaffold7764\n",
      "// processing Scaffold77891\n",
      "// processing Scaffold7793\n",
      "// processing Scaffold78\n",
      "// processing Scaffold7806\n",
      "// processing Scaffold78090\n",
      "// processing Scaffold7818\n",
      "// processing Scaffold782\n",
      "// processing Scaffold784\n",
      "// processing Scaffold78493\n",
      "// processing Scaffold7876\n",
      "// processing Scaffold78789\n",
      "// processing Scaffold78932\n",
      "// processing Scaffold79\n",
      "// processing Scaffold7904\n",
      "// processing Scaffold79263\n",
      "// processing Scaffold79443\n",
      "// processing Scaffold7945\n",
      "// processing Scaffold795\n",
      "// processing Scaffold7953\n",
      "// processing Scaffold79535\n",
      "// processing Scaffold79599\n",
      "// processing Scaffold7968\n",
      "// processing Scaffold79694\n",
      "// processing Scaffold79768\n",
      "// processing Scaffold80\n",
      "// processing Scaffold80179\n",
      "// processing Scaffold805\n",
      "// processing Scaffold80511\n",
      "// processing Scaffold80815\n",
      "// processing Scaffold80831\n",
      "// processing Scaffold809\n",
      "// processing Scaffold80909\n",
      "// processing Scaffold8093\n",
      "// processing Scaffold80961\n",
      "// processing Scaffold81\n",
      "// processing Scaffold810\n",
      "// processing Scaffold81082\n",
      "// processing Scaffold81182\n",
      "// processing Scaffold8121\n",
      "// processing Scaffold8130\n",
      "// processing Scaffold81358\n",
      "// processing Scaffold8163\n",
      "// processing Scaffold817\n",
      "// processing Scaffold81783\n",
      "// processing Scaffold818\n",
      "// processing Scaffold8187\n",
      "// processing Scaffold8192\n",
      "// processing Scaffold82\n",
      "// processing Scaffold8200\n",
      "// processing Scaffold82004\n",
      "// processing Scaffold822\n",
      "// processing Scaffold82235\n",
      "// processing Scaffold8226\n",
      "// processing Scaffold823\n",
      "// processing Scaffold82413\n",
      "// processing Scaffold8251\n",
      "// processing Scaffold8264\n",
      "// processing Scaffold82650\n",
      "// processing Scaffold82697\n",
      "// processing Scaffold8274\n",
      "// processing Scaffold82744\n",
      "// processing Scaffold8276\n",
      "// processing Scaffold8285\n",
      "// processing Scaffold82861\n",
      "// processing Scaffold82927\n",
      "// processing Scaffold83\n",
      "// processing Scaffold83062\n",
      "// processing Scaffold833\n",
      "// processing Scaffold83341\n",
      "// processing Scaffold83531\n",
      "// processing Scaffold83572\n",
      "// processing Scaffold8386\n",
      "// processing Scaffold83862\n",
      "// processing Scaffold8387\n",
      "// processing Scaffold8390\n",
      "// processing Scaffold83924\n",
      "// processing Scaffold8393\n",
      "// processing Scaffold8395\n",
      "// processing Scaffold84\n",
      "// processing Scaffold84152\n",
      "// processing Scaffold8432\n",
      "// processing Scaffold845\n",
      "// processing Scaffold8468\n",
      "// processing Scaffold84776\n",
      "// processing Scaffold8486\n",
      "// processing Scaffold8487\n",
      "// processing Scaffold84918\n",
      "// processing Scaffold8497\n",
      "// processing Scaffold85\n",
      "// processing Scaffold850\n",
      "// processing Scaffold85039\n",
      "// processing Scaffold8507\n",
      "// processing Scaffold85122\n",
      "// processing Scaffold8531\n",
      "// processing Scaffold8536\n",
      "// processing Scaffold854\n",
      "// processing Scaffold8543\n",
      "// processing Scaffold8568\n",
      "// processing Scaffold8590\n",
      "// processing Scaffold85921\n",
      "// processing Scaffold86\n",
      "// processing Scaffold86354\n",
      "// processing Scaffold86598\n",
      "// processing Scaffold8660\n",
      "// processing Scaffold867\n",
      "// processing Scaffold8673\n",
      "// processing Scaffold8693\n",
      "// processing Scaffold86989\n",
      "// processing Scaffold87\n",
      "// processing Scaffold87108\n",
      "// processing Scaffold8720\n",
      "// processing Scaffold8730\n",
      "// processing Scaffold87434\n",
      "// processing Scaffold8766\n",
      "// processing Scaffold87874\n",
      "// processing Scaffold8793\n",
      "// processing Scaffold87996\n",
      "// processing Scaffold88\n",
      "// processing Scaffold881\n",
      "// processing Scaffold8825\n",
      "// processing Scaffold88322\n",
      "// processing Scaffold886\n",
      "// processing Scaffold887\n",
      "// processing Scaffold8878\n",
      "// processing Scaffold8897\n",
      "// processing Scaffold89\n",
      "// processing Scaffold89048\n",
      "// processing Scaffold8927\n",
      "// processing Scaffold89480\n",
      "// processing Scaffold895\n",
      "// processing Scaffold8956\n",
      "// processing Scaffold89596\n",
      "// processing Scaffold89948\n",
      "// processing Scaffold90\n",
      "// processing Scaffold9041\n",
      "// processing Scaffold90563\n",
      "// processing Scaffold90756\n",
      "// processing Scaffold91\n",
      "// processing Scaffold91072\n",
      "// processing Scaffold9115\n",
      "// processing Scaffold9123\n",
      "// processing Scaffold9165\n",
      "// processing Scaffold91864\n",
      "// processing Scaffold9189\n",
      "// processing Scaffold9193\n",
      "// processing Scaffold92\n",
      "// processing Scaffold92007\n",
      "// processing Scaffold92090\n",
      "// processing Scaffold9211\n",
      "// processing Scaffold922\n",
      "// processing Scaffold9232\n",
      "// processing Scaffold926\n",
      "// processing Scaffold93\n",
      "// processing Scaffold930\n",
      "// processing Scaffold934\n",
      "// processing Scaffold9363\n",
      "// processing Scaffold93630\n",
      "// processing Scaffold93892\n",
      "// processing Scaffold939\n",
      "// processing Scaffold94\n",
      "// processing Scaffold9403\n",
      "// processing Scaffold94051\n",
      "// processing Scaffold9414\n",
      "// processing Scaffold942\n",
      "// processing Scaffold94400\n",
      "// processing Scaffold94444\n",
      "// processing Scaffold945\n",
      "// processing Scaffold94521\n",
      "// processing Scaffold94614\n",
      "// processing Scaffold9468\n",
      "// processing Scaffold94698\n",
      "// processing Scaffold9494\n",
      "// processing Scaffold94976\n",
      "// processing Scaffold95\n",
      "// processing Scaffold950\n",
      "// processing Scaffold9503\n",
      "// processing Scaffold9508\n",
      "// processing Scaffold95291\n",
      "// processing Scaffold953\n",
      "// processing Scaffold9558\n",
      "// processing Scaffold9591\n",
      "// processing Scaffold9592\n",
      "// processing Scaffold9597\n",
      "// processing Scaffold9598\n",
      "// processing Scaffold96\n",
      "// processing Scaffold9635\n",
      "// processing Scaffold96718\n",
      "// processing Scaffold97\n",
      "// processing Scaffold9729\n",
      "// processing Scaffold9758\n",
      "// processing Scaffold9761\n",
      "// processing Scaffold979\n",
      "// processing Scaffold98\n",
      "// processing Scaffold980\n",
      "// processing Scaffold98178\n",
      "// processing Scaffold983\n",
      "// processing Scaffold9837\n",
      "// processing Scaffold99\n",
      "// processing Scaffold992\n",
      "// processing Scaffold99228\n",
      "// processing Scaffold9939\n",
      "// processing Scaffold99458\n",
      "// processing Scaffold9949\n",
      "// processing Scaffold9972\n",
      "// processing Scaffold9978\n",
      "// processing chr1L\n",
      "// processing chr1S\n",
      "// processing chr2L\n",
      "// processing chr2S\n",
      "// processing chr3L\n",
      "// processing chr3S\n",
      "// processing chr4L\n",
      "// processing chr4S\n",
      "// processing chr5L\n",
      "// processing chr5S\n",
      "// processing chr6L\n",
      "// processing chr6S\n",
      "// processing chr7L\n",
      "// processing chr7S\n",
      "// processing chr8L\n",
      "// processing chr8S\n",
      "// processing chr9_10L\n",
      "// processing chr9_10S\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/compute_base_probs.pl ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa 0 > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//base_freqs.dat\n",
      "\n",
      "\n",
      "-first extracting base frequencies, we'll need them later.\n",
      "\n",
      "\n",
      "- extracting ORFs from transcripts.\n",
      "-total transcripts to examine: 88005\n",
      "[6100/88005] = 6.93% done    Sequence must code for at least a codon. Your seq_length is too short\n",
      "[18900/88005] = 21.48% done    Sequence must code for at least a codon. Your seq_length is too short\n",
      "[21700/88005] = 24.66% done    Sequence must code for at least a codon. Your seq_length is too short\n",
      "[44100/88005] = 50.11% done    Sequence must code for at least a codon. Your seq_length is too short\n",
      "[55700/88005] = 63.29% done    Sequence must code for at least a codon. Your seq_length is too short\n",
      "[58900/88005] = 66.93% done    Sequence must code for at least a codon. Your seq_length is too short\n",
      "[77000/88005] = 87.50% done    Sequence must code for at least a codon. Your seq_length is too short\n",
      "[85600/88005] = 97.27% done    Sequence must code for at least a codon. Your seq_length is too short\n",
      "[88000/88005] = 99.99% done    \n",
      "\n",
      "#################################\n",
      "### Done preparing long ORFs.  ###\n",
      "##################################\n",
      "\n",
      "\tUse file: ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.pep  for Pfam and/or BlastP searches to enable homology-based coding region identification.\n",
      "\n",
      "\tThen, run TransDecoder.Predict for your final coding region predictions.\n",
      "\n",
      "\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds 5000 5000 > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CMD: touch ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir/.__checkpoints_longorfs/TD.longorfs.ok\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/exclude_similar_proteins.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000 > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr\n",
      "-skipping training candidate: rna75476.p1, not unique enough\n",
      "-skipping training candidate: rna86172.p2, not unique enough\n",
      "-skipping training candidate: rna74302.p1, not unique enough\n",
      "-skipping training candidate: rna19792.p1, not unique enough\n",
      "-skipping training candidate: rna46980.p1, not unique enough\n",
      "-skipping training candidate: rna88326.p1, not unique enough\n",
      "-skipping training candidate: rna36869.p1, not unique enough\n",
      "-skipping training candidate: rna83367.p2, not unique enough\n",
      "-skipping training candidate: rna91229.p2, not unique enough\n",
      "-skipping training candidate: rna89026.p2, not unique enough\n",
      "-skipping training candidate: rna37880.p1, not unique enough\n",
      "-skipping training candidate: rna37881.p2, not unique enough\n",
      "-skipping training candidate: rna82052.p1, not unique enough\n",
      "-skipping training candidate: rna77444.p2, not unique enough\n",
      "-skipping training candidate: rna94846.p1, not unique enough\n",
      "-skipping training candidate: rna79825.p2, not unique enough\n",
      "-skipping training candidate: rna94846.p2, not unique enough\n",
      "-skipping training candidate: rna93484.p1, not unique enough\n",
      "-skipping training candidate: rna78862.p1, not unique enough\n",
      "-skipping training candidate: rna80496.p1, not unique enough\n",
      "-skipping training candidate: rna97263.p1, not unique enough\n",
      "-skipping training candidate: rna78862.p2, not unique enough\n",
      "-skipping training candidate: rna47745.p1, not unique enough\n",
      "-skipping training candidate: rna68243.p2, not unique enough\n",
      "-skipping training candidate: rna78640.p2, not unique enough\n",
      "-skipping training candidate: rna80496.p2, not unique enough\n",
      "-skipping training candidate: rna89567.p2, not unique enough\n",
      "-skipping training candidate: rna24428.p1, not unique enough\n",
      "-skipping training candidate: rna93595.p2, not unique enough\n",
      "-skipping training candidate: rna47745.p2, not unique enough\n",
      "-skipping training candidate: rna74027.p1, not unique enough\n",
      "-skipping training candidate: rna91277.p2, not unique enough\n",
      "-skipping training candidate: rna94842.p2, not unique enough\n",
      "-skipping training candidate: rna87194.p2, not unique enough\n",
      "-skipping training candidate: rna11301.p1, not unique enough\n",
      "-skipping training candidate: rna49181.p2, not unique enough\n",
      "-skipping training candidate: rna95487.p1, not unique enough\n",
      "-skipping training candidate: rna85075.p1, not unique enough\n",
      "-skipping training candidate: rna94960.p1, not unique enough\n",
      "-skipping training candidate: rna14392.p1, not unique enough\n",
      "-skipping training candidate: rna76866.p1, not unique enough\n",
      "-skipping training candidate: rna75698.p2, not unique enough\n",
      "-skipping training candidate: rna87555.p1, not unique enough\n",
      "-skipping training candidate: rna34868.p1, not unique enough\n",
      "-skipping training candidate: rna88765.p1, not unique enough\n",
      "-skipping training candidate: rna40446.p1, not unique enough\n",
      "-skipping training candidate: rna82105.p1, not unique enough\n",
      "-skipping training candidate: rna8469.p1, not unique enough\n",
      "-skipping training candidate: rna96477.p2, not unique enough\n",
      "-skipping training candidate: rna18315.p1, not unique enough\n",
      "-skipping training candidate: rna76364.p1, not unique enough\n",
      "-skipping training candidate: rna86146.p1, not unique enough\n",
      "-skipping training candidate: rna91184.p1, not unique enough\n",
      "-skipping training candidate: rna82255.p1, not unique enough\n",
      "-skipping training candidate: rna22282.p1, not unique enough\n",
      "-skipping training candidate: rna3655.p1, not unique enough\n",
      "-skipping training candidate: rna32289.p2, not unique enough\n",
      "-skipping training candidate: rna80392.p1, not unique enough\n",
      "-skipping training candidate: rna21709.p1, not unique enough\n",
      "-skipping training candidate: rna45822.p1, not unique enough\n",
      "-skipping training candidate: rna23013.p2, not unique enough\n",
      "-skipping training candidate: rna88541.p1, not unique enough\n",
      "-skipping training candidate: rna95749.p1, not unique enough\n",
      "-skipping training candidate: rna47257.p1, not unique enough\n",
      "-skipping training candidate: rna35649.p2, not unique enough\n",
      "-skipping training candidate: rna83905.p2, not unique enough\n",
      "-skipping training candidate: rna35102.p1, not unique enough\n",
      "-skipping training candidate: rna12044.p1, not unique enough\n",
      "-skipping training candidate: rna78161.p1, not unique enough\n",
      "-skipping training candidate: rna15153.p1, not unique enough\n",
      "-skipping training candidate: rna74913.p1, not unique enough\n",
      "-skipping training candidate: rna83226.p1, not unique enough\n",
      "-skipping training candidate: rna93385.p1, not unique enough\n",
      "-skipping training candidate: rna6521.p1, not unique enough\n",
      "-skipping training candidate: rna96499.p1, not unique enough\n",
      "-skipping training candidate: rna95343.p1, not unique enough\n",
      "-skipping training candidate: rna46253.p1, not unique enough\n",
      "-skipping training candidate: rna91203.p2, not unique enough\n",
      "-skipping training candidate: rna41572.p1, not unique enough\n",
      "-skipping training candidate: rna80845.p1, not unique enough\n",
      "-skipping training candidate: rna17833.p1, not unique enough\n",
      "-skipping training candidate: rna89155.p1, not unique enough\n",
      "-skipping training candidate: rna15188.p2, not unique enough\n",
      "-skipping training candidate: rna6795.p1, not unique enough\n",
      "-skipping training candidate: rna16012.p1, not unique enough\n",
      "-skipping training candidate: rna83047.p2, not unique enough\n",
      "-skipping training candidate: rna92393.p1, not unique enough\n",
      "-skipping training candidate: rna51588.p1, not unique enough\n",
      "-skipping training candidate: rna76597.p2, not unique enough\n",
      "-skipping training candidate: rna17248.p2, not unique enough\n",
      "-skipping training candidate: rna24717.p1, not unique enough\n",
      "-skipping training candidate: rna83321.p1, not unique enough\n",
      "-skipping training candidate: rna93553.p1, not unique enough\n",
      "-skipping training candidate: rna48366.p1, not unique enough\n",
      "-skipping training candidate: rna32283.p1, not unique enough\n",
      "-skipping training candidate: rna21783.p1, not unique enough\n",
      "-skipping training candidate: rna53955.p1, not unique enough\n",
      "-skipping training candidate: rna50135.p1, not unique enough\n",
      "-skipping training candidate: rna50137.p1, not unique enough\n",
      "-skipping training candidate: rna86908.p2, not unique enough\n",
      "-skipping training candidate: rna2901.p1, not unique enough\n",
      "-skipping training candidate: rna2901.p2, not unique enough\n",
      "-skipping training candidate: rna97372.p1, not unique enough\n",
      "-skipping training candidate: rna46918.p1, not unique enough\n",
      "-skipping training candidate: rna34549.p1, not unique enough\n",
      "-skipping training candidate: rna4755.p1, not unique enough\n",
      "-skipping training candidate: rna46779.p1, not unique enough\n",
      "-skipping training candidate: rna91031.p1, not unique enough\n",
      "-skipping training candidate: rna83447.p1, not unique enough\n",
      "-skipping training candidate: rna40811.p1, not unique enough\n",
      "-skipping training candidate: rna76101.p1, not unique enough\n",
      "-skipping training candidate: rna94676.p1, not unique enough\n",
      "-skipping training candidate: rna15924.p1, not unique enough\n",
      "-skipping training candidate: rna9240.p1, not unique enough\n",
      "-skipping training candidate: rna9240.p2, not unique enough\n",
      "-skipping training candidate: rna39211.p1, not unique enough\n",
      "-skipping training candidate: rna51768.p2, not unique enough\n",
      "-skipping training candidate: rna84114.p2, not unique enough\n",
      "-skipping training candidate: rna85855.p1, not unique enough\n",
      "-skipping training candidate: rna85695.p1, not unique enough\n",
      "-skipping training candidate: rna12069.p1, not unique enough\n",
      "-skipping training candidate: rna80415.p1, not unique enough\n",
      "-skipping training candidate: rna40950.p1, not unique enough\n",
      "-skipping training candidate: rna48939.p2, not unique enough\n",
      "-skipping training candidate: rna48934.p1, not unique enough\n",
      "-skipping training candidate: rna45523.p1, not unique enough\n",
      "-skipping training candidate: rna91713.p2, not unique enough\n",
      "-skipping training candidate: rna76745.p2, not unique enough\n",
      "-skipping training candidate: rna15546.p1, not unique enough\n",
      "-skipping training candidate: rna6117.p1, not unique enough\n",
      "-skipping training candidate: rna81593.p1, not unique enough\n",
      "-skipping training candidate: rna77403.p1, not unique enough\n",
      "-skipping training candidate: rna13057.p1, not unique enough\n",
      "-skipping training candidate: rna42534.p1, not unique enough\n",
      "-skipping training candidate: gene7673_t.p1, not unique enough\n",
      "-skipping training candidate: rna38967.p1, not unique enough\n",
      "-skipping training candidate: rna86481.p1, not unique enough\n",
      "-skipping training candidate: rna56254.p2, not unique enough\n",
      "-skipping training candidate: rna88805.p1, not unique enough\n",
      "-skipping training candidate: rna88805.p2, not unique enough\n",
      "-skipping training candidate: rna46421.p1, not unique enough\n",
      "-skipping training candidate: rna76246.p1, not unique enough\n",
      "-skipping training candidate: rna35269.p1, not unique enough\n",
      "-skipping training candidate: rna4571.p1, not unique enough\n",
      "-skipping training candidate: rna80020.p1, not unique enough\n",
      "-skipping training candidate: rna46303.p1, not unique enough\n",
      "-skipping training candidate: rna78531.p1, not unique enough\n",
      "-skipping training candidate: rna90440.p1, not unique enough\n",
      "-skipping training candidate: rna12766.p1, not unique enough\n",
      "-skipping training candidate: rna48729.p1, not unique enough\n",
      "-skipping training candidate: rna3353.p1, not unique enough\n",
      "-skipping training candidate: rna95398.p1, not unique enough\n",
      "-skipping training candidate: rna29324.p1, not unique enough\n",
      "-skipping training candidate: rna44389.p1, not unique enough\n",
      "-skipping training candidate: rna96283.p2, not unique enough\n",
      "-skipping training candidate: rna53850.p1, not unique enough\n",
      "-skipping training candidate: rna8568.p1, not unique enough\n",
      "-skipping training candidate: rna8653.p1, not unique enough\n",
      "-skipping training candidate: rna47394.p1, not unique enough\n",
      "-skipping training candidate: rna4776.p1, not unique enough\n",
      "-skipping training candidate: rna90938.p1, not unique enough\n",
      "-skipping training candidate: rna871.p1, not unique enough\n",
      "-skipping training candidate: rna32130.p1, not unique enough\n",
      "-skipping training candidate: rna89283.p1, not unique enough\n",
      "-skipping training candidate: rna52851.p1, not unique enough\n",
      "-skipping training candidate: rna74880.p1, not unique enough\n",
      "-skipping training candidate: rna68591.p1, not unique enough\n",
      "-skipping training candidate: rna11162.p1, not unique enough\n",
      "-skipping training candidate: rna76206.p1, not unique enough\n",
      "-skipping training candidate: rna76332.p1, not unique enough\n",
      "-skipping training candidate: rna42179.p1, not unique enough\n",
      "-skipping training candidate: rna91726.p1, not unique enough\n",
      "-skipping training candidate: rna13394.p1, not unique enough\n",
      "-skipping training candidate: rna78706.p1, not unique enough\n",
      "-skipping training candidate: rna38286.p2, not unique enough\n",
      "-skipping training candidate: rna86926.p1, not unique enough\n",
      "-skipping training candidate: rna54226.p1, not unique enough\n",
      "-skipping training candidate: rna54227.p2, not unique enough\n",
      "-skipping training candidate: rna36170.p2, not unique enough\n",
      "-skipping training candidate: rna36168.p1, not unique enough\n",
      "-skipping training candidate: rna53849.p2, not unique enough\n",
      "-skipping training candidate: rna32978.p2, not unique enough\n",
      "-skipping training candidate: rna32975.p1, not unique enough\n",
      "-skipping training candidate: rna78855.p1, not unique enough\n",
      "-skipping training candidate: rna47743.p1, not unique enough\n",
      "-skipping training candidate: rna33808.p1, not unique enough\n",
      "-skipping training candidate: rna94188.p1, not unique enough\n",
      "-skipping training candidate: rna74563.p1, not unique enough\n",
      "-skipping training candidate: rna82601.p1, not unique enough\n",
      "-skipping training candidate: rna37369.p1, not unique enough\n",
      "-skipping training candidate: rna84015.p1, not unique enough\n",
      "-skipping training candidate: rna10050.p1, not unique enough\n",
      "-skipping training candidate: rna97662.p1, not unique enough\n",
      "-skipping training candidate: rna32290.p2, not unique enough\n",
      "-skipping training candidate: rna708.p1, not unique enough\n",
      "-skipping training candidate: rna56471.p1, not unique enough\n",
      "-skipping training candidate: rna34374.p1, not unique enough\n",
      "-skipping training candidate: rna38864.p1, not unique enough\n",
      "-skipping training candidate: rna81649.p1, not unique enough\n",
      "-skipping training candidate: rna82703.p1, not unique enough\n",
      "-skipping training candidate: rna7116.p1, not unique enough\n",
      "-skipping training candidate: rna19641.p1, not unique enough\n",
      "-skipping training candidate: rna19639.p1, not unique enough\n",
      "-skipping training candidate: rna19639.p2, not unique enough\n",
      "-skipping training candidate: rna32931.p2, not unique enough\n",
      "-skipping training candidate: rna32930.p1, not unique enough\n",
      "-skipping training candidate: rna30639.p1, not unique enough\n",
      "-skipping training candidate: rna78195.p1, not unique enough\n",
      "-skipping training candidate: rna79535.p1, not unique enough\n",
      "-skipping training candidate: rna9581.p1, not unique enough\n",
      "-skipping training candidate: rna83150.p1, not unique enough\n",
      "-skipping training candidate: rna87430.p1, not unique enough\n",
      "-skipping training candidate: rna94517.p2, not unique enough\n",
      "-skipping training candidate: rna35834.p1, not unique enough\n",
      "-skipping training candidate: rna80698.p1, not unique enough\n",
      "-skipping training candidate: rna82210.p2, not unique enough\n",
      "-skipping training candidate: rna31910.p1, not unique enough\n",
      "-skipping training candidate: rna28595.p2, not unique enough\n",
      "-skipping training candidate: rna55339.p1, not unique enough\n",
      "-skipping training candidate: rna74100.p2, not unique enough\n",
      "-skipping training candidate: rna42087.p1, not unique enough\n",
      "-skipping training candidate: rna58146.p1, not unique enough\n",
      "-skipping training candidate: rna8897.p1, not unique enough\n",
      "-skipping training candidate: rna6783.p1, not unique enough\n",
      "-skipping training candidate: rna80512.p1, not unique enough\n",
      "-skipping training candidate: rna52324.p1, not unique enough\n",
      "-skipping training candidate: rna52324.p2, not unique enough\n",
      "-skipping training candidate: rna80052.p2, not unique enough\n",
      "-skipping training candidate: rna94066.p2, not unique enough\n",
      "-skipping training candidate: rna81986.p2, not unique enough\n",
      "-skipping training candidate: rna15695.p1, not unique enough\n",
      "-skipping training candidate: rna13030.p1, not unique enough\n",
      "-skipping training candidate: rna79710.p1, not unique enough\n",
      "-skipping training candidate: rna81913.p1, not unique enough\n",
      "-skipping training candidate: rna74338.p1, not unique enough\n",
      "-skipping training candidate: rna14700.p1, not unique enough\n",
      "-skipping training candidate: rna1482.p1, not unique enough\n",
      "-skipping training candidate: rna33608.p1, not unique enough\n",
      "-skipping training candidate: rna79508.p1, not unique enough\n",
      "-skipping training candidate: rna1820.p1, not unique enough\n",
      "-skipping training candidate: rna57212.p1, not unique enough\n",
      "-skipping training candidate: rna50200.p1, not unique enough\n",
      "-skipping training candidate: rna50200.p2, not unique enough\n",
      "-skipping training candidate: rna22759.p1, not unique enough\n",
      "-skipping training candidate: rna20693.p1, not unique enough\n",
      "-skipping training candidate: rna10917.p1, not unique enough\n",
      "-skipping training candidate: rna77083.p1, not unique enough\n",
      "-skipping training candidate: rna39320.p1, not unique enough\n",
      "-skipping training candidate: rna53215.p1, not unique enough\n",
      "-skipping training candidate: rna91571.p1, not unique enough\n",
      "-skipping training candidate: rna85776.p2, not unique enough\n",
      "-skipping training candidate: rna87121.p1, not unique enough\n",
      "-skipping training candidate: rna5941.p1, not unique enough\n",
      "-skipping training candidate: rna90102.p1, not unique enough\n",
      "-skipping training candidate: rna46028.p1, not unique enough\n",
      "-skipping training candidate: rna87694.p1, not unique enough\n",
      "-skipping training candidate: rna77313.p2, not unique enough\n",
      "-skipping training candidate: rna39183.p1, not unique enough\n",
      "-skipping training candidate: rna39183.p2, not unique enough\n",
      "-skipping training candidate: rna91233.p1, not unique enough\n",
      "-skipping training candidate: rna87038.p1, not unique enough\n",
      "-skipping training candidate: rna78535.p1, not unique enough\n",
      "-skipping training candidate: rna79219.p2, not unique enough\n",
      "-skipping training candidate: rna90917.p1, not unique enough\n",
      "-skipping training candidate: rna94539.p1, not unique enough\n",
      "-skipping training candidate: rna22898.p1, not unique enough\n",
      "-skipping training candidate: rna14336.p1, not unique enough\n",
      "-skipping training candidate: rna7784.p1, not unique enough\n",
      "-skipping training candidate: rna80898.p2, not unique enough\n",
      "-skipping training candidate: rna8521.p1, not unique enough\n",
      "-skipping training candidate: rna6686.p1, not unique enough\n",
      "-skipping training candidate: rna73881.p1, not unique enough\n",
      "-skipping training candidate: rna81151.p1, not unique enough\n",
      "-skipping training candidate: rna48686.p2, not unique enough\n",
      "-skipping training candidate: rna22898.p2, not unique enough\n",
      "-skipping training candidate: rna42940.p1, not unique enough\n",
      "-skipping training candidate: rna91330.p1, not unique enough\n",
      "-skipping training candidate: rna95035.p1, not unique enough\n",
      "-skipping training candidate: rna51803.p1, not unique enough\n",
      "-skipping training candidate: rna79294.p1, not unique enough\n",
      "-skipping training candidate: rna38956.p1, not unique enough\n",
      "-skipping training candidate: rna79072.p1, not unique enough\n",
      "-skipping training candidate: rna11634.p1, not unique enough\n",
      "-skipping training candidate: rna42743.p1, not unique enough\n",
      "-skipping training candidate: rna23131.p1, not unique enough\n",
      "-skipping training candidate: rna73961.p1, not unique enough\n",
      "-skipping training candidate: rna36107.p1, not unique enough\n",
      "-skipping training candidate: rna38644.p1, not unique enough\n",
      "-skipping training candidate: rna36185.p1, not unique enough\n",
      "-skipping training candidate: rna21420.p1, not unique enough\n",
      "-skipping training candidate: rna9613.p1, not unique enough\n",
      "-skipping training candidate: rna57374.p1, not unique enough\n",
      "-skipping training candidate: rna80292.p1, not unique enough\n",
      "-skipping training candidate: rna79738.p2, not unique enough\n",
      "-skipping training candidate: rna20338.p1, not unique enough\n",
      "-skipping training candidate: rna91738.p1, not unique enough\n",
      "-skipping training candidate: rna18588.p1, not unique enough\n",
      "-skipping training candidate: rna7129.p1, not unique enough\n",
      "-skipping training candidate: rna11426.p1, not unique enough\n",
      "-skipping training candidate: rna11426.p2, not unique enough\n",
      "-skipping training candidate: rna95714.p1, not unique enough\n",
      "-skipping training candidate: rna5789.p1, not unique enough\n",
      "-skipping training candidate: rna85782.p2, not unique enough\n",
      "-skipping training candidate: rna9263.p1, not unique enough\n",
      "-skipping training candidate: rna9875.p1, not unique enough\n",
      "-skipping training candidate: rna77089.p1, not unique enough\n",
      "-skipping training candidate: rna78894.p2, not unique enough\n",
      "-skipping training candidate: rna53939.p1, not unique enough\n",
      "-skipping training candidate: rna89032.p2, not unique enough\n",
      "-skipping training candidate: rna44011.p1, not unique enough\n",
      "-skipping training candidate: rna6672.p1, not unique enough\n",
      "-skipping training candidate: rna44011.p2, not unique enough\n",
      "-skipping training candidate: rna76326.p1, not unique enough\n",
      "-skipping training candidate: rna16124.p1, not unique enough\n",
      "-skipping training candidate: rna81521.p1, not unique enough\n",
      "-skipping training candidate: rna82177.p2, not unique enough\n",
      "-skipping training candidate: rna21004.p1, not unique enough\n",
      "-skipping training candidate: rna44041.p1, not unique enough\n",
      "-skipping training candidate: rna6280.p1, not unique enough\n",
      "-skipping training candidate: rna82289.p1, not unique enough\n",
      "-skipping training candidate: rna7261.p1, not unique enough\n",
      "-skipping training candidate: rna28238.p1, not unique enough\n",
      "-skipping training candidate: rna34186.p1, not unique enough\n",
      "-skipping training candidate: rna37602.p1, not unique enough\n",
      "-skipping training candidate: rna47182.p1, not unique enough\n",
      "-skipping training candidate: rna33436.p2, not unique enough\n",
      "-skipping training candidate: gene11512_t.p1, not unique enough\n",
      "-skipping training candidate: rna311.p1, not unique enough\n",
      "-skipping training candidate: rna312.p2, not unique enough\n",
      "-skipping training candidate: rna19245.p1, not unique enough\n",
      "-skipping training candidate: rna9840.p1, not unique enough\n",
      "-skipping training candidate: rna91796.p1, not unique enough\n",
      "-skipping training candidate: rna53326.p1, not unique enough\n",
      "-skipping training candidate: rna86733.p1, not unique enough\n",
      "-skipping training candidate: rna29516.p1, not unique enough\n",
      "-skipping training candidate: rna84635.p1, not unique enough\n",
      "-skipping training candidate: rna18159.p1, not unique enough\n",
      "-skipping training candidate: rna6112.p1, not unique enough\n",
      "-skipping training candidate: rna25305.p1, not unique enough\n",
      "-skipping training candidate: rna5866.p1, not unique enough\n",
      "-skipping training candidate: rna11801.p1, not unique enough\n",
      "-skipping training candidate: rna7690.p2, not unique enough\n",
      "-skipping training candidate: rna16081.p1, not unique enough\n",
      "-skipping training candidate: rna27916.p1, not unique enough\n",
      "-skipping training candidate: rna44287.p2, not unique enough\n",
      "-skipping training candidate: rna59932.p2, not unique enough\n",
      "-skipping training candidate: rna95225.p1, not unique enough\n",
      "-skipping training candidate: rna73948.p1, not unique enough\n",
      "-skipping training candidate: rna88270.p1, not unique enough\n",
      "-skipping training candidate: rna87426.p1, not unique enough\n",
      "-skipping training candidate: rna74986.p1, not unique enough\n",
      "-skipping training candidate: rna10128.p1, not unique enough\n",
      "-skipping training candidate: rna91557.p2, not unique enough\n",
      "-skipping training candidate: rna97335.p1, not unique enough\n",
      "-skipping training candidate: rna12971.p1, not unique enough\n",
      "-skipping training candidate: rna94898.p2, not unique enough\n",
      "-skipping training candidate: rna11817.p1, not unique enough\n",
      "-skipping training candidate: rna83660.p1, not unique enough\n",
      "-skipping training candidate: rna7803.p1, not unique enough\n",
      "-skipping training candidate: rna20896.p1, not unique enough\n",
      "-skipping training candidate: rna3812.p1, not unique enough\n",
      "-skipping training candidate: rna24424.p1, not unique enough\n",
      "-skipping training candidate: rna56043.p1, not unique enough\n",
      "-skipping training candidate: rna35170.p1, not unique enough\n",
      "-skipping training candidate: rna92300.p1, not unique enough\n",
      "-skipping training candidate: rna88700.p1, not unique enough\n",
      "-skipping training candidate: rna17740.p1, not unique enough\n",
      "-skipping training candidate: rna85241.p2, not unique enough\n",
      "-skipping training candidate: rna3588.p2, not unique enough\n",
      "-skipping training candidate: rna19905.p1, not unique enough\n",
      "-skipping training candidate: rna12695.p1, not unique enough\n",
      "-skipping training candidate: rna20941.p1, not unique enough\n",
      "-skipping training candidate: rna46204.p1, not unique enough\n",
      "-skipping training candidate: rna46204.p2, not unique enough\n",
      "-skipping training candidate: rna91280.p2, not unique enough\n",
      "-skipping training candidate: rna54791.p1, not unique enough\n",
      "-skipping training candidate: rna54791.p2, not unique enough\n",
      "-skipping training candidate: rna75306.p1, not unique enough\n",
      "-skipping training candidate: rna87852.p2, not unique enough\n",
      "-skipping training candidate: rna84025.p1, not unique enough\n",
      "-skipping training candidate: rna87792.p2, not unique enough\n",
      "-skipping training candidate: rna76799.p1, not unique enough\n",
      "-skipping training candidate: rna15290.p1, not unique enough\n",
      "-skipping training candidate: rna41528.p1, not unique enough\n",
      "-skipping training candidate: rna16761.p1, not unique enough\n",
      "-skipping training candidate: rna28227.p1, not unique enough\n",
      "-skipping training candidate: rna58582.p1, not unique enough\n",
      "-skipping training candidate: rna16527.p1, not unique enough\n",
      "-skipping training candidate: rna5524.p1, not unique enough\n",
      "-skipping training candidate: rna10474.p1, not unique enough\n",
      "-skipping training candidate: rna81554.p1, not unique enough\n",
      "-skipping training candidate: rna16239.p1, not unique enough\n",
      "-skipping training candidate: rna38145.p1, not unique enough\n",
      "-skipping training candidate: rna6767.p1, not unique enough\n",
      "-skipping training candidate: rna15286.p1, not unique enough\n",
      "-skipping training candidate: rna97736.p1, not unique enough\n",
      "-skipping training candidate: rna31598.p1, not unique enough\n",
      "-skipping training candidate: rna26404.p1, not unique enough\n",
      "-skipping training candidate: rna89033.p1, not unique enough\n",
      "-skipping training candidate: rna22758.p1, not unique enough\n",
      "-skipping training candidate: rna29557.p2, not unique enough\n",
      "-skipping training candidate: rna29559.p1, not unique enough\n",
      "-skipping training candidate: rna38466.p1, not unique enough\n",
      "-skipping training candidate: rna52416.p1, not unique enough\n",
      "-skipping training candidate: rna1822.p2, not unique enough\n",
      "-skipping training candidate: rna23613.p1, not unique enough\n",
      "-skipping training candidate: rna32282.p2, not unique enough\n",
      "-skipping training candidate: rna92040.p1, not unique enough\n",
      "-skipping training candidate: rna91014.p1, not unique enough\n",
      "-skipping training candidate: rna87490.p1, not unique enough\n",
      "-skipping training candidate: rna84691.p1, not unique enough\n",
      "-skipping training candidate: rna30513.p1, not unique enough\n",
      "-skipping training candidate: rna32616.p1, not unique enough\n",
      "-skipping training candidate: rna9992.p1, not unique enough\n",
      "-skipping training candidate: rna94113.p2, not unique enough\n",
      "-skipping training candidate: rna49215.p2, not unique enough\n",
      "-skipping training candidate: rna85814.p1, not unique enough\n",
      "-skipping training candidate: rna18670.p1, not unique enough\n",
      "-skipping training candidate: rna18095.p1, not unique enough\n",
      "-skipping training candidate: rna95039.p1, not unique enough\n",
      "-skipping training candidate: rna21258.p1, not unique enough\n",
      "-skipping training candidate: rna25097.p2, not unique enough\n",
      "-skipping training candidate: rna22506.p1, not unique enough\n",
      "-skipping training candidate: rna59986.p1, not unique enough\n",
      "-skipping training candidate: rna17144.p1, not unique enough\n",
      "-skipping training candidate: rna681.p1, not unique enough\n",
      "-skipping training candidate: rna19600.p1, not unique enough\n",
      "-skipping training candidate: rna90525.p1, not unique enough\n",
      "-skipping training candidate: rna699.p1, not unique enough\n",
      "-skipping training candidate: rna29972.p1, not unique enough\n",
      "-skipping training candidate: rna59128.p1, not unique enough\n",
      "-skipping training candidate: rna91097.p1, not unique enough\n",
      "-skipping training candidate: rna78028.p2, not unique enough\n",
      "-skipping training candidate: rna53417.p1, not unique enough\n",
      "-skipping training candidate: rna88761.p2, not unique enough\n",
      "-skipping training candidate: rna24547.p1, not unique enough\n",
      "-skipping training candidate: rna21368.p1, not unique enough\n",
      "-skipping training candidate: rna10586.p1, not unique enough\n",
      "-skipping training candidate: rna90057.p1, not unique enough\n",
      "-skipping training candidate: rna95235.p1, not unique enough\n",
      "-skipping training candidate: rna87727.p1, not unique enough\n",
      "-skipping training candidate: rna86966.p1, not unique enough\n",
      "-skipping training candidate: rna93049.p2, not unique enough\n",
      "-skipping training candidate: rna91636.p1, not unique enough\n",
      "-skipping training candidate: rna97205.p1, not unique enough\n",
      "-skipping training candidate: rna25597.p1, not unique enough\n",
      "-skipping training candidate: rna30721.p1, not unique enough\n",
      "-skipping training candidate: rna19190.p1, not unique enough\n",
      "-skipping training candidate: rna32290.p3, not unique enough\n",
      "-skipping training candidate: rna91604.p2, not unique enough\n",
      "-skipping training candidate: rna41700.p1, not unique enough\n",
      "-skipping training candidate: rna78412.p1, not unique enough\n",
      "-skipping training candidate: rna83839.p1, not unique enough\n",
      "-skipping training candidate: rna83839.p2, not unique enough\n",
      "-skipping training candidate: gene38276_t.p1, not unique enough\n",
      "-skipping training candidate: rna52416.p2, not unique enough\n",
      "-skipping training candidate: rna3314.p2, not unique enough\n",
      "-skipping training candidate: rna3315.p1, not unique enough\n",
      "-skipping training candidate: rna39244.p1, not unique enough\n",
      "-skipping training candidate: rna39244.p2, not unique enough\n",
      "-skipping training candidate: rna45986.p1, not unique enough\n",
      "-skipping training candidate: rna59081.p2, not unique enough\n",
      "-skipping training candidate: rna59080.p1, not unique enough\n",
      "-skipping training candidate: rna43839.p1, not unique enough\n",
      "-skipping training candidate: rna51666.p1, not unique enough\n",
      "-skipping training candidate: rna22556.p1, not unique enough\n",
      "-skipping training candidate: rna28648.p1, not unique enough\n",
      "-skipping training candidate: rna45232.p1, not unique enough\n",
      "-skipping training candidate: rna48862.p1, not unique enough\n",
      "-skipping training candidate: rna3291.p1, not unique enough\n",
      "-skipping training candidate: rna8831.p2, not unique enough\n",
      "-skipping training candidate: rna74593.p2, not unique enough\n",
      "-skipping training candidate: rna3941.p2, not unique enough\n",
      "-skipping training candidate: rna39556.p1, not unique enough\n",
      "-skipping training candidate: rna95364.p1, not unique enough\n",
      "-skipping training candidate: rna41222.p1, not unique enough\n",
      "-skipping training candidate: rna86895.p2, not unique enough\n",
      "-skipping training candidate: rna21548.p1, not unique enough\n",
      "-skipping training candidate: rna81171.p1, not unique enough\n",
      "-skipping training candidate: rna24173.p1, not unique enough\n",
      "-skipping training candidate: rna39975.p1, not unique enough\n",
      "-skipping training candidate: rna79863.p1, not unique enough\n",
      "-skipping training candidate: rna92181.p2, not unique enough\n",
      "-skipping training candidate: rna82032.p2, not unique enough\n",
      "-skipping training candidate: rna25993.p2, not unique enough\n",
      "-skipping training candidate: rna92692.p2, not unique enough\n",
      "-skipping training candidate: rna77762.p1, not unique enough\n",
      "-skipping training candidate: rna77029.p1, not unique enough\n",
      "-skipping training candidate: rna47739.p1, not unique enough\n",
      "-skipping training candidate: rna87990.p1, not unique enough\n",
      "-skipping training candidate: rna91171.p2, not unique enough\n",
      "-skipping training candidate: rna5032.p1, not unique enough\n",
      "-skipping training candidate: rna5033.p2, not unique enough\n",
      "-skipping training candidate: rna16342.p1, not unique enough\n",
      "-skipping training candidate: rna4284.p1, not unique enough\n",
      "-skipping training candidate: rna4285.p1, not unique enough\n",
      "-skipping training candidate: rna75706.p1, not unique enough\n",
      "-skipping training candidate: rna12612.p1, not unique enough\n",
      "-skipping training candidate: rna76958.p1, not unique enough\n",
      "-skipping training candidate: rna90833.p2, not unique enough\n",
      "-skipping training candidate: rna54421.p1, not unique enough\n",
      "-skipping training candidate: rna54421.p2, not unique enough\n",
      "-skipping training candidate: rna48716.p1, not unique enough\n",
      "-skipping training candidate: rna48715.p1, not unique enough\n",
      "-skipping training candidate: rna48715.p2, not unique enough\n",
      "-skipping training candidate: rna48763.p1, not unique enough\n",
      "-skipping training candidate: rna30275.p1, not unique enough\n",
      "-skipping training candidate: rna10169.p1, not unique enough\n",
      "-skipping training candidate: rna87598.p2, not unique enough\n",
      "-skipping training candidate: rna18524.p1, not unique enough\n",
      "-skipping training candidate: rna54889.p1, not unique enough\n",
      "-skipping training candidate: rna78194.p1, not unique enough\n",
      "-skipping training candidate: rna88844.p1, not unique enough\n",
      "-skipping training candidate: rna13766.p1, not unique enough\n",
      "-skipping training candidate: rna13766.p2, not unique enough\n",
      "-skipping training candidate: rna22115.p1, not unique enough\n",
      "-skipping training candidate: rna8429.p1, not unique enough\n",
      "-skipping training candidate: rna8429.p2, not unique enough\n",
      "-skipping training candidate: rna75447.p2, not unique enough\n",
      "-skipping training candidate: rna25723.p1, not unique enough\n",
      "-skipping training candidate: rna907.p1, not unique enough\n",
      "-skipping training candidate: rna90535.p2, not unique enough\n",
      "-skipping training candidate: rna19175.p1, not unique enough\n",
      "-skipping training candidate: rna19176.p1, not unique enough\n",
      "-skipping training candidate: rna87763.p1, not unique enough\n",
      "-skipping training candidate: rna15974.p1, not unique enough\n",
      "-skipping training candidate: rna56920.p1, not unique enough\n",
      "-skipping training candidate: rna56920.p2, not unique enough\n",
      "-skipping training candidate: rna7224.p1, not unique enough\n",
      "-skipping training candidate: rna68082.p1, not unique enough\n",
      "-skipping training candidate: rna89431.p1, not unique enough\n",
      "-skipping training candidate: rna10962.p1, not unique enough\n",
      "-skipping training candidate: rna8013.p1, not unique enough\n",
      "-skipping training candidate: rna18940.p1, not unique enough\n",
      "-skipping training candidate: rna29055.p1, not unique enough\n",
      "-skipping training candidate: rna51908.p1, not unique enough\n",
      "-skipping training candidate: rna74408.p1, not unique enough\n",
      "-skipping training candidate: rna76097.p1, not unique enough\n",
      "-skipping training candidate: rna81477.p1, not unique enough\n",
      "-skipping training candidate: rna43825.p1, not unique enough\n",
      "-skipping training candidate: rna89118.p2, not unique enough\n",
      "-skipping training candidate: rna29651.p1, not unique enough\n",
      "-skipping training candidate: rna38238.p2, not unique enough\n",
      "-skipping training candidate: rna32283.p2, not unique enough\n",
      "-skipping training candidate: rna86164.p1, not unique enough\n",
      "-skipping training candidate: rna75360.p2, not unique enough\n",
      "-skipping training candidate: rna21267.p1, not unique enough\n",
      "-skipping training candidate: rna87720.p1, not unique enough\n",
      "-skipping training candidate: rna46883.p1, not unique enough\n",
      "-skipping training candidate: rna74989.p1, not unique enough\n",
      "-skipping training candidate: rna56989.p1, not unique enough\n",
      "-skipping training candidate: rna91873.p2, not unique enough\n",
      "-skipping training candidate: rna20627.p1, not unique enough\n",
      "-skipping training candidate: rna79454.p1, not unique enough\n",
      "-skipping training candidate: rna91772.p2, not unique enough\n",
      "-skipping training candidate: rna59761.p1, not unique enough\n",
      "-skipping training candidate: rna31298.p2, not unique enough\n",
      "-skipping training candidate: rna31299.p1, not unique enough\n",
      "-skipping training candidate: rna90750.p1, not unique enough\n",
      "-skipping training candidate: rna3192.p1, not unique enough\n",
      "-skipping training candidate: rna33362.p1, not unique enough\n",
      "-skipping training candidate: rna97783.p2, not unique enough\n",
      "-skipping training candidate: rna48021.p1, not unique enough\n",
      "-skipping training candidate: rna48021.p2, not unique enough\n",
      "-skipping training candidate: rna70132.p1, not unique enough\n",
      "-skipping training candidate: rna30632.p1, not unique enough\n",
      "-skipping training candidate: rna88847.p1, not unique enough\n",
      "-skipping training candidate: rna78018.p2, not unique enough\n",
      "-skipping training candidate: rna90853.p2, not unique enough\n",
      "-skipping training candidate: rna79704.p2, not unique enough\n",
      "-skipping training candidate: rna456.p1, not unique enough\n",
      "-skipping training candidate: rna51704.p2, not unique enough\n",
      "-skipping training candidate: rna81011.p1, not unique enough\n",
      "-skipping training candidate: rna31239.p1, not unique enough\n",
      "-skipping training candidate: rna623.p2, not unique enough\n",
      "-skipping training candidate: rna5457.p2, not unique enough\n",
      "-skipping training candidate: rna80774.p2, not unique enough\n",
      "-skipping training candidate: rna38239.p1, not unique enough\n",
      "-skipping training candidate: rna29869.p1, not unique enough\n",
      "-skipping training candidate: rna93358.p2, not unique enough\n",
      "-skipping training candidate: rna83338.p2, not unique enough\n",
      "-skipping training candidate: rna80756.p1, not unique enough\n",
      "-skipping training candidate: rna94457.p1, not unique enough\n",
      "-skipping training candidate: rna51223.p2, not unique enough\n",
      "-skipping training candidate: rna27599.p1, not unique enough\n",
      "-skipping training candidate: rna92140.p1, not unique enough\n",
      "-skipping training candidate: rna34188.p1, not unique enough\n",
      "-skipping training candidate: rna34188.p2, not unique enough\n",
      "-skipping training candidate: rna82217.p2, not unique enough\n",
      "-skipping training candidate: rna93323.p1, not unique enough\n",
      "-skipping training candidate: rna93626.p1, not unique enough\n",
      "-skipping training candidate: rna38470.p1, not unique enough\n",
      "-skipping training candidate: rna79731.p2, not unique enough\n",
      "-skipping training candidate: rna85145.p2, not unique enough\n",
      "-skipping training candidate: rna43825.p2, not unique enough\n",
      "-skipping training candidate: rna89112.p1, not unique enough\n",
      "-skipping training candidate: rna95391.p2, not unique enough\n",
      "-skipping training candidate: rna80007.p2, not unique enough\n",
      "-skipping training candidate: rna89435.p2, not unique enough\n",
      "-skipping training candidate: rna82778.p1, not unique enough\n",
      "-skipping training candidate: rna4438.p2, not unique enough\n",
      "-skipping training candidate: rna4437.p1, not unique enough\n",
      "-skipping training candidate: rna80610.p1, not unique enough\n",
      "-skipping training candidate: rna310.p1, not unique enough\n",
      "-skipping training candidate: rna310.p2, not unique enough\n",
      "-skipping training candidate: rna88230.p1, not unique enough\n",
      "-skipping training candidate: rna53606.p1, not unique enough\n",
      "-skipping training candidate: rna82992.p2, not unique enough\n",
      "-skipping training candidate: rna93625.p1, not unique enough\n",
      "\n",
      "\t-redundancy-minimized set includes 4391 / 5000 = 87.82%\n",
      "\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr 500 > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest\n",
      "PCT_GC: 43.6\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/seq_n_baseprobs_to_loglikelihood_vals.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//base_freqs.dat > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//hexamer.scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/score_CDS_likelihood_all_6_frames.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//hexamer.scores > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/select_best_ORFs_per_transcript.pl --gff3_file ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.gff3 --cds_scores ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.scores  --min_length_auto_accept 645  > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 \n",
      "Selecting best orfs\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/train_start_PWM.pl --transcripts ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa --selected_orfs ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest --out_prefix ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement\n",
      "Training start codon pattern recognition* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/build_atgPWM_+-.pl  --transcripts ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa  --selected_orfs ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest  --out_prefix ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement --pwm_left 20 --pwm_right 10 \n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scoring.+-.pl  --features_plus ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.+.features  --features_minus ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.-.features  --atg_position 20  > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.feature.scores\n",
      "-round: 1\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "-round: 2\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "-round: 3\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "-round: 4\n",
      "-round: 5\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scores_to_ROC.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.feature.scores > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.feature.scores.roc\n",
      "-parsing scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/plot_ROC.Rscript ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.feature.scores.roc || :\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/compute_AUC.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.feature.scores.roc\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.+.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.-.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/deplete_feature_noise.pl  --features_plus ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.+.features  --pwm_minus ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.-.pwm  --out_prefix ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced\n",
      "num features: 394\tnum_incorporate: 118\n",
      "-feature swap of score: 0.282211042182629 instead of -2.10559846425201\n",
      "-feature swap of score: -1.16188308206509 instead of -2.28119328826824\n",
      "-feature swap of score: -1.36349372230503 instead of -2.37308013140263\n",
      "-feature swap of score: -1.5288154115419 instead of -2.38633864247541\n",
      "-feature swap of score: 1.30080256968223 instead of -2.12562639345566\n",
      "-feature swap of score: 2.58302338351591 instead of -2.1542417313035\n",
      "-feature swap of score: 1.03923917311702 instead of -2.15704712686731\n",
      "-feature swap of score: -0.759392418234196 instead of -1.96604994606952\n",
      "-feature swap of score: 1.19952871541642 instead of -2.0819275811987\n",
      "-feature swap of score: 1.62853952810211 instead of -1.79898664301526\n",
      "-feature swap of score: 1.82999954603619 instead of -1.60093084783994\n",
      "-feature swap of score: -0.409880358663875 instead of -1.48921771076515\n",
      "-feature swap of score: 2.14172589070934 instead of -1.50611235524916\n",
      "-feature swap of score: -1.27774293922662 instead of -1.35177364813391\n",
      "-feature swap of score: 0.406226976848686 instead of -1.28510457319706\n",
      "-feature swap of score: 0.441226568709701 instead of -1.25373972198834\n",
      "-feature swap of score: 0.626491054257062 instead of -1.26927450741765\n",
      "-feature swap of score: 1.93790701418717 instead of -1.2201140845988\n",
      "-feature swap of score: -0.109984239857404 instead of -1.08721029604963\n",
      "-feature swap of score: 2.18519899500483 instead of -1.20033734228051\n",
      "-feature swap of score: -0.282977604136504 instead of -1.42097881097281\n",
      "-feature swap of score: 1.60700240266365 instead of -1.17990234296841\n",
      "-feature swap of score: 0.280403271850439 instead of -1.25025464775653\n",
      "-feature swap of score: -0.0316551469448385 instead of -1.25575204194636\n",
      "-feature swap of score: -0.0414662588933564 instead of -0.927727164090748\n",
      "-feature swap of score: 2.58354495247625 instead of -0.871376023819319\n",
      "-feature swap of score: 2.42482128667312 instead of -0.638047212212269\n",
      "-feature swap of score: 0.802272207381882 instead of -0.553378558509306\n",
      "-feature swap of score: 1.96637774623844 instead of -0.591697075919005\n",
      "-feature swap of score: 2.50089740026352 instead of -0.614720806116368\n",
      "-feature swap of score: 1.31547905507288 instead of -0.681073940143409\n",
      "-feature swap of score: 2.0503127549551 instead of -0.800074526921327\n",
      "-feature swap of score: -0.540794902377789 instead of -0.943146284757201\n",
      "-feature swap of score: 3.21936153880605 instead of -0.529303620805585\n",
      "-feature swap of score: -0.280206070799899 instead of -0.923114463589244\n",
      "-feature swap of score: 0.492333981155279 instead of -0.566978714560265\n",
      "-feature swap of score: -0.0379615068123152 instead of -0.615554486218062\n",
      "-feature swap of score: 1.47384850982698 instead of -0.688945354099117\n",
      "-feature swap of score: 1.54264297230436 instead of -0.603318664781251\n",
      "-feature swap of score: 2.31213297609682 instead of -0.586164448643114\n",
      "-feature swap of score: 1.28399873498245 instead of -0.523392581335056\n",
      "-feature swap of score: 4.32356403148463 instead of -0.457738966643939\n",
      "-feature swap of score: 0.919129208478484 instead of -0.16288645904931\n",
      "-feature swap of score: 0.945008351476351 instead of -0.188167672189052\n",
      "-feature swap of score: 0.521855507732424 instead of 0.00137206864495978\n",
      "-feature swap of score: 2.00909268588197 instead of 0.0508602142857672\n",
      "-feature swap of score: 3.8789868461693 instead of -0.409265424427188\n",
      "-feature swap of score: 3.30482527345489 instead of 0.054441546784111\n",
      "-feature swap of score: 2.89407013561405 instead of -0.259339332745416\n",
      "-feature swap of score: 1.42694817626995 instead of 0.15533258362599\n",
      "-feature swap of score: 2.61829963556781 instead of 0.0577885480580716\n",
      "-feature swap of score: 1.46032640400844 instead of -0.759081159332202\n",
      "-feature swap of score: 3.1987968325987 instead of 0.130406036293526\n",
      "-feature swap of score: 1.64591626217056 instead of 0.136518768048064\n",
      "-feature swap of score: 0.734441609097138 instead of -0.0471976760525726\n",
      "-feature swap of score: 6.26848552992193 instead of 0.0974932784683117\n",
      "-feature swap of score: 0.504067100327014 instead of 0.0981934882213432\n",
      "-feature swap of score: 1.76313046249357 instead of 0.125936857115619\n",
      "-feature swap of score: 2.80856965960322 instead of -0.0580718492674737\n",
      "-feature swap of score: 1.47511676968584 instead of -0.592738023855157\n",
      "-feature swap of score: 0.667598564789021 instead of -0.260516904637941\n",
      "-feature swap of score: 2.29231410789857 instead of 0.216407455611756\n",
      "-feature swap of score: 1.33864385621757 instead of 0.310067355290087\n",
      "-feature swap of score: 3.49114139115248 instead of 0.221778112540442\n",
      "-feature swap of score: 0.823047205017706 instead of 0.147206600999125\n",
      "-feature swap of score: 0.834557697944349 instead of 0.0441167131500077\n",
      "-feature swap of score: 0.916174804261084 instead of 0.237551544024209\n",
      "-feature swap of score: 0.600722186759774 instead of 0.194054492371085\n",
      "-feature swap of score: 3.62798172998081 instead of 0.349873938045235\n",
      "-feature swap of score: 0.848718626512436 instead of 0.500482715303364\n",
      "-feature swap of score: 3.86679779014841 instead of 0.507184679247535\n",
      "-feature swap of score: 3.05415416486409 instead of 0.534081732772071\n",
      "-feature swap of score: 1.79653234145429 instead of 0.472572764296392\n",
      "-feature swap of score: 2.5909757190806 instead of 0.571540252195026\n",
      "-feature swap of score: 2.11006870070995 instead of 0.7831925943114\n",
      "-feature swap of score: 2.06560046334988 instead of 0.642201748126931\n",
      "-feature swap of score: 3.84106436968357 instead of 0.564815099178643\n",
      "-feature swap of score: 1.8304233027878 instead of 0.531740663218364\n",
      "-feature swap of score: 3.62935035945454 instead of 0.425594374622966\n",
      "-feature swap of score: 4.05588361061295 instead of 0.875062034749462\n",
      "-feature swap of score: 2.52617220838208 instead of 0.738027466992721\n",
      "-feature swap of score: 2.59648121270437 instead of 0.541739284843936\n",
      "-feature swap of score: 1.82483644465211 instead of 0.674551095987181\n",
      "-feature swap of score: 2.43328755041758 instead of 0.853248380154304\n",
      "-feature swap of score: 2.44080616941883 instead of 0.731183495140924\n",
      "-feature swap of score: 1.24916808158317 instead of 0.77213595016404\n",
      "-feature swap of score: 1.55097838423933 instead of 0.850175167366395\n",
      "-feature swap of score: 4.45325473981762 instead of 0.876831010138358\n",
      "-feature swap of score: 2.53336886429454 instead of 0.733810196778028\n",
      "-feature swap of score: 1.11602285022214 instead of 0.887437107244516\n",
      "-feature swap of score: 4.95851072455217 instead of 0.573626073776679\n",
      "-feature swap of score: 2.63879673496215 instead of 0.946222204216221\n",
      "-feature swap of score: 2.90306087537137 instead of 1.0596939471365\n",
      "-feature swap of score: 4.69454726695578 instead of 0.925737879895029\n",
      "-feature swap of score: 1.76748172426744 instead of 0.849640011141634\n",
      "-feature swap of score: 2.88097063597134 instead of 0.853185342403306\n",
      "-feature swap of score: 3.63690266701346 instead of 0.903386371484112\n",
      "-feature swap of score: 2.01489257622004 instead of 1.12181976425328\n",
      "-feature swap of score: 3.00125240686338 instead of 1.20027986303452\n",
      "-feature swap of score: 2.02276500345907 instead of 1.18475632675944\n",
      "-feature swap of score: 0.862602147240493 instead of 0.706180128952248\n",
      "-feature swap of score: 2.07106269219922 instead of 1.08323812424414\n",
      "-feature swap of score: 4.34474298095723 instead of 1.20106403164539\n",
      "-feature swap of score: 6.98863345907748 instead of 1.34672721352409\n",
      "-feature swap of score: 3.02216436983975 instead of 1.29201495888945\n",
      "-feature swap of score: 1.51165236841433 instead of 1.04456769599599\n",
      "-feature swap of score: 3.27124590599817 instead of 1.09904350340881\n",
      "-feature swap of score: 4.64304288914368 instead of 1.17510007153173\n",
      "-num feature swaps: 108\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scoring.+-.pl  --features_plus ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.+.features  --features_minus ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.-.features  --atg_position 20  > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores\n",
      "-round: 1\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "-round: 2\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "-round: 3\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "-round: 4\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "-round: 5\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scores_to_ROC.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc\n",
      "-parsing scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/plot_ROC.Rscript ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc || :\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/compute_AUC.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.+.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/start_codon_refinement.pl --transcripts ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa --gff3_file ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 --workdir ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir/ > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3\n",
      "Refining start codon selections.\n",
      "-number of revised start positions: 3105\n",
      "* Running CMD: cp ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3 XENLA_9.2_genome_cDNA.fa.transdecoder.gff3\n",
      "copying output to final output file: XENLA_9.2_genome_cDNA.fa.transdecoder.gff3* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_bed.pl XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 > XENLA_9.2_genome_cDNA.fa.transdecoder.bed\n",
      "Making bed file: XENLA_9.2_genome_cDNA.fa.transdecoder.bed\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 --fasta ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa  --genetic_code Universal > XENLA_9.2_genome_cDNA.fa.transdecoder.pep\n",
      "Making pep file: XENLA_9.2_genome_cDNA.fa.transdecoder.pep\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 --fasta ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa --seqType CDS  --genetic_code Universal > XENLA_9.2_genome_cDNA.fa.transdecoder.cds\n",
      "Making cds file: XENLA_9.2_genome_cDNA.fa.transdecoder.cds\n",
      "transdecoder is finished.  See output files XENLA_9.2_genome_cDNA.fa.transdecoder.*\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cdna = sample_Docket.genome_fasta.get_transdecoder_cdna_gtf(models_asgtf_updated, TRANSDECODER_LOC)\n",
    "sample_Docket.add_keyfile(cdna, 'cdna')\n",
    "\n",
    "transdecoder_files = sample_Docket.cdna.to_pep_files(TDLONGORF_LOC, TDPREDICT_LOC)\n",
    "sample_Docket.add_keyfiles(transdecoder_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "iEug6TQ6Cc9D"
   },
   "source": [
    "# 12. Push files to AWS S3\n",
    "\n",
    "Iteratively moves through the file_set and file_dict variables and populates files into the right place in AWS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 13. Pickle the `sample_Docket` variable for use by the next script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "dill_filename = output_folder + '_'.join([species_prefix, conditions, 'sample_Docket.pkl'])\n",
    "\n",
    "with open(dill_filename, 'wb') as file:\n",
    "    dill.dump(sample_Docket, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "Xtro_bro.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
