{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "L56fjA27DmhA"
   },
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "t0BTCsz7Y48y",
    "outputId": "f6510678-f977-4ec8-ce30-1f2acd4fe666"
   },
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import os\n",
    "import dill\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "import sys\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Download and describe data\n",
    "\n",
    "This notebook collects data from the [Liao et al. 2022](https://www.sciencedirect.com/science/article/pii/S0092867418301168#sec4) Xenopus laevis adult cell atlas.\n",
    "\n",
    "## Dataset description\n",
    "\n",
    "- Data can be found at the [GEO Accession GSE195790](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE195790).\n",
    "- This notebook previously collects data from one sample, [\"Xenopus_brain_COL65\"](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM6214268), for the purposes of data analysis and exploration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  150M  100  150M    0     0  28.7M      0  0:00:05  0:00:05 --:--:-- 31.3M\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'species': 'Xenopus_laevis',\n",
       " 'conditions': 'adultbrain',\n",
       " 'directory': '../../output/Xlae_adultbrain/',\n",
       " 'files': {},\n",
       " 'annot': <biofile_handling.GenomeGffFile at 0x7f4629a2e260>,\n",
       " 'genome_fasta': <biofile_handling.GenomeFastaFile at 0x7f4629a2e230>,\n",
       " 'gxc': <biofile_handling.GxcFile at 0x7f46daa72530>}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify the name of the species folder in Amazon S3\n",
    "species = 'Xenopus_laevis'\n",
    "\n",
    "# Specify any particular identifying conditions, eg tissue type:\n",
    "conditions = 'adultbrain'\n",
    "\n",
    "# Specify url and other variables\n",
    "genome_fasta_url = 'https://ftp.xenbase.org/pub/Genomics/JGI/Xenla9.2/XENLA_9.2_genome.fa.gz'\n",
    "genome_version = 'XENLA_9.2'\n",
    "\n",
    "annot_url = 'https://ftp.xenbase.org/pub/Genomics/JGI/Xenla9.2/XENLA_9.2_GCA.gff3'\n",
    "gxc_url = 'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM6214nnn/GSM6214268/suppl/GSM6214268_Xenopus_brain_COL65_dge.txt.gz'\n",
    "\n",
    "###########\n",
    "# runtime #\n",
    "###########\n",
    "\n",
    "protocol = 'curl'\n",
    "\n",
    "species_prefix = prefixify(species)\n",
    "\n",
    "# Specify folder as destination for file downloads\n",
    "output_folder = '../../output/' + prefixify(species) + '_' + conditions + '/'\n",
    "\n",
    "if not os.path.exists(output_folder):\n",
    "    os.mkdir(output_folder)\n",
    "    \n",
    "species_SampleDict = SampleDict(species, conditions, output_folder)\n",
    "\n",
    "genome_fasta = GenomeFastaFile(\n",
    "    filename = '',\n",
    "    sampledict = species_SampleDict,\n",
    "    version = genome_version,\n",
    "    url = genome_fasta_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "annot = GenomeGffFile(\n",
    "    filename = '',\n",
    "    sampledict = species_SampleDict,\n",
    "    GenomeFastaFile = genome_fasta,\n",
    "    url = annot_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "gxc = GxcFile(\n",
    "    filename = '',\n",
    "    sampledict = species_SampleDict,\n",
    "    GenomeFastaFile = genome_fasta,\n",
    "    GenomeAnnotFile = annot,\n",
    "    url = gxc_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "\n",
    "sample_Docket = Docket(species_SampleDict)\n",
    "keyfiles = {\n",
    "    'annot': annot,\n",
    "    'genome_fasta': genome_fasta,\n",
    "    'gxc': gxc\n",
    "}\n",
    "sample_Docket.add_keyfiles(keyfiles)\n",
    "\n",
    "display(vars(sample_Docket))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Load in the gxc matrix and get gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>AACCTATTCATATAAGGG</th>\n",
       "      <th>CTCGCATCAAAGTTAACT</th>\n",
       "      <th>AACCTAGTATACTTCCGC</th>\n",
       "      <th>AACCTAAAAGTTCTGAAA</th>\n",
       "      <th>CTCGCACGCACCCTCCAT</th>\n",
       "      <th>ACGTTGTATTGTAGCGAG</th>\n",
       "      <th>ACGAGCATGCTTTAGTCG</th>\n",
       "      <th>AACCTAGTCCCGCCATCT</th>\n",
       "      <th>AACCTAGCGAATTAGAGA</th>\n",
       "      <th>...</th>\n",
       "      <th>TCACTTGTTGCCATGCTT</th>\n",
       "      <th>TCGGGTTGTCACACTTAT</th>\n",
       "      <th>TCGTAATCGTAAGTTGCC</th>\n",
       "      <th>TGTCACGAATTACACAAG</th>\n",
       "      <th>TGTGCGTACTTCTAGTCG</th>\n",
       "      <th>TTAACTATACAGTGGATG</th>\n",
       "      <th>TTGGACACTTATGATCTT</th>\n",
       "      <th>AAAGTTACTTATGCCCTC</th>\n",
       "      <th>AACCTACGCACCTGCGGA</th>\n",
       "      <th>AACCTATAGTCGCTGTGT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.S</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>42Sp43.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42Sp50.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AK6.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AK6.S</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25907</th>\n",
       "      <td>zyg11b.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25908</th>\n",
       "      <td>zyg11b.S</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25909</th>\n",
       "      <td>zzef1.S</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25910</th>\n",
       "      <td>zzz3.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25911</th>\n",
       "      <td>zzz3.S</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25912 rows × 30001 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GENE  AACCTATTCATATAAGGG  CTCGCATCAAAGTTAACT  AACCTAGTATACTTCCGC  \\\n",
       "0           3.S                   0                   0                   0   \n",
       "1      42Sp43.L                   0                   0                   0   \n",
       "2      42Sp50.L                   0                   0                   0   \n",
       "3         AK6.L                   0                   0                   0   \n",
       "4         AK6.S                   1                   0                   0   \n",
       "...         ...                 ...                 ...                 ...   \n",
       "25907  zyg11b.L                   0                   0                   0   \n",
       "25908  zyg11b.S                   0                   0                   0   \n",
       "25909   zzef1.S                   0                   0                   0   \n",
       "25910    zzz3.L                   0                   0                   0   \n",
       "25911    zzz3.S                   0                   0                   0   \n",
       "\n",
       "       AACCTAAAAGTTCTGAAA  CTCGCACGCACCCTCCAT  ACGTTGTATTGTAGCGAG  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "25907                   1                   0                   0   \n",
       "25908                   0                   0                   0   \n",
       "25909                   0                   1                   0   \n",
       "25910                   0                   0                   0   \n",
       "25911                   0                   0                   0   \n",
       "\n",
       "       ACGAGCATGCTTTAGTCG  AACCTAGTCCCGCCATCT  AACCTAGCGAATTAGAGA  ...  \\\n",
       "0                       0                   0                   0  ...   \n",
       "1                       0                   0                   0  ...   \n",
       "2                       0                   0                   0  ...   \n",
       "3                       1                   0                   0  ...   \n",
       "4                       0                   0                   0  ...   \n",
       "...                   ...                 ...                 ...  ...   \n",
       "25907                   0                   0                   0  ...   \n",
       "25908                   0                   0                   0  ...   \n",
       "25909                   0                   0                   0  ...   \n",
       "25910                   2                   0                   0  ...   \n",
       "25911                   0                   0                   0  ...   \n",
       "\n",
       "       TCACTTGTTGCCATGCTT  TCGGGTTGTCACACTTAT  TCGTAATCGTAAGTTGCC  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "25907                   0                   0                   0   \n",
       "25908                   0                   0                   0   \n",
       "25909                   0                   0                   0   \n",
       "25910                   0                   0                   0   \n",
       "25911                   0                   0                   0   \n",
       "\n",
       "       TGTCACGAATTACACAAG  TGTGCGTACTTCTAGTCG  TTAACTATACAGTGGATG  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "25907                   0                   0                   0   \n",
       "25908                   0                   0                   0   \n",
       "25909                   0                   0                   0   \n",
       "25910                   0                   0                   0   \n",
       "25911                   0                   0                   0   \n",
       "\n",
       "       TTGGACACTTATGATCTT  AAAGTTACTTATGCCCTC  AACCTACGCACCTGCGGA  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "25907                   0                   0                   0   \n",
       "25908                   0                   0                   0   \n",
       "25909                   0                   0                   0   \n",
       "25910                   0                   0                   0   \n",
       "25911                   0                   0                   0   \n",
       "\n",
       "       AACCTATAGTCGCTGTGT  \n",
       "0                       0  \n",
       "1                       0  \n",
       "2                       0  \n",
       "3                       0  \n",
       "4                       0  \n",
       "...                   ...  \n",
       "25907                   0  \n",
       "25908                   0  \n",
       "25909                   0  \n",
       "25910                   0  \n",
       "25911                   0  \n",
       "\n",
       "[25912 rows x 30001 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>42Sp43.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42Sp50.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AK6.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AK6.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25907</th>\n",
       "      <td>zyg11b.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25908</th>\n",
       "      <td>zyg11b.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25909</th>\n",
       "      <td>zzef1.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25910</th>\n",
       "      <td>zzz3.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25911</th>\n",
       "      <td>zzz3.S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25912 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name\n",
       "0           3.S\n",
       "1      42Sp43.L\n",
       "2      42Sp50.L\n",
       "3         AK6.L\n",
       "4         AK6.S\n",
       "...         ...\n",
       "25907  zyg11b.L\n",
       "25908  zyg11b.S\n",
       "25909   zzef1.S\n",
       "25910    zzz3.L\n",
       "25911    zzz3.S\n",
       "\n",
       "[25912 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes_matrix = pd.read_csv(sample_Docket.gxc.path, sep = '\\t')\n",
    "display(genes_matrix)\n",
    "\n",
    "gxc_genes_list = pd.DataFrame({'gene_name':genes_matrix['GENE']})\n",
    "display(gxc_genes_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jvfcFNT2FuaZ"
   },
   "source": [
    "# 3. Get mapping identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Name=Xelaev18004747m;end_range=18399,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Note=transcript XELAEV_18...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>17924</td>\n",
       "      <td>18243</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809901</th>\n",
       "      <td>Scaffold94051</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8</td>\n",
       "      <td>205</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809902</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene45941;Name=Xelaev18004691m;end_range=23...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809903</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna47731;Parent=gene45941;Note=transcript X...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809904</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id452950;Parent=rna47731;Note=transcript XE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809905</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>809906 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0        1     2      3      4  5  6  7  \\\n",
       "0               chr1L  Genbank  gene  17924  18399  .  -  .   \n",
       "1               chr1L  Genbank  mRNA  17924  18399  .  -  .   \n",
       "2               chr1L  Genbank  exon  18336  18399  .  -  .   \n",
       "3               chr1L  Genbank  exon  17924  18243  .  -  .   \n",
       "4               chr1L  Genbank   CDS  18336  18399  .  -  0   \n",
       "...               ...      ...   ...    ...    ... .. .. ..   \n",
       "809901  Scaffold94051  Genbank   CDS      8    205  .  +  0   \n",
       "809902  Scaffold95291  Genbank  gene      9    236  .  -  .   \n",
       "809903  Scaffold95291  Genbank  mRNA      9    236  .  -  .   \n",
       "809904  Scaffold95291  Genbank  exon      9    236  .  -  .   \n",
       "809905  Scaffold95291  Genbank   CDS      9    236  .  -  0   \n",
       "\n",
       "                                                        8  \n",
       "0       ID=gene0;Name=Xelaev18004747m;end_range=18399,...  \n",
       "1       ID=rna0;Parent=gene0;Note=transcript XELAEV_18...  \n",
       "2       ID=id1;Parent=rna0;Note=transcript XELAEV_1800...  \n",
       "3       ID=id2;Parent=rna0;Note=transcript XELAEV_1800...  \n",
       "4       ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...  \n",
       "...                                                   ...  \n",
       "809901  ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...  \n",
       "809902  ID=gene45941;Name=Xelaev18004691m;end_range=23...  \n",
       "809903  ID=rna47731;Parent=gene45941;Note=transcript X...  \n",
       "809904  ID=id452950;Parent=rna47731;Note=transcript XE...  \n",
       "809905  ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...  \n",
       "\n",
       "[809906 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'ID=gene0;Name=Xelaev18004747m;end_range=18399,.;gbkey=Gene;gene_biotype=protein_coding;locus_tag=XELAEV_18004747mg;partial=true;start_range=.,17924;Alias=XB-GENE-5942444'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load in the original GFF-based annotation\n",
    "models = pd.read_csv(sample_Docket.annot.path, skiprows = 6, header = None, sep = '\\t', on_bad_lines = 'skip', comment='#')\n",
    "display(models)\n",
    "\n",
    "attributes_column = 8\n",
    "\n",
    "# Check the structure of fields in the GFF additional fields section\n",
    "display(models[attributes_column][0])\n",
    "\n",
    "# Remove any rows with NaNs\n",
    "models.dropna(inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Xenbase</th>\n",
       "      <th>Dbxref</th>\n",
       "      <th>dbxref_dict</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Name=Xelaev18004747m;end_range=18399,...</td>\n",
       "      <td>{'ID': 'gene0', 'Name': 'Xelaev18004747m', 'en...</td>\n",
       "      <td>Xelaev18004747m</td>\n",
       "      <td>XB-GENE-5942444</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>gene0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Note=transcript XELAEV_18...</td>\n",
       "      <td>{'ID': 'rna0', 'Parent': 'gene0', 'Note': 'tra...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rna0</td>\n",
       "      <td>gene0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "      <td>{'ID': 'id1', 'Parent': 'rna0', 'Note': 'trans...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>id1</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>17924</td>\n",
       "      <td>18243</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "      <td>{'ID': 'id2', 'Parent': 'rna0', 'Note': 'trans...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>id2</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...</td>\n",
       "      <td>{'ID': 'cds0', 'Parent': 'rna0', 'Dbxref': 'Ph...</td>\n",
       "      <td>OCT98948.1</td>\n",
       "      <td>None</td>\n",
       "      <td>Phytozome:Xelaev18004747m,NCBI_GP:OCT98948.1</td>\n",
       "      <td>{'Phytozome': 'Xelaev18004747m', 'NCBI_GP': 'O...</td>\n",
       "      <td>cds0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809901</th>\n",
       "      <td>Scaffold94051</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8</td>\n",
       "      <td>205</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...</td>\n",
       "      <td>{'ID': 'cds47730', 'Parent': 'rna47730', 'Dbxr...</td>\n",
       "      <td>OCT55143.1</td>\n",
       "      <td>None</td>\n",
       "      <td>Phytozome:Xelaev18004680m,NCBI_GP:OCT55143.1</td>\n",
       "      <td>{'Phytozome': 'Xelaev18004680m', 'NCBI_GP': 'O...</td>\n",
       "      <td>cds47730</td>\n",
       "      <td>rna47730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809902</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene45941;Name=Xelaev18004691m;end_range=23...</td>\n",
       "      <td>{'ID': 'gene45941', 'Name': 'Xelaev18004691m',...</td>\n",
       "      <td>Xelaev18004691m</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>gene45941</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809903</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna47731;Parent=gene45941;Note=transcript X...</td>\n",
       "      <td>{'ID': 'rna47731', 'Parent': 'gene45941', 'Not...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rna47731</td>\n",
       "      <td>gene45941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809904</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id452950;Parent=rna47731;Note=transcript XE...</td>\n",
       "      <td>{'ID': 'id452950', 'Parent': 'rna47731', 'Note...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>id452950</td>\n",
       "      <td>rna47731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809905</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...</td>\n",
       "      <td>{'ID': 'cds47731', 'Parent': 'rna47731', 'Dbxr...</td>\n",
       "      <td>OCT55142.1</td>\n",
       "      <td>None</td>\n",
       "      <td>Phytozome:Xelaev18004691m,NCBI_GP:OCT55142.1</td>\n",
       "      <td>{'Phytozome': 'Xelaev18004691m', 'NCBI_GP': 'O...</td>\n",
       "      <td>cds47731</td>\n",
       "      <td>rna47731</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>809906 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0        1     2      3      4  5  6  7  \\\n",
       "0               chr1L  Genbank  gene  17924  18399  .  -  .   \n",
       "1               chr1L  Genbank  mRNA  17924  18399  .  -  .   \n",
       "2               chr1L  Genbank  exon  18336  18399  .  -  .   \n",
       "3               chr1L  Genbank  exon  17924  18243  .  -  .   \n",
       "4               chr1L  Genbank   CDS  18336  18399  .  -  0   \n",
       "...               ...      ...   ...    ...    ... .. .. ..   \n",
       "809901  Scaffold94051  Genbank   CDS      8    205  .  +  0   \n",
       "809902  Scaffold95291  Genbank  gene      9    236  .  -  .   \n",
       "809903  Scaffold95291  Genbank  mRNA      9    236  .  -  .   \n",
       "809904  Scaffold95291  Genbank  exon      9    236  .  -  .   \n",
       "809905  Scaffold95291  Genbank   CDS      9    236  .  -  0   \n",
       "\n",
       "                                                        8  \\\n",
       "0       ID=gene0;Name=Xelaev18004747m;end_range=18399,...   \n",
       "1       ID=rna0;Parent=gene0;Note=transcript XELAEV_18...   \n",
       "2       ID=id1;Parent=rna0;Note=transcript XELAEV_1800...   \n",
       "3       ID=id2;Parent=rna0;Note=transcript XELAEV_1800...   \n",
       "4       ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...   \n",
       "...                                                   ...   \n",
       "809901  ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...   \n",
       "809902  ID=gene45941;Name=Xelaev18004691m;end_range=23...   \n",
       "809903  ID=rna47731;Parent=gene45941;Note=transcript X...   \n",
       "809904  ID=id452950;Parent=rna47731;Note=transcript XE...   \n",
       "809905  ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...   \n",
       "\n",
       "                                         field_dictionary        gene_name  \\\n",
       "0       {'ID': 'gene0', 'Name': 'Xelaev18004747m', 'en...  Xelaev18004747m   \n",
       "1       {'ID': 'rna0', 'Parent': 'gene0', 'Note': 'tra...             None   \n",
       "2       {'ID': 'id1', 'Parent': 'rna0', 'Note': 'trans...             None   \n",
       "3       {'ID': 'id2', 'Parent': 'rna0', 'Note': 'trans...             None   \n",
       "4       {'ID': 'cds0', 'Parent': 'rna0', 'Dbxref': 'Ph...       OCT98948.1   \n",
       "...                                                   ...              ...   \n",
       "809901  {'ID': 'cds47730', 'Parent': 'rna47730', 'Dbxr...       OCT55143.1   \n",
       "809902  {'ID': 'gene45941', 'Name': 'Xelaev18004691m',...  Xelaev18004691m   \n",
       "809903  {'ID': 'rna47731', 'Parent': 'gene45941', 'Not...             None   \n",
       "809904  {'ID': 'id452950', 'Parent': 'rna47731', 'Note...             None   \n",
       "809905  {'ID': 'cds47731', 'Parent': 'rna47731', 'Dbxr...       OCT55142.1   \n",
       "\n",
       "                Xenbase                                        Dbxref  \\\n",
       "0       XB-GENE-5942444                                          None   \n",
       "1                  None                                          None   \n",
       "2                  None                                          None   \n",
       "3                  None                                          None   \n",
       "4                  None  Phytozome:Xelaev18004747m,NCBI_GP:OCT98948.1   \n",
       "...                 ...                                           ...   \n",
       "809901             None  Phytozome:Xelaev18004680m,NCBI_GP:OCT55143.1   \n",
       "809902             None                                          None   \n",
       "809903             None                                          None   \n",
       "809904             None                                          None   \n",
       "809905             None  Phytozome:Xelaev18004691m,NCBI_GP:OCT55142.1   \n",
       "\n",
       "                                              dbxref_dict         ID  \\\n",
       "0                                                    None      gene0   \n",
       "1                                                    None       rna0   \n",
       "2                                                    None        id1   \n",
       "3                                                    None        id2   \n",
       "4       {'Phytozome': 'Xelaev18004747m', 'NCBI_GP': 'O...       cds0   \n",
       "...                                                   ...        ...   \n",
       "809901  {'Phytozome': 'Xelaev18004680m', 'NCBI_GP': 'O...   cds47730   \n",
       "809902                                               None  gene45941   \n",
       "809903                                               None   rna47731   \n",
       "809904                                               None   id452950   \n",
       "809905  {'Phytozome': 'Xelaev18004691m', 'NCBI_GP': 'O...   cds47731   \n",
       "\n",
       "           Parent  \n",
       "0            None  \n",
       "1           gene0  \n",
       "2            rna0  \n",
       "3            rna0  \n",
       "4            rna0  \n",
       "...           ...  \n",
       "809901   rna47730  \n",
       "809902       None  \n",
       "809903  gene45941  \n",
       "809904   rna47731  \n",
       "809905   rna47731  \n",
       "\n",
       "[809906 rows x 16 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract field and database cross-ref (dbxref) information into columns\n",
    "models['field_dictionary'] = models[attributes_column].apply(convert_fields_to_dict_gff)\n",
    "models['gene_name'] = [d.get('Name') for d in models['field_dictionary']]\n",
    "models['Xenbase'] = [d.get('Alias') for d in models['field_dictionary']]\n",
    "models['Dbxref'] = [d.get('Dbxref') for d in models['field_dictionary']]\n",
    "models['dbxref_dict'] = models['Dbxref'].apply(convert_dbxref_to_dict)\n",
    "\n",
    "display(models)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Xenbase</th>\n",
       "      <th>Dbxref</th>\n",
       "      <th>dbxref_dict</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Name=Xelaev18004747m;end_range=18399,...</td>\n",
       "      <td>{'ID': 'gene0', 'Name': 'Xelaev18004747m', 'en...</td>\n",
       "      <td>Xelaev18004747m</td>\n",
       "      <td>XB-GENE-5942444</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>gene0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Note=transcript XELAEV_18...</td>\n",
       "      <td>{'ID': 'rna0', 'Parent': 'gene0', 'Note': 'tra...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rna0</td>\n",
       "      <td>gene0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "      <td>{'ID': 'id1', 'Parent': 'rna0', 'Note': 'trans...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>id1</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>17924</td>\n",
       "      <td>18243</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "      <td>{'ID': 'id2', 'Parent': 'rna0', 'Note': 'trans...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>id2</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...</td>\n",
       "      <td>{'ID': 'cds0', 'Parent': 'rna0', 'Dbxref': 'Ph...</td>\n",
       "      <td>OCT98948.1</td>\n",
       "      <td>None</td>\n",
       "      <td>Phytozome:Xelaev18004747m,NCBI_GP:OCT98948.1</td>\n",
       "      <td>{'Phytozome': 'Xelaev18004747m', 'NCBI_GP': 'O...</td>\n",
       "      <td>cds0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809901</th>\n",
       "      <td>Scaffold94051</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8</td>\n",
       "      <td>205</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...</td>\n",
       "      <td>{'ID': 'cds47730', 'Parent': 'rna47730', 'Dbxr...</td>\n",
       "      <td>OCT55143.1</td>\n",
       "      <td>None</td>\n",
       "      <td>Phytozome:Xelaev18004680m,NCBI_GP:OCT55143.1</td>\n",
       "      <td>{'Phytozome': 'Xelaev18004680m', 'NCBI_GP': 'O...</td>\n",
       "      <td>cds47730</td>\n",
       "      <td>rna47730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809902</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene45941;Name=Xelaev18004691m;end_range=23...</td>\n",
       "      <td>{'ID': 'gene45941', 'Name': 'Xelaev18004691m',...</td>\n",
       "      <td>Xelaev18004691m</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>gene45941</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809903</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna47731;Parent=gene45941;Note=transcript X...</td>\n",
       "      <td>{'ID': 'rna47731', 'Parent': 'gene45941', 'Not...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rna47731</td>\n",
       "      <td>gene45941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809904</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id452950;Parent=rna47731;Note=transcript XE...</td>\n",
       "      <td>{'ID': 'id452950', 'Parent': 'rna47731', 'Note...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>id452950</td>\n",
       "      <td>rna47731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809905</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...</td>\n",
       "      <td>{'ID': 'cds47731', 'Parent': 'rna47731', 'Dbxr...</td>\n",
       "      <td>OCT55142.1</td>\n",
       "      <td>None</td>\n",
       "      <td>Phytozome:Xelaev18004691m,NCBI_GP:OCT55142.1</td>\n",
       "      <td>{'Phytozome': 'Xelaev18004691m', 'NCBI_GP': 'O...</td>\n",
       "      <td>cds47731</td>\n",
       "      <td>rna47731</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>809906 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0        1     2      3      4  5  6  7  \\\n",
       "0               chr1L  Genbank  gene  17924  18399  .  -  .   \n",
       "1               chr1L  Genbank  mRNA  17924  18399  .  -  .   \n",
       "2               chr1L  Genbank  exon  18336  18399  .  -  .   \n",
       "3               chr1L  Genbank  exon  17924  18243  .  -  .   \n",
       "4               chr1L  Genbank   CDS  18336  18399  .  -  0   \n",
       "...               ...      ...   ...    ...    ... .. .. ..   \n",
       "809901  Scaffold94051  Genbank   CDS      8    205  .  +  0   \n",
       "809902  Scaffold95291  Genbank  gene      9    236  .  -  .   \n",
       "809903  Scaffold95291  Genbank  mRNA      9    236  .  -  .   \n",
       "809904  Scaffold95291  Genbank  exon      9    236  .  -  .   \n",
       "809905  Scaffold95291  Genbank   CDS      9    236  .  -  0   \n",
       "\n",
       "                                                        8  \\\n",
       "0       ID=gene0;Name=Xelaev18004747m;end_range=18399,...   \n",
       "1       ID=rna0;Parent=gene0;Note=transcript XELAEV_18...   \n",
       "2       ID=id1;Parent=rna0;Note=transcript XELAEV_1800...   \n",
       "3       ID=id2;Parent=rna0;Note=transcript XELAEV_1800...   \n",
       "4       ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...   \n",
       "...                                                   ...   \n",
       "809901  ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...   \n",
       "809902  ID=gene45941;Name=Xelaev18004691m;end_range=23...   \n",
       "809903  ID=rna47731;Parent=gene45941;Note=transcript X...   \n",
       "809904  ID=id452950;Parent=rna47731;Note=transcript XE...   \n",
       "809905  ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...   \n",
       "\n",
       "                                         field_dictionary        gene_name  \\\n",
       "0       {'ID': 'gene0', 'Name': 'Xelaev18004747m', 'en...  Xelaev18004747m   \n",
       "1       {'ID': 'rna0', 'Parent': 'gene0', 'Note': 'tra...             None   \n",
       "2       {'ID': 'id1', 'Parent': 'rna0', 'Note': 'trans...             None   \n",
       "3       {'ID': 'id2', 'Parent': 'rna0', 'Note': 'trans...             None   \n",
       "4       {'ID': 'cds0', 'Parent': 'rna0', 'Dbxref': 'Ph...       OCT98948.1   \n",
       "...                                                   ...              ...   \n",
       "809901  {'ID': 'cds47730', 'Parent': 'rna47730', 'Dbxr...       OCT55143.1   \n",
       "809902  {'ID': 'gene45941', 'Name': 'Xelaev18004691m',...  Xelaev18004691m   \n",
       "809903  {'ID': 'rna47731', 'Parent': 'gene45941', 'Not...             None   \n",
       "809904  {'ID': 'id452950', 'Parent': 'rna47731', 'Note...             None   \n",
       "809905  {'ID': 'cds47731', 'Parent': 'rna47731', 'Dbxr...       OCT55142.1   \n",
       "\n",
       "                Xenbase                                        Dbxref  \\\n",
       "0       XB-GENE-5942444                                          None   \n",
       "1                  None                                          None   \n",
       "2                  None                                          None   \n",
       "3                  None                                          None   \n",
       "4                  None  Phytozome:Xelaev18004747m,NCBI_GP:OCT98948.1   \n",
       "...                 ...                                           ...   \n",
       "809901             None  Phytozome:Xelaev18004680m,NCBI_GP:OCT55143.1   \n",
       "809902             None                                          None   \n",
       "809903             None                                          None   \n",
       "809904             None                                          None   \n",
       "809905             None  Phytozome:Xelaev18004691m,NCBI_GP:OCT55142.1   \n",
       "\n",
       "                                              dbxref_dict         ID  \\\n",
       "0                                                    None      gene0   \n",
       "1                                                    None       rna0   \n",
       "2                                                    None        id1   \n",
       "3                                                    None        id2   \n",
       "4       {'Phytozome': 'Xelaev18004747m', 'NCBI_GP': 'O...       cds0   \n",
       "...                                                   ...        ...   \n",
       "809901  {'Phytozome': 'Xelaev18004680m', 'NCBI_GP': 'O...   cds47730   \n",
       "809902                                               None  gene45941   \n",
       "809903                                               None   rna47731   \n",
       "809904                                               None   id452950   \n",
       "809905  {'Phytozome': 'Xelaev18004691m', 'NCBI_GP': 'O...   cds47731   \n",
       "\n",
       "           Parent  \n",
       "0            None  \n",
       "1           gene0  \n",
       "2            rna0  \n",
       "3            rna0  \n",
       "4            rna0  \n",
       "...           ...  \n",
       "809901   rna47730  \n",
       "809902       None  \n",
       "809903  gene45941  \n",
       "809904   rna47731  \n",
       "809905   rna47731  \n",
       "\n",
       "[809906 rows x 16 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### Exception ###\n",
    "# Generate a new .gtf file that includes a gene name in the \"gene\" field for gffread to use\n",
    "models['ID'] = [d.get('ID') for d in models['field_dictionary']]\n",
    "models['Parent'] = [d.get('Parent') for d in models['field_dictionary']]\n",
    "display(models)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneid</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene0</td>\n",
       "      <td>Xelaev18004747m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>gene1</td>\n",
       "      <td>Xetrov90028798m.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>gene2</td>\n",
       "      <td>Xelaev18004749m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>gene3</td>\n",
       "      <td>Xelaev18004750m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>gene4</td>\n",
       "      <td>Xelaev18004751m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809886</th>\n",
       "      <td>gene45937</td>\n",
       "      <td>rpl37.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809890</th>\n",
       "      <td>gene45938</td>\n",
       "      <td>Xelaev18004664m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809894</th>\n",
       "      <td>gene45939</td>\n",
       "      <td>Xelaev18004665m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809898</th>\n",
       "      <td>gene45940</td>\n",
       "      <td>Xelaev18004680m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809902</th>\n",
       "      <td>gene45941</td>\n",
       "      <td>Xelaev18004691m</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>42889 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           geneid          gene_name\n",
       "0           gene0    Xelaev18004747m\n",
       "6           gene1  Xetrov90028798m.L\n",
       "12          gene2    Xelaev18004749m\n",
       "18          gene3    Xelaev18004750m\n",
       "24          gene4    Xelaev18004751m\n",
       "...           ...                ...\n",
       "809886  gene45937            rpl37.S\n",
       "809890  gene45938    Xelaev18004664m\n",
       "809894  gene45939    Xelaev18004665m\n",
       "809898  gene45940    Xelaev18004680m\n",
       "809902  gene45941    Xelaev18004691m\n",
       "\n",
       "[42889 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1937/845663448.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  rna_keys.rename(columns = {'ID': 'rnaid'}, inplace = True)\n",
      "/tmp/ipykernel_1937/845663448.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  id_keys.rename(columns = {'ID': 'idid'}, inplace = True)\n",
      "/tmp/ipykernel_1937/845663448.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  cds_keys.rename(columns = {'ID': 'cdsid'}, inplace = True)\n",
      "/tmp/ipykernel_1937/845663448.py:23: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  keypair_collector = gene_keypairs.append(rna_keypairs).append(id_keypairs).append(cds_keypairs)\n",
      "/tmp/ipykernel_1937/845663448.py:23: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  keypair_collector = gene_keypairs.append(rna_keypairs).append(id_keypairs).append(cds_keypairs)\n",
      "/tmp/ipykernel_1937/845663448.py:23: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  keypair_collector = gene_keypairs.append(rna_keypairs).append(id_keypairs).append(cds_keypairs)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>new_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene0</td>\n",
       "      <td>Xelaev18004747m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>gene1</td>\n",
       "      <td>Xetrov90028798m.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>gene2</td>\n",
       "      <td>Xelaev18004749m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>gene3</td>\n",
       "      <td>Xelaev18004750m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>gene4</td>\n",
       "      <td>Xelaev18004751m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331470</th>\n",
       "      <td>cds47727</td>\n",
       "      <td>rpl37.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331471</th>\n",
       "      <td>cds47728</td>\n",
       "      <td>Xelaev18004664m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331472</th>\n",
       "      <td>cds47729</td>\n",
       "      <td>Xelaev18004665m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331473</th>\n",
       "      <td>cds47730</td>\n",
       "      <td>Xelaev18004680m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331474</th>\n",
       "      <td>cds47731</td>\n",
       "      <td>Xelaev18004691m</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>763385 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID      new_gene_name\n",
       "0          gene0    Xelaev18004747m\n",
       "6          gene1  Xetrov90028798m.L\n",
       "12         gene2    Xelaev18004749m\n",
       "18         gene3    Xelaev18004750m\n",
       "24         gene4    Xelaev18004751m\n",
       "...          ...                ...\n",
       "331470  cds47727            rpl37.S\n",
       "331471  cds47728    Xelaev18004664m\n",
       "331472  cds47729    Xelaev18004665m\n",
       "331473  cds47730    Xelaev18004680m\n",
       "331474  cds47731    Xelaev18004691m\n",
       "\n",
       "[763385 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gene_name_keys = models[['ID', 'gene_name']].dropna(subset = 'gene_name')\n",
    "gene_name_keys = gene_name_keys[~gene_name_keys['gene_name'].str.contains('OCT*.*')]\n",
    "gene_name_keys.rename(columns = {'ID': 'geneid'}, inplace = True)\n",
    "display(gene_name_keys)\n",
    "\n",
    "all_keys = models[['ID', 'Parent']]\n",
    "rna_keys = all_keys[all_keys['ID'].str.contains('rna')]\n",
    "rna_keys.rename(columns = {'ID': 'rnaid'}, inplace = True)\n",
    "id_keys = all_keys[all_keys['ID'].str.contains('id')]\n",
    "id_keys.rename(columns = {'ID': 'idid'}, inplace = True)\n",
    "cds_keys = all_keys[all_keys['ID'].str.contains('cds')]\n",
    "cds_keys.rename(columns = {'ID': 'cdsid'}, inplace = True)\n",
    "\n",
    "key_aggregator = gene_name_keys.merge(rna_keys, left_on = 'geneid', right_on = 'Parent')\n",
    "id_key_aggregator = key_aggregator.merge(id_keys, left_on = 'rnaid', right_on = 'Parent')\n",
    "cds_key_aggregator = key_aggregator.merge(cds_keys, left_on = 'rnaid', right_on = 'Parent')\n",
    "\n",
    "gene_keypairs = gene_name_keys[['geneid', 'gene_name']].rename(columns = {'geneid':'ID'})\n",
    "rna_keypairs = key_aggregator[['rnaid', 'gene_name']].rename(columns = {'rnaid':'ID'})\n",
    "id_keypairs = id_key_aggregator[['idid', 'gene_name']].rename(columns = {'idid':'ID'})\n",
    "cds_keypairs = cds_key_aggregator[['cdsid', 'gene_name']].rename(columns = {'cdsid':'ID'})\n",
    "\n",
    "keypair_collector = gene_keypairs.append(rna_keypairs).append(id_keypairs).append(cds_keypairs)\n",
    "keypair_collector.rename(columns = {'gene_name': 'new_gene_name'}, inplace = True)\n",
    "display(keypair_collector)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Xenbase</th>\n",
       "      <th>Dbxref</th>\n",
       "      <th>dbxref_dict</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>new_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Name=Xelaev18004747m;end_range=18399,...</td>\n",
       "      <td>{'ID': 'gene0', 'Name': 'Xelaev18004747m', 'en...</td>\n",
       "      <td>Xelaev18004747m</td>\n",
       "      <td>XB-GENE-5942444</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>gene0</td>\n",
       "      <td>None</td>\n",
       "      <td>Xelaev18004747m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Note=transcript XELAEV_18...</td>\n",
       "      <td>{'ID': 'rna0', 'Parent': 'gene0', 'Note': 'tra...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rna0</td>\n",
       "      <td>gene0</td>\n",
       "      <td>Xelaev18004747m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "      <td>{'ID': 'id1', 'Parent': 'rna0', 'Note': 'trans...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>id1</td>\n",
       "      <td>rna0</td>\n",
       "      <td>Xelaev18004747m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>17924</td>\n",
       "      <td>18243</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "      <td>{'ID': 'id2', 'Parent': 'rna0', 'Note': 'trans...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>id2</td>\n",
       "      <td>rna0</td>\n",
       "      <td>Xelaev18004747m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...</td>\n",
       "      <td>{'ID': 'cds0', 'Parent': 'rna0', 'Dbxref': 'Ph...</td>\n",
       "      <td>OCT98948.1</td>\n",
       "      <td>None</td>\n",
       "      <td>Phytozome:Xelaev18004747m,NCBI_GP:OCT98948.1</td>\n",
       "      <td>{'Phytozome': 'Xelaev18004747m', 'NCBI_GP': 'O...</td>\n",
       "      <td>cds0</td>\n",
       "      <td>rna0</td>\n",
       "      <td>Xelaev18004747m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6375865</th>\n",
       "      <td>Scaffold94051</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8</td>\n",
       "      <td>205</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...</td>\n",
       "      <td>{'ID': 'cds47730', 'Parent': 'rna47730', 'Dbxr...</td>\n",
       "      <td>OCT55143.1</td>\n",
       "      <td>None</td>\n",
       "      <td>Phytozome:Xelaev18004680m,NCBI_GP:OCT55143.1</td>\n",
       "      <td>{'Phytozome': 'Xelaev18004680m', 'NCBI_GP': 'O...</td>\n",
       "      <td>cds47730</td>\n",
       "      <td>rna47730</td>\n",
       "      <td>Xelaev18004680m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6375866</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene45941;Name=Xelaev18004691m;end_range=23...</td>\n",
       "      <td>{'ID': 'gene45941', 'Name': 'Xelaev18004691m',...</td>\n",
       "      <td>Xelaev18004691m</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>gene45941</td>\n",
       "      <td>None</td>\n",
       "      <td>Xelaev18004691m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6375867</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna47731;Parent=gene45941;Note=transcript X...</td>\n",
       "      <td>{'ID': 'rna47731', 'Parent': 'gene45941', 'Not...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>rna47731</td>\n",
       "      <td>gene45941</td>\n",
       "      <td>Xelaev18004691m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6375868</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id452950;Parent=rna47731;Note=transcript XE...</td>\n",
       "      <td>{'ID': 'id452950', 'Parent': 'rna47731', 'Note...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>id452950</td>\n",
       "      <td>rna47731</td>\n",
       "      <td>Xelaev18004691m</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6375869</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...</td>\n",
       "      <td>{'ID': 'cds47731', 'Parent': 'rna47731', 'Dbxr...</td>\n",
       "      <td>OCT55142.1</td>\n",
       "      <td>None</td>\n",
       "      <td>Phytozome:Xelaev18004691m,NCBI_GP:OCT55142.1</td>\n",
       "      <td>{'Phytozome': 'Xelaev18004691m', 'NCBI_GP': 'O...</td>\n",
       "      <td>cds47731</td>\n",
       "      <td>rna47731</td>\n",
       "      <td>Xelaev18004691m</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6375870 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     0        1     2      3      4  5  6  7  \\\n",
       "0                chr1L  Genbank  gene  17924  18399  .  -  .   \n",
       "1                chr1L  Genbank  mRNA  17924  18399  .  -  .   \n",
       "2                chr1L  Genbank  exon  18336  18399  .  -  .   \n",
       "3                chr1L  Genbank  exon  17924  18243  .  -  .   \n",
       "4                chr1L  Genbank   CDS  18336  18399  .  -  0   \n",
       "...                ...      ...   ...    ...    ... .. .. ..   \n",
       "6375865  Scaffold94051  Genbank   CDS      8    205  .  +  0   \n",
       "6375866  Scaffold95291  Genbank  gene      9    236  .  -  .   \n",
       "6375867  Scaffold95291  Genbank  mRNA      9    236  .  -  .   \n",
       "6375868  Scaffold95291  Genbank  exon      9    236  .  -  .   \n",
       "6375869  Scaffold95291  Genbank   CDS      9    236  .  -  0   \n",
       "\n",
       "                                                         8  \\\n",
       "0        ID=gene0;Name=Xelaev18004747m;end_range=18399,...   \n",
       "1        ID=rna0;Parent=gene0;Note=transcript XELAEV_18...   \n",
       "2        ID=id1;Parent=rna0;Note=transcript XELAEV_1800...   \n",
       "3        ID=id2;Parent=rna0;Note=transcript XELAEV_1800...   \n",
       "4        ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...   \n",
       "...                                                    ...   \n",
       "6375865  ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...   \n",
       "6375866  ID=gene45941;Name=Xelaev18004691m;end_range=23...   \n",
       "6375867  ID=rna47731;Parent=gene45941;Note=transcript X...   \n",
       "6375868  ID=id452950;Parent=rna47731;Note=transcript XE...   \n",
       "6375869  ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...   \n",
       "\n",
       "                                          field_dictionary        gene_name  \\\n",
       "0        {'ID': 'gene0', 'Name': 'Xelaev18004747m', 'en...  Xelaev18004747m   \n",
       "1        {'ID': 'rna0', 'Parent': 'gene0', 'Note': 'tra...             None   \n",
       "2        {'ID': 'id1', 'Parent': 'rna0', 'Note': 'trans...             None   \n",
       "3        {'ID': 'id2', 'Parent': 'rna0', 'Note': 'trans...             None   \n",
       "4        {'ID': 'cds0', 'Parent': 'rna0', 'Dbxref': 'Ph...       OCT98948.1   \n",
       "...                                                    ...              ...   \n",
       "6375865  {'ID': 'cds47730', 'Parent': 'rna47730', 'Dbxr...       OCT55143.1   \n",
       "6375866  {'ID': 'gene45941', 'Name': 'Xelaev18004691m',...  Xelaev18004691m   \n",
       "6375867  {'ID': 'rna47731', 'Parent': 'gene45941', 'Not...             None   \n",
       "6375868  {'ID': 'id452950', 'Parent': 'rna47731', 'Note...             None   \n",
       "6375869  {'ID': 'cds47731', 'Parent': 'rna47731', 'Dbxr...       OCT55142.1   \n",
       "\n",
       "                 Xenbase                                        Dbxref  \\\n",
       "0        XB-GENE-5942444                                          None   \n",
       "1                   None                                          None   \n",
       "2                   None                                          None   \n",
       "3                   None                                          None   \n",
       "4                   None  Phytozome:Xelaev18004747m,NCBI_GP:OCT98948.1   \n",
       "...                  ...                                           ...   \n",
       "6375865             None  Phytozome:Xelaev18004680m,NCBI_GP:OCT55143.1   \n",
       "6375866             None                                          None   \n",
       "6375867             None                                          None   \n",
       "6375868             None                                          None   \n",
       "6375869             None  Phytozome:Xelaev18004691m,NCBI_GP:OCT55142.1   \n",
       "\n",
       "                                               dbxref_dict         ID  \\\n",
       "0                                                     None      gene0   \n",
       "1                                                     None       rna0   \n",
       "2                                                     None        id1   \n",
       "3                                                     None        id2   \n",
       "4        {'Phytozome': 'Xelaev18004747m', 'NCBI_GP': 'O...       cds0   \n",
       "...                                                    ...        ...   \n",
       "6375865  {'Phytozome': 'Xelaev18004680m', 'NCBI_GP': 'O...   cds47730   \n",
       "6375866                                               None  gene45941   \n",
       "6375867                                               None   rna47731   \n",
       "6375868                                               None   id452950   \n",
       "6375869  {'Phytozome': 'Xelaev18004691m', 'NCBI_GP': 'O...   cds47731   \n",
       "\n",
       "            Parent    new_gene_name  \n",
       "0             None  Xelaev18004747m  \n",
       "1            gene0  Xelaev18004747m  \n",
       "2             rna0  Xelaev18004747m  \n",
       "3             rna0  Xelaev18004747m  \n",
       "4             rna0  Xelaev18004747m  \n",
       "...            ...              ...  \n",
       "6375865   rna47730  Xelaev18004680m  \n",
       "6375866       None  Xelaev18004691m  \n",
       "6375867  gene45941  Xelaev18004691m  \n",
       "6375868   rna47731  Xelaev18004691m  \n",
       "6375869   rna47731  Xelaev18004691m  \n",
       "\n",
       "[6375870 rows x 17 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Name=Xelaev18004747m;end_range=18399,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Note=transcript XELAEV_18...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>17924</td>\n",
       "      <td>18243</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6375865</th>\n",
       "      <td>Scaffold94051</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8</td>\n",
       "      <td>205</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6375866</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene45941;Name=Xelaev18004691m;end_range=23...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6375867</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna47731;Parent=gene45941;Note=transcript X...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6375868</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id452950;Parent=rna47731;Note=transcript XE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6375869</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6375870 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     0        1     2      3      4  5  6  7  \\\n",
       "0                chr1L  Genbank  gene  17924  18399  .  -  .   \n",
       "1                chr1L  Genbank  mRNA  17924  18399  .  -  .   \n",
       "2                chr1L  Genbank  exon  18336  18399  .  -  .   \n",
       "3                chr1L  Genbank  exon  17924  18243  .  -  .   \n",
       "4                chr1L  Genbank   CDS  18336  18399  .  -  0   \n",
       "...                ...      ...   ...    ...    ... .. .. ..   \n",
       "6375865  Scaffold94051  Genbank   CDS      8    205  .  +  0   \n",
       "6375866  Scaffold95291  Genbank  gene      9    236  .  -  .   \n",
       "6375867  Scaffold95291  Genbank  mRNA      9    236  .  -  .   \n",
       "6375868  Scaffold95291  Genbank  exon      9    236  .  -  .   \n",
       "6375869  Scaffold95291  Genbank   CDS      9    236  .  -  0   \n",
       "\n",
       "                                                         8  \n",
       "0        ID=gene0;Name=Xelaev18004747m;end_range=18399,...  \n",
       "1        ID=rna0;Parent=gene0;Note=transcript XELAEV_18...  \n",
       "2        ID=id1;Parent=rna0;Note=transcript XELAEV_1800...  \n",
       "3        ID=id2;Parent=rna0;Note=transcript XELAEV_1800...  \n",
       "4        ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...  \n",
       "...                                                    ...  \n",
       "6375865  ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...  \n",
       "6375866  ID=gene45941;Name=Xelaev18004691m;end_range=23...  \n",
       "6375867  ID=rna47731;Parent=gene45941;Note=transcript X...  \n",
       "6375868  ID=id452950;Parent=rna47731;Note=transcript XE...  \n",
       "6375869  ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...  \n",
       "\n",
       "[6375870 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'ID=gene0;Name=Xelaev18004747m;end_range=18399,.;gbkey=Gene;gene_biotype=protein_coding;locus_tag=XELAEV_18004747mg;partial=true;start_range=.,17924;Alias=XB-GENE-5942444;gene=Xelaev18004747m'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "models = models.merge(keypair_collector, on = 'ID', how = 'left')\n",
    "\n",
    "models.apply(lambda x: x['field_dictionary'].update({'gene': x['new_gene_name']}), axis = 1)\n",
    "display(models)\n",
    "models.dropna(subset = ['field_dictionary'], inplace = True)\n",
    "models[attributes_column] = models['field_dictionary'].apply(convert_dict_to_fields_gff)\n",
    "\n",
    "new_annot = GenomeGffFile(\n",
    "    filename = sample_Docket.annot.filename.replace('.gff', '_new.gff'),\n",
    "    sampledict = species_SampleDict,\n",
    "    GenomeFastaFile = genome_fasta,\n",
    ")\n",
    "\n",
    "sample_Docket.add_keyfile(new_annot, 'new_annot')\n",
    "new_models = models[[0, 1, 2, 3, 4, 5, 6, 7, 8]]\n",
    "new_models.to_csv(sample_Docket.new_annot.path, sep = '\\t', header = None, index = None)\n",
    "\n",
    "display(new_models)\n",
    "display(new_models[attributes_column][0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Extract gene IDs for mapping to UniprotKB\n",
    "Specify which set of identifiers will be use to query the [Uniprot ID Mapping Tool](https://www.uniprot.org/id-mapping) via API.\n",
    "\n",
    "If using an identifier from the `dbxref_dict`, specify the name via string in the `dbxref_datafield` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Xenbase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Xelaev18004747m</td>\n",
       "      <td>XB-GENE-5942444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Xelaev18004750m</td>\n",
       "      <td>XB-GENE-5877070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Xelaev18004751m</td>\n",
       "      <td>XB-GENE-5758410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>loc652493.L</td>\n",
       "      <td>XB-GENE-5833449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>loc100130100.L</td>\n",
       "      <td>XB-GENE-6050773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6375130</th>\n",
       "      <td>trex2.L</td>\n",
       "      <td>XB-GENE-17336984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6375530</th>\n",
       "      <td>nmral1.L</td>\n",
       "      <td>XB-GENE-17339003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6375810</th>\n",
       "      <td>rps24.L</td>\n",
       "      <td>XB-GENE-967651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6375826</th>\n",
       "      <td>rpl37a.S</td>\n",
       "      <td>XB-GENE-17331914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6375850</th>\n",
       "      <td>rpl37.S</td>\n",
       "      <td>XB-GENE-17331711</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22002 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_name           Xenbase\n",
       "0        Xelaev18004747m   XB-GENE-5942444\n",
       "24       Xelaev18004750m   XB-GENE-5877070\n",
       "32       Xelaev18004751m   XB-GENE-5758410\n",
       "98           loc652493.L   XB-GENE-5833449\n",
       "110       loc100130100.L   XB-GENE-6050773\n",
       "...                  ...               ...\n",
       "6375130          trex2.L  XB-GENE-17336984\n",
       "6375530         nmral1.L  XB-GENE-17339003\n",
       "6375810          rps24.L    XB-GENE-967651\n",
       "6375826         rpl37a.S  XB-GENE-17331914\n",
       "6375850          rpl37.S  XB-GENE-17331711\n",
       "\n",
       "[22002 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dbxref_datafield = ''\n",
    "datafield = 'Xenbase'\n",
    "\n",
    "if dbxref_datafield != '':\n",
    "    models.dropna(axis = 0, subset = ['dbxref_dict'], inplace = True)\n",
    "    models[dbxref_datafield] = [d.get(dbxref_datafield) for d in models['dbxref_dict']]\n",
    "\n",
    "    models_subset = models[['gene_name', dbxref_datafield]].dropna().drop_duplicates()\n",
    "    display(models_subset)\n",
    "\n",
    "elif datafield == 'gene_name':\n",
    "    models_subset = models[['gene_name']].dropna().drop_duplicates()\n",
    "    display(models_subset)\n",
    "\n",
    "elif datafield != '':\n",
    "    models_subset = models[['gene_name', datafield]].dropna().drop_duplicates()\n",
    "    display(models_subset)\n",
    "\n",
    "else:\n",
    "    raise Exception('You must provide a data field for ID mapping.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Generate gene list file to query Uniprot ID Mapping API\n",
    "Generate a text file ending in `_ids.txt` for sending to the ID mapping API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 22001 gene ids to ../../output/Xlae_adultbrain/Xlae_adultbrain_Xenbase_ids.txt\n"
     ]
    }
   ],
   "source": [
    "datafield = dbxref_datafield if dbxref_datafield != '' else datafield\n",
    "\n",
    "gene_list = models_subset[datafield].unique()\n",
    "\n",
    "genelist_object = GeneListFile(\n",
    "    filename = '', # filename is automatically generated from sampledict\n",
    "    sampledict = species_SampleDict,\n",
    "    sources = [sample_Docket.annot],\n",
    "    genes = gene_list,\n",
    "    identifier = datafield\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Query Uniprot ID Mapping API\n",
    "Specify the `from_type` variable based on the Uniprot name of the identifier.\n",
    "The table below lists some databases and the `from_type` string that the API accepts for that datatype.\n",
    "\n",
    "For some reason, the ID mapping API usually returns \"resource not found\" the first time you use it.\n",
    "Try re-running this cell if you receive that error.\n",
    "\n",
    "\n",
    "| datatype | `from_type` string | description |\n",
    "| ---: | :--- | :--- |\n",
    "| Mouse Genome Informatics | `MGI` | ID starts with `MGI:` |\n",
    "| Zebrafish Information Network | `ZFIN` | ID starts with `ZDB-GENE-` |\n",
    "| NCBI Gene ID | `GeneID` | varies, usually numeric string |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  349k    0    52  100  348k     27   187k  0:00:01  0:00:01 --:--:--  187k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100    24    0    24    0     0     37      0 --:--:-- --:--:-- --:--:--    37\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>From</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XB-GENE-865154</td>\n",
       "      <td>Q63ZJ1</td>\n",
       "      <td>Q63ZJ1_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>LOC494796 protein (succinate-CoA ligase, alpha...</td>\n",
       "      <td>suclg1.L galpha LOC494796 mtdps9 sucla1 suclg1...</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XB-GENE-17336652</td>\n",
       "      <td>A0A1L8HSE5</td>\n",
       "      <td>A0A1L8HSE5_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>leucine-rich repeat transmembrane neuronal pro...</td>\n",
       "      <td>lrrtm1.L</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XB-GENE-5955229</td>\n",
       "      <td>Q6GLP0</td>\n",
       "      <td>CTNA2_XENLA</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Catenin alpha-2 (Alpha N-catenin)</td>\n",
       "      <td>ctnna2</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XB-GENE-17338166</td>\n",
       "      <td>A0A1L8HSD7</td>\n",
       "      <td>A0A1L8HSD7_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Alpha-1D adrenergic receptor (Alpha-1D adrenor...</td>\n",
       "      <td>adra1d.L</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XB-GENE-17330544</td>\n",
       "      <td>A0A1L8HSE9</td>\n",
       "      <td>A0A1L8HSE9_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>spermine oxidase isoform X1</td>\n",
       "      <td>smox.L</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18678</th>\n",
       "      <td>XB-GENE-865099</td>\n",
       "      <td>O42569</td>\n",
       "      <td>SOX2_XENLA</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Transcription factor Sox-2 (XSox2) (XlSox-2) (...</td>\n",
       "      <td>sox2</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18679</th>\n",
       "      <td>XB-GENE-865099</td>\n",
       "      <td>A0A4P2X5R0</td>\n",
       "      <td>A0A4P2X5R0_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Transcription factor Sox2.L</td>\n",
       "      <td>sox2.L</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18680</th>\n",
       "      <td>XB-GENE-977569</td>\n",
       "      <td>Q5PQ23</td>\n",
       "      <td>ODFP2_XENLA</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Outer dense fiber protein 2 (Cenexin) (Outer d...</td>\n",
       "      <td>odf2</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18681</th>\n",
       "      <td>XB-GENE-967651</td>\n",
       "      <td>P02377</td>\n",
       "      <td>RS24_XENLA</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>40S ribosomal protein S24 (S19)</td>\n",
       "      <td>rps24</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18682</th>\n",
       "      <td>XB-GENE-17331711</td>\n",
       "      <td>Q6GN79</td>\n",
       "      <td>Q6GN79_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Ribosomal protein L37</td>\n",
       "      <td>rpl37.S MGC82973 rpl37</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18683 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   From       Entry        Entry Name    Reviewed  \\\n",
       "0        XB-GENE-865154      Q63ZJ1      Q63ZJ1_XENLA  unreviewed   \n",
       "1      XB-GENE-17336652  A0A1L8HSE5  A0A1L8HSE5_XENLA  unreviewed   \n",
       "2       XB-GENE-5955229      Q6GLP0       CTNA2_XENLA    reviewed   \n",
       "3      XB-GENE-17338166  A0A1L8HSD7  A0A1L8HSD7_XENLA  unreviewed   \n",
       "4      XB-GENE-17330544  A0A1L8HSE9  A0A1L8HSE9_XENLA  unreviewed   \n",
       "...                 ...         ...               ...         ...   \n",
       "18678    XB-GENE-865099      O42569        SOX2_XENLA    reviewed   \n",
       "18679    XB-GENE-865099  A0A4P2X5R0  A0A4P2X5R0_XENLA  unreviewed   \n",
       "18680    XB-GENE-977569      Q5PQ23       ODFP2_XENLA    reviewed   \n",
       "18681    XB-GENE-967651      P02377        RS24_XENLA    reviewed   \n",
       "18682  XB-GENE-17331711      Q6GN79      Q6GN79_XENLA  unreviewed   \n",
       "\n",
       "                                           Protein names  \\\n",
       "0      LOC494796 protein (succinate-CoA ligase, alpha...   \n",
       "1      leucine-rich repeat transmembrane neuronal pro...   \n",
       "2                      Catenin alpha-2 (Alpha N-catenin)   \n",
       "3      Alpha-1D adrenergic receptor (Alpha-1D adrenor...   \n",
       "4                            spermine oxidase isoform X1   \n",
       "...                                                  ...   \n",
       "18678  Transcription factor Sox-2 (XSox2) (XlSox-2) (...   \n",
       "18679                        Transcription factor Sox2.L   \n",
       "18680  Outer dense fiber protein 2 (Cenexin) (Outer d...   \n",
       "18681                    40S ribosomal protein S24 (S19)   \n",
       "18682                              Ribosomal protein L37   \n",
       "\n",
       "                                              Gene Names  \\\n",
       "0      suclg1.L galpha LOC494796 mtdps9 sucla1 suclg1...   \n",
       "1                                               lrrtm1.L   \n",
       "2                                                 ctnna2   \n",
       "3                                               adra1d.L   \n",
       "4                                                 smox.L   \n",
       "...                                                  ...   \n",
       "18678                                               sox2   \n",
       "18679                                             sox2.L   \n",
       "18680                                               odf2   \n",
       "18681                                              rps24   \n",
       "18682                             rpl37.S MGC82973 rpl37   \n",
       "\n",
       "                                   Organism  Length  \n",
       "0      Xenopus laevis (African clawed frog)     195  \n",
       "1      Xenopus laevis (African clawed frog)     521  \n",
       "2      Xenopus laevis (African clawed frog)     966  \n",
       "3      Xenopus laevis (African clawed frog)     501  \n",
       "4      Xenopus laevis (African clawed frog)     538  \n",
       "...                                     ...     ...  \n",
       "18678  Xenopus laevis (African clawed frog)     311  \n",
       "18679  Xenopus laevis (African clawed frog)     311  \n",
       "18680  Xenopus laevis (African clawed frog)     649  \n",
       "18681  Xenopus laevis (African clawed frog)     132  \n",
       "18682  Xenopus laevis (African clawed frog)      97  \n",
       "\n",
       "[18683 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from_type = 'Xenbase'\n",
    "to_type = 'UniProtKB'\n",
    "\n",
    "uniprot_idmm_object = genelist_object.get_uniprot_ids(ID_MAPPER_LOC, from_type, to_type)\n",
    "sample_Docket.add_keyfiles({'uniprot_idmm': uniprot_idmm_object})\n",
    "\n",
    "uniprot_idmm = pd.read_csv(sample_Docket.uniprot_idmm.path, sep = '\\t')\n",
    "display(uniprot_idmm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Extract results and generate Uniprot IDMM\n",
    "Generates an idmm that links `gene_name`, the `dbxref_datafield` seleted above, and `uniprot_id` returned by API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1937/899271637.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  uniprot_idpairs.rename(columns = {'From': datafield, 'Entry': 'uniprot_id'}, inplace = True)\n",
      "/tmp/ipykernel_1937/899271637.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  uniprot_idpairs[datafield] = uniprot_idpairs[datafield].astype(str)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Xenbase</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XB-GENE-865154</td>\n",
       "      <td>Q63ZJ1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XB-GENE-17336652</td>\n",
       "      <td>A0A1L8HSE5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XB-GENE-5955229</td>\n",
       "      <td>Q6GLP0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XB-GENE-17338166</td>\n",
       "      <td>A0A1L8HSD7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XB-GENE-17330544</td>\n",
       "      <td>A0A1L8HSE9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18678</th>\n",
       "      <td>XB-GENE-865099</td>\n",
       "      <td>O42569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18679</th>\n",
       "      <td>XB-GENE-865099</td>\n",
       "      <td>A0A4P2X5R0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18680</th>\n",
       "      <td>XB-GENE-977569</td>\n",
       "      <td>Q5PQ23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18681</th>\n",
       "      <td>XB-GENE-967651</td>\n",
       "      <td>P02377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18682</th>\n",
       "      <td>XB-GENE-17331711</td>\n",
       "      <td>Q6GN79</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18683 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Xenbase  uniprot_id\n",
       "0        XB-GENE-865154      Q63ZJ1\n",
       "1      XB-GENE-17336652  A0A1L8HSE5\n",
       "2       XB-GENE-5955229      Q6GLP0\n",
       "3      XB-GENE-17338166  A0A1L8HSD7\n",
       "4      XB-GENE-17330544  A0A1L8HSE9\n",
       "...                 ...         ...\n",
       "18678    XB-GENE-865099      O42569\n",
       "18679    XB-GENE-865099  A0A4P2X5R0\n",
       "18680    XB-GENE-977569      Q5PQ23\n",
       "18681    XB-GENE-967651      P02377\n",
       "18682  XB-GENE-17331711      Q6GN79\n",
       "\n",
       "[18683 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Xenbase</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>suclg1.L</td>\n",
       "      <td>XB-GENE-865154</td>\n",
       "      <td>Q63ZJ1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lrrtm1.L</td>\n",
       "      <td>XB-GENE-17336652</td>\n",
       "      <td>A0A1L8HSE5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ctnna2.L</td>\n",
       "      <td>XB-GENE-5955229</td>\n",
       "      <td>Q6GLP0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>adra1d.L</td>\n",
       "      <td>XB-GENE-17338166</td>\n",
       "      <td>A0A1L8HSD7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>smox.L</td>\n",
       "      <td>XB-GENE-17330544</td>\n",
       "      <td>A0A1L8HSE9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18679</th>\n",
       "      <td>sox2.S</td>\n",
       "      <td>XB-GENE-865099</td>\n",
       "      <td>O42569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18680</th>\n",
       "      <td>sox2.S</td>\n",
       "      <td>XB-GENE-865099</td>\n",
       "      <td>A0A4P2X5R0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18681</th>\n",
       "      <td>odf2.L</td>\n",
       "      <td>XB-GENE-977569</td>\n",
       "      <td>Q5PQ23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18682</th>\n",
       "      <td>rps24.L</td>\n",
       "      <td>XB-GENE-967651</td>\n",
       "      <td>P02377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18683</th>\n",
       "      <td>rpl37.S</td>\n",
       "      <td>XB-GENE-17331711</td>\n",
       "      <td>Q6GN79</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18684 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name           Xenbase  uniprot_id\n",
       "0      suclg1.L    XB-GENE-865154      Q63ZJ1\n",
       "1      lrrtm1.L  XB-GENE-17336652  A0A1L8HSE5\n",
       "2      ctnna2.L   XB-GENE-5955229      Q6GLP0\n",
       "3      adra1d.L  XB-GENE-17338166  A0A1L8HSD7\n",
       "4        smox.L  XB-GENE-17330544  A0A1L8HSE9\n",
       "...         ...               ...         ...\n",
       "18679    sox2.S    XB-GENE-865099      O42569\n",
       "18680    sox2.S    XB-GENE-865099  A0A4P2X5R0\n",
       "18681    odf2.L    XB-GENE-977569      Q5PQ23\n",
       "18682   rps24.L    XB-GENE-967651      P02377\n",
       "18683   rpl37.S  XB-GENE-17331711      Q6GN79\n",
       "\n",
       "[18684 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "uniprot_idpairs = uniprot_idmm[['From', 'Entry']]\n",
    "uniprot_idpairs.rename(columns = {'From': datafield, 'Entry': 'uniprot_id'}, inplace = True)\n",
    "uniprot_idpairs[datafield] = uniprot_idpairs[datafield].astype(str)\n",
    "display(uniprot_idpairs)\n",
    "\n",
    "uniprot_output_idmm = models_subset.merge(uniprot_idpairs, on = datafield)\n",
    "display(uniprot_output_idmm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate a filename and file for the idmm\n",
    "uniprot_output_idmm_filename = '_'.join([species_prefix, conditions, 'uniprot-idmm.tsv'])\n",
    "uniprot_output_idmm_object = IdmmFile(uniprot_output_idmm_filename, species_SampleDict, kind = 'uniprot_idmm', sources = [sample_Docket.annot])\n",
    "\n",
    "# save to file and add to the Docket\n",
    "uniprot_output_idmm.to_csv(uniprot_output_idmm_object.path, sep = '\\t')\n",
    "sample_Docket.add_keyfile(uniprot_output_idmm_object, 'uniprot_idmm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8. Convert GFF to GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert the GFF file to GTF using gffread\n",
    "models_asgtf = sample_Docket.new_annot.to_gtf(GFFREAD_LOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>17924</td>\n",
       "      <td>18243</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>17924</td>\n",
       "      <td>18243</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763959</th>\n",
       "      <td>Scaffold94051</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>8</td>\n",
       "      <td>205</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna47730\"; gene_id \"gene45940\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763960</th>\n",
       "      <td>Scaffold94051</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8</td>\n",
       "      <td>205</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"rna47730\"; gene_id \"gene45940\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763961</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>transcript</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna47731\"; gene_id \"gene45941\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763962</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna47731\"; gene_id \"gene45941\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763963</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"rna47731\"; gene_id \"gene45941\";...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>763964 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0        1           2      3      4  5  6  7  \\\n",
       "0               chr1L  Genbank  transcript  17924  18399  .  -  .   \n",
       "1               chr1L  Genbank        exon  17924  18243  .  -  .   \n",
       "2               chr1L  Genbank        exon  18336  18399  .  -  .   \n",
       "3               chr1L  Genbank         CDS  17924  18243  .  -  2   \n",
       "4               chr1L  Genbank         CDS  18336  18399  .  -  0   \n",
       "...               ...      ...         ...    ...    ... .. .. ..   \n",
       "763959  Scaffold94051  Genbank        exon      8    205  .  +  .   \n",
       "763960  Scaffold94051  Genbank         CDS      8    205  .  +  0   \n",
       "763961  Scaffold95291  Genbank  transcript      9    236  .  -  .   \n",
       "763962  Scaffold95291  Genbank        exon      9    236  .  -  .   \n",
       "763963  Scaffold95291  Genbank         CDS      9    236  .  -  0   \n",
       "\n",
       "                                                        8  \n",
       "0       transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "1       transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "2       transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "3       transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "4       transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "...                                                   ...  \n",
       "763959  transcript_id \"rna47730\"; gene_id \"gene45940\";...  \n",
       "763960  transcript_id \"rna47730\"; gene_id \"gene45940\";...  \n",
       "763961  transcript_id \"rna47731\"; gene_id \"gene45941\";...  \n",
       "763962  transcript_id \"rna47731\"; gene_id \"gene45941\";...  \n",
       "763963  transcript_id \"rna47731\"; gene_id \"gene45941\";...  \n",
       "\n",
       "[763964 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'transcript_id \"rna0\"; gene_id \"gene0\"; gene_name \"Xelaev18004747m\";'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load the newly-generated GTF file as a dataframe\n",
    "models_asgtf_df = pd.read_csv(models_asgtf.path, skiprows = 0, header = None, sep = '\\t')\n",
    "\n",
    "display(models_asgtf_df)\n",
    "display(models_asgtf_df[attributes_column][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xelaev18004747m</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>17924</td>\n",
       "      <td>18243</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xelaev18004747m</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xelaev18004747m</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>transcript</td>\n",
       "      <td>35021</td>\n",
       "      <td>35381</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna1\"; gene_id \"gene1\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna1', 'gene_id': 'gene1', ...</td>\n",
       "      <td>Xetrov90028798m.L</td>\n",
       "      <td>gene1</td>\n",
       "      <td>rna1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>35021</td>\n",
       "      <td>35187</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna1\"; gene_id \"gene1\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna1', 'gene_id': 'gene1', ...</td>\n",
       "      <td>Xetrov90028798m.L</td>\n",
       "      <td>gene1</td>\n",
       "      <td>rna1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763956</th>\n",
       "      <td>Scaffold93892</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>2</td>\n",
       "      <td>241</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna47729\"; gene_id \"gene45939\";...</td>\n",
       "      <td>{'transcript_id': 'rna47729', 'gene_id': 'gene...</td>\n",
       "      <td>Xelaev18004665m</td>\n",
       "      <td>gene45939</td>\n",
       "      <td>rna47729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763958</th>\n",
       "      <td>Scaffold94051</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>transcript</td>\n",
       "      <td>8</td>\n",
       "      <td>205</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna47730\"; gene_id \"gene45940\";...</td>\n",
       "      <td>{'transcript_id': 'rna47730', 'gene_id': 'gene...</td>\n",
       "      <td>Xelaev18004680m</td>\n",
       "      <td>gene45940</td>\n",
       "      <td>rna47730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763959</th>\n",
       "      <td>Scaffold94051</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>8</td>\n",
       "      <td>205</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna47730\"; gene_id \"gene45940\";...</td>\n",
       "      <td>{'transcript_id': 'rna47730', 'gene_id': 'gene...</td>\n",
       "      <td>Xelaev18004680m</td>\n",
       "      <td>gene45940</td>\n",
       "      <td>rna47730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763961</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>transcript</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna47731\"; gene_id \"gene45941\";...</td>\n",
       "      <td>{'transcript_id': 'rna47731', 'gene_id': 'gene...</td>\n",
       "      <td>Xelaev18004691m</td>\n",
       "      <td>gene45941</td>\n",
       "      <td>rna47731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763962</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna47731\"; gene_id \"gene45941\";...</td>\n",
       "      <td>{'transcript_id': 'rna47731', 'gene_id': 'gene...</td>\n",
       "      <td>Xelaev18004691m</td>\n",
       "      <td>gene45941</td>\n",
       "      <td>rna47731</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>412823 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0        1           2      3      4  5  6  7  \\\n",
       "0               chr1L  Genbank  transcript  17924  18399  .  -  .   \n",
       "1               chr1L  Genbank        exon  17924  18243  .  -  .   \n",
       "2               chr1L  Genbank        exon  18336  18399  .  -  .   \n",
       "5               chr1L  Genbank  transcript  35021  35381  .  -  .   \n",
       "6               chr1L  Genbank        exon  35021  35187  .  -  .   \n",
       "...               ...      ...         ...    ...    ... .. .. ..   \n",
       "763956  Scaffold93892  Genbank        exon      2    241  .  +  .   \n",
       "763958  Scaffold94051  Genbank  transcript      8    205  .  +  .   \n",
       "763959  Scaffold94051  Genbank        exon      8    205  .  +  .   \n",
       "763961  Scaffold95291  Genbank  transcript      9    236  .  -  .   \n",
       "763962  Scaffold95291  Genbank        exon      9    236  .  -  .   \n",
       "\n",
       "                                                        8  \\\n",
       "0       transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "1       transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "2       transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "5       transcript_id \"rna1\"; gene_id \"gene1\"; gene_na...   \n",
       "6       transcript_id \"rna1\"; gene_id \"gene1\"; gene_na...   \n",
       "...                                                   ...   \n",
       "763956  transcript_id \"rna47729\"; gene_id \"gene45939\";...   \n",
       "763958  transcript_id \"rna47730\"; gene_id \"gene45940\";...   \n",
       "763959  transcript_id \"rna47730\"; gene_id \"gene45940\";...   \n",
       "763961  transcript_id \"rna47731\"; gene_id \"gene45941\";...   \n",
       "763962  transcript_id \"rna47731\"; gene_id \"gene45941\";...   \n",
       "\n",
       "                                         field_dictionary          gene_name  \\\n",
       "0       {'transcript_id': 'rna0', 'gene_id': 'gene0', ...    Xelaev18004747m   \n",
       "1       {'transcript_id': 'rna0', 'gene_id': 'gene0', ...    Xelaev18004747m   \n",
       "2       {'transcript_id': 'rna0', 'gene_id': 'gene0', ...    Xelaev18004747m   \n",
       "5       {'transcript_id': 'rna1', 'gene_id': 'gene1', ...  Xetrov90028798m.L   \n",
       "6       {'transcript_id': 'rna1', 'gene_id': 'gene1', ...  Xetrov90028798m.L   \n",
       "...                                                   ...                ...   \n",
       "763956  {'transcript_id': 'rna47729', 'gene_id': 'gene...    Xelaev18004665m   \n",
       "763958  {'transcript_id': 'rna47730', 'gene_id': 'gene...    Xelaev18004680m   \n",
       "763959  {'transcript_id': 'rna47730', 'gene_id': 'gene...    Xelaev18004680m   \n",
       "763961  {'transcript_id': 'rna47731', 'gene_id': 'gene...    Xelaev18004691m   \n",
       "763962  {'transcript_id': 'rna47731', 'gene_id': 'gene...    Xelaev18004691m   \n",
       "\n",
       "          gene_id transcript_id  \n",
       "0           gene0          rna0  \n",
       "1           gene0          rna0  \n",
       "2           gene0          rna0  \n",
       "5           gene1          rna1  \n",
       "6           gene1          rna1  \n",
       "...           ...           ...  \n",
       "763956  gene45939      rna47729  \n",
       "763958  gene45940      rna47730  \n",
       "763959  gene45940      rna47730  \n",
       "763961  gene45941      rna47731  \n",
       "763962  gene45941      rna47731  \n",
       "\n",
       "[412823 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Use a custom function to extract useful fields from the additional fields section (column 8)\n",
    "# Pull from that dict to fill in additional useful columns\n",
    "models_asgtf_df['field_dictionary'] = models_asgtf_df[attributes_column].apply(convert_fields_to_dict_gtf)\n",
    "models_asgtf_df['gene_name'] = [d.get('gene_name') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['gene_id'] = [d.get('gene_id') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['transcript_id'] = [d.get('transcript_id') for d in models_asgtf_df['field_dictionary']]\n",
    "\n",
    "# Remove CDS annotations because they interfere with TransDecoder cDNA generation\n",
    "models_asgtf_df = models_asgtf_df[models_asgtf_df[2] != 'CDS']\n",
    "display(models_asgtf_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9. Generate gtf-idmm\n",
    "This file maps the `gene_name` to `gene_id` and `transcript_id` fields generated by the conversion from GFF to GTF, which will be needed for downstream processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Xelaev18004747m</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Xetrov90028798m.L</td>\n",
       "      <td>gene1</td>\n",
       "      <td>rna1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Xelaev18004749m</td>\n",
       "      <td>gene2</td>\n",
       "      <td>rna2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Xelaev18004750m</td>\n",
       "      <td>gene3</td>\n",
       "      <td>rna3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Xelaev18004751m</td>\n",
       "      <td>gene4</td>\n",
       "      <td>rna4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763949</th>\n",
       "      <td>rpl37.S</td>\n",
       "      <td>gene45937</td>\n",
       "      <td>rna47727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763952</th>\n",
       "      <td>Xelaev18004664m</td>\n",
       "      <td>gene45938</td>\n",
       "      <td>rna47728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763955</th>\n",
       "      <td>Xelaev18004665m</td>\n",
       "      <td>gene45939</td>\n",
       "      <td>rna47729</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763958</th>\n",
       "      <td>Xelaev18004680m</td>\n",
       "      <td>gene45940</td>\n",
       "      <td>rna47730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>763961</th>\n",
       "      <td>Xelaev18004691m</td>\n",
       "      <td>gene45941</td>\n",
       "      <td>rna47731</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>47732 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_name    gene_id transcript_id\n",
       "0         Xelaev18004747m      gene0          rna0\n",
       "5       Xetrov90028798m.L      gene1          rna1\n",
       "10        Xelaev18004749m      gene2          rna2\n",
       "15        Xelaev18004750m      gene3          rna3\n",
       "20        Xelaev18004751m      gene4          rna4\n",
       "...                   ...        ...           ...\n",
       "763949            rpl37.S  gene45937      rna47727\n",
       "763952    Xelaev18004664m  gene45938      rna47728\n",
       "763955    Xelaev18004665m  gene45939      rna47729\n",
       "763958    Xelaev18004680m  gene45940      rna47730\n",
       "763961    Xelaev18004691m  gene45941      rna47731\n",
       "\n",
       "[47732 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract gene_name, gene_id, and transcript_id fields to generate an ID mapping matrix (idmm)\n",
    "idmm_df = models_asgtf_df[['gene_name', 'gene_id', 'transcript_id']].drop_duplicates()\n",
    "idmm_df.dropna(inplace = True)\n",
    "display(idmm_df)\n",
    "\n",
    "# generate a filename and file for the idmm\n",
    "idmm_filename = '_'.join([species_prefix, conditions, 'gtf-idmm.tsv'])\n",
    "idmm = IdmmFile(idmm_filename, species_SampleDict, kind = 'gtf_idmm', sources = [sample_Docket.annot])\n",
    "\n",
    "# save to file and add to the Docket\n",
    "idmm_df.to_csv(idmm.path, sep = '\\t')\n",
    "sample_Docket.add_keyfile(idmm, 'gtf_idmm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 10. Generate updated gtf\n",
    "Generated an updated GTF file using transcript_id as the key. For some datasets, transcripts do not consistently get gene names and gene IDs added, which causes Transdecoder to throw errors. This resolves that problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "models_asgtf_updated_df = models_asgtf_df.merge(idmm_df, on = 'transcript_id')\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_name': x['gene_name_y']}), axis = 1)\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_id': x['gene_id_y']}), axis = 1)\n",
    "models_asgtf_updated_df['new_fields'] = models_asgtf_updated_df['field_dictionary'].apply(convert_dict_to_fields_gtf)\n",
    "models_asgtf_updated_df = models_asgtf_updated_df[[0, 1, 2, 3, 4, 5, 6, 7, 'new_fields']]\n",
    "\n",
    "models_asgtf_updated_filename = models_asgtf.filename.replace('.gtf', '_updated.gtf')\n",
    "models_asgtf_updated = GenomeGtfFile(models_asgtf_updated_filename, species_SampleDict, GenomeFastaFile = sample_Docket.genome_fasta)\n",
    "\n",
    "models_asgtf_updated_df.to_csv(models_asgtf_updated.path, header = None, index = None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 11. Generate cDNA and peptide files\n",
    "Using the updated gtf file and genome file, generate cDNA sequence.\n",
    "\n",
    "Then, using the cDNA sequence, generate peptide sequences using transdecoder.\n",
    "\n",
    "Expect this step to take some time, probably ~20-30min."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/compute_base_probs.pl ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa 0 > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//base_freqs.dat\n",
      "\n",
      "\n",
      "-first extracting base frequencies, we'll need them later.\n",
      "\n",
      "\n",
      "- extracting ORFs from transcripts.\n",
      "-total transcripts to examine: 63852\n",
      "[63800/63852] = 99.92% done    "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CMD: touch ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir/.__checkpoints_longorfs/TD.longorfs.ok\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "#################################\n",
      "### Done preparing long ORFs.  ###\n",
      "##################################\n",
      "\n",
      "\tUse file: ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.pep  for Pfam and/or BlastP searches to enable homology-based coding region identification.\n",
      "\n",
      "\tThen, run TransDecoder.Predict for your final coding region predictions.\n",
      "\n",
      "\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds 5000 5000 > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/exclude_similar_proteins.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000 > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr\n",
      "-skipping training candidate: rna37997.p1, not unique enough\n",
      "-skipping training candidate: rna11650.p1, not unique enough\n",
      "-skipping training candidate: rna17540.p1, not unique enough\n",
      "-skipping training candidate: rna51849.p1, not unique enough\n",
      "-skipping training candidate: rna4236.p1, not unique enough\n",
      "-skipping training candidate: rna14787.p1, not unique enough\n",
      "-skipping training candidate: rna14226.p1, not unique enough\n",
      "-skipping training candidate: rna20810.p1, not unique enough\n",
      "-skipping training candidate: rna20737.p1, not unique enough\n",
      "-skipping training candidate: rna7611.p1, not unique enough\n",
      "-skipping training candidate: rna22616.p1, not unique enough\n",
      "-skipping training candidate: rna48388.p1, not unique enough\n",
      "-skipping training candidate: rna22596.p1, not unique enough\n",
      "-skipping training candidate: rna48387.p1, not unique enough\n",
      "-skipping training candidate: rna6017.p1, not unique enough\n",
      "-skipping training candidate: rna35717.p1, not unique enough\n",
      "-skipping training candidate: rna19971.p1, not unique enough\n",
      "-skipping training candidate: rna19970.p1, not unique enough\n",
      "-skipping training candidate: rna20033.p1, not unique enough\n",
      "-skipping training candidate: rna30884.p1, not unique enough\n",
      "-skipping training candidate: rna33489.p1, not unique enough\n",
      "-skipping training candidate: rna19972.p1, not unique enough\n",
      "-skipping training candidate: rna5364.p1, not unique enough\n",
      "-skipping training candidate: rna37998.p1, not unique enough\n",
      "-skipping training candidate: rna8342.p1, not unique enough\n",
      "-skipping training candidate: rna5365.p1, not unique enough\n",
      "-skipping training candidate: rna5366.p1, not unique enough\n",
      "-skipping training candidate: rna15162.p1, not unique enough\n",
      "-skipping training candidate: rna19968.p1, not unique enough\n",
      "-skipping training candidate: rna9511.p1, not unique enough\n",
      "-skipping training candidate: rna14228.p1, not unique enough\n",
      "-skipping training candidate: rna20098.p1, not unique enough\n",
      "-skipping training candidate: rna19969.p1, not unique enough\n",
      "-skipping training candidate: rna26425.p1, not unique enough\n",
      "-skipping training candidate: rna54545.p1, not unique enough\n",
      "-skipping training candidate: rna7610.p1, not unique enough\n",
      "-skipping training candidate: rna26427.p1, not unique enough\n",
      "-skipping training candidate: rna54543.p1, not unique enough\n",
      "-skipping training candidate: rna22597.p1, not unique enough\n",
      "-skipping training candidate: rna46549.p1, not unique enough\n",
      "-skipping training candidate: rna46293.p1, not unique enough\n",
      "-skipping training candidate: rna22600.p1, not unique enough\n",
      "-skipping training candidate: rna35721.p1, not unique enough\n",
      "-skipping training candidate: rna35723.p1, not unique enough\n",
      "-skipping training candidate: rna53397.p1, not unique enough\n",
      "-skipping training candidate: rna19974.p1, not unique enough\n",
      "-skipping training candidate: rna22599.p1, not unique enough\n",
      "-skipping training candidate: rna8852.p1, not unique enough\n",
      "-skipping training candidate: rna22595.p1, not unique enough\n",
      "-skipping training candidate: rna37995.p1, not unique enough\n",
      "-skipping training candidate: rna37994.p1, not unique enough\n",
      "-skipping training candidate: rna38001.p1, not unique enough\n",
      "-skipping training candidate: rna37993.p1, not unique enough\n",
      "-skipping training candidate: rna37996.p1, not unique enough\n",
      "-skipping training candidate: rna37999.p1, not unique enough\n",
      "-skipping training candidate: rna19566.p1, not unique enough\n",
      "-skipping training candidate: rna29621.p1, not unique enough\n",
      "-skipping training candidate: rna38891.p1, not unique enough\n",
      "-skipping training candidate: rna19567.p1, not unique enough\n",
      "-skipping training candidate: rna46294.p1, not unique enough\n",
      "-skipping training candidate: rna35719.p1, not unique enough\n",
      "-skipping training candidate: rna48386.p1, not unique enough\n",
      "-skipping training candidate: rna7612.p1, not unique enough\n",
      "-skipping training candidate: rna30885.p1, not unique enough\n",
      "-skipping training candidate: rna10520.p1, not unique enough\n",
      "-skipping training candidate: rna15163.p1, not unique enough\n",
      "-skipping training candidate: rna46450.p1, not unique enough\n",
      "-skipping training candidate: rna53157.p1, not unique enough\n",
      "-skipping training candidate: rna804.p1, not unique enough\n",
      "-skipping training candidate: rna7164.p1, not unique enough\n",
      "-skipping training candidate: rna46454.p1, not unique enough\n",
      "-skipping training candidate: rna5799.p1, not unique enough\n",
      "-skipping training candidate: rna16682.p1, not unique enough\n",
      "-skipping training candidate: rna20037.p1, not unique enough\n",
      "-skipping training candidate: rna26736.p1, not unique enough\n",
      "-skipping training candidate: rna26428.p1, not unique enough\n",
      "-skipping training candidate: rna26421.p1, not unique enough\n",
      "-skipping training candidate: rna26420.p1, not unique enough\n",
      "-skipping training candidate: rna33488.p1, not unique enough\n",
      "-skipping training candidate: rna17651.p1, not unique enough\n",
      "-skipping training candidate: rna22610.p1, not unique enough\n",
      "-skipping training candidate: rna22605.p1, not unique enough\n",
      "-skipping training candidate: rna14227.p1, not unique enough\n",
      "-skipping training candidate: rna53161.p1, not unique enough\n",
      "-skipping training candidate: rna57845.p1, not unique enough\n",
      "-skipping training candidate: rna3402.p1, not unique enough\n",
      "-skipping training candidate: rna13606.p1, not unique enough\n",
      "-skipping training candidate: rna13607.p1, not unique enough\n",
      "-skipping training candidate: rna46448.p1, not unique enough\n",
      "-skipping training candidate: rna498.p1, not unique enough\n",
      "-skipping training candidate: rna46451.p1, not unique enough\n",
      "-skipping training candidate: rna20101.p1, not unique enough\n",
      "-skipping training candidate: rna40205.p1, not unique enough\n",
      "-skipping training candidate: rna21821.p1, not unique enough\n",
      "-skipping training candidate: rna29819.p1, not unique enough\n",
      "-skipping training candidate: rna29821.p1, not unique enough\n",
      "-skipping training candidate: rna53162.p1, not unique enough\n",
      "-skipping training candidate: rna52672.p1, not unique enough\n",
      "-skipping training candidate: rna4710.p1, not unique enough\n",
      "-skipping training candidate: rna38000.p1, not unique enough\n",
      "-skipping training candidate: rna55319.p1, not unique enough\n",
      "-skipping training candidate: rna8397.p1, not unique enough\n",
      "-skipping training candidate: rna8396.p1, not unique enough\n",
      "-skipping training candidate: rna8400.p1, not unique enough\n",
      "-skipping training candidate: rna31056.p1, not unique enough\n",
      "-skipping training candidate: rna33076.p1, not unique enough\n",
      "-skipping training candidate: rna33078.p1, not unique enough\n",
      "-skipping training candidate: rna41433.p1, not unique enough\n",
      "-skipping training candidate: rna8399.p1, not unique enough\n",
      "-skipping training candidate: rna31055.p1, not unique enough\n",
      "-skipping training candidate: rna12161.p1, not unique enough\n",
      "-skipping training candidate: rna60860.p1, not unique enough\n",
      "-skipping training candidate: rna60861.p1, not unique enough\n",
      "-skipping training candidate: rna60856.p1, not unique enough\n",
      "-skipping training candidate: rna30504.p1, not unique enough\n",
      "-skipping training candidate: rna30501.p1, not unique enough\n",
      "-skipping training candidate: rna30503.p1, not unique enough\n",
      "-skipping training candidate: rna53734.p1, not unique enough\n",
      "-skipping training candidate: rna53133.p1, not unique enough\n",
      "-skipping training candidate: rna60857.p1, not unique enough\n",
      "-skipping training candidate: rna60858.p1, not unique enough\n",
      "-skipping training candidate: rna499.p1, not unique enough\n",
      "-skipping training candidate: rna3400.p1, not unique enough\n",
      "-skipping training candidate: rna3403.p1, not unique enough\n",
      "-skipping training candidate: rna41005.p1, not unique enough\n",
      "-skipping training candidate: rna41006.p1, not unique enough\n",
      "-skipping training candidate: rna47382.p1, not unique enough\n",
      "-skipping training candidate: rna3401.p1, not unique enough\n",
      "-skipping training candidate: rna33073.p1, not unique enough\n",
      "-skipping training candidate: rna33077.p1, not unique enough\n",
      "-skipping training candidate: rna46551.p1, not unique enough\n",
      "-skipping training candidate: rna16678.p1, not unique enough\n",
      "-skipping training candidate: rna19754.p1, not unique enough\n",
      "-skipping training candidate: rna26430.p1, not unique enough\n",
      "-skipping training candidate: rna50183.p1, not unique enough\n",
      "-skipping training candidate: rna57578.p1, not unique enough\n",
      "-skipping training candidate: rna60859.p1, not unique enough\n",
      "-skipping training candidate: rna8398.p1, not unique enough\n",
      "-skipping training candidate: rna10033.p1, not unique enough\n",
      "-skipping training candidate: rna13608.p1, not unique enough\n",
      "-skipping training candidate: rna26433.p1, not unique enough\n",
      "-skipping training candidate: rna26429.p1, not unique enough\n",
      "-skipping training candidate: rna26422.p1, not unique enough\n",
      "-skipping training candidate: rna20031.p1, not unique enough\n",
      "-skipping training candidate: rna26424.p1, not unique enough\n",
      "-skipping training candidate: rna30502.p1, not unique enough\n",
      "-skipping training candidate: rna22609.p1, not unique enough\n",
      "-skipping training candidate: rna22612.p1, not unique enough\n",
      "-skipping training candidate: rna46552.p1, not unique enough\n",
      "-skipping training candidate: rna22606.p1, not unique enough\n",
      "-skipping training candidate: rna22607.p1, not unique enough\n",
      "-skipping training candidate: rna50184.p1, not unique enough\n",
      "-skipping training candidate: rna60290.p1, not unique enough\n",
      "-skipping training candidate: rna60298.p1, not unique enough\n",
      "-skipping training candidate: rna60299.p1, not unique enough\n",
      "-skipping training candidate: rna60295.p1, not unique enough\n",
      "-skipping training candidate: rna60296.p1, not unique enough\n",
      "-skipping training candidate: rna50181.p1, not unique enough\n",
      "-skipping training candidate: rna60297.p1, not unique enough\n",
      "-skipping training candidate: rna60292.p1, not unique enough\n",
      "-skipping training candidate: rna3404.p1, not unique enough\n",
      "-skipping training candidate: rna8482.p1, not unique enough\n",
      "-skipping training candidate: rna47383.p1, not unique enough\n",
      "-skipping training candidate: rna10034.p1, not unique enough\n",
      "-skipping training candidate: rna47385.p1, not unique enough\n",
      "-skipping training candidate: rna56528.p1, not unique enough\n",
      "-skipping training candidate: rna37846.p1, not unique enough\n",
      "-skipping training candidate: rna46449.p1, not unique enough\n",
      "-skipping training candidate: rna46452.p1, not unique enough\n",
      "-skipping training candidate: rna49266.p1, not unique enough\n",
      "-skipping training candidate: rna49265.p1, not unique enough\n",
      "-skipping training candidate: rna7166.p1, not unique enough\n",
      "-skipping training candidate: rna20026.p1, not unique enough\n",
      "-skipping training candidate: rna20021.p1, not unique enough\n",
      "-skipping training candidate: rna20022.p1, not unique enough\n",
      "-skipping training candidate: rna20028.p1, not unique enough\n",
      "-skipping training candidate: rna20024.p1, not unique enough\n",
      "-skipping training candidate: rna20025.p1, not unique enough\n",
      "-skipping training candidate: rna22601.p1, not unique enough\n",
      "-skipping training candidate: rna22975.p1, not unique enough\n",
      "-skipping training candidate: rna46169.p1, not unique enough\n",
      "-skipping training candidate: rna40390.p1, not unique enough\n",
      "-skipping training candidate: rna40392.p1, not unique enough\n",
      "-skipping training candidate: rna4712.p1, not unique enough\n",
      "-skipping training candidate: rna20035.p1, not unique enough\n",
      "-skipping training candidate: rna55671.p1, not unique enough\n",
      "-skipping training candidate: rna2296.p1, not unique enough\n",
      "-skipping training candidate: rna20029.p1, not unique enough\n",
      "-skipping training candidate: rna50712.p1, not unique enough\n",
      "-skipping training candidate: rna56529.p1, not unique enough\n",
      "-skipping training candidate: rna414.p1, not unique enough\n",
      "-skipping training candidate: rna20027.p1, not unique enough\n",
      "-skipping training candidate: rna40203.p1, not unique enough\n",
      "-skipping training candidate: rna3407.p1, not unique enough\n",
      "-skipping training candidate: rna6991.p1, not unique enough\n",
      "-skipping training candidate: rna39686.p1, not unique enough\n",
      "-skipping training candidate: rna51642.p1, not unique enough\n",
      "-skipping training candidate: rna26431.p1, not unique enough\n",
      "-skipping training candidate: rna9523.p1, not unique enough\n",
      "-skipping training candidate: rna10035.p1, not unique enough\n",
      "-skipping training candidate: rna26434.p1, not unique enough\n",
      "-skipping training candidate: rna22611.p1, not unique enough\n",
      "-skipping training candidate: rna33074.p1, not unique enough\n",
      "-skipping training candidate: rna36050.p1, not unique enough\n",
      "-skipping training candidate: rna46163.p1, not unique enough\n",
      "-skipping training candidate: rna60291.p1, not unique enough\n",
      "-skipping training candidate: rna4237.p1, not unique enough\n",
      "-skipping training candidate: rna22613.p1, not unique enough\n",
      "-skipping training candidate: rna36384.p1, not unique enough\n",
      "-skipping training candidate: rna16802.p1, not unique enough\n",
      "-skipping training candidate: rna45026.p1, not unique enough\n",
      "-skipping training candidate: rna48035.p1, not unique enough\n",
      "-skipping training candidate: rna48034.p1, not unique enough\n",
      "-skipping training candidate: rna50186.p1, not unique enough\n",
      "-skipping training candidate: rna6994.p1, not unique enough\n",
      "-skipping training candidate: rna36201.p1, not unique enough\n",
      "-skipping training candidate: rna36202.p1, not unique enough\n",
      "-skipping training candidate: rna38387.p1, not unique enough\n",
      "-skipping training candidate: rna53158.p1, not unique enough\n",
      "-skipping training candidate: rna2660.p1, not unique enough\n",
      "-skipping training candidate: rna2656.p1, not unique enough\n",
      "-skipping training candidate: rna2655.p1, not unique enough\n",
      "-skipping training candidate: rna2659.p1, not unique enough\n",
      "-skipping training candidate: rna47386.p1, not unique enough\n",
      "-skipping training candidate: rna2658.p1, not unique enough\n",
      "-skipping training candidate: rna20808.p1, not unique enough\n",
      "-skipping training candidate: rna22602.p1, not unique enough\n",
      "-skipping training candidate: rna54659.p1, not unique enough\n",
      "-skipping training candidate: rna54657.p1, not unique enough\n",
      "-skipping training candidate: rna22603.p1, not unique enough\n",
      "-skipping training candidate: rna48730.p1, not unique enough\n",
      "-skipping training candidate: rna50711.p1, not unique enough\n",
      "-skipping training candidate: rna50664.p1, not unique enough\n",
      "-skipping training candidate: rna22598.p1, not unique enough\n",
      "-skipping training candidate: rna35245.p1, not unique enough\n",
      "-skipping training candidate: rna35242.p1, not unique enough\n",
      "-skipping training candidate: rna35244.p1, not unique enough\n",
      "-skipping training candidate: rna46172.p1, not unique enough\n",
      "-skipping training candidate: rna58661.p1, not unique enough\n",
      "-skipping training candidate: rna9524.p1, not unique enough\n",
      "-skipping training candidate: rna32522.p1, not unique enough\n",
      "-skipping training candidate: rna32521.p1, not unique enough\n",
      "-skipping training candidate: rna46783.p1, not unique enough\n",
      "-skipping training candidate: rna37853.p1, not unique enough\n",
      "-skipping training candidate: rna37852.p1, not unique enough\n",
      "-skipping training candidate: rna46168.p1, not unique enough\n",
      "-skipping training candidate: rna37851.p1, not unique enough\n",
      "-skipping training candidate: rna47954.p1, not unique enough\n",
      "-skipping training candidate: rna6992.p1, not unique enough\n",
      "-skipping training candidate: rna43152.p1, not unique enough\n",
      "-skipping training candidate: rna46162.p1, not unique enough\n",
      "-skipping training candidate: rna46167.p1, not unique enough\n",
      "-skipping training candidate: rna46160.p1, not unique enough\n",
      "-skipping training candidate: rna50182.p1, not unique enough\n",
      "-skipping training candidate: rna49365.p1, not unique enough\n",
      "-skipping training candidate: rna49361.p1, not unique enough\n",
      "-skipping training candidate: rna49363.p1, not unique enough\n",
      "-skipping training candidate: rna49364.p1, not unique enough\n",
      "-skipping training candidate: rna60293.p1, not unique enough\n",
      "-skipping training candidate: rna6993.p1, not unique enough\n",
      "-skipping training candidate: rna60294.p1, not unique enough\n",
      "-skipping training candidate: rna14511.p1, not unique enough\n",
      "-skipping training candidate: rna61423.p1, not unique enough\n",
      "-skipping training candidate: rna61422.p1, not unique enough\n",
      "-skipping training candidate: rna12223.p1, not unique enough\n",
      "-skipping training candidate: rna7936.p1, not unique enough\n",
      "-skipping training candidate: rna24525.p1, not unique enough\n",
      "-skipping training candidate: rna24526.p1, not unique enough\n",
      "-skipping training candidate: rna44546.p1, not unique enough\n",
      "-skipping training candidate: rna34528.p1, not unique enough\n",
      "-skipping training candidate: rna56547.p1, not unique enough\n",
      "-skipping training candidate: rna4653.p1, not unique enough\n",
      "-skipping training candidate: rna36048.p1, not unique enough\n",
      "-skipping training candidate: rna46171.p1, not unique enough\n",
      "-skipping training candidate: rna51609.p1, not unique enough\n",
      "-skipping training candidate: rna10096.p1, not unique enough\n",
      "-skipping training candidate: rna13663.p1, not unique enough\n",
      "-skipping training candidate: rna29581.p1, not unique enough\n",
      "-skipping training candidate: rna48731.p1, not unique enough\n",
      "-skipping training candidate: rna39684.p1, not unique enough\n",
      "-skipping training candidate: rna13511.p1, not unique enough\n",
      "-skipping training candidate: rna17498.p1, not unique enough\n",
      "-skipping training candidate: rna22604.p1, not unique enough\n",
      "-skipping training candidate: rna46159.p1, not unique enough\n",
      "-skipping training candidate: rna58662.p1, not unique enough\n",
      "-skipping training candidate: rna13513.p1, not unique enough\n",
      "-skipping training candidate: rna36049.p1, not unique enough\n",
      "-skipping training candidate: rna9613.p1, not unique enough\n",
      "-skipping training candidate: rna42496.p1, not unique enough\n",
      "-skipping training candidate: rna42495.p1, not unique enough\n",
      "-skipping training candidate: rna42500.p1, not unique enough\n",
      "-skipping training candidate: rna42497.p1, not unique enough\n",
      "-skipping training candidate: rna42499.p1, not unique enough\n",
      "-skipping training candidate: rna42498.p1, not unique enough\n",
      "-skipping training candidate: rna42595.p1, not unique enough\n",
      "-skipping training candidate: rna17337.p1, not unique enough\n",
      "-skipping training candidate: rna35162.p1, not unique enough\n",
      "-skipping training candidate: rna45025.p1, not unique enough\n",
      "-skipping training candidate: rna58663.p1, not unique enough\n",
      "-skipping training candidate: rna12221.p1, not unique enough\n",
      "-skipping training candidate: rna36608.p1, not unique enough\n",
      "-skipping training candidate: rna50717.p1, not unique enough\n",
      "-skipping training candidate: rna50716.p1, not unique enough\n",
      "-skipping training candidate: rna29664.p1, not unique enough\n",
      "-skipping training candidate: rna32410.p1, not unique enough\n",
      "-skipping training candidate: rna12219.p1, not unique enough\n",
      "-skipping training candidate: rna35563.p1, not unique enough\n",
      "-skipping training candidate: rna46295.p1, not unique enough\n",
      "-skipping training candidate: rna9851.p1, not unique enough\n",
      "-skipping training candidate: rna30500.p1, not unique enough\n",
      "-skipping training candidate: rna50715.p1, not unique enough\n",
      "-skipping training candidate: rna61421.p1, not unique enough\n",
      "-skipping training candidate: rna35803.p1, not unique enough\n",
      "-skipping training candidate: rna2657.p1, not unique enough\n",
      "-skipping training candidate: rna13510.p1, not unique enough\n",
      "-skipping training candidate: rna46166.p1, not unique enough\n",
      "-skipping training candidate: rna46165.p1, not unique enough\n",
      "-skipping training candidate: rna5367.p1, not unique enough\n",
      "-skipping training candidate: rna20034.p1, not unique enough\n",
      "-skipping training candidate: rna36607.p1, not unique enough\n",
      "-skipping training candidate: rna50714.p1, not unique enough\n",
      "-skipping training candidate: rna55318.p1, not unique enough\n",
      "-skipping training candidate: rna20030.p1, not unique enough\n",
      "-skipping training candidate: rna22302.p1, not unique enough\n",
      "-skipping training candidate: rna46155.p1, not unique enough\n",
      "-skipping training candidate: rna8381.p1, not unique enough\n",
      "-skipping training candidate: rna26423.p1, not unique enough\n",
      "-skipping training candidate: rna35619.p1, not unique enough\n",
      "-skipping training candidate: rna40204.p1, not unique enough\n",
      "-skipping training candidate: rna12220.p1, not unique enough\n",
      "-skipping training candidate: rna17334.p1, not unique enough\n",
      "-skipping training candidate: rna27798.p1, not unique enough\n",
      "-skipping training candidate: rna27797.p1, not unique enough\n",
      "-skipping training candidate: rna27800.p1, not unique enough\n",
      "-skipping training candidate: rna46170.p1, not unique enough\n",
      "-skipping training candidate: rna60825.p1, not unique enough\n",
      "-skipping training candidate: rna12218.p1, not unique enough\n",
      "-skipping training candidate: rna19659.p1, not unique enough\n",
      "-skipping training candidate: rna19658.p1, not unique enough\n",
      "-skipping training candidate: rna36610.p1, not unique enough\n",
      "-skipping training candidate: rna418.p1, not unique enough\n",
      "-skipping training candidate: rna5090.p1, not unique enough\n",
      "-skipping training candidate: rna22614.p1, not unique enough\n",
      "-skipping training candidate: rna43135.p1, not unique enough\n",
      "-skipping training candidate: rna43131.p1, not unique enough\n",
      "-skipping training candidate: rna7934.p1, not unique enough\n",
      "-skipping training candidate: rna20023.p1, not unique enough\n",
      "-skipping training candidate: rna20020.p1, not unique enough\n",
      "-skipping training candidate: rna26426.p1, not unique enough\n",
      "-skipping training candidate: rna43132.p1, not unique enough\n",
      "-skipping training candidate: rna43137.p1, not unique enough\n",
      "-skipping training candidate: rna42502.p1, not unique enough\n",
      "-skipping training candidate: rna55833.p1, not unique enough\n",
      "-skipping training candidate: rna60829.p1, not unique enough\n",
      "-skipping training candidate: rna32967.p1, not unique enough\n",
      "-skipping training candidate: rna32970.p1, not unique enough\n",
      "-skipping training candidate: rna32968.p1, not unique enough\n",
      "-skipping training candidate: rna32966.p1, not unique enough\n",
      "-skipping training candidate: rna36383.p1, not unique enough\n",
      "-skipping training candidate: rna60826.p1, not unique enough\n",
      "-skipping training candidate: rna22608.p1, not unique enough\n",
      "-skipping training candidate: rna32638.p1, not unique enough\n",
      "-skipping training candidate: rna38054.p1, not unique enough\n",
      "-skipping training candidate: rna36609.p1, not unique enough\n",
      "-skipping training candidate: rna38055.p1, not unique enough\n",
      "-skipping training candidate: rna29847.p1, not unique enough\n",
      "-skipping training candidate: rna29848.p1, not unique enough\n",
      "-skipping training candidate: rna29849.p1, not unique enough\n",
      "-skipping training candidate: rna29846.p1, not unique enough\n",
      "-skipping training candidate: rna17216.p1, not unique enough\n",
      "-skipping training candidate: rna30461.p1, not unique enough\n",
      "-skipping training candidate: rna30459.p1, not unique enough\n",
      "-skipping training candidate: rna29845.p1, not unique enough\n",
      "-skipping training candidate: rna36382.p1, not unique enough\n",
      "-skipping training candidate: rna36008.p1, not unique enough\n",
      "-skipping training candidate: rna38224.p1, not unique enough\n",
      "-skipping training candidate: rna12611.p1, not unique enough\n",
      "-skipping training candidate: rna36781.p1, not unique enough\n",
      "-skipping training candidate: rna48158.p1, not unique enough\n",
      "-skipping training candidate: rna31304.p1, not unique enough\n",
      "-skipping training candidate: rna8382.p1, not unique enough\n",
      "-skipping training candidate: rna9512.p1, not unique enough\n",
      "-skipping training candidate: rna56932.p1, not unique enough\n",
      "-skipping training candidate: rna2784.p1, not unique enough\n",
      "-skipping training candidate: rna4499.p1, not unique enough\n",
      "-skipping training candidate: rna17217.p1, not unique enough\n",
      "-skipping training candidate: rna41007.p1, not unique enough\n",
      "-skipping training candidate: rna2782.p1, not unique enough\n",
      "-skipping training candidate: rna21817.p1, not unique enough\n",
      "-skipping training candidate: rna53735.p1, not unique enough\n",
      "-skipping training candidate: rna58893.p1, not unique enough\n",
      "-skipping training candidate: rna60828.p1, not unique enough\n",
      "-skipping training candidate: rna50016.p1, not unique enough\n",
      "-skipping training candidate: rna50021.p1, not unique enough\n",
      "-skipping training candidate: rna17215.p1, not unique enough\n",
      "-skipping training candidate: rna35430.p1, not unique enough\n",
      "-skipping training candidate: rna60830.p1, not unique enough\n",
      "-skipping training candidate: rna4993.p1, not unique enough\n",
      "-skipping training candidate: rna13407.p1, not unique enough\n",
      "-skipping training candidate: rna26432.p1, not unique enough\n",
      "-skipping training candidate: rna46530.p1, not unique enough\n",
      "-skipping training candidate: rna38493.p1, not unique enough\n",
      "-skipping training candidate: rna38492.p1, not unique enough\n",
      "-skipping training candidate: rna38496.p1, not unique enough\n",
      "-skipping training candidate: rna38495.p1, not unique enough\n",
      "-skipping training candidate: rna51821.p1, not unique enough\n",
      "-skipping training candidate: rna51823.p1, not unique enough\n",
      "-skipping training candidate: rna22615.p1, not unique enough\n",
      "-skipping training candidate: rna51824.p1, not unique enough\n",
      "-skipping training candidate: rna276.p1, not unique enough\n",
      "-skipping training candidate: rna8692.p1, not unique enough\n",
      "-skipping training candidate: rna8693.p1, not unique enough\n",
      "-skipping training candidate: rna8691.p1, not unique enough\n",
      "-skipping training candidate: rna29580.p1, not unique enough\n",
      "-skipping training candidate: rna53743.p1, not unique enough\n",
      "-skipping training candidate: rna5861.p1, not unique enough\n",
      "-skipping training candidate: rna8346.p1, not unique enough\n",
      "-skipping training candidate: rna17218.p1, not unique enough\n",
      "-skipping training candidate: rna51822.p1, not unique enough\n",
      "-skipping training candidate: rna40570.p1, not unique enough\n",
      "-skipping training candidate: rna46164.p1, not unique enough\n",
      "-skipping training candidate: rna53744.p1, not unique enough\n",
      "-skipping training candidate: rna419.p1, not unique enough\n",
      "-skipping training candidate: rna40572.p1, not unique enough\n",
      "-skipping training candidate: rna43133.p1, not unique enough\n",
      "-skipping training candidate: rna48903.p1, not unique enough\n",
      "-skipping training candidate: rna48160.p1, not unique enough\n",
      "-skipping training candidate: rna9717.p1, not unique enough\n",
      "-skipping training candidate: rna8794.p1, not unique enough\n",
      "-skipping training candidate: rna33296.p1, not unique enough\n",
      "-skipping training candidate: rna37855.p1, not unique enough\n",
      "-skipping training candidate: rna43136.p1, not unique enough\n",
      "-skipping training candidate: rna60831.p1, not unique enough\n",
      "-skipping training candidate: rna8960.p1, not unique enough\n",
      "-skipping training candidate: rna30762.p1, not unique enough\n",
      "-skipping training candidate: rna30765.p1, not unique enough\n",
      "-skipping training candidate: rna36364.p1, not unique enough\n",
      "-skipping training candidate: rna36367.p1, not unique enough\n",
      "-skipping training candidate: rna36366.p1, not unique enough\n",
      "-skipping training candidate: rna40569.p1, not unique enough\n",
      "-skipping training candidate: rna44833.p1, not unique enough\n",
      "-skipping training candidate: rna32969.p1, not unique enough\n",
      "-skipping training candidate: rna45027.p1, not unique enough\n",
      "-skipping training candidate: rna55532.p1, not unique enough\n",
      "-skipping training candidate: rna55531.p1, not unique enough\n",
      "-skipping training candidate: rna5088.p1, not unique enough\n",
      "-skipping training candidate: rna8487.p1, not unique enough\n",
      "-skipping training candidate: rna12610.p1, not unique enough\n",
      "-skipping training candidate: rna44514.p1, not unique enough\n",
      "-skipping training candidate: rna2213.p1, not unique enough\n",
      "-skipping training candidate: rna3925.p1, not unique enough\n",
      "-skipping training candidate: rna17214.p1, not unique enough\n",
      "-skipping training candidate: rna35432.p1, not unique enough\n",
      "-skipping training candidate: rna44517.p1, not unique enough\n",
      "-skipping training candidate: rna3926.p1, not unique enough\n",
      "-skipping training candidate: rna17219.p1, not unique enough\n",
      "-skipping training candidate: rna30458.p1, not unique enough\n",
      "-skipping training candidate: rna50018.p1, not unique enough\n",
      "-skipping training candidate: rna53745.p1, not unique enough\n",
      "-skipping training candidate: rna5863.p1, not unique enough\n",
      "-skipping training candidate: rna7841.p1, not unique enough\n",
      "-skipping training candidate: rna7844.p1, not unique enough\n",
      "-skipping training candidate: rna7393.p1, not unique enough\n",
      "-skipping training candidate: rna7391.p1, not unique enough\n",
      "-skipping training candidate: rna31592.p1, not unique enough\n",
      "-skipping training candidate: rna40574.p1, not unique enough\n",
      "-skipping training candidate: rna51828.p1, not unique enough\n",
      "-skipping training candidate: rna54055.p1, not unique enough\n",
      "-skipping training candidate: rna7394.p1, not unique enough\n",
      "-skipping training candidate: rna8486.p1, not unique enough\n",
      "-skipping training candidate: rna31593.p1, not unique enough\n",
      "-skipping training candidate: rna46885.p1, not unique enough\n",
      "-skipping training candidate: rna44516.p1, not unique enough\n",
      "-skipping training candidate: rna16513.p1, not unique enough\n",
      "-skipping training candidate: rna51827.p1, not unique enough\n",
      "-skipping training candidate: rna4208.p1, not unique enough\n",
      "-skipping training candidate: rna5865.p1, not unique enough\n",
      "-skipping training candidate: rna40573.p1, not unique enough\n",
      "-skipping training candidate: rna14852.p1, not unique enough\n",
      "-skipping training candidate: rna14853.p1, not unique enough\n",
      "-skipping training candidate: rna17335.p1, not unique enough\n",
      "-skipping training candidate: rna21359.p1, not unique enough\n",
      "-skipping training candidate: rna44518.p1, not unique enough\n",
      "-skipping training candidate: rna48901.p1, not unique enough\n",
      "-skipping training candidate: rna56059.p1, not unique enough\n",
      "-skipping training candidate: rna417.p1, not unique enough\n",
      "-skipping training candidate: rna30464.p1, not unique enough\n",
      "-skipping training candidate: rna5368.p1, not unique enough\n",
      "-skipping training candidate: rna25460.p1, not unique enough\n",
      "-skipping training candidate: rna25462.p1, not unique enough\n",
      "-skipping training candidate: rna49366.p1, not unique enough\n",
      "-skipping training candidate: rna51826.p1, not unique enough\n",
      "-skipping training candidate: rna60837.p1, not unique enough\n",
      "-skipping training candidate: rna36051.p1, not unique enough\n",
      "-skipping training candidate: rna38285.p1, not unique enough\n",
      "-skipping training candidate: rna44519.p1, not unique enough\n",
      "-skipping training candidate: rna48210.p1, not unique enough\n",
      "-skipping training candidate: rna60833.p1, not unique enough\n",
      "-skipping training candidate: rna17220.p1, not unique enough\n",
      "-skipping training candidate: rna33920.p1, not unique enough\n",
      "-skipping training candidate: rna44520.p1, not unique enough\n",
      "-skipping training candidate: rna57161.p1, not unique enough\n",
      "-skipping training candidate: rna30474.p1, not unique enough\n",
      "-skipping training candidate: rna30477.p1, not unique enough\n",
      "-skipping training candidate: rna33921.p1, not unique enough\n",
      "-skipping training candidate: rna33917.p1, not unique enough\n",
      "-skipping training candidate: rna38225.p1, not unique enough\n",
      "-skipping training candidate: rna7843.p1, not unique enough\n",
      "-skipping training candidate: rna11233.p1, not unique enough\n",
      "-skipping training candidate: rna33918.p1, not unique enough\n",
      "-skipping training candidate: rna60827.p1, not unique enough\n",
      "-skipping training candidate: rna28303.p1, not unique enough\n",
      "-skipping training candidate: rna32637.p1, not unique enough\n",
      "-skipping training candidate: rna49082.p1, not unique enough\n",
      "-skipping training candidate: rna59535.p1, not unique enough\n",
      "-skipping training candidate: rna48900.p1, not unique enough\n",
      "-skipping training candidate: rna21530.p1, not unique enough\n",
      "-skipping training candidate: rna33293.p1, not unique enough\n",
      "-skipping training candidate: rna40577.p1, not unique enough\n",
      "-skipping training candidate: rna2024.p1, not unique enough\n",
      "-skipping training candidate: rna3395.p1, not unique enough\n",
      "-skipping training candidate: rna3396.p1, not unique enough\n",
      "-skipping training candidate: rna33297.p1, not unique enough\n",
      "-skipping training candidate: rna40576.p1, not unique enough\n",
      "-skipping training candidate: rna49093.p1, not unique enough\n",
      "-skipping training candidate: rna30476.p1, not unique enough\n",
      "-skipping training candidate: rna30763.p1, not unique enough\n",
      "-skipping training candidate: rna40219.p1, not unique enough\n",
      "-skipping training candidate: rna49088.p1, not unique enough\n",
      "-skipping training candidate: rna4994.p1, not unique enough\n",
      "-skipping training candidate: rna21358.p1, not unique enough\n",
      "-skipping training candidate: rna30766.p1, not unique enough\n",
      "-skipping training candidate: rna46528.p1, not unique enough\n",
      "-skipping training candidate: rna44521.p1, not unique enough\n",
      "-skipping training candidate: rna49087.p1, not unique enough\n",
      "-skipping training candidate: rna3927.p1, not unique enough\n",
      "-skipping training candidate: rna17212.p1, not unique enough\n",
      "-skipping training candidate: rna30457.p1, not unique enough\n",
      "-skipping training candidate: rna45317.p1, not unique enough\n",
      "-skipping training candidate: rna45319.p1, not unique enough\n",
      "-skipping training candidate: rna49086.p1, not unique enough\n",
      "-skipping training candidate: rna54709.p1, not unique enough\n",
      "-skipping training candidate: rna7395.p1, not unique enough\n",
      "-skipping training candidate: rna9845.p1, not unique enough\n",
      "-skipping training candidate: rna33298.p1, not unique enough\n",
      "-skipping training candidate: rna41128.p1, not unique enough\n",
      "-skipping training candidate: rna20036.p1, not unique enough\n",
      "-skipping training candidate: rna27224.p1, not unique enough\n",
      "-skipping training candidate: rna49605.p1, not unique enough\n",
      "-skipping training candidate: rna56273.p1, not unique enough\n",
      "-skipping training candidate: rna56271.p1, not unique enough\n",
      "-skipping training candidate: rna28506.p1, not unique enough\n",
      "-skipping training candidate: rna30768.p1, not unique enough\n",
      "-skipping training candidate: rna43370.p1, not unique enough\n",
      "-skipping training candidate: rna49085.p1, not unique enough\n",
      "-skipping training candidate: rna52425.p1, not unique enough\n",
      "-skipping training candidate: rna52427.p1, not unique enough\n",
      "-skipping training candidate: rna28503.p1, not unique enough\n",
      "-skipping training candidate: rna49084.p1, not unique enough\n",
      "-skipping training candidate: rna44522.p1, not unique enough\n",
      "-skipping training candidate: rna50982.p1, not unique enough\n",
      "-skipping training candidate: rna14851.p1, not unique enough\n",
      "-skipping training candidate: rna12333.p1, not unique enough\n",
      "-skipping training candidate: rna15857.p1, not unique enough\n",
      "-skipping training candidate: rna24349.p1, not unique enough\n",
      "-skipping training candidate: rna24348.p1, not unique enough\n",
      "-skipping training candidate: rna16612.p1, not unique enough\n",
      "-skipping training candidate: rna56388.p1, not unique enough\n",
      "-skipping training candidate: rna6998.p1, not unique enough\n",
      "-skipping training candidate: rna6997.p1, not unique enough\n",
      "-skipping training candidate: rna24651.p1, not unique enough\n",
      "-skipping training candidate: rna49092.p1, not unique enough\n",
      "-skipping training candidate: rna57148.p1, not unique enough\n",
      "-skipping training candidate: rna10525.p1, not unique enough\n",
      "-skipping training candidate: rna28302.p1, not unique enough\n",
      "-skipping training candidate: rna49083.p1, not unique enough\n",
      "-skipping training candidate: rna3876.p1, not unique enough\n",
      "-skipping training candidate: rna24096.p1, not unique enough\n",
      "-skipping training candidate: rna56934.p1, not unique enough\n",
      "-skipping training candidate: rna31594.p1, not unique enough\n",
      "-skipping training candidate: rna56055.p1, not unique enough\n",
      "-skipping training candidate: rna2021.p1, not unique enough\n",
      "-skipping training candidate: rna24097.p1, not unique enough\n",
      "-skipping training candidate: rna39551.p1, not unique enough\n",
      "-skipping training candidate: rna2019.p1, not unique enough\n",
      "-skipping training candidate: rna40567.p1, not unique enough\n",
      "-skipping training candidate: rna40571.p1, not unique enough\n",
      "-skipping training candidate: rna52068.p1, not unique enough\n",
      "-skipping training candidate: rna9720.p1, not unique enough\n",
      "-skipping training candidate: rna9718.p1, not unique enough\n",
      "-skipping training candidate: rna9719.p1, not unique enough\n",
      "-skipping training candidate: rna9722.p1, not unique enough\n",
      "-skipping training candidate: rna40515.p1, not unique enough\n",
      "-skipping training candidate: rna40513.p1, not unique enough\n",
      "-skipping training candidate: rna9721.p1, not unique enough\n",
      "-skipping training candidate: rna13408.p1, not unique enough\n",
      "-skipping training candidate: rna13409.p1, not unique enough\n",
      "-skipping training candidate: rna28507.p1, not unique enough\n",
      "-skipping training candidate: rna27225.p1, not unique enough\n",
      "-skipping training candidate: rna33711.p1, not unique enough\n",
      "-skipping training candidate: rna36012.p1, not unique enough\n",
      "-skipping training candidate: rna41552.p1, not unique enough\n",
      "-skipping training candidate: rna39552.p1, not unique enough\n",
      "-skipping training candidate: rna46154.p1, not unique enough\n",
      "-skipping training candidate: rna46158.p1, not unique enough\n",
      "-skipping training candidate: rna46161.p1, not unique enough\n",
      "-skipping training candidate: rna49091.p1, not unique enough\n",
      "-skipping training candidate: rna54022.p1, not unique enough\n",
      "-skipping training candidate: rna28504.p1, not unique enough\n",
      "-skipping training candidate: rna41551.p1, not unique enough\n",
      "-skipping training candidate: rna53494.p1, not unique enough\n",
      "-skipping training candidate: rna11654.p1, not unique enough\n",
      "-skipping training candidate: rna51212.p1, not unique enough\n",
      "-skipping training candidate: rna2022.p1, not unique enough\n",
      "-skipping training candidate: rna6194.p1, not unique enough\n",
      "-skipping training candidate: rna11655.p1, not unique enough\n",
      "-skipping training candidate: rna23675.p1, not unique enough\n",
      "-skipping training candidate: rna43783.p1, not unique enough\n",
      "-skipping training candidate: rna4556.p1, not unique enough\n",
      "-skipping training candidate: rna2020.p1, not unique enough\n",
      "-skipping training candidate: rna7842.p1, not unique enough\n",
      "-skipping training candidate: rna33919.p1, not unique enough\n",
      "-skipping training candidate: rna56060.p1, not unique enough\n",
      "-skipping training candidate: rna44483.p1, not unique enough\n",
      "-skipping training candidate: rna44523.p1, not unique enough\n",
      "-skipping training candidate: rna36011.p1, not unique enough\n",
      "-skipping training candidate: rna38226.p1, not unique enough\n",
      "-skipping training candidate: rna25775.p1, not unique enough\n",
      "-skipping training candidate: rna36010.p1, not unique enough\n",
      "-skipping training candidate: rna38227.p1, not unique enough\n",
      "-skipping training candidate: rna6193.p1, not unique enough\n",
      "-skipping training candidate: rna24101.p1, not unique enough\n",
      "-skipping training candidate: rna52430.p1, not unique enough\n",
      "-skipping training candidate: rna56066.p1, not unique enough\n",
      "-skipping training candidate: rna25779.p1, not unique enough\n",
      "-skipping training candidate: rna33294.p1, not unique enough\n",
      "-skipping training candidate: rna11539.p1, not unique enough\n",
      "-skipping training candidate: rna25782.p1, not unique enough\n",
      "-skipping training candidate: rna54057.p1, not unique enough\n",
      "-skipping training candidate: rna14769.p1, not unique enough\n",
      "-skipping training candidate: rna17959.p1, not unique enough\n",
      "-skipping training candidate: rna28545.p1, not unique enough\n",
      "-skipping training candidate: rna30764.p1, not unique enough\n",
      "-skipping training candidate: rna54875.p1, not unique enough\n",
      "-skipping training candidate: rna25781.p1, not unique enough\n",
      "-skipping training candidate: rna27889.p1, not unique enough\n",
      "-skipping training candidate: rna27891.p1, not unique enough\n",
      "-skipping training candidate: rna27226.p1, not unique enough\n",
      "-skipping training candidate: rna28547.p1, not unique enough\n",
      "-skipping training candidate: rna44524.p1, not unique enough\n",
      "-skipping training candidate: rna4294.p1, not unique enough\n",
      "-skipping training candidate: rna4295.p1, not unique enough\n",
      "-skipping training candidate: rna24926.p1, not unique enough\n",
      "-skipping training candidate: rna30475.p1, not unique enough\n",
      "-skipping training candidate: rna46910.p1, not unique enough\n",
      "-skipping training candidate: rna46906.p1, not unique enough\n",
      "-skipping training candidate: rna51850.p1, not unique enough\n",
      "-skipping training candidate: rna21046.p1, not unique enough\n",
      "-skipping training candidate: rna36150.p1, not unique enough\n",
      "-skipping training candidate: rna46904.p1, not unique enough\n",
      "-skipping training candidate: rna25367.p1, not unique enough\n",
      "-skipping training candidate: rna25365.p1, not unique enough\n",
      "-skipping training candidate: rna25783.p1, not unique enough\n",
      "-skipping training candidate: rna28502.p1, not unique enough\n",
      "-skipping training candidate: rna40512.p1, not unique enough\n",
      "-skipping training candidate: rna49090.p1, not unique enough\n",
      "-skipping training candidate: rna40511.p1, not unique enough\n",
      "-skipping training candidate: rna41055.p1, not unique enough\n",
      "-skipping training candidate: rna51453.p1, not unique enough\n",
      "-skipping training candidate: rna2023.p1, not unique enough\n",
      "-skipping training candidate: rna33916.p1, not unique enough\n",
      "-skipping training candidate: rna38494.p1, not unique enough\n",
      "-skipping training candidate: rna41054.p1, not unique enough\n",
      "-skipping training candidate: rna6784.p1, not unique enough\n",
      "-skipping training candidate: rna9639.p1, not unique enough\n",
      "-skipping training candidate: rna12529.p1, not unique enough\n",
      "-skipping training candidate: rna29298.p1, not unique enough\n",
      "-skipping training candidate: rna29300.p1, not unique enough\n",
      "-skipping training candidate: rna10097.p1, not unique enough\n",
      "-skipping training candidate: rna21075.p1, not unique enough\n",
      "-skipping training candidate: rna39130.p1, not unique enough\n",
      "-skipping training candidate: rna39132.p1, not unique enough\n",
      "-skipping training candidate: rna39128.p1, not unique enough\n",
      "-skipping training candidate: rna39131.p1, not unique enough\n",
      "-skipping training candidate: rna39129.p1, not unique enough\n",
      "-skipping training candidate: rna25780.p1, not unique enough\n",
      "-skipping training candidate: rna31322.p1, not unique enough\n",
      "-skipping training candidate: rna40575.p1, not unique enough\n",
      "-skipping training candidate: rna41052.p1, not unique enough\n",
      "-skipping training candidate: rna41053.p1, not unique enough\n",
      "-skipping training candidate: rna27849.p1, not unique enough\n",
      "-skipping training candidate: rna24098.p1, not unique enough\n",
      "-skipping training candidate: rna30528.p1, not unique enough\n",
      "-skipping training candidate: rna33710.p1, not unique enough\n",
      "-skipping training candidate: rna18429.p1, not unique enough\n",
      "-skipping training candidate: rna46153.p1, not unique enough\n",
      "-skipping training candidate: rna59407.p1, not unique enough\n",
      "-skipping training candidate: rna12827.p1, not unique enough\n",
      "-skipping training candidate: rna15693.p1, not unique enough\n",
      "-skipping training candidate: rna15689.p1, not unique enough\n",
      "-skipping training candidate: rna15691.p1, not unique enough\n",
      "-skipping training candidate: rna15695.p1, not unique enough\n",
      "-skipping training candidate: rna15694.p1, not unique enough\n",
      "-skipping training candidate: rna15692.p1, not unique enough\n",
      "-skipping training candidate: rna18426.p1, not unique enough\n",
      "-skipping training candidate: rna21069.p1, not unique enough\n",
      "-skipping training candidate: rna42501.p1, not unique enough\n",
      "-skipping training candidate: rna25778.p1, not unique enough\n",
      "-skipping training candidate: rna36368.p1, not unique enough\n",
      "-skipping training candidate: rna36365.p1, not unique enough\n",
      "-skipping training candidate: rna13429.p1, not unique enough\n",
      "-skipping training candidate: rna34576.p1, not unique enough\n",
      "-skipping training candidate: rna49130.p1, not unique enough\n",
      "-skipping training candidate: rna53627.p1, not unique enough\n",
      "-skipping training candidate: rna60832.p1, not unique enough\n",
      "-skipping training candidate: rna1723.p1, not unique enough\n",
      "-skipping training candidate: rna17303.p1, not unique enough\n",
      "-skipping training candidate: rna28501.p1, not unique enough\n",
      "-skipping training candidate: rna34574.p1, not unique enough\n",
      "-skipping training candidate: rna47238.p1, not unique enough\n",
      "-skipping training candidate: rna21068.p1, not unique enough\n",
      "-skipping training candidate: rna9849.p1, not unique enough\n",
      "-skipping training candidate: rna6566.p1, not unique enough\n",
      "-skipping training candidate: rna6567.p1, not unique enough\n",
      "-skipping training candidate: rna6564.p1, not unique enough\n",
      "-skipping training candidate: rna6565.p1, not unique enough\n",
      "-skipping training candidate: rna54037.p1, not unique enough\n",
      "-skipping training candidate: rna55487.p1, not unique enough\n",
      "-skipping training candidate: rna21004.p1, not unique enough\n",
      "-skipping training candidate: rna36881.p1, not unique enough\n",
      "-skipping training candidate: rna46157.p1, not unique enough\n",
      "-skipping training candidate: rna51209.p1, not unique enough\n",
      "-skipping training candidate: rna21074.p1, not unique enough\n",
      "-skipping training candidate: rna44916.p1, not unique enough\n",
      "-skipping training candidate: rna46626.p1, not unique enough\n",
      "-skipping training candidate: rna46627.p1, not unique enough\n",
      "-skipping training candidate: rna24102.p1, not unique enough\n",
      "-skipping training candidate: rna55578.p1, not unique enough\n",
      "-skipping training candidate: rna13430.p1, not unique enough\n",
      "-skipping training candidate: rna29850.p1, not unique enough\n",
      "-skipping training candidate: rna36880.p1, not unique enough\n",
      "-skipping training candidate: rna24100.p1, not unique enough\n",
      "-skipping training candidate: rna23678.p1, not unique enough\n",
      "-skipping training candidate: rna47542.p1, not unique enough\n",
      "-skipping training candidate: rna53496.p1, not unique enough\n",
      "-skipping training candidate: rna48029.p1, not unique enough\n",
      "-skipping training candidate: rna822.p1, not unique enough\n",
      "-skipping training candidate: rna2612.p1, not unique enough\n",
      "-skipping training candidate: rna2610.p1, not unique enough\n",
      "-skipping training candidate: rna2611.p1, not unique enough\n",
      "-skipping training candidate: rna52423.p1, not unique enough\n",
      "-skipping training candidate: rna15696.p1, not unique enough\n",
      "-skipping training candidate: rna23676.p1, not unique enough\n",
      "-skipping training candidate: rna6573.p1, not unique enough\n",
      "-skipping training candidate: rna6576.p1, not unique enough\n",
      "-skipping training candidate: rna7329.p1, not unique enough\n",
      "-skipping training candidate: rna27227.p1, not unique enough\n",
      "-skipping training candidate: rna54941.p1, not unique enough\n",
      "-skipping training candidate: rna20807.p1, not unique enough\n",
      "-skipping training candidate: rna20723.p1, not unique enough\n",
      "-skipping training candidate: rna6445.p1, not unique enough\n",
      "-skipping training candidate: rna7217.p1, not unique enough\n",
      "-skipping training candidate: rna11232.p1, not unique enough\n",
      "-skipping training candidate: rna21070.p1, not unique enough\n",
      "-skipping training candidate: rna40811.p1, not unique enough\n",
      "-skipping training candidate: rna3830.p1, not unique enough\n",
      "-skipping training candidate: rna6446.p1, not unique enough\n",
      "-skipping training candidate: rna21073.p1, not unique enough\n",
      "-skipping training candidate: rna46911.p1, not unique enough\n",
      "-skipping training candidate: rna51664.p1, not unique enough\n",
      "-skipping training candidate: rna43618.p1, not unique enough\n",
      "-skipping training candidate: rna57043.p1, not unique enough\n",
      "-skipping training candidate: rna60836.p1, not unique enough\n",
      "-skipping training candidate: rna5527.p1, not unique enough\n",
      "-skipping training candidate: rna5524.p1, not unique enough\n",
      "-skipping training candidate: rna5526.p1, not unique enough\n",
      "-skipping training candidate: rna5525.p1, not unique enough\n",
      "-skipping training candidate: rna17467.p1, not unique enough\n",
      "-skipping training candidate: rna27416.p1, not unique enough\n",
      "-skipping training candidate: rna40578.p1, not unique enough\n",
      "-skipping training candidate: rna59713.p1, not unique enough\n",
      "-skipping training candidate: rna59714.p1, not unique enough\n",
      "-skipping training candidate: rna8149.p1, not unique enough\n",
      "-skipping training candidate: rna30291.p1, not unique enough\n",
      "-skipping training candidate: rna30290.p1, not unique enough\n",
      "-skipping training candidate: rna40568.p1, not unique enough\n",
      "-skipping training candidate: rna46214.p1, not unique enough\n",
      "-skipping training candidate: rna46216.p1, not unique enough\n",
      "-skipping training candidate: rna46218.p1, not unique enough\n",
      "-skipping training candidate: rna46215.p1, not unique enough\n",
      "-skipping training candidate: rna10776.p1, not unique enough\n",
      "-skipping training candidate: rna20171.p1, not unique enough\n",
      "-skipping training candidate: rna46217.p1, not unique enough\n",
      "-skipping training candidate: rna50267.p1, not unique enough\n",
      "-skipping training candidate: rna7213.p1, not unique enough\n",
      "-skipping training candidate: rna12805.p1, not unique enough\n",
      "-skipping training candidate: rna51210.p1, not unique enough\n",
      "-skipping training candidate: rna20016.p1, not unique enough\n",
      "-skipping training candidate: rna20019.p1, not unique enough\n",
      "-skipping training candidate: rna24103.p1, not unique enough\n",
      "-skipping training candidate: rna34573.p1, not unique enough\n",
      "-skipping training candidate: rna17304.p1, not unique enough\n",
      "-skipping training candidate: rna23679.p1, not unique enough\n",
      "-skipping training candidate: rna23680.p1, not unique enough\n",
      "-skipping training candidate: rna53883.p1, not unique enough\n",
      "-skipping training candidate: rna5116.p1, not unique enough\n",
      "-skipping training candidate: rna24523.p1, not unique enough\n",
      "-skipping training candidate: rna35263.p1, not unique enough\n",
      "-skipping training candidate: rna2805.p1, not unique enough\n",
      "-skipping training candidate: rna21072.p1, not unique enough\n",
      "-skipping training candidate: rna27228.p1, not unique enough\n",
      "-skipping training candidate: rna3673.p1, not unique enough\n",
      "-skipping training candidate: rna31446.p1, not unique enough\n",
      "-skipping training candidate: rna34577.p1, not unique enough\n",
      "-skipping training candidate: rna39595.p1, not unique enough\n",
      "-skipping training candidate: rna3672.p1, not unique enough\n",
      "-skipping training candidate: rna10439.p1, not unique enough\n",
      "-skipping training candidate: rna12806.p1, not unique enough\n",
      "-skipping training candidate: rna20506.p1, not unique enough\n",
      "-skipping training candidate: rna41009.p1, not unique enough\n",
      "-skipping training candidate: rna46912.p1, not unique enough\n",
      "-skipping training candidate: rna46156.p1, not unique enough\n",
      "-skipping training candidate: rna16305.p1, not unique enough\n",
      "-skipping training candidate: rna20507.p1, not unique enough\n",
      "-skipping training candidate: rna21005.p1, not unique enough\n",
      "-skipping training candidate: rna40643.p1, not unique enough\n",
      "-skipping training candidate: rna7218.p1, not unique enough\n",
      "-skipping training candidate: rna9641.p1, not unique enough\n",
      "-skipping training candidate: rna30529.p1, not unique enough\n",
      "-skipping training candidate: rna40644.p1, not unique enough\n",
      "-skipping training candidate: rna40642.p1, not unique enough\n",
      "-skipping training candidate: rna10777.p1, not unique enough\n",
      "-skipping training candidate: rna31445.p1, not unique enough\n",
      "-skipping training candidate: rna3657.p1, not unique enough\n",
      "-skipping training candidate: rna41592.p1, not unique enough\n",
      "-skipping training candidate: rna59715.p1, not unique enough\n",
      "-skipping training candidate: rna4296.p1, not unique enough\n",
      "-skipping training candidate: rna24744.p1, not unique enough\n",
      "-skipping training candidate: rna43016.p1, not unique enough\n",
      "-skipping training candidate: rna43011.p1, not unique enough\n",
      "-skipping training candidate: rna43015.p1, not unique enough\n",
      "-skipping training candidate: rna7163.p1, not unique enough\n",
      "-skipping training candidate: rna7845.p1, not unique enough\n",
      "-skipping training candidate: rna20172.p1, not unique enough\n",
      "-skipping training candidate: rna43012.p1, not unique enough\n",
      "-skipping training candidate: rna43009.p1, not unique enough\n",
      "-skipping training candidate: rna48732.p1, not unique enough\n",
      "-skipping training candidate: rna3674.p1, not unique enough\n",
      "-skipping training candidate: rna6153.p1, not unique enough\n",
      "-skipping training candidate: rna13886.p1, not unique enough\n",
      "-skipping training candidate: rna21067.p1, not unique enough\n",
      "-skipping training candidate: rna28301.p1, not unique enough\n",
      "-skipping training candidate: rna36024.p1, not unique enough\n",
      "-skipping training candidate: rna43010.p1, not unique enough\n",
      "-skipping training candidate: rna279.p1, not unique enough\n",
      "-skipping training candidate: rna280.p1, not unique enough\n",
      "-skipping training candidate: rna281.p1, not unique enough\n",
      "-skipping training candidate: rna278.p1, not unique enough\n",
      "-skipping training candidate: rna23158.p2, not unique enough\n",
      "-skipping training candidate: rna23155.p2, not unique enough\n",
      "-skipping training candidate: rna13890.p1, not unique enough\n",
      "-skipping training candidate: rna23390.p1, not unique enough\n",
      "-skipping training candidate: rna32974.p1, not unique enough\n",
      "-skipping training candidate: rna25459.p1, not unique enough\n",
      "-skipping training candidate: rna34575.p1, not unique enough\n",
      "-skipping training candidate: rna4209.p1, not unique enough\n",
      "-skipping training candidate: rna10775.p1, not unique enough\n",
      "-skipping training candidate: rna17469.p1, not unique enough\n",
      "-skipping training candidate: rna15527.p1, not unique enough\n",
      "-skipping training candidate: rna54036.p1, not unique enough\n",
      "-skipping training candidate: rna20504.p1, not unique enough\n",
      "-skipping training candidate: rna23391.p1, not unique enough\n",
      "-skipping training candidate: rna35856.p1, not unique enough\n",
      "-skipping training candidate: rna35855.p1, not unique enough\n",
      "-skipping training candidate: rna40446.p1, not unique enough\n",
      "-skipping training candidate: rna55029.p1, not unique enough\n",
      "-skipping training candidate: rna55030.p1, not unique enough\n",
      "-skipping training candidate: rna55031.p1, not unique enough\n",
      "-skipping training candidate: rna4439.p1, not unique enough\n",
      "-skipping training candidate: rna7165.p1, not unique enough\n",
      "-skipping training candidate: rna30462.p1, not unique enough\n",
      "-skipping training candidate: rna46909.p1, not unique enough\n",
      "-skipping training candidate: rna50131.p1, not unique enough\n",
      "-skipping training candidate: rna50130.p1, not unique enough\n",
      "-skipping training candidate: rna56387.p1, not unique enough\n",
      "-skipping training candidate: rna61083.p1, not unique enough\n",
      "-skipping training candidate: rna3597.p1, not unique enough\n",
      "-skipping training candidate: rna4675.p1, not unique enough\n",
      "-skipping training candidate: rna4676.p1, not unique enough\n",
      "-skipping training candidate: rna4673.p1, not unique enough\n",
      "-skipping training candidate: rna18427.p1, not unique enough\n",
      "-skipping training candidate: rna34578.p1, not unique enough\n",
      "-skipping training candidate: rna46219.p1, not unique enough\n",
      "-skipping training candidate: rna5014.p1, not unique enough\n",
      "-skipping training candidate: rna5013.p1, not unique enough\n",
      "-skipping training candidate: rna10441.p1, not unique enough\n",
      "-skipping training candidate: rna12332.p1, not unique enough\n",
      "-skipping training candidate: rna18428.p1, not unique enough\n",
      "-skipping training candidate: rna20508.p1, not unique enough\n",
      "-skipping training candidate: rna27008.p1, not unique enough\n",
      "-skipping training candidate: rna27007.p1, not unique enough\n",
      "-skipping training candidate: rna60372.p1, not unique enough\n",
      "-skipping training candidate: rna17962.p1, not unique enough\n",
      "-skipping training candidate: rna20104.p1, not unique enough\n",
      "-skipping training candidate: rna26778.p1, not unique enough\n",
      "-skipping training candidate: rna36023.p1, not unique enough\n",
      "-skipping training candidate: rna49831.p1, not unique enough\n",
      "-skipping training candidate: rna37182.p1, not unique enough\n",
      "-skipping training candidate: rna32973.p1, not unique enough\n",
      "-skipping training candidate: rna32972.p1, not unique enough\n",
      "-skipping training candidate: rna39515.p1, not unique enough\n",
      "-skipping training candidate: rna7219.p1, not unique enough\n",
      "-skipping training candidate: rna17753.p1, not unique enough\n",
      "-skipping training candidate: rna60373.p1, not unique enough\n",
      "-skipping training candidate: rna9638.p1, not unique enough\n",
      "-skipping training candidate: rna29885.p1, not unique enough\n",
      "-skipping training candidate: rna33059.p1, not unique enough\n",
      "-skipping training candidate: rna496.p1, not unique enough\n",
      "-skipping training candidate: rna3879.p1, not unique enough\n",
      "-skipping training candidate: rna494.p1, not unique enough\n",
      "-skipping training candidate: rna4162.p1, not unique enough\n",
      "-skipping training candidate: rna7575.p1, not unique enough\n",
      "-skipping training candidate: rna15620.p1, not unique enough\n",
      "-skipping training candidate: rna23683.p1, not unique enough\n",
      "-skipping training candidate: rna23687.p1, not unique enough\n",
      "-skipping training candidate: rna23685.p1, not unique enough\n",
      "-skipping training candidate: rna30463.p1, not unique enough\n",
      "-skipping training candidate: rna60835.p1, not unique enough\n",
      "-skipping training candidate: rna19993.p1, not unique enough\n",
      "-skipping training candidate: rna25009.p1, not unique enough\n",
      "-skipping training candidate: rna32640.p1, not unique enough\n",
      "-skipping training candidate: rna33299.p1, not unique enough\n",
      "-skipping training candidate: rna13887.p1, not unique enough\n",
      "-skipping training candidate: rna22093.p1, not unique enough\n",
      "-skipping training candidate: rna30481.p1, not unique enough\n",
      "-skipping training candidate: rna49028.p1, not unique enough\n",
      "-skipping training candidate: rna7214.p1, not unique enough\n",
      "-skipping training candidate: rna30767.p1, not unique enough\n",
      "-skipping training candidate: rna49030.p1, not unique enough\n",
      "-skipping training candidate: rna49029.p1, not unique enough\n",
      "-skipping training candidate: rna1979.p1, not unique enough\n",
      "-skipping training candidate: rna4583.p1, not unique enough\n",
      "-skipping training candidate: rna10443.p1, not unique enough\n",
      "-skipping training candidate: rna26846.p1, not unique enough\n",
      "-skipping training candidate: rna46907.p1, not unique enough\n",
      "-skipping training candidate: rna6493.p1, not unique enough\n",
      "-skipping training candidate: rna6494.p1, not unique enough\n",
      "-skipping training candidate: rna46908.p1, not unique enough\n",
      "-skipping training candidate: rna9846.p1, not unique enough\n",
      "-skipping training candidate: rna60374.p1, not unique enough\n",
      "-skipping training candidate: rna20718.p1, not unique enough\n",
      "-skipping training candidate: rna33764.p1, not unique enough\n",
      "-skipping training candidate: rna36022.p1, not unique enough\n",
      "-skipping training candidate: rna48258.p1, not unique enough\n",
      "-skipping training candidate: rna26635.p1, not unique enough\n",
      "-skipping training candidate: rna33765.p1, not unique enough\n",
      "-skipping training candidate: rna48260.p1, not unique enough\n",
      "-skipping training candidate: rna48261.p1, not unique enough\n",
      "-skipping training candidate: rna49832.p1, not unique enough\n",
      "-skipping training candidate: rna9508.p1, not unique enough\n",
      "-skipping training candidate: rna40223.p1, not unique enough\n",
      "-skipping training candidate: rna40225.p1, not unique enough\n",
      "-skipping training candidate: rna49833.p1, not unique enough\n",
      "-skipping training candidate: rna11128.p1, not unique enough\n",
      "-skipping training candidate: rna14465.p1, not unique enough\n",
      "-skipping training candidate: rna19106.p1, not unique enough\n",
      "-skipping training candidate: rna19105.p1, not unique enough\n",
      "-skipping training candidate: rna19108.p1, not unique enough\n",
      "-skipping training candidate: rna37312.p1, not unique enough\n",
      "-skipping training candidate: rna44209.p1, not unique enough\n",
      "-skipping training candidate: rna51610.p1, not unique enough\n",
      "-skipping training candidate: rna20487.p1, not unique enough\n",
      "-skipping training candidate: rna22091.p1, not unique enough\n",
      "-skipping training candidate: rna47763.p1, not unique enough\n",
      "-skipping training candidate: rna47762.p1, not unique enough\n",
      "-skipping training candidate: rna50427.p1, not unique enough\n",
      "-skipping training candidate: rna49834.p1, not unique enough\n",
      "-skipping training candidate: rna13891.p1, not unique enough\n",
      "-skipping training candidate: rna2800.p1, not unique enough\n",
      "-skipping training candidate: rna18488.p1, not unique enough\n",
      "-skipping training candidate: rna18492.p1, not unique enough\n",
      "-skipping training candidate: rna18491.p1, not unique enough\n",
      "-skipping training candidate: rna34276.p1, not unique enough\n",
      "-skipping training candidate: rna37856.p1, not unique enough\n",
      "-skipping training candidate: rna50425.p1, not unique enough\n",
      "-skipping training candidate: rna4210.p1, not unique enough\n",
      "-skipping training candidate: rna17725.p1, not unique enough\n",
      "-skipping training candidate: rna18490.p1, not unique enough\n",
      "-skipping training candidate: rna18487.p1, not unique enough\n",
      "-skipping training candidate: rna20488.p1, not unique enough\n",
      "-skipping training candidate: rna4375.p1, not unique enough\n",
      "-skipping training candidate: rna51932.p1, not unique enough\n",
      "-skipping training candidate: rna51935.p1, not unique enough\n",
      "-skipping training candidate: rna51931.p1, not unique enough\n",
      "-skipping training candidate: rna37772.p1, not unique enough\n",
      "-skipping training candidate: rna48256.p1, not unique enough\n",
      "-skipping training candidate: rna6572.p1, not unique enough\n",
      "-skipping training candidate: rna7215.p1, not unique enough\n",
      "-skipping training candidate: rna17961.p1, not unique enough\n",
      "-skipping training candidate: rna30289.p1, not unique enough\n",
      "-skipping training candidate: rna35059.p1, not unique enough\n",
      "-skipping training candidate: rna48257.p1, not unique enough\n",
      "-skipping training candidate: rna43008.p1, not unique enough\n",
      "-skipping training candidate: rna50980.p1, not unique enough\n",
      "-skipping training candidate: rna52428.p1, not unique enough\n",
      "-skipping training candidate: rna52429.p1, not unique enough\n",
      "-skipping training candidate: rna7222.p1, not unique enough\n",
      "-skipping training candidate: rna19388.p1, not unique enough\n",
      "-skipping training candidate: rna38308.p1, not unique enough\n",
      "-skipping training candidate: rna40579.p1, not unique enough\n",
      "-skipping training candidate: rna45949.p1, not unique enough\n",
      "-skipping training candidate: rna45950.p1, not unique enough\n",
      "-skipping training candidate: rna45945.p1, not unique enough\n",
      "-skipping training candidate: rna53392.p1, not unique enough\n",
      "-skipping training candidate: rna53390.p1, not unique enough\n",
      "-skipping training candidate: rna3398.p1, not unique enough\n",
      "-skipping training candidate: rna17665.p1, not unique enough\n",
      "-skipping training candidate: rna21782.p1, not unique enough\n",
      "-skipping training candidate: rna21783.p1, not unique enough\n",
      "-skipping training candidate: rna27011.p1, not unique enough\n",
      "-skipping training candidate: rna27015.p1, not unique enough\n",
      "-skipping training candidate: rna27013.p1, not unique enough\n",
      "-skipping training candidate: rna37283.p1, not unique enough\n",
      "-skipping training candidate: rna47371.p1, not unique enough\n",
      "-skipping training candidate: rna49031.p1, not unique enough\n",
      "-skipping training candidate: rna60839.p1, not unique enough\n",
      "-skipping training candidate: rna60840.p1, not unique enough\n",
      "-skipping training candidate: rna60838.p1, not unique enough\n",
      "-skipping training candidate: rna60841.p1, not unique enough\n",
      "-skipping training candidate: rna14457.p1, not unique enough\n",
      "-skipping training candidate: rna47374.p1, not unique enough\n",
      "-skipping training candidate: rna53602.p1, not unique enough\n",
      "-skipping training candidate: rna20724.p1, not unique enough\n",
      "-skipping training candidate: rna37282.p1, not unique enough\n",
      "-skipping training candidate: rna25008.p1, not unique enough\n",
      "-skipping training candidate: rna42077.p1, not unique enough\n",
      "-skipping training candidate: rna48263.p1, not unique enough\n",
      "-skipping training candidate: rna33300.p1, not unique enough\n",
      "-skipping training candidate: rna34812.p1, not unique enough\n",
      "-skipping training candidate: rna37511.p1, not unique enough\n",
      "-skipping training candidate: rna43240.p1, not unique enough\n",
      "-skipping training candidate: rna20676.p1, not unique enough\n",
      "-skipping training candidate: rna20680.p1, not unique enough\n",
      "-skipping training candidate: rna20677.p1, not unique enough\n",
      "-skipping training candidate: rna20679.p1, not unique enough\n",
      "-skipping training candidate: rna20678.p1, not unique enough\n",
      "-skipping training candidate: rna33057.p1, not unique enough\n",
      "-skipping training candidate: rna37309.p1, not unique enough\n",
      "-skipping training candidate: rna43242.p1, not unique enough\n",
      "-skipping training candidate: rna53736.p1, not unique enough\n",
      "-skipping training candidate: rna53737.p1, not unique enough\n",
      "-skipping training candidate: rna56436.p1, not unique enough\n",
      "-skipping training candidate: rna55579.p1, not unique enough\n",
      "-skipping training candidate: rna12705.p1, not unique enough\n",
      "-skipping training candidate: rna12707.p1, not unique enough\n",
      "-skipping training candidate: rna26637.p1, not unique enough\n",
      "-skipping training candidate: rna30769.p1, not unique enough\n",
      "-skipping training candidate: rna32594.p1, not unique enough\n",
      "-skipping training candidate: rna37308.p1, not unique enough\n",
      "-skipping training candidate: rna20719.p1, not unique enough\n",
      "-skipping training candidate: rna24721.p1, not unique enough\n",
      "-skipping training candidate: rna24724.p1, not unique enough\n",
      "-skipping training candidate: rna24720.p1, not unique enough\n",
      "-skipping training candidate: rna24722.p1, not unique enough\n",
      "-skipping training candidate: rna9847.p1, not unique enough\n",
      "-skipping training candidate: rna20106.p1, not unique enough\n",
      "-skipping training candidate: rna20108.p1, not unique enough\n",
      "-skipping training candidate: rna20107.p1, not unique enough\n",
      "-skipping training candidate: rna20109.p1, not unique enough\n",
      "-skipping training candidate: rna40946.p1, not unique enough\n",
      "-skipping training candidate: rna43018.p1, not unique enough\n",
      "-skipping training candidate: rna49835.p1, not unique enough\n",
      "-skipping training candidate: rna33301.p1, not unique enough\n",
      "-skipping training candidate: rna34278.p1, not unique enough\n",
      "-skipping training candidate: rna51460.p1, not unique enough\n",
      "-skipping training candidate: rna54844.p1, not unique enough\n",
      "-skipping training candidate: rna20018.p1, not unique enough\n",
      "-skipping training candidate: rna34579.p1, not unique enough\n",
      "-skipping training candidate: rna10002.p1, not unique enough\n",
      "-skipping training candidate: rna10003.p1, not unique enough\n",
      "-skipping training candidate: rna30770.p1, not unique enough\n",
      "-skipping training candidate: rna52953.p1, not unique enough\n",
      "-skipping training candidate: rna52954.p1, not unique enough\n",
      "-skipping training candidate: rna10004.p1, not unique enough\n",
      "-skipping training candidate: rna33058.p1, not unique enough\n",
      "-skipping training candidate: rna25888.p1, not unique enough\n",
      "-skipping training candidate: rna34813.p1, not unique enough\n",
      "-skipping training candidate: rna34648.p1, not unique enough\n",
      "-skipping training candidate: rna4655.p1, not unique enough\n",
      "-skipping training candidate: rna10521.p1, not unique enough\n",
      "-skipping training candidate: rna8393.p1, not unique enough\n",
      "-skipping training candidate: rna8391.p1, not unique enough\n",
      "-skipping training candidate: rna47373.p1, not unique enough\n",
      "-skipping training candidate: rna57150.p1, not unique enough\n",
      "-skipping training candidate: rna10438.p1, not unique enough\n",
      "-skipping training candidate: rna28650.p1, not unique enough\n",
      "-skipping training candidate: rna40947.p1, not unique enough\n",
      "-skipping training candidate: rna40583.p1, not unique enough\n",
      "-skipping training candidate: rna44819.p1, not unique enough\n",
      "-skipping training candidate: rna53976.p1, not unique enough\n",
      "-skipping training candidate: rna7937.p1, not unique enough\n",
      "-skipping training candidate: rna25892.p1, not unique enough\n",
      "-skipping training candidate: rna30480.p1, not unique enough\n",
      "-skipping training candidate: rna57122.p1, not unique enough\n",
      "-skipping training candidate: rna60834.p1, not unique enough\n",
      "-skipping training candidate: rna1962.p1, not unique enough\n",
      "-skipping training candidate: rna2669.p1, not unique enough\n",
      "-skipping training candidate: rna2667.p1, not unique enough\n",
      "-skipping training candidate: rna14121.p1, not unique enough\n",
      "-skipping training candidate: rna23157.p2, not unique enough\n",
      "-skipping training candidate: rna26780.p1, not unique enough\n",
      "-skipping training candidate: rna36611.p1, not unique enough\n",
      "-skipping training candidate: rna4530.p1, not unique enough\n",
      "-skipping training candidate: rna4527.p1, not unique enough\n",
      "-skipping training candidate: rna4532.p1, not unique enough\n",
      "-skipping training candidate: rna4528.p1, not unique enough\n",
      "-skipping training candidate: rna4531.p1, not unique enough\n",
      "-skipping training candidate: rna9640.p1, not unique enough\n",
      "-skipping training candidate: rna12210.p1, not unique enough\n",
      "-skipping training candidate: rna12211.p1, not unique enough\n",
      "-skipping training candidate: rna38263.p1, not unique enough\n",
      "-skipping training candidate: rna7000.p1, not unique enough\n",
      "-skipping training candidate: rna7001.p1, not unique enough\n",
      "-skipping training candidate: rna6999.p1, not unique enough\n",
      "-skipping training candidate: rna52875.p1, not unique enough\n",
      "-skipping training candidate: rna53977.p1, not unique enough\n",
      "-skipping training candidate: rna17664.p1, not unique enough\n",
      "-skipping training candidate: rna1434.p1, not unique enough\n",
      "-skipping training candidate: rna6453.p1, not unique enough\n",
      "-skipping training candidate: rna6452.p1, not unique enough\n",
      "-skipping training candidate: rna17756.p1, not unique enough\n",
      "-skipping training candidate: rna17758.p1, not unique enough\n",
      "-skipping training candidate: rna17759.p1, not unique enough\n",
      "-skipping training candidate: rna17757.p1, not unique enough\n",
      "-skipping training candidate: rna17761.p1, not unique enough\n",
      "-skipping training candidate: rna17760.p1, not unique enough\n",
      "-skipping training candidate: rna44210.p1, not unique enough\n",
      "-skipping training candidate: rna44208.p1, not unique enough\n",
      "-skipping training candidate: rna44211.p1, not unique enough\n",
      "-skipping training candidate: rna51462.p1, not unique enough\n",
      "-skipping training candidate: rna53975.p1, not unique enough\n",
      "-skipping training candidate: rna57120.p1, not unique enough\n",
      "-skipping training candidate: rna2458.p1, not unique enough\n",
      "-skipping training candidate: rna3675.p1, not unique enough\n",
      "-skipping training candidate: rna25776.p1, not unique enough\n",
      "-skipping training candidate: rna51934.p1, not unique enough\n",
      "-skipping training candidate: rna2459.p1, not unique enough\n",
      "-skipping training candidate: rna2456.p1, not unique enough\n",
      "-skipping training candidate: rna1435.p1, not unique enough\n",
      "-skipping training candidate: rna13888.p1, not unique enough\n",
      "-skipping training candidate: rna25890.p1, not unique enough\n",
      "-skipping training candidate: rna28649.p1, not unique enough\n",
      "-skipping training candidate: rna35879.p1, not unique enough\n",
      "-skipping training candidate: rna52872.p1, not unique enough\n",
      "-skipping training candidate: rna52878.p1, not unique enough\n",
      "-skipping training candidate: rna57117.p1, not unique enough\n",
      "-skipping training candidate: rna1855.p1, not unique enough\n",
      "-skipping training candidate: rna1853.p1, not unique enough\n",
      "-skipping training candidate: rna1851.p1, not unique enough\n",
      "-skipping training candidate: rna1854.p1, not unique enough\n",
      "-skipping training candidate: rna1852.p1, not unique enough\n",
      "-skipping training candidate: rna329.p1, not unique enough\n",
      "-skipping training candidate: rna327.p1, not unique enough\n",
      "-skipping training candidate: rna25889.p1, not unique enough\n",
      "-skipping training candidate: rna31390.p1, not unique enough\n",
      "-skipping training candidate: rna51459.p1, not unique enough\n",
      "-skipping training candidate: rna8996.p1, not unique enough\n",
      "-skipping training candidate: rna10919.p1, not unique enough\n",
      "-skipping training candidate: rna14124.p1, not unique enough\n",
      "-skipping training candidate: rna21362.p1, not unique enough\n",
      "-skipping training candidate: rna46226.p1, not unique enough\n",
      "-skipping training candidate: rna53980.p1, not unique enough\n",
      "-skipping training candidate: rna15371.p1, not unique enough\n",
      "-skipping training candidate: rna26847.p1, not unique enough\n",
      "-skipping training candidate: rna20720.p1, not unique enough\n",
      "-skipping training candidate: rna37307.p1, not unique enough\n",
      "-skipping training candidate: rna57121.p1, not unique enough\n",
      "-skipping training candidate: rna11119.p1, not unique enough\n",
      "-skipping training candidate: rna14122.p1, not unique enough\n",
      "-skipping training candidate: rna20722.p1, not unique enough\n",
      "-skipping training candidate: rna25895.p1, not unique enough\n",
      "-skipping training candidate: rna38265.p1, not unique enough\n",
      "-skipping training candidate: rna42503.p1, not unique enough\n",
      "-skipping training candidate: rna30514.p1, not unique enough\n",
      "-skipping training candidate: rna45948.p1, not unique enough\n",
      "-skipping training candidate: rna45951.p1, not unique enough\n",
      "-skipping training candidate: rna45946.p1, not unique enough\n",
      "-skipping training candidate: rna55846.p1, not unique enough\n",
      "-skipping training candidate: rna5258.p1, not unique enough\n",
      "-skipping training candidate: rna25893.p1, not unique enough\n",
      "-skipping training candidate: rna36020.p1, not unique enough\n",
      "-skipping training candidate: rna46417.p1, not unique enough\n",
      "-skipping training candidate: rna7002.p1, not unique enough\n",
      "-skipping training candidate: rna10442.p1, not unique enough\n",
      "-skipping training candidate: rna33302.p1, not unique enough\n",
      "-skipping training candidate: rna497.p1, not unique enough\n",
      "-skipping training candidate: rna1981.p1, not unique enough\n",
      "-skipping training candidate: rna798.p1, not unique enough\n",
      "-skipping training candidate: rna14459.p1, not unique enough\n",
      "-skipping training candidate: rna32378.p1, not unique enough\n",
      "-skipping training candidate: rna37310.p1, not unique enough\n",
      "-skipping training candidate: rna8881.p1, not unique enough\n",
      "-skipping training candidate: rna8883.p1, not unique enough\n",
      "-skipping training candidate: rna16204.p1, not unique enough\n",
      "-skipping training candidate: rna28653.p1, not unique enough\n",
      "-skipping training candidate: rna30771.p1, not unique enough\n",
      "-skipping training candidate: rna37256.p1, not unique enough\n",
      "-skipping training candidate: rna37254.p1, not unique enough\n",
      "-skipping training candidate: rna18493.p1, not unique enough\n",
      "-skipping training candidate: rna25898.p1, not unique enough\n",
      "-skipping training candidate: rna36343.p1, not unique enough\n",
      "-skipping training candidate: rna35583.p1, not unique enough\n",
      "-skipping training candidate: rna40580.p1, not unique enough\n",
      "-skipping training candidate: rna10036.p1, not unique enough\n",
      "-skipping training candidate: rna19386.p1, not unique enough\n",
      "-skipping training candidate: rna37257.p1, not unique enough\n",
      "-skipping training candidate: rna44275.p1, not unique enough\n",
      "-skipping training candidate: rna57116.p1, not unique enough\n",
      "-skipping training candidate: rna8882.p1, not unique enough\n",
      "-skipping training candidate: rna28654.p1, not unique enough\n",
      "-skipping training candidate: rna50863.p1, not unique enough\n",
      "-skipping training candidate: rna50864.p1, not unique enough\n",
      "-skipping training candidate: rna25899.p1, not unique enough\n",
      "-skipping training candidate: rna34302.p1, not unique enough\n",
      "-skipping training candidate: rna34300.p1, not unique enough\n",
      "-skipping training candidate: rna42407.p1, not unique enough\n",
      "-skipping training candidate: rna46596.p1, not unique enough\n",
      "-skipping training candidate: rna15049.p1, not unique enough\n",
      "-skipping training candidate: rna17332.p1, not unique enough\n",
      "-skipping training candidate: rna44276.p1, not unique enough\n",
      "-skipping training candidate: rna13889.p1, not unique enough\n",
      "-skipping training candidate: rna32653.p1, not unique enough\n",
      "-skipping training candidate: rna17663.p1, not unique enough\n",
      "-skipping training candidate: rna26848.p1, not unique enough\n",
      "-skipping training candidate: rna44272.p1, not unique enough\n",
      "-skipping training candidate: rna53495.p1, not unique enough\n",
      "-skipping training candidate: rna4668.p1, not unique enough\n",
      "-skipping training candidate: rna4670.p1, not unique enough\n",
      "-skipping training candidate: rna4667.p1, not unique enough\n",
      "-skipping training candidate: rna17661.p1, not unique enough\n",
      "-skipping training candidate: rna39816.p1, not unique enough\n",
      "-skipping training candidate: rna50662.p1, not unique enough\n",
      "-skipping training candidate: rna55848.p1, not unique enough\n",
      "-skipping training candidate: rna56439.p1, not unique enough\n",
      "-skipping training candidate: rna5335.p1, not unique enough\n",
      "-skipping training candidate: rna37305.p1, not unique enough\n",
      "-skipping training candidate: rna46597.p1, not unique enough\n",
      "-skipping training candidate: rna53982.p1, not unique enough\n",
      "-skipping training candidate: rna14123.p1, not unique enough\n",
      "-skipping training candidate: rna17333.p1, not unique enough\n",
      "-skipping training candidate: rna20725.p1, not unique enough\n",
      "-skipping training candidate: rna25896.p1, not unique enough\n",
      "-skipping training candidate: rna46600.p1, not unique enough\n",
      "-skipping training candidate: rna53979.p1, not unique enough\n",
      "-skipping training candidate: rna15878.p1, not unique enough\n",
      "-skipping training candidate: rna15876.p1, not unique enough\n",
      "-skipping training candidate: rna15879.p1, not unique enough\n",
      "-skipping training candidate: rna15880.p1, not unique enough\n",
      "-skipping training candidate: rna24745.p1, not unique enough\n",
      "-skipping training candidate: rna27890.p1, not unique enough\n",
      "-skipping training candidate: rna34461.p1, not unique enough\n",
      "-skipping training candidate: rna57119.p1, not unique enough\n",
      "-skipping training candidate: rna15048.p1, not unique enough\n",
      "-skipping training candidate: rna20970.p1, not unique enough\n",
      "-skipping training candidate: rna24527.p1, not unique enough\n",
      "-skipping training candidate: rna39758.p1, not unique enough\n",
      "-skipping training candidate: rna39759.p1, not unique enough\n",
      "-skipping training candidate: rna56526.p1, not unique enough\n",
      "-skipping training candidate: rna20721.p1, not unique enough\n",
      "-skipping training candidate: rna37311.p1, not unique enough\n",
      "-skipping training candidate: rna52957.p1, not unique enough\n",
      "-skipping training candidate: rna12224.p1, not unique enough\n",
      "-skipping training candidate: rna42406.p1, not unique enough\n",
      "-skipping training candidate: rna25897.p1, not unique enough\n",
      "-skipping training candidate: rna27681.p1, not unique enough\n",
      "-skipping training candidate: rna44781.p1, not unique enough\n",
      "-skipping training candidate: rna44782.p1, not unique enough\n",
      "-skipping training candidate: rna53130.p1, not unique enough\n",
      "-skipping training candidate: rna2455.p1, not unique enough\n",
      "-skipping training candidate: rna10779.p1, not unique enough\n",
      "-skipping training candidate: rna14456.p1, not unique enough\n",
      "-skipping training candidate: rna12808.p1, not unique enough\n",
      "-skipping training candidate: rna27895.p1, not unique enough\n",
      "-skipping training candidate: rna38266.p1, not unique enough\n",
      "-skipping training candidate: rna38331.p1, not unique enough\n",
      "-skipping training candidate: rna53981.p1, not unique enough\n",
      "-skipping training candidate: rna53537.p1, not unique enough\n",
      "-skipping training candidate: rna14125.p1, not unique enough\n",
      "-skipping training candidate: rna34650.p1, not unique enough\n",
      "-skipping training candidate: rna34649.p1, not unique enough\n",
      "-skipping training candidate: rna38941.p1, not unique enough\n",
      "-skipping training candidate: rna42508.p1, not unique enough\n",
      "-skipping training candidate: rna48597.p1, not unique enough\n",
      "-skipping training candidate: rna25900.p1, not unique enough\n",
      "-skipping training candidate: rna34497.p1, not unique enough\n",
      "-skipping training candidate: rna38125.p1, not unique enough\n",
      "-skipping training candidate: rna44271.p1, not unique enough\n",
      "-skipping training candidate: rna57115.p1, not unique enough\n",
      "-skipping training candidate: rna8401.p1, not unique enough\n",
      "-skipping training candidate: rna12634.p1, not unique enough\n",
      "-skipping training candidate: rna26493.p1, not unique enough\n",
      "-skipping training candidate: rna38124.p1, not unique enough\n",
      "-skipping training candidate: rna38123.p1, not unique enough\n",
      "-skipping training candidate: rna26677.p1, not unique enough\n",
      "-skipping training candidate: rna36869.p1, not unique enough\n",
      "-skipping training candidate: rna37247.p1, not unique enough\n",
      "-skipping training candidate: rna51384.p1, not unique enough\n",
      "-skipping training candidate: rna5337.p1, not unique enough\n",
      "-skipping training candidate: rna9032.p1, not unique enough\n",
      "-skipping training candidate: rna28651.p1, not unique enough\n",
      "-skipping training candidate: rna56525.p1, not unique enough\n",
      "-skipping training candidate: rna221.p1, not unique enough\n",
      "-skipping training candidate: rna220.p1, not unique enough\n",
      "-skipping training candidate: rna219.p1, not unique enough\n",
      "-skipping training candidate: rna1430.p1, not unique enough\n",
      "-skipping training candidate: rna25891.p1, not unique enough\n",
      "-skipping training candidate: rna24528.p1, not unique enough\n",
      "-skipping training candidate: rna37246.p1, not unique enough\n",
      "-skipping training candidate: rna51461.p1, not unique enough\n",
      "-skipping training candidate: rna218.p1, not unique enough\n",
      "-skipping training candidate: rna13431.p1, not unique enough\n",
      "-skipping training candidate: rna25903.p1, not unique enough\n",
      "-skipping training candidate: rna42547.p1, not unique enough\n",
      "-skipping training candidate: rna46531.p1, not unique enough\n",
      "-skipping training candidate: rna53248.p1, not unique enough\n",
      "-skipping training candidate: rna5978.p1, not unique enough\n",
      "-skipping training candidate: rna5979.p1, not unique enough\n",
      "-skipping training candidate: rna28652.p1, not unique enough\n",
      "-skipping training candidate: rna38126.p1, not unique enough\n",
      "-skipping training candidate: rna48593.p1, not unique enough\n",
      "-skipping training candidate: rna58193.p1, not unique enough\n",
      "-skipping training candidate: rna27898.p1, not unique enough\n",
      "-skipping training candidate: rna32941.p1, not unique enough\n",
      "-skipping training candidate: rna39814.p1, not unique enough\n",
      "-skipping training candidate: rna15592.p1, not unique enough\n",
      "-skipping training candidate: rna17662.p1, not unique enough\n",
      "-skipping training candidate: rna53983.p1, not unique enough\n",
      "-skipping training candidate: rna8378.p1, not unique enough\n",
      "-skipping training candidate: rna17660.p1, not unique enough\n",
      "-skipping training candidate: rna34495.p1, not unique enough\n",
      "-skipping training candidate: rna1721.p1, not unique enough\n",
      "-skipping training candidate: rna8379.p1, not unique enough\n",
      "-skipping training candidate: rna13398.p1, not unique enough\n",
      "-skipping training candidate: rna27599.p1, not unique enough\n",
      "-skipping training candidate: rna30516.p1, not unique enough\n",
      "-skipping training candidate: rna30517.p1, not unique enough\n",
      "-skipping training candidate: rna30515.p1, not unique enough\n",
      "-skipping training candidate: rna36872.p1, not unique enough\n",
      "-skipping training candidate: rna36871.p1, not unique enough\n",
      "-skipping training candidate: rna43017.p1, not unique enough\n",
      "-skipping training candidate: rna56437.p1, not unique enough\n",
      "-skipping training candidate: rna57118.p1, not unique enough\n",
      "-skipping training candidate: rna1255.p1, not unique enough\n",
      "-skipping training candidate: rna17726.p1, not unique enough\n",
      "-skipping training candidate: rna20726.p1, not unique enough\n",
      "-skipping training candidate: rna44820.p1, not unique enough\n",
      "-skipping training candidate: rna1261.p1, not unique enough\n",
      "-skipping training candidate: rna1258.p1, not unique enough\n",
      "-skipping training candidate: rna15450.p1, not unique enough\n",
      "-skipping training candidate: rna17297.p1, not unique enough\n",
      "-skipping training candidate: rna50344.p1, not unique enough\n",
      "-skipping training candidate: rna48594.p1, not unique enough\n",
      "-skipping training candidate: rna52873.p1, not unique enough\n",
      "-skipping training candidate: rna52877.p1, not unique enough\n",
      "-skipping training candidate: rna52874.p1, not unique enough\n",
      "-skipping training candidate: rna54327.p1, not unique enough\n",
      "-skipping training candidate: rna1259.p1, not unique enough\n",
      "-skipping training candidate: rna5339.p1, not unique enough\n",
      "-skipping training candidate: rna50105.p1, not unique enough\n",
      "-skipping training candidate: rna50103.p1, not unique enough\n",
      "-skipping training candidate: rna50106.p1, not unique enough\n",
      "-skipping training candidate: rna54646.p1, not unique enough\n",
      "-skipping training candidate: rna4545.p1, not unique enough\n",
      "-skipping training candidate: rna4544.p1, not unique enough\n",
      "-skipping training candidate: rna20015.p1, not unique enough\n",
      "-skipping training candidate: rna27892.p1, not unique enough\n",
      "-skipping training candidate: rna41819.p1, not unique enough\n",
      "-skipping training candidate: rna58929.p1, not unique enough\n",
      "-skipping training candidate: rna14403.p1, not unique enough\n",
      "-skipping training candidate: rna32939.p1, not unique enough\n",
      "-skipping training candidate: rna41820.p1, not unique enough\n",
      "-skipping training candidate: rna41823.p1, not unique enough\n",
      "-skipping training candidate: rna7825.p1, not unique enough\n",
      "-skipping training candidate: rna32933.p1, not unique enough\n",
      "-skipping training candidate: rna36910.p1, not unique enough\n",
      "-skipping training candidate: rna36911.p1, not unique enough\n",
      "-skipping training candidate: rna48595.p1, not unique enough\n",
      "-skipping training candidate: rna52956.p1, not unique enough\n",
      "-skipping training candidate: rna7823.p1, not unique enough\n",
      "-skipping training candidate: rna5340.p1, not unique enough\n",
      "-skipping training candidate: rna34407.p1, not unique enough\n",
      "-skipping training candidate: rna34487.p1, not unique enough\n",
      "-skipping training candidate: rna36334.p1, not unique enough\n",
      "-skipping training candidate: rna36337.p1, not unique enough\n",
      "-skipping training candidate: rna46599.p1, not unique enough\n",
      "-skipping training candidate: rna48259.p1, not unique enough\n",
      "-skipping training candidate: rna50650.p1, not unique enough\n",
      "-skipping training candidate: rna57869.p1, not unique enough\n",
      "-skipping training candidate: rna21507.p1, not unique enough\n",
      "-skipping training candidate: rna36339.p1, not unique enough\n",
      "-skipping training candidate: rna36338.p1, not unique enough\n",
      "-skipping training candidate: rna50104.p1, not unique enough\n",
      "-skipping training candidate: rna50651.p1, not unique enough\n",
      "-skipping training candidate: rna97.p1, not unique enough\n",
      "-skipping training candidate: rna5336.p1, not unique enough\n",
      "-skipping training candidate: rna15453.p1, not unique enough\n",
      "-skipping training candidate: rna15452.p1, not unique enough\n",
      "-skipping training candidate: rna32938.p1, not unique enough\n",
      "-skipping training candidate: rna57870.p1, not unique enough\n",
      "-skipping training candidate: rna4546.p1, not unique enough\n",
      "-skipping training candidate: rna57867.p1, not unique enough\n",
      "-skipping training candidate: rna16206.p1, not unique enough\n",
      "-skipping training candidate: rna15758.p1, not unique enough\n",
      "-skipping training candidate: rna25894.p1, not unique enough\n",
      "-skipping training candidate: rna33541.p1, not unique enough\n",
      "-skipping training candidate: rna60819.p1, not unique enough\n",
      "-skipping training candidate: rna40593.p1, not unique enough\n",
      "-skipping training candidate: rna40581.p1, not unique enough\n",
      "-skipping training candidate: rna54328.p1, not unique enough\n",
      "-skipping training candidate: rna51720.p1, not unique enough\n",
      "-skipping training candidate: rna51721.p1, not unique enough\n",
      "-skipping training candidate: rna51722.p1, not unique enough\n",
      "-skipping training candidate: rna58930.p1, not unique enough\n",
      "-skipping training candidate: rna7824.p1, not unique enough\n",
      "-skipping training candidate: rna21360.p1, not unique enough\n",
      "-skipping training candidate: rna27602.p1, not unique enough\n",
      "-skipping training candidate: rna36340.p1, not unique enough\n",
      "-skipping training candidate: rna2457.p1, not unique enough\n",
      "-skipping training candidate: rna4876.p1, not unique enough\n",
      "-skipping training candidate: rna15455.p1, not unique enough\n",
      "-skipping training candidate: rna36341.p1, not unique enough\n",
      "-skipping training candidate: rna36335.p1, not unique enough\n",
      "-skipping training candidate: rna35880.p1, not unique enough\n",
      "-skipping training candidate: rna4875.p1, not unique enough\n",
      "-skipping training candidate: rna41822.p1, not unique enough\n",
      "-skipping training candidate: rna52301.p1, not unique enough\n",
      "-skipping training candidate: rna14290.p1, not unique enough\n",
      "-skipping training candidate: rna24218.p1, not unique enough\n",
      "-skipping training candidate: rna36204.p1, not unique enough\n",
      "-skipping training candidate: rna58931.p1, not unique enough\n",
      "-skipping training candidate: rna17449.p1, not unique enough\n",
      "-skipping training candidate: rna28655.p1, not unique enough\n",
      "-skipping training candidate: rna37184.p1, not unique enough\n",
      "-skipping training candidate: rna51339.p1, not unique enough\n",
      "-skipping training candidate: rna20869.p1, not unique enough\n",
      "-skipping training candidate: rna25901.p1, not unique enough\n",
      "-skipping training candidate: rna32937.p1, not unique enough\n",
      "-skipping training candidate: rna58926.p1, not unique enough\n",
      "-skipping training candidate: rna58927.p1, not unique enough\n",
      "-skipping training candidate: rna58928.p1, not unique enough\n",
      "-skipping training candidate: rna15052.p1, not unique enough\n",
      "-skipping training candidate: rna17266.p1, not unique enough\n",
      "-skipping training candidate: rna17267.p1, not unique enough\n",
      "-skipping training candidate: rna17264.p1, not unique enough\n",
      "-skipping training candidate: rna17265.p1, not unique enough\n",
      "-skipping training candidate: rna23432.p1, not unique enough\n",
      "-skipping training candidate: rna32936.p1, not unique enough\n",
      "-skipping training candidate: rna33837.p1, not unique enough\n",
      "-skipping training candidate: rna45616.p1, not unique enough\n",
      "-skipping training candidate: rna10098.p1, not unique enough\n",
      "-skipping training candidate: rna24624.p1, not unique enough\n",
      "-skipping training candidate: rna27603.p1, not unique enough\n",
      "-skipping training candidate: rna57265.p1, not unique enough\n",
      "-skipping training candidate: rna58589.p1, not unique enough\n",
      "-skipping training candidate: rna4548.p1, not unique enough\n",
      "-skipping training candidate: rna4877.p1, not unique enough\n",
      "-skipping training candidate: rna27894.p1, not unique enough\n",
      "-skipping training candidate: rna32395.p1, not unique enough\n",
      "-skipping training candidate: rna36648.p1, not unique enough\n",
      "-skipping training candidate: rna36647.p1, not unique enough\n",
      "-skipping training candidate: rna36650.p1, not unique enough\n",
      "-skipping training candidate: rna38730.p1, not unique enough\n",
      "-skipping training candidate: rna38732.p1, not unique enough\n",
      "-skipping training candidate: rna38731.p1, not unique enough\n",
      "-skipping training candidate: rna38728.p1, not unique enough\n",
      "-skipping training candidate: rna54224.p1, not unique enough\n",
      "-skipping training candidate: rna8349.p1, not unique enough\n",
      "-skipping training candidate: rna19290.p1, not unique enough\n",
      "-skipping training candidate: rna32394.p1, not unique enough\n",
      "-skipping training candidate: rna37306.p1, not unique enough\n",
      "-skipping training candidate: rna43413.p1, not unique enough\n",
      "-skipping training candidate: rna43416.p1, not unique enough\n",
      "-skipping training candidate: rna43415.p1, not unique enough\n",
      "-skipping training candidate: rna43414.p1, not unique enough\n",
      "-skipping training candidate: rna14276.p1, not unique enough\n",
      "-skipping training candidate: rna14278.p1, not unique enough\n",
      "-skipping training candidate: rna40041.p1, not unique enough\n",
      "-skipping training candidate: rna43412.p1, not unique enough\n",
      "-skipping training candidate: rna50346.p1, not unique enough\n",
      "-skipping training candidate: rna7828.p1, not unique enough\n",
      "-skipping training candidate: rna23435.p1, not unique enough\n",
      "-skipping training candidate: rna24923.p1, not unique enough\n",
      "-skipping training candidate: rna1260.p1, not unique enough\n",
      "-skipping training candidate: rna7822.p1, not unique enough\n",
      "-skipping training candidate: rna15050.p1, not unique enough\n",
      "-skipping training candidate: rna14034.p1, not unique enough\n",
      "-skipping training candidate: rna43013.p1, not unique enough\n",
      "-skipping training candidate: rna46628.p1, not unique enough\n",
      "-skipping training candidate: rna58932.p1, not unique enough\n",
      "-skipping training candidate: rna1262.p1, not unique enough\n",
      "-skipping training candidate: rna23403.p1, not unique enough\n",
      "-skipping training candidate: rna34498.p1, not unique enough\n",
      "-skipping training candidate: rna43359.p1, not unique enough\n",
      "-skipping training candidate: rna1263.p1, not unique enough\n",
      "-skipping training candidate: rna16912.p1, not unique enough\n",
      "-skipping training candidate: rna16913.p1, not unique enough\n",
      "-skipping training candidate: rna24971.p1, not unique enough\n",
      "-skipping training candidate: rna24972.p1, not unique enough\n",
      "-skipping training candidate: rna27604.p1, not unique enough\n",
      "-skipping training candidate: rna31285.p1, not unique enough\n",
      "-skipping training candidate: rna34914.p1, not unique enough\n",
      "-skipping training candidate: rna35186.p1, not unique enough\n",
      "-skipping training candidate: rna40646.p1, not unique enough\n",
      "-skipping training candidate: rna10918.p1, not unique enough\n",
      "-skipping training candidate: rna21304.p1, not unique enough\n",
      "-skipping training candidate: rna32641.p1, not unique enough\n",
      "-skipping training candidate: rna35265.p1, not unique enough\n",
      "-skipping training candidate: rna35185.p1, not unique enough\n",
      "-skipping training candidate: rna4547.p1, not unique enough\n",
      "-skipping training candidate: rna16911.p1, not unique enough\n",
      "-skipping training candidate: rna17887.p1, not unique enough\n",
      "-skipping training candidate: rna35878.p1, not unique enough\n",
      "-skipping training candidate: rna40645.p1, not unique enough\n",
      "-skipping training candidate: rna40397.p1, not unique enough\n",
      "-skipping training candidate: rna40398.p1, not unique enough\n",
      "-skipping training candidate: rna58590.p1, not unique enough\n",
      "-skipping training candidate: rna6027.p1, not unique enough\n",
      "-skipping training candidate: rna21306.p1, not unique enough\n",
      "-skipping training candidate: rna35877.p1, not unique enough\n",
      "-skipping training candidate: rna10911.p1, not unique enough\n",
      "-skipping training candidate: rna10912.p1, not unique enough\n",
      "-skipping training candidate: rna15242.p1, not unique enough\n",
      "-skipping training candidate: rna38920.p1, not unique enough\n",
      "-skipping training candidate: rna7827.p1, not unique enough\n",
      "-skipping training candidate: rna15454.p1, not unique enough\n",
      "-skipping training candidate: rna23738.p1, not unique enough\n",
      "-skipping training candidate: rna32935.p1, not unique enough\n",
      "-skipping training candidate: rna3738.p1, not unique enough\n",
      "-skipping training candidate: rna7821.p1, not unique enough\n",
      "-skipping training candidate: rna5338.p1, not unique enough\n",
      "-skipping training candidate: rna22320.p1, not unique enough\n",
      "-skipping training candidate: rna23736.p1, not unique enough\n",
      "-skipping training candidate: rna28243.p1, not unique enough\n",
      "-skipping training candidate: rna38159.p1, not unique enough\n",
      "-skipping training candidate: rna25902.p1, not unique enough\n",
      "-skipping training candidate: rna26449.p1, not unique enough\n",
      "-skipping training candidate: rna38157.p1, not unique enough\n",
      "-skipping training candidate: rna53215.p1, not unique enough\n",
      "-skipping training candidate: rna58585.p1, not unique enough\n",
      "-skipping training candidate: rna56660.p1, not unique enough\n",
      "-skipping training candidate: rna57874.p1, not unique enough\n",
      "-skipping training candidate: rna2447.p1, not unique enough\n",
      "-skipping training candidate: rna3739.p1, not unique enough\n",
      "-skipping training candidate: rna24955.p1, not unique enough\n",
      "-skipping training candidate: rna33241.p1, not unique enough\n",
      "-skipping training candidate: rna40582.p1, not unique enough\n",
      "-skipping training candidate: rna57873.p1, not unique enough\n",
      "-skipping training candidate: rna22321.p1, not unique enough\n",
      "-skipping training candidate: rna23682.p1, not unique enough\n",
      "-skipping training candidate: rna23686.p1, not unique enough\n",
      "-skipping training candidate: rna23681.p1, not unique enough\n",
      "-skipping training candidate: rna23684.p1, not unique enough\n",
      "-skipping training candidate: rna34494.p1, not unique enough\n",
      "-skipping training candidate: rna40409.p1, not unique enough\n",
      "-skipping training candidate: rna49054.p1, not unique enough\n",
      "-skipping training candidate: rna25002.p1, not unique enough\n",
      "-skipping training candidate: rna25003.p1, not unique enough\n",
      "-skipping training candidate: rna25001.p1, not unique enough\n",
      "-skipping training candidate: rna25000.p1, not unique enough\n",
      "-skipping training candidate: rna24924.p1, not unique enough\n",
      "-skipping training candidate: rna40413.p1, not unique enough\n",
      "-skipping training candidate: rna58721.p1, not unique enough\n",
      "-skipping training candidate: rna10920.p1, not unique enough\n",
      "-skipping training candidate: rna13951.p1, not unique enough\n",
      "-skipping training candidate: rna13953.p1, not unique enough\n",
      "-skipping training candidate: rna19906.p1, not unique enough\n",
      "-skipping training candidate: rna49263.p1, not unique enough\n",
      "-skipping training candidate: rna53216.p1, not unique enough\n",
      "-skipping training candidate: rna13397.p1, not unique enough\n",
      "-skipping training candidate: rna21361.p1, not unique enough\n",
      "-skipping training candidate: rna24775.p1, not unique enough\n",
      "-skipping training candidate: rna24774.p1, not unique enough\n",
      "-skipping training candidate: rna31912.p1, not unique enough\n",
      "-skipping training candidate: rna61204.p1, not unique enough\n",
      "-skipping training candidate: rna4873.p1, not unique enough\n",
      "-skipping training candidate: rna7833.p1, not unique enough\n",
      "-skipping training candidate: rna10541.p1, not unique enough\n",
      "-skipping training candidate: rna13954.p1, not unique enough\n",
      "-skipping training candidate: rna17659.p1, not unique enough\n",
      "-skipping training candidate: rna28530.p1, not unique enough\n",
      "-skipping training candidate: rna32940.p1, not unique enough\n",
      "-skipping training candidate: rna40412.p1, not unique enough\n",
      "-skipping training candidate: rna49052.p1, not unique enough\n",
      "-skipping training candidate: rna3386.p1, not unique enough\n",
      "-skipping training candidate: rna3389.p1, not unique enough\n",
      "-skipping training candidate: rna21508.p1, not unique enough\n",
      "-skipping training candidate: rna29785.p1, not unique enough\n",
      "-skipping training candidate: rna38158.p1, not unique enough\n",
      "-skipping training candidate: rna41824.p1, not unique enough\n",
      "-skipping training candidate: rna45952.p1, not unique enough\n",
      "-skipping training candidate: rna49055.p1, not unique enough\n",
      "-skipping training candidate: rna58664.p1, not unique enough\n",
      "-skipping training candidate: rna20727.p1, not unique enough\n",
      "-skipping training candidate: rna40414.p1, not unique enough\n",
      "-skipping training candidate: rna48305.p1, not unique enough\n",
      "-skipping training candidate: rna58588.p1, not unique enough\n",
      "-skipping training candidate: rna1256.p1, not unique enough\n",
      "-skipping training candidate: rna4084.p1, not unique enough\n",
      "-skipping training candidate: rna51338.p1, not unique enough\n",
      "-skipping training candidate: rna58587.p1, not unique enough\n",
      "-skipping training candidate: rna1027.p1, not unique enough\n",
      "-skipping training candidate: rna5141.p1, not unique enough\n",
      "-skipping training candidate: rna16637.p1, not unique enough\n",
      "-skipping training candidate: rna15456.p1, not unique enough\n",
      "-skipping training candidate: rna15451.p1, not unique enough\n",
      "-skipping training candidate: rna30692.p1, not unique enough\n",
      "-skipping training candidate: rna3388.p1, not unique enough\n",
      "-skipping training candidate: rna32393.p1, not unique enough\n",
      "-skipping training candidate: rna38918.p1, not unique enough\n",
      "-skipping training candidate: rna42964.p1, not unique enough\n",
      "-skipping training candidate: rna49057.p1, not unique enough\n",
      "-skipping training candidate: rna49053.p1, not unique enough\n",
      "-skipping training candidate: rna14279.p1, not unique enough\n",
      "-skipping training candidate: rna48906.p1, not unique enough\n",
      "-skipping training candidate: rna1858.p1, not unique enough\n",
      "-skipping training candidate: rna21509.p1, not unique enough\n",
      "-skipping training candidate: rna21133.p1, not unique enough\n",
      "-skipping training candidate: rna45614.p1, not unique enough\n",
      "-skipping training candidate: rna5333.p1, not unique enough\n",
      "-skipping training candidate: rna20484.p1, not unique enough\n",
      "-skipping training candidate: rna24104.p1, not unique enough\n",
      "-skipping training candidate: rna24106.p1, not unique enough\n",
      "-skipping training candidate: rna50019.p1, not unique enough\n",
      "-skipping training candidate: rna7826.p1, not unique enough\n",
      "-skipping training candidate: rna21453.p1, not unique enough\n",
      "-skipping training candidate: rna21454.p1, not unique enough\n",
      "-skipping training candidate: rna19998.p1, not unique enough\n",
      "-skipping training candidate: rna24108.p1, not unique enough\n",
      "-skipping training candidate: rna26733.p1, not unique enough\n",
      "-skipping training candidate: rna38811.p1, not unique enough\n",
      "-skipping training candidate: rna95.p1, not unique enough\n",
      "-skipping training candidate: rna3385.p1, not unique enough\n",
      "-skipping training candidate: rna14292.p1, not unique enough\n",
      "-skipping training candidate: rna4912.p1, not unique enough\n",
      "-skipping training candidate: rna10909.p1, not unique enough\n",
      "-skipping training candidate: rna23392.p1, not unique enough\n",
      "-skipping training candidate: rna35790.p1, not unique enough\n",
      "-skipping training candidate: rna41825.p1, not unique enough\n",
      "-skipping training candidate: rna44273.p1, not unique enough\n",
      "-skipping training candidate: rna44274.p1, not unique enough\n",
      "-skipping training candidate: rna1857.p1, not unique enough\n",
      "-skipping training candidate: rna2109.p1, not unique enough\n",
      "-skipping training candidate: rna6072.p1, not unique enough\n",
      "-skipping training candidate: rna8921.p1, not unique enough\n",
      "-skipping training candidate: rna12582.p1, not unique enough\n",
      "-skipping training candidate: rna34580.p1, not unique enough\n",
      "-skipping training candidate: rna38127.p1, not unique enough\n",
      "-skipping training candidate: rna42684.p1, not unique enough\n",
      "-skipping training candidate: rna6154.p1, not unique enough\n",
      "-skipping training candidate: rna9848.p1, not unique enough\n",
      "-skipping training candidate: rna14188.p1, not unique enough\n",
      "-skipping training candidate: rna25361.p1, not unique enough\n",
      "-skipping training candidate: rna25356.p1, not unique enough\n",
      "-skipping training candidate: rna25359.p1, not unique enough\n",
      "-skipping training candidate: rna25360.p1, not unique enough\n",
      "-skipping training candidate: rna25355.p1, not unique enough\n",
      "-skipping training candidate: rna25357.p1, not unique enough\n",
      "-skipping training candidate: rna45079.p1, not unique enough\n",
      "-skipping training candidate: rna49407.p1, not unique enough\n",
      "-skipping training candidate: rna14189.p1, not unique enough\n",
      "-skipping training candidate: rna20017.p1, not unique enough\n",
      "-skipping training candidate: rna22566.p1, not unique enough\n",
      "-skipping training candidate: rna26694.p1, not unique enough\n",
      "-skipping training candidate: rna31145.p1, not unique enough\n",
      "-skipping training candidate: rna44182.p1, not unique enough\n",
      "-skipping training candidate: rna9668.p1, not unique enough\n",
      "-skipping training candidate: rna27014.p1, not unique enough\n",
      "-skipping training candidate: rna27012.p1, not unique enough\n",
      "-skipping training candidate: rna27010.p1, not unique enough\n",
      "-skipping training candidate: rna27009.p1, not unique enough\n",
      "-skipping training candidate: rna27758.p1, not unique enough\n",
      "-skipping training candidate: rna39855.p1, not unique enough\n",
      "-skipping training candidate: rna4033.p1, not unique enough\n",
      "-skipping training candidate: rna19027.p1, not unique enough\n",
      "-skipping training candidate: rna19026.p1, not unique enough\n",
      "-skipping training candidate: rna19030.p1, not unique enough\n",
      "-skipping training candidate: rna56410.p1, not unique enough\n",
      "-skipping training candidate: rna94.p1, not unique enough\n",
      "-skipping training candidate: rna16272.p1, not unique enough\n",
      "-skipping training candidate: rna22167.p1, not unique enough\n",
      "-skipping training candidate: rna23433.p1, not unique enough\n",
      "-skipping training candidate: rna35293.p1, not unique enough\n",
      "-skipping training candidate: rna40207.p1, not unique enough\n",
      "-skipping training candidate: rna61156.p1, not unique enough\n",
      "-skipping training candidate: rna10017.p1, not unique enough\n",
      "-skipping training candidate: rna10015.p1, not unique enough\n",
      "-skipping training candidate: rna12677.p1, not unique enough\n",
      "-skipping training candidate: rna23657.p1, not unique enough\n",
      "-skipping training candidate: rna40949.p1, not unique enough\n",
      "-skipping training candidate: rna5259.p1, not unique enough\n",
      "-skipping training candidate: rna10016.p1, not unique enough\n",
      "-skipping training candidate: rna13923.p1, not unique enough\n",
      "-skipping training candidate: rna16699.p1, not unique enough\n",
      "-skipping training candidate: rna16698.p1, not unique enough\n",
      "-skipping training candidate: rna16700.p1, not unique enough\n",
      "-skipping training candidate: rna16697.p1, not unique enough\n",
      "-skipping training candidate: rna21709.p1, not unique enough\n",
      "-skipping training candidate: rna30695.p1, not unique enough\n",
      "-skipping training candidate: rna52797.p1, not unique enough\n",
      "-skipping training candidate: rna1520.p1, not unique enough\n",
      "-skipping training candidate: rna4037.p1, not unique enough\n",
      "-skipping training candidate: rna7573.p1, not unique enough\n",
      "-skipping training candidate: rna22568.p1, not unique enough\n",
      "-skipping training candidate: rna30693.p1, not unique enough\n",
      "-skipping training candidate: rna7126.p1, not unique enough\n",
      "-skipping training candidate: rna28518.p1, not unique enough\n",
      "-skipping training candidate: rna27757.p1, not unique enough\n",
      "-skipping training candidate: rna41528.p1, not unique enough\n",
      "-skipping training candidate: rna41529.p1, not unique enough\n",
      "-skipping training candidate: rna41527.p1, not unique enough\n",
      "-skipping training candidate: rna46556.p1, not unique enough\n",
      "-skipping training candidate: rna52798.p1, not unique enough\n",
      "-skipping training candidate: rna10005.p1, not unique enough\n",
      "-skipping training candidate: rna21583.p1, not unique enough\n",
      "-skipping training candidate: rna27756.p1, not unique enough\n",
      "-skipping training candidate: rna53180.p1, not unique enough\n",
      "-skipping training candidate: rna53181.p1, not unique enough\n",
      "-skipping training candidate: rna56514.p1, not unique enough\n",
      "-skipping training candidate: rna950.p1, not unique enough\n",
      "-skipping training candidate: rna7131.p1, not unique enough\n",
      "-skipping training candidate: rna20429.p1, not unique enough\n",
      "-skipping training candidate: rna24297.p1, not unique enough\n",
      "-skipping training candidate: rna24295.p1, not unique enough\n",
      "-skipping training candidate: rna25747.p1, not unique enough\n",
      "-skipping training candidate: rna23760.p1, not unique enough\n",
      "-skipping training candidate: rna32654.p1, not unique enough\n",
      "-skipping training candidate: rna51341.p1, not unique enough\n",
      "-skipping training candidate: rna56513.p1, not unique enough\n",
      "-skipping training candidate: rna3564.p1, not unique enough\n",
      "-skipping training candidate: rna3559.p1, not unique enough\n",
      "-skipping training candidate: rna9105.p1, not unique enough\n",
      "-skipping training candidate: rna10784.p1, not unique enough\n",
      "-skipping training candidate: rna10785.p1, not unique enough\n",
      "-skipping training candidate: rna26693.p1, not unique enough\n",
      "-skipping training candidate: rna29936.p1, not unique enough\n",
      "-skipping training candidate: rna32242.p1, not unique enough\n",
      "-skipping training candidate: rna35607.p1, not unique enough\n",
      "-skipping training candidate: rna35609.p1, not unique enough\n",
      "-skipping training candidate: rna10591.p1, not unique enough\n",
      "-skipping training candidate: rna34317.p1, not unique enough\n",
      "-skipping training candidate: rna47508.p1, not unique enough\n",
      "-skipping training candidate: rna50231.p1, not unique enough\n",
      "-skipping training candidate: rna3558.p1, not unique enough\n",
      "-skipping training candidate: rna6555.p1, not unique enough\n",
      "-skipping training candidate: rna14137.p1, not unique enough\n",
      "-skipping training candidate: rna20428.p1, not unique enough\n",
      "-skipping training candidate: rna23436.p1, not unique enough\n",
      "-skipping training candidate: rna23762.p1, not unique enough\n",
      "-skipping training candidate: rna23761.p1, not unique enough\n",
      "-skipping training candidate: rna32243.p1, not unique enough\n",
      "-skipping training candidate: rna53183.p1, not unique enough\n",
      "-skipping training candidate: rna53106.p1, not unique enough\n",
      "-skipping training candidate: rna27231.p1, not unique enough\n",
      "-skipping training candidate: rna27229.p1, not unique enough\n",
      "-skipping training candidate: rna44614.p1, not unique enough\n",
      "-skipping training candidate: rna56324.p1, not unique enough\n",
      "-skipping training candidate: rna58748.p1, not unique enough\n",
      "-skipping training candidate: rna7128.p1, not unique enough\n",
      "-skipping training candidate: rna6474.p1, not unique enough\n",
      "-skipping training candidate: rna8821.p1, not unique enough\n",
      "-skipping training candidate: rna27233.p1, not unique enough\n",
      "-skipping training candidate: rna27235.p1, not unique enough\n",
      "-skipping training candidate: rna30690.p1, not unique enough\n",
      "-skipping training candidate: rna1966.p1, not unique enough\n",
      "-skipping training candidate: rna1524.p1, not unique enough\n",
      "-skipping training candidate: rna3560.p1, not unique enough\n",
      "-skipping training candidate: rna36025.p1, not unique enough\n",
      "-skipping training candidate: rna35188.p1, not unique enough\n",
      "-skipping training candidate: rna44924.p1, not unique enough\n",
      "-skipping training candidate: rna56327.p1, not unique enough\n",
      "-skipping training candidate: rna56515.p1, not unique enough\n",
      "-skipping training candidate: rna12759.p1, not unique enough\n",
      "-skipping training candidate: rna12758.p1, not unique enough\n",
      "-skipping training candidate: rna12583.p1, not unique enough\n",
      "-skipping training candidate: rna22165.p1, not unique enough\n",
      "-skipping training candidate: rna23438.p1, not unique enough\n",
      "-skipping training candidate: rna58747.p1, not unique enough\n",
      "-skipping training candidate: rna58750.p1, not unique enough\n",
      "-skipping training candidate: rna58751.p1, not unique enough\n",
      "-skipping training candidate: rna3555.p1, not unique enough\n",
      "-skipping training candidate: rna9103.p1, not unique enough\n",
      "-skipping training candidate: rna10590.p1, not unique enough\n",
      "-skipping training candidate: rna12757.p1, not unique enough\n",
      "-skipping training candidate: rna12762.p1, not unique enough\n",
      "-skipping training candidate: rna41116.p1, not unique enough\n",
      "-skipping training candidate: rna22569.p1, not unique enough\n",
      "-skipping training candidate: rna23688.p1, not unique enough\n",
      "-skipping training candidate: rna27146.p1, not unique enough\n",
      "-skipping training candidate: rna38059.p1, not unique enough\n",
      "-skipping training candidate: rna56050.p1, not unique enough\n",
      "-skipping training candidate: rna1521.p1, not unique enough\n",
      "-skipping training candidate: rna7127.p1, not unique enough\n",
      "-skipping training candidate: rna14127.p1, not unique enough\n",
      "-skipping training candidate: rna23658.p1, not unique enough\n",
      "-skipping training candidate: rna32396.p1, not unique enough\n",
      "-skipping training candidate: rna33239.p1, not unique enough\n",
      "-skipping training candidate: rna58192.p1, not unique enough\n",
      "-skipping training candidate: rna21586.p1, not unique enough\n",
      "-skipping training candidate: rna29787.p1, not unique enough\n",
      "-skipping training candidate: rna35791.p1, not unique enough\n",
      "-skipping training candidate: rna38061.p1, not unique enough\n",
      "-skipping training candidate: rna46652.p1, not unique enough\n",
      "-skipping training candidate: rna56323.p1, not unique enough\n",
      "-skipping training candidate: rna21798.p1, not unique enough\n",
      "-skipping training candidate: rna21796.p1, not unique enough\n",
      "-skipping training candidate: rna21799.p1, not unique enough\n",
      "-skipping training candidate: rna22324.p1, not unique enough\n",
      "-skipping training candidate: rna22664.p1, not unique enough\n",
      "-skipping training candidate: rna26732.p1, not unique enough\n",
      "-skipping training candidate: rna28245.p1, not unique enough\n",
      "-skipping training candidate: rna29784.p1, not unique enough\n",
      "-skipping training candidate: rna46554.p1, not unique enough\n",
      "-skipping training candidate: rna44610.p1, not unique enough\n",
      "-skipping training candidate: rna51343.p1, not unique enough\n",
      "-skipping training candidate: rna58920.p1, not unique enough\n",
      "-skipping training candidate: rna10006.p1, not unique enough\n",
      "-skipping training candidate: rna1523.p1, not unique enough\n",
      "-skipping training candidate: rna14190.p1, not unique enough\n",
      "-skipping training candidate: rna22325.p1, not unique enough\n",
      "-skipping training candidate: rna26734.p1, not unique enough\n",
      "-skipping training candidate: rna37443.p1, not unique enough\n",
      "-skipping training candidate: rna7129.p1, not unique enough\n",
      "-skipping training candidate: rna26996.p1, not unique enough\n",
      "-skipping training candidate: rna26997.p1, not unique enough\n",
      "-skipping training candidate: rna29932.p1, not unique enough\n",
      "-skipping training candidate: rna4036.p1, not unique enough\n",
      "-skipping training candidate: rna29931.p1, not unique enough\n",
      "-skipping training candidate: rna29929.p1, not unique enough\n",
      "-skipping training candidate: rna30531.p1, not unique enough\n",
      "-skipping training candidate: rna32656.p1, not unique enough\n",
      "-skipping training candidate: rna34915.p1, not unique enough\n",
      "-skipping training candidate: rna57142.p1, not unique enough\n",
      "-skipping training candidate: rna6812.p1, not unique enough\n",
      "-skipping training candidate: rna6813.p1, not unique enough\n",
      "-skipping training candidate: rna29934.p1, not unique enough\n",
      "-skipping training candidate: rna29935.p1, not unique enough\n",
      "-skipping training candidate: rna29448.p1, not unique enough\n",
      "-skipping training candidate: rna42051.p1, not unique enough\n",
      "-skipping training candidate: rna57104.p1, not unique enough\n",
      "-skipping training candidate: rna59858.p1, not unique enough\n",
      "-skipping training candidate: rna60273.p1, not unique enough\n",
      "-skipping training candidate: rna22570.p1, not unique enough\n",
      "-skipping training candidate: rna29937.p1, not unique enough\n",
      "-skipping training candidate: rna36026.p1, not unique enough\n",
      "-skipping training candidate: rna57868.p1, not unique enough\n",
      "-skipping training candidate: rna2692.p1, not unique enough\n",
      "-skipping training candidate: rna26691.p1, not unique enough\n",
      "-skipping training candidate: rna26690.p1, not unique enough\n",
      "-skipping training candidate: rna53747.p1, not unique enough\n",
      "-skipping training candidate: rna1522.p1, not unique enough\n",
      "-skipping training candidate: rna4541.p1, not unique enough\n",
      "-skipping training candidate: rna6574.p1, not unique enough\n",
      "-skipping training candidate: rna6575.p1, not unique enough\n",
      "-skipping training candidate: rna8802.p1, not unique enough\n",
      "-skipping training candidate: rna14192.p1, not unique enough\n",
      "-skipping training candidate: rna21710.p1, not unique enough\n",
      "-skipping training candidate: rna24970.p1, not unique enough\n",
      "-skipping training candidate: rna31864.p1, not unique enough\n",
      "-skipping training candidate: rna38202.p1, not unique enough\n",
      "-skipping training candidate: rna38750.p1, not unique enough\n",
      "-skipping training candidate: rna52800.p1, not unique enough\n",
      "-skipping training candidate: rna7135.p1, not unique enough\n",
      "-skipping training candidate: rna26495.p1, not unique enough\n",
      "-skipping training candidate: rna32252.p1, not unique enough\n",
      "-skipping training candidate: rna32251.p1, not unique enough\n",
      "-skipping training candidate: rna32247.p1, not unique enough\n",
      "-skipping training candidate: rna32244.p1, not unique enough\n",
      "-skipping training candidate: rna32246.p1, not unique enough\n",
      "-skipping training candidate: rna32245.p1, not unique enough\n",
      "-skipping training candidate: rna33289.p1, not unique enough\n",
      "-skipping training candidate: rna33237.p1, not unique enough\n",
      "-skipping training candidate: rna34783.p1, not unique enough\n",
      "-skipping training candidate: rna45217.p1, not unique enough\n",
      "-skipping training candidate: rna45458.p1, not unique enough\n",
      "-skipping training candidate: rna45457.p1, not unique enough\n",
      "-skipping training candidate: rna49538.p1, not unique enough\n",
      "-skipping training candidate: rna7130.p1, not unique enough\n",
      "-skipping training candidate: rna8347.p1, not unique enough\n",
      "-skipping training candidate: rna8348.p1, not unique enough\n",
      "-skipping training candidate: rna15808.p1, not unique enough\n",
      "-skipping training candidate: rna15807.p1, not unique enough\n",
      "-skipping training candidate: rna22166.p1, not unique enough\n",
      "-skipping training candidate: rna22168.p1, not unique enough\n",
      "-skipping training candidate: rna30758.p1, not unique enough\n",
      "-skipping training candidate: rna44613.p1, not unique enough\n",
      "-skipping training candidate: rna52799.p1, not unique enough\n",
      "-skipping training candidate: rna22323.p1, not unique enough\n",
      "-skipping training candidate: rna27698.p1, not unique enough\n",
      "-skipping training candidate: rna53603.p1, not unique enough\n",
      "-skipping training candidate: rna53601.p1, not unique enough\n",
      "-skipping training candidate: rna61205.p1, not unique enough\n",
      "-skipping training candidate: rna2179.p1, not unique enough\n",
      "-skipping training candidate: rna3563.p1, not unique enough\n",
      "-skipping training candidate: rna14664.p1, not unique enough\n",
      "-skipping training candidate: rna14662.p1, not unique enough\n",
      "-skipping training candidate: rna14665.p1, not unique enough\n",
      "-skipping training candidate: rna14460.p1, not unique enough\n",
      "-skipping training candidate: rna23437.p1, not unique enough\n",
      "-skipping training candidate: rna36029.p1, not unique enough\n",
      "-skipping training candidate: rna49649.p1, not unique enough\n",
      "-skipping training candidate: rna53104.p1, not unique enough\n",
      "-skipping training candidate: rna3557.p1, not unique enough\n",
      "-skipping training candidate: rna5528.p1, not unique enough\n",
      "-skipping training candidate: rna10599.p1, not unique enough\n",
      "-skipping training candidate: rna21784.p1, not unique enough\n",
      "-skipping training candidate: rna26496.p1, not unique enough\n",
      "-skipping training candidate: rna27695.p1, not unique enough\n",
      "-skipping training candidate: rna35610.p1, not unique enough\n",
      "-skipping training candidate: rna39554.p1, not unique enough\n",
      "-skipping training candidate: rna58921.p1, not unique enough\n",
      "-skipping training candidate: rna7134.p1, not unique enough\n",
      "-skipping training candidate: rna6811.p1, not unique enough\n",
      "-skipping training candidate: rna11407.p1, not unique enough\n",
      "-skipping training candidate: rna8820.p1, not unique enough\n",
      "-skipping training candidate: rna17765.p1, not unique enough\n",
      "-skipping training candidate: rna17763.p1, not unique enough\n",
      "-skipping training candidate: rna17762.p1, not unique enough\n",
      "-skipping training candidate: rna45218.p1, not unique enough\n",
      "-skipping training candidate: rna49539.p1, not unique enough\n",
      "-skipping training candidate: rna59962.p1, not unique enough\n",
      "-skipping training candidate: rna60475.p1, not unique enough\n",
      "-skipping training candidate: rna30615.p1, not unique enough\n",
      "-skipping training candidate: rna55756.p1, not unique enough\n",
      "-skipping training candidate: rna7133.p1, not unique enough\n",
      "-skipping training candidate: rna12314.p1, not unique enough\n",
      "-skipping training candidate: rna24105.p1, not unique enough\n",
      "-skipping training candidate: rna24107.p1, not unique enough\n",
      "-skipping training candidate: rna26457.p1, not unique enough\n",
      "-skipping training candidate: rna33243.p1, not unique enough\n",
      "-skipping training candidate: rna33238.p1, not unique enough\n",
      "-skipping training candidate: rna36028.p1, not unique enough\n",
      "-skipping training candidate: rna56325.p1, not unique enough\n",
      "-skipping training candidate: rna3561.p1, not unique enough\n",
      "-skipping training candidate: rna5939.p1, not unique enough\n",
      "-skipping training candidate: rna8235.p1, not unique enough\n",
      "-skipping training candidate: rna10645.p1, not unique enough\n",
      "-skipping training candidate: rna10646.p1, not unique enough\n",
      "-skipping training candidate: rna12312.p1, not unique enough\n",
      "-skipping training candidate: rna12987.p1, not unique enough\n",
      "-skipping training candidate: rna23739.p1, not unique enough\n",
      "-skipping training candidate: rna24109.p1, not unique enough\n",
      "-skipping training candidate: rna58586.p1, not unique enough\n",
      "-skipping training candidate: rna3556.p1, not unique enough\n",
      "-skipping training candidate: rna7952.p1, not unique enough\n",
      "-skipping training candidate: rna8234.p1, not unique enough\n",
      "-skipping training candidate: rna30161.p1, not unique enough\n",
      "-skipping training candidate: rna56328.p1, not unique enough\n",
      "-skipping training candidate: rna59960.p1, not unique enough\n",
      "-skipping training candidate: rna3562.p1, not unique enough\n",
      "-skipping training candidate: rna12311.p1, not unique enough\n",
      "-skipping training candidate: rna29296.p1, not unique enough\n",
      "-skipping training candidate: rna32035.p1, not unique enough\n",
      "-skipping training candidate: rna32036.p1, not unique enough\n",
      "-skipping training candidate: rna32033.p1, not unique enough\n",
      "-skipping training candidate: rna48981.p1, not unique enough\n",
      "-skipping training candidate: rna51937.p1, not unique enough\n",
      "-skipping training candidate: rna1526.p1, not unique enough\n",
      "-skipping training candidate: rna12316.p1, not unique enough\n",
      "-skipping training candidate: rna16705.p1, not unique enough\n",
      "-skipping training candidate: rna29938.p1, not unique enough\n",
      "-skipping training candidate: rna29295.p1, not unique enough\n",
      "-skipping training candidate: rna29294.p1, not unique enough\n",
      "-skipping training candidate: rna28926.p1, not unique enough\n",
      "-skipping training candidate: rna28925.p1, not unique enough\n",
      "-skipping training candidate: rna31816.p1, not unique enough\n",
      "-skipping training candidate: rna31818.p1, not unique enough\n",
      "-skipping training candidate: rna39020.p1, not unique enough\n",
      "-skipping training candidate: rna52453.p1, not unique enough\n",
      "-skipping training candidate: rna52454.p1, not unique enough\n",
      "-skipping training candidate: rna57141.p1, not unique enough\n",
      "-skipping training candidate: rna7950.p1, not unique enough\n",
      "-skipping training candidate: rna10601.p1, not unique enough\n",
      "-skipping training candidate: rna10589.p1, not unique enough\n",
      "-skipping training candidate: rna8818.p1, not unique enough\n",
      "-skipping training candidate: rna12315.p1, not unique enough\n",
      "-skipping training candidate: rna13993.p1, not unique enough\n",
      "-skipping training candidate: rna31817.p1, not unique enough\n",
      "-skipping training candidate: rna61155.p1, not unique enough\n",
      "-skipping training candidate: rna4035.p1, not unique enough\n",
      "-skipping training candidate: rna7949.p1, not unique enough\n",
      "-skipping training candidate: rna7956.p1, not unique enough\n",
      "-skipping training candidate: rna8038.p1, not unique enough\n",
      "-skipping training candidate: rna20304.p1, not unique enough\n",
      "-skipping training candidate: rna16017.p1, not unique enough\n",
      "-skipping training candidate: rna31069.p1, not unique enough\n",
      "-skipping training candidate: rna36671.p1, not unique enough\n",
      "-skipping training candidate: rna44612.p1, not unique enough\n",
      "-skipping training candidate: rna8808.p1, not unique enough\n",
      "-skipping training candidate: rna8807.p1, not unique enough\n",
      "-skipping training candidate: rna8813.p1, not unique enough\n",
      "-skipping training candidate: rna8811.p1, not unique enough\n",
      "-skipping training candidate: rna8809.p1, not unique enough\n",
      "-skipping training candidate: rna8810.p1, not unique enough\n",
      "-skipping training candidate: rna36031.p1, not unique enough\n",
      "-skipping training candidate: rna38749.p1, not unique enough\n",
      "-skipping training candidate: rna53936.p1, not unique enough\n",
      "-skipping training candidate: rna4040.p1, not unique enough\n",
      "-skipping training candidate: rna7953.p1, not unique enough\n",
      "-skipping training candidate: rna9196.p1, not unique enough\n",
      "-skipping training candidate: rna9670.p1, not unique enough\n",
      "-skipping training candidate: rna20430.p1, not unique enough\n",
      "-skipping training candidate: rna27760.p1, not unique enough\n",
      "-skipping training candidate: rna46464.p1, not unique enough\n",
      "-skipping training candidate: rna44400.p1, not unique enough\n",
      "-skipping training candidate: rna54246.p1, not unique enough\n",
      "-skipping training candidate: rna7954.p1, not unique enough\n",
      "-skipping training candidate: rna12476.p1, not unique enough\n",
      "-skipping training candidate: rna12477.p1, not unique enough\n",
      "-skipping training candidate: rna20792.p1, not unique enough\n",
      "-skipping training candidate: rna20793.p1, not unique enough\n",
      "-skipping training candidate: rna20791.p1, not unique enough\n",
      "-skipping training candidate: rna39859.p1, not unique enough\n",
      "-skipping training candidate: rna39860.p1, not unique enough\n",
      "-skipping training candidate: rna52864.p1, not unique enough\n",
      "-skipping training candidate: rna1527.p1, not unique enough\n",
      "-skipping training candidate: rna15165.p1, not unique enough\n",
      "-skipping training candidate: rna12479.p1, not unique enough\n",
      "-skipping training candidate: rna52137.p1, not unique enough\n",
      "-skipping training candidate: rna467.p1, not unique enough\n",
      "-skipping training candidate: rna4542.p1, not unique enough\n",
      "-skipping training candidate: rna6472.p1, not unique enough\n",
      "-skipping training candidate: rna10593.p1, not unique enough\n",
      "-skipping training candidate: rna8812.p1, not unique enough\n",
      "-skipping training candidate: rna8806.p1, not unique enough\n",
      "-skipping training candidate: rna24951.p1, not unique enough\n",
      "-skipping training candidate: rna24944.p1, not unique enough\n",
      "-skipping training candidate: rna24947.p1, not unique enough\n",
      "-skipping training candidate: rna28262.p1, not unique enough\n",
      "-skipping training candidate: rna27618.p1, not unique enough\n",
      "-skipping training candidate: rna28689.p1, not unique enough\n",
      "-skipping training candidate: rna31068.p1, not unique enough\n",
      "-skipping training candidate: rna32248.p1, not unique enough\n",
      "-skipping training candidate: rna51342.p1, not unique enough\n",
      "-skipping training candidate: rna54244.p1, not unique enough\n",
      "-skipping training candidate: rna9195.p1, not unique enough\n",
      "-skipping training candidate: rna11481.p1, not unique enough\n",
      "-skipping training candidate: rna13356.p1, not unique enough\n",
      "-skipping training candidate: rna17571.p1, not unique enough\n",
      "-skipping training candidate: rna17572.p1, not unique enough\n",
      "-skipping training candidate: rna16702.p1, not unique enough\n",
      "-skipping training candidate: rna16704.p1, not unique enough\n",
      "-skipping training candidate: rna21451.p1, not unique enough\n",
      "-skipping training candidate: rna24950.p1, not unique enough\n",
      "-skipping training candidate: rna24948.p1, not unique enough\n",
      "-skipping training candidate: rna30694.p1, not unique enough\n",
      "-skipping training candidate: rna29828.p1, not unique enough\n",
      "-skipping training candidate: rna36670.p1, not unique enough\n",
      "-skipping training candidate: rna48112.p1, not unique enough\n",
      "-skipping training candidate: rna3297.p1, not unique enough\n",
      "-skipping training candidate: rna1532.p1, not unique enough\n",
      "-skipping training candidate: rna1432.p1, not unique enough\n",
      "-skipping training candidate: rna7957.p1, not unique enough\n",
      "-skipping training candidate: rna7955.p1, not unique enough\n",
      "-skipping training candidate: rna6995.p1, not unique enough\n",
      "-skipping training candidate: rna9194.p1, not unique enough\n",
      "-skipping training candidate: rna11482.p1, not unique enough\n",
      "-skipping training candidate: rna27700.p1, not unique enough\n",
      "-skipping training candidate: rna35891.p1, not unique enough\n",
      "-skipping training candidate: rna35893.p1, not unique enough\n",
      "-skipping training candidate: rna35892.p1, not unique enough\n",
      "-skipping training candidate: rna3676.p1, not unique enough\n",
      "-skipping training candidate: rna7820.p1, not unique enough\n",
      "-skipping training candidate: rna5260.p1, not unique enough\n",
      "-skipping training candidate: rna7958.p1, not unique enough\n",
      "-skipping training candidate: rna10597.p1, not unique enough\n",
      "-skipping training candidate: rna21585.p1, not unique enough\n",
      "-skipping training candidate: rna26497.p1, not unique enough\n",
      "-skipping training candidate: rna27616.p1, not unique enough\n",
      "-skipping training candidate: rna9669.p1, not unique enough\n",
      "-skipping training candidate: rna27615.p1, not unique enough\n",
      "-skipping training candidate: rna1750.p1, not unique enough\n",
      "-skipping training candidate: rna1748.p1, not unique enough\n",
      "-skipping training candidate: rna1749.p1, not unique enough\n",
      "-skipping training candidate: rna10595.p1, not unique enough\n",
      "-skipping training candidate: rna14193.p1, not unique enough\n",
      "-skipping training candidate: rna14194.p1, not unique enough\n",
      "-skipping training candidate: rna46039.p1, not unique enough\n",
      "-skipping training candidate: rna1746.p1, not unique enough\n",
      "-skipping training candidate: rna9108.p1, not unique enough\n",
      "-skipping training candidate: rna12070.p1, not unique enough\n",
      "-skipping training candidate: rna33712.p1, not unique enough\n",
      "-skipping training candidate: rna36030.p1, not unique enough\n",
      "-skipping training candidate: rna4717.p1, not unique enough\n",
      "-skipping training candidate: rna4716.p1, not unique enough\n",
      "-skipping training candidate: rna4715.p1, not unique enough\n",
      "-skipping training candidate: rna9107.p1, not unique enough\n",
      "-skipping training candidate: rna11492.p1, not unique enough\n",
      "-skipping training candidate: rna11503.p1, not unique enough\n",
      "-skipping training candidate: rna18781.p1, not unique enough\n",
      "-skipping training candidate: rna27232.p1, not unique enough\n",
      "-skipping training candidate: rna27230.p1, not unique enough\n",
      "-skipping training candidate: rna29930.p1, not unique enough\n",
      "-skipping training candidate: rna31146.p1, not unique enough\n",
      "-skipping training candidate: rna48700.p1, not unique enough\n",
      "-skipping training candidate: rna471.p1, not unique enough\n",
      "-skipping training candidate: rna4543.p1, not unique enough\n",
      "-skipping training candidate: rna7959.p1, not unique enough\n",
      "-skipping training candidate: rna11486.p1, not unique enough\n",
      "-skipping training candidate: rna11495.p1, not unique enough\n",
      "-skipping training candidate: rna11484.p1, not unique enough\n",
      "-skipping training candidate: rna11489.p1, not unique enough\n",
      "-skipping training candidate: rna11498.p1, not unique enough\n",
      "-skipping training candidate: rna11485.p1, not unique enough\n",
      "-skipping training candidate: rna27234.p1, not unique enough\n",
      "-skipping training candidate: rna30617.p1, not unique enough\n",
      "-skipping training candidate: rna41783.p1, not unique enough\n",
      "-skipping training candidate: rna58922.p1, not unique enough\n",
      "-skipping training candidate: rna16555.p1, not unique enough\n",
      "-skipping training candidate: rna21587.p1, not unique enough\n",
      "-skipping training candidate: rna25937.p1, not unique enough\n",
      "-skipping training candidate: rna25938.p1, not unique enough\n",
      "-skipping training candidate: rna25936.p1, not unique enough\n",
      "-skipping training candidate: rna26584.p1, not unique enough\n",
      "-skipping training candidate: rna26585.p1, not unique enough\n",
      "-skipping training candidate: rna33513.p1, not unique enough\n",
      "-skipping training candidate: rna32945.p1, not unique enough\n",
      "-skipping training candidate: rna51290.p1, not unique enough\n",
      "-skipping training candidate: rna465.p1, not unique enough\n",
      "-skipping training candidate: rna8816.p1, not unique enough\n",
      "-skipping training candidate: rna17885.p1, not unique enough\n",
      "-skipping training candidate: rna20699.p1, not unique enough\n",
      "-skipping training candidate: rna28246.p1, not unique enough\n",
      "-skipping training candidate: rna30688.p1, not unique enough\n",
      "-skipping training candidate: rna32074.p1, not unique enough\n",
      "-skipping training candidate: rna57872.p1, not unique enough\n",
      "-skipping training candidate: rna61073.p1, not unique enough\n",
      "-skipping training candidate: rna13597.p1, not unique enough\n",
      "-skipping training candidate: rna32494.p1, not unique enough\n",
      "-skipping training candidate: rna51289.p1, not unique enough\n",
      "-skipping training candidate: rna5868.p1, not unique enough\n",
      "-skipping training candidate: rna10600.p1, not unique enough\n",
      "-skipping training candidate: rna15164.p1, not unique enough\n",
      "-skipping training candidate: rna12760.p1, not unique enough\n",
      "-skipping training candidate: rna29657.p1, not unique enough\n",
      "-skipping training candidate: rna38810.p1, not unique enough\n",
      "-skipping training candidate: rna60277.p1, not unique enough\n",
      "-skipping training candidate: rna6428.p1, not unique enough\n",
      "-skipping training candidate: rna9102.p1, not unique enough\n",
      "-skipping training candidate: rna12763.p1, not unique enough\n",
      "-skipping training candidate: rna12764.p1, not unique enough\n",
      "-skipping training candidate: rna26675.p1, not unique enough\n",
      "-skipping training candidate: rna33807.p1, not unique enough\n",
      "-skipping training candidate: rna33808.p1, not unique enough\n",
      "-skipping training candidate: rna42954.p1, not unique enough\n",
      "-skipping training candidate: rna2690.p1, not unique enough\n",
      "-skipping training candidate: rna1096.p1, not unique enough\n",
      "-skipping training candidate: rna9106.p1, not unique enough\n",
      "-skipping training candidate: rna12761.p1, not unique enough\n",
      "-skipping training candidate: rna12310.p1, not unique enough\n",
      "-skipping training candidate: rna27843.p1, not unique enough\n",
      "-skipping training candidate: rna51294.p1, not unique enough\n",
      "-skipping training candidate: rna53911.p1, not unique enough\n",
      "-skipping training candidate: rna53912.p1, not unique enough\n",
      "-skipping training candidate: rna53913.p1, not unique enough\n",
      "-skipping training candidate: rna58776.p1, not unique enough\n",
      "-skipping training candidate: rna58774.p1, not unique enough\n",
      "-skipping training candidate: rna4034.p1, not unique enough\n",
      "-skipping training candidate: rna8233.p1, not unique enough\n",
      "-skipping training candidate: rna48982.p1, not unique enough\n",
      "-skipping training candidate: rna50363.p1, not unique enough\n",
      "-skipping training candidate: rna56587.p1, not unique enough\n",
      "-skipping training candidate: rna1967.p1, not unique enough\n",
      "-skipping training candidate: rna7960.p1, not unique enough\n",
      "-skipping training candidate: rna28247.p1, not unique enough\n",
      "-skipping training candidate: rna40251.p1, not unique enough\n",
      "-skipping training candidate: rna52167.p1, not unique enough\n",
      "-skipping training candidate: rna52801.p1, not unique enough\n",
      "-skipping training candidate: rna5935.p1, not unique enough\n",
      "-skipping training candidate: rna27682.p1, not unique enough\n",
      "-skipping training candidate: rna12990.p1, not unique enough\n",
      "-skipping training candidate: rna13756.p1, not unique enough\n",
      "-skipping training candidate: rna36032.p1, not unique enough\n",
      "-skipping training candidate: rna36027.p1, not unique enough\n",
      "-skipping training candidate: rna16689.p1, not unique enough\n",
      "-skipping training candidate: rna21839.p1, not unique enough\n",
      "-skipping training candidate: rna44401.p1, not unique enough\n",
      "-skipping training candidate: rna14666.p1, not unique enough\n",
      "-skipping training candidate: rna20700.p1, not unique enough\n",
      "-skipping training candidate: rna24949.p1, not unique enough\n",
      "-skipping training candidate: rna31434.p1, not unique enough\n",
      "-skipping training candidate: rna31433.p1, not unique enough\n",
      "-skipping training candidate: rna31435.p1, not unique enough\n",
      "-skipping training candidate: rna60376.p1, not unique enough\n",
      "-skipping training candidate: rna1468.p1, not unique enough\n",
      "-skipping training candidate: rna12309.p1, not unique enough\n",
      "-skipping training candidate: rna12480.p1, not unique enough\n",
      "-skipping training candidate: rna20565.p1, not unique enough\n",
      "-skipping training candidate: rna59365.p1, not unique enough\n",
      "-skipping training candidate: rna1528.p1, not unique enough\n",
      "-skipping training candidate: rna5176.p1, not unique enough\n",
      "-skipping training candidate: rna5172.p1, not unique enough\n",
      "-skipping training candidate: rna5174.p1, not unique enough\n",
      "-skipping training candidate: rna21007.p1, not unique enough\n",
      "-skipping training candidate: rna22407.p1, not unique enough\n",
      "-skipping training candidate: rna27694.p1, not unique enough\n",
      "-skipping training candidate: rna33511.p1, not unique enough\n",
      "-skipping training candidate: rna29927.p1, not unique enough\n",
      "-skipping training candidate: rna50359.p1, not unique enough\n",
      "-skipping training candidate: rna544.p1, not unique enough\n",
      "-skipping training candidate: rna543.p1, not unique enough\n",
      "-skipping training candidate: rna547.p1, not unique enough\n",
      "-skipping training candidate: rna468.p1, not unique enough\n",
      "-skipping training candidate: rna27605.p1, not unique enough\n",
      "-skipping training candidate: rna29926.p1, not unique enough\n",
      "-skipping training candidate: rna52166.p1, not unique enough\n",
      "-skipping training candidate: rna58777.p1, not unique enough\n",
      "-skipping training candidate: rna15210.p1, not unique enough\n",
      "-skipping training candidate: rna29297.p1, not unique enough\n",
      "-skipping training candidate: rna51072.p1, not unique enough\n",
      "-skipping training candidate: rna52211.p1, not unique enough\n",
      "-skipping training candidate: rna54711.p1, not unique enough\n",
      "-skipping training candidate: rna59366.p1, not unique enough\n",
      "-skipping training candidate: rna7009.p1, not unique enough\n",
      "-skipping training candidate: rna10598.p1, not unique enough\n",
      "-skipping training candidate: rna24772.p1, not unique enough\n",
      "-skipping training candidate: rna27617.p1, not unique enough\n",
      "-skipping training candidate: rna46201.p1, not unique enough\n",
      "-skipping training candidate: rna50366.p1, not unique enough\n",
      "-skipping training candidate: rna50362.p1, not unique enough\n",
      "-skipping training candidate: rna6923.p1, not unique enough\n",
      "-skipping training candidate: rna5261.p1, not unique enough\n",
      "-skipping training candidate: rna31087.p1, not unique enough\n",
      "-skipping training candidate: rna29203.p1, not unique enough\n",
      "-skipping training candidate: rna29656.p1, not unique enough\n",
      "-skipping training candidate: rna29667.p1, not unique enough\n",
      "-skipping training candidate: rna38794.p1, not unique enough\n",
      "-skipping training candidate: rna38792.p1, not unique enough\n",
      "-skipping training candidate: rna54245.p1, not unique enough\n",
      "-skipping training candidate: rna58746.p1, not unique enough\n",
      "-skipping training candidate: rna98.p1, not unique enough\n",
      "-skipping training candidate: rna3908.p1, not unique enough\n",
      "-skipping training candidate: rna3911.p1, not unique enough\n",
      "-skipping training candidate: rna3910.p1, not unique enough\n",
      "-skipping training candidate: rna8983.p1, not unique enough\n",
      "-skipping training candidate: rna11483.p1, not unique enough\n",
      "-skipping training candidate: rna10808.p1, not unique enough\n",
      "-skipping training candidate: rna24372.p1, not unique enough\n",
      "-skipping training candidate: rna24371.p1, not unique enough\n",
      "-skipping training candidate: rna27408.p1, not unique enough\n",
      "-skipping training candidate: rna56326.p1, not unique enough\n",
      "-skipping training candidate: rna57281.p1, not unique enough\n",
      "-skipping training candidate: rna1529.p1, not unique enough\n",
      "-skipping training candidate: rna13174.p1, not unique enough\n",
      "-skipping training candidate: rna12414.p1, not unique enough\n",
      "-skipping training candidate: rna30701.p1, not unique enough\n",
      "-skipping training candidate: rna36759.p1, not unique enough\n",
      "-skipping training candidate: rna48983.p1, not unique enough\n",
      "-skipping training candidate: rna12989.p1, not unique enough\n",
      "-skipping training candidate: rna19028.p1, not unique enough\n",
      "-skipping training candidate: rna21840.p1, not unique enough\n",
      "-skipping training candidate: rna34513.p1, not unique enough\n",
      "-skipping training candidate: rna34512.p1, not unique enough\n",
      "-skipping training candidate: rna34514.p1, not unique enough\n",
      "-skipping training candidate: rna4327.p1, not unique enough\n",
      "-skipping training candidate: rna8817.p1, not unique enough\n",
      "-skipping training candidate: rna17470.p1, not unique enough\n",
      "-skipping training candidate: rna20431.p1, not unique enough\n",
      "-skipping training candidate: rna31067.p1, not unique enough\n",
      "-skipping training candidate: rna31819.p1, not unique enough\n",
      "-skipping training candidate: rna50361.p1, not unique enough\n",
      "-skipping training candidate: rna51071.p1, not unique enough\n",
      "-skipping training candidate: rna1525.p1, not unique enough\n",
      "-skipping training candidate: rna10594.p1, not unique enough\n",
      "-skipping training candidate: rna14036.p1, not unique enough\n",
      "-skipping training candidate: rna51070.p1, not unique enough\n",
      "-skipping training candidate: rna61297.p1, not unique enough\n",
      "-skipping training candidate: rna8236.p1, not unique enough\n",
      "-skipping training candidate: rna20987.p1, not unique enough\n",
      "-skipping training candidate: rna24626.p1, not unique enough\n",
      "-skipping training candidate: rna23759.p1, not unique enough\n",
      "-skipping training candidate: rna35895.p1, not unique enough\n",
      "-skipping training candidate: rna38271.p1, not unique enough\n",
      "-skipping training candidate: rna40161.p1, not unique enough\n",
      "-skipping training candidate: rna42953.p1, not unique enough\n",
      "-skipping training candidate: rna4039.p1, not unique enough\n",
      "-skipping training candidate: rna12929.p1, not unique enough\n",
      "-skipping training candidate: rna12928.p1, not unique enough\n",
      "-skipping training candidate: rna12927.p1, not unique enough\n",
      "-skipping training candidate: rna12935.p1, not unique enough\n",
      "-skipping training candidate: rna12936.p1, not unique enough\n",
      "-skipping training candidate: rna12937.p1, not unique enough\n",
      "-skipping training candidate: rna32459.p1, not unique enough\n",
      "-skipping training candidate: rna32655.p1, not unique enough\n",
      "-skipping training candidate: rna34581.p1, not unique enough\n",
      "-skipping training candidate: rna39904.p1, not unique enough\n",
      "-skipping training candidate: rna47404.p1, not unique enough\n",
      "-skipping training candidate: rna44588.p1, not unique enough\n",
      "-skipping training candidate: rna51293.p1, not unique enough\n",
      "-skipping training candidate: rna52209.p1, not unique enough\n",
      "-skipping training candidate: rna99.p1, not unique enough\n",
      "-skipping training candidate: rna8099.p1, not unique enough\n",
      "-skipping training candidate: rna33510.p1, not unique enough\n",
      "-skipping training candidate: rna32942.p1, not unique enough\n",
      "-skipping training candidate: rna41785.p1, not unique enough\n",
      "-skipping training candidate: rna56572.p1, not unique enough\n",
      "-skipping training candidate: rna56558.p1, not unique enough\n",
      "-skipping training candidate: rna56560.p1, not unique enough\n",
      "-skipping training candidate: rna2802.p1, not unique enough\n",
      "-skipping training candidate: rna6577.p1, not unique enough\n",
      "-skipping training candidate: rna8815.p1, not unique enough\n",
      "-skipping training candidate: rna14837.p1, not unique enough\n",
      "-skipping training candidate: rna12942.p1, not unique enough\n",
      "-skipping training candidate: rna14929.p1, not unique enough\n",
      "-skipping training candidate: rna14928.p1, not unique enough\n",
      "-skipping training candidate: rna17688.p1, not unique enough\n",
      "-skipping training candidate: rna29653.p1, not unique enough\n",
      "-skipping training candidate: rna31066.p1, not unique enough\n",
      "-skipping training candidate: rna48698.p1, not unique enough\n",
      "-skipping training candidate: rna48624.p1, not unique enough\n",
      "-skipping training candidate: rna51295.p1, not unique enough\n",
      "-skipping training candidate: rna5177.p1, not unique enough\n",
      "-skipping training candidate: rna5175.p1, not unique enough\n",
      "-skipping training candidate: rna10807.p1, not unique enough\n",
      "-skipping training candidate: rna13176.p1, not unique enough\n",
      "-skipping training candidate: rna30616.p1, not unique enough\n",
      "-skipping training candidate: rna48697.p1, not unique enough\n",
      "-skipping training candidate: rna58749.p1, not unique enough\n",
      "-skipping training candidate: rna7678.p1, not unique enough\n",
      "-skipping training candidate: rna10561.p1, not unique enough\n",
      "-skipping training candidate: rna32288.p1, not unique enough\n",
      "-skipping training candidate: rna32456.p1, not unique enough\n",
      "-skipping training candidate: rna35074.p1, not unique enough\n",
      "-skipping training candidate: rna3665.p1, not unique enough\n",
      "-skipping training candidate: rna546.p1, not unique enough\n",
      "-skipping training candidate: rna18248.p1, not unique enough\n",
      "-skipping training candidate: rna43373.p1, not unique enough\n",
      "-skipping training candidate: rna48024.p1, not unique enough\n",
      "-skipping training candidate: rna56559.p1, not unique enough\n",
      "-skipping training candidate: rna5487.p1, not unique enough\n",
      "-skipping training candidate: rna11504.p1, not unique enough\n",
      "-skipping training candidate: rna11493.p1, not unique enough\n",
      "-skipping training candidate: rna11509.p1, not unique enough\n",
      "-skipping training candidate: rna8800.p1, not unique enough\n",
      "-skipping training candidate: rna8803.p1, not unique enough\n",
      "-skipping training candidate: rna12941.p1, not unique enough\n",
      "-skipping training candidate: rna29662.p1, not unique enough\n",
      "-skipping training candidate: rna36669.p1, not unique enough\n",
      "-skipping training candidate: rna41666.p1, not unique enough\n",
      "-skipping training candidate: rna41664.p1, not unique enough\n",
      "-skipping training candidate: rna54318.p1, not unique enough\n",
      "-skipping training candidate: rna11490.p1, not unique enough\n",
      "-skipping training candidate: rna11499.p1, not unique enough\n",
      "-skipping training candidate: rna11487.p1, not unique enough\n",
      "-skipping training candidate: rna11488.p1, not unique enough\n",
      "-skipping training candidate: rna28946.p1, not unique enough\n",
      "-skipping training candidate: rna30037.p1, not unique enough\n",
      "-skipping training candidate: rna30038.p1, not unique enough\n",
      "-skipping training candidate: rna30041.p1, not unique enough\n",
      "-skipping training candidate: rna30036.p1, not unique enough\n",
      "-skipping training candidate: rna36057.p1, not unique enough\n",
      "-skipping training candidate: rna39903.p1, not unique enough\n",
      "-skipping training candidate: rna56440.p1, not unique enough\n",
      "-skipping training candidate: rna59840.p1, not unique enough\n",
      "-skipping training candidate: rna10559.p1, not unique enough\n",
      "-skipping training candidate: rna15899.p1, not unique enough\n",
      "-skipping training candidate: rna20988.p1, not unique enough\n",
      "-skipping training candidate: rna31467.p1, not unique enough\n",
      "-skipping training candidate: rna52452.p1, not unique enough\n",
      "-skipping training candidate: rna10603.p1, not unique enough\n",
      "-skipping training candidate: rna24627.p1, not unique enough\n",
      "-skipping training candidate: rna27607.p1, not unique enough\n",
      "-skipping training candidate: rna29655.p1, not unique enough\n",
      "-skipping training candidate: rna30043.p1, not unique enough\n",
      "-skipping training candidate: rna30042.p1, not unique enough\n",
      "-skipping training candidate: rna33423.p1, not unique enough\n",
      "-skipping training candidate: rna7370.p1, not unique enough\n",
      "-skipping training candidate: rna9381.p1, not unique enough\n",
      "-skipping training candidate: rna8805.p1, not unique enough\n",
      "-skipping training candidate: rna8801.p1, not unique enough\n",
      "-skipping training candidate: rna16386.p1, not unique enough\n",
      "-skipping training candidate: rna15805.p1, not unique enough\n",
      "-skipping training candidate: rna24213.p1, not unique enough\n",
      "-skipping training candidate: rna28832.p1, not unique enough\n",
      "-skipping training candidate: rna31065.p1, not unique enough\n",
      "-skipping training candidate: rna53749.p1, not unique enough\n",
      "-skipping training candidate: rna56564.p1, not unique enough\n",
      "-skipping training candidate: rna58195.p1, not unique enough\n",
      "-skipping training candidate: rna3249.p1, not unique enough\n",
      "-skipping training candidate: rna3248.p1, not unique enough\n",
      "-skipping training candidate: rna5296.p1, not unique enough\n",
      "-skipping training candidate: rna6429.p1, not unique enough\n",
      "-skipping training candidate: rna10810.p1, not unique enough\n",
      "-skipping training candidate: rna14353.p1, not unique enough\n",
      "-skipping training candidate: rna21465.p1, not unique enough\n",
      "-skipping training candidate: rna21467.p1, not unique enough\n",
      "-skipping training candidate: rna21468.p1, not unique enough\n",
      "-skipping training candidate: rna27844.p1, not unique enough\n",
      "-skipping training candidate: rna32249.p1, not unique enough\n",
      "-skipping training candidate: rna32025.p1, not unique enough\n",
      "-skipping training candidate: rna38267.p1, not unique enough\n",
      "-skipping training candidate: rna6430.p1, not unique enough\n",
      "-skipping training candidate: rna8804.p1, not unique enough\n",
      "-skipping training candidate: rna20989.p1, not unique enough\n",
      "-skipping training candidate: rna27606.p1, not unique enough\n",
      "-skipping training candidate: rna38942.p1, not unique enough\n",
      "-skipping training candidate: rna56561.p1, not unique enough\n",
      "-skipping training candidate: rna8497.p1, not unique enough\n",
      "-skipping training candidate: rna16390.p1, not unique enough\n",
      "-skipping training candidate: rna29654.p1, not unique enough\n",
      "-skipping training candidate: rna53218.p1, not unique enough\n",
      "-skipping training candidate: rna27697.p1, not unique enough\n",
      "-skipping training candidate: rna32592.p1, not unique enough\n",
      "-skipping training candidate: rna41526.p1, not unique enough\n",
      "-skipping training candidate: rna44399.p1, not unique enough\n",
      "-skipping training candidate: rna59540.p1, not unique enough\n",
      "-skipping training candidate: rna60274.p1, not unique enough\n",
      "-skipping training candidate: rna7961.p1, not unique enough\n",
      "-skipping training candidate: rna7948.p1, not unique enough\n",
      "-skipping training candidate: rna9671.p1, not unique enough\n",
      "-skipping training candidate: rna15213.p1, not unique enough\n",
      "-skipping training candidate: rna15215.p1, not unique enough\n",
      "-skipping training candidate: rna15212.p1, not unique enough\n",
      "-skipping training candidate: rna15214.p1, not unique enough\n",
      "-skipping training candidate: rna16391.p1, not unique enough\n",
      "-skipping training candidate: rna21713.p1, not unique enough\n",
      "-skipping training candidate: rna22169.p1, not unique enough\n",
      "-skipping training candidate: rna29829.p1, not unique enough\n",
      "-skipping training candidate: rna32591.p1, not unique enough\n",
      "-skipping training candidate: rna41665.p1, not unique enough\n",
      "-skipping training candidate: rna40410.p1, not unique enough\n",
      "-skipping training candidate: rna44398.p1, not unique enough\n",
      "-skipping training candidate: rna54319.p1, not unique enough\n",
      "-skipping training candidate: rna19746.p1, not unique enough\n",
      "-skipping training candidate: rna45572.p1, not unique enough\n",
      "-skipping training candidate: rna50365.p1, not unique enough\n",
      "-skipping training candidate: rna53107.p1, not unique enough\n",
      "-skipping training candidate: rna51291.p1, not unique enough\n",
      "-skipping training candidate: rna100.p1, not unique enough\n",
      "-skipping training candidate: rna14821.p1, not unique enough\n",
      "-skipping training candidate: rna13880.p1, not unique enough\n",
      "-skipping training candidate: rna13882.p1, not unique enough\n",
      "-skipping training candidate: rna13883.p1, not unique enough\n",
      "-skipping training candidate: rna13879.p1, not unique enough\n",
      "-skipping training candidate: rna12153.p1, not unique enough\n",
      "-skipping training candidate: rna12154.p1, not unique enough\n",
      "-skipping training candidate: rna44591.p1, not unique enough\n",
      "-skipping training candidate: rna49779.p1, not unique enough\n",
      "-skipping training candidate: rna53108.p1, not unique enough\n",
      "-skipping training candidate: rna51074.p1, not unique enough\n",
      "-skipping training candidate: rna56562.p1, not unique enough\n",
      "-skipping training candidate: rna61072.p1, not unique enough\n",
      "-skipping training candidate: rna61071.p1, not unique enough\n",
      "-skipping training candidate: rna3669.p1, not unique enough\n",
      "-skipping training candidate: rna45575.p1, not unique enough\n",
      "-skipping training candidate: rna49780.p1, not unique enough\n",
      "-skipping training candidate: rna56520.p1, not unique enough\n",
      "-skipping training candidate: rna60275.p1, not unique enough\n",
      "-skipping training candidate: rna3041.p1, not unique enough\n",
      "-skipping training candidate: rna3040.p1, not unique enough\n",
      "-skipping training candidate: rna3044.p1, not unique enough\n",
      "-skipping training candidate: rna3042.p1, not unique enough\n",
      "-skipping training candidate: rna3045.p1, not unique enough\n",
      "-skipping training candidate: rna6757.p1, not unique enough\n",
      "-skipping training candidate: rna6756.p1, not unique enough\n",
      "-skipping training candidate: rna6759.p1, not unique enough\n",
      "-skipping training candidate: rna6758.p1, not unique enough\n",
      "-skipping training candidate: rna7007.p1, not unique enough\n",
      "-skipping training candidate: rna13994.p1, not unique enough\n",
      "-skipping training candidate: rna26450.p1, not unique enough\n",
      "-skipping training candidate: rna32458.p1, not unique enough\n",
      "-skipping training candidate: rna36059.p1, not unique enough\n",
      "-skipping training candidate: rna46069.p1, not unique enough\n",
      "-skipping training candidate: rna59691.p1, not unique enough\n",
      "-skipping training candidate: rna6475.p1, not unique enough\n",
      "-skipping training candidate: rna11614.p1, not unique enough\n",
      "-skipping training candidate: rna12934.p1, not unique enough\n",
      "-skipping training candidate: rna45682.p1, not unique enough\n",
      "-skipping training candidate: rna45684.p1, not unique enough\n",
      "-skipping training candidate: rna45681.p1, not unique enough\n",
      "-skipping training candidate: rna52567.p1, not unique enough\n",
      "-skipping training candidate: rna52571.p1, not unique enough\n",
      "-skipping training candidate: rna52572.p1, not unique enough\n",
      "-skipping training candidate: rna52566.p1, not unique enough\n",
      "-skipping training candidate: rna52568.p1, not unique enough\n",
      "-skipping training candidate: rna52570.p1, not unique enough\n",
      "-skipping training candidate: rna52565.p1, not unique enough\n",
      "-skipping training candidate: rna52569.p1, not unique enough\n",
      "-skipping training candidate: rna59368.p1, not unique enough\n",
      "-skipping training candidate: rna4032.p1, not unique enough\n",
      "-skipping training candidate: rna29928.p1, not unique enough\n",
      "-skipping training candidate: rna29661.p1, not unique enough\n",
      "-skipping training candidate: rna36060.p1, not unique enough\n",
      "-skipping training candidate: rna38861.p1, not unique enough\n",
      "-skipping training candidate: rna10560.p1, not unique enough\n",
      "-skipping training candidate: rna24645.p1, not unique enough\n",
      "-skipping training candidate: rna36358.p1, not unique enough\n",
      "-skipping training candidate: rna57283.p1, not unique enough\n",
      "-skipping training candidate: rna2622.p1, not unique enough\n",
      "-skipping training candidate: rna2624.p1, not unique enough\n",
      "-skipping training candidate: rna17886.p1, not unique enough\n",
      "-skipping training candidate: rna22120.p1, not unique enough\n",
      "-skipping training candidate: rna24647.p1, not unique enough\n",
      "-skipping training candidate: rna36813.p1, not unique enough\n",
      "-skipping training candidate: rna36812.p1, not unique enough\n",
      "-skipping training candidate: rna35072.p1, not unique enough\n",
      "-skipping training candidate: rna38293.p1, not unique enough\n",
      "-skipping training candidate: rna37725.p1, not unique enough\n",
      "-skipping training candidate: rna48847.p1, not unique enough\n",
      "-skipping training candidate: rna51292.p1, not unique enough\n",
      "-skipping training candidate: rna51075.p1, not unique enough\n",
      "-skipping training candidate: rna59369.p1, not unique enough\n",
      "-skipping training candidate: rna60378.p1, not unique enough\n",
      "-skipping training candidate: rna8519.p1, not unique enough\n",
      "-skipping training candidate: rna19801.p1, not unique enough\n",
      "-skipping training candidate: rna38484.p1, not unique enough\n",
      "-skipping training candidate: rna38485.p1, not unique enough\n",
      "-skipping training candidate: rna48848.p1, not unique enough\n",
      "-skipping training candidate: rna53119.p1, not unique enough\n",
      "-skipping training candidate: rna56521.p1, not unique enough\n",
      "-skipping training candidate: rna58116.p1, not unique enough\n",
      "-skipping training candidate: rna58115.p1, not unique enough\n",
      "-skipping training candidate: rna58113.p1, not unique enough\n",
      "-skipping training candidate: rna59370.p1, not unique enough\n",
      "-skipping training candidate: rna7679.p1, not unique enough\n",
      "-skipping training candidate: rna15457.p1, not unique enough\n",
      "-skipping training candidate: rna15458.p1, not unique enough\n",
      "-skipping training candidate: rna31876.p1, not unique enough\n",
      "-skipping training candidate: rna36033.p1, not unique enough\n",
      "-skipping training candidate: rna58112.p1, not unique enough\n",
      "-skipping training candidate: rna1745.p1, not unique enough\n",
      "-skipping training candidate: rna165.p1, not unique enough\n",
      "-skipping training candidate: rna12940.p1, not unique enough\n",
      "-skipping training candidate: rna12943.p1, not unique enough\n",
      "-skipping training candidate: rna20592.p1, not unique enough\n",
      "-skipping training candidate: rna59692.p1, not unique enough\n",
      "-skipping training candidate: rna2298.p1, not unique enough\n",
      "-skipping training candidate: rna6903.p1, not unique enough\n",
      "-skipping training candidate: rna14163.p1, not unique enough\n",
      "-skipping training candidate: rna13968.p1, not unique enough\n",
      "-skipping training candidate: rna27701.p1, not unique enough\n",
      "-skipping training candidate: rna31070.p1, not unique enough\n",
      "-skipping training candidate: rna31290.p1, not unique enough\n",
      "-skipping training candidate: rna31291.p1, not unique enough\n",
      "-skipping training candidate: rna33381.p1, not unique enough\n",
      "-skipping training candidate: rna35073.p1, not unique enough\n",
      "-skipping training candidate: rna37726.p1, not unique enough\n",
      "-skipping training candidate: rna53605.p1, not unique enough\n",
      "-skipping training candidate: rna53909.p1, not unique enough\n",
      "-skipping training candidate: rna56870.p1, not unique enough\n",
      "-skipping training candidate: rna3663.p1, not unique enough\n",
      "-skipping training candidate: rna6904.p1, not unique enough\n",
      "-skipping training candidate: rna12960.p1, not unique enough\n",
      "-skipping training candidate: rna31295.p1, not unique enough\n",
      "-skipping training candidate: rna35605.p1, not unique enough\n",
      "-skipping training candidate: rna34520.p1, not unique enough\n",
      "-skipping training candidate: rna53659.p1, not unique enough\n",
      "-skipping training candidate: rna58991.p1, not unique enough\n",
      "-skipping training candidate: rna21717.p1, not unique enough\n",
      "-skipping training candidate: rna23154.p2, not unique enough\n",
      "-skipping training candidate: rna27570.p1, not unique enough\n",
      "-skipping training candidate: rna56872.p1, not unique enough\n",
      "-skipping training candidate: rna55757.p1, not unique enough\n",
      "-skipping training candidate: rna13967.p1, not unique enough\n",
      "-skipping training candidate: rna13969.p1, not unique enough\n",
      "-skipping training candidate: rna23928.p1, not unique enough\n",
      "-skipping training candidate: rna39905.p1, not unique enough\n",
      "-skipping training candidate: rna43088.p1, not unique enough\n",
      "-skipping training candidate: rna44575.p1, not unique enough\n",
      "-skipping training candidate: rna48844.p1, not unique enough\n",
      "-skipping training candidate: rna59694.p1, not unique enough\n",
      "-skipping training candidate: rna60630.p1, not unique enough\n",
      "-skipping training candidate: rna469.p1, not unique enough\n",
      "-skipping training candidate: rna4943.p1, not unique enough\n",
      "-skipping training candidate: rna19800.p1, not unique enough\n",
      "-skipping training candidate: rna20433.p1, not unique enough\n",
      "-skipping training candidate: rna23849.p1, not unique enough\n",
      "-skipping training candidate: rna27567.p1, not unique enough\n",
      "-skipping training candidate: rna30559.p1, not unique enough\n",
      "-skipping training candidate: rna43086.p1, not unique enough\n",
      "-skipping training candidate: rna48845.p1, not unique enough\n",
      "-skipping training candidate: rna13881.p1, not unique enough\n",
      "-skipping training candidate: rna19372.p1, not unique enough\n",
      "-skipping training candidate: rna21711.p1, not unique enough\n",
      "-skipping training candidate: rna20432.p1, not unique enough\n",
      "-skipping training candidate: rna23850.p1, not unique enough\n",
      "-skipping training candidate: rna35219.p1, not unique enough\n",
      "-skipping training candidate: rna44576.p1, not unique enough\n",
      "-skipping training candidate: rna50364.p1, not unique enough\n",
      "-skipping training candidate: rna882.p1, not unique enough\n",
      "-skipping training candidate: rna11634.p1, not unique enough\n",
      "-skipping training candidate: rna11632.p1, not unique enough\n",
      "-skipping training candidate: rna10488.p1, not unique enough\n",
      "-skipping training candidate: rna13922.p1, not unique enough\n",
      "-skipping training candidate: rna22327.p1, not unique enough\n",
      "-skipping training candidate: rna23165.p2, not unique enough\n",
      "-skipping training candidate: rna23163.p2, not unique enough\n",
      "-skipping training candidate: rna24975.p1, not unique enough\n",
      "-skipping training candidate: rna22726.p1, not unique enough\n",
      "-skipping training candidate: rna25926.p1, not unique enough\n",
      "-skipping training candidate: rna31597.p1, not unique enough\n",
      "-skipping training candidate: rna56207.p1, not unique enough\n",
      "-skipping training candidate: rna56574.p1, not unique enough\n",
      "-skipping training candidate: rna3043.p1, not unique enough\n",
      "-skipping training candidate: rna10321.p1, not unique enough\n",
      "-skipping training candidate: rna14766.p1, not unique enough\n",
      "-skipping training candidate: rna19779.p1, not unique enough\n",
      "-skipping training candidate: rna32287.p1, not unique enough\n",
      "-skipping training candidate: rna40769.p1, not unique enough\n",
      "-skipping training candidate: rna45576.p1, not unique enough\n",
      "-skipping training candidate: rna60213.p1, not unique enough\n",
      "-skipping training candidate: rna9098.p1, not unique enough\n",
      "-skipping training candidate: rna16385.p1, not unique enough\n",
      "-skipping training candidate: rna19803.p1, not unique enough\n",
      "-skipping training candidate: rna26264.p1, not unique enough\n",
      "-skipping training candidate: rna26263.p1, not unique enough\n",
      "-skipping training candidate: rna26261.p1, not unique enough\n",
      "-skipping training candidate: rna26266.p1, not unique enough\n",
      "-skipping training candidate: rna26262.p1, not unique enough\n",
      "-skipping training candidate: rna32460.p1, not unique enough\n",
      "-skipping training candidate: rna43087.p1, not unique enough\n",
      "-skipping training candidate: rna45573.p1, not unique enough\n",
      "-skipping training candidate: rna4510.p1, not unique enough\n",
      "-skipping training candidate: rna8498.p1, not unique enough\n",
      "-skipping training candidate: rna37724.p1, not unique enough\n",
      "-skipping training candidate: rna49781.p1, not unique enough\n",
      "-skipping training candidate: rna3666.p1, not unique enough\n",
      "-skipping training candidate: rna6767.p1, not unique enough\n",
      "-skipping training candidate: rna10318.p1, not unique enough\n",
      "-skipping training candidate: rna9490.p1, not unique enough\n",
      "-skipping training candidate: rna25848.p1, not unique enough\n",
      "-skipping training candidate: rna29660.p1, not unique enough\n",
      "-skipping training candidate: rna37676.p1, not unique enough\n",
      "-skipping training candidate: rna50368.p1, not unique enough\n",
      "-skipping training candidate: rna49603.p1, not unique enough\n",
      "-skipping training candidate: rna57926.p1, not unique enough\n",
      "-skipping training candidate: rna57924.p1, not unique enough\n",
      "-skipping training candidate: rna2164.p1, not unique enough\n",
      "-skipping training candidate: rna5616.p1, not unique enough\n",
      "-skipping training candidate: rna10319.p1, not unique enough\n",
      "-skipping training candidate: rna9482.p1, not unique enough\n",
      "-skipping training candidate: rna25710.p1, not unique enough\n",
      "-skipping training candidate: rna27568.p1, not unique enough\n",
      "-skipping training candidate: rna59963.p1, not unique enough\n",
      "-skipping training candidate: rna1366.p1, not unique enough\n",
      "-skipping training candidate: rna11505.p1, not unique enough\n",
      "-skipping training candidate: rna11494.p1, not unique enough\n",
      "-skipping training candidate: rna12931.p1, not unique enough\n",
      "-skipping training candidate: rna21149.p1, not unique enough\n",
      "-skipping training candidate: rna32457.p1, not unique enough\n",
      "-skipping training candidate: rna33496.p1, not unique enough\n",
      "-skipping training candidate: rna37269.p1, not unique enough\n",
      "-skipping training candidate: rna49604.p1, not unique enough\n",
      "-skipping training candidate: rna53938.p1, not unique enough\n",
      "-skipping training candidate: rna51966.p1, not unique enough\n",
      "-skipping training candidate: rna11496.p1, not unique enough\n",
      "-skipping training candidate: rna11491.p1, not unique enough\n",
      "-skipping training candidate: rna10596.p1, not unique enough\n",
      "-skipping training candidate: rna24625.p1, not unique enough\n",
      "-skipping training candidate: rna32026.p1, not unique enough\n",
      "-skipping training candidate: rna37686.p1, not unique enough\n",
      "-skipping training candidate: rna41667.p1, not unique enough\n",
      "-skipping training candidate: rna43089.p1, not unique enough\n",
      "-skipping training candidate: rna56209.p1, not unique enough\n",
      "-skipping training candidate: rna3667.p1, not unique enough\n",
      "-skipping training candidate: rna3668.p1, not unique enough\n",
      "-skipping training candidate: rna3670.p1, not unique enough\n",
      "-skipping training candidate: rna10972.p1, not unique enough\n",
      "-skipping training candidate: rna20633.p1, not unique enough\n",
      "-skipping training candidate: rna29659.p1, not unique enough\n",
      "-skipping training candidate: rna48745.p1, not unique enough\n",
      "-skipping training candidate: rna21407.p1, not unique enough\n",
      "-skipping training candidate: rna21450.p1, not unique enough\n",
      "-skipping training candidate: rna36811.p1, not unique enough\n",
      "-skipping training candidate: rna48743.p1, not unique enough\n",
      "-skipping training candidate: rna52593.p1, not unique enough\n",
      "-skipping training candidate: rna59661.p1, not unique enough\n",
      "-skipping training candidate: rna3417.p1, not unique enough\n",
      "-skipping training candidate: rna11004.p1, not unique enough\n",
      "-skipping training candidate: rna9672.p1, not unique enough\n",
      "-skipping training candidate: rna28314.p1, not unique enough\n",
      "-skipping training candidate: rna40772.p1, not unique enough\n",
      "-skipping training candidate: rna44450.p1, not unique enough\n",
      "-skipping training candidate: rna44449.p1, not unique enough\n",
      "-skipping training candidate: rna5123.p1, not unique enough\n",
      "-skipping training candidate: rna29244.p1, not unique enough\n",
      "-skipping training candidate: rna29242.p1, not unique enough\n",
      "-skipping training candidate: rna40775.p1, not unique enough\n",
      "-skipping training candidate: rna40771.p1, not unique enough\n",
      "-skipping training candidate: rna44452.p1, not unique enough\n",
      "-skipping training candidate: rna53374.p1, not unique enough\n",
      "-skipping training candidate: rna56167.p1, not unique enough\n",
      "-skipping training candidate: rna56168.p1, not unique enough\n",
      "-skipping training candidate: rna1842.p1, not unique enough\n",
      "-skipping training candidate: rna7013.p1, not unique enough\n",
      "-skipping training candidate: rna7015.p1, not unique enough\n",
      "-skipping training candidate: rna7012.p1, not unique enough\n",
      "-skipping training candidate: rna10973.p1, not unique enough\n",
      "-skipping training candidate: rna10971.p1, not unique enough\n",
      "-skipping training candidate: rna14354.p1, not unique enough\n",
      "-skipping training candidate: rna17339.p1, not unique enough\n",
      "-skipping training candidate: rna19981.p1, not unique enough\n",
      "-skipping training candidate: rna44367.p1, not unique enough\n",
      "-skipping training candidate: rna56160.p1, not unique enough\n",
      "-skipping training candidate: rna4332.p1, not unique enough\n",
      "-skipping training candidate: rna4333.p1, not unique enough\n",
      "-skipping training candidate: rna12946.p1, not unique enough\n",
      "-skipping training candidate: rna18246.p1, not unique enough\n",
      "-skipping training candidate: rna18245.p1, not unique enough\n",
      "-skipping training candidate: rna22125.p1, not unique enough\n",
      "-skipping training candidate: rna30561.p1, not unique enough\n",
      "-skipping training candidate: rna35071.p1, not unique enough\n",
      "-skipping training candidate: rna42414.p1, not unique enough\n",
      "-skipping training candidate: rna48742.p1, not unique enough\n",
      "-skipping training candidate: rna53657.p1, not unique enough\n",
      "-skipping training candidate: rna50871.p1, not unique enough\n",
      "-skipping training candidate: rna59689.p1, not unique enough\n",
      "-skipping training candidate: rna5399.p1, not unique enough\n",
      "-skipping training candidate: rna5400.p1, not unique enough\n",
      "-skipping training candidate: rna16877.p1, not unique enough\n",
      "-skipping training candidate: rna26689.p1, not unique enough\n",
      "-skipping training candidate: rna29939.p1, not unique enough\n",
      "-skipping training candidate: rna33497.p1, not unique enough\n",
      "-skipping training candidate: rna34825.p1, not unique enough\n",
      "-skipping training candidate: rna48744.p1, not unique enough\n",
      "-skipping training candidate: rna53372.p1, not unique enough\n",
      "-skipping training candidate: rna56869.p1, not unique enough\n",
      "-skipping training candidate: rna60590.p1, not unique enough\n",
      "-skipping training candidate: rna1744.p1, not unique enough\n",
      "-skipping training candidate: rna16878.p1, not unique enough\n",
      "-skipping training candidate: rna33498.p1, not unique enough\n",
      "-skipping training candidate: rna37522.p1, not unique enough\n",
      "-skipping training candidate: rna48748.p1, not unique enough\n",
      "-skipping training candidate: rna50369.p1, not unique enough\n",
      "-skipping training candidate: rna53656.p1, not unique enough\n",
      "-skipping training candidate: rna56565.p1, not unique enough\n",
      "-skipping training candidate: rna12939.p1, not unique enough\n",
      "-skipping training candidate: rna24646.p1, not unique enough\n",
      "-skipping training candidate: rna30040.p1, not unique enough\n",
      "-skipping training candidate: rna32944.p1, not unique enough\n",
      "-skipping training candidate: rna41630.p1, not unique enough\n",
      "-skipping training candidate: rna45200.p1, not unique enough\n",
      "-skipping training candidate: rna44590.p1, not unique enough\n",
      "-skipping training candidate: rna51971.p1, not unique enough\n",
      "-skipping training candidate: rna56871.p1, not unique enough\n",
      "-skipping training candidate: rna4512.p1, not unique enough\n",
      "-skipping training candidate: rna20891.p1, not unique enough\n",
      "-skipping training candidate: rna27613.p1, not unique enough\n",
      "-skipping training candidate: rna25925.p1, not unique enough\n",
      "-skipping training candidate: rna45199.p1, not unique enough\n",
      "-skipping training candidate: rna44453.p1, not unique enough\n",
      "-skipping training candidate: rna52289.p1, not unique enough\n",
      "-skipping training candidate: rna50566.p1, not unique enough\n",
      "-skipping training candidate: rna51970.p1, not unique enough\n",
      "-skipping training candidate: rna54579.p1, not unique enough\n",
      "-skipping training candidate: rna12947.p1, not unique enough\n",
      "-skipping training candidate: rna19798.p1, not unique enough\n",
      "-skipping training candidate: rna31064.p1, not unique enough\n",
      "-skipping training candidate: rna32291.p1, not unique enough\n",
      "-skipping training candidate: rna43876.p1, not unique enough\n",
      "-skipping training candidate: rna45541.p1, not unique enough\n",
      "-skipping training candidate: rna50568.p1, not unique enough\n",
      "-skipping training candidate: rna51969.p1, not unique enough\n",
      "-skipping training candidate: rna53704.p1, not unique enough\n",
      "-skipping training candidate: rna54580.p1, not unique enough\n",
      "-skipping training candidate: rna57538.p1, not unique enough\n",
      "-skipping training candidate: rna824.p1, not unique enough\n",
      "-skipping training candidate: rna6489.p1, not unique enough\n",
      "-skipping training candidate: rna14059.p1, not unique enough\n",
      "-skipping training candidate: rna32290.p1, not unique enough\n",
      "-skipping training candidate: rna38268.p1, not unique enough\n",
      "-skipping training candidate: rna40752.p1, not unique enough\n",
      "-skipping training candidate: rna40748.p1, not unique enough\n",
      "-skipping training candidate: rna45015.p1, not unique enough\n",
      "-skipping training candidate: rna51660.p1, not unique enough\n",
      "-skipping training candidate: rna52757.p1, not unique enough\n",
      "-skipping training candidate: rna52564.p1, not unique enough\n",
      "-skipping training candidate: rna55482.p1, not unique enough\n",
      "-skipping training candidate: rna826.p1, not unique enough\n",
      "-skipping training candidate: rna9486.p1, not unique enough\n",
      "-skipping training candidate: rna9484.p1, not unique enough\n",
      "-skipping training candidate: rna28569.p1, not unique enough\n",
      "-skipping training candidate: rna28562.p1, not unique enough\n",
      "-skipping training candidate: rna28565.p1, not unique enough\n",
      "-skipping training candidate: rna28563.p1, not unique enough\n",
      "-skipping training candidate: rna45014.p1, not unique enough\n",
      "-skipping training candidate: rna50367.p1, not unique enough\n",
      "-skipping training candidate: rna52761.p1, not unique enough\n",
      "-skipping training candidate: rna52759.p1, not unique enough\n",
      "-skipping training candidate: rna60278.p1, not unique enough\n",
      "-skipping training candidate: rna4252.p1, not unique enough\n",
      "-skipping training candidate: rna6432.p1, not unique enough\n",
      "-skipping training candidate: rna6431.p1, not unique enough\n",
      "-skipping training candidate: rna6433.p1, not unique enough\n",
      "-skipping training candidate: rna6019.p1, not unique enough\n",
      "-skipping training candidate: rna28388.p1, not unique enough\n",
      "-skipping training candidate: rna31471.p1, not unique enough\n",
      "-skipping training candidate: rna52758.p1, not unique enough\n",
      "-skipping training candidate: rna55620.p1, not unique enough\n",
      "-skipping training candidate: rna57539.p1, not unique enough\n",
      "-skipping training candidate: rna9483.p1, not unique enough\n",
      "-skipping training candidate: rna18249.p1, not unique enough\n",
      "-skipping training candidate: rna21744.p1, not unique enough\n",
      "-skipping training candidate: rna21745.p1, not unique enough\n",
      "-skipping training candidate: rna21743.p1, not unique enough\n",
      "-skipping training candidate: rna1348.p1, not unique enough\n",
      "-skipping training candidate: rna5786.p1, not unique enough\n",
      "-skipping training candidate: rna16349.p1, not unique enough\n",
      "-skipping training candidate: rna21748.p1, not unique enough\n",
      "-skipping training candidate: rna30034.p1, not unique enough\n",
      "-skipping training candidate: rna43095.p1, not unique enough\n",
      "-skipping training candidate: rna60214.p1, not unique enough\n",
      "-skipping training candidate: rna7647.p1, not unique enough\n",
      "-skipping training candidate: rna5126.p1, not unique enough\n",
      "-skipping training candidate: rna49355.p1, not unique enough\n",
      "-skipping training candidate: rna49357.p1, not unique enough\n",
      "-skipping training candidate: rna3877.p1, not unique enough\n",
      "-skipping training candidate: rna16875.p1, not unique enough\n",
      "-skipping training candidate: rna21455.p1, not unique enough\n",
      "-skipping training candidate: rna49356.p1, not unique enough\n",
      "-skipping training candidate: rna53400.p1, not unique enough\n",
      "-skipping training candidate: rna36034.p1, not unique enough\n",
      "-skipping training candidate: rna42413.p1, not unique enough\n",
      "-skipping training candidate: rna45196.p1, not unique enough\n",
      "-skipping training candidate: rna52760.p1, not unique enough\n",
      "-skipping training candidate: rna53706.p1, not unique enough\n",
      "-skipping training candidate: rna55056.p1, not unique enough\n",
      "-skipping training candidate: rna55058.p1, not unique enough\n",
      "-skipping training candidate: rna55059.p1, not unique enough\n",
      "-skipping training candidate: rna59688.p1, not unique enough\n",
      "-skipping training candidate: rna10975.p1, not unique enough\n",
      "-skipping training candidate: rna18918.p1, not unique enough\n",
      "-skipping training candidate: rna52762.p1, not unique enough\n",
      "-skipping training candidate: rna52755.p1, not unique enough\n",
      "-skipping training candidate: rna50567.p1, not unique enough\n",
      "-skipping training candidate: rna54581.p1, not unique enough\n",
      "-skipping training candidate: rna166.p1, not unique enough\n",
      "-skipping training candidate: rna4945.p1, not unique enough\n",
      "-skipping training candidate: rna36617.p1, not unique enough\n",
      "-skipping training candidate: rna50569.p1, not unique enough\n",
      "-skipping training candidate: rna54582.p1, not unique enough\n",
      "-skipping training candidate: rna59693.p1, not unique enough\n",
      "-skipping training candidate: rna638.p1, not unique enough\n",
      "-skipping training candidate: rna3047.p1, not unique enough\n",
      "-skipping training candidate: rna6760.p1, not unique enough\n",
      "-skipping training candidate: rna5209.p1, not unique enough\n",
      "-skipping training candidate: rna15897.p1, not unique enough\n",
      "-skipping training candidate: rna19979.p1, not unique enough\n",
      "-skipping training candidate: rna21714.p1, not unique enough\n",
      "-skipping training candidate: rna26486.p1, not unique enough\n",
      "-skipping training candidate: rna27179.p1, not unique enough\n",
      "-skipping training candidate: rna27180.p1, not unique enough\n",
      "-skipping training candidate: rna60589.p1, not unique enough\n",
      "-skipping training candidate: rna564.p1, not unique enough\n",
      "-skipping training candidate: rna562.p1, not unique enough\n",
      "-skipping training candidate: rna563.p1, not unique enough\n",
      "-skipping training candidate: rna565.p1, not unique enough\n",
      "-skipping training candidate: rna2163.p1, not unique enough\n",
      "-skipping training candidate: rna5370.p1, not unique enough\n",
      "-skipping training candidate: rna16389.p1, not unique enough\n",
      "-skipping training candidate: rna21715.p1, not unique enough\n",
      "-skipping training candidate: rna27178.p1, not unique enough\n",
      "-skipping training candidate: rna33501.p1, not unique enough\n",
      "-skipping training candidate: rna34021.p1, not unique enough\n",
      "-skipping training candidate: rna36695.p1, not unique enough\n",
      "-skipping training candidate: rna36696.p1, not unique enough\n",
      "-skipping training candidate: rna36694.p1, not unique enough\n",
      "-skipping training candidate: rna46014.p1, not unique enough\n",
      "-skipping training candidate: rna60591.p1, not unique enough\n",
      "-skipping training candidate: rna1943.p1, not unique enough\n",
      "-skipping training candidate: rna5369.p1, not unique enough\n",
      "-skipping training candidate: rna16388.p1, not unique enough\n",
      "-skipping training candidate: rna17338.p1, not unique enough\n",
      "-skipping training candidate: rna19797.p1, not unique enough\n",
      "-skipping training candidate: rna41208.p1, not unique enough\n",
      "-skipping training candidate: rna46012.p1, not unique enough\n",
      "-skipping training candidate: rna45207.p1, not unique enough\n",
      "-skipping training candidate: rna52439.p1, not unique enough\n",
      "-skipping training candidate: rna56077.p1, not unique enough\n",
      "-skipping training candidate: rna612.p1, not unique enough\n",
      "-skipping training candidate: rna1351.p1, not unique enough\n",
      "-skipping training candidate: rna5785.p1, not unique enough\n",
      "-skipping training candidate: rna18062.p1, not unique enough\n",
      "-skipping training candidate: rna28368.p1, not unique enough\n",
      "-skipping training candidate: rna33181.p1, not unique enough\n",
      "-skipping training candidate: rna53200.p1, not unique enough\n",
      "-skipping training candidate: rna53186.p1, not unique enough\n",
      "-skipping training candidate: rna53700.p1, not unique enough\n",
      "-skipping training candidate: rna53702.p1, not unique enough\n",
      "-skipping training candidate: rna56910.p1, not unique enough\n",
      "-skipping training candidate: rna7723.p1, not unique enough\n",
      "-skipping training candidate: rna6730.p1, not unique enough\n",
      "-skipping training candidate: rna14940.p1, not unique enough\n",
      "-skipping training candidate: rna17536.p1, not unique enough\n",
      "-skipping training candidate: rna19802.p1, not unique enough\n",
      "-skipping training candidate: rna20434.p1, not unique enough\n",
      "-skipping training candidate: rna20249.p1, not unique enough\n",
      "-skipping training candidate: rna20250.p1, not unique enough\n",
      "-skipping training candidate: rna35064.p1, not unique enough\n",
      "-skipping training candidate: rna35066.p1, not unique enough\n",
      "-skipping training candidate: rna35063.p1, not unique enough\n",
      "-skipping training candidate: rna46415.p1, not unique enough\n",
      "-skipping training candidate: rna12885.p1, not unique enough\n",
      "-skipping training candidate: rna16393.p1, not unique enough\n",
      "-skipping training candidate: rna20816.p1, not unique enough\n",
      "-skipping training candidate: rna21348.p1, not unique enough\n",
      "-skipping training candidate: rna51449.p1, not unique enough\n",
      "-skipping training candidate: rna7294.p1, not unique enough\n",
      "-skipping training candidate: rna12891.p1, not unique enough\n",
      "-skipping training candidate: rna16392.p1, not unique enough\n",
      "-skipping training candidate: rna21716.p1, not unique enough\n",
      "-skipping training candidate: rna25615.p1, not unique enough\n",
      "-skipping training candidate: rna33494.p1, not unique enough\n",
      "-skipping training candidate: rna53401.p1, not unique enough\n",
      "-skipping training candidate: rna7290.p1, not unique enough\n",
      "-skipping training candidate: rna5561.p1, not unique enough\n",
      "-skipping training candidate: rna25538.p1, not unique enough\n",
      "-skipping training candidate: rna25616.p1, not unique enough\n",
      "-skipping training candidate: rna29061.p1, not unique enough\n",
      "-skipping training candidate: rna43875.p1, not unique enough\n",
      "-skipping training candidate: rna48702.p1, not unique enough\n",
      "-skipping training candidate: rna48701.p1, not unique enough\n",
      "-skipping training candidate: rna49969.p1, not unique enough\n",
      "-skipping training candidate: rna53189.p1, not unique enough\n",
      "-skipping training candidate: rna3781.p1, not unique enough\n",
      "-skipping training candidate: rna1350.p1, not unique enough\n",
      "-skipping training candidate: rna16849.p1, not unique enough\n",
      "-skipping training candidate: rna27696.p1, not unique enough\n",
      "-skipping training candidate: rna30675.p1, not unique enough\n",
      "-skipping training candidate: rna31940.p1, not unique enough\n",
      "-skipping training candidate: rna31939.p1, not unique enough\n",
      "-skipping training candidate: rna36518.p1, not unique enough\n",
      "-skipping training candidate: rna56915.p1, not unique enough\n",
      "-skipping training candidate: rna3790.p1, not unique enough\n",
      "-skipping training candidate: rna3785.p1, not unique enough\n",
      "-skipping training candidate: rna9488.p1, not unique enough\n",
      "-skipping training candidate: rna41207.p1, not unique enough\n",
      "-skipping training candidate: rna41206.p1, not unique enough\n",
      "-skipping training candidate: rna43877.p1, not unique enough\n",
      "-skipping training candidate: rna3788.p1, not unique enough\n",
      "-skipping training candidate: rna11506.p1, not unique enough\n",
      "-skipping training candidate: rna17919.p1, not unique enough\n",
      "-skipping training candidate: rna23348.p1, not unique enough\n",
      "-skipping training candidate: rna37261.p1, not unique enough\n",
      "-skipping training candidate: rna37260.p1, not unique enough\n",
      "-skipping training candidate: rna37673.p1, not unique enough\n",
      "-skipping training candidate: rna41774.p1, not unique enough\n",
      "-skipping training candidate: rna46640.p1, not unique enough\n",
      "-skipping training candidate: rna56570.p1, not unique enough\n",
      "-skipping training candidate: rna3786.p1, not unique enough\n",
      "-skipping training candidate: rna11500.p1, not unique enough\n",
      "-skipping training candidate: rna12899.p1, not unique enough\n",
      "-skipping training candidate: rna19684.p1, not unique enough\n",
      "-skipping training candidate: rna24641.p1, not unique enough\n",
      "-skipping training candidate: rna36526.p1, not unique enough\n",
      "-skipping training candidate: rna46639.p1, not unique enough\n",
      "-skipping training candidate: rna48747.p1, not unique enough\n",
      "-skipping training candidate: rna52637.p1, not unique enough\n",
      "-skipping training candidate: rna55616.p1, not unique enough\n",
      "-skipping training candidate: rna55617.p1, not unique enough\n",
      "-skipping training candidate: rna11201.p1, not unique enough\n",
      "-skipping training candidate: rna38171.p1, not unique enough\n",
      "-skipping training candidate: rna38292.p1, not unique enough\n",
      "-skipping training candidate: rna46414.p1, not unique enough\n",
      "-skipping training candidate: rna52950.p1, not unique enough\n",
      "-skipping training candidate: rna53192.p1, not unique enough\n",
      "-skipping training candidate: rna53701.p1, not unique enough\n",
      "-skipping training candidate: rna56164.p1, not unique enough\n",
      "-skipping training candidate: rna56434.p1, not unique enough\n",
      "-skipping training candidate: rna3724.p1, not unique enough\n",
      "-skipping training candidate: rna16850.p1, not unique enough\n",
      "-skipping training candidate: rna19570.p1, not unique enough\n",
      "-skipping training candidate: rna37270.p1, not unique enough\n",
      "-skipping training candidate: rna44954.p1, not unique enough\n",
      "-skipping training candidate: rna46543.p1, not unique enough\n",
      "-skipping training candidate: rna46544.p1, not unique enough\n",
      "-skipping training candidate: rna48072.p1, not unique enough\n",
      "-skipping training candidate: rna48073.p1, not unique enough\n",
      "-skipping training candidate: rna48071.p1, not unique enough\n",
      "-skipping training candidate: rna51968.p1, not unique enough\n",
      "-skipping training candidate: rna51972.p1, not unique enough\n",
      "-skipping training candidate: rna3723.p1, not unique enough\n",
      "-skipping training candidate: rna3727.p1, not unique enough\n",
      "-skipping training candidate: rna10007.p1, not unique enough\n",
      "-skipping training candidate: rna12543.p1, not unique enough\n",
      "-skipping training candidate: rna31468.p1, not unique enough\n",
      "-skipping training candidate: rna31469.p1, not unique enough\n",
      "\n",
      "\t-redundancy-minimized set includes 2014 / 5000 = 40.28%\n",
      "\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr 500 > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest\n",
      "PCT_GC: 43.8\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/seq_n_baseprobs_to_loglikelihood_vals.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//base_freqs.dat > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//hexamer.scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/score_CDS_likelihood_all_6_frames.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//hexamer.scores > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/select_best_ORFs_per_transcript.pl --gff3_file ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.gff3 --cds_scores ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.scores  --min_length_auto_accept 645  > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 \n",
      "Selecting best orfs\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/train_start_PWM.pl --transcripts ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa --selected_orfs ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest --out_prefix ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement\n",
      "Training start codon pattern recognition* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/build_atgPWM_+-.pl  --transcripts ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa  --selected_orfs ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest  --out_prefix ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement --pwm_left 20 --pwm_right 10 \n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scoring.+-.pl  --features_plus ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.+.features  --features_minus ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.-.features  --atg_position 20  > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.feature.scores\n",
      "-round: 1\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "-round: 2\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "-round: 3\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "-round: 4\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "-round: 5\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scores_to_ROC.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.feature.scores > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.feature.scores.roc\n",
      "-parsing scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/plot_ROC.Rscript ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.feature.scores.roc || :\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/compute_AUC.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.feature.scores.roc\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.+.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.-.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/deplete_feature_noise.pl  --features_plus ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.+.features  --pwm_minus ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.-.pwm  --out_prefix ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced\n",
      "num features: 400\tnum_incorporate: 120\n",
      "-feature swap of score: -1.40605264711408 instead of -3.75284147529333\n",
      "-feature swap of score: 3.09385333269827 instead of -2.97423186557724\n",
      "-feature swap of score: 2.84680725213525 instead of -2.77940110666865\n",
      "-feature swap of score: -1.10007579422027 instead of -2.6661729785303\n",
      "-feature swap of score: 2.77150528572296 instead of -2.25893589950959\n",
      "-feature swap of score: -1.63531444944453 instead of -2.23082525543057\n",
      "-feature swap of score: -0.980079349106674 instead of -2.10175534193288\n",
      "-feature swap of score: 0.104166354767867 instead of -2.13863332998193\n",
      "-feature swap of score: 0.907947524684922 instead of -1.94774725481962\n",
      "-feature swap of score: 0.546343758232545 instead of -2.03904737043408\n",
      "-feature swap of score: 4.04011689293189 instead of -1.9132885376474\n",
      "-feature swap of score: 3.1609456376325 instead of -2.03686903503235\n",
      "-feature swap of score: -1.27980649300313 instead of -1.58593894512908\n",
      "-feature swap of score: -0.74543555479243 instead of -1.37454178476353\n",
      "-feature swap of score: 0.404220481396799 instead of -1.51705196298263\n",
      "-feature swap of score: 0.458331710139308 instead of -1.57055596315814\n",
      "-feature swap of score: -0.0482089424172396 instead of -1.34338208670224\n",
      "-feature swap of score: 0.83400588989121 instead of -1.36366673268763\n",
      "-feature swap of score: 4.88310898397107 instead of -1.3024541488332\n",
      "-feature swap of score: 4.3298410514217 instead of -1.29573238718853\n",
      "-feature swap of score: 4.19238596342849 instead of -1.54705552602791\n",
      "-feature swap of score: 1.97754780412606 instead of -1.35698565110761\n",
      "-feature swap of score: 0.119367323635178 instead of -1.35797132764935\n",
      "-feature swap of score: 1.72456883590734 instead of -1.44190167512864\n",
      "-feature swap of score: -1.14827084396248 instead of -1.58008835229266\n",
      "-feature swap of score: -1.04333766062854 instead of -1.21931424729271\n",
      "-feature swap of score: -0.854194569955236 instead of -1.04147746543748\n",
      "-feature swap of score: 4.34523828940672 instead of -1.57670555355998\n",
      "-feature swap of score: 2.26689017186158 instead of -0.915540052242827\n",
      "-feature swap of score: 0.576071149585172 instead of -0.936086993691815\n",
      "-feature swap of score: 4.00471006982611 instead of -0.720920451279178\n",
      "-feature swap of score: -0.129255646282681 instead of -0.674614828461519\n",
      "-feature swap of score: -0.0740087664429986 instead of -0.515356006888088\n",
      "-feature swap of score: -0.479203337569863 instead of -0.530697024004942\n",
      "-feature swap of score: 0.540741926732706 instead of -0.64147904153089\n",
      "-feature swap of score: 1.26190930726619 instead of -0.659279917158608\n",
      "-feature swap of score: 0.766550356350002 instead of -1.41811461029606\n",
      "-feature swap of score: 0.836900774013357 instead of -0.821650526118257\n",
      "-feature swap of score: 1.66050542935772 instead of -0.688608416821223\n",
      "-feature swap of score: -0.267019407560664 instead of -0.776360753557054\n",
      "-feature swap of score: 1.26167990005787 instead of -0.675655277792918\n",
      "-feature swap of score: -0.0349858887064554 instead of -0.553624150145984\n",
      "-feature swap of score: 2.91801106621374 instead of -0.632758766232423\n",
      "-feature swap of score: 2.16875909019159 instead of -0.544836047237528\n",
      "-feature swap of score: 1.06448319614531 instead of -0.501665978611968\n",
      "-feature swap of score: 0.0928802166900943 instead of -0.293840381972011\n",
      "-feature swap of score: 0.997960797564876 instead of -0.282983539135728\n",
      "-feature swap of score: -0.272406092896159 instead of -0.273155210254983\n",
      "-feature swap of score: 3.5048584922666 instead of -0.306889388351233\n",
      "-feature swap of score: -0.126855710012692 instead of -0.169680912088122\n",
      "-feature swap of score: 1.4691196317887 instead of -0.0530003735672946\n",
      "-feature swap of score: 1.73680434582641 instead of -0.0423475956948795\n",
      "-feature swap of score: 0.776959839012662 instead of 0.0399378571599738\n",
      "-feature swap of score: 1.70830019473632 instead of 0.064626973719339\n",
      "-feature swap of score: 2.8256965293003 instead of 0.0876950918284519\n",
      "-feature swap of score: 0.013374518432918 instead of -0.609383995945565\n",
      "-feature swap of score: 2.12319794143152 instead of 0.0831221597264322\n",
      "-feature swap of score: 3.1149394870927 instead of 0.132072494339351\n",
      "-feature swap of score: 2.82243598644415 instead of 0.085068126453433\n",
      "-feature swap of score: 1.90485316494912 instead of -0.00656284357187269\n",
      "-feature swap of score: 2.34337747102025 instead of 0.154165510440591\n",
      "-feature swap of score: 2.64965077722897 instead of 0.123582583577812\n",
      "-feature swap of score: 0.941480443708941 instead of 0.148099926272716\n",
      "-feature swap of score: 0.240642975389002 instead of 0.210033800173625\n",
      "-feature swap of score: 2.54979999889659 instead of 0.219120958499336\n",
      "-feature swap of score: 2.67256400548485 instead of 0.160161892528213\n",
      "-feature swap of score: 1.90532524122611 instead of 0.284503337214137\n",
      "-feature swap of score: 2.20753120613019 instead of 0.0706086798170776\n",
      "-feature swap of score: 2.52117526458786 instead of 0.278957557892042\n",
      "-feature swap of score: 2.69603131019289 instead of 0.359132755406249\n",
      "-feature swap of score: 4.0221202851691 instead of 0.548429866945142\n",
      "-feature swap of score: 6.08617635604484 instead of 0.550087984572697\n",
      "-feature swap of score: 0.926902630120103 instead of 0.659558939623973\n",
      "-feature swap of score: 1.46263240472247 instead of 0.818367348961608\n",
      "-feature swap of score: 6.50525133253087 instead of 0.831191804323202\n",
      "-feature swap of score: 2.39488370514887 instead of 0.829871289559189\n",
      "-feature swap of score: 1.87826962015941 instead of 0.696336585424508\n",
      "-feature swap of score: 1.98792698046775 instead of 0.626766101624741\n",
      "-feature swap of score: 2.1560075332881 instead of 0.737295393559387\n",
      "-feature swap of score: 3.02933721425839 instead of 0.775022092000299\n",
      "-feature swap of score: 4.63398178465379 instead of 0.701684870445399\n",
      "-feature swap of score: 2.21421421264382 instead of 0.731392658676677\n",
      "-feature swap of score: 2.73108807026278 instead of 0.646531682799389\n",
      "-feature swap of score: 1.98118536761625 instead of 0.69567847847298\n",
      "-feature swap of score: 1.29126238262313 instead of 0.682472371948076\n",
      "-feature swap of score: 3.40404707455715 instead of 0.829917445333795\n",
      "-feature swap of score: 2.81615503387654 instead of 0.565503847900317\n",
      "-feature swap of score: 3.49470163142624 instead of 0.621761739741194\n",
      "-feature swap of score: 0.503068438743307 instead of 0.469686520397743\n",
      "-feature swap of score: 1.36199957007568 instead of 0.611380916888328\n",
      "-feature swap of score: 5.19850509946918 instead of 0.711214615589483\n",
      "-feature swap of score: 4.10130039405076 instead of 0.664629769400499\n",
      "-feature swap of score: 4.47911640028375 instead of 0.877574898815822\n",
      "-feature swap of score: 1.82281296414313 instead of 1.09816156622156\n",
      "-feature swap of score: 1.53626122019059 instead of 0.91680762110027\n",
      "-feature swap of score: 3.62275320823371 instead of 0.957305923587057\n",
      "-feature swap of score: 1.42847464725092 instead of 1.08462456320911\n",
      "-feature swap of score: 4.14249301611941 instead of 0.986415805944747\n",
      "-feature swap of score: 2.43031614227797 instead of 0.968875225801817\n",
      "-feature swap of score: 1.06590183725599 instead of 0.991436495562598\n",
      "-feature swap of score: 2.02632694790516 instead of 0.983529201479059\n",
      "-feature swap of score: 3.60080664754868 instead of 1.13250249332824\n",
      "-feature swap of score: 4.21030139609961 instead of 1.09801871887995\n",
      "-feature swap of score: 1.91100493284934 instead of 1.1717441372437\n",
      "-feature swap of score: 4.2480395869192 instead of 1.01796590768164\n",
      "-feature swap of score: 2.04704657800203 instead of 1.16336990975957\n",
      "-feature swap of score: 1.38065318518879 instead of 1.08666218772302\n",
      "-feature swap of score: 2.98682240391045 instead of 0.821589242876029\n",
      "-feature swap of score: 2.34727164531967 instead of 0.846092079042534\n",
      "-feature swap of score: 1.87514580844553 instead of 1.01784777470912\n",
      "-feature swap of score: 1.27327341425536 instead of 1.0117824359468\n",
      "-feature swap of score: 2.71678738183317 instead of 1.12721634953476\n",
      "-feature swap of score: 3.99340588272202 instead of 1.42680644920542\n",
      "-feature swap of score: 3.82304999800667 instead of 1.41648512819762\n",
      "-feature swap of score: 4.57833766254179 instead of 1.49243352891958\n",
      "-feature swap of score: 3.28852392023355 instead of 1.78632849992738\n",
      "-num feature swaps: 116\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scoring.+-.pl  --features_plus ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.+.features  --features_minus ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.-.features  --atg_position 20  > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores\n",
      "-round: 1\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "-round: 2\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "-round: 3\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "-round: 4\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "-round: 5\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scores_to_ROC.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc\n",
      "-parsing scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/plot_ROC.Rscript ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc || :\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/compute_AUC.pl ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.+.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/start_codon_refinement.pl --transcripts ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa --gff3_file ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 --workdir ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir/ > ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3\n",
      "Refining start codon selections.\n",
      "-number of revised start positions: 3129\n",
      "* Running CMD: cp ../../output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3 XENLA_9.2_genome_cDNA.fa.transdecoder.gff3\n",
      "copying output to final output file: XENLA_9.2_genome_cDNA.fa.transdecoder.gff3* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_bed.pl XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 > XENLA_9.2_genome_cDNA.fa.transdecoder.bed\n",
      "Making bed file: XENLA_9.2_genome_cDNA.fa.transdecoder.bed\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 --fasta ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa  --genetic_code Universal > XENLA_9.2_genome_cDNA.fa.transdecoder.pep\n",
      "Making pep file: XENLA_9.2_genome_cDNA.fa.transdecoder.pep\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 --fasta ../../output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa --seqType CDS  --genetic_code Universal > XENLA_9.2_genome_cDNA.fa.transdecoder.cds\n",
      "Making cds file: XENLA_9.2_genome_cDNA.fa.transdecoder.cds\n",
      "transdecoder is finished.  See output files XENLA_9.2_genome_cDNA.fa.transdecoder.*\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cdna = sample_Docket.genome_fasta.get_transdecoder_cdna_gtf(models_asgtf_updated, TRANSDECODER_LOC)\n",
    "sample_Docket.add_keyfile(cdna, 'cdna')\n",
    "\n",
    "transdecoder_files = sample_Docket.cdna.to_pep_files(TDLONGORF_LOC, TDPREDICT_LOC)\n",
    "sample_Docket.add_keyfiles(transdecoder_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "iEug6TQ6Cc9D"
   },
   "source": [
    "# 12. Push files to AWS S3\n",
    "\n",
    "Iteratively moves through the file_set and file_dict variables and populates files into the right place in AWS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 13. Pickle the `sample_Docket` variable for use by the next script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "dill_filename = output_folder + '_'.join([species_prefix, conditions, 'sample_Docket.pkl'])\n",
    "\n",
    "with open(dill_filename, 'wb') as file:\n",
    "    dill.dump(sample_Docket, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "Xtro_bro.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
