{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview\n",
    "This notebook collects data from the [Liao et al. 2022](https://www.nature.com/articles/s41467-022-31949-2) Xenopus laevis adult cell atlas.\n",
    "\n",
    "#### Dataset description\n",
    "\n",
    "- Data can be found at the [GEO Accession GSE195790](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE195790).\n",
    "- This notebook previously collects data from one sample, [\"Xenopus_brain_COL65\"](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM6214268), for the purposes of data analysis and exploration."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "L56fjA27DmhA"
   },
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "t0BTCsz7Y48y",
    "outputId": "f6510678-f977-4ec8-ce30-1f2acd4fe666"
   },
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess, os, dill, sys\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Download files\n",
    "\n",
    "#### BioFileDocket\n",
    "First, make a BioFileDocket for the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/glial-origins/output/Xlae_adultbrain/ already exists\n",
      "Files will be saved into /home/ec2-user/glial-origins/output/Xlae_adultbrain/\n"
     ]
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify the name of the species folder in Amazon S3\n",
    "species = 'Xenopus_laevis'\n",
    "\n",
    "# Specify any particular identifying conditions, eg tissue type:\n",
    "conditions = 'adultbrain'\n",
    "\n",
    "################\n",
    "################\n",
    "\n",
    "sample_BFD = BioFileDocket(species, conditions)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Download and add files\n",
    "Next, download the genome_fasta, annotation, and gxc files based on urls.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inferring file name as XENLA_9.2_genome.fa.gz\n",
      "file XENLA_9.2_genome.fa.gz already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome.fa.gz\n",
      "file XENLA_9.2_genome.fa.gz unzipped and object renamed to XENLA_9.2_genome.fa\n",
      "inferring file name as XENLA_9.2_Xenbase.gtf\n",
      "file XENLA_9.2_Xenbase.gtf already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_Xenbase.gtf\n",
      "inferring file name as XENLA_9.2_GCA.gff3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "gzip: /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome.fa: unknown suffix -- ignored\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      " 96  150M   96  144M    0     0  7523k      0  0:00:20  0:00:19  0:00:01 8992k"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "downloaded file /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_GCA.gff3\n",
      "Renaming GFF3 file XENLA_9.2_GCA.gff3 to XENLA_9.2_GCA.gff\n",
      "inferring file name as GSM6214268_Xenopus_brain_COL65_dge.txt.gz\n",
      "file GSM6214268_Xenopus_brain_COL65_dge.txt.gz already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge.txt.gz\n",
      "file GSM6214268_Xenopus_brain_COL65_dge.txt.gz unzipped and object renamed to GSM6214268_Xenopus_brain_COL65_dge.txt\n",
      "key \"annot\" already exists, ignoring\n",
      "key \"annot2\" already exists, ignoring\n",
      "key \"genome_fasta\" already exists, ignoring\n",
      "key \"gxc\" already exists, ignoring\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100  150M  100  150M    0     0  7581k      0  0:00:20  0:00:20 --:--:-- 9086k\n",
      "gzip: /home/ec2-user/glial-origins/output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge.txt: unknown suffix -- ignored\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'species': 'Xenopus_laevis',\n",
       " 'conditions': 'adultbrain',\n",
       " 'directory': '/home/ec2-user/glial-origins/output/Xlae_adultbrain/',\n",
       " 'files': {},\n",
       " 'metadata': <biofile_handling.metadata_object at 0x7f4d617eaef0>,\n",
       " 'annot': <biofile_handling.GenomeGtfFile at 0x7f4d617ebac0>,\n",
       " 'annot2': <biofile_handling.GenomeGffFile at 0x7f4d617ea080>,\n",
       " 'genome_fasta': <biofile_handling.GenomeFastaFile at 0x7f4cb0898be0>,\n",
       " 'gxc': <biofile_handling.GxcFile at 0x7f4d617eabc0>}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify url and other variables\n",
    "genome_fasta_url = 'https://ftp.xenbase.org/pub/Genomics/JGI/Xenla9.2/XENLA_9.2_genome.fa.gz'\n",
    "genome_version = 'XENLA_9.2'\n",
    "\n",
    "annot_url = 'https://ftp.xenbase.org/pub/Genomics/JGI/Xenla9.2/XENLA_9.2_Xenbase.gtf'\n",
    "\n",
    "# Need to get a second annotation in order to pull usable ID mapping information\n",
    "annot2_url = 'https://ftp.xenbase.org/pub/Genomics/JGI/Xenla9.2/XENLA_9.2_GCA.gff3'\n",
    "\n",
    "gxc_url = 'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM6214nnn/GSM6214268/suppl/GSM6214268_Xenopus_brain_COL65_dge.txt.gz'\n",
    "\n",
    "################\n",
    "################\n",
    "\n",
    "protocol = 'curl'\n",
    "    \n",
    "genome_fasta = GenomeFastaFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    version = genome_version,\n",
    "    url = genome_fasta_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "annot = GenomeGtfFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    reference_genome = genome_fasta,\n",
    "    url = annot_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "annot2 = GenomeGffFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    reference_genome = genome_fasta,\n",
    "    url = annot2_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "gxc = GxcFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    reference_genome  = genome_fasta,\n",
    "    reference_annot = annot,\n",
    "    url = gxc_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "keyfiles = {\n",
    "    'annot': annot,\n",
    "    'annot2': annot2,\n",
    "    'genome_fasta': genome_fasta,\n",
    "    'gxc': gxc\n",
    "}\n",
    "sample_BFD.add_keyfiles(keyfiles)\n",
    "\n",
    "display(vars(sample_BFD))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Extract gxc information\n",
    "\n",
    "#### Preview gxc matrix\n",
    "- Show the first 10 rows of the gxc matrix.  \n",
    "- Extract the first column as a 'gene_name' dataframe.\n",
    "- Record the number of cells and genes in the gxc to the `sample_BFD.metadata` attribute."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>AACCTATTCATATAAGGG</th>\n",
       "      <th>CTCGCATCAAAGTTAACT</th>\n",
       "      <th>AACCTAGTATACTTCCGC</th>\n",
       "      <th>AACCTAAAAGTTCTGAAA</th>\n",
       "      <th>CTCGCACGCACCCTCCAT</th>\n",
       "      <th>ACGTTGTATTGTAGCGAG</th>\n",
       "      <th>ACGAGCATGCTTTAGTCG</th>\n",
       "      <th>AACCTAGTCCCGCCATCT</th>\n",
       "      <th>AACCTAGCGAATTAGAGA</th>\n",
       "      <th>...</th>\n",
       "      <th>TCACTTGTTGCCATGCTT</th>\n",
       "      <th>TCGGGTTGTCACACTTAT</th>\n",
       "      <th>TCGTAATCGTAAGTTGCC</th>\n",
       "      <th>TGTCACGAATTACACAAG</th>\n",
       "      <th>TGTGCGTACTTCTAGTCG</th>\n",
       "      <th>TTAACTATACAGTGGATG</th>\n",
       "      <th>TTGGACACTTATGATCTT</th>\n",
       "      <th>AAAGTTACTTATGCCCTC</th>\n",
       "      <th>AACCTACGCACCTGCGGA</th>\n",
       "      <th>AACCTATAGTCGCTGTGT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.S</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>42Sp43.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42Sp50.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AK6.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AK6.S</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MGC107841.L</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MGC107851.L</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MGC107876.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MGC108117.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MGC108429.L</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 30001 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GENE  AACCTATTCATATAAGGG  CTCGCATCAAAGTTAACT  AACCTAGTATACTTCCGC  \\\n",
       "0          3.S                   0                   0                   0   \n",
       "1     42Sp43.L                   0                   0                   0   \n",
       "2     42Sp50.L                   0                   0                   0   \n",
       "3        AK6.L                   0                   0                   0   \n",
       "4        AK6.S                   1                   0                   0   \n",
       "5  MGC107841.L                   1                   0                   1   \n",
       "6  MGC107851.L                   0                   1                   0   \n",
       "7  MGC107876.L                   0                   0                   0   \n",
       "8  MGC108117.L                   0                   0                   0   \n",
       "9  MGC108429.L                   0                   1                   3   \n",
       "\n",
       "   AACCTAAAAGTTCTGAAA  CTCGCACGCACCCTCCAT  ACGTTGTATTGTAGCGAG  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   1                   1                   0   \n",
       "\n",
       "   ACGAGCATGCTTTAGTCG  AACCTAGTCCCGCCATCT  AACCTAGCGAATTAGAGA  ...  \\\n",
       "0                   0                   0                   0  ...   \n",
       "1                   0                   0                   0  ...   \n",
       "2                   0                   0                   0  ...   \n",
       "3                   1                   0                   0  ...   \n",
       "4                   0                   0                   0  ...   \n",
       "5                   0                   0                   0  ...   \n",
       "6                   0                   1                   0  ...   \n",
       "7                   0                   1                   0  ...   \n",
       "8                   0                   0                   0  ...   \n",
       "9                   1                   0                   0  ...   \n",
       "\n",
       "   TCACTTGTTGCCATGCTT  TCGGGTTGTCACACTTAT  TCGTAATCGTAAGTTGCC  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   TGTCACGAATTACACAAG  TGTGCGTACTTCTAGTCG  TTAACTATACAGTGGATG  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   TTGGACACTTATGATCTT  AAAGTTACTTATGCCCTC  AACCTACGCACCTGCGGA  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   AACCTATAGTCGCTGTGT  \n",
       "0                   0  \n",
       "1                   0  \n",
       "2                   0  \n",
       "3                   0  \n",
       "4                   0  \n",
       "5                   0  \n",
       "6                   0  \n",
       "7                   0  \n",
       "8                   0  \n",
       "9                   0  \n",
       "\n",
       "[10 rows x 30001 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GENE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42Sp43.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>42Sp50.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AK6.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25908</th>\n",
       "      <td>zyg11b.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25909</th>\n",
       "      <td>zyg11b.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25910</th>\n",
       "      <td>zzef1.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25911</th>\n",
       "      <td>zzz3.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25912</th>\n",
       "      <td>zzz3.S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25913 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name\n",
       "0          GENE\n",
       "1           3.S\n",
       "2      42Sp43.L\n",
       "3      42Sp50.L\n",
       "4         AK6.L\n",
       "...         ...\n",
       "25908  zyg11b.L\n",
       "25909  zyg11b.S\n",
       "25910   zzef1.S\n",
       "25911    zzz3.L\n",
       "25912    zzz3.S\n",
       "\n",
       "[25913 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes_matrix = pd.read_csv(sample_BFD.gxc.path, sep = '\\t', nrows = 10)\n",
    "display(genes_matrix)\n",
    "\n",
    "gxc_genes_list = pd.read_csv(sample_BFD.gxc.path, sep = '\\t', usecols=[0], names = ['gene_name'])\n",
    "display(gxc_genes_list)\n",
    "\n",
    "sample_BFD.metadata.add('num_cells', len(genes_matrix.columns) - 1)\n",
    "sample_BFD.metadata.add('num_genes', len(gxc_genes_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jvfcFNT2FuaZ"
   },
   "source": [
    "# 3. Get mapping identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2136</td>\n",
       "      <td>2204</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2205</td>\n",
       "      <td>3023</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3024</td>\n",
       "      <td>3092</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3093</td>\n",
       "      <td>4723</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>4724</td>\n",
       "      <td>4798</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771995</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104527008</td>\n",
       "      <td>104527068</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771996</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771997</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771998</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771999</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>772000 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0        1     2          3          4  5  6  7  \\\n",
       "0             MT  Xenbase  exon       2136       2204  .  +  .   \n",
       "1             MT  Xenbase  exon       2205       3023  .  +  .   \n",
       "2             MT  Xenbase  exon       3024       3092  .  +  .   \n",
       "3             MT  Xenbase  exon       3093       4723  .  +  .   \n",
       "4             MT  Xenbase  exon       4724       4798  .  +  .   \n",
       "...          ...      ...   ...        ...        ... .. .. ..   \n",
       "771995  chr9_10S  Xenbase   CDS  104527008  104527068  .  -  0   \n",
       "771996  chr9_10S  Xenbase  exon  104530653  104530753  .  -  2   \n",
       "771997  chr9_10S  Xenbase   CDS  104530653  104530753  .  -  2   \n",
       "771998  chr9_10S  Xenbase  exon  104531027  104531054  .  -  0   \n",
       "771999  chr9_10S  Xenbase   CDS  104531027  104531054  .  -  0   \n",
       "\n",
       "                                                        8  \n",
       "0       gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...  \n",
       "1       gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...  \n",
       "2       gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...  \n",
       "3       gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...  \n",
       "4       gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...  \n",
       "...                                                   ...  \n",
       "771995  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771996  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771997  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771998  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771999  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "\n",
       "[772000 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"; transcript_id \"gene42065_t\"; transcript_name \"mt-trna-phe.L\";'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load in the original GFF-based annotation\n",
    "models = pd.read_csv(sample_BFD.annot.path, skiprows = 0, header = None, sep = '\\t', on_bad_lines = 'skip', comment='#')\n",
    "display(models)\n",
    "\n",
    "attributes_column = 8\n",
    "\n",
    "# Check the structure of fields in the GFF additional fields section\n",
    "display(models[attributes_column][0])\n",
    "\n",
    "# Remove any rows with NaNs\n",
    "models.dropna(inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2136</td>\n",
       "      <td>2204</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...</td>\n",
       "      <td>{'gene_id': 'gene42065', 'gene_name': 'mt-trna...</td>\n",
       "      <td>mt-trna-phe.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2205</td>\n",
       "      <td>3023</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...</td>\n",
       "      <td>{'gene_id': 'gene34778', 'gene_name': 'mt-rnr1...</td>\n",
       "      <td>mt-rnr1.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3024</td>\n",
       "      <td>3092</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...</td>\n",
       "      <td>{'gene_id': 'gene48202', 'gene_name': 'mt-trna...</td>\n",
       "      <td>mt-trna-val.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3093</td>\n",
       "      <td>4723</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...</td>\n",
       "      <td>{'gene_id': 'gene44770', 'gene_name': 'mt-rnr2...</td>\n",
       "      <td>mt-rnr2.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>4724</td>\n",
       "      <td>4798</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...</td>\n",
       "      <td>{'gene_id': 'gene43253', 'gene_name': 'mt-trna...</td>\n",
       "      <td>mt-trna-leu1.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771995</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104527008</td>\n",
       "      <td>104527068</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771996</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771997</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771998</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771999</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>772000 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0        1     2          3          4  5  6  7  \\\n",
       "0             MT  Xenbase  exon       2136       2204  .  +  .   \n",
       "1             MT  Xenbase  exon       2205       3023  .  +  .   \n",
       "2             MT  Xenbase  exon       3024       3092  .  +  .   \n",
       "3             MT  Xenbase  exon       3093       4723  .  +  .   \n",
       "4             MT  Xenbase  exon       4724       4798  .  +  .   \n",
       "...          ...      ...   ...        ...        ... .. .. ..   \n",
       "771995  chr9_10S  Xenbase   CDS  104527008  104527068  .  -  0   \n",
       "771996  chr9_10S  Xenbase  exon  104530653  104530753  .  -  2   \n",
       "771997  chr9_10S  Xenbase   CDS  104530653  104530753  .  -  2   \n",
       "771998  chr9_10S  Xenbase  exon  104531027  104531054  .  -  0   \n",
       "771999  chr9_10S  Xenbase   CDS  104531027  104531054  .  -  0   \n",
       "\n",
       "                                                        8  \\\n",
       "0       gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...   \n",
       "1       gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...   \n",
       "2       gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...   \n",
       "3       gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...   \n",
       "4       gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...   \n",
       "...                                                   ...   \n",
       "771995  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "771996  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "771997  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "771998  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "771999  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "\n",
       "                                         field_dictionary          gene_name  \n",
       "0       {'gene_id': 'gene42065', 'gene_name': 'mt-trna...      mt-trna-phe.L  \n",
       "1       {'gene_id': 'gene34778', 'gene_name': 'mt-rnr1...          mt-rnr1.L  \n",
       "2       {'gene_id': 'gene48202', 'gene_name': 'mt-trna...      mt-trna-val.L  \n",
       "3       {'gene_id': 'gene44770', 'gene_name': 'mt-rnr2...          mt-rnr2.L  \n",
       "4       {'gene_id': 'gene43253', 'gene_name': 'mt-trna...     mt-trna-leu1.L  \n",
       "...                                                   ...                ...  \n",
       "771995  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S  \n",
       "771996  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S  \n",
       "771997  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S  \n",
       "771998  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S  \n",
       "771999  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S  \n",
       "\n",
       "[772000 rows x 11 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "49282"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract field and database cross-ref (dbxref) information into columns\n",
    "models['field_dictionary'] = models[attributes_column].apply(convert_fields_to_dict_gtf)\n",
    "models['gene_name'] = [d.get('gene_name') for d in models['field_dictionary']]\n",
    "display(models)\n",
    "\n",
    "sample_BFD.metadata.add('num_annot_models', len(models['gene_name'].unique()))\n",
    "display(sample_BFD.metadata.num_annot_models)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Name=Xelaev18004747m;end_range=18399,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Note=transcript XELAEV_18...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>17924</td>\n",
       "      <td>18243</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809901</th>\n",
       "      <td>Scaffold94051</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8</td>\n",
       "      <td>205</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809902</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene45941;Name=Xelaev18004691m;end_range=23...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809903</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna47731;Parent=gene45941;Note=transcript X...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809904</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id452950;Parent=rna47731;Note=transcript XE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809905</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>809906 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0        1     2      3      4  5  6  7  \\\n",
       "0               chr1L  Genbank  gene  17924  18399  .  -  .   \n",
       "1               chr1L  Genbank  mRNA  17924  18399  .  -  .   \n",
       "2               chr1L  Genbank  exon  18336  18399  .  -  .   \n",
       "3               chr1L  Genbank  exon  17924  18243  .  -  .   \n",
       "4               chr1L  Genbank   CDS  18336  18399  .  -  0   \n",
       "...               ...      ...   ...    ...    ... .. .. ..   \n",
       "809901  Scaffold94051  Genbank   CDS      8    205  .  +  0   \n",
       "809902  Scaffold95291  Genbank  gene      9    236  .  -  .   \n",
       "809903  Scaffold95291  Genbank  mRNA      9    236  .  -  .   \n",
       "809904  Scaffold95291  Genbank  exon      9    236  .  -  .   \n",
       "809905  Scaffold95291  Genbank   CDS      9    236  .  -  0   \n",
       "\n",
       "                                                        8  \n",
       "0       ID=gene0;Name=Xelaev18004747m;end_range=18399,...  \n",
       "1       ID=rna0;Parent=gene0;Note=transcript XELAEV_18...  \n",
       "2       ID=id1;Parent=rna0;Note=transcript XELAEV_1800...  \n",
       "3       ID=id2;Parent=rna0;Note=transcript XELAEV_1800...  \n",
       "4       ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...  \n",
       "...                                                   ...  \n",
       "809901  ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...  \n",
       "809902  ID=gene45941;Name=Xelaev18004691m;end_range=23...  \n",
       "809903  ID=rna47731;Parent=gene45941;Note=transcript X...  \n",
       "809904  ID=id452950;Parent=rna47731;Note=transcript XE...  \n",
       "809905  ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...  \n",
       "\n",
       "[809906 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Need to load another annotation to be able to extract a Xenbase ID\n",
    "models2 = pd.read_csv(sample_BFD.annot2.path, skiprows = 0, header = None, sep = '\\t', on_bad_lines = 'skip', comment='#')\n",
    "display(models2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Xenbase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Name=Xelaev18004747m;end_range=18399,...</td>\n",
       "      <td>{'ID': 'gene0', 'Name': 'Xelaev18004747m', 'en...</td>\n",
       "      <td>Xelaev18004747m</td>\n",
       "      <td>XB-GENE-5942444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>17924</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Note=transcript XELAEV_18...</td>\n",
       "      <td>{'ID': 'rna0', 'Parent': 'gene0', 'Note': 'tra...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "      <td>{'ID': 'id1', 'Parent': 'rna0', 'Note': 'trans...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>17924</td>\n",
       "      <td>18243</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Note=transcript XELAEV_1800...</td>\n",
       "      <td>{'ID': 'id2', 'Parent': 'rna0', 'Note': 'trans...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1L</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>18336</td>\n",
       "      <td>18399</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...</td>\n",
       "      <td>{'ID': 'cds0', 'Parent': 'rna0', 'Dbxref': 'Ph...</td>\n",
       "      <td>OCT98948.1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809901</th>\n",
       "      <td>Scaffold94051</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8</td>\n",
       "      <td>205</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...</td>\n",
       "      <td>{'ID': 'cds47730', 'Parent': 'rna47730', 'Dbxr...</td>\n",
       "      <td>OCT55143.1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809902</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>gene</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene45941;Name=Xelaev18004691m;end_range=23...</td>\n",
       "      <td>{'ID': 'gene45941', 'Name': 'Xelaev18004691m',...</td>\n",
       "      <td>Xelaev18004691m</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809903</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna47731;Parent=gene45941;Note=transcript X...</td>\n",
       "      <td>{'ID': 'rna47731', 'Parent': 'gene45941', 'Not...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809904</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>exon</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id452950;Parent=rna47731;Note=transcript XE...</td>\n",
       "      <td>{'ID': 'id452950', 'Parent': 'rna47731', 'Note...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809905</th>\n",
       "      <td>Scaffold95291</td>\n",
       "      <td>Genbank</td>\n",
       "      <td>CDS</td>\n",
       "      <td>9</td>\n",
       "      <td>236</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...</td>\n",
       "      <td>{'ID': 'cds47731', 'Parent': 'rna47731', 'Dbxr...</td>\n",
       "      <td>OCT55142.1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>809906 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0        1     2      3      4  5  6  7  \\\n",
       "0               chr1L  Genbank  gene  17924  18399  .  -  .   \n",
       "1               chr1L  Genbank  mRNA  17924  18399  .  -  .   \n",
       "2               chr1L  Genbank  exon  18336  18399  .  -  .   \n",
       "3               chr1L  Genbank  exon  17924  18243  .  -  .   \n",
       "4               chr1L  Genbank   CDS  18336  18399  .  -  0   \n",
       "...               ...      ...   ...    ...    ... .. .. ..   \n",
       "809901  Scaffold94051  Genbank   CDS      8    205  .  +  0   \n",
       "809902  Scaffold95291  Genbank  gene      9    236  .  -  .   \n",
       "809903  Scaffold95291  Genbank  mRNA      9    236  .  -  .   \n",
       "809904  Scaffold95291  Genbank  exon      9    236  .  -  .   \n",
       "809905  Scaffold95291  Genbank   CDS      9    236  .  -  0   \n",
       "\n",
       "                                                        8  \\\n",
       "0       ID=gene0;Name=Xelaev18004747m;end_range=18399,...   \n",
       "1       ID=rna0;Parent=gene0;Note=transcript XELAEV_18...   \n",
       "2       ID=id1;Parent=rna0;Note=transcript XELAEV_1800...   \n",
       "3       ID=id2;Parent=rna0;Note=transcript XELAEV_1800...   \n",
       "4       ID=cds0;Parent=rna0;Dbxref=Phytozome:Xelaev180...   \n",
       "...                                                   ...   \n",
       "809901  ID=cds47730;Parent=rna47730;Dbxref=Phytozome:X...   \n",
       "809902  ID=gene45941;Name=Xelaev18004691m;end_range=23...   \n",
       "809903  ID=rna47731;Parent=gene45941;Note=transcript X...   \n",
       "809904  ID=id452950;Parent=rna47731;Note=transcript XE...   \n",
       "809905  ID=cds47731;Parent=rna47731;Dbxref=Phytozome:X...   \n",
       "\n",
       "                                         field_dictionary        gene_name  \\\n",
       "0       {'ID': 'gene0', 'Name': 'Xelaev18004747m', 'en...  Xelaev18004747m   \n",
       "1       {'ID': 'rna0', 'Parent': 'gene0', 'Note': 'tra...             None   \n",
       "2       {'ID': 'id1', 'Parent': 'rna0', 'Note': 'trans...             None   \n",
       "3       {'ID': 'id2', 'Parent': 'rna0', 'Note': 'trans...             None   \n",
       "4       {'ID': 'cds0', 'Parent': 'rna0', 'Dbxref': 'Ph...       OCT98948.1   \n",
       "...                                                   ...              ...   \n",
       "809901  {'ID': 'cds47730', 'Parent': 'rna47730', 'Dbxr...       OCT55143.1   \n",
       "809902  {'ID': 'gene45941', 'Name': 'Xelaev18004691m',...  Xelaev18004691m   \n",
       "809903  {'ID': 'rna47731', 'Parent': 'gene45941', 'Not...             None   \n",
       "809904  {'ID': 'id452950', 'Parent': 'rna47731', 'Note...             None   \n",
       "809905  {'ID': 'cds47731', 'Parent': 'rna47731', 'Dbxr...       OCT55142.1   \n",
       "\n",
       "                Xenbase  \n",
       "0       XB-GENE-5942444  \n",
       "1                  None  \n",
       "2                  None  \n",
       "3                  None  \n",
       "4                  None  \n",
       "...                 ...  \n",
       "809901             None  \n",
       "809902             None  \n",
       "809903             None  \n",
       "809904             None  \n",
       "809905             None  \n",
       "\n",
       "[809906 rows x 12 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Xenbase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>42sp43.L</td>\n",
       "      <td>XB-GENE-6252610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>42sp50.L</td>\n",
       "      <td>XB-GENE-5853356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LOC100036702.L</td>\n",
       "      <td>XB-GENE-18005853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LOC100037907.L</td>\n",
       "      <td>XB-GENE-18005786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LOC100037907.S</td>\n",
       "      <td>XB-GENE-18005873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21951</th>\n",
       "      <td>zyx.L</td>\n",
       "      <td>XB-GENE-6252950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21952</th>\n",
       "      <td>zyx.S</td>\n",
       "      <td>XB-GENE-6253946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21953</th>\n",
       "      <td>zzef1.S</td>\n",
       "      <td>XB-GENE-6488145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21954</th>\n",
       "      <td>zzz3.L</td>\n",
       "      <td>XB-GENE-6486634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21955</th>\n",
       "      <td>zzz3.S</td>\n",
       "      <td>XB-GENE-17341117</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21956 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_name           Xenbase\n",
       "0            42sp43.L   XB-GENE-6252610\n",
       "1            42sp50.L   XB-GENE-5853356\n",
       "2      LOC100036702.L  XB-GENE-18005853\n",
       "3      LOC100037907.L  XB-GENE-18005786\n",
       "4      LOC100037907.S  XB-GENE-18005873\n",
       "...               ...               ...\n",
       "21951           zyx.L   XB-GENE-6252950\n",
       "21952           zyx.S   XB-GENE-6253946\n",
       "21953         zzef1.S   XB-GENE-6488145\n",
       "21954          zzz3.L   XB-GENE-6486634\n",
       "21955          zzz3.S  XB-GENE-17341117\n",
       "\n",
       "[21956 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract field and database cross-ref (dbxref) information into columns\n",
    "models2['field_dictionary'] = models2[attributes_column].apply(convert_fields_to_dict_gff)\n",
    "models2['gene_name'] = [d.get('Name') for d in models2['field_dictionary']]\n",
    "models2['Xenbase'] = [d.get('Alias') for d in models2['field_dictionary']]\n",
    "display(models2)\n",
    "\n",
    "# Extract gene_name and Xenbase unique pairs to get id mapping between two annotations\n",
    "xenbase_keys = models2[['gene_name', 'Xenbase']].dropna().drop_duplicates()\n",
    "xenbase_keys = xenbase_keys.groupby('gene_name').agg({'gene_name': 'first', 'Xenbase': ','.join}).reset_index(drop = True)\n",
    "display(xenbase_keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19286"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>new_fields</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2136</td>\n",
       "      <td>2204</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2205</td>\n",
       "      <td>3023</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3024</td>\n",
       "      <td>3092</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3093</td>\n",
       "      <td>4723</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>4724</td>\n",
       "      <td>4798</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771995</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104527008</td>\n",
       "      <td>104527068</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771996</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771997</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771998</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771999</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>772000 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0        1     2          3          4  5  6  7  \\\n",
       "0             MT  Xenbase  exon       2136       2204  .  +  .   \n",
       "1             MT  Xenbase  exon       2205       3023  .  +  .   \n",
       "2             MT  Xenbase  exon       3024       3092  .  +  .   \n",
       "3             MT  Xenbase  exon       3093       4723  .  +  .   \n",
       "4             MT  Xenbase  exon       4724       4798  .  +  .   \n",
       "...          ...      ...   ...        ...        ... .. .. ..   \n",
       "771995  chr9_10S  Xenbase   CDS  104527008  104527068  .  -  0   \n",
       "771996  chr9_10S  Xenbase  exon  104530653  104530753  .  -  2   \n",
       "771997  chr9_10S  Xenbase   CDS  104530653  104530753  .  -  2   \n",
       "771998  chr9_10S  Xenbase  exon  104531027  104531054  .  -  0   \n",
       "771999  chr9_10S  Xenbase   CDS  104531027  104531054  .  -  0   \n",
       "\n",
       "                                               new_fields  \n",
       "0       gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...  \n",
       "1       gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...  \n",
       "2       gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...  \n",
       "3       gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...  \n",
       "4       gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...  \n",
       "...                                                   ...  \n",
       "771995  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771996  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771997  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771998  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771999  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "\n",
       "[772000 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "models = models.merge(xenbase_keys, on = 'gene_name', how = 'left')\n",
    "new_models = models.copy(deep = True)\n",
    "\n",
    "# Collect information about the number of unique gene_name + Xenbase pairs\n",
    "sample_BFD.metadata.add('num_xenbase_models', len(new_models[['gene_name', 'Xenbase']].dropna().drop_duplicates())) \n",
    "display(sample_BFD.metadata.num_xenbase_models)\n",
    "\n",
    "# Create a new field dictionary based on new fields\n",
    "new_models['Xenbase'] = new_models['Xenbase'].astype(str)\n",
    "new_models.apply(lambda x: x['field_dictionary'].update({'Xenbase': x['Xenbase']}), axis = 1)\n",
    "new_models['new_fields'] = new_models['field_dictionary'].apply(convert_dict_to_fields_gtf)\n",
    "\n",
    "new_models = new_models[[0, 1, 2, 3, 4, 5, 6, 7, 'new_fields']]\n",
    "\n",
    "models_asgtf_updated_filename = sample_BFD.annot.filename.replace('.gtf', '_updated.gtf')\n",
    "models_asgtf_updated = GenomeGtfFile(\n",
    "    filename = models_asgtf_updated_filename, \n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    reference_genome = sample_BFD.genome_fasta)\n",
    "\n",
    "new_models.to_csv(models_asgtf_updated.path, sep = '\\t', index = None, header = None)\n",
    "display(new_models)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Extract gene IDs for mapping to UniprotKB\n",
    "Specify which set of identifiers will be use to query the [Uniprot ID Mapping Tool](https://www.uniprot.org/id-mapping) via API.\n",
    "\n",
    "If using an identifier from the `dbxref_dict`, specify the name via string in the `dbxref_datafield` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Xenbase</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>vdac2.S</td>\n",
       "      <td>XB-GENE-954844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>samd8.S</td>\n",
       "      <td>XB-GENE-17343426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>227</th>\n",
       "      <td>dusp13.S</td>\n",
       "      <td>XB-GENE-17337763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258</th>\n",
       "      <td>adk.S</td>\n",
       "      <td>XB-GENE-997231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>281</th>\n",
       "      <td>vcl.S</td>\n",
       "      <td>XB-GENE-5759162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771494</th>\n",
       "      <td>nif3l1.S</td>\n",
       "      <td>XB-GENE-6078750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771661</th>\n",
       "      <td>tpsg1.S</td>\n",
       "      <td>XB-GENE-5893017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771750</th>\n",
       "      <td>loc100491461.S</td>\n",
       "      <td>XB-GENE-6489054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771854</th>\n",
       "      <td>rnf112.1.S</td>\n",
       "      <td>XB-GENE-5960855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771885</th>\n",
       "      <td>rnf112.S</td>\n",
       "      <td>XB-GENE-17336033</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19286 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_name           Xenbase\n",
       "192            vdac2.S    XB-GENE-954844\n",
       "212            samd8.S  XB-GENE-17343426\n",
       "227           dusp13.S  XB-GENE-17337763\n",
       "258              adk.S    XB-GENE-997231\n",
       "281              vcl.S   XB-GENE-5759162\n",
       "...                ...               ...\n",
       "771494        nif3l1.S   XB-GENE-6078750\n",
       "771661         tpsg1.S   XB-GENE-5893017\n",
       "771750  loc100491461.S   XB-GENE-6489054\n",
       "771854      rnf112.1.S   XB-GENE-5960855\n",
       "771885        rnf112.S  XB-GENE-17336033\n",
       "\n",
       "[19286 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dbxref_datafield = ''\n",
    "datafield = 'Xenbase'\n",
    "\n",
    "if dbxref_datafield != '':\n",
    "    models.dropna(axis = 0, subset = ['dbxref_dict'], inplace = True)\n",
    "    models[dbxref_datafield] = [d.get(dbxref_datafield) for d in models['dbxref_dict']]\n",
    "\n",
    "    models_subset = models[['gene_name', dbxref_datafield]].dropna().drop_duplicates()\n",
    "    display(models_subset)\n",
    "\n",
    "elif datafield == 'gene_name':\n",
    "    models_subset = models[['gene_name']].dropna().drop_duplicates()\n",
    "    display(models_subset)\n",
    "\n",
    "elif datafield != '':\n",
    "    models_subset = models[['gene_name', datafield]].dropna().drop_duplicates()\n",
    "    display(models_subset)\n",
    "\n",
    "else:\n",
    "    raise Exception('You must provide a data field for ID mapping.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Generate gene list file to query Uniprot ID Mapping API\n",
    "Generate a text file ending in `_ids.txt` for sending to the ID mapping API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 19285 gene ids to /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_Xenbase_ids.txt\n"
     ]
    }
   ],
   "source": [
    "datafield = dbxref_datafield if dbxref_datafield != '' else datafield\n",
    "\n",
    "gene_list = models_subset[datafield].unique()\n",
    "\n",
    "genelist_object = GeneListFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    sources = [sample_BFD.annot],\n",
    "    genes = gene_list,\n",
    "    identifier = datafield\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Query Uniprot ID Mapping API\n",
    "Specify the `from_type` variable based on the Uniprot name of the identifier.  \n",
    "The table below lists some databases and the `from_type` string that the API accepts for that datatype.  \n",
    "\n",
    "| datatype | `from_type` string | description |\n",
    "| ---: | :--- | :--- |\n",
    "| Mouse Genome Informatics | `MGI` | ID starts with `MGI:` |\n",
    "| Zebrafish Information Network | `ZFIN` | ID starts with `ZDB-GENE-` |\n",
    "| Xenbase | `Xenbase` | ID starts with `XB-GENE-` |\n",
    "\n",
    "__NOTE:__ You may have to run the cell below twice - UniProt sometimes throws an \"Resource not found\" message on the first query to the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  305k    0    52  100  305k     20   118k  0:00:02  0:00:02 --:--:--  118k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100    24    0    24    0     0     28      0 --:--:-- --:--:-- --:--:--    28\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>From</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XB-GENE-954844</td>\n",
       "      <td>Q52KY1</td>\n",
       "      <td>Q52KY1_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Voltage-dependent anion-selective channel prot...</td>\n",
       "      <td>vdac2.S vdac2 vdac2.L</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XB-GENE-997231</td>\n",
       "      <td>Q6DJM1</td>\n",
       "      <td>Q6DJM1_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Adenosine kinase (AK) (EC 2.7.1.20) (Adenosine...</td>\n",
       "      <td>adk.S adk MGC82032</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XB-GENE-5910051</td>\n",
       "      <td>Q6GM53</td>\n",
       "      <td>IKKA_XENLA</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Inhibitor of nuclear factor kappa-B kinase sub...</td>\n",
       "      <td>chuk ikka</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XB-GENE-958583</td>\n",
       "      <td>Q641H0</td>\n",
       "      <td>Q641H0_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Annexin</td>\n",
       "      <td>anxa11.S anx11 anxa11 cap50 MGC81584</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XB-GENE-6251658</td>\n",
       "      <td>Q5FWP7</td>\n",
       "      <td>Q5FWP7_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>MGC84996 protein (nuclear receptor coactivator...</td>\n",
       "      <td>ncoa4.S ara70 ele1 MGC84996 ncoa4 ncoa4-b ptc3...</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17505</th>\n",
       "      <td>XB-GENE-6078750</td>\n",
       "      <td>A0A1L8ERE3</td>\n",
       "      <td>A0A1L8ERE3_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>NIF3-like protein 1</td>\n",
       "      <td>nif3l1.S nif3l1 nif3l1.L</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17506</th>\n",
       "      <td>XB-GENE-6078750</td>\n",
       "      <td>Q0IHC9</td>\n",
       "      <td>Q0IHC9_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>NIF3-like protein 1</td>\n",
       "      <td>nif3l1.S MGC154449 nif3l1 nif3l1.L</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17507</th>\n",
       "      <td>XB-GENE-5893017</td>\n",
       "      <td>A0A1L8ERF3</td>\n",
       "      <td>A0A1L8ERF3_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>serine protease 27</td>\n",
       "      <td>tpsg1.S</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17508</th>\n",
       "      <td>XB-GENE-5960855</td>\n",
       "      <td>A0A1L8ERI6</td>\n",
       "      <td>A0A1L8ERI6_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>RING finger protein 112</td>\n",
       "      <td>rnf112.1.S</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17509</th>\n",
       "      <td>XB-GENE-17336033</td>\n",
       "      <td>A0A1L8ERH8</td>\n",
       "      <td>A0A1L8ERH8_XENLA</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>RING finger protein 112</td>\n",
       "      <td>rnf112.S</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "      <td>481</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17510 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   From       Entry        Entry Name    Reviewed  \\\n",
       "0        XB-GENE-954844      Q52KY1      Q52KY1_XENLA  unreviewed   \n",
       "1        XB-GENE-997231      Q6DJM1      Q6DJM1_XENLA  unreviewed   \n",
       "2       XB-GENE-5910051      Q6GM53        IKKA_XENLA    reviewed   \n",
       "3        XB-GENE-958583      Q641H0      Q641H0_XENLA  unreviewed   \n",
       "4       XB-GENE-6251658      Q5FWP7      Q5FWP7_XENLA  unreviewed   \n",
       "...                 ...         ...               ...         ...   \n",
       "17505   XB-GENE-6078750  A0A1L8ERE3  A0A1L8ERE3_XENLA  unreviewed   \n",
       "17506   XB-GENE-6078750      Q0IHC9      Q0IHC9_XENLA  unreviewed   \n",
       "17507   XB-GENE-5893017  A0A1L8ERF3  A0A1L8ERF3_XENLA  unreviewed   \n",
       "17508   XB-GENE-5960855  A0A1L8ERI6  A0A1L8ERI6_XENLA  unreviewed   \n",
       "17509  XB-GENE-17336033  A0A1L8ERH8  A0A1L8ERH8_XENLA  unreviewed   \n",
       "\n",
       "                                           Protein names  \\\n",
       "0      Voltage-dependent anion-selective channel prot...   \n",
       "1      Adenosine kinase (AK) (EC 2.7.1.20) (Adenosine...   \n",
       "2      Inhibitor of nuclear factor kappa-B kinase sub...   \n",
       "3                                                Annexin   \n",
       "4      MGC84996 protein (nuclear receptor coactivator...   \n",
       "...                                                  ...   \n",
       "17505                                NIF3-like protein 1   \n",
       "17506                                NIF3-like protein 1   \n",
       "17507                                 serine protease 27   \n",
       "17508                            RING finger protein 112   \n",
       "17509                            RING finger protein 112   \n",
       "\n",
       "                                              Gene Names  \\\n",
       "0                                  vdac2.S vdac2 vdac2.L   \n",
       "1                                     adk.S adk MGC82032   \n",
       "2                                              chuk ikka   \n",
       "3                   anxa11.S anx11 anxa11 cap50 MGC81584   \n",
       "4      ncoa4.S ara70 ele1 MGC84996 ncoa4 ncoa4-b ptc3...   \n",
       "...                                                  ...   \n",
       "17505                           nif3l1.S nif3l1 nif3l1.L   \n",
       "17506                 nif3l1.S MGC154449 nif3l1 nif3l1.L   \n",
       "17507                                            tpsg1.S   \n",
       "17508                                         rnf112.1.S   \n",
       "17509                                           rnf112.S   \n",
       "\n",
       "                                   Organism  Length  \n",
       "0      Xenopus laevis (African clawed frog)     283  \n",
       "1      Xenopus laevis (African clawed frog)     361  \n",
       "2      Xenopus laevis (African clawed frog)     743  \n",
       "3      Xenopus laevis (African clawed frog)     502  \n",
       "4      Xenopus laevis (African clawed frog)     625  \n",
       "...                                     ...     ...  \n",
       "17505  Xenopus laevis (African clawed frog)     369  \n",
       "17506  Xenopus laevis (African clawed frog)     344  \n",
       "17507  Xenopus laevis (African clawed frog)     280  \n",
       "17508  Xenopus laevis (African clawed frog)     625  \n",
       "17509  Xenopus laevis (African clawed frog)     481  \n",
       "\n",
       "[17510 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from_type = 'Xenbase'\n",
    "to_type = 'UniProtKB'\n",
    "\n",
    "uniprot_idmm_object = genelist_object.get_uniprot_ids(ID_MAPPER_LOC, from_type, to_type)\n",
    "sample_BFD.add_keyfile('uniprot_idmm', uniprot_idmm_object, overwrite = True)\n",
    "\n",
    "uniprot_idmm = pd.read_csv(sample_BFD.uniprot_idmm.path, sep = '\\t')\n",
    "display(uniprot_idmm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Extract results and generate Uniprot IDMM\n",
    "Generates an idmm that links `gene_name`, the `dbxref_datafield` seleted above, and `uniprot_id` returned by API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_22280/899271637.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  uniprot_idpairs.rename(columns = {'From': datafield, 'Entry': 'uniprot_id'}, inplace = True)\n",
      "/tmp/ipykernel_22280/899271637.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  uniprot_idpairs[datafield] = uniprot_idpairs[datafield].astype(str)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Xenbase</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XB-GENE-954844</td>\n",
       "      <td>Q52KY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XB-GENE-997231</td>\n",
       "      <td>Q6DJM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XB-GENE-5910051</td>\n",
       "      <td>Q6GM53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XB-GENE-958583</td>\n",
       "      <td>Q641H0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XB-GENE-6251658</td>\n",
       "      <td>Q5FWP7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17505</th>\n",
       "      <td>XB-GENE-6078750</td>\n",
       "      <td>A0A1L8ERE3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17506</th>\n",
       "      <td>XB-GENE-6078750</td>\n",
       "      <td>Q0IHC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17507</th>\n",
       "      <td>XB-GENE-5893017</td>\n",
       "      <td>A0A1L8ERF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17508</th>\n",
       "      <td>XB-GENE-5960855</td>\n",
       "      <td>A0A1L8ERI6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17509</th>\n",
       "      <td>XB-GENE-17336033</td>\n",
       "      <td>A0A1L8ERH8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17510 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Xenbase  uniprot_id\n",
       "0        XB-GENE-954844      Q52KY1\n",
       "1        XB-GENE-997231      Q6DJM1\n",
       "2       XB-GENE-5910051      Q6GM53\n",
       "3        XB-GENE-958583      Q641H0\n",
       "4       XB-GENE-6251658      Q5FWP7\n",
       "...                 ...         ...\n",
       "17505   XB-GENE-6078750  A0A1L8ERE3\n",
       "17506   XB-GENE-6078750      Q0IHC9\n",
       "17507   XB-GENE-5893017  A0A1L8ERF3\n",
       "17508   XB-GENE-5960855  A0A1L8ERI6\n",
       "17509  XB-GENE-17336033  A0A1L8ERH8\n",
       "\n",
       "[17510 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Xenbase</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>vdac2.S</td>\n",
       "      <td>XB-GENE-954844</td>\n",
       "      <td>Q52KY1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>adk.S</td>\n",
       "      <td>XB-GENE-997231</td>\n",
       "      <td>Q6DJM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chuk.S</td>\n",
       "      <td>XB-GENE-5910051</td>\n",
       "      <td>Q6GM53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>anxa11.S</td>\n",
       "      <td>XB-GENE-958583</td>\n",
       "      <td>Q641H0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ncoa4.S</td>\n",
       "      <td>XB-GENE-6251658</td>\n",
       "      <td>Q5FWP7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17474</th>\n",
       "      <td>nif3l1.S</td>\n",
       "      <td>XB-GENE-6078750</td>\n",
       "      <td>A0A1L8ERE3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17475</th>\n",
       "      <td>nif3l1.S</td>\n",
       "      <td>XB-GENE-6078750</td>\n",
       "      <td>Q0IHC9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17476</th>\n",
       "      <td>tpsg1.S</td>\n",
       "      <td>XB-GENE-5893017</td>\n",
       "      <td>A0A1L8ERF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17477</th>\n",
       "      <td>rnf112.1.S</td>\n",
       "      <td>XB-GENE-5960855</td>\n",
       "      <td>A0A1L8ERI6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17478</th>\n",
       "      <td>rnf112.S</td>\n",
       "      <td>XB-GENE-17336033</td>\n",
       "      <td>A0A1L8ERH8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17479 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name           Xenbase  uniprot_id\n",
       "0         vdac2.S    XB-GENE-954844      Q52KY1\n",
       "1           adk.S    XB-GENE-997231      Q6DJM1\n",
       "2          chuk.S   XB-GENE-5910051      Q6GM53\n",
       "3        anxa11.S    XB-GENE-958583      Q641H0\n",
       "4         ncoa4.S   XB-GENE-6251658      Q5FWP7\n",
       "...           ...               ...         ...\n",
       "17474    nif3l1.S   XB-GENE-6078750  A0A1L8ERE3\n",
       "17475    nif3l1.S   XB-GENE-6078750      Q0IHC9\n",
       "17476     tpsg1.S   XB-GENE-5893017  A0A1L8ERF3\n",
       "17477  rnf112.1.S   XB-GENE-5960855  A0A1L8ERI6\n",
       "17478    rnf112.S  XB-GENE-17336033  A0A1L8ERH8\n",
       "\n",
       "[17479 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "uniprot_idpairs = uniprot_idmm[['From', 'Entry']]\n",
    "uniprot_idpairs.rename(columns = {'From': datafield, 'Entry': 'uniprot_id'}, inplace = True)\n",
    "uniprot_idpairs[datafield] = uniprot_idpairs[datafield].astype(str)\n",
    "display(uniprot_idpairs)\n",
    "\n",
    "uniprot_output_idmm = models_subset.merge(uniprot_idpairs, on = datafield)\n",
    "display(uniprot_output_idmm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "key \"uniprot_idmm\" already exists, ignoring\n"
     ]
    }
   ],
   "source": [
    "# generate a filename and file for the idmm\n",
    "uniprot_output_idmm_filename = '_'.join([sample_BFD.species_prefix, conditions, 'uniprot-idmm.tsv'])\n",
    "uniprot_output_idmm_object = IdmmFile(\n",
    "    filename = uniprot_output_idmm_filename, \n",
    "    sampledict = sample_BFD.sampledict, \n",
    "    kind = 'uniprot_idmm', \n",
    "    sources = [sample_BFD.annot]\n",
    ")\n",
    "\n",
    "# save to file and add to the BioFileDocket\n",
    "uniprot_output_idmm.to_csv(uniprot_output_idmm_object.path, sep = '\\t')\n",
    "sample_BFD.add_keyfile('uniprot_idmm', uniprot_output_idmm_object)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9. Generate gtf-idmm\n",
    "This file maps the `gene_name` to `gene_id` and `transcript_id` fields generated by the conversion from GFF to GTF, which will be needed for downstream processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2136</td>\n",
       "      <td>2204</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2205</td>\n",
       "      <td>3023</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3024</td>\n",
       "      <td>3092</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3093</td>\n",
       "      <td>4723</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>4724</td>\n",
       "      <td>4798</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771995</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104527008</td>\n",
       "      <td>104527068</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771996</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771997</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771998</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771999</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>772000 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0        1     2          3          4  5  6  7  \\\n",
       "0             MT  Xenbase  exon       2136       2204  .  +  .   \n",
       "1             MT  Xenbase  exon       2205       3023  .  +  .   \n",
       "2             MT  Xenbase  exon       3024       3092  .  +  .   \n",
       "3             MT  Xenbase  exon       3093       4723  .  +  .   \n",
       "4             MT  Xenbase  exon       4724       4798  .  +  .   \n",
       "...          ...      ...   ...        ...        ... .. .. ..   \n",
       "771995  chr9_10S  Xenbase   CDS  104527008  104527068  .  -  0   \n",
       "771996  chr9_10S  Xenbase  exon  104530653  104530753  .  -  2   \n",
       "771997  chr9_10S  Xenbase   CDS  104530653  104530753  .  -  2   \n",
       "771998  chr9_10S  Xenbase  exon  104531027  104531054  .  -  0   \n",
       "771999  chr9_10S  Xenbase   CDS  104531027  104531054  .  -  0   \n",
       "\n",
       "                                                        8  \n",
       "0       gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...  \n",
       "1       gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...  \n",
       "2       gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...  \n",
       "3       gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...  \n",
       "4       gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...  \n",
       "...                                                   ...  \n",
       "771995  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771996  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771997  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771998  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "771999  gene_id \"gene24857\"; gene_name \"Xetrov90026942...  \n",
       "\n",
       "[772000 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; transcript_id \"gene34778_t\"; transcript_name \"mt-rnr1.L\"; Xenbase \"nan\"'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "models_asgtf_df = pd.read_csv(models_asgtf_updated.path, sep = '\\t', header = None)\n",
    "\n",
    "display(models_asgtf_df)\n",
    "display(models_asgtf_df[attributes_column][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2136</td>\n",
       "      <td>2204</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...</td>\n",
       "      <td>{'gene_id': 'gene42065', 'gene_name': 'mt-trna...</td>\n",
       "      <td>mt-trna-phe.L</td>\n",
       "      <td>gene42065</td>\n",
       "      <td>gene42065_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>2205</td>\n",
       "      <td>3023</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...</td>\n",
       "      <td>{'gene_id': 'gene34778', 'gene_name': 'mt-rnr1...</td>\n",
       "      <td>mt-rnr1.L</td>\n",
       "      <td>gene34778</td>\n",
       "      <td>gene34778_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3024</td>\n",
       "      <td>3092</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...</td>\n",
       "      <td>{'gene_id': 'gene48202', 'gene_name': 'mt-trna...</td>\n",
       "      <td>mt-trna-val.L</td>\n",
       "      <td>gene48202</td>\n",
       "      <td>gene48202_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>3093</td>\n",
       "      <td>4723</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...</td>\n",
       "      <td>{'gene_id': 'gene44770', 'gene_name': 'mt-rnr2...</td>\n",
       "      <td>mt-rnr2.L</td>\n",
       "      <td>gene44770</td>\n",
       "      <td>gene44770_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MT</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>4724</td>\n",
       "      <td>4798</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...</td>\n",
       "      <td>{'gene_id': 'gene43253', 'gene_name': 'mt-trna...</td>\n",
       "      <td>mt-trna-leu1.L</td>\n",
       "      <td>gene43253</td>\n",
       "      <td>gene43253_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771990</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104524309</td>\n",
       "      <td>104524351</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "      <td>gene24857</td>\n",
       "      <td>rna64354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771992</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104525747</td>\n",
       "      <td>104525868</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "      <td>gene24857</td>\n",
       "      <td>rna64354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771994</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104527008</td>\n",
       "      <td>104527068</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "      <td>gene24857</td>\n",
       "      <td>rna64354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771996</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104530653</td>\n",
       "      <td>104530753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "      <td>gene24857</td>\n",
       "      <td>rna64354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771998</th>\n",
       "      <td>chr9_10S</td>\n",
       "      <td>Xenbase</td>\n",
       "      <td>exon</td>\n",
       "      <td>104531027</td>\n",
       "      <td>104531054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"gene24857\"; gene_name \"Xetrov90026942...</td>\n",
       "      <td>{'gene_id': 'gene24857', 'gene_name': 'Xetrov9...</td>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "      <td>gene24857</td>\n",
       "      <td>rna64354</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>453943 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               0        1     2          3          4  5  6  7  \\\n",
       "0             MT  Xenbase  exon       2136       2204  .  +  .   \n",
       "1             MT  Xenbase  exon       2205       3023  .  +  .   \n",
       "2             MT  Xenbase  exon       3024       3092  .  +  .   \n",
       "3             MT  Xenbase  exon       3093       4723  .  +  .   \n",
       "4             MT  Xenbase  exon       4724       4798  .  +  .   \n",
       "...          ...      ...   ...        ...        ... .. .. ..   \n",
       "771990  chr9_10S  Xenbase  exon  104524309  104524351  .  -  0   \n",
       "771992  chr9_10S  Xenbase  exon  104525747  104525868  .  -  2   \n",
       "771994  chr9_10S  Xenbase  exon  104527008  104527068  .  -  0   \n",
       "771996  chr9_10S  Xenbase  exon  104530653  104530753  .  -  2   \n",
       "771998  chr9_10S  Xenbase  exon  104531027  104531054  .  -  0   \n",
       "\n",
       "                                                        8  \\\n",
       "0       gene_id \"gene42065\"; gene_name \"mt-trna-phe.L\"...   \n",
       "1       gene_id \"gene34778\"; gene_name \"mt-rnr1.L\"; tr...   \n",
       "2       gene_id \"gene48202\"; gene_name \"mt-trna-val.L\"...   \n",
       "3       gene_id \"gene44770\"; gene_name \"mt-rnr2.L\"; tr...   \n",
       "4       gene_id \"gene43253\"; gene_name \"mt-trna-leu1.L...   \n",
       "...                                                   ...   \n",
       "771990  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "771992  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "771994  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "771996  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "771998  gene_id \"gene24857\"; gene_name \"Xetrov90026942...   \n",
       "\n",
       "                                         field_dictionary          gene_name  \\\n",
       "0       {'gene_id': 'gene42065', 'gene_name': 'mt-trna...      mt-trna-phe.L   \n",
       "1       {'gene_id': 'gene34778', 'gene_name': 'mt-rnr1...          mt-rnr1.L   \n",
       "2       {'gene_id': 'gene48202', 'gene_name': 'mt-trna...      mt-trna-val.L   \n",
       "3       {'gene_id': 'gene44770', 'gene_name': 'mt-rnr2...          mt-rnr2.L   \n",
       "4       {'gene_id': 'gene43253', 'gene_name': 'mt-trna...     mt-trna-leu1.L   \n",
       "...                                                   ...                ...   \n",
       "771990  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S   \n",
       "771992  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S   \n",
       "771994  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S   \n",
       "771996  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S   \n",
       "771998  {'gene_id': 'gene24857', 'gene_name': 'Xetrov9...  Xetrov90026942m.S   \n",
       "\n",
       "          gene_id transcript_id  \n",
       "0       gene42065   gene42065_t  \n",
       "1       gene34778   gene34778_t  \n",
       "2       gene48202   gene48202_t  \n",
       "3       gene44770   gene44770_t  \n",
       "4       gene43253   gene43253_t  \n",
       "...           ...           ...  \n",
       "771990  gene24857      rna64354  \n",
       "771992  gene24857      rna64354  \n",
       "771994  gene24857      rna64354  \n",
       "771996  gene24857      rna64354  \n",
       "771998  gene24857      rna64354  \n",
       "\n",
       "[453943 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Use a custom function to extract useful fields from the additional fields section (column 8)\n",
    "# Pull from that dict to fill in additional useful columns\n",
    "models_asgtf_df['field_dictionary'] = models_asgtf_df[attributes_column].apply(convert_fields_to_dict_gtf)\n",
    "models_asgtf_df['gene_name'] = [d.get('gene_name') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['gene_id'] = [d.get('gene_id') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['transcript_id'] = [d.get('transcript_id') for d in models_asgtf_df['field_dictionary']]\n",
    "\n",
    "# Remove CDS annotations because they interfere with TransDecoder cDNA generation\n",
    "models_asgtf_df = models_asgtf_df[models_asgtf_df[2] != 'CDS']\n",
    "display(models_asgtf_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mt-trna-phe.L</td>\n",
       "      <td>gene42065</td>\n",
       "      <td>gene42065_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mt-rnr1.L</td>\n",
       "      <td>gene34778</td>\n",
       "      <td>gene34778_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mt-trna-val.L</td>\n",
       "      <td>gene48202</td>\n",
       "      <td>gene48202_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mt-rnr2.L</td>\n",
       "      <td>gene44770</td>\n",
       "      <td>gene44770_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mt-trna-leu1.L</td>\n",
       "      <td>gene43253</td>\n",
       "      <td>gene43253_t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771955</th>\n",
       "      <td>vkorc1.S</td>\n",
       "      <td>gene13974</td>\n",
       "      <td>rna40672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771956</th>\n",
       "      <td>vkorc1.S</td>\n",
       "      <td>gene13974</td>\n",
       "      <td>rna40673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771962</th>\n",
       "      <td>LOC108703288</td>\n",
       "      <td>gene37157</td>\n",
       "      <td>rna77797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771977</th>\n",
       "      <td>LOC108702356 [provisional:urgcp]</td>\n",
       "      <td>gene34762</td>\n",
       "      <td>rna74435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771984</th>\n",
       "      <td>Xetrov90026942m.S</td>\n",
       "      <td>gene24857</td>\n",
       "      <td>rna64354</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88005 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               gene_name    gene_id transcript_id\n",
       "0                          mt-trna-phe.L  gene42065   gene42065_t\n",
       "1                              mt-rnr1.L  gene34778   gene34778_t\n",
       "2                          mt-trna-val.L  gene48202   gene48202_t\n",
       "3                              mt-rnr2.L  gene44770   gene44770_t\n",
       "4                         mt-trna-leu1.L  gene43253   gene43253_t\n",
       "...                                  ...        ...           ...\n",
       "771955                          vkorc1.S  gene13974      rna40672\n",
       "771956                          vkorc1.S  gene13974      rna40673\n",
       "771962                      LOC108703288  gene37157      rna77797\n",
       "771977  LOC108702356 [provisional:urgcp]  gene34762      rna74435\n",
       "771984                 Xetrov90026942m.S  gene24857      rna64354\n",
       "\n",
       "[88005 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract gene_name, gene_id, and transcript_id fields to generate an ID mapping matrix (idmm)\n",
    "idmm_df = models_asgtf_df[['gene_name', 'gene_id', 'transcript_id']].drop_duplicates()\n",
    "idmm_df.dropna(inplace = True)\n",
    "display(idmm_df)\n",
    "\n",
    "# generate a filename and file for the idmm\n",
    "idmm_filename = '_'.join([sample_BFD.species_prefix, conditions, 'gtf-idmm.tsv'])\n",
    "idmm = IdmmFile(\n",
    "    filename = idmm_filename, \n",
    "    sampledict = sample_BFD.sampledict, \n",
    "    kind = 'gtf_idmm', \n",
    "    sources = [sample_BFD.annot]\n",
    ")\n",
    "\n",
    "# save to file and add to the BioFileDocket\n",
    "idmm_df.to_csv(idmm.path, sep = '\\t')\n",
    "sample_BFD.add_keyfile('gtf_idmm', idmm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 11. Generate cDNA and peptide files\n",
    "Using the updated gtf file and genome file, generate cDNA sequence.\n",
    "\n",
    "Then, using the cDNA sequence, generate peptide sequences using transdecoder.\n",
    "\n",
    "Expect this step to take some time, probably ~20-30min."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/compute_base_probs.pl /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa 0 > /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//base_freqs.dat, checkpoint [/home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir/.__checkpoints_longorfs/base_freqs_file.ok] exists.\n",
      "-skipping long orf extraction, already completed earlier as per checkpoint: /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir/.__checkpoints_longorfs/TD.longorfs.ok\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds 5000 5000 > /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000, checkpoint [/home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir/.__checkpoints/get_longest_orfs.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/exclude_similar_proteins.pl /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000 > /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr, checkpoint [/home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir/.__checkpoints/nr.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr 500 > /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest, checkpoint [/home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir/.__checkpoints/top_train_select.ok] exists.\n",
      "PCT_GC: 43.6\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/seq_n_baseprobs_to_loglikelihood_vals.pl /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//base_freqs.dat > /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//hexamer.scores, checkpoint [/home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir/.__checkpoints/hexamer_scores.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/score_CDS_likelihood_all_6_frames.pl /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//hexamer.scores > /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.scores, checkpoint [/home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir/.__checkpoints/score_cdss.ok] exists.\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/select_best_ORFs_per_transcript.pl --gff3_file /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.gff3 --cds_scores /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.scores  --min_length_auto_accept 645  > /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 \n",
      "Selecting best orfs\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/train_start_PWM.pl --transcripts /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa --selected_orfs /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest --out_prefix /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement\n",
      "Training start codon pattern recognition* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/build_atgPWM_+-.pl  --transcripts /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa  --selected_orfs /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest  --out_prefix /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement --pwm_left 20 --pwm_right 10 \n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scoring.+-.pl  --features_plus /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.+.features  --features_minus /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.-.features  --atg_position 20  > /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.feature.scores\n",
      "-round: 1\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "-round: 2\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "-round: 3\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "-round: 4\n",
      "-round: 5\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scores_to_ROC.pl /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.feature.scores > /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.feature.scores.roc\n",
      "-parsing scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/plot_ROC.Rscript /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.feature.scores.roc || :\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/compute_AUC.pl /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.feature.scores.roc\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.+.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.-.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/deplete_feature_noise.pl  --features_plus /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.+.features  --pwm_minus /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.-.pwm  --out_prefix /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced\n",
      "num features: 394\tnum_incorporate: 118\n",
      "-feature swap of score: 0.282211042182629 instead of -2.10559846425201\n",
      "-feature swap of score: -1.16188308206509 instead of -2.28119328826824\n",
      "-feature swap of score: -1.36349372230503 instead of -2.37308013140263\n",
      "-feature swap of score: -1.5288154115419 instead of -2.38633864247541\n",
      "-feature swap of score: 1.30080256968223 instead of -2.12562639345566\n",
      "-feature swap of score: 2.58302338351591 instead of -2.1542417313035\n",
      "-feature swap of score: 1.03923917311702 instead of -2.15704712686731\n",
      "-feature swap of score: -0.759392418234196 instead of -1.96604994606952\n",
      "-feature swap of score: 1.19952871541642 instead of -2.0819275811987\n",
      "-feature swap of score: 1.62853952810211 instead of -1.79898664301526\n",
      "-feature swap of score: 1.82999954603619 instead of -1.60093084783994\n",
      "-feature swap of score: -0.409880358663875 instead of -1.48921771076515\n",
      "-feature swap of score: 2.14172589070934 instead of -1.50611235524916\n",
      "-feature swap of score: -1.27774293922662 instead of -1.35177364813391\n",
      "-feature swap of score: 0.406226976848686 instead of -1.28510457319706\n",
      "-feature swap of score: 0.441226568709701 instead of -1.25373972198834\n",
      "-feature swap of score: 0.626491054257062 instead of -1.26927450741765\n",
      "-feature swap of score: 1.93790701418717 instead of -1.2201140845988\n",
      "-feature swap of score: -0.109984239857404 instead of -1.08721029604963\n",
      "-feature swap of score: 2.18519899500483 instead of -1.20033734228051\n",
      "-feature swap of score: -0.282977604136504 instead of -1.42097881097281\n",
      "-feature swap of score: 1.60700240266365 instead of -1.17990234296841\n",
      "-feature swap of score: 0.280403271850439 instead of -1.25025464775653\n",
      "-feature swap of score: -0.0316551469448385 instead of -1.25575204194636\n",
      "-feature swap of score: -0.0414662588933564 instead of -0.927727164090748\n",
      "-feature swap of score: 2.58354495247625 instead of -0.871376023819319\n",
      "-feature swap of score: 2.42482128667312 instead of -0.638047212212269\n",
      "-feature swap of score: 0.802272207381882 instead of -0.553378558509306\n",
      "-feature swap of score: 1.96637774623844 instead of -0.591697075919005\n",
      "-feature swap of score: 2.50089740026352 instead of -0.614720806116368\n",
      "-feature swap of score: 1.31547905507288 instead of -0.681073940143409\n",
      "-feature swap of score: 2.0503127549551 instead of -0.800074526921327\n",
      "-feature swap of score: -0.540794902377789 instead of -0.943146284757201\n",
      "-feature swap of score: 3.21936153880605 instead of -0.529303620805585\n",
      "-feature swap of score: -0.280206070799899 instead of -0.923114463589244\n",
      "-feature swap of score: 0.492333981155279 instead of -0.566978714560265\n",
      "-feature swap of score: -0.0379615068123152 instead of -0.615554486218062\n",
      "-feature swap of score: 1.47384850982698 instead of -0.688945354099117\n",
      "-feature swap of score: 1.54264297230436 instead of -0.603318664781251\n",
      "-feature swap of score: 2.31213297609682 instead of -0.586164448643114\n",
      "-feature swap of score: 1.28399873498245 instead of -0.523392581335056\n",
      "-feature swap of score: 4.32356403148463 instead of -0.457738966643939\n",
      "-feature swap of score: 0.919129208478484 instead of -0.16288645904931\n",
      "-feature swap of score: 0.945008351476351 instead of -0.188167672189052\n",
      "-feature swap of score: 0.521855507732424 instead of 0.00137206864495978\n",
      "-feature swap of score: 2.00909268588197 instead of 0.0508602142857672\n",
      "-feature swap of score: 3.8789868461693 instead of -0.409265424427188\n",
      "-feature swap of score: 3.30482527345489 instead of 0.054441546784111\n",
      "-feature swap of score: 2.89407013561405 instead of -0.259339332745416\n",
      "-feature swap of score: 1.42694817626995 instead of 0.15533258362599\n",
      "-feature swap of score: 2.61829963556781 instead of 0.0577885480580716\n",
      "-feature swap of score: 1.46032640400844 instead of -0.759081159332202\n",
      "-feature swap of score: 3.1987968325987 instead of 0.130406036293526\n",
      "-feature swap of score: 1.64591626217056 instead of 0.136518768048064\n",
      "-feature swap of score: 0.734441609097138 instead of -0.0471976760525726\n",
      "-feature swap of score: 6.26848552992193 instead of 0.0974932784683117\n",
      "-feature swap of score: 0.504067100327014 instead of 0.0981934882213432\n",
      "-feature swap of score: 1.76313046249357 instead of 0.125936857115619\n",
      "-feature swap of score: 2.80856965960322 instead of -0.0580718492674737\n",
      "-feature swap of score: 1.47511676968584 instead of -0.592738023855157\n",
      "-feature swap of score: 0.667598564789021 instead of -0.260516904637941\n",
      "-feature swap of score: 2.29231410789857 instead of 0.216407455611756\n",
      "-feature swap of score: 1.33864385621757 instead of 0.310067355290087\n",
      "-feature swap of score: 3.49114139115248 instead of 0.221778112540442\n",
      "-feature swap of score: 0.823047205017706 instead of 0.147206600999125\n",
      "-feature swap of score: 0.834557697944349 instead of 0.0441167131500077\n",
      "-feature swap of score: 0.916174804261084 instead of 0.237551544024209\n",
      "-feature swap of score: 0.600722186759774 instead of 0.194054492371085\n",
      "-feature swap of score: 3.62798172998081 instead of 0.349873938045235\n",
      "-feature swap of score: 0.848718626512436 instead of 0.500482715303364\n",
      "-feature swap of score: 3.86679779014841 instead of 0.507184679247535\n",
      "-feature swap of score: 3.05415416486409 instead of 0.534081732772071\n",
      "-feature swap of score: 1.79653234145429 instead of 0.472572764296392\n",
      "-feature swap of score: 2.5909757190806 instead of 0.571540252195026\n",
      "-feature swap of score: 2.11006870070995 instead of 0.7831925943114\n",
      "-feature swap of score: 2.06560046334988 instead of 0.642201748126931\n",
      "-feature swap of score: 3.84106436968357 instead of 0.564815099178643\n",
      "-feature swap of score: 1.8304233027878 instead of 0.531740663218364\n",
      "-feature swap of score: 3.62935035945454 instead of 0.425594374622966\n",
      "-feature swap of score: 4.05588361061295 instead of 0.875062034749462\n",
      "-feature swap of score: 2.52617220838208 instead of 0.738027466992721\n",
      "-feature swap of score: 2.59648121270437 instead of 0.541739284843936\n",
      "-feature swap of score: 1.82483644465211 instead of 0.674551095987181\n",
      "-feature swap of score: 2.43328755041758 instead of 0.853248380154304\n",
      "-feature swap of score: 2.44080616941883 instead of 0.731183495140924\n",
      "-feature swap of score: 1.24916808158317 instead of 0.77213595016404\n",
      "-feature swap of score: 1.55097838423933 instead of 0.850175167366395\n",
      "-feature swap of score: 4.45325473981762 instead of 0.876831010138358\n",
      "-feature swap of score: 2.53336886429454 instead of 0.733810196778028\n",
      "-feature swap of score: 1.11602285022214 instead of 0.887437107244516\n",
      "-feature swap of score: 4.95851072455217 instead of 0.573626073776679\n",
      "-feature swap of score: 2.63879673496215 instead of 0.946222204216221\n",
      "-feature swap of score: 2.90306087537137 instead of 1.0596939471365\n",
      "-feature swap of score: 4.69454726695578 instead of 0.925737879895029\n",
      "-feature swap of score: 1.76748172426744 instead of 0.849640011141634\n",
      "-feature swap of score: 2.88097063597134 instead of 0.853185342403306\n",
      "-feature swap of score: 3.63690266701346 instead of 0.903386371484112\n",
      "-feature swap of score: 2.01489257622004 instead of 1.12181976425328\n",
      "-feature swap of score: 3.00125240686338 instead of 1.20027986303452\n",
      "-feature swap of score: 2.02276500345907 instead of 1.18475632675944\n",
      "-feature swap of score: 0.862602147240493 instead of 0.706180128952248\n",
      "-feature swap of score: 2.07106269219922 instead of 1.08323812424414\n",
      "-feature swap of score: 4.34474298095723 instead of 1.20106403164539\n",
      "-feature swap of score: 6.98863345907748 instead of 1.34672721352409\n",
      "-feature swap of score: 3.02216436983975 instead of 1.29201495888945\n",
      "-feature swap of score: 1.51165236841433 instead of 1.04456769599599\n",
      "-feature swap of score: 3.27124590599817 instead of 1.09904350340881\n",
      "-feature swap of score: 4.64304288914368 instead of 1.17510007153173\n",
      "-num feature swaps: 108\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scoring.+-.pl  --features_plus /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.+.features  --features_minus /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.-.features  --atg_position 20  > /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores\n",
      "-round: 1\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "-round: 2\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "-round: 3\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "-round: 4\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNCACAACCTCCCTTGCATATGGCACACCAAC contains non-GATC chars... skipping\n",
      "-round: 5\n",
      "Error, feature_seq: TAATATTTGGTGGAGTTGAAATGNNNNNNNNNN contains non-GATC chars... skipping\n",
      "Error, feature_seq: NNNNNNNNTATTTAATACCAATGTTTTCAGTGA contains non-GATC chars... skipping\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scores_to_ROC.pl /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores > /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc\n",
      "-parsing scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/plot_ROC.Rscript /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc || :\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/compute_AUC.pl /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//start_refinement.enhanced.+.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/start_codon_refinement.pl --transcripts /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa --gff3_file /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 --workdir /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir/ > /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3\n",
      "Refining start codon selections.\n",
      "-number of revised start positions: 3105\n",
      "* Running CMD: cp /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xenopus_laevis_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3 XENLA_9.2_genome_cDNA.fa.transdecoder.gff3\n",
      "copying output to final output file: XENLA_9.2_genome_cDNA.fa.transdecoder.gff3* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_bed.pl XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 > XENLA_9.2_genome_cDNA.fa.transdecoder.bed\n",
      "Making bed file: XENLA_9.2_genome_cDNA.fa.transdecoder.bed\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 --fasta /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa  --genetic_code Universal > XENLA_9.2_genome_cDNA.fa.transdecoder.pep\n",
      "Making pep file: XENLA_9.2_genome_cDNA.fa.transdecoder.pep\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 --fasta /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa --seqType CDS  --genetic_code Universal > XENLA_9.2_genome_cDNA.fa.transdecoder.cds\n",
      "Making cds file: XENLA_9.2_genome_cDNA.fa.transdecoder.cds\n",
      "transdecoder is finished.  See output files XENLA_9.2_genome_cDNA.fa.transdecoder.*\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cdna = sample_BFD.genome_fasta.get_transdecoder_cdna_gtf(models_asgtf_updated, TRANSDECODER_LOC)\n",
    "sample_BFD.add_keyfile('cdna', cdna)\n",
    "\n",
    "transdecoder_files = sample_BFD.cdna.to_pep_files(TDLONGORF_LOC, TDPREDICT_LOC)\n",
    "sample_BFD.add_keyfiles(transdecoder_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "iEug6TQ6Cc9D"
   },
   "source": [
    "# 12. Push files to AWS S3\n",
    "\n",
    "Iteratively moves through the file_set and file_dict variables and populates files into the right place in AWS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "XENLA_9.2_Xenbase.gtf already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_GCA.gff already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome.fa already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM6214268_Xenopus_brain_COL65_dge.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "upload: ../../output/Xlae_adultbrain/Xlae_adultbrain_Xenbase_UniProtIDs.txt to s3://arcadia-reference-datasets/organisms/Xenopus_laevis/genomics_reference/mapping_file/Xlae_adultbrain_Xenbase_UniProtIDs.txt\n",
      "Xlae_adultbrain_gtf-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.bed already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.cds already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.pep already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n"
     ]
    }
   ],
   "source": [
    "sample_BFD.local_to_s3()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 13. Pickle the `sample_BFD` variable for use by the next script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upload: ../../output/Xlae_adultbrain/Xlae_adultbrain_BioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/Xlae_adultbrain_BioFileDocket.pkl\n"
     ]
    }
   ],
   "source": [
    "# Generate a .pkl file for the Docket\n",
    "sample_BFD.pickle()\n",
    "\n",
    "# Push to S3, optionally overwriting existing file\n",
    "sample_BFD.push_to_s3(overwrite = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "Xtro_bro.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
