{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7387935a-1e42-451e-b862-7e15eb7a0bbd",
   "metadata": {},
   "source": [
    "# Overview\n",
    "\n",
    "This notebook downloads cell annotation from the [Liao et al. 2022](https://www.sciencedirect.com/science/article/pii/S0092867418301168#sec4) Xenopus laevis adult cell atlas for the sample `Brain3_4`, marked in GEO as [\"Xenopus_brain_COL65\"](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM6214268).  \n",
    "This Batch ID was determined by empirically comparing the cell names between datasets.\n",
    "\n",
    "## Guidelines\n",
    "\n",
    "Due to the wide variability in potential formats for these filetypes, it is difficult to generate a standardized analysis pipeline for these files.  \n",
    "For downstream analysis purposes, we'll coerce celltype annotation information into a standard format.\n",
    "\n",
    "The final output generated by this script should be a .tsv file with the following format:\n",
    "\n",
    "| cell_barcode | celltype |\n",
    "|--------------|----------|\n",
    "| ATGCAGATACAC | pyramidal cell |\n",
    "| CCATACAGACTA | neuroblast |\n",
    "| TTTCAAAGACAG | astrocyte |\n",
    "| ... | ... |\n",
    "\n",
    "The final datatype expects two columns: `cell_barcode` and `celltype`.  \n",
    "The `cell_barcode` names should be properly matched to the same format as the `gxc` file.  \n",
    "The `celltype` can have any format, but should include cell identity information.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7684cd5b-58fb-4cd7-bfbb-941ff0d5e867",
   "metadata": {},
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "58f78eca-7ccb-4e60-b6ca-bdc3acb26d7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import os\n",
    "import dill\n",
    "import datetime\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "import sys\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e69c9550-66f7-467d-b626-af08dca0611b",
   "metadata": {},
   "source": [
    "# 1. Load BioFileDocket\n",
    "\n",
    "Load in the BioFileDocket file for this dataset using `BioFileDocket.unpickle()`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8ddbd0a5-f4e0-4321-9e8d-5a12c5bb386b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/glial-origins/output/Xlae_adultbrain/ already exists\n",
      "Files will be saved into /home/ec2-user/glial-origins/output/Xlae_adultbrain/\n",
      "file Xlae_adultbrain_BioFileDocket.pkl already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_BioFileDocket.pkl\n",
      "file XENLA_9.2_Xenbase.gtf already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_Xenbase.gtf\n",
      "file XENLA_9.2_GCA.gff already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_GCA.gff\n",
      "file XENLA_9.2_genome.fa already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome.fa\n",
      "file GSM6214268_Xenopus_brain_COL65_dge.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge.txt\n",
      "file Xlae_adultbrain_Xenbase_UniProtIDs.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_Xenbase_UniProtIDs.txt\n",
      "file Xlae_adultbrain_uniprot-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_uniprot-idmm.tsv\n",
      "file Xlae_adultbrain_gtf-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_gtf-idmm.tsv\n",
      "file XENLA_9.2_genome_cDNA.fa already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.bed already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.bed\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.cds already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.cds\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.gff3\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.pep already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.pep\n",
      "file Xlae_adultbrain_og_idmm.tsv already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_og_idmm.tsv\n",
      "file OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv\n",
      "file GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup.txt\n",
      "file DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv\n",
      "file Xlae_taxid_genes.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_taxid_genes.txt\n",
      "file Xlae_adultbrain_sc_idmm.tsv already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_sc_idmm.tsv\n",
      "file DrerMmusXlae_adultbrain_struclusters_file.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_struclusters_file.tsv\n",
      "file GSM6214268_Xenopus_brain_COL65_dge_asStruCluster.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge_asStruCluster.txt\n",
      "file DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv\n"
     ]
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify the name of the species folder in Amazon S3\n",
    "species = 'Xenopus_laevis'\n",
    "\n",
    "# Specify any particular identifying conditions, eg tissue type:\n",
    "conditions = 'adultbrain'\n",
    "\n",
    "################\n",
    "\n",
    "species_BFD = BioFileDocket(species, conditions).get_from_s3().unpickle()\n",
    "species_BFD.s3_to_local()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d28c6a1-f800-4b67-aa43-c5228ab5fb89",
   "metadata": {},
   "source": [
    "# 2. Download cell annotation file and load in raw file\n",
    "\n",
    "Using `subprocess.run`, download the file using an appropriate protocol and load it as a pandas DataFrame.  \n",
    "For .csv and .tsv files, use `pd.read_csv`.  \n",
    "For .xlsx and .xls files, use `pd.read_excel`.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d604e913-34cd-49ba-bef0-287d0217ac9f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>tSNE_1</th>\n",
       "      <th>tSNE_2</th>\n",
       "      <th>cluster</th>\n",
       "      <th>stage</th>\n",
       "      <th>celltype</th>\n",
       "      <th>cellID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Brain1_3.25.CTGTGTGATCTTGTGGTA</td>\n",
       "      <td>-25.083085</td>\n",
       "      <td>-1.615592</td>\n",
       "      <td>6</td>\n",
       "      <td>Adult</td>\n",
       "      <td>GABAergic neuron_tac1 high</td>\n",
       "      <td>Brain1_3.25.CTGTGTGATCTTGTGGTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Brain1_3.25.ACAATACCTAGAAATAAA</td>\n",
       "      <td>-41.017949</td>\n",
       "      <td>17.559271</td>\n",
       "      <td>2</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Myelinating oligodendrocyte progenitor cell</td>\n",
       "      <td>Brain1_3.25.ACAATACCTAGAAATAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Brain1_3.25.TGCGGATAGAGACCTTTC</td>\n",
       "      <td>-12.190198</td>\n",
       "      <td>-3.974513</td>\n",
       "      <td>19</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Excitatory neuron_zbtb18 high</td>\n",
       "      <td>Brain1_3.25.TGCGGATAGAGACCTTTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Brain1_3.25.GAGGAGATGGCGCGAGTA</td>\n",
       "      <td>-10.871158</td>\n",
       "      <td>-0.184952</td>\n",
       "      <td>19</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Excitatory neuron_zbtb18 high</td>\n",
       "      <td>Brain1_3.25.GAGGAGATGGCGCGAGTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Brain1_3.25.TTAACTAAAGTTAACGCC</td>\n",
       "      <td>-10.116023</td>\n",
       "      <td>-0.098181</td>\n",
       "      <td>19</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Excitatory neuron_zbtb18 high</td>\n",
       "      <td>Brain1_3.25.TTAACTAAAGTTAACGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23901</th>\n",
       "      <td>Brain4_4.25.ATCAACTGAAGCGCGTGC</td>\n",
       "      <td>46.084537</td>\n",
       "      <td>-6.320718</td>\n",
       "      <td>12</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Thyrotroph cell</td>\n",
       "      <td>Brain4_4.25.ATCAACTGAAGCGCGTGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23902</th>\n",
       "      <td>Brain4_4.25.TGCAATGCGTGCAAAGTT</td>\n",
       "      <td>16.626571</td>\n",
       "      <td>9.189624</td>\n",
       "      <td>15</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Growth hormone cell_Rp high</td>\n",
       "      <td>Brain4_4.25.TGCAATGCGTGCAAAGTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23903</th>\n",
       "      <td>Brain4_4.25.GCTCAAAACCTAGCGTGC</td>\n",
       "      <td>-10.361278</td>\n",
       "      <td>23.647850</td>\n",
       "      <td>1</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Gonadotroph cell_cga high</td>\n",
       "      <td>Brain4_4.25.GCTCAAAACCTAGCGTGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23904</th>\n",
       "      <td>Brain4_4.25.TCGTAACGTATTGCGTGC</td>\n",
       "      <td>-14.885559</td>\n",
       "      <td>18.001137</td>\n",
       "      <td>1</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Gonadotroph cell_cga high</td>\n",
       "      <td>Brain4_4.25.TCGTAACGTATTGCGTGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23905</th>\n",
       "      <td>Brain4_4.25.CCGACGCTGAAAGCGTGC</td>\n",
       "      <td>3.604502</td>\n",
       "      <td>38.282519</td>\n",
       "      <td>1</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Gonadotroph cell_cga high</td>\n",
       "      <td>Brain4_4.25.CCGACGCTGAAAGCGTGC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23906 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Unnamed: 0     tSNE_1     tSNE_2  cluster  stage  \\\n",
       "0      Brain1_3.25.CTGTGTGATCTTGTGGTA -25.083085  -1.615592        6  Adult   \n",
       "1      Brain1_3.25.ACAATACCTAGAAATAAA -41.017949  17.559271        2  Adult   \n",
       "2      Brain1_3.25.TGCGGATAGAGACCTTTC -12.190198  -3.974513       19  Adult   \n",
       "3      Brain1_3.25.GAGGAGATGGCGCGAGTA -10.871158  -0.184952       19  Adult   \n",
       "4      Brain1_3.25.TTAACTAAAGTTAACGCC -10.116023  -0.098181       19  Adult   \n",
       "...                               ...        ...        ...      ...    ...   \n",
       "23901  Brain4_4.25.ATCAACTGAAGCGCGTGC  46.084537  -6.320718       12  Adult   \n",
       "23902  Brain4_4.25.TGCAATGCGTGCAAAGTT  16.626571   9.189624       15  Adult   \n",
       "23903  Brain4_4.25.GCTCAAAACCTAGCGTGC -10.361278  23.647850        1  Adult   \n",
       "23904  Brain4_4.25.TCGTAACGTATTGCGTGC -14.885559  18.001137        1  Adult   \n",
       "23905  Brain4_4.25.CCGACGCTGAAAGCGTGC   3.604502  38.282519        1  Adult   \n",
       "\n",
       "                                          celltype  \\\n",
       "0                       GABAergic neuron_tac1 high   \n",
       "1      Myelinating oligodendrocyte progenitor cell   \n",
       "2                    Excitatory neuron_zbtb18 high   \n",
       "3                    Excitatory neuron_zbtb18 high   \n",
       "4                    Excitatory neuron_zbtb18 high   \n",
       "...                                            ...   \n",
       "23901                              Thyrotroph cell   \n",
       "23902                  Growth hormone cell_Rp high   \n",
       "23903                    Gonadotroph cell_cga high   \n",
       "23904                    Gonadotroph cell_cga high   \n",
       "23905                    Gonadotroph cell_cga high   \n",
       "\n",
       "                               cellID  \n",
       "0      Brain1_3.25.CTGTGTGATCTTGTGGTA  \n",
       "1      Brain1_3.25.ACAATACCTAGAAATAAA  \n",
       "2      Brain1_3.25.TGCGGATAGAGACCTTTC  \n",
       "3      Brain1_3.25.GAGGAGATGGCGCGAGTA  \n",
       "4      Brain1_3.25.TTAACTAAAGTTAACGCC  \n",
       "...                               ...  \n",
       "23901  Brain4_4.25.ATCAACTGAAGCGCGTGC  \n",
       "23902  Brain4_4.25.TGCAATGCGTGCAAAGTT  \n",
       "23903  Brain4_4.25.GCTCAAAACCTAGCGTGC  \n",
       "23904  Brain4_4.25.TCGTAACGTATTGCGTGC  \n",
       "23905  Brain4_4.25.CCGACGCTGAAAGCGTGC  \n",
       "\n",
       "[23906 rows x 7 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>AACCTATTCATATAAGGG</th>\n",
       "      <th>CTCGCATCAAAGTTAACT</th>\n",
       "      <th>AACCTAGTATACTTCCGC</th>\n",
       "      <th>AACCTAAAAGTTCTGAAA</th>\n",
       "      <th>CTCGCACGCACCCTCCAT</th>\n",
       "      <th>ACGTTGTATTGTAGCGAG</th>\n",
       "      <th>ACGAGCATGCTTTAGTCG</th>\n",
       "      <th>AACCTAGTCCCGCCATCT</th>\n",
       "      <th>AACCTAGCGAATTAGAGA</th>\n",
       "      <th>...</th>\n",
       "      <th>TCACTTGTTGCCATGCTT</th>\n",
       "      <th>TCGGGTTGTCACACTTAT</th>\n",
       "      <th>TCGTAATCGTAAGTTGCC</th>\n",
       "      <th>TGTCACGAATTACACAAG</th>\n",
       "      <th>TGTGCGTACTTCTAGTCG</th>\n",
       "      <th>TTAACTATACAGTGGATG</th>\n",
       "      <th>TTGGACACTTATGATCTT</th>\n",
       "      <th>AAAGTTACTTATGCCCTC</th>\n",
       "      <th>AACCTACGCACCTGCGGA</th>\n",
       "      <th>AACCTATAGTCGCTGTGT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3.S</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>42Sp43.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>42Sp50.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AK6.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AK6.S</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MGC107841.L</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MGC107851.L</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MGC107876.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MGC108117.L</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MGC108429.L</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 30001 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name  AACCTATTCATATAAGGG  CTCGCATCAAAGTTAACT  AACCTAGTATACTTCCGC  \\\n",
       "0          3.S                   0                   0                   0   \n",
       "1     42Sp43.L                   0                   0                   0   \n",
       "2     42Sp50.L                   0                   0                   0   \n",
       "3        AK6.L                   0                   0                   0   \n",
       "4        AK6.S                   1                   0                   0   \n",
       "5  MGC107841.L                   1                   0                   1   \n",
       "6  MGC107851.L                   0                   1                   0   \n",
       "7  MGC107876.L                   0                   0                   0   \n",
       "8  MGC108117.L                   0                   0                   0   \n",
       "9  MGC108429.L                   0                   1                   3   \n",
       "\n",
       "   AACCTAAAAGTTCTGAAA  CTCGCACGCACCCTCCAT  ACGTTGTATTGTAGCGAG  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   1                   1                   0   \n",
       "\n",
       "   ACGAGCATGCTTTAGTCG  AACCTAGTCCCGCCATCT  AACCTAGCGAATTAGAGA  ...  \\\n",
       "0                   0                   0                   0  ...   \n",
       "1                   0                   0                   0  ...   \n",
       "2                   0                   0                   0  ...   \n",
       "3                   1                   0                   0  ...   \n",
       "4                   0                   0                   0  ...   \n",
       "5                   0                   0                   0  ...   \n",
       "6                   0                   1                   0  ...   \n",
       "7                   0                   1                   0  ...   \n",
       "8                   0                   0                   0  ...   \n",
       "9                   1                   0                   0  ...   \n",
       "\n",
       "   TCACTTGTTGCCATGCTT  TCGGGTTGTCACACTTAT  TCGTAATCGTAAGTTGCC  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   TGTCACGAATTACACAAG  TGTGCGTACTTCTAGTCG  TTAACTATACAGTGGATG  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   TTGGACACTTATGATCTT  AAAGTTACTTATGCCCTC  AACCTACGCACCTGCGGA  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   AACCTATAGTCGCTGTGT  \n",
       "0                   0  \n",
       "1                   0  \n",
       "2                   0  \n",
       "3                   0  \n",
       "4                   0  \n",
       "5                   0  \n",
       "6                   0  \n",
       "7                   0  \n",
       "8                   0  \n",
       "9                   0  \n",
       "\n",
       "[10 rows x 30001 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Specify file url and output information\n",
    "file_url = 'https://figshare.com/ndownloader/files/34026644'\n",
    "output_filename = 'Xlae_cellannot.zip'\n",
    "output_filepath = species_BFD.sampledict.directory + output_filename\n",
    "annot_filepath = species_BFD.sampledict.directory + 'dge_cell_info/' + 'Brain_cell_info.csv'\n",
    "\n",
    "subprocess.run(['wget', file_url, '-O', output_filepath],\n",
    "               stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)\n",
    "subprocess.run(['unzip', output_filepath, '-d', species_BFD.sampledict.directory],\n",
    "               stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)\n",
    "\n",
    "cell_annots = pd.read_csv(annot_filepath)\n",
    "display(cell_annots)\n",
    "\n",
    "# Load in gxc matrix to check proper cell names\n",
    "gxc = pd.read_csv(species_BFD.gxc.path, sep = '\\t', nrows = 10)\n",
    "display(gxc)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "969d8546-a048-458b-8565-0631d95e1e98",
   "metadata": {},
   "source": [
    "# 3. Extract and coerce data into the correct format. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1e1acbb5-5c14-470b-b51f-6b1e368a2b66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>celltype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8238</th>\n",
       "      <td>AACCTAGCGAATTAGAGA</td>\n",
       "      <td>Growth hormone progenitor cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8239</th>\n",
       "      <td>CTGTGTGATCTTCCGCTA</td>\n",
       "      <td>Thyrotroph cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8240</th>\n",
       "      <td>AAGCGGGGTACAATGGCG</td>\n",
       "      <td>GABAergic neuron_tac1 high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8241</th>\n",
       "      <td>CTCGCACGTATTCTGAAA</td>\n",
       "      <td>Myelinating oligodendrocyte progenitor cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8242</th>\n",
       "      <td>CTGTGTTTGGACTATTGT</td>\n",
       "      <td>Myelinating oligodendrocyte progenitor cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16322</th>\n",
       "      <td>CTCGCATGCGGAGCGTGC</td>\n",
       "      <td>Gonadotroph cell_cga high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16323</th>\n",
       "      <td>GAGGAGAGCGAGGCGTGC</td>\n",
       "      <td>Gonadotroph cell_cga high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16324</th>\n",
       "      <td>TCGTAAGCGTGCCTGTGT</td>\n",
       "      <td>Thyrotroph cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16325</th>\n",
       "      <td>TATGTAATACAGGCGTGC</td>\n",
       "      <td>Gonadotroph cell_cga high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16326</th>\n",
       "      <td>ATTCCAGACACTGCGTGC</td>\n",
       "      <td>Gonadotroph cell_cyp21a2.1 high</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8089 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             cell_barcode                                     celltype\n",
       "8238   AACCTAGCGAATTAGAGA               Growth hormone progenitor cell\n",
       "8239   CTGTGTGATCTTCCGCTA                              Thyrotroph cell\n",
       "8240   AAGCGGGGTACAATGGCG                   GABAergic neuron_tac1 high\n",
       "8241   CTCGCACGTATTCTGAAA  Myelinating oligodendrocyte progenitor cell\n",
       "8242   CTGTGTTTGGACTATTGT  Myelinating oligodendrocyte progenitor cell\n",
       "...                   ...                                          ...\n",
       "16322  CTCGCATGCGGAGCGTGC                    Gonadotroph cell_cga high\n",
       "16323  GAGGAGAGCGAGGCGTGC                    Gonadotroph cell_cga high\n",
       "16324  TCGTAAGCGTGCCTGTGT                              Thyrotroph cell\n",
       "16325  TATGTAATACAGGCGTGC                    Gonadotroph cell_cga high\n",
       "16326  ATTCCAGACACTGCGTGC              Gonadotroph cell_cyp21a2.1 high\n",
       "\n",
       "[8089 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Specify sample name for filtering\n",
    "sample_name = 'Brain3_4'\n",
    "\n",
    "# Extract cell barcode and celltype info\n",
    "cell_annots['Batch'] = cell_annots['Unnamed: 0'].str.split('.', expand = True)[0]\n",
    "cell_annots['cell_barcode'] = cell_annots['Unnamed: 0'].str.split('.', expand = True)[2]\n",
    "\n",
    "# Get cells from the Batch matching the sample_name\n",
    "sample_cell_annots = cell_annots[cell_annots['Batch'] == sample_name]\n",
    "sample_cell_annots = sample_cell_annots[['cell_barcode', 'celltype']]\n",
    "display(sample_cell_annots)\n",
    "\n",
    "# Generate filename\n",
    "output_filename = '_'.join([prefixify(species_BFD.species), species_BFD.conditions, sample_name, 'cellannot.tsv'])\n",
    "\n",
    "# Generate BioFile object\n",
    "output_CellAnnotFile = CellAnnotFile(\n",
    "    filename = output_filename,\n",
    "    sampledict = species_BFD.sampledict,\n",
    "    sources = [species_BFD.gxc]\n",
    ")\n",
    "\n",
    "# Export file and add to BioFileDocket\n",
    "sample_cell_annots.to_csv(output_CellAnnotFile.path, sep = '\\t', index = None)\n",
    "species_BFD.add_keyfile('cellannot', output_CellAnnotFile)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ead72526-9415-4023-ba27-a5cc3f23ff6f",
   "metadata": {},
   "source": [
    "# 4. Update BioFileDocket and push files to S3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "709a46f7-256e-4c8f-909f-edaea94620f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "XENLA_9.2_Xenbase.gtf already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_GCA.gff already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome.fa already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM6214268_Xenopus_brain_COL65_dge.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_Xenbase_UniProtIDs.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_uniprot-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_gtf-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.bed already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.cds already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.pep already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_og_idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_taxid_genes.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_sc_idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_struclusters_file.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM6214268_Xenopus_brain_COL65_dge_asStruCluster.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_Brain3_4_cellannot.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_BioFileDocket.pkl already exists in S3 bucket; overwriting.\n",
      "upload: ../../output/Xlae_adultbrain/Xlae_adultbrain_BioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/Xlae_adultbrain_BioFileDocket.pkl\n"
     ]
    }
   ],
   "source": [
    "species_BFD.local_to_s3()\n",
    "species_BFD.pickle()\n",
    "species_BFD.push_to_s3(overwrite = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f82e1eb6-0030-4efd-9013-34362cb83f39",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fe15bdd-e6fa-40d2-b73f-0aaf788a63cd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
