{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview\n",
    "\n",
    "This notebook collects data from the [Raj et al. 2020](https://www.frontiersin.org/articles/10.3389/fcell.2021.743421/full) zebrafish brain developmental cell atlas.\n",
    "\n",
    "#### Dataset description\n",
    "\n",
    "- Data can be found at the [GEO Accession GSE158142](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE158142).\n",
    "- Genome annotation used was at [GEO Accession GSE105010](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE105010)\n",
    "- This notebook previously collects data from one sample, [\"zBr15dpf1_S1\"](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM4793235), for the purposes of data analysis and exploration."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "L56fjA27DmhA"
   },
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "t0BTCsz7Y48y",
    "outputId": "f6510678-f977-4ec8-ce30-1f2acd4fe666"
   },
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess, os, dill, sys\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Download and describe data\n",
    "\n",
    "#### BioFileDocket\n",
    "First, make a BioFileDocket for the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/glial-origins/output/Drer_juvbrain/ already exists\n",
      "Files will be saved into /home/ec2-user/glial-origins/output/Drer_juvbrain/\n"
     ]
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify the name of the species folder in Amazon S3\n",
    "species = 'Danio_rerio'\n",
    "\n",
    "# Specify any particular identifying conditions, eg tissue type:\n",
    "conditions = 'juvbrain'\n",
    "\n",
    "################\n",
    "################\n",
    "\n",
    "sample_BFD = BioFileDocket(species, conditions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inferring file name as GCA_000002035.3_GRCz10_genomic.fna.gz\n",
      "file GCA_000002035.3_GRCz10_genomic.fna.gz already exists at /home/ec2-user/glial-origins/output/Drer_juvbrain/GCA_000002035.3_GRCz10_genomic.fna.gz\n",
      "file GCA_000002035.3_GRCz10_genomic.fna.gz unzipped and object renamed to GCA_000002035.3_GRCz10_genomic.fna\n",
      "inferring file name as GSE105010_Danio_rerio.GRCz10.86.modified.gtf.gz\n",
      "file GSE105010_Danio_rerio.GRCz10.86.modified.gtf.gz already exists at /home/ec2-user/glial-origins/output/Drer_juvbrain/GSE105010_Danio_rerio.GRCz10.86.modified.gtf.gz\n",
      "file GSE105010_Danio_rerio.GRCz10.86.modified.gtf.gz unzipped and object renamed to GSE105010_Danio_rerio.GRCz10.86.modified.gtf\n",
      "inferring file name as GSM4793240_comb-zBr15dpf6_S1-barcodes.tsv.gz\n",
      "file GSM4793240_comb-zBr15dpf6_S1-barcodes.tsv.gz already exists at /home/ec2-user/glial-origins/output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-barcodes.tsv.gz\n",
      "file GSM4793240_comb-zBr15dpf6_S1-barcodes.tsv.gz unzipped and object renamed to GSM4793240_comb-zBr15dpf6_S1-barcodes.tsv\n",
      "inferring file name as GSM4793240_comb-zBr15dpf6_S1-matrix.mtx.gz\n",
      "file GSM4793240_comb-zBr15dpf6_S1-matrix.mtx.gz already exists at /home/ec2-user/glial-origins/output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-matrix.mtx.gz\n",
      "file GSM4793240_comb-zBr15dpf6_S1-matrix.mtx.gz unzipped and object renamed to GSM4793240_comb-zBr15dpf6_S1-matrix.mtx\n",
      "inferring file name as GSM4793240_comb-zBr15dpf6_S1-genes.tsv.gz\n",
      "file GSM4793240_comb-zBr15dpf6_S1-genes.tsv.gz already exists at /home/ec2-user/glial-origins/output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-genes.tsv.gz\n",
      "file GSM4793240_comb-zBr15dpf6_S1-genes.tsv.gz unzipped and object renamed to GSM4793240_comb-zBr15dpf6_S1-genes.tsv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "gzip: /home/ec2-user/glial-origins/output/Drer_juvbrain/GCA_000002035.3_GRCz10_genomic.fna: unknown suffix -- ignored\n",
      "gzip: /home/ec2-user/glial-origins/output/Drer_juvbrain/GSE105010_Danio_rerio.GRCz10.86.modified.gtf: unknown suffix -- ignored\n",
      "gzip: /home/ec2-user/glial-origins/output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-barcodes.tsv: unknown suffix -- ignored\n",
      "gzip: /home/ec2-user/glial-origins/output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-matrix.mtx: unknown suffix -- ignored\n",
      "gzip: /home/ec2-user/glial-origins/output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-genes.tsv: unknown suffix -- ignored\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>AAACCTGAGAAGCCCA-1</th>\n",
       "      <th>AAACCTGAGGATGGAA-1</th>\n",
       "      <th>AAACCTGAGTCTTGCA-1</th>\n",
       "      <th>AAACCTGCACAGACAG-1</th>\n",
       "      <th>AAACCTGCACTTGGAT-1</th>\n",
       "      <th>AAACCTGCATGGGAAC-1</th>\n",
       "      <th>AAACCTGCATTCCTCG-1</th>\n",
       "      <th>AAACCTGGTACATCCA-1</th>\n",
       "      <th>AAACCTGGTGCTGTAT-1</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGTCAAGCGCTTAT-1</th>\n",
       "      <th>TTTGTCAAGCTAAACA-1</th>\n",
       "      <th>TTTGTCACACAAGACG-1</th>\n",
       "      <th>TTTGTCACACGCTTTC-1</th>\n",
       "      <th>TTTGTCACATCCTTGC-1</th>\n",
       "      <th>TTTGTCAGTATAGTAG-1</th>\n",
       "      <th>TTTGTCAGTATATGAG-1</th>\n",
       "      <th>TTTGTCAGTCGACTAT-1</th>\n",
       "      <th>TTTGTCATCATTGCGA-1</th>\n",
       "      <th>TTTGTCATCCTAAGTG-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PARP12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>syn3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ptpro</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>eps8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>si:ch1073-395m9.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tbk1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>gpr19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>crebl2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>dusp16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 5291 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_name  AAACCTGAGAAGCCCA-1  AAACCTGAGGATGGAA-1  \\\n",
       "0   si:ch73-252i11.3                   0                   0   \n",
       "1             PARP12                   0                   0   \n",
       "2               syn3                   0                   0   \n",
       "3              ptpro                   0                   0   \n",
       "4               eps8                   0                   0   \n",
       "5  si:ch1073-395m9.3                   0                   0   \n",
       "6               tbk1                   0                   0   \n",
       "7              gpr19                   0                   0   \n",
       "8             crebl2                   0                   0   \n",
       "9             dusp16                   0                   0   \n",
       "\n",
       "   AAACCTGAGTCTTGCA-1  AAACCTGCACAGACAG-1  AAACCTGCACTTGGAT-1  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   AAACCTGCATGGGAAC-1  AAACCTGCATTCCTCG-1  AAACCTGGTACATCCA-1  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   1   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   AAACCTGGTGCTGTAT-1  ...  TTTGTCAAGCGCTTAT-1  TTTGTCAAGCTAAACA-1  \\\n",
       "0                   0  ...                   0                   0   \n",
       "1                   0  ...                   0                   0   \n",
       "2                   0  ...                   0                   0   \n",
       "3                   0  ...                   0                   0   \n",
       "4                   0  ...                   0                   0   \n",
       "5                   0  ...                   0                   0   \n",
       "6                   0  ...                   0                   0   \n",
       "7                   0  ...                   0                   0   \n",
       "8                   0  ...                   0                   0   \n",
       "9                   0  ...                   0                   0   \n",
       "\n",
       "   TTTGTCACACAAGACG-1  TTTGTCACACGCTTTC-1  TTTGTCACATCCTTGC-1  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   TTTGTCAGTATAGTAG-1  TTTGTCAGTATATGAG-1  TTTGTCAGTCGACTAT-1  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   TTTGTCATCATTGCGA-1  TTTGTCATCCTAAGTG-1  \n",
       "0                   0                   0  \n",
       "1                   0                   0  \n",
       "2                   0                   0  \n",
       "3                   0                   0  \n",
       "4                   0                   0  \n",
       "5                   0                   0  \n",
       "6                   0                   0  \n",
       "7                   0                   0  \n",
       "8                   0                   0  \n",
       "9                   0                   0  \n",
       "\n",
       "[10 rows x 5291 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File already exists at /home/ec2-user/glial-origins/output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-gxc.tsv . Set overwrite = True to overwrite.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'species': 'Danio_rerio',\n",
       " 'conditions': 'juvbrain',\n",
       " 'directory': '/home/ec2-user/glial-origins/output/Drer_juvbrain/',\n",
       " 'files': {},\n",
       " 'metadata': <biofile_handling.metadata_object at 0x7f61ef1f7ee0>,\n",
       " 'annot': <biofile_handling.GenomeGffFile at 0x7f61eecc7a30>,\n",
       " 'genome_fasta': <biofile_handling.GenomeFastaFile at 0x7f613dd7f8b0>,\n",
       " 'genome_fasta_original': <biofile_handling.GenomeFastaFile at 0x7f613dd7e920>,\n",
       " 'gxc': <biofile_handling.GxcFile at 0x7f61ef1f58d0>,\n",
       " 'cellranger_filegroup': <biofile_handling.CellRangerFileGroup at 0x7f61ef1f4e20>,\n",
       " 'cellranger_features': <biofile_handling.CellRangerFeaturesFile at 0x7f613db38d90>,\n",
       " 'cellranger_matrix': <biofile_handling.CellRangerMatrixFile at 0x7f613dd4d120>,\n",
       " 'cellranger_barcodes': <biofile_handling.CellRangerBarcodesFile at 0x7f61ef1f7d60>}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify url and other variables\n",
    "genome_fasta_url = 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/002/035/GCA_000002035.3_GRCz10/GCA_000002035.3_GRCz10_genomic.fna.gz'\n",
    "genome_version = 'GRCz10-86'\n",
    "\n",
    "annot_url = 'https://ftp.ncbi.nlm.nih.gov/geo/series/GSE105nnn/GSE105010/suppl/GSE105010_Danio_rerio.GRCz10.86.modified.gtf.gz'\n",
    "\n",
    "barcodes_address = 'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4793nnn/GSM4793240/suppl/GSM4793240_comb-zBr15dpf6_S1-barcodes.tsv.gz'\n",
    "features_address = 'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4793nnn/GSM4793240/suppl/GSM4793240_comb-zBr15dpf6_S1-genes.tsv.gz'\n",
    "matrix_address = 'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4793nnn/GSM4793240/suppl/GSM4793240_comb-zBr15dpf6_S1-matrix.mtx.gz'\n",
    "\n",
    "###########\n",
    "# runtime #\n",
    "###########\n",
    "\n",
    "protocol = 'curl'\n",
    "\n",
    "genome_fasta_original = GenomeFastaFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    version = genome_version,\n",
    "    url = genome_fasta_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "# Renames RefSeq chromosomes, returning a new object\n",
    "genome_fasta = genome_fasta_original.rename_RefSeq_chromosomes()\n",
    "\n",
    "annot = GenomeGffFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    reference_genome = genome_fasta,\n",
    "    url = annot_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "# Download CellRanger files from urls\n",
    "cellranger_filegroup = CellRangerFileGroup(\n",
    "    sampledict = sample_BFD.sampledict, \n",
    "    barcodes_address = barcodes_address,\n",
    "    features_address = features_address,\n",
    "    matrix_address = matrix_address,\n",
    "    how = 'url',\n",
    "    protocol = protocol)\n",
    "\n",
    "# Use method to generate a gxc file from cellranger file\n",
    "# Will display first 10 rows of file before saving\n",
    "# For some reason this takes a long time?\n",
    "gxc = cellranger_filegroup.to_gxc(\n",
    "    filename = 'GSM4793240_comb-zBr15dpf6_S1-gxc.tsv',\n",
    "    reference_genome = genome_fasta,\n",
    "    reference_annot = annot\n",
    ")\n",
    "\n",
    "keyfiles = {\n",
    "    'annot': annot,\n",
    "    'genome_fasta': genome_fasta,\n",
    "    'genome_fasta_original': genome_fasta_original,\n",
    "    'gxc': gxc,\n",
    "    'cellranger_filegroup': cellranger_filegroup,\n",
    "    'cellranger_features': cellranger_filegroup.features,\n",
    "    'cellranger_matrix': cellranger_filegroup.matrix,\n",
    "    'cellranger_barcodes': cellranger_filegroup.barcodes\n",
    "}\n",
    "\n",
    "sample_BFD.add_keyfiles(keyfiles)\n",
    "display(vars(sample_BFD))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Load in the gxc matrix and get gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>AAACCTGAGAAGCCCA-1</th>\n",
       "      <th>AAACCTGAGGATGGAA-1</th>\n",
       "      <th>AAACCTGAGTCTTGCA-1</th>\n",
       "      <th>AAACCTGCACAGACAG-1</th>\n",
       "      <th>AAACCTGCACTTGGAT-1</th>\n",
       "      <th>AAACCTGCATGGGAAC-1</th>\n",
       "      <th>AAACCTGCATTCCTCG-1</th>\n",
       "      <th>AAACCTGGTACATCCA-1</th>\n",
       "      <th>AAACCTGGTGCTGTAT-1</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGTCAAGCGCTTAT-1</th>\n",
       "      <th>TTTGTCAAGCTAAACA-1</th>\n",
       "      <th>TTTGTCACACAAGACG-1</th>\n",
       "      <th>TTTGTCACACGCTTTC-1</th>\n",
       "      <th>TTTGTCACATCCTTGC-1</th>\n",
       "      <th>TTTGTCAGTATAGTAG-1</th>\n",
       "      <th>TTTGTCAGTATATGAG-1</th>\n",
       "      <th>TTTGTCAGTCGACTAT-1</th>\n",
       "      <th>TTTGTCATCATTGCGA-1</th>\n",
       "      <th>TTTGTCATCCTAAGTG-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PARP12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>syn3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ptpro</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>eps8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>si:ch1073-395m9.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tbk1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>gpr19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>crebl2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>dusp16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 5291 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_name  AAACCTGAGAAGCCCA-1  AAACCTGAGGATGGAA-1  \\\n",
       "0   si:ch73-252i11.3                   0                   0   \n",
       "1             PARP12                   0                   0   \n",
       "2               syn3                   0                   0   \n",
       "3              ptpro                   0                   0   \n",
       "4               eps8                   0                   0   \n",
       "5  si:ch1073-395m9.3                   0                   0   \n",
       "6               tbk1                   0                   0   \n",
       "7              gpr19                   0                   0   \n",
       "8             crebl2                   0                   0   \n",
       "9             dusp16                   0                   0   \n",
       "\n",
       "   AAACCTGAGTCTTGCA-1  AAACCTGCACAGACAG-1  AAACCTGCACTTGGAT-1  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   AAACCTGCATGGGAAC-1  AAACCTGCATTCCTCG-1  AAACCTGGTACATCCA-1  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   1   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   AAACCTGGTGCTGTAT-1  ...  TTTGTCAAGCGCTTAT-1  TTTGTCAAGCTAAACA-1  \\\n",
       "0                   0  ...                   0                   0   \n",
       "1                   0  ...                   0                   0   \n",
       "2                   0  ...                   0                   0   \n",
       "3                   0  ...                   0                   0   \n",
       "4                   0  ...                   0                   0   \n",
       "5                   0  ...                   0                   0   \n",
       "6                   0  ...                   0                   0   \n",
       "7                   0  ...                   0                   0   \n",
       "8                   0  ...                   0                   0   \n",
       "9                   0  ...                   0                   0   \n",
       "\n",
       "   TTTGTCACACAAGACG-1  TTTGTCACACGCTTTC-1  TTTGTCACATCCTTGC-1  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   TTTGTCAGTATAGTAG-1  TTTGTCAGTATATGAG-1  TTTGTCAGTCGACTAT-1  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   TTTGTCATCATTGCGA-1  TTTGTCATCCTAAGTG-1  \n",
       "0                   0                   0  \n",
       "1                   0                   0  \n",
       "2                   0                   0  \n",
       "3                   0                   0  \n",
       "4                   0                   0  \n",
       "5                   0                   0  \n",
       "6                   0                   0  \n",
       "7                   0                   0  \n",
       "8                   0                   0  \n",
       "9                   0                   0  \n",
       "\n",
       "[10 rows x 5291 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PARP12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>syn3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ptpro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32187</th>\n",
       "      <td>CABZ01109843.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32188</th>\n",
       "      <td>CABZ01111913.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32189</th>\n",
       "      <td>CABZ01079745.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32190</th>\n",
       "      <td>NP5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32191</th>\n",
       "      <td>NP33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32192 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_name\n",
       "0             gene_name\n",
       "1      si:ch73-252i11.3\n",
       "2                PARP12\n",
       "3                  syn3\n",
       "4                 ptpro\n",
       "...                 ...\n",
       "32187    CABZ01109843.1\n",
       "32188    CABZ01111913.1\n",
       "32189    CABZ01079745.1\n",
       "32190               NP5\n",
       "32191              NP33\n",
       "\n",
       "[32192 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes_matrix = pd.read_csv(sample_BFD.gxc.path, sep = '\\t', nrows = 10)\n",
    "display(genes_matrix)\n",
    "\n",
    "gxc_genes_list = pd.read_csv(sample_BFD.gxc.path, sep = '\\t', usecols=[0], names = ['gene_name'])\n",
    "display(gxc_genes_list)\n",
    "\n",
    "sample_BFD.metadata.add('num_cells', len(genes_matrix.columns) - 1)\n",
    "sample_BFD.metadata.add('num_genes', len(genes_matrix))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jvfcFNT2FuaZ"
   },
   "source": [
    "# 3. Get mapping identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_106482/746866034.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  models = pd.read_csv(sample_BFD.annot.path, header = None, sep = '\\t', comment = '#')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6733</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>6733</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>52002</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>48613</td>\n",
       "      <td>48740</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>13755</td>\n",
       "      <td>13811</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149247</th>\n",
       "      <td>22</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>598076</td>\n",
       "      <td>599835</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149248</th>\n",
       "      <td>22</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>598076</td>\n",
       "      <td>599835</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149249</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; gene_nam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149250</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149251</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1149252 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0               1           2         3         4  5  6  7  \\\n",
       "0         4  ensembl_havana        gene      6733     52120  .  -  .   \n",
       "1         4         ensembl  transcript      6733     52120  .  -  .   \n",
       "2         4         ensembl        exon     52002     52120  .  -  .   \n",
       "3         4         ensembl        exon     48613     48740  .  -  .   \n",
       "4         4         ensembl        exon     13755     13811  .  -  .   \n",
       "...      ..             ...         ...       ...       ... .. .. ..   \n",
       "1149247  22         ensembl        gene    598076    599835  .  +  .   \n",
       "1149248  22         ensembl        gene    598076    599835  .  +  .   \n",
       "1149249  16         ensembl        gene  47449630  47452948  .  +  .   \n",
       "1149250  16         ensembl  transcript  47449630  47452948  .  +  .   \n",
       "1149251  16         ensembl        exon  47449630  47452948  .  +  .   \n",
       "\n",
       "                                                         8  \n",
       "0        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...  \n",
       "1        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...  \n",
       "2        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...  \n",
       "3        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...  \n",
       "4        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...  \n",
       "...                                                    ...  \n",
       "1149247  gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...  \n",
       "1149248  gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...  \n",
       "1149249  gene_id \"ENSGNP33\"; gene_version \"1\"; gene_nam...  \n",
       "1149250  gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...  \n",
       "1149251  gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...  \n",
       "\n",
       "[1149252 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'gene_id \"ENSDARG00000104632\"; gene_version \"1\"; transcript_id \"ENSDART00000166186\"; transcript_version \"1\"; exon_number \"2\"; gene_name \"si:ch73-252i11.3\"; gene_source \"ensembl_havana\"; gene_biotype \"lincRNA\"; havana_gene \"OTTDARG00000037780\"; havana_gene_version \"1\"; transcript_name \"si:ch73-252i11.3-201\"; transcript_source \"ensembl\"; transcript_biotype \"lincRNA\"; exon_id \"ENSDARE00001217519\"; exon_version \"1\"; tag \"basic\";'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load in the original GFF-based annotation\n",
    "models = pd.read_csv(sample_BFD.annot.path, header = None, sep = '\\t', comment = '#')\n",
    "display(models)\n",
    "\n",
    "attributes_column = 8\n",
    "\n",
    "# Check the structure of fields in the GFF additional fields section\n",
    "display(models[attributes_column][3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6733</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "      <td>{'gene_id': 'ENSDARG00000104632', 'gene_versio...</td>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>6733</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "      <td>{'gene_id': 'ENSDARG00000104632', 'gene_versio...</td>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>52002</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "      <td>{'gene_id': 'ENSDARG00000104632', 'gene_versio...</td>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>48613</td>\n",
       "      <td>48740</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "      <td>{'gene_id': 'ENSDARG00000104632', 'gene_versio...</td>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>13755</td>\n",
       "      <td>13811</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "      <td>{'gene_id': 'ENSDARG00000104632', 'gene_versio...</td>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149247</th>\n",
       "      <td>22</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>598076</td>\n",
       "      <td>599835</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...</td>\n",
       "      <td>{'gene_id': 'ENSGNP31', 'gene_version': '1', '...</td>\n",
       "      <td>NP31</td>\n",
       "      <td>ENSGNP31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149248</th>\n",
       "      <td>22</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>598076</td>\n",
       "      <td>599835</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...</td>\n",
       "      <td>{'gene_id': 'ENSGNP31', 'gene_version': '1', '...</td>\n",
       "      <td>NP31</td>\n",
       "      <td>ENSGNP31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149249</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; gene_nam...</td>\n",
       "      <td>{'gene_id': 'ENSGNP33', 'gene_version': '1', '...</td>\n",
       "      <td>NP33</td>\n",
       "      <td>ENSGNP33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149250</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...</td>\n",
       "      <td>{'gene_id': 'ENSGNP33', 'gene_version': '1', '...</td>\n",
       "      <td>NP33</td>\n",
       "      <td>ENSGNP33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149251</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...</td>\n",
       "      <td>{'gene_id': 'ENSGNP33', 'gene_version': '1', '...</td>\n",
       "      <td>NP33</td>\n",
       "      <td>ENSGNP33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1149252 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0               1           2         3         4  5  6  7  \\\n",
       "0         4  ensembl_havana        gene      6733     52120  .  -  .   \n",
       "1         4         ensembl  transcript      6733     52120  .  -  .   \n",
       "2         4         ensembl        exon     52002     52120  .  -  .   \n",
       "3         4         ensembl        exon     48613     48740  .  -  .   \n",
       "4         4         ensembl        exon     13755     13811  .  -  .   \n",
       "...      ..             ...         ...       ...       ... .. .. ..   \n",
       "1149247  22         ensembl        gene    598076    599835  .  +  .   \n",
       "1149248  22         ensembl        gene    598076    599835  .  +  .   \n",
       "1149249  16         ensembl        gene  47449630  47452948  .  +  .   \n",
       "1149250  16         ensembl  transcript  47449630  47452948  .  +  .   \n",
       "1149251  16         ensembl        exon  47449630  47452948  .  +  .   \n",
       "\n",
       "                                                         8  \\\n",
       "0        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...   \n",
       "1        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...   \n",
       "2        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...   \n",
       "3        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...   \n",
       "4        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...   \n",
       "...                                                    ...   \n",
       "1149247  gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...   \n",
       "1149248  gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...   \n",
       "1149249  gene_id \"ENSGNP33\"; gene_version \"1\"; gene_nam...   \n",
       "1149250  gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...   \n",
       "1149251  gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...   \n",
       "\n",
       "                                          field_dictionary         gene_name  \\\n",
       "0        {'gene_id': 'ENSDARG00000104632', 'gene_versio...  si:ch73-252i11.3   \n",
       "1        {'gene_id': 'ENSDARG00000104632', 'gene_versio...  si:ch73-252i11.3   \n",
       "2        {'gene_id': 'ENSDARG00000104632', 'gene_versio...  si:ch73-252i11.3   \n",
       "3        {'gene_id': 'ENSDARG00000104632', 'gene_versio...  si:ch73-252i11.3   \n",
       "4        {'gene_id': 'ENSDARG00000104632', 'gene_versio...  si:ch73-252i11.3   \n",
       "...                                                    ...               ...   \n",
       "1149247  {'gene_id': 'ENSGNP31', 'gene_version': '1', '...              NP31   \n",
       "1149248  {'gene_id': 'ENSGNP31', 'gene_version': '1', '...              NP31   \n",
       "1149249  {'gene_id': 'ENSGNP33', 'gene_version': '1', '...              NP33   \n",
       "1149250  {'gene_id': 'ENSGNP33', 'gene_version': '1', '...              NP33   \n",
       "1149251  {'gene_id': 'ENSGNP33', 'gene_version': '1', '...              NP33   \n",
       "\n",
       "                    gene_id  \n",
       "0        ENSDARG00000104632  \n",
       "1        ENSDARG00000104632  \n",
       "2        ENSDARG00000104632  \n",
       "3        ENSDARG00000104632  \n",
       "4        ENSDARG00000104632  \n",
       "...                     ...  \n",
       "1149247            ENSGNP31  \n",
       "1149248            ENSGNP31  \n",
       "1149249            ENSGNP33  \n",
       "1149250            ENSGNP33  \n",
       "1149251            ENSGNP33  \n",
       "\n",
       "[1149252 rows x 12 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Remove any rows with NaNs\n",
    "models.dropna(inplace = True)\n",
    "\n",
    "# Extract field and database cross-ref (dbxref) information into columns\n",
    "models['field_dictionary'] = models[attributes_column].apply(convert_fields_to_dict_gtf)\n",
    "models['gene_name'] = [d.get('gene_name') for d in models['field_dictionary']]\n",
    "models['gene_id'] = [d.get('gene_id') for d in models['field_dictionary']]\n",
    "\n",
    "display(models)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Extract gene IDs for mapping to UniprotKB\n",
    "Specify which set of identifiers will be use to query the [Uniprot ID Mapping Tool](https://www.uniprot.org/id-mapping) via API.\n",
    "\n",
    "If using an identifier from the `dbxref_dict`, specify the name via string in the `dbxref_datafield` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PARP12</td>\n",
       "      <td>ENSDARG00000100660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>syn3</td>\n",
       "      <td>ENSDARG00000098417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>ptpro</td>\n",
       "      <td>ENSDARG00000100422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>eps8</td>\n",
       "      <td>ENSDARG00000102128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149224</th>\n",
       "      <td>CABZ01111913.1</td>\n",
       "      <td>ENSDARG00000100589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149229</th>\n",
       "      <td>CABZ01079745.1</td>\n",
       "      <td>ENSDARG00000099551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149240</th>\n",
       "      <td>NP5</td>\n",
       "      <td>ENSGNP5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149246</th>\n",
       "      <td>NP31</td>\n",
       "      <td>ENSGNP31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149249</th>\n",
       "      <td>NP33</td>\n",
       "      <td>ENSGNP33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32192 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_name             gene_id\n",
       "0        si:ch73-252i11.3  ENSDARG00000104632\n",
       "10                 PARP12  ENSDARG00000100660\n",
       "57                   syn3  ENSDARG00000098417\n",
       "67                  ptpro  ENSDARG00000100422\n",
       "182                  eps8  ENSDARG00000102128\n",
       "...                   ...                 ...\n",
       "1149224    CABZ01111913.1  ENSDARG00000100589\n",
       "1149229    CABZ01079745.1  ENSDARG00000099551\n",
       "1149240               NP5             ENSGNP5\n",
       "1149246              NP31            ENSGNP31\n",
       "1149249              NP33            ENSGNP33\n",
       "\n",
       "[32192 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dbxref_datafield = 'gene_id'\n",
    "\n",
    "models_subset = models[['gene_name', dbxref_datafield]].dropna().drop_duplicates()\n",
    "\n",
    "display(models_subset)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Generate gene list file to query Uniprot ID Mapping API\n",
    "Generate a text file ending in `_ids.txt` for sending to the ID mapping API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 32192 gene ids to /home/ec2-user/glial-origins/output/Drer_juvbrain/Drer_juvbrain_gene_id_ids.txt\n"
     ]
    }
   ],
   "source": [
    "gene_list = models_subset[dbxref_datafield].unique()\n",
    "\n",
    "genelist_object = GeneListFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    sources = [sample_BFD.annot],\n",
    "    genes = gene_list,\n",
    "    identifier = dbxref_datafield\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Query Uniprot ID Mapping API\n",
    "Specify the `from_type` variable based on the Uniprot name of the identifier.  \n",
    "The table below lists some databases and the `from_type` string that the API accepts for that datatype.  \n",
    "\n",
    "| datatype | `from_type` string | description |\n",
    "| ---: | :--- | :--- |\n",
    "| Mouse Genome Informatics | `MGI` | ID starts with `MGI:` |\n",
    "| Zebrafish Information Network | `ZFIN` | ID starts with `ZDB-GENE-` |\n",
    "| Xenbase | `Xenbase` | ID starts with `XB-GENE-` |\n",
    "\n",
    "__NOTE:__ You may have to run the cell below twice - UniProt sometimes throws an \"Resource not found\" message on the first query to the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "  0     0    0     0    0     0      0      0 --:--:--  0:02:09 --:--:--     0\n",
      "curl: (28) Failed to connect to rest.uniprot.org port 443 after 129845 ms: Connection timed out\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100   119    0   119    0     0    194      0 --:--:-- --:--:-- --:--:--   194\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Error messages</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Resource not found</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Error messages\n",
       "0  Resource not found"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from_type = 'Ensembl'\n",
    "to_type = 'UniProtKB'\n",
    "\n",
    "uniprot_idmm_object = genelist_object.get_uniprot_ids(ID_MAPPER_LOC, from_type, to_type)\n",
    "sample_BFD.add_keyfile('uniprot_output_idmm', uniprot_idmm_object, overwrite = True)\n",
    "\n",
    "uniprot_idmm = pd.read_csv(sample_BFD.uniprot_output_idmm.path, sep = '\\t')\n",
    "display(uniprot_idmm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Extract results and generate Uniprot IDMM\n",
    "Generates an idmm that links `gene_name`, the `dbxref_datafield` seleted above, and `uniprot_id` returned by API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"None of [Index(['From', 'Entry'], dtype='object')] are in the [columns]\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [11], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m uniprot_idpairs \u001b[38;5;241m=\u001b[39m \u001b[43muniprot_idmm\u001b[49m\u001b[43m[\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mFrom\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mEntry\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m      2\u001b[0m uniprot_idpairs\u001b[38;5;241m.\u001b[39mrename(columns \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mFrom\u001b[39m\u001b[38;5;124m'\u001b[39m: dbxref_datafield, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mEntry\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124muniprot_id\u001b[39m\u001b[38;5;124m'\u001b[39m}, inplace \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m      3\u001b[0m display(uniprot_idpairs)\n",
      "File \u001b[0;32m~/miniconda3/envs/glial_origins/lib/python3.10/site-packages/pandas/core/frame.py:3511\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3509\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m is_iterator(key):\n\u001b[1;32m   3510\u001b[0m         key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(key)\n\u001b[0;32m-> 3511\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_indexer_strict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcolumns\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m   3513\u001b[0m \u001b[38;5;66;03m# take() does not accept boolean indexers\u001b[39;00m\n\u001b[1;32m   3514\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(indexer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mbool\u001b[39m:\n",
      "File \u001b[0;32m~/miniconda3/envs/glial_origins/lib/python3.10/site-packages/pandas/core/indexes/base.py:5796\u001b[0m, in \u001b[0;36mIndex._get_indexer_strict\u001b[0;34m(self, key, axis_name)\u001b[0m\n\u001b[1;32m   5793\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   5794\u001b[0m     keyarr, indexer, new_indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reindex_non_unique(keyarr)\n\u001b[0;32m-> 5796\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_raise_if_missing\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeyarr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis_name\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   5798\u001b[0m keyarr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtake(indexer)\n\u001b[1;32m   5799\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, Index):\n\u001b[1;32m   5800\u001b[0m     \u001b[38;5;66;03m# GH 42790 - Preserve name from an Index\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/glial_origins/lib/python3.10/site-packages/pandas/core/indexes/base.py:5856\u001b[0m, in \u001b[0;36mIndex._raise_if_missing\u001b[0;34m(self, key, indexer, axis_name)\u001b[0m\n\u001b[1;32m   5854\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m use_interval_msg:\n\u001b[1;32m   5855\u001b[0m         key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(key)\n\u001b[0;32m-> 5856\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNone of [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m] are in the [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00maxis_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   5858\u001b[0m not_found \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(ensure_index(key)[missing_mask\u001b[38;5;241m.\u001b[39mnonzero()[\u001b[38;5;241m0\u001b[39m]]\u001b[38;5;241m.\u001b[39munique())\n\u001b[1;32m   5859\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnot_found\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not in index\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mKeyError\u001b[0m: \"None of [Index(['From', 'Entry'], dtype='object')] are in the [columns]\""
     ]
    }
   ],
   "source": [
    "uniprot_idpairs = uniprot_idmm[['From', 'Entry']]\n",
    "uniprot_idpairs.rename(columns = {'From': dbxref_datafield, 'Entry': 'uniprot_id'}, inplace = True)\n",
    "display(uniprot_idpairs)\n",
    "\n",
    "uniprot_output_idmm = models_subset.merge(uniprot_idpairs, on = dbxref_datafield)\n",
    "display(uniprot_output_idmm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate a filename and file for the idmm\n",
    "uniprot_output_idmm_filename = '_'.join([sample_BFD.species_prefix, conditions, 'uniprot-idmm.tsv'])\n",
    "uniprot_output_idmm_object = IdmmFile(\n",
    "    filename = uniprot_output_idmm_filename, \n",
    "    sampledict = sample_BFD.sampledict, \n",
    "    kind = 'uniprot_idmm', \n",
    "    sources = [sample_BFD.annot]\n",
    ")\n",
    "\n",
    "# save to file and add to the BioFileDocket\n",
    "uniprot_output_idmm.to_csv(uniprot_output_idmm_object.path, sep = '\\t')\n",
    "sample_BFD.add_keyfile('uniprot_idmm', uniprot_output_idmm_object)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8. Convert GFF to GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load the newly-generated GTF file as a dataframe\n",
    "models_asgtf_df = pd.read_csv(sample_BFD.annot.path, skiprows = 5, header = None, sep = '\\t')\n",
    "\n",
    "display(models_asgtf_df)\n",
    "display(models_asgtf_df[attributes_column][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use a custom function to extract useful fields from the additional fields section (column 8)\n",
    "# Pull from that dict to fill in additional useful columns\n",
    "models_asgtf_df['field_dictionary'] = models_asgtf_df[attributes_column].apply(convert_fields_to_dict_gtf)\n",
    "models_asgtf_df['gene_name'] = [d.get('gene_name') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['gene_id'] = [d.get('gene_id') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['transcript_id'] = [d.get('transcript_id') for d in models_asgtf_df['field_dictionary']]\n",
    "\n",
    "# Remove CDS annotations because they interfere with TransDecoder cDNA generation\n",
    "models_asgtf_df = models_asgtf_df[models_asgtf_df[2] != 'CDS']\n",
    "display(models_asgtf_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9. Generate gtf-idmm\n",
    "This file maps the `gene_name` to `gene_id` and `transcript_id` fields generated by the conversion from GFF to GTF, which will be needed for downstream processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract gene_name, gene_id, and transcript_id fields to generate an ID mapping matrix (idmm)\n",
    "idmm_df = models_asgtf_df[['gene_name', 'gene_id', 'transcript_id']].drop_duplicates()\n",
    "idmm_df.dropna(inplace = True)\n",
    "display(idmm_df)\n",
    "\n",
    "# generate a filename and file for the idmm\n",
    "idmm_filename = '_'.join([sample_BFD.species_prefix, conditions, 'gtf-idmm.tsv'])\n",
    "idmm = IdmmFile(\n",
    "    filename = idmm_filename, \n",
    "    sampledict = sample_BFD.sampledict, \n",
    "    kind = 'gtf_idmm', \n",
    "    sources = [sample_BFD.annot]\n",
    ")\n",
    "\n",
    "# save to file and add to the BioFileDocket\n",
    "idmm_df.to_csv(idmm.path, sep = '\\t')\n",
    "sample_BFD.add_keyfile('gtf_idmm', idmm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 10. Generate updated gtf\n",
    "Generated an updated GTF file using transcript_id as the key. For some datasets, transcripts do not consistently get gene names and gene IDs added, which causes Transdecoder to throw errors. This resolves that problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "models_asgtf_updated_df = models_asgtf_df.merge(idmm_df, on = 'transcript_id')\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_name': x['gene_name_y']}), axis = 1)\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_id': x['gene_id_y']}), axis = 1)\n",
    "models_asgtf_updated_df['new_fields'] = models_asgtf_updated_df['field_dictionary'].apply(convert_dict_to_fields_gtf)\n",
    "models_asgtf_updated_df = models_asgtf_updated_df[[0, 1, 2, 3, 4, 5, 6, 7, 'new_fields']]\n",
    "\n",
    "models_asgtf_updated_filename = sample_BFD.annot.filename.replace('.gtf', '_updated.gtf')\n",
    "models_asgtf_updated = GenomeGtfFile(\n",
    "    filename = models_asgtf_updated_filename, \n",
    "    sampledict = sample_BFD.sampledict, \n",
    "    reference_genome = sample_BFD.genome_fasta\n",
    ")\n",
    "\n",
    "models_asgtf_updated_df.to_csv(models_asgtf_updated.path, header = None, index = None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 11. Generate cDNA and peptide files\n",
    "Using the updated gtf file and genome file, generate cDNA sequence.\n",
    "\n",
    "Then, using the cDNA sequence, generate peptide sequences using transdecoder.\n",
    "\n",
    "Expect this step to take some time, probably ~20-30min."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cdna = sample_BFD.genome_fasta.get_transdecoder_cdna_gtf(models_asgtf_updated, TRANSDECODER_LOC)\n",
    "sample_BFD.add_keyfile('cdna', cdna)\n",
    "\n",
    "transdecoder_files = sample_BFD.cdna.to_pep_files(TDLONGORF_LOC, TDPREDICT_LOC)\n",
    "sample_BFD.add_keyfiles(transdecoder_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "iEug6TQ6Cc9D"
   },
   "source": [
    "# 12. Push files to AWS S3\n",
    "\n",
    "Iteratively moves through BioFileDocket BioFile objects and pushes to the right place in AWS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_BFD.local_to_s3()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 13. Pickle the `sample_BFD` variable for use by the next script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate a .pkl file for the Docket\n",
    "sample_BFD.pickle()\n",
    "\n",
    "# Push to S3, optionally overwriting existing file\n",
    "sample_BFD.push_to_s3(overwrite = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "Xtro_bro.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
