{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview\n",
    "\n",
    "This notebook collects data from the [Raj et al. 2020](https://www.frontiersin.org/articles/10.3389/fcell.2021.743421/full) zebrafish brain developmental cell atlas.\n",
    "\n",
    "#### Dataset description\n",
    "\n",
    "- Data can be found at the [GEO Accession GSE158142](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE158142).\n",
    "- Genome annotation used was at [GEO Accession GSE105010](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE105010)\n",
    "- This notebook previously collects data from one sample, [\"zBr15dpf1_S1\"](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM4793235), for the purposes of data analysis and exploration."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "L56fjA27DmhA"
   },
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "t0BTCsz7Y48y",
    "outputId": "f6510678-f977-4ec8-ce30-1f2acd4fe666"
   },
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess, os, dill, sys\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Download and describe data\n",
    "\n",
    "#### BioFileDocket\n",
    "First, make a BioFileDocket for the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/glial-origins/output/Drer_juvbrain/ already exists\n",
      "Files will be saved into /home/ec2-user/glial-origins/output/Drer_juvbrain/\n"
     ]
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify the name of the species folder in Amazon S3\n",
    "species = 'Danio_rerio'\n",
    "\n",
    "# Specify any particular identifying conditions, eg tissue type:\n",
    "conditions = 'juvbrain'\n",
    "\n",
    "################\n",
    "################\n",
    "\n",
    "sample_BFD = BioFileDocket(species, conditions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inferring file name as GCA_000002035.3_GRCz10_genomic.fna.gz\n",
      "file GCA_000002035.3_GRCz10_genomic.fna.gz already exists at /home/ec2-user/glial-origins/output/Drer_juvbrain/GCA_000002035.3_GRCz10_genomic.fna.gz\n",
      "file GCA_000002035.3_GRCz10_genomic.fna.gz unzipped and object renamed to GCA_000002035.3_GRCz10_genomic.fna\n",
      "inferring file name as GSE105010_Danio_rerio.GRCz10.86.modified.gtf.gz\n",
      "file GSE105010_Danio_rerio.GRCz10.86.modified.gtf.gz already exists at /home/ec2-user/glial-origins/output/Drer_juvbrain/GSE105010_Danio_rerio.GRCz10.86.modified.gtf.gz\n",
      "file GSE105010_Danio_rerio.GRCz10.86.modified.gtf.gz unzipped and object renamed to GSE105010_Danio_rerio.GRCz10.86.modified.gtf\n",
      "inferring file name as GSM4793240_comb-zBr15dpf6_S1-barcodes.tsv.gz\n",
      "file GSM4793240_comb-zBr15dpf6_S1-barcodes.tsv.gz already exists at /home/ec2-user/glial-origins/output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-barcodes.tsv.gz\n",
      "file GSM4793240_comb-zBr15dpf6_S1-barcodes.tsv.gz unzipped and object renamed to GSM4793240_comb-zBr15dpf6_S1-barcodes.tsv\n",
      "inferring file name as GSM4793240_comb-zBr15dpf6_S1-matrix.mtx.gz\n",
      "file GSM4793240_comb-zBr15dpf6_S1-matrix.mtx.gz already exists at /home/ec2-user/glial-origins/output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-matrix.mtx.gz\n",
      "file GSM4793240_comb-zBr15dpf6_S1-matrix.mtx.gz unzipped and object renamed to GSM4793240_comb-zBr15dpf6_S1-matrix.mtx\n",
      "inferring file name as GSM4793240_comb-zBr15dpf6_S1-genes.tsv.gz\n",
      "file GSM4793240_comb-zBr15dpf6_S1-genes.tsv.gz already exists at /home/ec2-user/glial-origins/output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-genes.tsv.gz\n",
      "file GSM4793240_comb-zBr15dpf6_S1-genes.tsv.gz unzipped and object renamed to GSM4793240_comb-zBr15dpf6_S1-genes.tsv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "gzip: /home/ec2-user/glial-origins/output/Drer_juvbrain/GCA_000002035.3_GRCz10_genomic.fna: unknown suffix -- ignored\n",
      "gzip: /home/ec2-user/glial-origins/output/Drer_juvbrain/GSE105010_Danio_rerio.GRCz10.86.modified.gtf: unknown suffix -- ignored\n",
      "gzip: /home/ec2-user/glial-origins/output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-barcodes.tsv: unknown suffix -- ignored\n",
      "gzip: /home/ec2-user/glial-origins/output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-matrix.mtx: unknown suffix -- ignored\n",
      "gzip: /home/ec2-user/glial-origins/output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-genes.tsv: unknown suffix -- ignored\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>AAACCTGAGAAGCCCA-1</th>\n",
       "      <th>AAACCTGAGGATGGAA-1</th>\n",
       "      <th>AAACCTGAGTCTTGCA-1</th>\n",
       "      <th>AAACCTGCACAGACAG-1</th>\n",
       "      <th>AAACCTGCACTTGGAT-1</th>\n",
       "      <th>AAACCTGCATGGGAAC-1</th>\n",
       "      <th>AAACCTGCATTCCTCG-1</th>\n",
       "      <th>AAACCTGGTACATCCA-1</th>\n",
       "      <th>AAACCTGGTGCTGTAT-1</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGTCAAGCGCTTAT-1</th>\n",
       "      <th>TTTGTCAAGCTAAACA-1</th>\n",
       "      <th>TTTGTCACACAAGACG-1</th>\n",
       "      <th>TTTGTCACACGCTTTC-1</th>\n",
       "      <th>TTTGTCACATCCTTGC-1</th>\n",
       "      <th>TTTGTCAGTATAGTAG-1</th>\n",
       "      <th>TTTGTCAGTATATGAG-1</th>\n",
       "      <th>TTTGTCAGTCGACTAT-1</th>\n",
       "      <th>TTTGTCATCATTGCGA-1</th>\n",
       "      <th>TTTGTCATCCTAAGTG-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PARP12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>syn3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ptpro</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>eps8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>si:ch1073-395m9.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tbk1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>gpr19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>crebl2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>dusp16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 5291 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_name  AAACCTGAGAAGCCCA-1  AAACCTGAGGATGGAA-1  \\\n",
       "0   si:ch73-252i11.3                   0                   0   \n",
       "1             PARP12                   0                   0   \n",
       "2               syn3                   0                   0   \n",
       "3              ptpro                   0                   0   \n",
       "4               eps8                   0                   0   \n",
       "5  si:ch1073-395m9.3                   0                   0   \n",
       "6               tbk1                   0                   0   \n",
       "7              gpr19                   0                   0   \n",
       "8             crebl2                   0                   0   \n",
       "9             dusp16                   0                   0   \n",
       "\n",
       "   AAACCTGAGTCTTGCA-1  AAACCTGCACAGACAG-1  AAACCTGCACTTGGAT-1  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   AAACCTGCATGGGAAC-1  AAACCTGCATTCCTCG-1  AAACCTGGTACATCCA-1  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   1   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   AAACCTGGTGCTGTAT-1  ...  TTTGTCAAGCGCTTAT-1  TTTGTCAAGCTAAACA-1  \\\n",
       "0                   0  ...                   0                   0   \n",
       "1                   0  ...                   0                   0   \n",
       "2                   0  ...                   0                   0   \n",
       "3                   0  ...                   0                   0   \n",
       "4                   0  ...                   0                   0   \n",
       "5                   0  ...                   0                   0   \n",
       "6                   0  ...                   0                   0   \n",
       "7                   0  ...                   0                   0   \n",
       "8                   0  ...                   0                   0   \n",
       "9                   0  ...                   0                   0   \n",
       "\n",
       "   TTTGTCACACAAGACG-1  TTTGTCACACGCTTTC-1  TTTGTCACATCCTTGC-1  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   TTTGTCAGTATAGTAG-1  TTTGTCAGTATATGAG-1  TTTGTCAGTCGACTAT-1  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   TTTGTCATCATTGCGA-1  TTTGTCATCCTAAGTG-1  \n",
       "0                   0                   0  \n",
       "1                   0                   0  \n",
       "2                   0                   0  \n",
       "3                   0                   0  \n",
       "4                   0                   0  \n",
       "5                   0                   0  \n",
       "6                   0                   0  \n",
       "7                   0                   0  \n",
       "8                   0                   0  \n",
       "9                   0                   0  \n",
       "\n",
       "[10 rows x 5291 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File already exists at /home/ec2-user/glial-origins/output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-gxc.tsv . Set overwrite = True to overwrite.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'species': 'Danio_rerio',\n",
       " 'conditions': 'juvbrain',\n",
       " 'directory': '/home/ec2-user/glial-origins/output/Drer_juvbrain/',\n",
       " 'files': {},\n",
       " 'metadata': <biofile_handling.metadata_object at 0x7fe58c0ef880>,\n",
       " 'annot': <biofile_handling.GenomeGffFile at 0x7fe58c0efbb0>,\n",
       " 'genome_fasta': <biofile_handling.GenomeFastaFile at 0x7fe58c0ed0f0>,\n",
       " 'genome_fasta_original': <biofile_handling.GenomeFastaFile at 0x7fe4db1b1b40>,\n",
       " 'gxc': <biofile_handling.GxcFile at 0x7fe4db1e8a30>,\n",
       " 'cellranger_filegroup': <biofile_handling.CellRangerFileGroup at 0x7fe58c0efa30>,\n",
       " 'cellranger_features': <biofile_handling.CellRangerFeaturesFile at 0x7fe4daf64d60>,\n",
       " 'cellranger_matrix': <biofile_handling.CellRangerMatrixFile at 0x7fe4db184af0>,\n",
       " 'cellranger_barcodes': <biofile_handling.CellRangerBarcodesFile at 0x7fe58c0ee860>}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify url and other variables\n",
    "genome_fasta_url = 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/002/035/GCA_000002035.3_GRCz10/GCA_000002035.3_GRCz10_genomic.fna.gz'\n",
    "genome_version = 'GRCz10-86'\n",
    "\n",
    "annot_url = 'https://ftp.ncbi.nlm.nih.gov/geo/series/GSE105nnn/GSE105010/suppl/GSE105010_Danio_rerio.GRCz10.86.modified.gtf.gz'\n",
    "\n",
    "barcodes_address = 'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4793nnn/GSM4793240/suppl/GSM4793240_comb-zBr15dpf6_S1-barcodes.tsv.gz'\n",
    "features_address = 'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4793nnn/GSM4793240/suppl/GSM4793240_comb-zBr15dpf6_S1-genes.tsv.gz'\n",
    "matrix_address = 'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM4793nnn/GSM4793240/suppl/GSM4793240_comb-zBr15dpf6_S1-matrix.mtx.gz'\n",
    "\n",
    "###########\n",
    "# runtime #\n",
    "###########\n",
    "\n",
    "protocol = 'curl'\n",
    "\n",
    "genome_fasta_original = GenomeFastaFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    version = genome_version,\n",
    "    url = genome_fasta_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "# Renames RefSeq chromosomes, returning a new object\n",
    "genome_fasta = genome_fasta_original.rename_RefSeq_chromosomes()\n",
    "\n",
    "annot = GenomeGffFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    reference_genome = genome_fasta,\n",
    "    url = annot_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "# Download CellRanger files from urls\n",
    "cellranger_filegroup = CellRangerFileGroup(\n",
    "    sampledict = sample_BFD.sampledict, \n",
    "    barcodes_address = barcodes_address,\n",
    "    features_address = features_address,\n",
    "    matrix_address = matrix_address,\n",
    "    how = 'url',\n",
    "    protocol = protocol)\n",
    "\n",
    "# Use method to generate a gxc file from cellranger file\n",
    "# Will display first 10 rows of file before saving\n",
    "# For some reason this takes a long time?\n",
    "gxc = cellranger_filegroup.to_gxc(\n",
    "    filename = 'GSM4793240_comb-zBr15dpf6_S1-gxc.tsv',\n",
    "    reference_genome = genome_fasta,\n",
    "    reference_annot = annot\n",
    ")\n",
    "\n",
    "keyfiles = {\n",
    "    'annot': annot,\n",
    "    'genome_fasta': genome_fasta,\n",
    "    'genome_fasta_original': genome_fasta_original,\n",
    "    'gxc': gxc,\n",
    "    'cellranger_filegroup': cellranger_filegroup,\n",
    "    'cellranger_features': cellranger_filegroup.features,\n",
    "    'cellranger_matrix': cellranger_filegroup.matrix,\n",
    "    'cellranger_barcodes': cellranger_filegroup.barcodes\n",
    "}\n",
    "\n",
    "sample_BFD.add_keyfiles(keyfiles)\n",
    "display(vars(sample_BFD))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Load in the gxc matrix and get gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>AAACCTGAGAAGCCCA-1</th>\n",
       "      <th>AAACCTGAGGATGGAA-1</th>\n",
       "      <th>AAACCTGAGTCTTGCA-1</th>\n",
       "      <th>AAACCTGCACAGACAG-1</th>\n",
       "      <th>AAACCTGCACTTGGAT-1</th>\n",
       "      <th>AAACCTGCATGGGAAC-1</th>\n",
       "      <th>AAACCTGCATTCCTCG-1</th>\n",
       "      <th>AAACCTGGTACATCCA-1</th>\n",
       "      <th>AAACCTGGTGCTGTAT-1</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGTCAAGCGCTTAT-1</th>\n",
       "      <th>TTTGTCAAGCTAAACA-1</th>\n",
       "      <th>TTTGTCACACAAGACG-1</th>\n",
       "      <th>TTTGTCACACGCTTTC-1</th>\n",
       "      <th>TTTGTCACATCCTTGC-1</th>\n",
       "      <th>TTTGTCAGTATAGTAG-1</th>\n",
       "      <th>TTTGTCAGTATATGAG-1</th>\n",
       "      <th>TTTGTCAGTCGACTAT-1</th>\n",
       "      <th>TTTGTCATCATTGCGA-1</th>\n",
       "      <th>TTTGTCATCCTAAGTG-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PARP12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>syn3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ptpro</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>eps8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>si:ch1073-395m9.3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tbk1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>gpr19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>crebl2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>dusp16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 5291 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_name  AAACCTGAGAAGCCCA-1  AAACCTGAGGATGGAA-1  \\\n",
       "0   si:ch73-252i11.3                   0                   0   \n",
       "1             PARP12                   0                   0   \n",
       "2               syn3                   0                   0   \n",
       "3              ptpro                   0                   0   \n",
       "4               eps8                   0                   0   \n",
       "5  si:ch1073-395m9.3                   0                   0   \n",
       "6               tbk1                   0                   0   \n",
       "7              gpr19                   0                   0   \n",
       "8             crebl2                   0                   0   \n",
       "9             dusp16                   0                   0   \n",
       "\n",
       "   AAACCTGAGTCTTGCA-1  AAACCTGCACAGACAG-1  AAACCTGCACTTGGAT-1  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   AAACCTGCATGGGAAC-1  AAACCTGCATTCCTCG-1  AAACCTGGTACATCCA-1  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   1   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   AAACCTGGTGCTGTAT-1  ...  TTTGTCAAGCGCTTAT-1  TTTGTCAAGCTAAACA-1  \\\n",
       "0                   0  ...                   0                   0   \n",
       "1                   0  ...                   0                   0   \n",
       "2                   0  ...                   0                   0   \n",
       "3                   0  ...                   0                   0   \n",
       "4                   0  ...                   0                   0   \n",
       "5                   0  ...                   0                   0   \n",
       "6                   0  ...                   0                   0   \n",
       "7                   0  ...                   0                   0   \n",
       "8                   0  ...                   0                   0   \n",
       "9                   0  ...                   0                   0   \n",
       "\n",
       "   TTTGTCACACAAGACG-1  TTTGTCACACGCTTTC-1  TTTGTCACATCCTTGC-1  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   TTTGTCAGTATAGTAG-1  TTTGTCAGTATATGAG-1  TTTGTCAGTCGACTAT-1  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   TTTGTCATCATTGCGA-1  TTTGTCATCCTAAGTG-1  \n",
       "0                   0                   0  \n",
       "1                   0                   0  \n",
       "2                   0                   0  \n",
       "3                   0                   0  \n",
       "4                   0                   0  \n",
       "5                   0                   0  \n",
       "6                   0                   0  \n",
       "7                   0                   0  \n",
       "8                   0                   0  \n",
       "9                   0                   0  \n",
       "\n",
       "[10 rows x 5291 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PARP12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>syn3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ptpro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32187</th>\n",
       "      <td>CABZ01109843.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32188</th>\n",
       "      <td>CABZ01111913.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32189</th>\n",
       "      <td>CABZ01079745.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32190</th>\n",
       "      <td>NP5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32191</th>\n",
       "      <td>NP33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32192 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_name\n",
       "0             gene_name\n",
       "1      si:ch73-252i11.3\n",
       "2                PARP12\n",
       "3                  syn3\n",
       "4                 ptpro\n",
       "...                 ...\n",
       "32187    CABZ01109843.1\n",
       "32188    CABZ01111913.1\n",
       "32189    CABZ01079745.1\n",
       "32190               NP5\n",
       "32191              NP33\n",
       "\n",
       "[32192 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes_matrix = pd.read_csv(sample_BFD.gxc.path, sep = '\\t', nrows = 10)\n",
    "display(genes_matrix)\n",
    "\n",
    "gxc_genes_list = pd.read_csv(sample_BFD.gxc.path, sep = '\\t', usecols=[0], names = ['gene_name'])\n",
    "display(gxc_genes_list)\n",
    "\n",
    "sample_BFD.metadata.add('num_cells', len(genes_matrix.columns) - 1)\n",
    "sample_BFD.metadata.add('num_genes', len(genes_matrix))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jvfcFNT2FuaZ"
   },
   "source": [
    "# 3. Get mapping identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_14128/746866034.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  models = pd.read_csv(sample_BFD.annot.path, header = None, sep = '\\t', comment = '#')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6733</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>6733</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>52002</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>48613</td>\n",
       "      <td>48740</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>13755</td>\n",
       "      <td>13811</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149247</th>\n",
       "      <td>22</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>598076</td>\n",
       "      <td>599835</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149248</th>\n",
       "      <td>22</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>598076</td>\n",
       "      <td>599835</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149249</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; gene_nam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149250</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149251</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1149252 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0               1           2         3         4  5  6  7  \\\n",
       "0         4  ensembl_havana        gene      6733     52120  .  -  .   \n",
       "1         4         ensembl  transcript      6733     52120  .  -  .   \n",
       "2         4         ensembl        exon     52002     52120  .  -  .   \n",
       "3         4         ensembl        exon     48613     48740  .  -  .   \n",
       "4         4         ensembl        exon     13755     13811  .  -  .   \n",
       "...      ..             ...         ...       ...       ... .. .. ..   \n",
       "1149247  22         ensembl        gene    598076    599835  .  +  .   \n",
       "1149248  22         ensembl        gene    598076    599835  .  +  .   \n",
       "1149249  16         ensembl        gene  47449630  47452948  .  +  .   \n",
       "1149250  16         ensembl  transcript  47449630  47452948  .  +  .   \n",
       "1149251  16         ensembl        exon  47449630  47452948  .  +  .   \n",
       "\n",
       "                                                         8  \n",
       "0        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...  \n",
       "1        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...  \n",
       "2        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...  \n",
       "3        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...  \n",
       "4        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...  \n",
       "...                                                    ...  \n",
       "1149247  gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...  \n",
       "1149248  gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...  \n",
       "1149249  gene_id \"ENSGNP33\"; gene_version \"1\"; gene_nam...  \n",
       "1149250  gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...  \n",
       "1149251  gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...  \n",
       "\n",
       "[1149252 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'gene_id \"ENSDARG00000104632\"; gene_version \"1\"; transcript_id \"ENSDART00000166186\"; transcript_version \"1\"; exon_number \"2\"; gene_name \"si:ch73-252i11.3\"; gene_source \"ensembl_havana\"; gene_biotype \"lincRNA\"; havana_gene \"OTTDARG00000037780\"; havana_gene_version \"1\"; transcript_name \"si:ch73-252i11.3-201\"; transcript_source \"ensembl\"; transcript_biotype \"lincRNA\"; exon_id \"ENSDARE00001217519\"; exon_version \"1\"; tag \"basic\";'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load in the original GFF-based annotation\n",
    "models = pd.read_csv(sample_BFD.annot.path, header = None, sep = '\\t', comment = '#')\n",
    "display(models)\n",
    "\n",
    "attributes_column = 8\n",
    "\n",
    "# Check the structure of fields in the GFF additional fields section\n",
    "display(models[attributes_column][3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6733</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "      <td>{'gene_id': 'ENSDARG00000104632', 'gene_versio...</td>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>6733</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "      <td>{'gene_id': 'ENSDARG00000104632', 'gene_versio...</td>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>52002</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "      <td>{'gene_id': 'ENSDARG00000104632', 'gene_versio...</td>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>48613</td>\n",
       "      <td>48740</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "      <td>{'gene_id': 'ENSDARG00000104632', 'gene_versio...</td>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>13755</td>\n",
       "      <td>13811</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "      <td>{'gene_id': 'ENSDARG00000104632', 'gene_versio...</td>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149247</th>\n",
       "      <td>22</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>598076</td>\n",
       "      <td>599835</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...</td>\n",
       "      <td>{'gene_id': 'ENSGNP31', 'gene_version': '1', '...</td>\n",
       "      <td>NP31</td>\n",
       "      <td>ENSGNP31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149248</th>\n",
       "      <td>22</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>598076</td>\n",
       "      <td>599835</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...</td>\n",
       "      <td>{'gene_id': 'ENSGNP31', 'gene_version': '1', '...</td>\n",
       "      <td>NP31</td>\n",
       "      <td>ENSGNP31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149249</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; gene_nam...</td>\n",
       "      <td>{'gene_id': 'ENSGNP33', 'gene_version': '1', '...</td>\n",
       "      <td>NP33</td>\n",
       "      <td>ENSGNP33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149250</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...</td>\n",
       "      <td>{'gene_id': 'ENSGNP33', 'gene_version': '1', '...</td>\n",
       "      <td>NP33</td>\n",
       "      <td>ENSGNP33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149251</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...</td>\n",
       "      <td>{'gene_id': 'ENSGNP33', 'gene_version': '1', '...</td>\n",
       "      <td>NP33</td>\n",
       "      <td>ENSGNP33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1149252 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0               1           2         3         4  5  6  7  \\\n",
       "0         4  ensembl_havana        gene      6733     52120  .  -  .   \n",
       "1         4         ensembl  transcript      6733     52120  .  -  .   \n",
       "2         4         ensembl        exon     52002     52120  .  -  .   \n",
       "3         4         ensembl        exon     48613     48740  .  -  .   \n",
       "4         4         ensembl        exon     13755     13811  .  -  .   \n",
       "...      ..             ...         ...       ...       ... .. .. ..   \n",
       "1149247  22         ensembl        gene    598076    599835  .  +  .   \n",
       "1149248  22         ensembl        gene    598076    599835  .  +  .   \n",
       "1149249  16         ensembl        gene  47449630  47452948  .  +  .   \n",
       "1149250  16         ensembl  transcript  47449630  47452948  .  +  .   \n",
       "1149251  16         ensembl        exon  47449630  47452948  .  +  .   \n",
       "\n",
       "                                                         8  \\\n",
       "0        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...   \n",
       "1        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...   \n",
       "2        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...   \n",
       "3        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...   \n",
       "4        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...   \n",
       "...                                                    ...   \n",
       "1149247  gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...   \n",
       "1149248  gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...   \n",
       "1149249  gene_id \"ENSGNP33\"; gene_version \"1\"; gene_nam...   \n",
       "1149250  gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...   \n",
       "1149251  gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...   \n",
       "\n",
       "                                          field_dictionary         gene_name  \\\n",
       "0        {'gene_id': 'ENSDARG00000104632', 'gene_versio...  si:ch73-252i11.3   \n",
       "1        {'gene_id': 'ENSDARG00000104632', 'gene_versio...  si:ch73-252i11.3   \n",
       "2        {'gene_id': 'ENSDARG00000104632', 'gene_versio...  si:ch73-252i11.3   \n",
       "3        {'gene_id': 'ENSDARG00000104632', 'gene_versio...  si:ch73-252i11.3   \n",
       "4        {'gene_id': 'ENSDARG00000104632', 'gene_versio...  si:ch73-252i11.3   \n",
       "...                                                    ...               ...   \n",
       "1149247  {'gene_id': 'ENSGNP31', 'gene_version': '1', '...              NP31   \n",
       "1149248  {'gene_id': 'ENSGNP31', 'gene_version': '1', '...              NP31   \n",
       "1149249  {'gene_id': 'ENSGNP33', 'gene_version': '1', '...              NP33   \n",
       "1149250  {'gene_id': 'ENSGNP33', 'gene_version': '1', '...              NP33   \n",
       "1149251  {'gene_id': 'ENSGNP33', 'gene_version': '1', '...              NP33   \n",
       "\n",
       "                    gene_id  \n",
       "0        ENSDARG00000104632  \n",
       "1        ENSDARG00000104632  \n",
       "2        ENSDARG00000104632  \n",
       "3        ENSDARG00000104632  \n",
       "4        ENSDARG00000104632  \n",
       "...                     ...  \n",
       "1149247            ENSGNP31  \n",
       "1149248            ENSGNP31  \n",
       "1149249            ENSGNP33  \n",
       "1149250            ENSGNP33  \n",
       "1149251            ENSGNP33  \n",
       "\n",
       "[1149252 rows x 12 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Remove any rows with NaNs\n",
    "models.dropna(inplace = True)\n",
    "\n",
    "# Extract field and database cross-ref (dbxref) information into columns\n",
    "models['field_dictionary'] = models[attributes_column].apply(convert_fields_to_dict_gtf)\n",
    "models['gene_name'] = [d.get('gene_name') for d in models['field_dictionary']]\n",
    "models['gene_id'] = [d.get('gene_id') for d in models['field_dictionary']]\n",
    "\n",
    "display(models)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Extract gene IDs for mapping to UniprotKB\n",
    "Specify which set of identifiers will be use to query the [Uniprot ID Mapping Tool](https://www.uniprot.org/id-mapping) via API.\n",
    "\n",
    "If using an identifier from the `dbxref_dict`, specify the name via string in the `dbxref_datafield` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PARP12</td>\n",
       "      <td>ENSDARG00000100660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>syn3</td>\n",
       "      <td>ENSDARG00000098417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>ptpro</td>\n",
       "      <td>ENSDARG00000100422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>eps8</td>\n",
       "      <td>ENSDARG00000102128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149224</th>\n",
       "      <td>CABZ01111913.1</td>\n",
       "      <td>ENSDARG00000100589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149229</th>\n",
       "      <td>CABZ01079745.1</td>\n",
       "      <td>ENSDARG00000099551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149240</th>\n",
       "      <td>NP5</td>\n",
       "      <td>ENSGNP5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149246</th>\n",
       "      <td>NP31</td>\n",
       "      <td>ENSGNP31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149249</th>\n",
       "      <td>NP33</td>\n",
       "      <td>ENSGNP33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32192 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_name             gene_id\n",
       "0        si:ch73-252i11.3  ENSDARG00000104632\n",
       "10                 PARP12  ENSDARG00000100660\n",
       "57                   syn3  ENSDARG00000098417\n",
       "67                  ptpro  ENSDARG00000100422\n",
       "182                  eps8  ENSDARG00000102128\n",
       "...                   ...                 ...\n",
       "1149224    CABZ01111913.1  ENSDARG00000100589\n",
       "1149229    CABZ01079745.1  ENSDARG00000099551\n",
       "1149240               NP5             ENSGNP5\n",
       "1149246              NP31            ENSGNP31\n",
       "1149249              NP33            ENSGNP33\n",
       "\n",
       "[32192 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dbxref_datafield = 'gene_id'\n",
    "\n",
    "models_subset = models[['gene_name', dbxref_datafield]].dropna().drop_duplicates()\n",
    "\n",
    "display(models_subset)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Generate gene list file to query Uniprot ID Mapping API\n",
    "Generate a text file ending in `_ids.txt` for sending to the ID mapping API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 32192 gene ids to /home/ec2-user/glial-origins/output/Drer_juvbrain/Drer_juvbrain_gene_id_ids.txt\n"
     ]
    }
   ],
   "source": [
    "gene_list = models_subset[dbxref_datafield].unique()\n",
    "\n",
    "genelist_object = GeneListFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    sources = [sample_BFD.annot],\n",
    "    genes = gene_list,\n",
    "    identifier = dbxref_datafield\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Query Uniprot ID Mapping API\n",
    "Specify the `from_type` variable based on the Uniprot name of the identifier.  \n",
    "The table below lists some databases and the `from_type` string that the API accepts for that datatype.  \n",
    "\n",
    "| datatype | `from_type` string | description |\n",
    "| ---: | :--- | :--- |\n",
    "| Mouse Genome Informatics | `MGI` | ID starts with `MGI:` |\n",
    "| Zebrafish Information Network | `ZFIN` | ID starts with `ZDB-GENE-` |\n",
    "| Xenbase | `Xenbase` | ID starts with `XB-GENE-` |\n",
    "\n",
    "__NOTE:__ You may have to run the cell below twice - UniProt sometimes throws an \"Resource not found\" message on the first query to the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  597k    0    52  100  597k     20   230k  0:00:02  0:00:02 --:--:--  230k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100    24    0    24    0     0     31      0 --:--:-- --:--:-- --:--:--    31\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>From</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "      <td>A0A1L1QZP7</td>\n",
       "      <td>A0A1L1QZP7_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Si:cabz01085950.1</td>\n",
       "      <td>si:cabz01085950.1</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSDARG00000100660</td>\n",
       "      <td>A0A0R4IH74</td>\n",
       "      <td>A0A0R4IH74_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Si:ch73-252i11.1</td>\n",
       "      <td>si:ch73-252i11.1</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSDARG00000100422</td>\n",
       "      <td>A0A0R4IIG9</td>\n",
       "      <td>A0A0R4IIG9_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Protein tyrosine phosphatase receptor type O</td>\n",
       "      <td>ptpro</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSDARG00000102128</td>\n",
       "      <td>A0A0R4IID9</td>\n",
       "      <td>A0A0R4IID9_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Epidermal growth factor receptor pathway subst...</td>\n",
       "      <td>eps8 zgc:56041 eps8a</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSDARG00000103095</td>\n",
       "      <td>Q0GKD6</td>\n",
       "      <td>Q0GKD6_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>TANK-binding kinase 1 (serine/threonine-protei...</td>\n",
       "      <td>tbk1 wu:fk70c05 zgc:136548 tbk1 SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39477</th>\n",
       "      <td>ENSDARG00000104659</td>\n",
       "      <td>A0A0G2KLM5</td>\n",
       "      <td>A0A0G2KLM5_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>tryptophan--tRNA ligase (EC 6.1.1.2) (Tryptoph...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39478</th>\n",
       "      <td>ENSDARG00000101098</td>\n",
       "      <td>A0A0G2KE88</td>\n",
       "      <td>A0A0G2KE88_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Lysosomal associated membrane protein family m...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39479</th>\n",
       "      <td>ENSDARG00000101098</td>\n",
       "      <td>A0A0G2L379</td>\n",
       "      <td>A0A0G2L379_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Lysosomal associated membrane protein family m...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39480</th>\n",
       "      <td>ENSDARG00000103574</td>\n",
       "      <td>H9GYN8</td>\n",
       "      <td>H9GYN8_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Rhodopsin</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39481</th>\n",
       "      <td>ENSDARG00000101185</td>\n",
       "      <td>A0A0G2L330</td>\n",
       "      <td>A0A0G2L330_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>RIC3 domain-containing protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>237</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39482 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     From       Entry        Entry Name    Reviewed  \\\n",
       "0      ENSDARG00000104632  A0A1L1QZP7  A0A1L1QZP7_DANRE  unreviewed   \n",
       "1      ENSDARG00000100660  A0A0R4IH74  A0A0R4IH74_DANRE  unreviewed   \n",
       "2      ENSDARG00000100422  A0A0R4IIG9  A0A0R4IIG9_DANRE  unreviewed   \n",
       "3      ENSDARG00000102128  A0A0R4IID9  A0A0R4IID9_DANRE  unreviewed   \n",
       "4      ENSDARG00000103095      Q0GKD6      Q0GKD6_DANRE  unreviewed   \n",
       "...                   ...         ...               ...         ...   \n",
       "39477  ENSDARG00000104659  A0A0G2KLM5  A0A0G2KLM5_DANRE  unreviewed   \n",
       "39478  ENSDARG00000101098  A0A0G2KE88  A0A0G2KE88_DANRE  unreviewed   \n",
       "39479  ENSDARG00000101098  A0A0G2L379  A0A0G2L379_DANRE  unreviewed   \n",
       "39480  ENSDARG00000103574      H9GYN8      H9GYN8_DANRE  unreviewed   \n",
       "39481  ENSDARG00000101185  A0A0G2L330  A0A0G2L330_DANRE  unreviewed   \n",
       "\n",
       "                                           Protein names  \\\n",
       "0                                      Si:cabz01085950.1   \n",
       "1                                       Si:ch73-252i11.1   \n",
       "2           Protein tyrosine phosphatase receptor type O   \n",
       "3      Epidermal growth factor receptor pathway subst...   \n",
       "4      TANK-binding kinase 1 (serine/threonine-protei...   \n",
       "...                                                  ...   \n",
       "39477  tryptophan--tRNA ligase (EC 6.1.1.2) (Tryptoph...   \n",
       "39478  Lysosomal associated membrane protein family m...   \n",
       "39479  Lysosomal associated membrane protein family m...   \n",
       "39480                                          Rhodopsin   \n",
       "39481                     RIC3 domain-containing protein   \n",
       "\n",
       "                                       Gene Names  \\\n",
       "0                               si:cabz01085950.1   \n",
       "1                                si:ch73-252i11.1   \n",
       "2                                           ptpro   \n",
       "3                            eps8 zgc:56041 eps8a   \n",
       "4      tbk1 wu:fk70c05 zgc:136548 tbk1 SO:0001217   \n",
       "...                                           ...   \n",
       "39477                                         NaN   \n",
       "39478                                         NaN   \n",
       "39479                                         NaN   \n",
       "39480                                         NaN   \n",
       "39481                                         NaN   \n",
       "\n",
       "                                          Organism  Length  \n",
       "0      Danio rerio (Zebrafish) (Brachydanio rerio)     184  \n",
       "1      Danio rerio (Zebrafish) (Brachydanio rerio)     520  \n",
       "2      Danio rerio (Zebrafish) (Brachydanio rerio)     866  \n",
       "3      Danio rerio (Zebrafish) (Brachydanio rerio)     804  \n",
       "4      Danio rerio (Zebrafish) (Brachydanio rerio)     727  \n",
       "...                                            ...     ...  \n",
       "39477  Danio rerio (Zebrafish) (Brachydanio rerio)     170  \n",
       "39478  Danio rerio (Zebrafish) (Brachydanio rerio)     213  \n",
       "39479  Danio rerio (Zebrafish) (Brachydanio rerio)     213  \n",
       "39480  Danio rerio (Zebrafish) (Brachydanio rerio)     168  \n",
       "39481  Danio rerio (Zebrafish) (Brachydanio rerio)     237  \n",
       "\n",
       "[39482 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from_type = 'Ensembl'\n",
    "to_type = 'UniProtKB'\n",
    "\n",
    "uniprot_idmm_object = genelist_object.get_uniprot_ids(ID_MAPPER_LOC, from_type, to_type)\n",
    "sample_BFD.add_keyfile('uniprot_output_idmm', uniprot_idmm_object, overwrite = True)\n",
    "\n",
    "uniprot_idmm = pd.read_csv(sample_BFD.uniprot_output_idmm.path, sep = '\\t')\n",
    "display(uniprot_idmm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Extract results and generate Uniprot IDMM\n",
    "Generates an idmm that links `gene_name`, the `dbxref_datafield` seleted above, and `uniprot_id` returned by API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_14128/3311337425.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  uniprot_idpairs.rename(columns = {'From': dbxref_datafield, 'Entry': 'uniprot_id'}, inplace = True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "      <td>A0A1L1QZP7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSDARG00000100660</td>\n",
       "      <td>A0A0R4IH74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSDARG00000100422</td>\n",
       "      <td>A0A0R4IIG9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSDARG00000102128</td>\n",
       "      <td>A0A0R4IID9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSDARG00000103095</td>\n",
       "      <td>Q0GKD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39477</th>\n",
       "      <td>ENSDARG00000104659</td>\n",
       "      <td>A0A0G2KLM5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39478</th>\n",
       "      <td>ENSDARG00000101098</td>\n",
       "      <td>A0A0G2KE88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39479</th>\n",
       "      <td>ENSDARG00000101098</td>\n",
       "      <td>A0A0G2L379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39480</th>\n",
       "      <td>ENSDARG00000103574</td>\n",
       "      <td>H9GYN8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39481</th>\n",
       "      <td>ENSDARG00000101185</td>\n",
       "      <td>A0A0G2L330</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39482 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_id  uniprot_id\n",
       "0      ENSDARG00000104632  A0A1L1QZP7\n",
       "1      ENSDARG00000100660  A0A0R4IH74\n",
       "2      ENSDARG00000100422  A0A0R4IIG9\n",
       "3      ENSDARG00000102128  A0A0R4IID9\n",
       "4      ENSDARG00000103095      Q0GKD6\n",
       "...                   ...         ...\n",
       "39477  ENSDARG00000104659  A0A0G2KLM5\n",
       "39478  ENSDARG00000101098  A0A0G2KE88\n",
       "39479  ENSDARG00000101098  A0A0G2L379\n",
       "39480  ENSDARG00000103574      H9GYN8\n",
       "39481  ENSDARG00000101185  A0A0G2L330\n",
       "\n",
       "[39482 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "      <td>A0A1L1QZP7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PARP12</td>\n",
       "      <td>ENSDARG00000100660</td>\n",
       "      <td>A0A0R4IH74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ptpro</td>\n",
       "      <td>ENSDARG00000100422</td>\n",
       "      <td>A0A0R4IIG9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>eps8</td>\n",
       "      <td>ENSDARG00000102128</td>\n",
       "      <td>A0A0R4IID9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tbk1</td>\n",
       "      <td>ENSDARG00000103095</td>\n",
       "      <td>Q0GKD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39477</th>\n",
       "      <td>CABZ01109604.1</td>\n",
       "      <td>ENSDARG00000104659</td>\n",
       "      <td>A0A0G2KLM5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39478</th>\n",
       "      <td>LAMP5</td>\n",
       "      <td>ENSDARG00000101098</td>\n",
       "      <td>A0A0G2KE88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39479</th>\n",
       "      <td>LAMP5</td>\n",
       "      <td>ENSDARG00000101098</td>\n",
       "      <td>A0A0G2L379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39480</th>\n",
       "      <td>CABZ01109058.1</td>\n",
       "      <td>ENSDARG00000103574</td>\n",
       "      <td>H9GYN8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39481</th>\n",
       "      <td>CABZ01095001.1</td>\n",
       "      <td>ENSDARG00000101185</td>\n",
       "      <td>A0A0G2L330</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39482 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_name             gene_id  uniprot_id\n",
       "0      si:ch73-252i11.3  ENSDARG00000104632  A0A1L1QZP7\n",
       "1                PARP12  ENSDARG00000100660  A0A0R4IH74\n",
       "2                 ptpro  ENSDARG00000100422  A0A0R4IIG9\n",
       "3                  eps8  ENSDARG00000102128  A0A0R4IID9\n",
       "4                  tbk1  ENSDARG00000103095      Q0GKD6\n",
       "...                 ...                 ...         ...\n",
       "39477    CABZ01109604.1  ENSDARG00000104659  A0A0G2KLM5\n",
       "39478             LAMP5  ENSDARG00000101098  A0A0G2KE88\n",
       "39479             LAMP5  ENSDARG00000101098  A0A0G2L379\n",
       "39480    CABZ01109058.1  ENSDARG00000103574      H9GYN8\n",
       "39481    CABZ01095001.1  ENSDARG00000101185  A0A0G2L330\n",
       "\n",
       "[39482 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "uniprot_idpairs = uniprot_idmm[['From', 'Entry']]\n",
    "uniprot_idpairs.rename(columns = {'From': dbxref_datafield, 'Entry': 'uniprot_id'}, inplace = True)\n",
    "display(uniprot_idpairs)\n",
    "\n",
    "uniprot_output_idmm = models_subset.merge(uniprot_idpairs, on = dbxref_datafield)\n",
    "display(uniprot_output_idmm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate a filename and file for the idmm\n",
    "uniprot_output_idmm_filename = '_'.join([sample_BFD.species_prefix, conditions, 'uniprot-idmm.tsv'])\n",
    "uniprot_output_idmm_object = IdmmFile(\n",
    "    filename = uniprot_output_idmm_filename, \n",
    "    sampledict = sample_BFD.sampledict, \n",
    "    kind = 'uniprot_idmm', \n",
    "    sources = [sample_BFD.annot]\n",
    ")\n",
    "\n",
    "# save to file and add to the BioFileDocket\n",
    "uniprot_output_idmm.to_csv(uniprot_output_idmm_object.path, sep = '\\t')\n",
    "sample_BFD.add_keyfile('uniprot_idmm', uniprot_output_idmm_object)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8. Convert GFF to GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_14128/853027043.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  models_asgtf_df = pd.read_csv(sample_BFD.annot.path, skiprows = 5, header = None, sep = '\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6733</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>6733</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>52002</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>48613</td>\n",
       "      <td>48740</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>13755</td>\n",
       "      <td>13811</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149247</th>\n",
       "      <td>22</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>598076</td>\n",
       "      <td>599835</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149248</th>\n",
       "      <td>22</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>598076</td>\n",
       "      <td>599835</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149249</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; gene_nam...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149250</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149251</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1149252 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0               1           2         3         4  5  6  7  \\\n",
       "0         4  ensembl_havana        gene      6733     52120  .  -  .   \n",
       "1         4         ensembl  transcript      6733     52120  .  -  .   \n",
       "2         4         ensembl        exon     52002     52120  .  -  .   \n",
       "3         4         ensembl        exon     48613     48740  .  -  .   \n",
       "4         4         ensembl        exon     13755     13811  .  -  .   \n",
       "...      ..             ...         ...       ...       ... .. .. ..   \n",
       "1149247  22         ensembl        gene    598076    599835  .  +  .   \n",
       "1149248  22         ensembl        gene    598076    599835  .  +  .   \n",
       "1149249  16         ensembl        gene  47449630  47452948  .  +  .   \n",
       "1149250  16         ensembl  transcript  47449630  47452948  .  +  .   \n",
       "1149251  16         ensembl        exon  47449630  47452948  .  +  .   \n",
       "\n",
       "                                                         8  \n",
       "0        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...  \n",
       "1        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...  \n",
       "2        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...  \n",
       "3        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...  \n",
       "4        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...  \n",
       "...                                                    ...  \n",
       "1149247  gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...  \n",
       "1149248  gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...  \n",
       "1149249  gene_id \"ENSGNP33\"; gene_version \"1\"; gene_nam...  \n",
       "1149250  gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...  \n",
       "1149251  gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...  \n",
       "\n",
       "[1149252 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'gene_id \"ENSDARG00000104632\"; gene_version \"1\"; transcript_id \"ENSDART00000166186\"; transcript_version \"1\"; gene_name \"si:ch73-252i11.3\"; gene_source \"ensembl_havana\"; gene_biotype \"lincRNA\"; havana_gene \"OTTDARG00000037780\"; havana_gene_version \"1\"; transcript_name \"si:ch73-252i11.3-201\"; transcript_source \"ensembl\"; transcript_biotype \"lincRNA\"; tag \"basic\";'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load the newly-generated GTF file as a dataframe\n",
    "models_asgtf_df = pd.read_csv(sample_BFD.annot.path, skiprows = 5, header = None, sep = '\\t')\n",
    "\n",
    "display(models_asgtf_df)\n",
    "display(models_asgtf_df[attributes_column][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6733</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "      <td>{'gene_id': 'ENSDARG00000104632', 'gene_versio...</td>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>6733</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "      <td>{'gene_id': 'ENSDARG00000104632', 'gene_versio...</td>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "      <td>ENSDART00000166186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>52002</td>\n",
       "      <td>52120</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "      <td>{'gene_id': 'ENSDARG00000104632', 'gene_versio...</td>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "      <td>ENSDART00000166186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>48613</td>\n",
       "      <td>48740</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "      <td>{'gene_id': 'ENSDARG00000104632', 'gene_versio...</td>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "      <td>ENSDART00000166186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>13755</td>\n",
       "      <td>13811</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSDARG00000104632\"; gene_version \"1\"...</td>\n",
       "      <td>{'gene_id': 'ENSDARG00000104632', 'gene_versio...</td>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "      <td>ENSDART00000166186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149247</th>\n",
       "      <td>22</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>598076</td>\n",
       "      <td>599835</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...</td>\n",
       "      <td>{'gene_id': 'ENSGNP31', 'gene_version': '1', '...</td>\n",
       "      <td>NP31</td>\n",
       "      <td>ENSGNP31</td>\n",
       "      <td>ENSTNP31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149248</th>\n",
       "      <td>22</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>598076</td>\n",
       "      <td>599835</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...</td>\n",
       "      <td>{'gene_id': 'ENSGNP31', 'gene_version': '1', '...</td>\n",
       "      <td>NP31</td>\n",
       "      <td>ENSGNP31</td>\n",
       "      <td>ENSTNP31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149249</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; gene_nam...</td>\n",
       "      <td>{'gene_id': 'ENSGNP33', 'gene_version': '1', '...</td>\n",
       "      <td>NP33</td>\n",
       "      <td>ENSGNP33</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149250</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...</td>\n",
       "      <td>{'gene_id': 'ENSGNP33', 'gene_version': '1', '...</td>\n",
       "      <td>NP33</td>\n",
       "      <td>ENSGNP33</td>\n",
       "      <td>ENSTNP33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149251</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>47449630</td>\n",
       "      <td>47452948</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...</td>\n",
       "      <td>{'gene_id': 'ENSGNP33', 'gene_version': '1', '...</td>\n",
       "      <td>NP33</td>\n",
       "      <td>ENSGNP33</td>\n",
       "      <td>ENSTNP33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>733994 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0               1           2         3         4  5  6  7  \\\n",
       "0         4  ensembl_havana        gene      6733     52120  .  -  .   \n",
       "1         4         ensembl  transcript      6733     52120  .  -  .   \n",
       "2         4         ensembl        exon     52002     52120  .  -  .   \n",
       "3         4         ensembl        exon     48613     48740  .  -  .   \n",
       "4         4         ensembl        exon     13755     13811  .  -  .   \n",
       "...      ..             ...         ...       ...       ... .. .. ..   \n",
       "1149247  22         ensembl        gene    598076    599835  .  +  .   \n",
       "1149248  22         ensembl        gene    598076    599835  .  +  .   \n",
       "1149249  16         ensembl        gene  47449630  47452948  .  +  .   \n",
       "1149250  16         ensembl  transcript  47449630  47452948  .  +  .   \n",
       "1149251  16         ensembl        exon  47449630  47452948  .  +  .   \n",
       "\n",
       "                                                         8  \\\n",
       "0        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...   \n",
       "1        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...   \n",
       "2        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...   \n",
       "3        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...   \n",
       "4        gene_id \"ENSDARG00000104632\"; gene_version \"1\"...   \n",
       "...                                                    ...   \n",
       "1149247  gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...   \n",
       "1149248  gene_id \"ENSGNP31\"; gene_version \"1\"; transcri...   \n",
       "1149249  gene_id \"ENSGNP33\"; gene_version \"1\"; gene_nam...   \n",
       "1149250  gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...   \n",
       "1149251  gene_id \"ENSGNP33\"; gene_version \"1\"; transcri...   \n",
       "\n",
       "                                          field_dictionary         gene_name  \\\n",
       "0        {'gene_id': 'ENSDARG00000104632', 'gene_versio...  si:ch73-252i11.3   \n",
       "1        {'gene_id': 'ENSDARG00000104632', 'gene_versio...  si:ch73-252i11.3   \n",
       "2        {'gene_id': 'ENSDARG00000104632', 'gene_versio...  si:ch73-252i11.3   \n",
       "3        {'gene_id': 'ENSDARG00000104632', 'gene_versio...  si:ch73-252i11.3   \n",
       "4        {'gene_id': 'ENSDARG00000104632', 'gene_versio...  si:ch73-252i11.3   \n",
       "...                                                    ...               ...   \n",
       "1149247  {'gene_id': 'ENSGNP31', 'gene_version': '1', '...              NP31   \n",
       "1149248  {'gene_id': 'ENSGNP31', 'gene_version': '1', '...              NP31   \n",
       "1149249  {'gene_id': 'ENSGNP33', 'gene_version': '1', '...              NP33   \n",
       "1149250  {'gene_id': 'ENSGNP33', 'gene_version': '1', '...              NP33   \n",
       "1149251  {'gene_id': 'ENSGNP33', 'gene_version': '1', '...              NP33   \n",
       "\n",
       "                    gene_id       transcript_id  \n",
       "0        ENSDARG00000104632                None  \n",
       "1        ENSDARG00000104632  ENSDART00000166186  \n",
       "2        ENSDARG00000104632  ENSDART00000166186  \n",
       "3        ENSDARG00000104632  ENSDART00000166186  \n",
       "4        ENSDARG00000104632  ENSDART00000166186  \n",
       "...                     ...                 ...  \n",
       "1149247            ENSGNP31            ENSTNP31  \n",
       "1149248            ENSGNP31            ENSTNP31  \n",
       "1149249            ENSGNP33                None  \n",
       "1149250            ENSGNP33            ENSTNP33  \n",
       "1149251            ENSGNP33            ENSTNP33  \n",
       "\n",
       "[733994 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Use a custom function to extract useful fields from the additional fields section (column 8)\n",
    "# Pull from that dict to fill in additional useful columns\n",
    "models_asgtf_df['field_dictionary'] = models_asgtf_df[attributes_column].apply(convert_fields_to_dict_gtf)\n",
    "models_asgtf_df['gene_name'] = [d.get('gene_name') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['gene_id'] = [d.get('gene_id') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['transcript_id'] = [d.get('transcript_id') for d in models_asgtf_df['field_dictionary']]\n",
    "\n",
    "# Remove CDS annotations because they interfere with TransDecoder cDNA generation\n",
    "models_asgtf_df = models_asgtf_df[models_asgtf_df[2] != 'CDS']\n",
    "display(models_asgtf_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9. Generate gtf-idmm\n",
    "This file maps the `gene_name` to `gene_id` and `transcript_id` fields generated by the conversion from GFF to GTF, which will be needed for downstream processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "      <td>ENSDART00000166186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>si:ch73-252i11.3</td>\n",
       "      <td>ENSDARG00000104632</td>\n",
       "      <td>ENSDART00000170987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>PARP12</td>\n",
       "      <td>ENSDARG00000100660</td>\n",
       "      <td>ENSDART00000166693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>PARP12</td>\n",
       "      <td>ENSDARG00000100660</td>\n",
       "      <td>ENSDART00000166174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>syn3</td>\n",
       "      <td>ENSDARG00000098417</td>\n",
       "      <td>ENSDART00000157825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149225</th>\n",
       "      <td>CABZ01111913.1</td>\n",
       "      <td>ENSDARG00000100589</td>\n",
       "      <td>ENSDART00000163136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149230</th>\n",
       "      <td>CABZ01079745.1</td>\n",
       "      <td>ENSDARG00000099551</td>\n",
       "      <td>ENSDART00000165750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149241</th>\n",
       "      <td>NP5</td>\n",
       "      <td>ENSGNP5</td>\n",
       "      <td>ENSTNP5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149247</th>\n",
       "      <td>NP31</td>\n",
       "      <td>ENSGNP31</td>\n",
       "      <td>ENSTNP31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1149250</th>\n",
       "      <td>NP33</td>\n",
       "      <td>ENSGNP33</td>\n",
       "      <td>ENSTNP33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58277 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene_name             gene_id       transcript_id\n",
       "1        si:ch73-252i11.3  ENSDARG00000104632  ENSDART00000166186\n",
       "7        si:ch73-252i11.3  ENSDARG00000104632  ENSDART00000170987\n",
       "11                 PARP12  ENSDARG00000100660  ENSDART00000166693\n",
       "38                 PARP12  ENSDARG00000100660  ENSDART00000166174\n",
       "58                   syn3  ENSDARG00000098417  ENSDART00000157825\n",
       "...                   ...                 ...                 ...\n",
       "1149225    CABZ01111913.1  ENSDARG00000100589  ENSDART00000163136\n",
       "1149230    CABZ01079745.1  ENSDARG00000099551  ENSDART00000165750\n",
       "1149241               NP5             ENSGNP5             ENSTNP5\n",
       "1149247              NP31            ENSGNP31            ENSTNP31\n",
       "1149250              NP33            ENSGNP33            ENSTNP33\n",
       "\n",
       "[58277 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract gene_name, gene_id, and transcript_id fields to generate an ID mapping matrix (idmm)\n",
    "idmm_df = models_asgtf_df[['gene_name', 'gene_id', 'transcript_id']].drop_duplicates()\n",
    "idmm_df.dropna(inplace = True)\n",
    "display(idmm_df)\n",
    "\n",
    "# generate a filename and file for the idmm\n",
    "idmm_filename = '_'.join([sample_BFD.species_prefix, conditions, 'gtf-idmm.tsv'])\n",
    "idmm = IdmmFile(\n",
    "    filename = idmm_filename, \n",
    "    sampledict = sample_BFD.sampledict, \n",
    "    kind = 'gtf_idmm', \n",
    "    sources = [sample_BFD.annot]\n",
    ")\n",
    "\n",
    "# save to file and add to the BioFileDocket\n",
    "idmm_df.to_csv(idmm.path, sep = '\\t')\n",
    "sample_BFD.add_keyfile('gtf_idmm', idmm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 10. Generate updated gtf\n",
    "Generated an updated GTF file using transcript_id as the key. For some datasets, transcripts do not consistently get gene names and gene IDs added, which causes Transdecoder to throw errors. This resolves that problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "models_asgtf_updated_df = models_asgtf_df.merge(idmm_df, on = 'transcript_id')\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_name': x['gene_name_y']}), axis = 1)\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_id': x['gene_id_y']}), axis = 1)\n",
    "models_asgtf_updated_df['new_fields'] = models_asgtf_updated_df['field_dictionary'].apply(convert_dict_to_fields_gtf)\n",
    "models_asgtf_updated_df = models_asgtf_updated_df[[0, 1, 2, 3, 4, 5, 6, 7, 'new_fields']]\n",
    "\n",
    "models_asgtf_updated_filename = sample_BFD.annot.filename.replace('.gtf', '_updated.gtf')\n",
    "models_asgtf_updated = GenomeGtfFile(\n",
    "    filename = models_asgtf_updated_filename, \n",
    "    sampledict = sample_BFD.sampledict, \n",
    "    reference_genome = sample_BFD.genome_fasta\n",
    ")\n",
    "\n",
    "models_asgtf_updated_df.to_csv(models_asgtf_updated.path, header = None, index = None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 11. Generate cDNA and peptide files\n",
    "Using the updated gtf file and genome file, generate cDNA sequence.\n",
    "\n",
    "Then, using the cDNA sequence, generate peptide sequences using transdecoder.\n",
    "\n",
    "Expect this step to take some time, probably ~20-30min."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/compute_base_probs.pl /home/ec2-user/glial-origins/output/Drer_juvbrain/GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa 0 > /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//base_freqs.dat, checkpoint [/home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir/.__checkpoints_longorfs/base_freqs_file.ok] exists.\n",
      "-skipping long orf extraction, already completed earlier as per checkpoint: /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir/.__checkpoints_longorfs/TD.longorfs.ok\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//longest_orfs.cds 5000 5000 > /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000, checkpoint [/home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir/.__checkpoints/get_longest_orfs.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/exclude_similar_proteins.pl /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000 > /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr, checkpoint [/home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir/.__checkpoints/nr.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr 500 > /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//longest_orfs.cds.top_500_longest, checkpoint [/home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir/.__checkpoints/top_train_select.ok] exists.\n",
      "PCT_GC: 46\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/seq_n_baseprobs_to_loglikelihood_vals.pl /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//longest_orfs.cds.top_500_longest /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//base_freqs.dat > /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//hexamer.scores, checkpoint [/home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir/.__checkpoints/hexamer_scores.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/score_CDS_likelihood_all_6_frames.pl /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//longest_orfs.cds /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//hexamer.scores > /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//longest_orfs.cds.scores, checkpoint [/home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir/.__checkpoints/score_cdss.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/select_best_ORFs_per_transcript.pl --gff3_file /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//longest_orfs.gff3 --cds_scores /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//longest_orfs.cds.scores  --min_length_auto_accept 749  > /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 , checkpoint [/home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir/.__checkpoints/select_best_orfs.5f11f170332a038f4a01e61feb8882f7.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/train_start_PWM.pl --transcripts /home/ec2-user/glial-origins/output/Drer_juvbrain/GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa --selected_orfs /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//longest_orfs.cds.top_500_longest --out_prefix /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//start_refinement, checkpoint [/home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir/.__checkpoints/train_start_PWM.5f11f170332a038f4a01e61feb8882f7.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/start_codon_refinement.pl --transcripts /home/ec2-user/glial-origins/output/Drer_juvbrain/GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa --gff3_file /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 --workdir /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir/ > /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3, checkpoint [/home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir/.__checkpoints/revise_starts.5f11f170332a038f4a01e61feb8882f7.ok] exists.\n",
      "-- Skipping CMD: cp /home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3 GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa.transdecoder.gff3, checkpoint [/home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir/.__checkpoints/copy_to_final_outputfile.5f11f170332a038f4a01e61feb8882f7.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_bed.pl GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa.transdecoder.gff3 > GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa.transdecoder.bed, checkpoint [/home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir/.__checkpoints/make_final_bed.5f11f170332a038f4a01e61feb8882f7.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa.transdecoder.gff3 --fasta /home/ec2-user/glial-origins/output/Drer_juvbrain/GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa  --genetic_code Universal > GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa.transdecoder.pep, checkpoint [/home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir/.__checkpoints/make_final_pep.5f11f170332a038f4a01e61feb8882f7.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa.transdecoder.gff3 --fasta /home/ec2-user/glial-origins/output/Drer_juvbrain/GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa --seqType CDS  --genetic_code Universal > GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa.transdecoder.cds, checkpoint [/home/ec2-user/glial-origins/output/Drer_juvbrain/Danio_rerio_juvbrain.transdecoder_dir/.__checkpoints/make_final_cds.5f11f170332a038f4a01e61feb8882f7.ok] exists.\n",
      "transdecoder is finished.  See output files GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa.transdecoder.*\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cdna = sample_BFD.genome_fasta.get_transdecoder_cdna_gtf(models_asgtf_updated, TRANSDECODER_LOC)\n",
    "sample_BFD.add_keyfile('cdna', cdna)\n",
    "\n",
    "transdecoder_files = sample_BFD.cdna.to_pep_files(TDLONGORF_LOC, TDPREDICT_LOC)\n",
    "sample_BFD.add_keyfiles(transdecoder_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "iEug6TQ6Cc9D"
   },
   "source": [
    "# 12. Push files to AWS S3\n",
    "\n",
    "Iteratively moves through BioFileDocket BioFile objects and pushes to the right place in AWS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upload: ../../output/Drer_juvbrain/GSE105010_Danio_rerio.GRCz10.86.modified.gtf to s3://arcadia-reference-datasets/organisms/Danio_rerio/genomics_reference/annotation/GSE105010_Danio_rerio.GRCz10.86.modified.gtf\n",
      "upload: ../../output/Drer_juvbrain/GCA_000002035.3_GRCz10_genomic.fna.renamed.fa to s3://arcadia-reference-datasets/organisms/Danio_rerio/genomics_reference/genome/GCA_000002035.3_GRCz10_genomic.fna.renamed.fa\n",
      "upload: ../../output/Drer_juvbrain/GCA_000002035.3_GRCz10_genomic.fna to s3://arcadia-reference-datasets/organisms/Danio_rerio/genomics_reference/genome/GCA_000002035.3_GRCz10_genomic.fna\n",
      "upload: ../../output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-gxc.tsv to s3://arcadia-reference-datasets/organisms/Danio_rerio/functional_sequencing/scRNA-Seq/GSM4793240_comb-zBr15dpf6_S1-gxc.tsv\n",
      "upload: ../../output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-genes.tsv to s3://arcadia-reference-datasets/organisms/Danio_rerio/functional_sequencing/scRNA-Seq/GSM4793240_comb-zBr15dpf6_S1-genes.tsv\n",
      "upload: ../../output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-matrix.mtx to s3://arcadia-reference-datasets/organisms/Danio_rerio/functional_sequencing/scRNA-Seq/GSM4793240_comb-zBr15dpf6_S1-matrix.mtx\n",
      "upload: ../../output/Drer_juvbrain/GSM4793240_comb-zBr15dpf6_S1-barcodes.tsv to s3://arcadia-reference-datasets/organisms/Danio_rerio/functional_sequencing/scRNA-Seq/GSM4793240_comb-zBr15dpf6_S1-barcodes.tsv\n",
      "upload: ../../output/Drer_juvbrain/Drer_juvbrain_gene_id_UniProtIDs.txt to s3://arcadia-reference-datasets/organisms/Danio_rerio/genomics_reference/mapping_file/Drer_juvbrain_gene_id_UniProtIDs.txt\n",
      "upload: ../../output/Drer_juvbrain/Drer_juvbrain_uniprot-idmm.tsv to s3://arcadia-reference-datasets/organisms/Danio_rerio/genomics_reference/mapping_file/Drer_juvbrain_uniprot-idmm.tsv\n",
      "upload: ../../output/Drer_juvbrain/Drer_juvbrain_gtf-idmm.tsv to s3://arcadia-reference-datasets/organisms/Danio_rerio/genomics_reference/mapping_file/Drer_juvbrain_gtf-idmm.tsv\n",
      "upload: ../../output/Drer_juvbrain/GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa to s3://arcadia-reference-datasets/organisms/Danio_rerio/genomics_reference/transcriptome/GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa\n",
      "upload: ../../output/Drer_juvbrain/GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa.transdecoder.bed to s3://arcadia-reference-datasets/organisms/Danio_rerio/genomics_reference/proteome/GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa.transdecoder.bed\n",
      "upload: ../../output/Drer_juvbrain/GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa.transdecoder.cds to s3://arcadia-reference-datasets/organisms/Danio_rerio/genomics_reference/proteome/GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa.transdecoder.cds\n",
      "upload: ../../output/Drer_juvbrain/GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa.transdecoder.gff3 to s3://arcadia-reference-datasets/organisms/Danio_rerio/genomics_reference/proteome/GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa.transdecoder.gff3\n",
      "upload: ../../output/Drer_juvbrain/GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa.transdecoder.pep to s3://arcadia-reference-datasets/organisms/Danio_rerio/genomics_reference/proteome/GCA_000002035.3_GRCz10_genomic.fna.renamed_cDNA.fa.transdecoder.pep\n"
     ]
    }
   ],
   "source": [
    "sample_BFD.local_to_s3()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 13. Pickle the `sample_BFD` variable for use by the next script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upload: ../../output/Drer_juvbrain/Drer_juvbrain_BioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/Drer_juvbrain_BioFileDocket.pkl\n"
     ]
    }
   ],
   "source": [
    "# Generate a .pkl file for the Docket\n",
    "sample_BFD.pickle()\n",
    "\n",
    "# Push to S3, optionally overwriting existing file\n",
    "sample_BFD.push_to_s3(overwrite = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "Xtro_bro.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
