{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "L56fjA27DmhA"
   },
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "t0BTCsz7Y48y",
    "outputId": "f6510678-f977-4ec8-ce30-1f2acd4fe666"
   },
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import os\n",
    "import dill\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "import sys\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Download and describe data\n",
    "\n",
    "\n",
    "This notebook collects data from the  .\n",
    "\n",
    "Dataset description\n",
    "Data can be found at the .\n",
    "This notebook only collects data from one sample, \"Brain1\", for the purposes of data analysis and exploration.\n",
    "\n",
    "This notebook collects data from the [Jiang et al. 2021](https://www.frontiersin.org/articles/10.3389/fcell.2021.743421/full) zebrafish cell atlas.\n",
    "\n",
    "## Dataset description\n",
    "\n",
    "- Data can be found at the [GEO Accession GSE130487](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE130487).\n",
    "- This notebook previously collects data from one sample, [\"Brain8\"](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3768152), for the purposes of data analysis and exploration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'species': 'Danio_rerio',\n",
       " 'conditions': 'adultbrain',\n",
       " 'directory': '../../output/Drer_adultbrain/',\n",
       " 'files': {},\n",
       " 'annot': <biofile_handling.GenomeGffFile at 0x7f73d40ea8f0>,\n",
       " 'genome_fasta': <biofile_handling.GenomeFastaFile at 0x7f73d40eb850>,\n",
       " 'gxc': <biofile_handling.GxcFile at 0x7f73d40eba60>}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify the name of the species folder in Amazon S3\n",
    "species = 'Danio_rerio'\n",
    "\n",
    "# Specify any particular identifying conditions, eg tissue type:\n",
    "conditions = 'adultbrain'\n",
    "\n",
    "# Specify url and other variables\n",
    "genome_fasta_url = 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/035/GCF_000002035.5_GRCz10/GCF_000002035.5_GRCz10_genomic.fna.gz'\n",
    "genome_version = 'GRCz10'\n",
    "\n",
    "annot_url = 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/035/GCF_000002035.5_GRCz10/GCF_000002035.5_GRCz10_genomic.gff.gz'\n",
    "gxc_url = 'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3768nnn/GSM3768152/suppl/GSM3768152_Brain_8_dge.txt.gz'\n",
    "\n",
    "###########\n",
    "# runtime #\n",
    "###########\n",
    "\n",
    "protocol = 'curl'\n",
    "\n",
    "species_prefix = prefixify(species)\n",
    "\n",
    "# Specify folder as destination for file downloads\n",
    "output_folder = '../../output/' + prefixify(species) + '_' + conditions + '/'\n",
    "\n",
    "if not os.path.exists(output_folder):\n",
    "    os.mkdir(output_folder)\n",
    "    \n",
    "species_SampleDict = SampleDict(species, conditions, output_folder)\n",
    "\n",
    "genome_fasta = GenomeFastaFile(\n",
    "    filename = '',\n",
    "    sampledict = species_SampleDict,\n",
    "    version = genome_version,\n",
    "    url = genome_fasta_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "annot = GenomeGffFile(\n",
    "    filename = '',\n",
    "    sampledict = species_SampleDict,\n",
    "    GenomeFastaFile = genome_fasta,\n",
    "    url = annot_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "gxc = GxcFile(\n",
    "    filename = '',\n",
    "    sampledict = species_SampleDict,\n",
    "    GenomeFastaFile = genome_fasta,\n",
    "    GenomeAnnotFile = annot,\n",
    "    url = gxc_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "\n",
    "sample_BioFileDocket = BioFileDocket(species_SampleDict)\n",
    "keyfiles = {\n",
    "    'annot': annot,\n",
    "    'genome_fasta': genome_fasta,\n",
    "    'gxc': gxc\n",
    "}\n",
    "sample_BioFileDocket.add_keyfiles(keyfiles)\n",
    "\n",
    "display(vars(sample_BioFileDocket))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Load in the gxc matrix and get gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>ACAATATATTGTACCTGA</th>\n",
       "      <th>ACGTTGATGGCGTAGAGA</th>\n",
       "      <th>AACCTAACCTGAATTTGC</th>\n",
       "      <th>CTCGCAGCCCTCTATGTA</th>\n",
       "      <th>ACGTTGCGTATTTAGTCG</th>\n",
       "      <th>AACCTATAGAGACCGACG</th>\n",
       "      <th>ACGAGCGCTGTGGCCTAG</th>\n",
       "      <th>GCGAATGGACATGGACAT</th>\n",
       "      <th>TCTACCGCTCAAGCTCAA</th>\n",
       "      <th>...</th>\n",
       "      <th>CGGCAGTCAAAGATCTCT</th>\n",
       "      <th>GACACTGCGAATCTGTGT</th>\n",
       "      <th>GCAGGAGGCTGCTAAGGG</th>\n",
       "      <th>TATGTATACTTCCGCACC</th>\n",
       "      <th>TGGATGTTCCGCACAATA</th>\n",
       "      <th>AACCTATGGATGGGGTTT</th>\n",
       "      <th>AAGCGGAGGACTCTCCAT</th>\n",
       "      <th>ACCTGACTCGCAAGCGAG</th>\n",
       "      <th>ACGTTGCAAAGTTTCATA</th>\n",
       "      <th>ATCAACTGCAATTTCCGC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCF3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACOT12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACSF3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACTC1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACVR1C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20802</th>\n",
       "      <td>zwilch</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20803</th>\n",
       "      <td>zyg11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20804</th>\n",
       "      <td>zyx</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20805</th>\n",
       "      <td>zzef1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20806</th>\n",
       "      <td>zzz3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20807 rows × 10001 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GENE  ACAATATATTGTACCTGA  ACGTTGATGGCGTAGAGA  AACCTAACCTGAATTTGC  \\\n",
       "0       ABCF3                   0                   0                   0   \n",
       "1      ACOT12                   0                   0                   0   \n",
       "2       ACSF3                   0                   0                   0   \n",
       "3       ACTC1                   0                   0                   0   \n",
       "4      ACVR1C                   0                   0                   0   \n",
       "...       ...                 ...                 ...                 ...   \n",
       "20802  zwilch                   0                   0                   0   \n",
       "20803   zyg11                   0                   0                   0   \n",
       "20804     zyx                   0                   0                   0   \n",
       "20805   zzef1                   0                   0                   0   \n",
       "20806    zzz3                   0                   0                   0   \n",
       "\n",
       "       CTCGCAGCCCTCTATGTA  ACGTTGCGTATTTAGTCG  AACCTATAGAGACCGACG  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "20802                   0                   0                   0   \n",
       "20803                   0                   0                   0   \n",
       "20804                   0                   0                   0   \n",
       "20805                   0                   0                   0   \n",
       "20806                   0                   0                   0   \n",
       "\n",
       "       ACGAGCGCTGTGGCCTAG  GCGAATGGACATGGACAT  TCTACCGCTCAAGCTCAA  ...  \\\n",
       "0                       0                   0                   0  ...   \n",
       "1                       0                   0                   0  ...   \n",
       "2                       0                   0                   0  ...   \n",
       "3                       0                   0                   0  ...   \n",
       "4                       0                   0                   0  ...   \n",
       "...                   ...                 ...                 ...  ...   \n",
       "20802                   0                   0                   0  ...   \n",
       "20803                   0                   1                   0  ...   \n",
       "20804                   0                   0                   0  ...   \n",
       "20805                   0                   0                   0  ...   \n",
       "20806                   0                   0                   0  ...   \n",
       "\n",
       "       CGGCAGTCAAAGATCTCT  GACACTGCGAATCTGTGT  GCAGGAGGCTGCTAAGGG  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "20802                   0                   0                   0   \n",
       "20803                   0                   1                   1   \n",
       "20804                   0                   0                   0   \n",
       "20805                   0                   0                   0   \n",
       "20806                   0                   0                   0   \n",
       "\n",
       "       TATGTATACTTCCGCACC  TGGATGTTCCGCACAATA  AACCTATGGATGGGGTTT  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "20802                   0                   0                   0   \n",
       "20803                   0                   0                   0   \n",
       "20804                   0                   0                   0   \n",
       "20805                   0                   0                   0   \n",
       "20806                   0                   0                   0   \n",
       "\n",
       "       AAGCGGAGGACTCTCCAT  ACCTGACTCGCAAGCGAG  ACGTTGCAAAGTTTCATA  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "20802                   0                   0                   0   \n",
       "20803                   0                   0                   0   \n",
       "20804                   0                   0                   0   \n",
       "20805                   0                   0                   0   \n",
       "20806                   0                   0                   0   \n",
       "\n",
       "       ATCAACTGCAATTTCCGC  \n",
       "0                       0  \n",
       "1                       0  \n",
       "2                       0  \n",
       "3                       0  \n",
       "4                       0  \n",
       "...                   ...  \n",
       "20802                   0  \n",
       "20803                   0  \n",
       "20804                   0  \n",
       "20805                   0  \n",
       "20806                   0  \n",
       "\n",
       "[20807 rows x 10001 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACOT12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACSF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACTC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACVR1C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20802</th>\n",
       "      <td>zwilch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20803</th>\n",
       "      <td>zyg11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20804</th>\n",
       "      <td>zyx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20805</th>\n",
       "      <td>zzef1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20806</th>\n",
       "      <td>zzz3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20807 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name\n",
       "0         ABCF3\n",
       "1        ACOT12\n",
       "2         ACSF3\n",
       "3         ACTC1\n",
       "4        ACVR1C\n",
       "...         ...\n",
       "20802    zwilch\n",
       "20803     zyg11\n",
       "20804       zyx\n",
       "20805     zzef1\n",
       "20806      zzz3\n",
       "\n",
       "[20807 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes_matrix = pd.read_csv(sample_BioFileDocket.gxc.path, sep = '\\t')\n",
    "display(genes_matrix)\n",
    "\n",
    "gxc_genes_list = pd.DataFrame({'gene_name':genes_matrix['GENE']})\n",
    "display(gxc_genes_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jvfcFNT2FuaZ"
   },
   "source": [
    "# 3. Get mapping identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1.0</td>\n",
       "      <td>58871917.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id0;Dbxref=taxon:7955;Name=1;chromosome=1;g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>6642.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Dbxref=GeneID:192301,ZFIN:ZDB-GENE-02...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>6642.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Dbxref=GeneID:192301,Genb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11751.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Dbxref=GeneID:192301,Genban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11550.0</td>\n",
       "      <td>11625.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Dbxref=GeneID:192301,Genban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362356</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16449.0</td>\n",
       "      <td>16520.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id634507;Parent=rna65516;Dbxref=GeneID:1405...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362357</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene42297;Dbxref=GeneID:140511,ZFIN:ZDB-GEN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362358</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna65517;Parent=gene42297;Dbxref=GeneID:140...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362359</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id634508;Parent=rna65517;Dbxref=GeneID:1405...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362360</th>\n",
       "      <td>###</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1362361 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0           1       2        3           4    5    6    7  \\\n",
       "0        NC_007112.6      RefSeq  region      1.0  58871917.0    .    +    .   \n",
       "1        NC_007112.6  BestRefSeq    gene   6642.0     11878.0    .    -    .   \n",
       "2        NC_007112.6  BestRefSeq    mRNA   6642.0     11878.0    .    -    .   \n",
       "3        NC_007112.6  BestRefSeq    exon  11751.0     11878.0    .    -    .   \n",
       "4        NC_007112.6  BestRefSeq    exon  11550.0     11625.0    .    -    .   \n",
       "...              ...         ...     ...      ...         ...  ...  ...  ...   \n",
       "1362356  NC_002333.2      RefSeq    exon  16449.0     16520.0    .    +    .   \n",
       "1362357  NC_002333.2      RefSeq    gene  16527.0     16596.0    .    -    .   \n",
       "1362358  NC_002333.2      RefSeq    tRNA  16527.0     16596.0    .    -    .   \n",
       "1362359  NC_002333.2      RefSeq    exon  16527.0     16596.0    .    -    .   \n",
       "1362360          ###         NaN     NaN      NaN         NaN  NaN  NaN  NaN   \n",
       "\n",
       "                                                         8  \n",
       "0        ID=id0;Dbxref=taxon:7955;Name=1;chromosome=1;g...  \n",
       "1        ID=gene0;Dbxref=GeneID:192301,ZFIN:ZDB-GENE-02...  \n",
       "2        ID=rna0;Parent=gene0;Dbxref=GeneID:192301,Genb...  \n",
       "3        ID=id1;Parent=rna0;Dbxref=GeneID:192301,Genban...  \n",
       "4        ID=id2;Parent=rna0;Dbxref=GeneID:192301,Genban...  \n",
       "...                                                    ...  \n",
       "1362356  ID=id634507;Parent=rna65516;Dbxref=GeneID:1405...  \n",
       "1362357  ID=gene42297;Dbxref=GeneID:140511,ZFIN:ZDB-GEN...  \n",
       "1362358  ID=rna65517;Parent=gene42297;Dbxref=GeneID:140...  \n",
       "1362359  ID=id634508;Parent=rna65517;Dbxref=GeneID:1405...  \n",
       "1362360                                                NaN  \n",
       "\n",
       "[1362361 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'ID=id1;Parent=rna0;Dbxref=GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GENE-020419-25;gbkey=mRNA;gene=rpl24;product=ribosomal protein L24;transcript_id=NM_173235.3'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load in the original GFF-based annotation\n",
    "models = pd.read_csv(sample_BioFileDocket.annot.path, skiprows = 7, header = None, sep = '\\t')\n",
    "display(models)\n",
    "\n",
    "attributes_column = 8\n",
    "\n",
    "# Check the structure of fields in the GFF additional fields section\n",
    "display(models[attributes_column][3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Dbxref</th>\n",
       "      <th>dbxref_dict</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1.0</td>\n",
       "      <td>58871917.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id0;Dbxref=taxon:7955;Name=1;chromosome=1;g...</td>\n",
       "      <td>{'ID': 'id0', 'Dbxref': 'taxon:7955', 'Name': ...</td>\n",
       "      <td>None</td>\n",
       "      <td>taxon:7955</td>\n",
       "      <td>{'taxon': '7955'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>6642.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Dbxref=GeneID:192301,ZFIN:ZDB-GENE-02...</td>\n",
       "      <td>{'ID': 'gene0', 'Dbxref': 'GeneID:192301,ZFIN:...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>GeneID:192301,ZFIN:ZDB-GENE-020419-25</td>\n",
       "      <td>{'GeneID': '192301', 'ZFIN': 'ZDB-GENE-020419-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>6642.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Dbxref=GeneID:192301,Genb...</td>\n",
       "      <td>{'ID': 'rna0', 'Parent': 'gene0', 'Dbxref': 'G...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...</td>\n",
       "      <td>{'GeneID': '192301', 'Genbank': 'NM_173235.3',...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11751.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Dbxref=GeneID:192301,Genban...</td>\n",
       "      <td>{'ID': 'id1', 'Parent': 'rna0', 'Dbxref': 'Gen...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...</td>\n",
       "      <td>{'GeneID': '192301', 'Genbank': 'NM_173235.3',...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11550.0</td>\n",
       "      <td>11625.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Dbxref=GeneID:192301,Genban...</td>\n",
       "      <td>{'ID': 'id2', 'Parent': 'rna0', 'Dbxref': 'Gen...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...</td>\n",
       "      <td>{'GeneID': '192301', 'Genbank': 'NM_173235.3',...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362355</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>16449.0</td>\n",
       "      <td>16520.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna65516;Parent=gene42296;Dbxref=GeneID:140...</td>\n",
       "      <td>{'ID': 'rna65516', 'Parent': 'gene42296', 'Dbx...</td>\n",
       "      <td>trnT</td>\n",
       "      <td>GeneID:140518,ZFIN:ZDB-GENE-011205-39</td>\n",
       "      <td>{'GeneID': '140518', 'ZFIN': 'ZDB-GENE-011205-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362356</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16449.0</td>\n",
       "      <td>16520.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id634507;Parent=rna65516;Dbxref=GeneID:1405...</td>\n",
       "      <td>{'ID': 'id634507', 'Parent': 'rna65516', 'Dbxr...</td>\n",
       "      <td>trnT</td>\n",
       "      <td>GeneID:140518,ZFIN:ZDB-GENE-011205-39</td>\n",
       "      <td>{'GeneID': '140518', 'ZFIN': 'ZDB-GENE-011205-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362357</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene42297;Dbxref=GeneID:140511,ZFIN:ZDB-GEN...</td>\n",
       "      <td>{'ID': 'gene42297', 'Dbxref': 'GeneID:140511,Z...</td>\n",
       "      <td>trnP</td>\n",
       "      <td>GeneID:140511,ZFIN:ZDB-GENE-011205-38</td>\n",
       "      <td>{'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362358</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna65517;Parent=gene42297;Dbxref=GeneID:140...</td>\n",
       "      <td>{'ID': 'rna65517', 'Parent': 'gene42297', 'Dbx...</td>\n",
       "      <td>trnP</td>\n",
       "      <td>GeneID:140511,ZFIN:ZDB-GENE-011205-38</td>\n",
       "      <td>{'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362359</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id634508;Parent=rna65517;Dbxref=GeneID:1405...</td>\n",
       "      <td>{'ID': 'id634508', 'Parent': 'rna65517', 'Dbxr...</td>\n",
       "      <td>trnP</td>\n",
       "      <td>GeneID:140511,ZFIN:ZDB-GENE-011205-38</td>\n",
       "      <td>{'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1360240 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0           1       2        3           4  5  6  7  \\\n",
       "0        NC_007112.6      RefSeq  region      1.0  58871917.0  .  +  .   \n",
       "1        NC_007112.6  BestRefSeq    gene   6642.0     11878.0  .  -  .   \n",
       "2        NC_007112.6  BestRefSeq    mRNA   6642.0     11878.0  .  -  .   \n",
       "3        NC_007112.6  BestRefSeq    exon  11751.0     11878.0  .  -  .   \n",
       "4        NC_007112.6  BestRefSeq    exon  11550.0     11625.0  .  -  .   \n",
       "...              ...         ...     ...      ...         ... .. .. ..   \n",
       "1362355  NC_002333.2      RefSeq    tRNA  16449.0     16520.0  .  +  .   \n",
       "1362356  NC_002333.2      RefSeq    exon  16449.0     16520.0  .  +  .   \n",
       "1362357  NC_002333.2      RefSeq    gene  16527.0     16596.0  .  -  .   \n",
       "1362358  NC_002333.2      RefSeq    tRNA  16527.0     16596.0  .  -  .   \n",
       "1362359  NC_002333.2      RefSeq    exon  16527.0     16596.0  .  -  .   \n",
       "\n",
       "                                                         8  \\\n",
       "0        ID=id0;Dbxref=taxon:7955;Name=1;chromosome=1;g...   \n",
       "1        ID=gene0;Dbxref=GeneID:192301,ZFIN:ZDB-GENE-02...   \n",
       "2        ID=rna0;Parent=gene0;Dbxref=GeneID:192301,Genb...   \n",
       "3        ID=id1;Parent=rna0;Dbxref=GeneID:192301,Genban...   \n",
       "4        ID=id2;Parent=rna0;Dbxref=GeneID:192301,Genban...   \n",
       "...                                                    ...   \n",
       "1362355  ID=rna65516;Parent=gene42296;Dbxref=GeneID:140...   \n",
       "1362356  ID=id634507;Parent=rna65516;Dbxref=GeneID:1405...   \n",
       "1362357  ID=gene42297;Dbxref=GeneID:140511,ZFIN:ZDB-GEN...   \n",
       "1362358  ID=rna65517;Parent=gene42297;Dbxref=GeneID:140...   \n",
       "1362359  ID=id634508;Parent=rna65517;Dbxref=GeneID:1405...   \n",
       "\n",
       "                                          field_dictionary gene_name  \\\n",
       "0        {'ID': 'id0', 'Dbxref': 'taxon:7955', 'Name': ...      None   \n",
       "1        {'ID': 'gene0', 'Dbxref': 'GeneID:192301,ZFIN:...     rpl24   \n",
       "2        {'ID': 'rna0', 'Parent': 'gene0', 'Dbxref': 'G...     rpl24   \n",
       "3        {'ID': 'id1', 'Parent': 'rna0', 'Dbxref': 'Gen...     rpl24   \n",
       "4        {'ID': 'id2', 'Parent': 'rna0', 'Dbxref': 'Gen...     rpl24   \n",
       "...                                                    ...       ...   \n",
       "1362355  {'ID': 'rna65516', 'Parent': 'gene42296', 'Dbx...      trnT   \n",
       "1362356  {'ID': 'id634507', 'Parent': 'rna65516', 'Dbxr...      trnT   \n",
       "1362357  {'ID': 'gene42297', 'Dbxref': 'GeneID:140511,Z...      trnP   \n",
       "1362358  {'ID': 'rna65517', 'Parent': 'gene42297', 'Dbx...      trnP   \n",
       "1362359  {'ID': 'id634508', 'Parent': 'rna65517', 'Dbxr...      trnP   \n",
       "\n",
       "                                                    Dbxref  \\\n",
       "0                                               taxon:7955   \n",
       "1                    GeneID:192301,ZFIN:ZDB-GENE-020419-25   \n",
       "2        GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...   \n",
       "3        GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...   \n",
       "4        GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...   \n",
       "...                                                    ...   \n",
       "1362355              GeneID:140518,ZFIN:ZDB-GENE-011205-39   \n",
       "1362356              GeneID:140518,ZFIN:ZDB-GENE-011205-39   \n",
       "1362357              GeneID:140511,ZFIN:ZDB-GENE-011205-38   \n",
       "1362358              GeneID:140511,ZFIN:ZDB-GENE-011205-38   \n",
       "1362359              GeneID:140511,ZFIN:ZDB-GENE-011205-38   \n",
       "\n",
       "                                               dbxref_dict  \n",
       "0                                        {'taxon': '7955'}  \n",
       "1        {'GeneID': '192301', 'ZFIN': 'ZDB-GENE-020419-...  \n",
       "2        {'GeneID': '192301', 'Genbank': 'NM_173235.3',...  \n",
       "3        {'GeneID': '192301', 'Genbank': 'NM_173235.3',...  \n",
       "4        {'GeneID': '192301', 'Genbank': 'NM_173235.3',...  \n",
       "...                                                    ...  \n",
       "1362355  {'GeneID': '140518', 'ZFIN': 'ZDB-GENE-011205-...  \n",
       "1362356  {'GeneID': '140518', 'ZFIN': 'ZDB-GENE-011205-...  \n",
       "1362357  {'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...  \n",
       "1362358  {'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...  \n",
       "1362359  {'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...  \n",
       "\n",
       "[1360240 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Remove any rows with NaNs\n",
    "models.dropna(inplace = True)\n",
    "\n",
    "# Extract field and database cross-ref (dbxref) information into columns\n",
    "models['field_dictionary'] = models[attributes_column].apply(convert_fields_to_dict_gff)\n",
    "models['gene_name'] = [d.get('gene') for d in models['field_dictionary']]\n",
    "models['Dbxref'] = [d.get('Dbxref') for d in models['field_dictionary']]\n",
    "models['dbxref_dict'] = models['Dbxref'].apply(convert_dbxref_to_dict)\n",
    "\n",
    "display(models)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Extract gene IDs for mapping to UniprotKB\n",
    "Specify which set of identifiers will be use to query the [Uniprot ID Mapping Tool](https://www.uniprot.org/id-mapping) via API.\n",
    "\n",
    "If using an identifier from the `dbxref_dict`, specify the name via string in the `dbxref_datafield` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ZFIN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rpl24</td>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>cep97</td>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>nfkbiz</td>\n",
       "      <td>ZDB-GENE-071024-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>eed</td>\n",
       "      <td>ZDB-GENE-050417-287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>zgc:110091</td>\n",
       "      <td>ZDB-GENE-050417-34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362347</th>\n",
       "      <td>ND6</td>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362349</th>\n",
       "      <td>trnE</td>\n",
       "      <td>ZDB-GENE-011205-37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362352</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362354</th>\n",
       "      <td>trnT</td>\n",
       "      <td>ZDB-GENE-011205-39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362357</th>\n",
       "      <td>trnP</td>\n",
       "      <td>ZDB-GENE-011205-38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22031 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_name                 ZFIN\n",
       "1             rpl24   ZDB-GENE-020419-25\n",
       "15            cep97   ZDB-GENE-031030-11\n",
       "74           nfkbiz    ZDB-GENE-071024-1\n",
       "102             eed  ZDB-GENE-050417-287\n",
       "128      zgc:110091   ZDB-GENE-050417-34\n",
       "...             ...                  ...\n",
       "1362347         ND6   ZDB-GENE-011205-13\n",
       "1362349        trnE   ZDB-GENE-011205-37\n",
       "1362352        CYTB   ZDB-GENE-011205-17\n",
       "1362354        trnT   ZDB-GENE-011205-39\n",
       "1362357        trnP   ZDB-GENE-011205-38\n",
       "\n",
       "[22031 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dbxref_datafield = 'ZFIN'\n",
    "\n",
    "\n",
    "models.dropna(axis = 0, subset = ['dbxref_dict'], inplace = True)\n",
    "models[dbxref_datafield] = [d.get(dbxref_datafield) for d in models['dbxref_dict']]\n",
    "\n",
    "models_subset = models[['gene_name', dbxref_datafield]].dropna().drop_duplicates()\n",
    "\n",
    "\n",
    "display(models_subset)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Generate gene list file to query Uniprot ID Mapping API\n",
    "Generate a text file ending in `_ids.txt` for sending to the ID mapping API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "filename is ignored and generated by input as  Drer_adultbrain_ZFIN_ids.txt\n",
      "Wrote 22031 gene ids to ../../output/Drer_adultbrain/Drer_adultbrain_ZFIN_ids.txt\n"
     ]
    }
   ],
   "source": [
    "gene_list = models_subset[dbxref_datafield].unique()\n",
    "\n",
    "genelist_object = GeneListFile(\n",
    "    filename = '', # filename is automatically generated from sampledict\n",
    "    sampledict = species_SampleDict,\n",
    "    sources = [sample_BioFileDocket.annot],\n",
    "    genes = gene_list,\n",
    "    identifier = dbxref_datafield\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Query Uniprot ID Mapping API\n",
    "Specify the `from_type` variable based on the Uniprot name of the identifier.\n",
    "\n",
    "The table below lists some databases and the `from_type` string that the API accepts for that datatype.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "| datatype | `from_type` string | description |\n",
    "| ---: | :--- | :--- |\n",
    "| Mouse Genome Informatics | `MGI` | ID starts with `MGI:` |\n",
    "| Zebrafish Information Network | `ZFIN` | ID starts with `ZDB-GENE-` |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  419k    0    52  100  418k     23   191k  0:00:02  0:00:02 --:--:--  191k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100    24    0    24    0     0     33      0 --:--:-- --:--:-- --:--:--    33\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>From</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>Q8JGR4</td>\n",
       "      <td>RL24_DANRE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>60S ribosomal protein L24</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>A0A0R4IMS3</td>\n",
       "      <td>A0A0R4IMS3_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>60S ribosomal protein L24</td>\n",
       "      <td>rpl24 SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ICF0</td>\n",
       "      <td>A0A0R4ICF0_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Centrosomal protein 97</td>\n",
       "      <td>cep97 SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ISF5</td>\n",
       "      <td>A0A0R4ISF5_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Centrosomal protein 97</td>\n",
       "      <td>cep97 SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4IX25</td>\n",
       "      <td>A0A0R4IX25_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Centrosomal protein 97</td>\n",
       "      <td>cep97 SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54705</th>\n",
       "      <td>ZDB-GENE-011205-12</td>\n",
       "      <td>Q9MIY0</td>\n",
       "      <td>NU5M_DANRE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NADH-ubiquinone oxidoreductase chain 5 (EC 7.1...</td>\n",
       "      <td>mt-nd5 mtnd5 nd5</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54706</th>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>Q9MIX9</td>\n",
       "      <td>NU6M_DANRE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NADH-ubiquinone oxidoreductase chain 6 (EC 7.1...</td>\n",
       "      <td>mt-nd6 mtnd6 nd6</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54707</th>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>A0A0A0VG13</td>\n",
       "      <td>A0A0A0VG13_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>NADH-ubiquinone oxidoreductase chain 6 (EC 7.1...</td>\n",
       "      <td>ND6 mt-nd6 SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54708</th>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>Q9MIX8</td>\n",
       "      <td>CYB_DANRE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Cytochrome b (Complex III subunit 3) (Complex ...</td>\n",
       "      <td>mt-cyb cob cytb mtcyb</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54709</th>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>A0A0A0VDB5</td>\n",
       "      <td>A0A0A0VDB5_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Cytochrome b</td>\n",
       "      <td>CYTB mt-cyb SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>380</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54710 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     From       Entry        Entry Name    Reviewed  \\\n",
       "0      ZDB-GENE-020419-25      Q8JGR4        RL24_DANRE    reviewed   \n",
       "1      ZDB-GENE-020419-25  A0A0R4IMS3  A0A0R4IMS3_DANRE  unreviewed   \n",
       "2      ZDB-GENE-031030-11  A0A0R4ICF0  A0A0R4ICF0_DANRE  unreviewed   \n",
       "3      ZDB-GENE-031030-11  A0A0R4ISF5  A0A0R4ISF5_DANRE  unreviewed   \n",
       "4      ZDB-GENE-031030-11  A0A0R4IX25  A0A0R4IX25_DANRE  unreviewed   \n",
       "...                   ...         ...               ...         ...   \n",
       "54705  ZDB-GENE-011205-12      Q9MIY0        NU5M_DANRE    reviewed   \n",
       "54706  ZDB-GENE-011205-13      Q9MIX9        NU6M_DANRE    reviewed   \n",
       "54707  ZDB-GENE-011205-13  A0A0A0VG13  A0A0A0VG13_DANRE  unreviewed   \n",
       "54708  ZDB-GENE-011205-17      Q9MIX8         CYB_DANRE    reviewed   \n",
       "54709  ZDB-GENE-011205-17  A0A0A0VDB5  A0A0A0VDB5_DANRE  unreviewed   \n",
       "\n",
       "                                           Protein names  \\\n",
       "0                              60S ribosomal protein L24   \n",
       "1                              60S ribosomal protein L24   \n",
       "2                                 Centrosomal protein 97   \n",
       "3                                 Centrosomal protein 97   \n",
       "4                                 Centrosomal protein 97   \n",
       "...                                                  ...   \n",
       "54705  NADH-ubiquinone oxidoreductase chain 5 (EC 7.1...   \n",
       "54706  NADH-ubiquinone oxidoreductase chain 6 (EC 7.1...   \n",
       "54707  NADH-ubiquinone oxidoreductase chain 6 (EC 7.1...   \n",
       "54708  Cytochrome b (Complex III subunit 3) (Complex ...   \n",
       "54709                                       Cytochrome b   \n",
       "\n",
       "                   Gene Names                                     Organism  \\\n",
       "0                       rpl24  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "1            rpl24 SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "2            cep97 SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "3            cep97 SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "4            cep97 SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "...                       ...                                          ...   \n",
       "54705        mt-nd5 mtnd5 nd5  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "54706        mt-nd6 mtnd6 nd6  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "54707   ND6 mt-nd6 SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "54708   mt-cyb cob cytb mtcyb  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "54709  CYTB mt-cyb SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "\n",
       "       Length  \n",
       "0         157  \n",
       "1         157  \n",
       "2         599  \n",
       "3          40  \n",
       "4         240  \n",
       "...       ...  \n",
       "54705     606  \n",
       "54706     172  \n",
       "54707     172  \n",
       "54708     380  \n",
       "54709     380  \n",
       "\n",
       "[54710 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from_type = 'ZFIN'\n",
    "to_type = 'UniProtKB'\n",
    "\n",
    "uniprot_idmm_object = genelist_object.get_uniprot_ids(ID_MAPPER_LOC, from_type, to_type)\n",
    "sample_BioFileDocket.add_keyfiles({'uniprot_idmm': uniprot_idmm_object})\n",
    "\n",
    "uniprot_idmm = pd.read_csv(sample_BioFileDocket.uniprot_idmm.path, sep = '\\t')\n",
    "display(uniprot_idmm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Extract results and generate Uniprot IDMM\n",
    "Generates an idmm that links `gene_name`, the `dbxref_datafield` seleted above, and `uniprot_id` returned by API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_58660/3311337425.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  uniprot_idpairs.rename(columns = {'From': dbxref_datafield, 'Entry': 'uniprot_id'}, inplace = True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ZFIN</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>Q8JGR4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>A0A0R4IMS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ICF0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ISF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4IX25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54705</th>\n",
       "      <td>ZDB-GENE-011205-12</td>\n",
       "      <td>Q9MIY0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54706</th>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>Q9MIX9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54707</th>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>A0A0A0VG13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54708</th>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>Q9MIX8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54709</th>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>A0A0A0VDB5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54710 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     ZFIN  uniprot_id\n",
       "0      ZDB-GENE-020419-25      Q8JGR4\n",
       "1      ZDB-GENE-020419-25  A0A0R4IMS3\n",
       "2      ZDB-GENE-031030-11  A0A0R4ICF0\n",
       "3      ZDB-GENE-031030-11  A0A0R4ISF5\n",
       "4      ZDB-GENE-031030-11  A0A0R4IX25\n",
       "...                   ...         ...\n",
       "54705  ZDB-GENE-011205-12      Q9MIY0\n",
       "54706  ZDB-GENE-011205-13      Q9MIX9\n",
       "54707  ZDB-GENE-011205-13  A0A0A0VG13\n",
       "54708  ZDB-GENE-011205-17      Q9MIX8\n",
       "54709  ZDB-GENE-011205-17  A0A0A0VDB5\n",
       "\n",
       "[54710 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ZFIN</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rpl24</td>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>Q8JGR4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rpl24</td>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>A0A0R4IMS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cep97</td>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ICF0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cep97</td>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ISF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cep97</td>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4IX25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54705</th>\n",
       "      <td>ND5</td>\n",
       "      <td>ZDB-GENE-011205-12</td>\n",
       "      <td>Q9MIY0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54706</th>\n",
       "      <td>ND6</td>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>Q9MIX9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54707</th>\n",
       "      <td>ND6</td>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>A0A0A0VG13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54708</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>Q9MIX8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54709</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>A0A0A0VDB5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54710 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name                ZFIN  uniprot_id\n",
       "0         rpl24  ZDB-GENE-020419-25      Q8JGR4\n",
       "1         rpl24  ZDB-GENE-020419-25  A0A0R4IMS3\n",
       "2         cep97  ZDB-GENE-031030-11  A0A0R4ICF0\n",
       "3         cep97  ZDB-GENE-031030-11  A0A0R4ISF5\n",
       "4         cep97  ZDB-GENE-031030-11  A0A0R4IX25\n",
       "...         ...                 ...         ...\n",
       "54705       ND5  ZDB-GENE-011205-12      Q9MIY0\n",
       "54706       ND6  ZDB-GENE-011205-13      Q9MIX9\n",
       "54707       ND6  ZDB-GENE-011205-13  A0A0A0VG13\n",
       "54708      CYTB  ZDB-GENE-011205-17      Q9MIX8\n",
       "54709      CYTB  ZDB-GENE-011205-17  A0A0A0VDB5\n",
       "\n",
       "[54710 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "uniprot_idpairs = uniprot_idmm[['From', 'Entry']]\n",
    "uniprot_idpairs.rename(columns = {'From': dbxref_datafield, 'Entry': 'uniprot_id'}, inplace = True)\n",
    "display(uniprot_idpairs)\n",
    "\n",
    "uniprot_output_idmm = models_subset.merge(uniprot_idpairs, on = dbxref_datafield)\n",
    "display(uniprot_output_idmm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate a filename and file for the idmm\n",
    "uniprot_output_idmm_filename = '_'.join([species_prefix, conditions, 'uniprot-idmm.tsv'])\n",
    "uniprot_output_idmm_object = IdmmFile(uniprot_output_idmm_filename, species_SampleDict, kind = 'uniprot_idmm', sources = [sample_BioFileDocket.annot])\n",
    "\n",
    "# save to file and add to the BioFileDocket\n",
    "uniprot_output_idmm.to_csv(uniprot_output_idmm_object.path, sep = '\\t')\n",
    "sample_BioFileDocket.add_keyfile(uniprot_output_idmm_object, 'uniprot_idmm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8. Convert GFF to GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted file GCF_000002035.5_GRCz10_genomic.gtf already exists at:\n",
      " ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.gtf\n"
     ]
    }
   ],
   "source": [
    "# convert the GFF file to GTF using gffread\n",
    "models_asgtf = sample_BioFileDocket.annot.to_gtf(GFFREAD_LOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>6642</td>\n",
       "      <td>11878</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>6642</td>\n",
       "      <td>6760</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>6892</td>\n",
       "      <td>6955</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>9558</td>\n",
       "      <td>9694</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>10081</td>\n",
       "      <td>10191</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258174</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>15308</td>\n",
       "      <td>16448</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"gene42295\"; gene_name \"CYTB\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258175</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>16449</td>\n",
       "      <td>16520</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65516\"; gene_id \"gene42296\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258176</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16449</td>\n",
       "      <td>16520</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65516\"; gene_id \"gene42296\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258177</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>16527</td>\n",
       "      <td>16596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65517\"; gene_id \"gene42297\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258178</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16527</td>\n",
       "      <td>16596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65517\"; gene_id \"gene42297\";...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1258179 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0           1           2      3      4  5  6  7  \\\n",
       "0        NC_007112.6  BestRefSeq  transcript   6642  11878  .  -  .   \n",
       "1        NC_007112.6  BestRefSeq        exon   6642   6760  .  -  .   \n",
       "2        NC_007112.6  BestRefSeq        exon   6892   6955  .  -  .   \n",
       "3        NC_007112.6  BestRefSeq        exon   9558   9694  .  -  .   \n",
       "4        NC_007112.6  BestRefSeq        exon  10081  10191  .  -  .   \n",
       "...              ...         ...         ...    ...    ... .. .. ..   \n",
       "1258174  NC_002333.2      RefSeq         CDS  15308  16448  .  +  0   \n",
       "1258175  NC_002333.2      RefSeq  transcript  16449  16520  .  +  .   \n",
       "1258176  NC_002333.2      RefSeq        exon  16449  16520  .  +  .   \n",
       "1258177  NC_002333.2      RefSeq  transcript  16527  16596  .  -  .   \n",
       "1258178  NC_002333.2      RefSeq        exon  16527  16596  .  -  .   \n",
       "\n",
       "                                                         8  \n",
       "0        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "1        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "2        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "3        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "4        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "...                                                    ...  \n",
       "1258174       transcript_id \"gene42295\"; gene_name \"CYTB\";  \n",
       "1258175  transcript_id \"rna65516\"; gene_id \"gene42296\";...  \n",
       "1258176  transcript_id \"rna65516\"; gene_id \"gene42296\";...  \n",
       "1258177  transcript_id \"rna65517\"; gene_id \"gene42297\";...  \n",
       "1258178  transcript_id \"rna65517\"; gene_id \"gene42297\";...  \n",
       "\n",
       "[1258179 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'transcript_id \"rna0\"; gene_id \"gene0\"; gene_name \"rpl24\";'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load the newly-generated GTF file as a dataframe\n",
    "models_asgtf_df = pd.read_csv(models_asgtf.path, skiprows = 0, header = None, sep = '\\t')\n",
    "\n",
    "display(models_asgtf_df)\n",
    "display(models_asgtf_df[attributes_column][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>6642</td>\n",
       "      <td>11878</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>6642</td>\n",
       "      <td>6760</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>6892</td>\n",
       "      <td>6955</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>9558</td>\n",
       "      <td>9694</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>10081</td>\n",
       "      <td>10191</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258173</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15308</td>\n",
       "      <td>16448</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"gene42295\"; gene_id \"gene42295\"...</td>\n",
       "      <td>{'transcript_id': 'gene42295', 'gene_id': 'gen...</td>\n",
       "      <td>CYTB</td>\n",
       "      <td>gene42295</td>\n",
       "      <td>gene42295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258175</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>16449</td>\n",
       "      <td>16520</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65516\"; gene_id \"gene42296\";...</td>\n",
       "      <td>{'transcript_id': 'rna65516', 'gene_id': 'gene...</td>\n",
       "      <td>trnT</td>\n",
       "      <td>gene42296</td>\n",
       "      <td>rna65516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258176</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16449</td>\n",
       "      <td>16520</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65516\"; gene_id \"gene42296\";...</td>\n",
       "      <td>{'transcript_id': 'rna65516', 'gene_id': 'gene...</td>\n",
       "      <td>trnT</td>\n",
       "      <td>gene42296</td>\n",
       "      <td>rna65516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258177</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>16527</td>\n",
       "      <td>16596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65517\"; gene_id \"gene42297\";...</td>\n",
       "      <td>{'transcript_id': 'rna65517', 'gene_id': 'gene...</td>\n",
       "      <td>trnP</td>\n",
       "      <td>gene42297</td>\n",
       "      <td>rna65517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258178</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16527</td>\n",
       "      <td>16596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65517\"; gene_id \"gene42297\";...</td>\n",
       "      <td>{'transcript_id': 'rna65517', 'gene_id': 'gene...</td>\n",
       "      <td>trnP</td>\n",
       "      <td>gene42297</td>\n",
       "      <td>rna65517</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>699236 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0           1           2      3      4  5  6  7  \\\n",
       "0        NC_007112.6  BestRefSeq  transcript   6642  11878  .  -  .   \n",
       "1        NC_007112.6  BestRefSeq        exon   6642   6760  .  -  .   \n",
       "2        NC_007112.6  BestRefSeq        exon   6892   6955  .  -  .   \n",
       "3        NC_007112.6  BestRefSeq        exon   9558   9694  .  -  .   \n",
       "4        NC_007112.6  BestRefSeq        exon  10081  10191  .  -  .   \n",
       "...              ...         ...         ...    ...    ... .. .. ..   \n",
       "1258173  NC_002333.2      RefSeq  transcript  15308  16448  .  +  .   \n",
       "1258175  NC_002333.2      RefSeq  transcript  16449  16520  .  +  .   \n",
       "1258176  NC_002333.2      RefSeq        exon  16449  16520  .  +  .   \n",
       "1258177  NC_002333.2      RefSeq  transcript  16527  16596  .  -  .   \n",
       "1258178  NC_002333.2      RefSeq        exon  16527  16596  .  -  .   \n",
       "\n",
       "                                                         8  \\\n",
       "0        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "1        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "2        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "3        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "4        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "...                                                    ...   \n",
       "1258173  transcript_id \"gene42295\"; gene_id \"gene42295\"...   \n",
       "1258175  transcript_id \"rna65516\"; gene_id \"gene42296\";...   \n",
       "1258176  transcript_id \"rna65516\"; gene_id \"gene42296\";...   \n",
       "1258177  transcript_id \"rna65517\"; gene_id \"gene42297\";...   \n",
       "1258178  transcript_id \"rna65517\"; gene_id \"gene42297\";...   \n",
       "\n",
       "                                          field_dictionary gene_name  \\\n",
       "0        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...     rpl24   \n",
       "1        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...     rpl24   \n",
       "2        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...     rpl24   \n",
       "3        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...     rpl24   \n",
       "4        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...     rpl24   \n",
       "...                                                    ...       ...   \n",
       "1258173  {'transcript_id': 'gene42295', 'gene_id': 'gen...      CYTB   \n",
       "1258175  {'transcript_id': 'rna65516', 'gene_id': 'gene...      trnT   \n",
       "1258176  {'transcript_id': 'rna65516', 'gene_id': 'gene...      trnT   \n",
       "1258177  {'transcript_id': 'rna65517', 'gene_id': 'gene...      trnP   \n",
       "1258178  {'transcript_id': 'rna65517', 'gene_id': 'gene...      trnP   \n",
       "\n",
       "           gene_id transcript_id  \n",
       "0            gene0          rna0  \n",
       "1            gene0          rna0  \n",
       "2            gene0          rna0  \n",
       "3            gene0          rna0  \n",
       "4            gene0          rna0  \n",
       "...            ...           ...  \n",
       "1258173  gene42295     gene42295  \n",
       "1258175  gene42296      rna65516  \n",
       "1258176  gene42296      rna65516  \n",
       "1258177  gene42297      rna65517  \n",
       "1258178  gene42297      rna65517  \n",
       "\n",
       "[699236 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Use a custom function to extract useful fields from the additional fields section (column 8)\n",
    "# Pull from that dict to fill in additional useful columns\n",
    "models_asgtf_df['field_dictionary'] = models_asgtf_df[attributes_column].apply(convert_fields_to_dict_gtf)\n",
    "models_asgtf_df['gene_name'] = [d.get('gene_name') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['gene_id'] = [d.get('gene_id') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['transcript_id'] = [d.get('transcript_id') for d in models_asgtf_df['field_dictionary']]\n",
    "\n",
    "# Remove CDS annotations because they interfere with TransDecoder cDNA generation\n",
    "models_asgtf_df = models_asgtf_df[models_asgtf_df[2] != 'CDS']\n",
    "display(models_asgtf_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9. Generate gtf-idmm\n",
    "This file maps the `gene_name` to `gene_id` and `transcript_id` fields generated by the conversion from GFF to GTF, which will be needed for downstream processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cep97</td>\n",
       "      <td>gene1</td>\n",
       "      <td>rna2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>cep97</td>\n",
       "      <td>gene1</td>\n",
       "      <td>rna1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>cep97</td>\n",
       "      <td>gene1</td>\n",
       "      <td>rna3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>nfkbiz</td>\n",
       "      <td>gene2</td>\n",
       "      <td>rna4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258169</th>\n",
       "      <td>ND6</td>\n",
       "      <td>gene42293</td>\n",
       "      <td>gene42293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258171</th>\n",
       "      <td>trnE</td>\n",
       "      <td>gene42294</td>\n",
       "      <td>rna65515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258173</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>gene42295</td>\n",
       "      <td>gene42295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258175</th>\n",
       "      <td>trnT</td>\n",
       "      <td>gene42296</td>\n",
       "      <td>rna65516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258177</th>\n",
       "      <td>trnP</td>\n",
       "      <td>gene42297</td>\n",
       "      <td>rna65517</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65914 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name    gene_id transcript_id\n",
       "0           rpl24      gene0          rna0\n",
       "13          cep97      gene1          rna2\n",
       "25          cep97      gene1          rna1\n",
       "48          cep97      gene1          rna3\n",
       "71         nfkbiz      gene2          rna4\n",
       "...           ...        ...           ...\n",
       "1258169       ND6  gene42293     gene42293\n",
       "1258171      trnE  gene42294      rna65515\n",
       "1258173      CYTB  gene42295     gene42295\n",
       "1258175      trnT  gene42296      rna65516\n",
       "1258177      trnP  gene42297      rna65517\n",
       "\n",
       "[65914 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract gene_name, gene_id, and transcript_id fields to generate an ID mapping matrix (idmm)\n",
    "idmm_df = models_asgtf_df[['gene_name', 'gene_id', 'transcript_id']].drop_duplicates()\n",
    "idmm_df.dropna(inplace = True)\n",
    "display(idmm_df)\n",
    "\n",
    "# generate a filename and file for the idmm\n",
    "idmm_filename = '_'.join([species_prefix, conditions, 'gtf-idmm.tsv'])\n",
    "idmm = IdmmFile(idmm_filename, species_SampleDict, kind = 'gtf_idmm', sources = [sample_BioFileDocket.annot])\n",
    "\n",
    "# save to file and add to the BioFileDocket\n",
    "idmm_df.to_csv(idmm.path, sep = '\\t')\n",
    "sample_BioFileDocket.add_keyfile(idmm, 'gtf_idmm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 10. Generate updated gtf\n",
    "Generated an updated GTF file using transcript_id as the key. For some datasets, transcripts do not consistently get gene names and gene IDs added, which causes Transdecoder to throw errors. This resolves that problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "models_asgtf_updated_df = models_asgtf_df.merge(idmm_df, on = 'transcript_id')\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_name': x['gene_name_y']}), axis = 1)\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_id': x['gene_id_y']}), axis = 1)\n",
    "models_asgtf_updated_df['new_fields'] = models_asgtf_updated_df['field_dictionary'].apply(convert_dict_to_fields_gtf)\n",
    "models_asgtf_updated_df = models_asgtf_updated_df[[0, 1, 2, 3, 4, 5, 6, 7, 'new_fields']]\n",
    "\n",
    "models_asgtf_updated_filename = models_asgtf.filename.replace('.gtf', '_updated.gtf')\n",
    "models_asgtf_updated = GenomeGtfFile(models_asgtf_updated_filename, species_SampleDict, GenomeFastaFile = sample_BioFileDocket.genome_fasta)\n",
    "\n",
    "models_asgtf_updated_df.to_csv(models_asgtf_updated.path, header = None, index = None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 11. Generate cDNA and peptide files\n",
    "Using the updated gtf file and genome file, generate cDNA sequence.\n",
    "\n",
    "Then, using the cDNA sequence, generate peptide sequences using transdecoder.\n",
    "\n",
    "Expect this step to take some time, probably ~20-30min."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/compute_base_probs.pl ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna 0 > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//base_freqs.dat, checkpoint [/home/ec2-user/glial-origins/notebooks/Drer_adultbrain/../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints_longorfs/base_freqs_file.ok] exists.\n",
      "-skipping long orf extraction, already completed earlier as per checkpoint: ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints_longorfs/TD.longorfs.ok\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds 5000 5000 > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000, checkpoint [/home/ec2-user/glial-origins/notebooks/Drer_adultbrain/../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/get_longest_orfs.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/exclude_similar_proteins.pl ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000 > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr, checkpoint [/home/ec2-user/glial-origins/notebooks/Drer_adultbrain/../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/nr.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr 500 > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest, checkpoint [/home/ec2-user/glial-origins/notebooks/Drer_adultbrain/../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/top_train_select.ok] exists.\n",
      "PCT_GC: 46.2\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/seq_n_baseprobs_to_loglikelihood_vals.pl ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//base_freqs.dat > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//hexamer.scores, checkpoint [/home/ec2-user/glial-origins/notebooks/Drer_adultbrain/../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/hexamer_scores.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/score_CDS_likelihood_all_6_frames.pl ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//hexamer.scores > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.scores, checkpoint [/home/ec2-user/glial-origins/notebooks/Drer_adultbrain/../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/score_cdss.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/select_best_ORFs_per_transcript.pl --gff3_file ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.gff3 --cds_scores ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.scores  --min_length_auto_accept 749  > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 , checkpoint [/home/ec2-user/glial-origins/notebooks/Drer_adultbrain/../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/select_best_orfs.8a11739f1dfde1dc35170019af827e97.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/train_start_PWM.pl --transcripts ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna --selected_orfs ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest --out_prefix ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement, checkpoint [/home/ec2-user/glial-origins/notebooks/Drer_adultbrain/../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/train_start_PWM.8a11739f1dfde1dc35170019af827e97.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/start_codon_refinement.pl --transcripts ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna --gff3_file ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 --workdir ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/ > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3, checkpoint [/home/ec2-user/glial-origins/notebooks/Drer_adultbrain/../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/revise_starts.8a11739f1dfde1dc35170019af827e97.ok] exists.\n",
      "-- Skipping CMD: cp ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3 GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3, checkpoint [/home/ec2-user/glial-origins/notebooks/Drer_adultbrain/../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/copy_to_final_outputfile.8a11739f1dfde1dc35170019af827e97.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_bed.pl GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 > GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed, checkpoint [/home/ec2-user/glial-origins/notebooks/Drer_adultbrain/../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/make_final_bed.8a11739f1dfde1dc35170019af827e97.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 --fasta ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna  --genetic_code Universal > GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep, checkpoint [/home/ec2-user/glial-origins/notebooks/Drer_adultbrain/../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/make_final_pep.8a11739f1dfde1dc35170019af827e97.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 --fasta ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna --seqType CDS  --genetic_code Universal > GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds, checkpoint [/home/ec2-user/glial-origins/notebooks/Drer_adultbrain/../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/make_final_cds.8a11739f1dfde1dc35170019af827e97.ok] exists.\n",
      "transdecoder is finished.  See output files GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.*\n",
      "\n",
      "\n",
      "mv: cannot stat ‘GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed’: No such file or directory\n",
      "mv: cannot stat ‘GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds’: No such file or directory\n",
      "mv: cannot stat ‘GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3’: No such file or directory\n",
      "mv: cannot stat ‘GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep’: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "cdna = sample_BioFileDocket.genome_fasta.get_transdecoder_cdna_gtf(models_asgtf_updated, TRANSDECODER_LOC)\n",
    "sample_BioFileDocket.add_keyfile(cdna, 'cdna')\n",
    "\n",
    "transdecoder_files = sample_BioFileDocket.cdna.to_pep_files(TDLONGORF_LOC, TDPREDICT_LOC)\n",
    "sample_BioFileDocket.add_keyfiles(transdecoder_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "iEug6TQ6Cc9D"
   },
   "source": [
    "# 12. Push files to AWS S3\n",
    "\n",
    "Iteratively moves through BioFileDocket BioFile objects and pushes to the right place in AWS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"AcceptRanges\": \"bytes\", \n",
      "    \"ContentType\": \"binary/octet-stream\", \n",
      "    \"LastModified\": \"Wed, 31 Aug 2022 16:53:44 GMT\", \n",
      "    \"ContentLength\": 404109353, \n",
      "    \"VersionId\": \"6YfNqzr1SyHj1mGlHke1D4CNrCEp8aVR\", \n",
      "    \"ETag\": \"\\\"97f90265ca3e1b21f7c13e4ddfb41584-49\\\"\", \n",
      "    \"ServerSideEncryption\": \"AES256\", \n",
      "    \"Metadata\": {}\n",
      "}\n",
      "GCF_000002035.5_GRCz10_genomic.gff already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "{\n",
      "    \"AcceptRanges\": \"bytes\", \n",
      "    \"ContentType\": \"binary/octet-stream\", \n",
      "    \"LastModified\": \"Tue, 11 Oct 2022 23:09:05 GMT\", \n",
      "    \"ContentLength\": 1388963222, \n",
      "    \"VersionId\": \"kZEWjktEIBOm0Z3sTNGRCe_CjgcEOWud\", \n",
      "    \"ETag\": \"\\\"8dba2a5b1b850bfc0cd3e7b3152903c3-166\\\"\", \n",
      "    \"ServerSideEncryption\": \"AES256\", \n",
      "    \"Metadata\": {}\n",
      "}\n",
      "GCF_000002035.5_GRCz10_genomic.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "{\n",
      "    \"AcceptRanges\": \"bytes\", \n",
      "    \"ContentType\": \"text/plain\", \n",
      "    \"LastModified\": \"Tue, 11 Oct 2022 23:09:20 GMT\", \n",
      "    \"ContentLength\": 416652730, \n",
      "    \"VersionId\": \"u5f64DPmbInDNHXBbU2gNvI1BeamAy2.\", \n",
      "    \"ETag\": \"\\\"94546c2513c4b49723560c59b468646f-50\\\"\", \n",
      "    \"ServerSideEncryption\": \"AES256\", \n",
      "    \"Metadata\": {}\n",
      "}\n",
      "GSM3768152_Brain_8_dge.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "{\n",
      "    \"AcceptRanges\": \"bytes\", \n",
      "    \"ContentType\": \"text/tab-separated-values\", \n",
      "    \"LastModified\": \"Tue, 11 Oct 2022 23:09:27 GMT\", \n",
      "    \"ContentLength\": 2235565, \n",
      "    \"VersionId\": \"TZVf6gqET3Ixg101t0M_YmfH1808sVX9\", \n",
      "    \"ETag\": \"\\\"9559de3a7d9591747a28c0e2da10793c\\\"\", \n",
      "    \"ServerSideEncryption\": \"AES256\", \n",
      "    \"Metadata\": {}\n",
      "}\n",
      "Drer_adultbrain_uniprot-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "{\n",
      "    \"AcceptRanges\": \"bytes\", \n",
      "    \"ContentType\": \"text/tab-separated-values\", \n",
      "    \"LastModified\": \"Tue, 11 Oct 2022 23:09:29 GMT\", \n",
      "    \"ContentLength\": 2317545, \n",
      "    \"VersionId\": \"CvthMioewNpw.PGQWcs0wfZuOclL2D3u\", \n",
      "    \"ETag\": \"\\\"30fb375218d9be3707c874ad0a9d03a4\\\"\", \n",
      "    \"ServerSideEncryption\": \"AES256\", \n",
      "    \"Metadata\": {}\n",
      "}\n",
      "Drer_adultbrain_gtf-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "{\n",
      "    \"AcceptRanges\": \"bytes\", \n",
      "    \"ContentType\": \"binary/octet-stream\", \n",
      "    \"LastModified\": \"Tue, 11 Oct 2022 23:09:30 GMT\", \n",
      "    \"ContentLength\": 151626002, \n",
      "    \"VersionId\": \"ilIwuWP5sONPSAH6U8x07xkXo25G1q4P\", \n",
      "    \"ETag\": \"\\\"4559a642dd98273784b09d7e26fe2b57-19\\\"\", \n",
      "    \"ServerSideEncryption\": \"AES256\", \n",
      "    \"Metadata\": {}\n",
      "}\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "{\n",
      "    \"AcceptRanges\": \"bytes\", \n",
      "    \"ContentType\": \"application/vnd.realvnc.bed\", \n",
      "    \"LastModified\": \"Tue, 11 Oct 2022 23:09:32 GMT\", \n",
      "    \"ContentLength\": 6289609, \n",
      "    \"VersionId\": \"mii1Igh8Ok1L9DytiGi95wiLIBaDe3ZQ\", \n",
      "    \"ETag\": \"\\\"3e033f2e4925f5944b7de026fd506d4c\\\"\", \n",
      "    \"ServerSideEncryption\": \"AES256\", \n",
      "    \"Metadata\": {}\n",
      "}\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "{\n",
      "    \"AcceptRanges\": \"bytes\", \n",
      "    \"ContentType\": \"binary/octet-stream\", \n",
      "    \"LastModified\": \"Tue, 11 Oct 2022 23:09:34 GMT\", \n",
      "    \"ContentLength\": 105218681, \n",
      "    \"VersionId\": \"MfRM5HEdBgtCyaw5dSdWGyECzVdou0si\", \n",
      "    \"ETag\": \"\\\"ba0b03e88f570a45c322e044410e0aa2-13\\\"\", \n",
      "    \"ServerSideEncryption\": \"AES256\", \n",
      "    \"Metadata\": {}\n",
      "}\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "{\n",
      "    \"AcceptRanges\": \"bytes\", \n",
      "    \"ContentType\": \"binary/octet-stream\", \n",
      "    \"LastModified\": \"Tue, 11 Oct 2022 23:09:36 GMT\", \n",
      "    \"ContentLength\": 31215175, \n",
      "    \"VersionId\": \"ojfjU_9zDMjZ9Lmc4G5jFJRZHjma_saa\", \n",
      "    \"ETag\": \"\\\"3ca0375bf2b4c7ac1a551074bf0e60a1-4\\\"\", \n",
      "    \"ServerSideEncryption\": \"AES256\", \n",
      "    \"Metadata\": {}\n",
      "}\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "{\n",
      "    \"AcceptRanges\": \"bytes\", \n",
      "    \"ContentType\": \"binary/octet-stream\", \n",
      "    \"LastModified\": \"Tue, 11 Oct 2022 23:09:37 GMT\", \n",
      "    \"ContentLength\": 38611068, \n",
      "    \"VersionId\": \"b1qlzIAuHUG41HgZ._gyq50nJC116J9s\", \n",
      "    \"ETag\": \"\\\"5109e8268676895598589a9ea499612a-5\\\"\", \n",
      "    \"ServerSideEncryption\": \"AES256\", \n",
      "    \"Metadata\": {}\n",
      "}\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n"
     ]
    }
   ],
   "source": [
    "sample_BioFileDocket.local_to_s3()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 13. Pickle the `sample_BioFileDocket` variable for use by the next script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upload: ../../output/Drer_adultbrain/Drer_adultbrain_sample_BioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/Drer_adultbrain_sample_BioFileDocket.pkl\n"
     ]
    }
   ],
   "source": [
    "sample_BioFileDocket.pickle()\n",
    "sample_BioFileDocket.push_to_s3()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "Xtro_bro.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
