{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview\n",
    "\n",
    "This notebook collects data from the [Jiang et al. 2021](https://www.frontiersin.org/articles/10.3389/fcell.2021.743421/full) zebrafish cell atlas.\n",
    "\n",
    "#### Dataset description\n",
    "\n",
    "- Data can be found at the [GEO Accession GSE130487](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE130487).\n",
    "- This notebook previously collects data from one sample, [\"Brain8\"](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3768152), for the purposes of data analysis and exploration."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "L56fjA27DmhA"
   },
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "t0BTCsz7Y48y",
    "outputId": "f6510678-f977-4ec8-ce30-1f2acd4fe666"
   },
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess, os, dill, sys\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Download files\n",
    "\n",
    "#### BioFileDocket\n",
    "First, make a BioFileDocket for the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/glial-origins/output/Drer_adultbrain/ already exists\n",
      "Files will be saved into /home/ec2-user/glial-origins/output/Drer_adultbrain/\n"
     ]
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify the name of the species folder in Amazon S3\n",
    "species = 'Danio_rerio'\n",
    "\n",
    "# Specify any particular identifying conditions, eg tissue type:\n",
    "conditions = 'adultbrain'\n",
    "\n",
    "################\n",
    "################\n",
    "\n",
    "sample_BFD = BioFileDocket(species, conditions)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Download and add files\n",
    "Next, download the genome_fasta, annotation, and gxc files based on urls.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inferring file name as GCF_000002035.5_GRCz10_genomic.fna.gz\n",
      "file GCF_000002035.5_GRCz10_genomic.fna.gz already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.fna.gz\n",
      "file GCF_000002035.5_GRCz10_genomic.fna.gz unzipped and object renamed to GCF_000002035.5_GRCz10_genomic.fna\n",
      "inferring file name as GCF_000002035.5_GRCz10_genomic.gff.gz\n",
      "file GCF_000002035.5_GRCz10_genomic.gff.gz already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.gff.gz\n",
      "file GCF_000002035.5_GRCz10_genomic.gff.gz unzipped and object renamed to GCF_000002035.5_GRCz10_genomic.gff\n",
      "inferring file name as GSM3768152_Brain_8_dge.txt.gz\n",
      "file GSM3768152_Brain_8_dge.txt.gz already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GSM3768152_Brain_8_dge.txt.gz\n",
      "file GSM3768152_Brain_8_dge.txt.gz unzipped and object renamed to GSM3768152_Brain_8_dge.txt\n",
      "key \"annot\" already exists, ignoring\n",
      "key \"genome_fasta\" already exists, ignoring\n",
      "key \"gxc\" already exists, ignoring\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "gzip: /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.fna: unknown suffix -- ignored\n",
      "gzip: /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.gff: unknown suffix -- ignored\n",
      "gzip: /home/ec2-user/glial-origins/output/Drer_adultbrain/GSM3768152_Brain_8_dge.txt: unknown suffix -- ignored\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'species': 'Danio_rerio',\n",
       " 'conditions': 'adultbrain',\n",
       " 'directory': '/home/ec2-user/glial-origins/output/Drer_adultbrain/',\n",
       " 'files': {},\n",
       " 'metadata': <biofile_handling.metadata_object at 0x7fe17d01bf40>,\n",
       " 'annot': <biofile_handling.GenomeGffFile at 0x7fe17d019030>,\n",
       " 'genome_fasta': <biofile_handling.GenomeFastaFile at 0x7fe0cc0ecd00>,\n",
       " 'gxc': <biofile_handling.GxcFile at 0x7fe17d01b400>}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify url and other variables\n",
    "genome_fasta_url = 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/035/GCF_000002035.5_GRCz10/GCF_000002035.5_GRCz10_genomic.fna.gz'\n",
    "genome_version = 'GRCz10'\n",
    "\n",
    "annot_url = 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/035/GCF_000002035.5_GRCz10/GCF_000002035.5_GRCz10_genomic.gff.gz'\n",
    "gxc_url = 'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3768nnn/GSM3768152/suppl/GSM3768152_Brain_8_dge.txt.gz'\n",
    "\n",
    "###########\n",
    "# runtime #\n",
    "###########\n",
    "\n",
    "protocol = 'curl'\n",
    "\n",
    "genome_fasta = GenomeFastaFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    version = genome_version,\n",
    "    url = genome_fasta_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "annot = GenomeGffFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    reference_genome = genome_fasta,\n",
    "    url = annot_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "gxc = GxcFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    reference_genome = genome_fasta,\n",
    "    reference_annot = annot,\n",
    "    url = gxc_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "keyfiles = {\n",
    "    'annot': annot,\n",
    "    'genome_fasta': genome_fasta,\n",
    "    'gxc': gxc\n",
    "}\n",
    "sample_BFD.add_keyfiles(keyfiles)\n",
    "\n",
    "display(vars(sample_BFD))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Extract gxc information\n",
    "\n",
    "#### Preview gxc matrix\n",
    "- Show the first 10 rows of the gxc matrix.  \n",
    "- Extract the first column as a 'gene_name' dataframe.\n",
    "- Record the number of cells and genes in the gxc to the `sample_BFD.metadata` attribute."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>ACAATATATTGTACCTGA</th>\n",
       "      <th>ACGTTGATGGCGTAGAGA</th>\n",
       "      <th>AACCTAACCTGAATTTGC</th>\n",
       "      <th>CTCGCAGCCCTCTATGTA</th>\n",
       "      <th>ACGTTGCGTATTTAGTCG</th>\n",
       "      <th>AACCTATAGAGACCGACG</th>\n",
       "      <th>ACGAGCGCTGTGGCCTAG</th>\n",
       "      <th>GCGAATGGACATGGACAT</th>\n",
       "      <th>TCTACCGCTCAAGCTCAA</th>\n",
       "      <th>...</th>\n",
       "      <th>CGGCAGTCAAAGATCTCT</th>\n",
       "      <th>GACACTGCGAATCTGTGT</th>\n",
       "      <th>GCAGGAGGCTGCTAAGGG</th>\n",
       "      <th>TATGTATACTTCCGCACC</th>\n",
       "      <th>TGGATGTTCCGCACAATA</th>\n",
       "      <th>AACCTATGGATGGGGTTT</th>\n",
       "      <th>AAGCGGAGGACTCTCCAT</th>\n",
       "      <th>ACCTGACTCGCAAGCGAG</th>\n",
       "      <th>ACGTTGCAAAGTTTCATA</th>\n",
       "      <th>ATCAACTGCAATTTCCGC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCF3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACOT12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACSF3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACTC1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACVR1C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ADAM12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ADAMTSL4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ADGRL3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AKAP13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AL590134.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 10001 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GENE  ACAATATATTGTACCTGA  ACGTTGATGGCGTAGAGA  AACCTAACCTGAATTTGC  \\\n",
       "0       ABCF3                   0                   0                   0   \n",
       "1      ACOT12                   0                   0                   0   \n",
       "2       ACSF3                   0                   0                   0   \n",
       "3       ACTC1                   0                   0                   0   \n",
       "4      ACVR1C                   0                   0                   0   \n",
       "5      ADAM12                   0                   0                   0   \n",
       "6    ADAMTSL4                   0                   0                   0   \n",
       "7      ADGRL3                   0                   0                   0   \n",
       "8      AKAP13                   0                   0                   0   \n",
       "9  AL590134.1                   0                   0                   0   \n",
       "\n",
       "   CTCGCAGCCCTCTATGTA  ACGTTGCGTATTTAGTCG  AACCTATAGAGACCGACG  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   ACGAGCGCTGTGGCCTAG  GCGAATGGACATGGACAT  TCTACCGCTCAAGCTCAA  ...  \\\n",
       "0                   0                   0                   0  ...   \n",
       "1                   0                   0                   0  ...   \n",
       "2                   0                   0                   0  ...   \n",
       "3                   0                   0                   0  ...   \n",
       "4                   0                   0                   0  ...   \n",
       "5                   0                   0                   0  ...   \n",
       "6                   0                   0                   0  ...   \n",
       "7                   0                   0                   0  ...   \n",
       "8                   0                   0                   0  ...   \n",
       "9                   1                   0                   0  ...   \n",
       "\n",
       "   CGGCAGTCAAAGATCTCT  GACACTGCGAATCTGTGT  GCAGGAGGCTGCTAAGGG  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   TATGTATACTTCCGCACC  TGGATGTTCCGCACAATA  AACCTATGGATGGGGTTT  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   AAGCGGAGGACTCTCCAT  ACCTGACTCGCAAGCGAG  ACGTTGCAAAGTTTCATA  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   ATCAACTGCAATTTCCGC  \n",
       "0                   0  \n",
       "1                   0  \n",
       "2                   0  \n",
       "3                   0  \n",
       "4                   0  \n",
       "5                   0  \n",
       "6                   0  \n",
       "7                   0  \n",
       "8                   0  \n",
       "9                   0  \n",
       "\n",
       "[10 rows x 10001 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GENE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACOT12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACSF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACTC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20803</th>\n",
       "      <td>zwilch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20804</th>\n",
       "      <td>zyg11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20805</th>\n",
       "      <td>zyx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20806</th>\n",
       "      <td>zzef1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20807</th>\n",
       "      <td>zzz3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20808 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name\n",
       "0          GENE\n",
       "1         ABCF3\n",
       "2        ACOT12\n",
       "3         ACSF3\n",
       "4         ACTC1\n",
       "...         ...\n",
       "20803    zwilch\n",
       "20804     zyg11\n",
       "20805       zyx\n",
       "20806     zzef1\n",
       "20807      zzz3\n",
       "\n",
       "[20808 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes_matrix = pd.read_csv(sample_BFD.gxc.path, sep = '\\t', nrows = 10)\n",
    "display(genes_matrix)\n",
    "\n",
    "gxc_genes_list = pd.read_csv(sample_BFD.gxc.path, sep = '\\t', usecols=[0], names = ['gene_name'])\n",
    "display(gxc_genes_list)\n",
    "\n",
    "sample_BFD.metadata.add('num_cells', len(genes_matrix.columns) - 1)\n",
    "sample_BFD.metadata.add('num_genes', len(genes_matrix))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jvfcFNT2FuaZ"
   },
   "source": [
    "# 3. Get mapping identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1.0</td>\n",
       "      <td>58871917.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id0;Dbxref=taxon:7955;Name=1;chromosome=1;g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>6642.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Dbxref=GeneID:192301,ZFIN:ZDB-GENE-02...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>6642.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Dbxref=GeneID:192301,Genb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11751.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Dbxref=GeneID:192301,Genban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11550.0</td>\n",
       "      <td>11625.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Dbxref=GeneID:192301,Genban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362356</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16449.0</td>\n",
       "      <td>16520.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id634507;Parent=rna65516;Dbxref=GeneID:1405...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362357</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene42297;Dbxref=GeneID:140511,ZFIN:ZDB-GEN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362358</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna65517;Parent=gene42297;Dbxref=GeneID:140...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362359</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id634508;Parent=rna65517;Dbxref=GeneID:1405...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362360</th>\n",
       "      <td>###</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1362361 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0           1       2        3           4    5    6    7  \\\n",
       "0        NC_007112.6      RefSeq  region      1.0  58871917.0    .    +    .   \n",
       "1        NC_007112.6  BestRefSeq    gene   6642.0     11878.0    .    -    .   \n",
       "2        NC_007112.6  BestRefSeq    mRNA   6642.0     11878.0    .    -    .   \n",
       "3        NC_007112.6  BestRefSeq    exon  11751.0     11878.0    .    -    .   \n",
       "4        NC_007112.6  BestRefSeq    exon  11550.0     11625.0    .    -    .   \n",
       "...              ...         ...     ...      ...         ...  ...  ...  ...   \n",
       "1362356  NC_002333.2      RefSeq    exon  16449.0     16520.0    .    +    .   \n",
       "1362357  NC_002333.2      RefSeq    gene  16527.0     16596.0    .    -    .   \n",
       "1362358  NC_002333.2      RefSeq    tRNA  16527.0     16596.0    .    -    .   \n",
       "1362359  NC_002333.2      RefSeq    exon  16527.0     16596.0    .    -    .   \n",
       "1362360          ###         NaN     NaN      NaN         NaN  NaN  NaN  NaN   \n",
       "\n",
       "                                                         8  \n",
       "0        ID=id0;Dbxref=taxon:7955;Name=1;chromosome=1;g...  \n",
       "1        ID=gene0;Dbxref=GeneID:192301,ZFIN:ZDB-GENE-02...  \n",
       "2        ID=rna0;Parent=gene0;Dbxref=GeneID:192301,Genb...  \n",
       "3        ID=id1;Parent=rna0;Dbxref=GeneID:192301,Genban...  \n",
       "4        ID=id2;Parent=rna0;Dbxref=GeneID:192301,Genban...  \n",
       "...                                                    ...  \n",
       "1362356  ID=id634507;Parent=rna65516;Dbxref=GeneID:1405...  \n",
       "1362357  ID=gene42297;Dbxref=GeneID:140511,ZFIN:ZDB-GEN...  \n",
       "1362358  ID=rna65517;Parent=gene42297;Dbxref=GeneID:140...  \n",
       "1362359  ID=id634508;Parent=rna65517;Dbxref=GeneID:1405...  \n",
       "1362360                                                NaN  \n",
       "\n",
       "[1362361 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'ID=id1;Parent=rna0;Dbxref=GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GENE-020419-25;gbkey=mRNA;gene=rpl24;product=ribosomal protein L24;transcript_id=NM_173235.3'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load in the original GFF-based annotation\n",
    "models = pd.read_csv(sample_BFD.annot.path, skiprows = 7, header = None, sep = '\\t')\n",
    "display(models)\n",
    "\n",
    "attributes_column = 8\n",
    "\n",
    "# Check the structure of fields in the GFF additional fields section\n",
    "display(models[attributes_column][3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Dbxref</th>\n",
       "      <th>dbxref_dict</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1.0</td>\n",
       "      <td>58871917.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id0;Dbxref=taxon:7955;Name=1;chromosome=1;g...</td>\n",
       "      <td>{'ID': 'id0', 'Dbxref': 'taxon:7955', 'Name': ...</td>\n",
       "      <td>None</td>\n",
       "      <td>taxon:7955</td>\n",
       "      <td>{'taxon': '7955'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>6642.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Dbxref=GeneID:192301,ZFIN:ZDB-GENE-02...</td>\n",
       "      <td>{'ID': 'gene0', 'Dbxref': 'GeneID:192301,ZFIN:...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>GeneID:192301,ZFIN:ZDB-GENE-020419-25</td>\n",
       "      <td>{'GeneID': '192301', 'ZFIN': 'ZDB-GENE-020419-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>6642.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Dbxref=GeneID:192301,Genb...</td>\n",
       "      <td>{'ID': 'rna0', 'Parent': 'gene0', 'Dbxref': 'G...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...</td>\n",
       "      <td>{'GeneID': '192301', 'Genbank': 'NM_173235.3',...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11751.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Dbxref=GeneID:192301,Genban...</td>\n",
       "      <td>{'ID': 'id1', 'Parent': 'rna0', 'Dbxref': 'Gen...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...</td>\n",
       "      <td>{'GeneID': '192301', 'Genbank': 'NM_173235.3',...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11550.0</td>\n",
       "      <td>11625.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Dbxref=GeneID:192301,Genban...</td>\n",
       "      <td>{'ID': 'id2', 'Parent': 'rna0', 'Dbxref': 'Gen...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...</td>\n",
       "      <td>{'GeneID': '192301', 'Genbank': 'NM_173235.3',...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362355</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>16449.0</td>\n",
       "      <td>16520.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna65516;Parent=gene42296;Dbxref=GeneID:140...</td>\n",
       "      <td>{'ID': 'rna65516', 'Parent': 'gene42296', 'Dbx...</td>\n",
       "      <td>trnT</td>\n",
       "      <td>GeneID:140518,ZFIN:ZDB-GENE-011205-39</td>\n",
       "      <td>{'GeneID': '140518', 'ZFIN': 'ZDB-GENE-011205-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362356</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16449.0</td>\n",
       "      <td>16520.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id634507;Parent=rna65516;Dbxref=GeneID:1405...</td>\n",
       "      <td>{'ID': 'id634507', 'Parent': 'rna65516', 'Dbxr...</td>\n",
       "      <td>trnT</td>\n",
       "      <td>GeneID:140518,ZFIN:ZDB-GENE-011205-39</td>\n",
       "      <td>{'GeneID': '140518', 'ZFIN': 'ZDB-GENE-011205-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362357</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene42297;Dbxref=GeneID:140511,ZFIN:ZDB-GEN...</td>\n",
       "      <td>{'ID': 'gene42297', 'Dbxref': 'GeneID:140511,Z...</td>\n",
       "      <td>trnP</td>\n",
       "      <td>GeneID:140511,ZFIN:ZDB-GENE-011205-38</td>\n",
       "      <td>{'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362358</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna65517;Parent=gene42297;Dbxref=GeneID:140...</td>\n",
       "      <td>{'ID': 'rna65517', 'Parent': 'gene42297', 'Dbx...</td>\n",
       "      <td>trnP</td>\n",
       "      <td>GeneID:140511,ZFIN:ZDB-GENE-011205-38</td>\n",
       "      <td>{'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362359</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id634508;Parent=rna65517;Dbxref=GeneID:1405...</td>\n",
       "      <td>{'ID': 'id634508', 'Parent': 'rna65517', 'Dbxr...</td>\n",
       "      <td>trnP</td>\n",
       "      <td>GeneID:140511,ZFIN:ZDB-GENE-011205-38</td>\n",
       "      <td>{'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1360240 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0           1       2        3           4  5  6  7  \\\n",
       "0        NC_007112.6      RefSeq  region      1.0  58871917.0  .  +  .   \n",
       "1        NC_007112.6  BestRefSeq    gene   6642.0     11878.0  .  -  .   \n",
       "2        NC_007112.6  BestRefSeq    mRNA   6642.0     11878.0  .  -  .   \n",
       "3        NC_007112.6  BestRefSeq    exon  11751.0     11878.0  .  -  .   \n",
       "4        NC_007112.6  BestRefSeq    exon  11550.0     11625.0  .  -  .   \n",
       "...              ...         ...     ...      ...         ... .. .. ..   \n",
       "1362355  NC_002333.2      RefSeq    tRNA  16449.0     16520.0  .  +  .   \n",
       "1362356  NC_002333.2      RefSeq    exon  16449.0     16520.0  .  +  .   \n",
       "1362357  NC_002333.2      RefSeq    gene  16527.0     16596.0  .  -  .   \n",
       "1362358  NC_002333.2      RefSeq    tRNA  16527.0     16596.0  .  -  .   \n",
       "1362359  NC_002333.2      RefSeq    exon  16527.0     16596.0  .  -  .   \n",
       "\n",
       "                                                         8  \\\n",
       "0        ID=id0;Dbxref=taxon:7955;Name=1;chromosome=1;g...   \n",
       "1        ID=gene0;Dbxref=GeneID:192301,ZFIN:ZDB-GENE-02...   \n",
       "2        ID=rna0;Parent=gene0;Dbxref=GeneID:192301,Genb...   \n",
       "3        ID=id1;Parent=rna0;Dbxref=GeneID:192301,Genban...   \n",
       "4        ID=id2;Parent=rna0;Dbxref=GeneID:192301,Genban...   \n",
       "...                                                    ...   \n",
       "1362355  ID=rna65516;Parent=gene42296;Dbxref=GeneID:140...   \n",
       "1362356  ID=id634507;Parent=rna65516;Dbxref=GeneID:1405...   \n",
       "1362357  ID=gene42297;Dbxref=GeneID:140511,ZFIN:ZDB-GEN...   \n",
       "1362358  ID=rna65517;Parent=gene42297;Dbxref=GeneID:140...   \n",
       "1362359  ID=id634508;Parent=rna65517;Dbxref=GeneID:1405...   \n",
       "\n",
       "                                          field_dictionary gene_name  \\\n",
       "0        {'ID': 'id0', 'Dbxref': 'taxon:7955', 'Name': ...      None   \n",
       "1        {'ID': 'gene0', 'Dbxref': 'GeneID:192301,ZFIN:...     rpl24   \n",
       "2        {'ID': 'rna0', 'Parent': 'gene0', 'Dbxref': 'G...     rpl24   \n",
       "3        {'ID': 'id1', 'Parent': 'rna0', 'Dbxref': 'Gen...     rpl24   \n",
       "4        {'ID': 'id2', 'Parent': 'rna0', 'Dbxref': 'Gen...     rpl24   \n",
       "...                                                    ...       ...   \n",
       "1362355  {'ID': 'rna65516', 'Parent': 'gene42296', 'Dbx...      trnT   \n",
       "1362356  {'ID': 'id634507', 'Parent': 'rna65516', 'Dbxr...      trnT   \n",
       "1362357  {'ID': 'gene42297', 'Dbxref': 'GeneID:140511,Z...      trnP   \n",
       "1362358  {'ID': 'rna65517', 'Parent': 'gene42297', 'Dbx...      trnP   \n",
       "1362359  {'ID': 'id634508', 'Parent': 'rna65517', 'Dbxr...      trnP   \n",
       "\n",
       "                                                    Dbxref  \\\n",
       "0                                               taxon:7955   \n",
       "1                    GeneID:192301,ZFIN:ZDB-GENE-020419-25   \n",
       "2        GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...   \n",
       "3        GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...   \n",
       "4        GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...   \n",
       "...                                                    ...   \n",
       "1362355              GeneID:140518,ZFIN:ZDB-GENE-011205-39   \n",
       "1362356              GeneID:140518,ZFIN:ZDB-GENE-011205-39   \n",
       "1362357              GeneID:140511,ZFIN:ZDB-GENE-011205-38   \n",
       "1362358              GeneID:140511,ZFIN:ZDB-GENE-011205-38   \n",
       "1362359              GeneID:140511,ZFIN:ZDB-GENE-011205-38   \n",
       "\n",
       "                                               dbxref_dict  \n",
       "0                                        {'taxon': '7955'}  \n",
       "1        {'GeneID': '192301', 'ZFIN': 'ZDB-GENE-020419-...  \n",
       "2        {'GeneID': '192301', 'Genbank': 'NM_173235.3',...  \n",
       "3        {'GeneID': '192301', 'Genbank': 'NM_173235.3',...  \n",
       "4        {'GeneID': '192301', 'Genbank': 'NM_173235.3',...  \n",
       "...                                                    ...  \n",
       "1362355  {'GeneID': '140518', 'ZFIN': 'ZDB-GENE-011205-...  \n",
       "1362356  {'GeneID': '140518', 'ZFIN': 'ZDB-GENE-011205-...  \n",
       "1362357  {'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...  \n",
       "1362358  {'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...  \n",
       "1362359  {'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...  \n",
       "\n",
       "[1360240 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Remove any rows with NaNs\n",
    "models.dropna(inplace = True)\n",
    "\n",
    "# Extract field and database cross-ref (dbxref) information into columns\n",
    "models['field_dictionary'] = models[attributes_column].apply(convert_fields_to_dict_gff)\n",
    "models['gene_name'] = [d.get('gene') for d in models['field_dictionary']]\n",
    "models['Dbxref'] = [d.get('Dbxref') for d in models['field_dictionary']]\n",
    "models['dbxref_dict'] = models['Dbxref'].apply(convert_dbxref_to_dict)\n",
    "\n",
    "display(models)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Extract gene IDs for mapping to UniprotKB\n",
    "Specify which set of identifiers will be use to query the [Uniprot ID Mapping Tool](https://www.uniprot.org/id-mapping) via API.\n",
    "\n",
    "If using an identifier from the `dbxref_dict`, specify the name via string in the `dbxref_datafield` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ZFIN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rpl24</td>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>cep97</td>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>nfkbiz</td>\n",
       "      <td>ZDB-GENE-071024-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>eed</td>\n",
       "      <td>ZDB-GENE-050417-287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>zgc:110091</td>\n",
       "      <td>ZDB-GENE-050417-34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362347</th>\n",
       "      <td>ND6</td>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362349</th>\n",
       "      <td>trnE</td>\n",
       "      <td>ZDB-GENE-011205-37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362352</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362354</th>\n",
       "      <td>trnT</td>\n",
       "      <td>ZDB-GENE-011205-39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362357</th>\n",
       "      <td>trnP</td>\n",
       "      <td>ZDB-GENE-011205-38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22031 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_name                 ZFIN\n",
       "1             rpl24   ZDB-GENE-020419-25\n",
       "15            cep97   ZDB-GENE-031030-11\n",
       "74           nfkbiz    ZDB-GENE-071024-1\n",
       "102             eed  ZDB-GENE-050417-287\n",
       "128      zgc:110091   ZDB-GENE-050417-34\n",
       "...             ...                  ...\n",
       "1362347         ND6   ZDB-GENE-011205-13\n",
       "1362349        trnE   ZDB-GENE-011205-37\n",
       "1362352        CYTB   ZDB-GENE-011205-17\n",
       "1362354        trnT   ZDB-GENE-011205-39\n",
       "1362357        trnP   ZDB-GENE-011205-38\n",
       "\n",
       "[22031 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dbxref_datafield = 'ZFIN'\n",
    "\n",
    "models.dropna(axis = 0, subset = ['dbxref_dict'], inplace = True)\n",
    "models[dbxref_datafield] = [d.get(dbxref_datafield) for d in models['dbxref_dict']]\n",
    "\n",
    "models_subset = models[['gene_name', dbxref_datafield]].dropna().drop_duplicates()\n",
    "\n",
    "display(models_subset)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Generate gene list file to query Uniprot ID Mapping API\n",
    "Generate a text file ending in `_ids.txt` for sending to the ID mapping API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 22031 gene ids to /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_ZFIN_ids.txt\n"
     ]
    }
   ],
   "source": [
    "gene_list = models_subset[dbxref_datafield].unique()\n",
    "\n",
    "genelist_object = GeneListFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    sources = [sample_BFD.annot],\n",
    "    genes = gene_list,\n",
    "    identifier = dbxref_datafield\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Query Uniprot ID Mapping API\n",
    "Specify the `from_type` variable based on the Uniprot name of the identifier.  \n",
    "The table below lists some databases and the `from_type` string that the API accepts for that datatype.  \n",
    "\n",
    "| datatype | `from_type` string | description |\n",
    "| ---: | :--- | :--- |\n",
    "| Mouse Genome Informatics | `MGI` | ID starts with `MGI:` |\n",
    "| Zebrafish Information Network | `ZFIN` | ID starts with `ZDB-GENE-` |\n",
    "| Xenbase | `Xenbase` | ID starts with `XB-GENE-` |\n",
    "\n",
    "__NOTE:__ You may have to run the cell below twice - UniProt sometimes throws an \"Resource not found\" message on the first query to the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  419k    0    52  100  418k     26   212k  0:00:01  0:00:01 --:--:--  212k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100    24    0    24    0     0     28      0 --:--:-- --:--:-- --:--:--    28\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>From</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>Q8JGR4</td>\n",
       "      <td>RL24_DANRE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>60S ribosomal protein L24</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>A0A0R4IMS3</td>\n",
       "      <td>A0A0R4IMS3_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>60S ribosomal protein L24</td>\n",
       "      <td>rpl24 SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ICF0</td>\n",
       "      <td>A0A0R4ICF0_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Centrosomal protein 97</td>\n",
       "      <td>cep97 SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ISF5</td>\n",
       "      <td>A0A0R4ISF5_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Centrosomal protein 97</td>\n",
       "      <td>cep97 SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4IX25</td>\n",
       "      <td>A0A0R4IX25_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Centrosomal protein 97</td>\n",
       "      <td>cep97 SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54701</th>\n",
       "      <td>ZDB-GENE-011205-12</td>\n",
       "      <td>Q9MIY0</td>\n",
       "      <td>NU5M_DANRE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NADH-ubiquinone oxidoreductase chain 5 (EC 7.1...</td>\n",
       "      <td>mt-nd5 mtnd5 nd5</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54702</th>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>Q9MIX9</td>\n",
       "      <td>NU6M_DANRE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NADH-ubiquinone oxidoreductase chain 6 (EC 7.1...</td>\n",
       "      <td>mt-nd6 mtnd6 nd6</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54703</th>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>A0A0A0VG13</td>\n",
       "      <td>A0A0A0VG13_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>NADH-ubiquinone oxidoreductase chain 6 (EC 7.1...</td>\n",
       "      <td>ND6 mt-nd6 SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54704</th>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>Q9MIX8</td>\n",
       "      <td>CYB_DANRE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Cytochrome b (Complex III subunit 3) (Complex ...</td>\n",
       "      <td>mt-cyb cob cytb mtcyb</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54705</th>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>A0A0A0VDB5</td>\n",
       "      <td>A0A0A0VDB5_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Cytochrome b</td>\n",
       "      <td>CYTB mt-cyb SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>380</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54706 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     From       Entry        Entry Name    Reviewed  \\\n",
       "0      ZDB-GENE-020419-25      Q8JGR4        RL24_DANRE    reviewed   \n",
       "1      ZDB-GENE-020419-25  A0A0R4IMS3  A0A0R4IMS3_DANRE  unreviewed   \n",
       "2      ZDB-GENE-031030-11  A0A0R4ICF0  A0A0R4ICF0_DANRE  unreviewed   \n",
       "3      ZDB-GENE-031030-11  A0A0R4ISF5  A0A0R4ISF5_DANRE  unreviewed   \n",
       "4      ZDB-GENE-031030-11  A0A0R4IX25  A0A0R4IX25_DANRE  unreviewed   \n",
       "...                   ...         ...               ...         ...   \n",
       "54701  ZDB-GENE-011205-12      Q9MIY0        NU5M_DANRE    reviewed   \n",
       "54702  ZDB-GENE-011205-13      Q9MIX9        NU6M_DANRE    reviewed   \n",
       "54703  ZDB-GENE-011205-13  A0A0A0VG13  A0A0A0VG13_DANRE  unreviewed   \n",
       "54704  ZDB-GENE-011205-17      Q9MIX8         CYB_DANRE    reviewed   \n",
       "54705  ZDB-GENE-011205-17  A0A0A0VDB5  A0A0A0VDB5_DANRE  unreviewed   \n",
       "\n",
       "                                           Protein names  \\\n",
       "0                              60S ribosomal protein L24   \n",
       "1                              60S ribosomal protein L24   \n",
       "2                                 Centrosomal protein 97   \n",
       "3                                 Centrosomal protein 97   \n",
       "4                                 Centrosomal protein 97   \n",
       "...                                                  ...   \n",
       "54701  NADH-ubiquinone oxidoreductase chain 5 (EC 7.1...   \n",
       "54702  NADH-ubiquinone oxidoreductase chain 6 (EC 7.1...   \n",
       "54703  NADH-ubiquinone oxidoreductase chain 6 (EC 7.1...   \n",
       "54704  Cytochrome b (Complex III subunit 3) (Complex ...   \n",
       "54705                                       Cytochrome b   \n",
       "\n",
       "                   Gene Names                                     Organism  \\\n",
       "0                       rpl24  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "1            rpl24 SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "2            cep97 SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "3            cep97 SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "4            cep97 SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "...                       ...                                          ...   \n",
       "54701        mt-nd5 mtnd5 nd5  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "54702        mt-nd6 mtnd6 nd6  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "54703   ND6 mt-nd6 SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "54704   mt-cyb cob cytb mtcyb  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "54705  CYTB mt-cyb SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "\n",
       "       Length  \n",
       "0         157  \n",
       "1         157  \n",
       "2         599  \n",
       "3          40  \n",
       "4         240  \n",
       "...       ...  \n",
       "54701     606  \n",
       "54702     172  \n",
       "54703     172  \n",
       "54704     380  \n",
       "54705     380  \n",
       "\n",
       "[54706 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from_type = 'ZFIN'\n",
    "to_type = 'UniProtKB'\n",
    "\n",
    "uniprot_idmm_object = genelist_object.get_uniprot_ids(ID_MAPPER_LOC, from_type, to_type)\n",
    "sample_BFD.add_keyfile('uniprot_idmm', uniprot_idmm_object, overwrite = True)\n",
    "\n",
    "uniprot_idmm = pd.read_csv(sample_BFD.uniprot_idmm.path, sep = '\\t')\n",
    "display(uniprot_idmm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Extract results and generate Uniprot IDMM\n",
    "Generates an idmm that links `gene_name`, the `dbxref_datafield` seleted above, and `uniprot_id` returned by API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_30283/3311337425.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  uniprot_idpairs.rename(columns = {'From': dbxref_datafield, 'Entry': 'uniprot_id'}, inplace = True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ZFIN</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>Q8JGR4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>A0A0R4IMS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ICF0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ISF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4IX25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54701</th>\n",
       "      <td>ZDB-GENE-011205-12</td>\n",
       "      <td>Q9MIY0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54702</th>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>Q9MIX9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54703</th>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>A0A0A0VG13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54704</th>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>Q9MIX8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54705</th>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>A0A0A0VDB5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54706 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     ZFIN  uniprot_id\n",
       "0      ZDB-GENE-020419-25      Q8JGR4\n",
       "1      ZDB-GENE-020419-25  A0A0R4IMS3\n",
       "2      ZDB-GENE-031030-11  A0A0R4ICF0\n",
       "3      ZDB-GENE-031030-11  A0A0R4ISF5\n",
       "4      ZDB-GENE-031030-11  A0A0R4IX25\n",
       "...                   ...         ...\n",
       "54701  ZDB-GENE-011205-12      Q9MIY0\n",
       "54702  ZDB-GENE-011205-13      Q9MIX9\n",
       "54703  ZDB-GENE-011205-13  A0A0A0VG13\n",
       "54704  ZDB-GENE-011205-17      Q9MIX8\n",
       "54705  ZDB-GENE-011205-17  A0A0A0VDB5\n",
       "\n",
       "[54706 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ZFIN</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rpl24</td>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>Q8JGR4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rpl24</td>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>A0A0R4IMS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cep97</td>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ICF0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cep97</td>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ISF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cep97</td>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4IX25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54701</th>\n",
       "      <td>ND5</td>\n",
       "      <td>ZDB-GENE-011205-12</td>\n",
       "      <td>Q9MIY0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54702</th>\n",
       "      <td>ND6</td>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>Q9MIX9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54703</th>\n",
       "      <td>ND6</td>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>A0A0A0VG13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54704</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>Q9MIX8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54705</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>A0A0A0VDB5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54706 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name                ZFIN  uniprot_id\n",
       "0         rpl24  ZDB-GENE-020419-25      Q8JGR4\n",
       "1         rpl24  ZDB-GENE-020419-25  A0A0R4IMS3\n",
       "2         cep97  ZDB-GENE-031030-11  A0A0R4ICF0\n",
       "3         cep97  ZDB-GENE-031030-11  A0A0R4ISF5\n",
       "4         cep97  ZDB-GENE-031030-11  A0A0R4IX25\n",
       "...         ...                 ...         ...\n",
       "54701       ND5  ZDB-GENE-011205-12      Q9MIY0\n",
       "54702       ND6  ZDB-GENE-011205-13      Q9MIX9\n",
       "54703       ND6  ZDB-GENE-011205-13  A0A0A0VG13\n",
       "54704      CYTB  ZDB-GENE-011205-17      Q9MIX8\n",
       "54705      CYTB  ZDB-GENE-011205-17  A0A0A0VDB5\n",
       "\n",
       "[54706 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "uniprot_idpairs = uniprot_idmm[['From', 'Entry']]\n",
    "uniprot_idpairs.rename(columns = {'From': dbxref_datafield, 'Entry': 'uniprot_id'}, inplace = True)\n",
    "display(uniprot_idpairs)\n",
    "\n",
    "uniprot_output_idmm = models_subset.merge(uniprot_idpairs, on = dbxref_datafield)\n",
    "display(uniprot_output_idmm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate a filename and file for the idmm\n",
    "uniprot_output_idmm_filename = '_'.join([sample_BFD.species_prefix, conditions, 'uniprot-idmm.tsv'])\n",
    "uniprot_output_idmm_object = IdmmFile(\n",
    "    filename = uniprot_output_idmm_filename, \n",
    "    sampledict = sample_BFD.sampledict, \n",
    "    kind = 'uniprot_idmm', \n",
    "    sources = [sample_BFD.annot]\n",
    ")\n",
    "\n",
    "# save to file and add to the BioFileDocket\n",
    "uniprot_output_idmm.to_csv(uniprot_output_idmm_object.path, sep = '\\t')\n",
    "sample_BFD.add_keyfile('uniprot_idmm', uniprot_output_idmm_object)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8. Convert GFF to GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted file GCF_000002035.5_GRCz10_genomic.gtf already exists at:\n",
      " /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.gtf\n"
     ]
    }
   ],
   "source": [
    "# convert the GFF file to GTF using gffread\n",
    "models_asgtf = sample_BFD.annot.to_gtf(GFFREAD_LOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>6642</td>\n",
       "      <td>11878</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>6642</td>\n",
       "      <td>6760</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>6892</td>\n",
       "      <td>6955</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>9558</td>\n",
       "      <td>9694</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>10081</td>\n",
       "      <td>10191</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258174</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>15308</td>\n",
       "      <td>16448</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"gene42295\"; gene_name \"CYTB\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258175</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>16449</td>\n",
       "      <td>16520</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65516\"; gene_id \"gene42296\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258176</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16449</td>\n",
       "      <td>16520</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65516\"; gene_id \"gene42296\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258177</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>16527</td>\n",
       "      <td>16596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65517\"; gene_id \"gene42297\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258178</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16527</td>\n",
       "      <td>16596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65517\"; gene_id \"gene42297\";...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1258179 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0           1           2      3      4  5  6  7  \\\n",
       "0        NC_007112.6  BestRefSeq  transcript   6642  11878  .  -  .   \n",
       "1        NC_007112.6  BestRefSeq        exon   6642   6760  .  -  .   \n",
       "2        NC_007112.6  BestRefSeq        exon   6892   6955  .  -  .   \n",
       "3        NC_007112.6  BestRefSeq        exon   9558   9694  .  -  .   \n",
       "4        NC_007112.6  BestRefSeq        exon  10081  10191  .  -  .   \n",
       "...              ...         ...         ...    ...    ... .. .. ..   \n",
       "1258174  NC_002333.2      RefSeq         CDS  15308  16448  .  +  0   \n",
       "1258175  NC_002333.2      RefSeq  transcript  16449  16520  .  +  .   \n",
       "1258176  NC_002333.2      RefSeq        exon  16449  16520  .  +  .   \n",
       "1258177  NC_002333.2      RefSeq  transcript  16527  16596  .  -  .   \n",
       "1258178  NC_002333.2      RefSeq        exon  16527  16596  .  -  .   \n",
       "\n",
       "                                                         8  \n",
       "0        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "1        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "2        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "3        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "4        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "...                                                    ...  \n",
       "1258174       transcript_id \"gene42295\"; gene_name \"CYTB\";  \n",
       "1258175  transcript_id \"rna65516\"; gene_id \"gene42296\";...  \n",
       "1258176  transcript_id \"rna65516\"; gene_id \"gene42296\";...  \n",
       "1258177  transcript_id \"rna65517\"; gene_id \"gene42297\";...  \n",
       "1258178  transcript_id \"rna65517\"; gene_id \"gene42297\";...  \n",
       "\n",
       "[1258179 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'transcript_id \"rna0\"; gene_id \"gene0\"; gene_name \"rpl24\";'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load the newly-generated GTF file as a dataframe\n",
    "models_asgtf_df = pd.read_csv(models_asgtf.path, skiprows = 0, header = None, sep = '\\t')\n",
    "\n",
    "display(models_asgtf_df)\n",
    "display(models_asgtf_df[attributes_column][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>6642</td>\n",
       "      <td>11878</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>6642</td>\n",
       "      <td>6760</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>6892</td>\n",
       "      <td>6955</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>9558</td>\n",
       "      <td>9694</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>10081</td>\n",
       "      <td>10191</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258173</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15308</td>\n",
       "      <td>16448</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"gene42295\"; gene_id \"gene42295\"...</td>\n",
       "      <td>{'transcript_id': 'gene42295', 'gene_id': 'gen...</td>\n",
       "      <td>CYTB</td>\n",
       "      <td>gene42295</td>\n",
       "      <td>gene42295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258175</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>16449</td>\n",
       "      <td>16520</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65516\"; gene_id \"gene42296\";...</td>\n",
       "      <td>{'transcript_id': 'rna65516', 'gene_id': 'gene...</td>\n",
       "      <td>trnT</td>\n",
       "      <td>gene42296</td>\n",
       "      <td>rna65516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258176</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16449</td>\n",
       "      <td>16520</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65516\"; gene_id \"gene42296\";...</td>\n",
       "      <td>{'transcript_id': 'rna65516', 'gene_id': 'gene...</td>\n",
       "      <td>trnT</td>\n",
       "      <td>gene42296</td>\n",
       "      <td>rna65516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258177</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>16527</td>\n",
       "      <td>16596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65517\"; gene_id \"gene42297\";...</td>\n",
       "      <td>{'transcript_id': 'rna65517', 'gene_id': 'gene...</td>\n",
       "      <td>trnP</td>\n",
       "      <td>gene42297</td>\n",
       "      <td>rna65517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258178</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16527</td>\n",
       "      <td>16596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65517\"; gene_id \"gene42297\";...</td>\n",
       "      <td>{'transcript_id': 'rna65517', 'gene_id': 'gene...</td>\n",
       "      <td>trnP</td>\n",
       "      <td>gene42297</td>\n",
       "      <td>rna65517</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>699236 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0           1           2      3      4  5  6  7  \\\n",
       "0        NC_007112.6  BestRefSeq  transcript   6642  11878  .  -  .   \n",
       "1        NC_007112.6  BestRefSeq        exon   6642   6760  .  -  .   \n",
       "2        NC_007112.6  BestRefSeq        exon   6892   6955  .  -  .   \n",
       "3        NC_007112.6  BestRefSeq        exon   9558   9694  .  -  .   \n",
       "4        NC_007112.6  BestRefSeq        exon  10081  10191  .  -  .   \n",
       "...              ...         ...         ...    ...    ... .. .. ..   \n",
       "1258173  NC_002333.2      RefSeq  transcript  15308  16448  .  +  .   \n",
       "1258175  NC_002333.2      RefSeq  transcript  16449  16520  .  +  .   \n",
       "1258176  NC_002333.2      RefSeq        exon  16449  16520  .  +  .   \n",
       "1258177  NC_002333.2      RefSeq  transcript  16527  16596  .  -  .   \n",
       "1258178  NC_002333.2      RefSeq        exon  16527  16596  .  -  .   \n",
       "\n",
       "                                                         8  \\\n",
       "0        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "1        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "2        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "3        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "4        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "...                                                    ...   \n",
       "1258173  transcript_id \"gene42295\"; gene_id \"gene42295\"...   \n",
       "1258175  transcript_id \"rna65516\"; gene_id \"gene42296\";...   \n",
       "1258176  transcript_id \"rna65516\"; gene_id \"gene42296\";...   \n",
       "1258177  transcript_id \"rna65517\"; gene_id \"gene42297\";...   \n",
       "1258178  transcript_id \"rna65517\"; gene_id \"gene42297\";...   \n",
       "\n",
       "                                          field_dictionary gene_name  \\\n",
       "0        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...     rpl24   \n",
       "1        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...     rpl24   \n",
       "2        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...     rpl24   \n",
       "3        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...     rpl24   \n",
       "4        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...     rpl24   \n",
       "...                                                    ...       ...   \n",
       "1258173  {'transcript_id': 'gene42295', 'gene_id': 'gen...      CYTB   \n",
       "1258175  {'transcript_id': 'rna65516', 'gene_id': 'gene...      trnT   \n",
       "1258176  {'transcript_id': 'rna65516', 'gene_id': 'gene...      trnT   \n",
       "1258177  {'transcript_id': 'rna65517', 'gene_id': 'gene...      trnP   \n",
       "1258178  {'transcript_id': 'rna65517', 'gene_id': 'gene...      trnP   \n",
       "\n",
       "           gene_id transcript_id  \n",
       "0            gene0          rna0  \n",
       "1            gene0          rna0  \n",
       "2            gene0          rna0  \n",
       "3            gene0          rna0  \n",
       "4            gene0          rna0  \n",
       "...            ...           ...  \n",
       "1258173  gene42295     gene42295  \n",
       "1258175  gene42296      rna65516  \n",
       "1258176  gene42296      rna65516  \n",
       "1258177  gene42297      rna65517  \n",
       "1258178  gene42297      rna65517  \n",
       "\n",
       "[699236 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Use a custom function to extract useful fields from the additional fields section (column 8)\n",
    "# Pull from that dict to fill in additional useful columns\n",
    "models_asgtf_df['field_dictionary'] = models_asgtf_df[attributes_column].apply(convert_fields_to_dict_gtf)\n",
    "models_asgtf_df['gene_name'] = [d.get('gene_name') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['gene_id'] = [d.get('gene_id') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['transcript_id'] = [d.get('transcript_id') for d in models_asgtf_df['field_dictionary']]\n",
    "\n",
    "# Remove CDS annotations because they interfere with TransDecoder cDNA generation\n",
    "models_asgtf_df = models_asgtf_df[models_asgtf_df[2] != 'CDS']\n",
    "display(models_asgtf_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9. Generate gtf-idmm\n",
    "This file maps the `gene_name` to `gene_id` and `transcript_id` fields generated by the conversion from GFF to GTF, which will be needed for downstream processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cep97</td>\n",
       "      <td>gene1</td>\n",
       "      <td>rna2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>cep97</td>\n",
       "      <td>gene1</td>\n",
       "      <td>rna1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>cep97</td>\n",
       "      <td>gene1</td>\n",
       "      <td>rna3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>nfkbiz</td>\n",
       "      <td>gene2</td>\n",
       "      <td>rna4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258169</th>\n",
       "      <td>ND6</td>\n",
       "      <td>gene42293</td>\n",
       "      <td>gene42293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258171</th>\n",
       "      <td>trnE</td>\n",
       "      <td>gene42294</td>\n",
       "      <td>rna65515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258173</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>gene42295</td>\n",
       "      <td>gene42295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258175</th>\n",
       "      <td>trnT</td>\n",
       "      <td>gene42296</td>\n",
       "      <td>rna65516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258177</th>\n",
       "      <td>trnP</td>\n",
       "      <td>gene42297</td>\n",
       "      <td>rna65517</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65914 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name    gene_id transcript_id\n",
       "0           rpl24      gene0          rna0\n",
       "13          cep97      gene1          rna2\n",
       "25          cep97      gene1          rna1\n",
       "48          cep97      gene1          rna3\n",
       "71         nfkbiz      gene2          rna4\n",
       "...           ...        ...           ...\n",
       "1258169       ND6  gene42293     gene42293\n",
       "1258171      trnE  gene42294      rna65515\n",
       "1258173      CYTB  gene42295     gene42295\n",
       "1258175      trnT  gene42296      rna65516\n",
       "1258177      trnP  gene42297      rna65517\n",
       "\n",
       "[65914 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract gene_name, gene_id, and transcript_id fields to generate an ID mapping matrix (idmm)\n",
    "idmm_df = models_asgtf_df[['gene_name', 'gene_id', 'transcript_id']].drop_duplicates()\n",
    "idmm_df.dropna(inplace = True)\n",
    "display(idmm_df)\n",
    "\n",
    "# generate a filename and file for the idmm\n",
    "idmm_filename = '_'.join([sample_BFD.species_prefix, conditions, 'gtf-idmm.tsv'])\n",
    "idmm = IdmmFile(\n",
    "    filename = idmm_filename, \n",
    "    sampledict = sample_BFD.sampledict, \n",
    "    kind = 'gtf_idmm', \n",
    "    sources = [sample_BFD.annot]\n",
    ")\n",
    "\n",
    "# save to file and add to the BioFileDocket\n",
    "idmm_df.to_csv(idmm.path, sep = '\\t')\n",
    "sample_BFD.add_keyfile('gtf_idmm', idmm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 10. Generate updated gtf\n",
    "Generated an updated GTF file using transcript_id as the key. For some datasets, transcripts do not consistently get gene names and gene IDs added, which causes Transdecoder to throw errors. This resolves that problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "models_asgtf_updated_df = models_asgtf_df.merge(idmm_df, on = 'transcript_id')\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_name': x['gene_name_y']}), axis = 1)\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_id': x['gene_id_y']}), axis = 1)\n",
    "models_asgtf_updated_df['new_fields'] = models_asgtf_updated_df['field_dictionary'].apply(convert_dict_to_fields_gtf)\n",
    "models_asgtf_updated_df = models_asgtf_updated_df[[0, 1, 2, 3, 4, 5, 6, 7, 'new_fields']]\n",
    "\n",
    "models_asgtf_updated_filename = models_asgtf.filename.replace('.gtf', '_updated.gtf')\n",
    "models_asgtf_updated = GenomeGtfFile(\n",
    "    filename = models_asgtf_updated_filename, \n",
    "    sampledict = sample_BFD.sampledict, \n",
    "    reference_genome = sample_BFD.genome_fasta\n",
    ")\n",
    "\n",
    "models_asgtf_updated_df.to_csv(models_asgtf_updated.path, header = None, index = None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 11. Generate cDNA and peptide files\n",
    "Using the updated gtf file and genome file, generate cDNA sequence.\n",
    "\n",
    "Then, using the cDNA sequence, generate peptide sequences using transdecoder.\n",
    "\n",
    "Expect this step to take some time, probably ~20-30min."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/compute_base_probs.pl /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna 0 > /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//base_freqs.dat, checkpoint [/home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints_longorfs/base_freqs_file.ok] exists.\n",
      "-skipping long orf extraction, already completed earlier as per checkpoint: /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints_longorfs/TD.longorfs.ok\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds 5000 5000 > /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000, checkpoint [/home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/get_longest_orfs.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/exclude_similar_proteins.pl /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000 > /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr, checkpoint [/home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/nr.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr 500 > /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest, checkpoint [/home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/top_train_select.ok] exists.\n",
      "PCT_GC: 46.2\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/seq_n_baseprobs_to_loglikelihood_vals.pl /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//base_freqs.dat > /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//hexamer.scores, checkpoint [/home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/hexamer_scores.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/score_CDS_likelihood_all_6_frames.pl /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//hexamer.scores > /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.scores, checkpoint [/home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints/score_cdss.ok] exists.\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/select_best_ORFs_per_transcript.pl --gff3_file /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.gff3 --cds_scores /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.scores  --min_length_auto_accept 749  > /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 \n",
      "Selecting best orfs\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/train_start_PWM.pl --transcripts /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna --selected_orfs /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest --out_prefix /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement\n",
      "Training start codon pattern recognition* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/build_atgPWM_+-.pl  --transcripts /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna  --selected_orfs /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest  --out_prefix /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement --pwm_left 20 --pwm_right 10 \n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scoring.+-.pl  --features_plus /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.+.features  --features_minus /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.-.features  --atg_position 20  > /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.feature.scores\n",
      "-round: 1\n",
      "-round: 2\n",
      "-round: 3\n",
      "-round: 4\n",
      "-round: 5\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scores_to_ROC.pl /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.feature.scores > /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.feature.scores.roc\n",
      "-parsing scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/plot_ROC.Rscript /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.feature.scores.roc || :\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/compute_AUC.pl /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.feature.scores.roc\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.+.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.-.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/deplete_feature_noise.pl  --features_plus /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.+.features  --pwm_minus /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.-.pwm  --out_prefix /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.enhanced\n",
      "num features: 397\tnum_incorporate: 119\n",
      "-feature swap of score: -2.21692784225968 instead of -2.82480127800276\n",
      "-feature swap of score: -0.696828751663083 instead of -2.21669338229439\n",
      "-feature swap of score: 0.763599817269247 instead of -2.08002913214012\n",
      "-feature swap of score: -1.11519786089795 instead of -2.13198058788708\n",
      "-feature swap of score: 3.53408341556213 instead of -2.26417417643694\n",
      "-feature swap of score: 1.86101588908586 instead of -2.22369485081157\n",
      "-feature swap of score: -0.218665863983171 instead of -2.28581831991281\n",
      "-feature swap of score: 2.02886036026552 instead of -1.9859531223204\n",
      "-feature swap of score: 1.10702231475703 instead of -2.17554202724522\n",
      "-feature swap of score: 2.5383655822014 instead of -2.20399741332803\n",
      "-feature swap of score: 2.31679310064595 instead of -1.82450535806742\n",
      "-feature swap of score: -0.659719692413243 instead of -1.78167925724792\n",
      "-feature swap of score: 1.94987485463704 instead of -1.7706989701408\n",
      "-feature swap of score: 2.83377523996852 instead of -1.76507044820695\n",
      "-feature swap of score: 2.04112913248544 instead of -1.6645444711762\n",
      "-feature swap of score: 2.23140986944903 instead of -1.43593801295698\n",
      "-feature swap of score: -0.468853215613425 instead of -1.3662057797844\n",
      "-feature swap of score: 2.8501343048918 instead of -1.33918648637577\n",
      "-feature swap of score: -0.82297065064561 instead of -1.26654294850041\n",
      "-feature swap of score: 1.74847096833361 instead of -1.23636035760393\n",
      "-feature swap of score: 2.35820960262574 instead of -1.33451927243332\n",
      "-feature swap of score: 0.174525520245121 instead of -1.33831867169981\n",
      "-feature swap of score: -0.589301270618693 instead of -1.32496614856519\n",
      "-feature swap of score: 0.253617627828586 instead of -1.21867443551391\n",
      "-feature swap of score: 1.52673011593545 instead of -1.0299387224734\n",
      "-feature swap of score: -0.213045721889497 instead of -0.85861641458173\n",
      "-feature swap of score: 0.958395940861797 instead of -0.820298163482403\n",
      "-feature swap of score: -0.17652808834708 instead of -0.927096043151601\n",
      "-feature swap of score: -0.713001792305057 instead of -0.943313963296813\n",
      "-feature swap of score: 0.569552634718937 instead of -1.13043289339032\n",
      "-feature swap of score: 0.786660533945287 instead of -1.06490212259205\n",
      "-feature swap of score: -0.864984377267684 instead of -1.03253596898551\n",
      "-feature swap of score: 0.526490575345434 instead of -0.914478403205129\n",
      "-feature swap of score: -0.784147873225043 instead of -1.00113771518839\n",
      "-feature swap of score: 1.3138356220685 instead of -0.64912022078675\n",
      "-feature swap of score: 2.79690373635375 instead of -0.601493959624492\n",
      "-feature swap of score: 1.76065980912418 instead of -0.365952128306905\n",
      "-feature swap of score: 2.26891484656023 instead of -0.378676603838264\n",
      "-feature swap of score: -0.210905168398584 instead of -0.446565302579578\n",
      "-feature swap of score: 1.89889203183362 instead of -0.358936596302073\n",
      "-feature swap of score: 1.00688750298899 instead of -0.353916354013543\n",
      "-feature swap of score: 1.79484698552421 instead of -0.65989441137551\n",
      "-feature swap of score: 1.19827463110793 instead of -0.514872410851988\n",
      "-feature swap of score: 5.42964733568679 instead of -0.500999227608865\n",
      "-feature swap of score: 0.79055074424589 instead of -0.305912642936535\n",
      "-feature swap of score: 0.504275415107053 instead of -0.0829589300527132\n",
      "-feature swap of score: 1.6071457004952 instead of 0.0798266020192396\n",
      "-feature swap of score: 2.55300576811539 instead of 0.00962164936263132\n",
      "-feature swap of score: 1.6198141794378 instead of -0.275832437734224\n",
      "-feature swap of score: 0.923907367442004 instead of 0.0616025828884617\n",
      "-feature swap of score: 3.5634443444382 instead of 0.0244157032017905\n",
      "-feature swap of score: 4.46845866730229 instead of 0.0398402559058269\n",
      "-feature swap of score: 1.60001406794808 instead of -0.0230489817681658\n",
      "-feature swap of score: 1.49782608245693 instead of 0.0873432674626391\n",
      "-feature swap of score: 2.33907393406013 instead of -0.0916375721415191\n",
      "-feature swap of score: 1.86427405105594 instead of -0.0766140301656006\n",
      "-feature swap of score: 4.33559769833044 instead of -0.204162854577705\n",
      "-feature swap of score: 4.89324968832146 instead of -0.0290853636308332\n",
      "-feature swap of score: 0.630765880399525 instead of -0.0687196902425364\n",
      "-feature swap of score: 1.95903724520605 instead of 0.126086386012773\n",
      "-feature swap of score: 2.60481753120288 instead of 0.191334999026586\n",
      "-feature swap of score: 4.3608068902744 instead of -0.156283516370599\n",
      "-feature swap of score: 1.92944996439441 instead of 0.290193244748163\n",
      "-feature swap of score: 1.86578603701568 instead of 0.225229275180886\n",
      "-feature swap of score: 1.94018317547513 instead of 0.116538243846511\n",
      "-feature swap of score: 2.38816702354996 instead of 0.28047302730733\n",
      "-feature swap of score: 1.81725122339717 instead of 0.386253265777538\n",
      "-feature swap of score: 1.92520130938505 instead of 0.503716109145222\n",
      "-feature swap of score: 1.5452210972483 instead of 0.53356735390578\n",
      "-feature swap of score: 2.88801770441359 instead of 0.492389607438148\n",
      "-feature swap of score: 1.27405643003176 instead of 0.432658714686395\n",
      "-feature swap of score: 6.05437493391006 instead of 0.194328395891367\n",
      "-feature swap of score: 0.616082456763985 instead of 0.413620079700086\n",
      "-feature swap of score: 4.78033553337365 instead of 0.384021327172111\n",
      "-feature swap of score: 1.86738731571186 instead of 0.313622161722505\n",
      "-feature swap of score: 4.88405640273726 instead of 0.532053576165131\n",
      "-feature swap of score: 1.75822800393298 instead of 0.311359625224178\n",
      "-feature swap of score: 1.19262966750173 instead of 0.578854442017519\n",
      "-feature swap of score: 3.24795027257256 instead of 0.711014404123423\n",
      "-feature swap of score: 4.23546031605856 instead of 0.799980961414935\n",
      "-feature swap of score: 1.69783338063326 instead of 0.790876916862567\n",
      "-feature swap of score: 1.81965502104673 instead of 0.665324393576442\n",
      "-feature swap of score: 2.69788577770424 instead of 0.644831256446135\n",
      "-feature swap of score: 3.40645518604735 instead of 0.664482830522306\n",
      "-feature swap of score: 2.96028090277577 instead of 0.610638586788851\n",
      "-feature swap of score: 4.82604103608634 instead of 1.03340711309669\n",
      "-feature swap of score: 2.50275871642906 instead of 1.07193423555004\n",
      "-feature swap of score: 3.60957312686455 instead of 0.956754186985914\n",
      "-feature swap of score: 4.4433301270039 instead of 1.09402741546021\n",
      "-feature swap of score: 3.796273651055 instead of 1.02467993649294\n",
      "-feature swap of score: 3.89046488250719 instead of 1.05325815845505\n",
      "-feature swap of score: 2.0867183861946 instead of 1.08090722177749\n",
      "-feature swap of score: 1.57796425823252 instead of 0.922201284791306\n",
      "-feature swap of score: 5.33850009445087 instead of 0.928992085852854\n",
      "-feature swap of score: 2.32343864603735 instead of 0.898484973715832\n",
      "-feature swap of score: 5.91450664599095 instead of 1.22525033588609\n",
      "-feature swap of score: 2.11672909445256 instead of 1.11491412545647\n",
      "-feature swap of score: 1.9099999944661 instead of 1.33961086227587\n",
      "-feature swap of score: 1.32614707249276 instead of 1.10819707516468\n",
      "-feature swap of score: 2.18649035304408 instead of 1.08677612853226\n",
      "-feature swap of score: 3.19770920810262 instead of 1.18217498300053\n",
      "-feature swap of score: 3.87207599357921 instead of 1.38330442423035\n",
      "-feature swap of score: 3.34411985762808 instead of 1.35902255088362\n",
      "-feature swap of score: 1.64762094990613 instead of 1.41633056149429\n",
      "-feature swap of score: 6.85681092414695 instead of 1.49804798673973\n",
      "-feature swap of score: 5.33997426279726 instead of 1.48377281057682\n",
      "-feature swap of score: 5.42090313439803 instead of 1.53651511977846\n",
      "-feature swap of score: 2.93580997720439 instead of 1.4588028225823\n",
      "-feature swap of score: 2.28270524102015 instead of 1.44457420528128\n",
      "-feature swap of score: 3.81157422004727 instead of 1.62346847737815\n",
      "-feature swap of score: 5.54701787577771 instead of 1.4646282098063\n",
      "-feature swap of score: 4.10280722740544 instead of 1.70740657298957\n",
      "-feature swap of score: 2.52849153508133 instead of 1.62755282396467\n",
      "-feature swap of score: 4.93042683261043 instead of 1.57956790241985\n",
      "-feature swap of score: 4.4909098849646 instead of 1.54536669803315\n",
      "-feature swap of score: 7.14916040206657 instead of 1.57809830201934\n",
      "-feature swap of score: 1.60342145364952 instead of 1.45793981560608\n",
      "-num feature swaps: 117\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scoring.+-.pl  --features_plus /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.enhanced.+.features  --features_minus /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.-.features  --atg_position 20  > /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores\n",
      "-round: 1\n",
      "-round: 2\n",
      "-round: 3\n",
      "-round: 4\n",
      "-round: 5\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scores_to_ROC.pl /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores > /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc\n",
      "-parsing scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/plot_ROC.Rscript /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc || :\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/compute_AUC.pl /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.enhanced.+.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/start_codon_refinement.pl --transcripts /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna --gff3_file /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 --workdir /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/ > /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3\n",
      "Refining start codon selections.\n",
      "-number of revised start positions: 2462\n",
      "* Running CMD: cp /home/ec2-user/glial-origins/output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3 GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3\n",
      "copying output to final output file: GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_bed.pl GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 > GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed\n",
      "Making bed file: GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 --fasta /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna  --genetic_code Universal > GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep\n",
      "Making pep file: GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 --fasta /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna --seqType CDS  --genetic_code Universal > GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds\n",
      "Making cds file: GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds\n",
      "transdecoder is finished.  See output files GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.*\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cdna = sample_BFD.genome_fasta.get_transdecoder_cdna_gtf(models_asgtf_updated, TRANSDECODER_LOC)\n",
    "sample_BFD.add_keyfile('cdna', cdna)\n",
    "\n",
    "transdecoder_files = sample_BFD.cdna.to_pep_files(TDLONGORF_LOC, TDPREDICT_LOC)\n",
    "sample_BFD.add_keyfiles(transdecoder_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "iEug6TQ6Cc9D"
   },
   "source": [
    "# 12. Push files to AWS S3\n",
    "\n",
    "Iteratively moves through BioFileDocket BioFile objects and pushes to the right place in AWS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_000002035.5_GRCz10_genomic.gff already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM3768152_Brain_8_dge.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_uniprot-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_gtf-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n"
     ]
    }
   ],
   "source": [
    "sample_BFD.local_to_s3()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 13. Pickle the `sample_BFD` variable for use by the next script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upload: ../../output/Drer_adultbrain/Drer_adultbrain_BioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/Drer_adultbrain_BioFileDocket.pkl\n"
     ]
    }
   ],
   "source": [
    "# Generate a .pkl file for the Docket\n",
    "sample_BFD.pickle()\n",
    "\n",
    "# Push to S3, optionally overwriting existing file\n",
    "sample_BFD.push_to_s3(overwrite = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "Xtro_bro.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
