{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "L56fjA27DmhA"
   },
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "t0BTCsz7Y48y",
    "outputId": "f6510678-f977-4ec8-ce30-1f2acd4fe666"
   },
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import os\n",
    "import dill\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "import sys\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Download and describe data\n",
    "\n",
    "\n",
    "This notebook collects data from the  .\n",
    "\n",
    "Dataset description\n",
    "Data can be found at the .\n",
    "This notebook only collects data from one sample, \"Brain1\", for the purposes of data analysis and exploration.\n",
    "\n",
    "This notebook collects data from the [Jiang et al. 2021](https://www.frontiersin.org/articles/10.3389/fcell.2021.743421/full) zebrafish cell atlas.\n",
    "\n",
    "## Dataset description\n",
    "\n",
    "- Data can be found at the [GEO Accession GSE130487](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE130487).\n",
    "- This notebook previously collects data from one sample, [\"Brain8\"](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3768152), for the purposes of data analysis and exploration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  410M  100  410M    0     0  6681k      0  0:01:02  0:01:02 --:--:-- 6608k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 18.1M  100 18.1M    0     0  2717k      0  0:00:06  0:00:06 --:--:-- 3454k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 7875k  100 7875k    0     0  1191k      0  0:00:06  0:00:06 --:--:-- 1800k\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'species': 'Danio_rerio',\n",
       " 'conditions': 'adultbrain',\n",
       " 'directory': '../../output/Drer_adultbrain/',\n",
       " 'files': {},\n",
       " 'annot': <biofile_handling.GenomeGffFile at 0x7fea27247490>,\n",
       " 'genome_fasta': <biofile_handling.GenomeFastaFile at 0x7fea27247d60>,\n",
       " 'gxc': <biofile_handling.GxcFile at 0x7fea27245c00>}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify the name of the species folder in Amazon S3\n",
    "species = 'Danio_rerio'\n",
    "\n",
    "# Specify any particular identifying conditions, eg tissue type:\n",
    "conditions = 'adultbrain'\n",
    "\n",
    "# Specify url and other variables\n",
    "genome_fasta_url = 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/035/GCF_000002035.5_GRCz10/GCF_000002035.5_GRCz10_genomic.fna.gz'\n",
    "genome_version = 'GRCz10'\n",
    "\n",
    "annot_url = 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/002/035/GCF_000002035.5_GRCz10/GCF_000002035.5_GRCz10_genomic.gff.gz'\n",
    "gxc_url = 'https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3768nnn/GSM3768152/suppl/GSM3768152_Brain_8_dge.txt.gz'\n",
    "\n",
    "###########\n",
    "# runtime #\n",
    "###########\n",
    "\n",
    "protocol = 'curl'\n",
    "\n",
    "species_prefix = prefixify(species)\n",
    "\n",
    "# Specify folder as destination for file downloads\n",
    "output_folder = '../../output/' + prefixify(species) + '_' + conditions + '/'\n",
    "\n",
    "if not os.path.exists(output_folder):\n",
    "    os.mkdir(output_folder)\n",
    "    \n",
    "species_SampleDict = SampleDict(species, conditions, output_folder)\n",
    "\n",
    "genome_fasta = GenomeFastaFile(\n",
    "    filename = '',\n",
    "    sampledict = species_SampleDict,\n",
    "    version = genome_version,\n",
    "    url = genome_fasta_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "annot = GenomeGffFile(\n",
    "    filename = '',\n",
    "    sampledict = species_SampleDict,\n",
    "    GenomeFastaFile = genome_fasta,\n",
    "    url = annot_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "gxc = GxcFile(\n",
    "    filename = '',\n",
    "    sampledict = species_SampleDict,\n",
    "    GenomeFastaFile = genome_fasta,\n",
    "    GenomeAnnotFile = annot,\n",
    "    url = gxc_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "\n",
    "sample_Docket = Docket(species_SampleDict)\n",
    "keyfiles = {\n",
    "    'annot': annot,\n",
    "    'genome_fasta': genome_fasta,\n",
    "    'gxc': gxc\n",
    "}\n",
    "sample_Docket.add_keyfiles(keyfiles)\n",
    "\n",
    "display(vars(sample_Docket))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Load in the gxc matrix and get gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>ACAATATATTGTACCTGA</th>\n",
       "      <th>ACGTTGATGGCGTAGAGA</th>\n",
       "      <th>AACCTAACCTGAATTTGC</th>\n",
       "      <th>CTCGCAGCCCTCTATGTA</th>\n",
       "      <th>ACGTTGCGTATTTAGTCG</th>\n",
       "      <th>AACCTATAGAGACCGACG</th>\n",
       "      <th>ACGAGCGCTGTGGCCTAG</th>\n",
       "      <th>GCGAATGGACATGGACAT</th>\n",
       "      <th>TCTACCGCTCAAGCTCAA</th>\n",
       "      <th>...</th>\n",
       "      <th>CGGCAGTCAAAGATCTCT</th>\n",
       "      <th>GACACTGCGAATCTGTGT</th>\n",
       "      <th>GCAGGAGGCTGCTAAGGG</th>\n",
       "      <th>TATGTATACTTCCGCACC</th>\n",
       "      <th>TGGATGTTCCGCACAATA</th>\n",
       "      <th>AACCTATGGATGGGGTTT</th>\n",
       "      <th>AAGCGGAGGACTCTCCAT</th>\n",
       "      <th>ACCTGACTCGCAAGCGAG</th>\n",
       "      <th>ACGTTGCAAAGTTTCATA</th>\n",
       "      <th>ATCAACTGCAATTTCCGC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCF3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACOT12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACSF3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACTC1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACVR1C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20802</th>\n",
       "      <td>zwilch</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20803</th>\n",
       "      <td>zyg11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20804</th>\n",
       "      <td>zyx</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20805</th>\n",
       "      <td>zzef1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20806</th>\n",
       "      <td>zzz3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20807 rows × 10001 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GENE  ACAATATATTGTACCTGA  ACGTTGATGGCGTAGAGA  AACCTAACCTGAATTTGC  \\\n",
       "0       ABCF3                   0                   0                   0   \n",
       "1      ACOT12                   0                   0                   0   \n",
       "2       ACSF3                   0                   0                   0   \n",
       "3       ACTC1                   0                   0                   0   \n",
       "4      ACVR1C                   0                   0                   0   \n",
       "...       ...                 ...                 ...                 ...   \n",
       "20802  zwilch                   0                   0                   0   \n",
       "20803   zyg11                   0                   0                   0   \n",
       "20804     zyx                   0                   0                   0   \n",
       "20805   zzef1                   0                   0                   0   \n",
       "20806    zzz3                   0                   0                   0   \n",
       "\n",
       "       CTCGCAGCCCTCTATGTA  ACGTTGCGTATTTAGTCG  AACCTATAGAGACCGACG  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "20802                   0                   0                   0   \n",
       "20803                   0                   0                   0   \n",
       "20804                   0                   0                   0   \n",
       "20805                   0                   0                   0   \n",
       "20806                   0                   0                   0   \n",
       "\n",
       "       ACGAGCGCTGTGGCCTAG  GCGAATGGACATGGACAT  TCTACCGCTCAAGCTCAA  ...  \\\n",
       "0                       0                   0                   0  ...   \n",
       "1                       0                   0                   0  ...   \n",
       "2                       0                   0                   0  ...   \n",
       "3                       0                   0                   0  ...   \n",
       "4                       0                   0                   0  ...   \n",
       "...                   ...                 ...                 ...  ...   \n",
       "20802                   0                   0                   0  ...   \n",
       "20803                   0                   1                   0  ...   \n",
       "20804                   0                   0                   0  ...   \n",
       "20805                   0                   0                   0  ...   \n",
       "20806                   0                   0                   0  ...   \n",
       "\n",
       "       CGGCAGTCAAAGATCTCT  GACACTGCGAATCTGTGT  GCAGGAGGCTGCTAAGGG  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "20802                   0                   0                   0   \n",
       "20803                   0                   1                   1   \n",
       "20804                   0                   0                   0   \n",
       "20805                   0                   0                   0   \n",
       "20806                   0                   0                   0   \n",
       "\n",
       "       TATGTATACTTCCGCACC  TGGATGTTCCGCACAATA  AACCTATGGATGGGGTTT  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "20802                   0                   0                   0   \n",
       "20803                   0                   0                   0   \n",
       "20804                   0                   0                   0   \n",
       "20805                   0                   0                   0   \n",
       "20806                   0                   0                   0   \n",
       "\n",
       "       AAGCGGAGGACTCTCCAT  ACCTGACTCGCAAGCGAG  ACGTTGCAAAGTTTCATA  \\\n",
       "0                       0                   0                   0   \n",
       "1                       0                   0                   0   \n",
       "2                       0                   0                   0   \n",
       "3                       0                   0                   0   \n",
       "4                       0                   0                   0   \n",
       "...                   ...                 ...                 ...   \n",
       "20802                   0                   0                   0   \n",
       "20803                   0                   0                   0   \n",
       "20804                   0                   0                   0   \n",
       "20805                   0                   0                   0   \n",
       "20806                   0                   0                   0   \n",
       "\n",
       "       ATCAACTGCAATTTCCGC  \n",
       "0                       0  \n",
       "1                       0  \n",
       "2                       0  \n",
       "3                       0  \n",
       "4                       0  \n",
       "...                   ...  \n",
       "20802                   0  \n",
       "20803                   0  \n",
       "20804                   0  \n",
       "20805                   0  \n",
       "20806                   0  \n",
       "\n",
       "[20807 rows x 10001 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACOT12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACSF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACTC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACVR1C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20802</th>\n",
       "      <td>zwilch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20803</th>\n",
       "      <td>zyg11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20804</th>\n",
       "      <td>zyx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20805</th>\n",
       "      <td>zzef1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20806</th>\n",
       "      <td>zzz3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20807 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name\n",
       "0         ABCF3\n",
       "1        ACOT12\n",
       "2         ACSF3\n",
       "3         ACTC1\n",
       "4        ACVR1C\n",
       "...         ...\n",
       "20802    zwilch\n",
       "20803     zyg11\n",
       "20804       zyx\n",
       "20805     zzef1\n",
       "20806      zzz3\n",
       "\n",
       "[20807 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes_matrix = pd.read_csv(sample_Docket.gxc.path, sep = '\\t')\n",
    "display(genes_matrix)\n",
    "\n",
    "gxc_genes_list = pd.DataFrame({'gene_name':genes_matrix['GENE']})\n",
    "display(gxc_genes_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jvfcFNT2FuaZ"
   },
   "source": [
    "# 3. Get mapping identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1.0</td>\n",
       "      <td>58871917.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id0;Dbxref=taxon:7955;Name=1;chromosome=1;g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>6642.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Dbxref=GeneID:192301,ZFIN:ZDB-GENE-02...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>6642.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Dbxref=GeneID:192301,Genb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11751.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Dbxref=GeneID:192301,Genban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11550.0</td>\n",
       "      <td>11625.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Dbxref=GeneID:192301,Genban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362356</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16449.0</td>\n",
       "      <td>16520.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id634507;Parent=rna65516;Dbxref=GeneID:1405...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362357</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene42297;Dbxref=GeneID:140511,ZFIN:ZDB-GEN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362358</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna65517;Parent=gene42297;Dbxref=GeneID:140...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362359</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id634508;Parent=rna65517;Dbxref=GeneID:1405...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362360</th>\n",
       "      <td>###</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1362361 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0           1       2        3           4    5    6    7  \\\n",
       "0        NC_007112.6      RefSeq  region      1.0  58871917.0    .    +    .   \n",
       "1        NC_007112.6  BestRefSeq    gene   6642.0     11878.0    .    -    .   \n",
       "2        NC_007112.6  BestRefSeq    mRNA   6642.0     11878.0    .    -    .   \n",
       "3        NC_007112.6  BestRefSeq    exon  11751.0     11878.0    .    -    .   \n",
       "4        NC_007112.6  BestRefSeq    exon  11550.0     11625.0    .    -    .   \n",
       "...              ...         ...     ...      ...         ...  ...  ...  ...   \n",
       "1362356  NC_002333.2      RefSeq    exon  16449.0     16520.0    .    +    .   \n",
       "1362357  NC_002333.2      RefSeq    gene  16527.0     16596.0    .    -    .   \n",
       "1362358  NC_002333.2      RefSeq    tRNA  16527.0     16596.0    .    -    .   \n",
       "1362359  NC_002333.2      RefSeq    exon  16527.0     16596.0    .    -    .   \n",
       "1362360          ###         NaN     NaN      NaN         NaN  NaN  NaN  NaN   \n",
       "\n",
       "                                                         8  \n",
       "0        ID=id0;Dbxref=taxon:7955;Name=1;chromosome=1;g...  \n",
       "1        ID=gene0;Dbxref=GeneID:192301,ZFIN:ZDB-GENE-02...  \n",
       "2        ID=rna0;Parent=gene0;Dbxref=GeneID:192301,Genb...  \n",
       "3        ID=id1;Parent=rna0;Dbxref=GeneID:192301,Genban...  \n",
       "4        ID=id2;Parent=rna0;Dbxref=GeneID:192301,Genban...  \n",
       "...                                                    ...  \n",
       "1362356  ID=id634507;Parent=rna65516;Dbxref=GeneID:1405...  \n",
       "1362357  ID=gene42297;Dbxref=GeneID:140511,ZFIN:ZDB-GEN...  \n",
       "1362358  ID=rna65517;Parent=gene42297;Dbxref=GeneID:140...  \n",
       "1362359  ID=id634508;Parent=rna65517;Dbxref=GeneID:1405...  \n",
       "1362360                                                NaN  \n",
       "\n",
       "[1362361 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'ID=id1;Parent=rna0;Dbxref=GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GENE-020419-25;gbkey=mRNA;gene=rpl24;product=ribosomal protein L24;transcript_id=NM_173235.3'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load in the original GFF-based annotation\n",
    "models = pd.read_csv(sample_Docket.annot.path, skiprows = 7, header = None, sep = '\\t')\n",
    "display(models)\n",
    "\n",
    "attributes_column = 8\n",
    "\n",
    "# Check the structure of fields in the GFF additional fields section\n",
    "display(models[attributes_column][3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Dbxref</th>\n",
       "      <th>dbxref_dict</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1.0</td>\n",
       "      <td>58871917.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id0;Dbxref=taxon:7955;Name=1;chromosome=1;g...</td>\n",
       "      <td>{'ID': 'id0', 'Dbxref': 'taxon:7955', 'Name': ...</td>\n",
       "      <td>None</td>\n",
       "      <td>taxon:7955</td>\n",
       "      <td>{'taxon': '7955'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>6642.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Dbxref=GeneID:192301,ZFIN:ZDB-GENE-02...</td>\n",
       "      <td>{'ID': 'gene0', 'Dbxref': 'GeneID:192301,ZFIN:...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>GeneID:192301,ZFIN:ZDB-GENE-020419-25</td>\n",
       "      <td>{'GeneID': '192301', 'ZFIN': 'ZDB-GENE-020419-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>6642.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Dbxref=GeneID:192301,Genb...</td>\n",
       "      <td>{'ID': 'rna0', 'Parent': 'gene0', 'Dbxref': 'G...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...</td>\n",
       "      <td>{'GeneID': '192301', 'Genbank': 'NM_173235.3',...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11751.0</td>\n",
       "      <td>11878.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Dbxref=GeneID:192301,Genban...</td>\n",
       "      <td>{'ID': 'id1', 'Parent': 'rna0', 'Dbxref': 'Gen...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...</td>\n",
       "      <td>{'GeneID': '192301', 'Genbank': 'NM_173235.3',...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>11550.0</td>\n",
       "      <td>11625.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Dbxref=GeneID:192301,Genban...</td>\n",
       "      <td>{'ID': 'id2', 'Parent': 'rna0', 'Dbxref': 'Gen...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...</td>\n",
       "      <td>{'GeneID': '192301', 'Genbank': 'NM_173235.3',...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362355</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>16449.0</td>\n",
       "      <td>16520.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna65516;Parent=gene42296;Dbxref=GeneID:140...</td>\n",
       "      <td>{'ID': 'rna65516', 'Parent': 'gene42296', 'Dbx...</td>\n",
       "      <td>trnT</td>\n",
       "      <td>GeneID:140518,ZFIN:ZDB-GENE-011205-39</td>\n",
       "      <td>{'GeneID': '140518', 'ZFIN': 'ZDB-GENE-011205-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362356</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16449.0</td>\n",
       "      <td>16520.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id634507;Parent=rna65516;Dbxref=GeneID:1405...</td>\n",
       "      <td>{'ID': 'id634507', 'Parent': 'rna65516', 'Dbxr...</td>\n",
       "      <td>trnT</td>\n",
       "      <td>GeneID:140518,ZFIN:ZDB-GENE-011205-39</td>\n",
       "      <td>{'GeneID': '140518', 'ZFIN': 'ZDB-GENE-011205-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362357</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>gene</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene42297;Dbxref=GeneID:140511,ZFIN:ZDB-GEN...</td>\n",
       "      <td>{'ID': 'gene42297', 'Dbxref': 'GeneID:140511,Z...</td>\n",
       "      <td>trnP</td>\n",
       "      <td>GeneID:140511,ZFIN:ZDB-GENE-011205-38</td>\n",
       "      <td>{'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362358</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna65517;Parent=gene42297;Dbxref=GeneID:140...</td>\n",
       "      <td>{'ID': 'rna65517', 'Parent': 'gene42297', 'Dbx...</td>\n",
       "      <td>trnP</td>\n",
       "      <td>GeneID:140511,ZFIN:ZDB-GENE-011205-38</td>\n",
       "      <td>{'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362359</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16527.0</td>\n",
       "      <td>16596.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id634508;Parent=rna65517;Dbxref=GeneID:1405...</td>\n",
       "      <td>{'ID': 'id634508', 'Parent': 'rna65517', 'Dbxr...</td>\n",
       "      <td>trnP</td>\n",
       "      <td>GeneID:140511,ZFIN:ZDB-GENE-011205-38</td>\n",
       "      <td>{'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1360240 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0           1       2        3           4  5  6  7  \\\n",
       "0        NC_007112.6      RefSeq  region      1.0  58871917.0  .  +  .   \n",
       "1        NC_007112.6  BestRefSeq    gene   6642.0     11878.0  .  -  .   \n",
       "2        NC_007112.6  BestRefSeq    mRNA   6642.0     11878.0  .  -  .   \n",
       "3        NC_007112.6  BestRefSeq    exon  11751.0     11878.0  .  -  .   \n",
       "4        NC_007112.6  BestRefSeq    exon  11550.0     11625.0  .  -  .   \n",
       "...              ...         ...     ...      ...         ... .. .. ..   \n",
       "1362355  NC_002333.2      RefSeq    tRNA  16449.0     16520.0  .  +  .   \n",
       "1362356  NC_002333.2      RefSeq    exon  16449.0     16520.0  .  +  .   \n",
       "1362357  NC_002333.2      RefSeq    gene  16527.0     16596.0  .  -  .   \n",
       "1362358  NC_002333.2      RefSeq    tRNA  16527.0     16596.0  .  -  .   \n",
       "1362359  NC_002333.2      RefSeq    exon  16527.0     16596.0  .  -  .   \n",
       "\n",
       "                                                         8  \\\n",
       "0        ID=id0;Dbxref=taxon:7955;Name=1;chromosome=1;g...   \n",
       "1        ID=gene0;Dbxref=GeneID:192301,ZFIN:ZDB-GENE-02...   \n",
       "2        ID=rna0;Parent=gene0;Dbxref=GeneID:192301,Genb...   \n",
       "3        ID=id1;Parent=rna0;Dbxref=GeneID:192301,Genban...   \n",
       "4        ID=id2;Parent=rna0;Dbxref=GeneID:192301,Genban...   \n",
       "...                                                    ...   \n",
       "1362355  ID=rna65516;Parent=gene42296;Dbxref=GeneID:140...   \n",
       "1362356  ID=id634507;Parent=rna65516;Dbxref=GeneID:1405...   \n",
       "1362357  ID=gene42297;Dbxref=GeneID:140511,ZFIN:ZDB-GEN...   \n",
       "1362358  ID=rna65517;Parent=gene42297;Dbxref=GeneID:140...   \n",
       "1362359  ID=id634508;Parent=rna65517;Dbxref=GeneID:1405...   \n",
       "\n",
       "                                          field_dictionary gene_name  \\\n",
       "0        {'ID': 'id0', 'Dbxref': 'taxon:7955', 'Name': ...      None   \n",
       "1        {'ID': 'gene0', 'Dbxref': 'GeneID:192301,ZFIN:...     rpl24   \n",
       "2        {'ID': 'rna0', 'Parent': 'gene0', 'Dbxref': 'G...     rpl24   \n",
       "3        {'ID': 'id1', 'Parent': 'rna0', 'Dbxref': 'Gen...     rpl24   \n",
       "4        {'ID': 'id2', 'Parent': 'rna0', 'Dbxref': 'Gen...     rpl24   \n",
       "...                                                    ...       ...   \n",
       "1362355  {'ID': 'rna65516', 'Parent': 'gene42296', 'Dbx...      trnT   \n",
       "1362356  {'ID': 'id634507', 'Parent': 'rna65516', 'Dbxr...      trnT   \n",
       "1362357  {'ID': 'gene42297', 'Dbxref': 'GeneID:140511,Z...      trnP   \n",
       "1362358  {'ID': 'rna65517', 'Parent': 'gene42297', 'Dbx...      trnP   \n",
       "1362359  {'ID': 'id634508', 'Parent': 'rna65517', 'Dbxr...      trnP   \n",
       "\n",
       "                                                    Dbxref  \\\n",
       "0                                               taxon:7955   \n",
       "1                    GeneID:192301,ZFIN:ZDB-GENE-020419-25   \n",
       "2        GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...   \n",
       "3        GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...   \n",
       "4        GeneID:192301,Genbank:NM_173235.3,ZFIN:ZDB-GEN...   \n",
       "...                                                    ...   \n",
       "1362355              GeneID:140518,ZFIN:ZDB-GENE-011205-39   \n",
       "1362356              GeneID:140518,ZFIN:ZDB-GENE-011205-39   \n",
       "1362357              GeneID:140511,ZFIN:ZDB-GENE-011205-38   \n",
       "1362358              GeneID:140511,ZFIN:ZDB-GENE-011205-38   \n",
       "1362359              GeneID:140511,ZFIN:ZDB-GENE-011205-38   \n",
       "\n",
       "                                               dbxref_dict  \n",
       "0                                        {'taxon': '7955'}  \n",
       "1        {'GeneID': '192301', 'ZFIN': 'ZDB-GENE-020419-...  \n",
       "2        {'GeneID': '192301', 'Genbank': 'NM_173235.3',...  \n",
       "3        {'GeneID': '192301', 'Genbank': 'NM_173235.3',...  \n",
       "4        {'GeneID': '192301', 'Genbank': 'NM_173235.3',...  \n",
       "...                                                    ...  \n",
       "1362355  {'GeneID': '140518', 'ZFIN': 'ZDB-GENE-011205-...  \n",
       "1362356  {'GeneID': '140518', 'ZFIN': 'ZDB-GENE-011205-...  \n",
       "1362357  {'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...  \n",
       "1362358  {'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...  \n",
       "1362359  {'GeneID': '140511', 'ZFIN': 'ZDB-GENE-011205-...  \n",
       "\n",
       "[1360240 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Remove any rows with NaNs\n",
    "models.dropna(inplace = True)\n",
    "\n",
    "# Extract field and database cross-ref (dbxref) information into columns\n",
    "models['field_dictionary'] = models[attributes_column].apply(convert_fields_to_dict_gff)\n",
    "models['gene_name'] = [d.get('gene') for d in models['field_dictionary']]\n",
    "models['Dbxref'] = [d.get('Dbxref') for d in models['field_dictionary']]\n",
    "models['dbxref_dict'] = models['Dbxref'].apply(convert_dbxref_to_dict)\n",
    "\n",
    "display(models)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Extract gene IDs for mapping to UniprotKB\n",
    "Specify which set of identifiers will be use to query the [Uniprot ID Mapping Tool](https://www.uniprot.org/id-mapping) via API.\n",
    "\n",
    "If using an identifier from the `dbxref_dict`, specify the name via string in the `dbxref_datafield` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ZFIN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rpl24</td>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>cep97</td>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>nfkbiz</td>\n",
       "      <td>ZDB-GENE-071024-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>eed</td>\n",
       "      <td>ZDB-GENE-050417-287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>zgc:110091</td>\n",
       "      <td>ZDB-GENE-050417-34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362347</th>\n",
       "      <td>ND6</td>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362349</th>\n",
       "      <td>trnE</td>\n",
       "      <td>ZDB-GENE-011205-37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362352</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362354</th>\n",
       "      <td>trnT</td>\n",
       "      <td>ZDB-GENE-011205-39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1362357</th>\n",
       "      <td>trnP</td>\n",
       "      <td>ZDB-GENE-011205-38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22031 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_name                 ZFIN\n",
       "1             rpl24   ZDB-GENE-020419-25\n",
       "15            cep97   ZDB-GENE-031030-11\n",
       "74           nfkbiz    ZDB-GENE-071024-1\n",
       "102             eed  ZDB-GENE-050417-287\n",
       "128      zgc:110091   ZDB-GENE-050417-34\n",
       "...             ...                  ...\n",
       "1362347         ND6   ZDB-GENE-011205-13\n",
       "1362349        trnE   ZDB-GENE-011205-37\n",
       "1362352        CYTB   ZDB-GENE-011205-17\n",
       "1362354        trnT   ZDB-GENE-011205-39\n",
       "1362357        trnP   ZDB-GENE-011205-38\n",
       "\n",
       "[22031 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dbxref_datafield = 'ZFIN'\n",
    "\n",
    "\n",
    "models.dropna(axis = 0, subset = ['dbxref_dict'], inplace = True)\n",
    "models[dbxref_datafield] = [d.get(dbxref_datafield) for d in models['dbxref_dict']]\n",
    "\n",
    "models_subset = models[['gene_name', dbxref_datafield]].dropna().drop_duplicates()\n",
    "\n",
    "\n",
    "display(models_subset)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Generate gene list file to query Uniprot ID Mapping API\n",
    "Generate a text file ending in `_ids.txt` for sending to the ID mapping API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 22031 gene ids to ../../output/Drer_adultbrain/Drer_adultbrain_ZFIN_ids.txt\n"
     ]
    }
   ],
   "source": [
    "gene_list = models_subset[dbxref_datafield].unique()\n",
    "\n",
    "genelist_object = GeneListFile(\n",
    "    filename = '', # filename is automatically generated from sampledict\n",
    "    sampledict = species_SampleDict,\n",
    "    sources = [sample_Docket.annot],\n",
    "    genes = gene_list,\n",
    "    identifier = dbxref_datafield\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Query Uniprot ID Mapping API\n",
    "Specify the `from_type` variable based on the Uniprot name of the identifier.\n",
    "\n",
    "The table below lists some databases and the `from_type` string that the API accepts for that datatype.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "| datatype | `from_type` string | description |\n",
    "| ---: | :--- | :--- |\n",
    "| Mouse Genome Informatics | `MGI` | ID starts with `MGI:` |\n",
    "| Zebrafish Information Network | `ZFIN` | ID starts with `ZDB-GENE-` |"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  419k    0    52  100  418k     19   157k  0:00:02  0:00:02 --:--:--  157k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100    24    0    24    0     0     37      0 --:--:-- --:--:-- --:--:--    37\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>From</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>Q8JGR4</td>\n",
       "      <td>RL24_DANRE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>60S ribosomal protein L24</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>A0A0R4IMS3</td>\n",
       "      <td>A0A0R4IMS3_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>60S ribosomal protein L24</td>\n",
       "      <td>rpl24 SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ICF0</td>\n",
       "      <td>A0A0R4ICF0_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Centrosomal protein 97</td>\n",
       "      <td>cep97 SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ISF5</td>\n",
       "      <td>A0A0R4ISF5_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Centrosomal protein 97</td>\n",
       "      <td>cep97 SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4IX25</td>\n",
       "      <td>A0A0R4IX25_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Centrosomal protein 97</td>\n",
       "      <td>cep97 SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54705</th>\n",
       "      <td>ZDB-GENE-011205-12</td>\n",
       "      <td>Q9MIY0</td>\n",
       "      <td>NU5M_DANRE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NADH-ubiquinone oxidoreductase chain 5 (EC 7.1...</td>\n",
       "      <td>mt-nd5 mtnd5 nd5</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54706</th>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>Q9MIX9</td>\n",
       "      <td>NU6M_DANRE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>NADH-ubiquinone oxidoreductase chain 6 (EC 7.1...</td>\n",
       "      <td>mt-nd6 mtnd6 nd6</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54707</th>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>A0A0A0VG13</td>\n",
       "      <td>A0A0A0VG13_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>NADH-ubiquinone oxidoreductase chain 6 (EC 7.1...</td>\n",
       "      <td>ND6 mt-nd6 SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54708</th>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>Q9MIX8</td>\n",
       "      <td>CYB_DANRE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Cytochrome b (Complex III subunit 3) (Complex ...</td>\n",
       "      <td>mt-cyb cob cytb mtcyb</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54709</th>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>A0A0A0VDB5</td>\n",
       "      <td>A0A0A0VDB5_DANRE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Cytochrome b</td>\n",
       "      <td>CYTB mt-cyb SO:0001217</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "      <td>380</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54710 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     From       Entry        Entry Name    Reviewed  \\\n",
       "0      ZDB-GENE-020419-25      Q8JGR4        RL24_DANRE    reviewed   \n",
       "1      ZDB-GENE-020419-25  A0A0R4IMS3  A0A0R4IMS3_DANRE  unreviewed   \n",
       "2      ZDB-GENE-031030-11  A0A0R4ICF0  A0A0R4ICF0_DANRE  unreviewed   \n",
       "3      ZDB-GENE-031030-11  A0A0R4ISF5  A0A0R4ISF5_DANRE  unreviewed   \n",
       "4      ZDB-GENE-031030-11  A0A0R4IX25  A0A0R4IX25_DANRE  unreviewed   \n",
       "...                   ...         ...               ...         ...   \n",
       "54705  ZDB-GENE-011205-12      Q9MIY0        NU5M_DANRE    reviewed   \n",
       "54706  ZDB-GENE-011205-13      Q9MIX9        NU6M_DANRE    reviewed   \n",
       "54707  ZDB-GENE-011205-13  A0A0A0VG13  A0A0A0VG13_DANRE  unreviewed   \n",
       "54708  ZDB-GENE-011205-17      Q9MIX8         CYB_DANRE    reviewed   \n",
       "54709  ZDB-GENE-011205-17  A0A0A0VDB5  A0A0A0VDB5_DANRE  unreviewed   \n",
       "\n",
       "                                           Protein names  \\\n",
       "0                              60S ribosomal protein L24   \n",
       "1                              60S ribosomal protein L24   \n",
       "2                                 Centrosomal protein 97   \n",
       "3                                 Centrosomal protein 97   \n",
       "4                                 Centrosomal protein 97   \n",
       "...                                                  ...   \n",
       "54705  NADH-ubiquinone oxidoreductase chain 5 (EC 7.1...   \n",
       "54706  NADH-ubiquinone oxidoreductase chain 6 (EC 7.1...   \n",
       "54707  NADH-ubiquinone oxidoreductase chain 6 (EC 7.1...   \n",
       "54708  Cytochrome b (Complex III subunit 3) (Complex ...   \n",
       "54709                                       Cytochrome b   \n",
       "\n",
       "                   Gene Names                                     Organism  \\\n",
       "0                       rpl24  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "1            rpl24 SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "2            cep97 SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "3            cep97 SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "4            cep97 SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "...                       ...                                          ...   \n",
       "54705        mt-nd5 mtnd5 nd5  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "54706        mt-nd6 mtnd6 nd6  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "54707   ND6 mt-nd6 SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "54708   mt-cyb cob cytb mtcyb  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "54709  CYTB mt-cyb SO:0001217  Danio rerio (Zebrafish) (Brachydanio rerio)   \n",
       "\n",
       "       Length  \n",
       "0         157  \n",
       "1         157  \n",
       "2         599  \n",
       "3          40  \n",
       "4         240  \n",
       "...       ...  \n",
       "54705     606  \n",
       "54706     172  \n",
       "54707     172  \n",
       "54708     380  \n",
       "54709     380  \n",
       "\n",
       "[54710 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from_type = 'ZFIN'\n",
    "to_type = 'UniProtKB'\n",
    "\n",
    "uniprot_idmm_object = genelist_object.get_uniprot_ids(ID_MAPPER_LOC, from_type, to_type)\n",
    "sample_Docket.add_keyfiles({'uniprot_idmm': uniprot_idmm_object})\n",
    "\n",
    "uniprot_idmm = pd.read_csv(sample_Docket.uniprot_idmm.path, sep = '\\t')\n",
    "display(uniprot_idmm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Extract results and generate Uniprot IDMM\n",
    "Generates an idmm that links `gene_name`, the `dbxref_datafield` seleted above, and `uniprot_id` returned by API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_93348/3311337425.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  uniprot_idpairs.rename(columns = {'From': dbxref_datafield, 'Entry': 'uniprot_id'}, inplace = True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ZFIN</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>Q8JGR4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>A0A0R4IMS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ICF0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ISF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4IX25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54705</th>\n",
       "      <td>ZDB-GENE-011205-12</td>\n",
       "      <td>Q9MIY0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54706</th>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>Q9MIX9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54707</th>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>A0A0A0VG13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54708</th>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>Q9MIX8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54709</th>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>A0A0A0VDB5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54710 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     ZFIN  uniprot_id\n",
       "0      ZDB-GENE-020419-25      Q8JGR4\n",
       "1      ZDB-GENE-020419-25  A0A0R4IMS3\n",
       "2      ZDB-GENE-031030-11  A0A0R4ICF0\n",
       "3      ZDB-GENE-031030-11  A0A0R4ISF5\n",
       "4      ZDB-GENE-031030-11  A0A0R4IX25\n",
       "...                   ...         ...\n",
       "54705  ZDB-GENE-011205-12      Q9MIY0\n",
       "54706  ZDB-GENE-011205-13      Q9MIX9\n",
       "54707  ZDB-GENE-011205-13  A0A0A0VG13\n",
       "54708  ZDB-GENE-011205-17      Q9MIX8\n",
       "54709  ZDB-GENE-011205-17  A0A0A0VDB5\n",
       "\n",
       "[54710 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ZFIN</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rpl24</td>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>Q8JGR4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rpl24</td>\n",
       "      <td>ZDB-GENE-020419-25</td>\n",
       "      <td>A0A0R4IMS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cep97</td>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ICF0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cep97</td>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4ISF5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cep97</td>\n",
       "      <td>ZDB-GENE-031030-11</td>\n",
       "      <td>A0A0R4IX25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54705</th>\n",
       "      <td>ND5</td>\n",
       "      <td>ZDB-GENE-011205-12</td>\n",
       "      <td>Q9MIY0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54706</th>\n",
       "      <td>ND6</td>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>Q9MIX9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54707</th>\n",
       "      <td>ND6</td>\n",
       "      <td>ZDB-GENE-011205-13</td>\n",
       "      <td>A0A0A0VG13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54708</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>Q9MIX8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54709</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>ZDB-GENE-011205-17</td>\n",
       "      <td>A0A0A0VDB5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54710 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name                ZFIN  uniprot_id\n",
       "0         rpl24  ZDB-GENE-020419-25      Q8JGR4\n",
       "1         rpl24  ZDB-GENE-020419-25  A0A0R4IMS3\n",
       "2         cep97  ZDB-GENE-031030-11  A0A0R4ICF0\n",
       "3         cep97  ZDB-GENE-031030-11  A0A0R4ISF5\n",
       "4         cep97  ZDB-GENE-031030-11  A0A0R4IX25\n",
       "...         ...                 ...         ...\n",
       "54705       ND5  ZDB-GENE-011205-12      Q9MIY0\n",
       "54706       ND6  ZDB-GENE-011205-13      Q9MIX9\n",
       "54707       ND6  ZDB-GENE-011205-13  A0A0A0VG13\n",
       "54708      CYTB  ZDB-GENE-011205-17      Q9MIX8\n",
       "54709      CYTB  ZDB-GENE-011205-17  A0A0A0VDB5\n",
       "\n",
       "[54710 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "uniprot_idpairs = uniprot_idmm[['From', 'Entry']]\n",
    "uniprot_idpairs.rename(columns = {'From': dbxref_datafield, 'Entry': 'uniprot_id'}, inplace = True)\n",
    "display(uniprot_idpairs)\n",
    "\n",
    "uniprot_output_idmm = models_subset.merge(uniprot_idpairs, on = dbxref_datafield)\n",
    "display(uniprot_output_idmm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate a filename and file for the idmm\n",
    "uniprot_output_idmm_filename = '_'.join([species_prefix, conditions, 'uniprot-idmm.tsv'])\n",
    "uniprot_output_idmm_object = IdmmFile(uniprot_output_idmm_filename, species_SampleDict, kind = 'uniprot_idmm', sources = [sample_Docket.annot])\n",
    "\n",
    "# save to file and add to the Docket\n",
    "uniprot_output_idmm.to_csv(uniprot_output_idmm_object.path, sep = '\\t')\n",
    "sample_Docket.add_keyfile(uniprot_output_idmm_object, 'uniprot_idmm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8. Convert GFF to GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted file GCF_000002035.5_GRCz10_genomic.gtf already exists at:\n",
      " ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.gtf\n"
     ]
    }
   ],
   "source": [
    "# convert the GFF file to GTF using gffread\n",
    "models_asgtf = sample_Docket.annot.to_gtf(GFFREAD_LOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>6642</td>\n",
       "      <td>11878</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>6642</td>\n",
       "      <td>6760</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>6892</td>\n",
       "      <td>6955</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>9558</td>\n",
       "      <td>9694</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>10081</td>\n",
       "      <td>10191</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258174</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>15308</td>\n",
       "      <td>16448</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"gene42295\"; gene_name \"CYTB\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258175</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>16449</td>\n",
       "      <td>16520</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65516\"; gene_id \"gene42296\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258176</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16449</td>\n",
       "      <td>16520</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65516\"; gene_id \"gene42296\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258177</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>16527</td>\n",
       "      <td>16596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65517\"; gene_id \"gene42297\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258178</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16527</td>\n",
       "      <td>16596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65517\"; gene_id \"gene42297\";...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1258179 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0           1           2      3      4  5  6  7  \\\n",
       "0        NC_007112.6  BestRefSeq  transcript   6642  11878  .  -  .   \n",
       "1        NC_007112.6  BestRefSeq        exon   6642   6760  .  -  .   \n",
       "2        NC_007112.6  BestRefSeq        exon   6892   6955  .  -  .   \n",
       "3        NC_007112.6  BestRefSeq        exon   9558   9694  .  -  .   \n",
       "4        NC_007112.6  BestRefSeq        exon  10081  10191  .  -  .   \n",
       "...              ...         ...         ...    ...    ... .. .. ..   \n",
       "1258174  NC_002333.2      RefSeq         CDS  15308  16448  .  +  0   \n",
       "1258175  NC_002333.2      RefSeq  transcript  16449  16520  .  +  .   \n",
       "1258176  NC_002333.2      RefSeq        exon  16449  16520  .  +  .   \n",
       "1258177  NC_002333.2      RefSeq  transcript  16527  16596  .  -  .   \n",
       "1258178  NC_002333.2      RefSeq        exon  16527  16596  .  -  .   \n",
       "\n",
       "                                                         8  \n",
       "0        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "1        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "2        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "3        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "4        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "...                                                    ...  \n",
       "1258174       transcript_id \"gene42295\"; gene_name \"CYTB\";  \n",
       "1258175  transcript_id \"rna65516\"; gene_id \"gene42296\";...  \n",
       "1258176  transcript_id \"rna65516\"; gene_id \"gene42296\";...  \n",
       "1258177  transcript_id \"rna65517\"; gene_id \"gene42297\";...  \n",
       "1258178  transcript_id \"rna65517\"; gene_id \"gene42297\";...  \n",
       "\n",
       "[1258179 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'transcript_id \"rna0\"; gene_id \"gene0\"; gene_name \"rpl24\";'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load the newly-generated GTF file as a dataframe\n",
    "models_asgtf_df = pd.read_csv(models_asgtf.path, skiprows = 0, header = None, sep = '\\t')\n",
    "\n",
    "display(models_asgtf_df)\n",
    "display(models_asgtf_df[attributes_column][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>6642</td>\n",
       "      <td>11878</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>6642</td>\n",
       "      <td>6760</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>6892</td>\n",
       "      <td>6955</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>9558</td>\n",
       "      <td>9694</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_007112.6</td>\n",
       "      <td>BestRefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>10081</td>\n",
       "      <td>10191</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258173</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15308</td>\n",
       "      <td>16448</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"gene42295\"; gene_id \"gene42295\"...</td>\n",
       "      <td>{'transcript_id': 'gene42295', 'gene_id': 'gen...</td>\n",
       "      <td>CYTB</td>\n",
       "      <td>gene42295</td>\n",
       "      <td>gene42295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258175</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>16449</td>\n",
       "      <td>16520</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65516\"; gene_id \"gene42296\";...</td>\n",
       "      <td>{'transcript_id': 'rna65516', 'gene_id': 'gene...</td>\n",
       "      <td>trnT</td>\n",
       "      <td>gene42296</td>\n",
       "      <td>rna65516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258176</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16449</td>\n",
       "      <td>16520</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65516\"; gene_id \"gene42296\";...</td>\n",
       "      <td>{'transcript_id': 'rna65516', 'gene_id': 'gene...</td>\n",
       "      <td>trnT</td>\n",
       "      <td>gene42296</td>\n",
       "      <td>rna65516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258177</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>16527</td>\n",
       "      <td>16596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65517\"; gene_id \"gene42297\";...</td>\n",
       "      <td>{'transcript_id': 'rna65517', 'gene_id': 'gene...</td>\n",
       "      <td>trnP</td>\n",
       "      <td>gene42297</td>\n",
       "      <td>rna65517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258178</th>\n",
       "      <td>NC_002333.2</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>16527</td>\n",
       "      <td>16596</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna65517\"; gene_id \"gene42297\";...</td>\n",
       "      <td>{'transcript_id': 'rna65517', 'gene_id': 'gene...</td>\n",
       "      <td>trnP</td>\n",
       "      <td>gene42297</td>\n",
       "      <td>rna65517</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>699236 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0           1           2      3      4  5  6  7  \\\n",
       "0        NC_007112.6  BestRefSeq  transcript   6642  11878  .  -  .   \n",
       "1        NC_007112.6  BestRefSeq        exon   6642   6760  .  -  .   \n",
       "2        NC_007112.6  BestRefSeq        exon   6892   6955  .  -  .   \n",
       "3        NC_007112.6  BestRefSeq        exon   9558   9694  .  -  .   \n",
       "4        NC_007112.6  BestRefSeq        exon  10081  10191  .  -  .   \n",
       "...              ...         ...         ...    ...    ... .. .. ..   \n",
       "1258173  NC_002333.2      RefSeq  transcript  15308  16448  .  +  .   \n",
       "1258175  NC_002333.2      RefSeq  transcript  16449  16520  .  +  .   \n",
       "1258176  NC_002333.2      RefSeq        exon  16449  16520  .  +  .   \n",
       "1258177  NC_002333.2      RefSeq  transcript  16527  16596  .  -  .   \n",
       "1258178  NC_002333.2      RefSeq        exon  16527  16596  .  -  .   \n",
       "\n",
       "                                                         8  \\\n",
       "0        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "1        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "2        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "3        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "4        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "...                                                    ...   \n",
       "1258173  transcript_id \"gene42295\"; gene_id \"gene42295\"...   \n",
       "1258175  transcript_id \"rna65516\"; gene_id \"gene42296\";...   \n",
       "1258176  transcript_id \"rna65516\"; gene_id \"gene42296\";...   \n",
       "1258177  transcript_id \"rna65517\"; gene_id \"gene42297\";...   \n",
       "1258178  transcript_id \"rna65517\"; gene_id \"gene42297\";...   \n",
       "\n",
       "                                          field_dictionary gene_name  \\\n",
       "0        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...     rpl24   \n",
       "1        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...     rpl24   \n",
       "2        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...     rpl24   \n",
       "3        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...     rpl24   \n",
       "4        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...     rpl24   \n",
       "...                                                    ...       ...   \n",
       "1258173  {'transcript_id': 'gene42295', 'gene_id': 'gen...      CYTB   \n",
       "1258175  {'transcript_id': 'rna65516', 'gene_id': 'gene...      trnT   \n",
       "1258176  {'transcript_id': 'rna65516', 'gene_id': 'gene...      trnT   \n",
       "1258177  {'transcript_id': 'rna65517', 'gene_id': 'gene...      trnP   \n",
       "1258178  {'transcript_id': 'rna65517', 'gene_id': 'gene...      trnP   \n",
       "\n",
       "           gene_id transcript_id  \n",
       "0            gene0          rna0  \n",
       "1            gene0          rna0  \n",
       "2            gene0          rna0  \n",
       "3            gene0          rna0  \n",
       "4            gene0          rna0  \n",
       "...            ...           ...  \n",
       "1258173  gene42295     gene42295  \n",
       "1258175  gene42296      rna65516  \n",
       "1258176  gene42296      rna65516  \n",
       "1258177  gene42297      rna65517  \n",
       "1258178  gene42297      rna65517  \n",
       "\n",
       "[699236 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Use a custom function to extract useful fields from the additional fields section (column 8)\n",
    "# Pull from that dict to fill in additional useful columns\n",
    "models_asgtf_df['field_dictionary'] = models_asgtf_df[attributes_column].apply(convert_fields_to_dict_gtf)\n",
    "models_asgtf_df['gene_name'] = [d.get('gene_name') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['gene_id'] = [d.get('gene_id') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['transcript_id'] = [d.get('transcript_id') for d in models_asgtf_df['field_dictionary']]\n",
    "\n",
    "# Remove CDS annotations because they interfere with TransDecoder cDNA generation\n",
    "models_asgtf_df = models_asgtf_df[models_asgtf_df[2] != 'CDS']\n",
    "display(models_asgtf_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9. Generate gtf-idmm\n",
    "This file maps the `gene_name` to `gene_id` and `transcript_id` fields generated by the conversion from GFF to GTF, which will be needed for downstream processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rpl24</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cep97</td>\n",
       "      <td>gene1</td>\n",
       "      <td>rna2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>cep97</td>\n",
       "      <td>gene1</td>\n",
       "      <td>rna1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>cep97</td>\n",
       "      <td>gene1</td>\n",
       "      <td>rna3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>nfkbiz</td>\n",
       "      <td>gene2</td>\n",
       "      <td>rna4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258169</th>\n",
       "      <td>ND6</td>\n",
       "      <td>gene42293</td>\n",
       "      <td>gene42293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258171</th>\n",
       "      <td>trnE</td>\n",
       "      <td>gene42294</td>\n",
       "      <td>rna65515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258173</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>gene42295</td>\n",
       "      <td>gene42295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258175</th>\n",
       "      <td>trnT</td>\n",
       "      <td>gene42296</td>\n",
       "      <td>rna65516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258177</th>\n",
       "      <td>trnP</td>\n",
       "      <td>gene42297</td>\n",
       "      <td>rna65517</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>65914 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name    gene_id transcript_id\n",
       "0           rpl24      gene0          rna0\n",
       "13          cep97      gene1          rna2\n",
       "25          cep97      gene1          rna1\n",
       "48          cep97      gene1          rna3\n",
       "71         nfkbiz      gene2          rna4\n",
       "...           ...        ...           ...\n",
       "1258169       ND6  gene42293     gene42293\n",
       "1258171      trnE  gene42294      rna65515\n",
       "1258173      CYTB  gene42295     gene42295\n",
       "1258175      trnT  gene42296      rna65516\n",
       "1258177      trnP  gene42297      rna65517\n",
       "\n",
       "[65914 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract gene_name, gene_id, and transcript_id fields to generate an ID mapping matrix (idmm)\n",
    "idmm_df = models_asgtf_df[['gene_name', 'gene_id', 'transcript_id']].drop_duplicates()\n",
    "idmm_df.dropna(inplace = True)\n",
    "display(idmm_df)\n",
    "\n",
    "# generate a filename and file for the idmm\n",
    "idmm_filename = '_'.join([species_prefix, conditions, 'gtf-idmm.tsv'])\n",
    "idmm = IdmmFile(idmm_filename, species_SampleDict, kind = 'gtf_idmm', sources = [sample_Docket.annot])\n",
    "\n",
    "# save to file and add to the Docket\n",
    "idmm_df.to_csv(idmm.path, sep = '\\t')\n",
    "sample_Docket.add_keyfile(idmm, 'gtf_idmm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 10. Generate updated gtf\n",
    "Generated an updated GTF file using transcript_id as the key. For some datasets, transcripts do not consistently get gene names and gene IDs added, which causes Transdecoder to throw errors. This resolves that problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "models_asgtf_updated_df = models_asgtf_df.merge(idmm_df, on = 'transcript_id')\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_name': x['gene_name_y']}), axis = 1)\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_id': x['gene_id_y']}), axis = 1)\n",
    "models_asgtf_updated_df['new_fields'] = models_asgtf_updated_df['field_dictionary'].apply(convert_dict_to_fields_gtf)\n",
    "models_asgtf_updated_df = models_asgtf_updated_df[[0, 1, 2, 3, 4, 5, 6, 7, 'new_fields']]\n",
    "\n",
    "models_asgtf_updated_filename = models_asgtf.filename.replace('.gtf', '_updated.gtf')\n",
    "models_asgtf_updated = GenomeGtfFile(models_asgtf_updated_filename, species_SampleDict, GenomeFastaFile = sample_Docket.genome_fasta)\n",
    "\n",
    "models_asgtf_updated_df.to_csv(models_asgtf_updated.path, header = None, index = None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 11. Generate cDNA and peptide files\n",
    "Using the updated gtf file and genome file, generate cDNA sequence.\n",
    "\n",
    "Then, using the cDNA sequence, generate peptide sequences using transdecoder.\n",
    "\n",
    "Expect this step to take some time, probably ~20-30min."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "-parsing cufflinks output: ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_updated.gtf\n",
      "-parsing genome fasta: ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.fna\n",
      "-done parsing genome.\n",
      "// processing NC_002333.2\n",
      "// processing NC_007112.6\n",
      "// processing NC_007113.6\n",
      "// processing NC_007114.6\n",
      "// processing NC_007115.6\n",
      "// processing NC_007116.6\n",
      "// processing NC_007117.6\n",
      "// processing NC_007118.6\n",
      "// processing NC_007119.6\n",
      "// processing NC_007120.6\n",
      "// processing NC_007121.6\n",
      "// processing NC_007122.6\n",
      "// processing NC_007123.6\n",
      "// processing NC_007124.6\n",
      "// processing NC_007125.6\n",
      "// processing NC_007126.6\n",
      "// processing NC_007127.6\n",
      "// processing NC_007128.6\n",
      "// processing NC_007129.6\n",
      "// processing NC_007130.6\n",
      "// processing NC_007131.6\n",
      "// processing NC_007132.6\n",
      "// processing NC_007133.6\n",
      "// processing NC_007134.6\n",
      "// processing NC_007135.6\n",
      "// processing NC_007136.6\n",
      "// processing NW_001878118.4\n",
      "// processing NW_001878128.4\n",
      "// processing NW_001878251.4\n",
      "// processing NW_001879289.4\n",
      "// processing NW_001884381.4\n",
      "// processing NW_001884386.4\n",
      "// processing NW_001884400.4\n",
      "// processing NW_001884413.3\n",
      "// processing NW_001884427.4\n",
      "// processing NW_001884444.4\n",
      "// processing NW_003039384.3\n",
      "// processing NW_003039389.3\n",
      "// processing NW_003043344.3\n",
      "// processing NW_003044888.3\n",
      "// processing NW_003051371.3\n",
      "// processing NW_003334026.1\n",
      "// processing NW_003334027.1\n",
      "// processing NW_003334029.1\n",
      "// processing NW_003334033.1\n",
      "// processing NW_003334035.2\n",
      "// processing NW_003334071.2\n",
      "// processing NW_003334077.1\n",
      "// processing NW_003334086.1\n",
      "// processing NW_003334099.1\n",
      "// processing NW_003334114.1\n",
      "// processing NW_003334119.1\n",
      "// processing NW_003334150.2\n",
      "// processing NW_003334166.1\n",
      "// processing NW_003334203.1\n",
      "// processing NW_003334214.1\n",
      "// processing NW_003334215.1\n",
      "// processing NW_003334219.2\n",
      "// processing NW_003334233.2\n",
      "// processing NW_003334236.1\n",
      "// processing NW_003334245.1\n",
      "// processing NW_003334256.1\n",
      "// processing NW_003334261.1\n",
      "// processing NW_003334263.2\n",
      "// processing NW_003334265.1\n",
      "// processing NW_003334281.1\n",
      "// processing NW_003334318.1\n",
      "// processing NW_003334319.1\n",
      "// processing NW_003334334.2\n",
      "// processing NW_003334369.1\n",
      "// processing NW_003334432.1\n",
      "// processing NW_003334441.1\n",
      "// processing NW_003334443.1\n",
      "// processing NW_003334458.2\n",
      "// processing NW_003334505.1\n",
      "// processing NW_003334506.1\n",
      "// processing NW_003334536.1\n",
      "// processing NW_003334548.2\n",
      "// processing NW_003334656.2\n",
      "// processing NW_003334669.1\n",
      "// processing NW_003334672.2\n",
      "// processing NW_003334673.2\n",
      "// processing NW_003334680.2\n",
      "// processing NW_003334692.2\n",
      "// processing NW_003334718.1\n",
      "// processing NW_003334756.2\n",
      "// processing NW_003334810.2\n",
      "// processing NW_003334832.2\n",
      "// processing NW_003334841.1\n",
      "// processing NW_003334844.1\n",
      "// processing NW_003334875.1\n",
      "// processing NW_003334899.1\n",
      "// processing NW_003334902.1\n",
      "// processing NW_003334912.1\n",
      "// processing NW_003335000.1\n",
      "// processing NW_003335005.1\n",
      "// processing NW_003335113.1\n",
      "// processing NW_003335139.1\n",
      "// processing NW_003335143.1\n",
      "// processing NW_003335146.1\n",
      "// processing NW_003335148.1\n",
      "// processing NW_003335149.2\n",
      "// processing NW_003335153.1\n",
      "// processing NW_003335154.1\n",
      "// processing NW_003335158.1\n",
      "// processing NW_003335161.1\n",
      "// processing NW_003335163.1\n",
      "// processing NW_003335166.2\n",
      "// processing NW_003335231.1\n",
      "// processing NW_003335249.1\n",
      "// processing NW_003335266.1\n",
      "// processing NW_003335278.1\n",
      "// processing NW_003335280.1\n",
      "// processing NW_003335281.1\n",
      "// processing NW_003335289.1\n",
      "// processing NW_003335313.1\n",
      "// processing NW_003335314.1\n",
      "// processing NW_003335315.1\n",
      "// processing NW_003335321.1\n",
      "// processing NW_003335332.2\n",
      "// processing NW_003335337.2\n",
      "// processing NW_003335350.1\n",
      "// processing NW_003335355.2\n",
      "// processing NW_003335359.2\n",
      "// processing NW_003335383.2\n",
      "// processing NW_003335384.1\n",
      "// processing NW_003335385.1\n",
      "// processing NW_003335386.2\n",
      "// processing NW_003335387.1\n",
      "// processing NW_003335395.1\n",
      "// processing NW_003335408.1\n",
      "// processing NW_003335494.1\n",
      "// processing NW_003335521.2\n",
      "// processing NW_003335549.1\n",
      "// processing NW_003335558.1\n",
      "// processing NW_003335559.1\n",
      "// processing NW_003335567.1\n",
      "// processing NW_003335573.1\n",
      "// processing NW_003335577.1\n",
      "// processing NW_003335638.2\n",
      "// processing NW_003335647.1\n",
      "// processing NW_003335650.1\n",
      "// processing NW_003335654.1\n",
      "// processing NW_003335663.1\n",
      "// processing NW_003335757.1\n",
      "// processing NW_003335764.1\n",
      "// processing NW_003335779.1\n",
      "// processing NW_003335796.1\n",
      "// processing NW_003335800.1\n",
      "// processing NW_003335880.2\n",
      "// processing NW_003335884.2\n",
      "// processing NW_003335889.1\n",
      "// processing NW_003335891.1\n",
      "// processing NW_003335906.1\n",
      "// processing NW_003335907.1\n",
      "// processing NW_003335960.1\n",
      "// processing NW_003336064.1\n",
      "// processing NW_003336066.1\n",
      "// processing NW_003336073.1\n",
      "// processing NW_003336088.1\n",
      "// processing NW_003336129.1\n",
      "// processing NW_003336152.2\n",
      "// processing NW_003336157.1\n",
      "// processing NW_003336166.1\n",
      "// processing NW_003336169.1\n",
      "// processing NW_003336182.2\n",
      "// processing NW_003336186.1\n",
      "// processing NW_003336196.1\n",
      "// processing NW_003336226.1\n",
      "// processing NW_003336232.1\n",
      "// processing NW_003336241.2\n",
      "// processing NW_003336254.1\n",
      "// processing NW_003336264.2\n",
      "// processing NW_003336274.1\n",
      "// processing NW_003336276.1\n",
      "// processing NW_003336277.2\n",
      "// processing NW_003336283.1\n",
      "// processing NW_003336286.1\n",
      "// processing NW_003336287.1\n",
      "// processing NW_003336289.1\n",
      "// processing NW_003336294.1\n",
      "// processing NW_003336298.1\n",
      "// processing NW_003336299.1\n",
      "// processing NW_003336300.2\n",
      "// processing NW_003336308.1\n",
      "// processing NW_003336310.1\n",
      "// processing NW_003336316.1\n",
      "// processing NW_003336321.1\n",
      "// processing NW_003336326.1\n",
      "// processing NW_003336328.1\n",
      "// processing NW_003336331.2\n",
      "// processing NW_003336332.1\n",
      "// processing NW_003336333.2\n",
      "// processing NW_003336336.1\n",
      "// processing NW_003336337.1\n",
      "// processing NW_003336341.1\n",
      "// processing NW_003336345.1\n",
      "// processing NW_003336353.1\n",
      "// processing NW_003336354.1\n",
      "// processing NW_003336356.1\n",
      "// processing NW_003336358.1\n",
      "// processing NW_003336370.1\n",
      "// processing NW_003336371.1\n",
      "// processing NW_003336375.1\n",
      "// processing NW_003336377.1\n",
      "// processing NW_003336381.1\n",
      "// processing NW_003336388.1\n",
      "// processing NW_003336391.1\n",
      "// processing NW_003336392.1\n",
      "// processing NW_003336396.1\n",
      "// processing NW_003336398.1\n",
      "// processing NW_003336403.1\n",
      "// processing NW_003336404.2\n",
      "// processing NW_003336406.1\n",
      "// processing NW_003336407.1\n",
      "// processing NW_003336410.1\n",
      "// processing NW_003336418.1\n",
      "// processing NW_003336419.1\n",
      "// processing NW_003336420.1\n",
      "// processing NW_003336426.1\n",
      "// processing NW_003336430.1\n",
      "// processing NW_003336436.1\n",
      "// processing NW_003336441.1\n",
      "// processing NW_003336445.1\n",
      "// processing NW_003336446.1\n",
      "// processing NW_003336449.1\n",
      "// processing NW_003336453.1\n",
      "// processing NW_003336455.2\n",
      "// processing NW_003336465.1\n",
      "// processing NW_003336466.1\n",
      "// processing NW_003336468.1\n",
      "// processing NW_003336470.1\n",
      "// processing NW_003336471.1\n",
      "// processing NW_003336473.1\n",
      "// processing NW_003336475.1\n",
      "// processing NW_003336480.1\n",
      "// processing NW_003336483.1\n",
      "// processing NW_003336484.1\n",
      "// processing NW_003336490.1\n",
      "// processing NW_003336492.1\n",
      "// processing NW_003336493.1\n",
      "// processing NW_003336495.1\n",
      "// processing NW_003336501.1\n",
      "// processing NW_003336506.1\n",
      "// processing NW_003336507.1\n",
      "// processing NW_003336509.1\n",
      "// processing NW_003336511.1\n",
      "// processing NW_003336512.1\n",
      "// processing NW_003336513.1\n",
      "// processing NW_003336515.1\n",
      "// processing NW_003336519.1\n",
      "// processing NW_003336521.1\n",
      "// processing NW_003336523.1\n",
      "// processing NW_003336533.1\n",
      "// processing NW_003336534.1\n",
      "// processing NW_003336535.1\n",
      "// processing NW_003336537.1\n",
      "// processing NW_003336540.1\n",
      "// processing NW_003336543.1\n",
      "// processing NW_003336549.2\n",
      "// processing NW_003336550.1\n",
      "// processing NW_003336554.1\n",
      "// processing NW_003336559.1\n",
      "// processing NW_003336563.1\n",
      "// processing NW_003336566.1\n",
      "// processing NW_003336567.1\n",
      "// processing NW_003336572.2\n",
      "// processing NW_003336573.1\n",
      "// processing NW_003336575.1\n",
      "// processing NW_003336579.1\n",
      "// processing NW_003336582.1\n",
      "// processing NW_003336583.1\n",
      "// processing NW_003336586.1\n",
      "// processing NW_003336589.1\n",
      "// processing NW_003336591.1\n",
      "// processing NW_003336599.1\n",
      "// processing NW_003336600.1\n",
      "// processing NW_003336603.1\n",
      "// processing NW_003336606.2\n",
      "// processing NW_003336612.1\n",
      "// processing NW_003336613.1\n",
      "// processing NW_003336614.1\n",
      "// processing NW_003336623.1\n",
      "// processing NW_003336627.1\n",
      "// processing NW_003336633.1\n",
      "// processing NW_003336636.1\n",
      "// processing NW_003336639.1\n",
      "// processing NW_003336641.1\n",
      "// processing NW_003336643.1\n",
      "// processing NW_003336645.1\n",
      "// processing NW_003336646.1\n",
      "// processing NW_003336647.1\n",
      "// processing NW_003336650.1\n",
      "// processing NW_003336659.1\n",
      "// processing NW_003336661.1\n",
      "// processing NW_003336662.1\n",
      "// processing NW_003336663.1\n",
      "// processing NW_003336668.1\n",
      "// processing NW_003336671.1\n",
      "// processing NW_003336672.2\n",
      "// processing NW_003336678.1\n",
      "// processing NW_003336679.1\n",
      "// processing NW_003336682.1\n",
      "// processing NW_003336684.1\n",
      "// processing NW_003336686.1\n",
      "// processing NW_003336687.2\n",
      "// processing NW_003336689.1\n",
      "// processing NW_003336693.1\n",
      "// processing NW_003336698.1\n",
      "// processing NW_003336699.1\n",
      "// processing NW_003336701.2\n",
      "// processing NW_003336703.1\n",
      "// processing NW_003336711.1\n",
      "// processing NW_003336712.1\n",
      "// processing NW_003336714.2\n",
      "// processing NW_003336715.1\n",
      "// processing NW_003336720.1\n",
      "// processing NW_003336724.1\n",
      "// processing NW_003336729.1\n",
      "// processing NW_003336730.1\n",
      "// processing NW_003336732.1\n",
      "// processing NW_003336733.1\n",
      "// processing NW_003336739.1\n",
      "// processing NW_003336743.1\n",
      "// processing NW_003336746.1\n",
      "// processing NW_003336748.2\n",
      "// processing NW_003336751.1\n",
      "// processing NW_003336757.1\n",
      "// processing NW_003336758.1\n",
      "// processing NW_003336762.1\n",
      "// processing NW_003336772.2\n",
      "// processing NW_003336774.2\n",
      "// processing NW_003336777.1\n",
      "// processing NW_003336781.2\n",
      "// processing NW_003336784.1\n",
      "// processing NW_003336786.1\n",
      "// processing NW_003336787.1\n",
      "// processing NW_003336789.1\n",
      "// processing NW_003336792.1\n",
      "// processing NW_003336806.1\n",
      "// processing NW_003336808.1\n",
      "// processing NW_003336810.1\n",
      "// processing NW_003336812.1\n",
      "// processing NW_003336823.1\n",
      "// processing NW_003336831.1\n",
      "// processing NW_003336837.1\n",
      "// processing NW_003336838.1\n",
      "// processing NW_003336845.1\n",
      "// processing NW_003336847.1\n",
      "// processing NW_003336851.1\n",
      "// processing NW_003336855.1\n",
      "// processing NW_003336858.1\n",
      "// processing NW_003336859.1\n",
      "// processing NW_003336860.1\n",
      "// processing NW_003336867.1\n",
      "// processing NW_003336868.1\n",
      "// processing NW_003336869.1\n",
      "// processing NW_003336871.1\n",
      "// processing NW_003336874.1\n",
      "// processing NW_003336876.1\n",
      "// processing NW_003336877.1\n",
      "// processing NW_003336883.1\n",
      "// processing NW_003336886.1\n",
      "// processing NW_003336893.1\n",
      "// processing NW_003336894.1\n",
      "// processing NW_003336897.1\n",
      "// processing NW_003336900.1\n",
      "// processing NW_003336903.1\n",
      "// processing NW_003336907.1\n",
      "// processing NW_003336908.1\n",
      "// processing NW_003336909.1\n",
      "// processing NW_003336912.2\n",
      "// processing NW_003336914.1\n",
      "// processing NW_003336915.1\n",
      "// processing NW_003336917.1\n",
      "// processing NW_003336918.1\n",
      "// processing NW_003336919.1\n",
      "// processing NW_003336921.1\n",
      "// processing NW_003336925.1\n",
      "// processing NW_003336928.1\n",
      "// processing NW_003336929.1\n",
      "// processing NW_003336930.1\n",
      "// processing NW_003336934.1\n",
      "// processing NW_003336936.1\n",
      "// processing NW_003336937.1\n",
      "// processing NW_003336939.1\n",
      "// processing NW_003336940.1\n",
      "// processing NW_003336941.1\n",
      "// processing NW_003336946.1\n",
      "// processing NW_003336949.2\n",
      "// processing NW_003336950.1\n",
      "// processing NW_003336951.2\n",
      "// processing NW_003336961.1\n",
      "// processing NW_003336964.1\n",
      "// processing NW_003336967.1\n",
      "// processing NW_003336968.1\n",
      "// processing NW_003336969.1\n",
      "// processing NW_003336970.1\n",
      "// processing NW_003336972.2\n",
      "// processing NW_003336973.1\n",
      "// processing NW_003336974.1\n",
      "// processing NW_003336976.1\n",
      "// processing NW_003336977.1\n",
      "// processing NW_003336978.1\n",
      "// processing NW_003336979.1\n",
      "// processing NW_003336980.1\n",
      "// processing NW_003336991.1\n",
      "// processing NW_003336994.1\n",
      "// processing NW_003336996.1\n",
      "// processing NW_003336997.1\n",
      "// processing NW_003337002.1\n",
      "// processing NW_003337006.1\n",
      "// processing NW_003337010.1\n",
      "// processing NW_003337013.1\n",
      "// processing NW_003337014.1\n",
      "// processing NW_003337016.1\n",
      "// processing NW_003337018.1\n",
      "// processing NW_003337019.1\n",
      "// processing NW_003337027.1\n",
      "// processing NW_003337028.1\n",
      "// processing NW_003337033.1\n",
      "// processing NW_003337038.1\n",
      "// processing NW_003337040.2\n",
      "// processing NW_003337042.1\n",
      "// processing NW_003337045.1\n",
      "// processing NW_003337046.1\n",
      "// processing NW_003337049.1\n",
      "// processing NW_003337059.1\n",
      "// processing NW_003337060.1\n",
      "// processing NW_003337066.1\n",
      "// processing NW_003337067.1\n",
      "// processing NW_003337078.1\n",
      "// processing NW_003337080.1\n",
      "// processing NW_003337083.1\n",
      "// processing NW_003337085.1\n",
      "// processing NW_003337086.1\n",
      "// processing NW_003337088.1\n",
      "// processing NW_003337090.1\n",
      "// processing NW_003337091.2\n",
      "// processing NW_003337093.1\n",
      "// processing NW_003337097.1\n",
      "// processing NW_003337107.1\n",
      "// processing NW_003337109.1\n",
      "// processing NW_003337115.1\n",
      "// processing NW_003337116.1\n",
      "// processing NW_003337117.1\n",
      "// processing NW_003337119.1\n",
      "// processing NW_003337120.1\n",
      "// processing NW_003337121.1\n",
      "// processing NW_003337124.1\n",
      "// processing NW_003337125.1\n",
      "// processing NW_003337129.1\n",
      "// processing NW_003337130.1\n",
      "// processing NW_003337131.2\n",
      "// processing NW_003337135.1\n",
      "// processing NW_003337138.1\n",
      "// processing NW_003337143.1\n",
      "// processing NW_003337145.1\n",
      "// processing NW_003337146.1\n",
      "// processing NW_003337151.1\n",
      "// processing NW_003337153.2\n",
      "// processing NW_003337155.1\n",
      "// processing NW_003337159.1\n",
      "// processing NW_003337161.1\n",
      "// processing NW_003337164.1\n",
      "// processing NW_003337165.1\n",
      "// processing NW_003337169.1\n",
      "// processing NW_003337172.1\n",
      "// processing NW_003337175.1\n",
      "// processing NW_003337176.1\n",
      "// processing NW_003337183.1\n",
      "// processing NW_003337185.1\n",
      "// processing NW_003337188.1\n",
      "// processing NW_003337189.1\n",
      "// processing NW_003337190.1\n",
      "// processing NW_003337193.1\n",
      "// processing NW_003337194.1\n",
      "// processing NW_003337197.1\n",
      "// processing NW_003337199.1\n",
      "// processing NW_003337203.1\n",
      "// processing NW_003337204.1\n",
      "// processing NW_003337207.1\n",
      "// processing NW_003337218.1\n",
      "// processing NW_003337220.1\n",
      "// processing NW_003337224.1\n",
      "// processing NW_003337228.1\n",
      "// processing NW_003337230.1\n",
      "// processing NW_003337232.1\n",
      "// processing NW_003337233.1\n",
      "// processing NW_003337234.1\n",
      "// processing NW_003337239.1\n",
      "// processing NW_003337240.1\n",
      "// processing NW_003337241.1\n",
      "// processing NW_003337242.1\n",
      "// processing NW_003337253.1\n",
      "// processing NW_003337254.2\n",
      "// processing NW_003337255.1\n",
      "// processing NW_003337256.1\n",
      "// processing NW_003337258.1\n",
      "// processing NW_003337259.1\n",
      "// processing NW_003337260.1\n",
      "// processing NW_003337261.1\n",
      "// processing NW_003337263.1\n",
      "// processing NW_003337266.1\n",
      "// processing NW_008805346.1\n",
      "// processing NW_008805347.1\n",
      "// processing NW_008805349.1\n",
      "// processing NW_008805350.1\n",
      "// processing NW_008805351.1\n",
      "// processing NW_008805352.1\n",
      "// processing NW_008805353.1\n",
      "// processing NW_008805355.1\n",
      "// processing NW_008805356.1\n",
      "// processing NW_008805357.1\n",
      "// processing NW_008805358.1\n",
      "// processing NW_008805359.1\n",
      "// processing NW_008805360.1\n",
      "// processing NW_008805362.1\n",
      "// processing NW_008805365.1\n",
      "// processing NW_008805369.1\n",
      "// processing NW_008805370.1\n",
      "// processing NW_008805371.1\n",
      "// processing NW_008805374.1\n",
      "// processing NW_008805375.1\n",
      "// processing NW_008805376.1\n",
      "// processing NW_008805377.1\n",
      "// processing NW_008805379.1\n",
      "// processing NW_008805380.1\n",
      "// processing NW_008805381.1\n",
      "// processing NW_008805388.1\n",
      "// processing NW_008805389.1\n",
      "// processing NW_008805390.1\n",
      "// processing NW_008805391.1\n",
      "// processing NW_008805393.1\n",
      "// processing NW_008805394.1\n",
      "// processing NW_008805395.1\n",
      "// processing NW_008805399.1\n",
      "// processing NW_008805400.1\n",
      "// processing NW_008805402.1\n",
      "// processing NW_008805404.1\n",
      "// processing NW_008805405.1\n",
      "// processing NW_008805406.1\n",
      "// processing NW_008805408.1\n",
      "// processing NW_008805410.1\n",
      "// processing NW_008805411.1\n",
      "// processing NW_008805412.1\n",
      "// processing NW_008805413.1\n",
      "// processing NW_008805414.1\n",
      "// processing NW_008805415.1\n",
      "// processing NW_008805416.1\n",
      "// processing NW_008805417.1\n",
      "// processing NW_008805420.1\n",
      "// processing NW_008805421.1\n",
      "// processing NW_008805422.1\n",
      "// processing NW_008805424.1\n",
      "// processing NW_008805425.1\n",
      "// processing NW_008805426.1\n",
      "// processing NW_008805429.1\n",
      "// processing NW_008805430.1\n",
      "// processing NW_008805431.1\n",
      "// processing NW_008805436.1\n",
      "// processing NW_008805438.1\n",
      "// processing NW_008805439.1\n",
      "// processing NW_008805440.1\n",
      "// processing NW_008805441.1\n",
      "// processing NW_008805442.1\n",
      "// processing NW_008805443.1\n",
      "// processing NW_008805444.1\n",
      "// processing NW_008805445.1\n",
      "// processing NW_008805446.1\n",
      "// processing NW_008805447.1\n",
      "// processing NW_008805450.1\n",
      "// processing NW_008805451.1\n",
      "// processing NW_008805454.1\n",
      "// processing NW_008805455.1\n",
      "// processing NW_008805456.1\n",
      "// processing NW_008805457.1\n",
      "// processing NW_008805458.1\n",
      "// processing NW_008805459.1\n",
      "// processing NW_008805460.1\n",
      "// processing NW_008805462.1\n",
      "// processing NW_008805465.1\n",
      "// processing NW_008805467.1\n",
      "// processing NW_008805468.1\n",
      "// processing NW_008805470.1\n",
      "// processing NW_008805473.1\n",
      "// processing NW_008805474.1\n",
      "// processing NW_008805475.1\n",
      "// processing NW_008805476.1\n",
      "// processing NW_008805477.1\n",
      "// processing NW_008805478.1\n",
      "// processing NW_008805479.1\n",
      "// processing NW_008805481.1\n",
      "// processing NW_008805482.1\n",
      "// processing NW_008805483.1\n",
      "// processing NW_008805484.1\n",
      "// processing NW_008805489.1\n",
      "// processing NW_008805491.1\n",
      "// processing NW_008805492.1\n",
      "// processing NW_008805493.1\n",
      "// processing NW_008805495.1\n",
      "// processing NW_008805496.1\n",
      "// processing NW_008805497.1\n",
      "// processing NW_008805498.1\n",
      "// processing NW_008805499.1\n",
      "// processing NW_008805500.1\n",
      "// processing NW_008805501.1\n",
      "// processing NW_008805502.1\n",
      "// processing NW_008805503.1\n",
      "// processing NW_008805506.1\n",
      "// processing NW_008805507.1\n",
      "// processing NW_008805509.1\n",
      "// processing NW_008805512.1\n",
      "// processing NW_008805514.1\n",
      "// processing NW_008805515.1\n",
      "// processing NW_008805518.1\n",
      "// processing NW_008805520.1\n",
      "// processing NW_008805521.1\n",
      "// processing NW_008805523.1\n",
      "// processing NW_008805524.1\n",
      "// processing NW_008805526.1\n",
      "// processing NW_008805527.1\n",
      "// processing NW_008805528.1\n",
      "// processing NW_008805531.1\n",
      "// processing NW_008805533.1\n",
      "// processing NW_008805534.1\n",
      "// processing NW_008805536.1\n",
      "// processing NW_008805537.1\n",
      "// processing NW_008805538.1\n",
      "// processing NW_008805539.1\n",
      "// processing NW_008805540.1\n",
      "// processing NW_008805541.1\n",
      "// processing NW_008805542.1\n",
      "// processing NW_008805543.1\n",
      "// processing NW_008805545.1\n",
      "// processing NW_008805547.1\n",
      "// processing NW_008805550.1\n",
      "// processing NW_008805551.1\n",
      "// processing NW_008805552.1\n",
      "// processing NW_008805553.1\n",
      "// processing NW_008805554.1\n",
      "// processing NW_008805555.1\n",
      "// processing NW_008805558.1\n",
      "// processing NW_008805560.1\n",
      "// processing NW_008805562.1\n",
      "// processing NW_008805563.1\n",
      "// processing NW_008805564.1\n",
      "// processing NW_008805565.1\n",
      "// processing NW_008805568.1\n",
      "// processing NW_008805569.1\n",
      "// processing NW_008805571.1\n",
      "// processing NW_008805572.1\n",
      "// processing NW_008805573.1\n",
      "// processing NW_008805574.1\n",
      "// processing NW_008805576.1\n",
      "// processing NW_008805577.1\n",
      "// processing NW_008805579.1\n",
      "// processing NW_008805580.1\n",
      "// processing NW_008805582.1\n",
      "// processing NW_008805584.1\n",
      "// processing NW_008805585.1\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/compute_base_probs.pl ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna 0 > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//base_freqs.dat\n",
      "\n",
      "\n",
      "-first extracting base frequencies, we'll need them later.\n",
      "\n",
      "\n",
      "- extracting ORFs from transcripts.\n",
      "-total transcripts to examine: 65901\n",
      "[65900/65901] = 100.00% done    \n",
      "\n",
      "#################################\n",
      "### Done preparing long ORFs.  ###\n",
      "##################################\n",
      "\n",
      "\tUse file: ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.pep  for Pfam and/or BlastP searches to enable homology-based coding region identification.\n",
      "\n",
      "\tThen, run TransDecoder.Predict for your final coding region predictions.\n",
      "\n",
      "\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds 5000 5000 > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CMD: touch ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/.__checkpoints_longorfs/TD.longorfs.ok\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/exclude_similar_proteins.pl ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000 > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr\n",
      "-skipping training candidate: rna40168.p1, not unique enough\n",
      "-skipping training candidate: rna25405.p1, not unique enough\n",
      "-skipping training candidate: rna3897.p1, not unique enough\n",
      "-skipping training candidate: rna44478.p1, not unique enough\n",
      "-skipping training candidate: rna27051.p1, not unique enough\n",
      "-skipping training candidate: rna55607.p1, not unique enough\n",
      "-skipping training candidate: rna9862.p1, not unique enough\n",
      "-skipping training candidate: rna9857.p1, not unique enough\n",
      "-skipping training candidate: rna27052.p1, not unique enough\n",
      "-skipping training candidate: rna55137.p1, not unique enough\n",
      "-skipping training candidate: rna44428.p1, not unique enough\n",
      "-skipping training candidate: rna44430.p1, not unique enough\n",
      "-skipping training candidate: rna42817.p1, not unique enough\n",
      "-skipping training candidate: rna52108.p1, not unique enough\n",
      "-skipping training candidate: rna27053.p1, not unique enough\n",
      "-skipping training candidate: rna27050.p1, not unique enough\n",
      "-skipping training candidate: rna25293.p1, not unique enough\n",
      "-skipping training candidate: rna25406.p1, not unique enough\n",
      "-skipping training candidate: rna27055.p1, not unique enough\n",
      "-skipping training candidate: rna27054.p1, not unique enough\n",
      "-skipping training candidate: rna42814.p1, not unique enough\n",
      "-skipping training candidate: rna46840.p1, not unique enough\n",
      "-skipping training candidate: rna26735.p1, not unique enough\n",
      "-skipping training candidate: rna42808.p1, not unique enough\n",
      "-skipping training candidate: rna25570.p1, not unique enough\n",
      "-skipping training candidate: rna9037.p1, not unique enough\n",
      "-skipping training candidate: rna9038.p1, not unique enough\n",
      "-skipping training candidate: rna45707.p1, not unique enough\n",
      "-skipping training candidate: rna44429.p1, not unique enough\n",
      "-skipping training candidate: rna26731.p1, not unique enough\n",
      "-skipping training candidate: rna25572.p1, not unique enough\n",
      "-skipping training candidate: rna62075.p1, not unique enough\n",
      "-skipping training candidate: rna52107.p1, not unique enough\n",
      "-skipping training candidate: rna26730.p1, not unique enough\n",
      "-skipping training candidate: rna34852.p1, not unique enough\n",
      "-skipping training candidate: rna9034.p1, not unique enough\n",
      "-skipping training candidate: rna26733.p1, not unique enough\n",
      "-skipping training candidate: rna26732.p1, not unique enough\n",
      "-skipping training candidate: rna9035.p1, not unique enough\n",
      "-skipping training candidate: rna33746.p1, not unique enough\n",
      "-skipping training candidate: rna9861.p1, not unique enough\n",
      "-skipping training candidate: rna25588.p1, not unique enough\n",
      "-skipping training candidate: rna27851.p1, not unique enough\n",
      "-skipping training candidate: rna40872.p1, not unique enough\n",
      "-skipping training candidate: rna25578.p1, not unique enough\n",
      "-skipping training candidate: rna40869.p1, not unique enough\n",
      "-skipping training candidate: rna45819.p1, not unique enough\n",
      "-skipping training candidate: rna6106.p1, not unique enough\n",
      "-skipping training candidate: rna29813.p1, not unique enough\n",
      "-skipping training candidate: rna54069.p1, not unique enough\n",
      "-skipping training candidate: rna54073.p1, not unique enough\n",
      "-skipping training candidate: rna40870.p1, not unique enough\n",
      "-skipping training candidate: rna45820.p1, not unique enough\n",
      "-skipping training candidate: rna54072.p1, not unique enough\n",
      "-skipping training candidate: rna54070.p1, not unique enough\n",
      "-skipping training candidate: rna18522.p1, not unique enough\n",
      "-skipping training candidate: rna27056.p1, not unique enough\n",
      "-skipping training candidate: rna27059.p1, not unique enough\n",
      "-skipping training candidate: rna21114.p1, not unique enough\n",
      "-skipping training candidate: rna25560.p1, not unique enough\n",
      "-skipping training candidate: rna25569.p1, not unique enough\n",
      "-skipping training candidate: rna34746.p1, not unique enough\n",
      "-skipping training candidate: rna34287.p1, not unique enough\n",
      "-skipping training candidate: rna5552.p1, not unique enough\n",
      "-skipping training candidate: rna9858.p1, not unique enough\n",
      "-skipping training candidate: rna25586.p1, not unique enough\n",
      "-skipping training candidate: rna25559.p1, not unique enough\n",
      "-skipping training candidate: rna32559.p1, not unique enough\n",
      "-skipping training candidate: rna25568.p1, not unique enough\n",
      "-skipping training candidate: rna25566.p1, not unique enough\n",
      "-skipping training candidate: rna32289.p1, not unique enough\n",
      "-skipping training candidate: rna33854.p1, not unique enough\n",
      "-skipping training candidate: rna34288.p1, not unique enough\n",
      "-skipping training candidate: rna39237.p1, not unique enough\n",
      "-skipping training candidate: rna59534.p1, not unique enough\n",
      "-skipping training candidate: rna40868.p1, not unique enough\n",
      "-skipping training candidate: rna44477.p1, not unique enough\n",
      "-skipping training candidate: rna25292.p1, not unique enough\n",
      "-skipping training candidate: rna27057.p1, not unique enough\n",
      "-skipping training candidate: rna27058.p1, not unique enough\n",
      "-skipping training candidate: rna61133.p1, not unique enough\n",
      "-skipping training candidate: rna34286.p1, not unique enough\n",
      "-skipping training candidate: rna50075.p1, not unique enough\n",
      "-skipping training candidate: rna62158.p1, not unique enough\n",
      "-skipping training candidate: rna26541.p1, not unique enough\n",
      "-skipping training candidate: rna26542.p1, not unique enough\n",
      "-skipping training candidate: rna36590.p1, not unique enough\n",
      "-skipping training candidate: rna40873.p1, not unique enough\n",
      "-skipping training candidate: rna36152.p1, not unique enough\n",
      "-skipping training candidate: rna38096.p1, not unique enough\n",
      "-skipping training candidate: rna38098.p1, not unique enough\n",
      "-skipping training candidate: rna23132.p1, not unique enough\n",
      "-skipping training candidate: rna23134.p1, not unique enough\n",
      "-skipping training candidate: rna32595.p1, not unique enough\n",
      "-skipping training candidate: rna34745.p1, not unique enough\n",
      "-skipping training candidate: rna61136.p1, not unique enough\n",
      "-skipping training candidate: rna23135.p1, not unique enough\n",
      "-skipping training candidate: rna40875.p1, not unique enough\n",
      "-skipping training candidate: rna25575.p1, not unique enough\n",
      "-skipping training candidate: rna58591.p1, not unique enough\n",
      "-skipping training candidate: rna61135.p1, not unique enough\n",
      "-skipping training candidate: rna62684.p1, not unique enough\n",
      "-skipping training candidate: rna9032.p1, not unique enough\n",
      "-skipping training candidate: rna44767.p1, not unique enough\n",
      "-skipping training candidate: rna25557.p1, not unique enough\n",
      "-skipping training candidate: rna32594.p1, not unique enough\n",
      "-skipping training candidate: rna40876.p1, not unique enough\n",
      "-skipping training candidate: rna46637.p1, not unique enough\n",
      "-skipping training candidate: rna38097.p1, not unique enough\n",
      "-skipping training candidate: rna23133.p1, not unique enough\n",
      "-skipping training candidate: rna25585.p1, not unique enough\n",
      "-skipping training candidate: rna40874.p1, not unique enough\n",
      "-skipping training candidate: rna25562.p1, not unique enough\n",
      "-skipping training candidate: rna25555.p1, not unique enough\n",
      "-skipping training candidate: rna52433.p1, not unique enough\n",
      "-skipping training candidate: rna5963.p1, not unique enough\n",
      "-skipping training candidate: rna5959.p1, not unique enough\n",
      "-skipping training candidate: rna25009.p1, not unique enough\n",
      "-skipping training candidate: rna25561.p1, not unique enough\n",
      "-skipping training candidate: rna32593.p1, not unique enough\n",
      "-skipping training candidate: rna52432.p1, not unique enough\n",
      "-skipping training candidate: rna9860.p1, not unique enough\n",
      "-skipping training candidate: rna44765.p1, not unique enough\n",
      "-skipping training candidate: rna58592.p1, not unique enough\n",
      "-skipping training candidate: rna58593.p1, not unique enough\n",
      "-skipping training candidate: rna58594.p1, not unique enough\n",
      "-skipping training candidate: rna63286.p1, not unique enough\n",
      "-skipping training candidate: rna32596.p1, not unique enough\n",
      "-skipping training candidate: rna32589.p1, not unique enough\n",
      "-skipping training candidate: rna61137.p1, not unique enough\n",
      "-skipping training candidate: rna22796.p1, not unique enough\n",
      "-skipping training candidate: rna42797.p1, not unique enough\n",
      "-skipping training candidate: rna32225.p1, not unique enough\n",
      "-skipping training candidate: rna46247.p1, not unique enough\n",
      "-skipping training candidate: rna30996.p1, not unique enough\n",
      "-skipping training candidate: rna50811.p1, not unique enough\n",
      "-skipping training candidate: rna58589.p1, not unique enough\n",
      "-skipping training candidate: rna38316.p1, not unique enough\n",
      "-skipping training candidate: rna42801.p1, not unique enough\n",
      "-skipping training candidate: rna63289.p1, not unique enough\n",
      "-skipping training candidate: rna5950.p1, not unique enough\n",
      "-skipping training candidate: rna42795.p1, not unique enough\n",
      "-skipping training candidate: rna3125.p1, not unique enough\n",
      "-skipping training candidate: rna38318.p1, not unique enough\n",
      "-skipping training candidate: rna30997.p1, not unique enough\n",
      "-skipping training candidate: rna23494.p1, not unique enough\n",
      "-skipping training candidate: rna33745.p1, not unique enough\n",
      "-skipping training candidate: rna34443.p1, not unique enough\n",
      "-skipping training candidate: rna50809.p1, not unique enough\n",
      "-skipping training candidate: rna42798.p1, not unique enough\n",
      "-skipping training candidate: rna59018.p1, not unique enough\n",
      "-skipping training candidate: rna3278.p1, not unique enough\n",
      "-skipping training candidate: rna4441.p1, not unique enough\n",
      "-skipping training candidate: rna31002.p1, not unique enough\n",
      "-skipping training candidate: rna30994.p1, not unique enough\n",
      "-skipping training candidate: rna44770.p1, not unique enough\n",
      "-skipping training candidate: rna59359.p1, not unique enough\n",
      "-skipping training candidate: rna30995.p1, not unique enough\n",
      "-skipping training candidate: rna9387.p1, not unique enough\n",
      "-skipping training candidate: rna9386.p1, not unique enough\n",
      "-skipping training candidate: rna32227.p1, not unique enough\n",
      "-skipping training candidate: rna32591.p1, not unique enough\n",
      "-skipping training candidate: rna32590.p1, not unique enough\n",
      "-skipping training candidate: rna42113.p1, not unique enough\n",
      "-skipping training candidate: rna42802.p1, not unique enough\n",
      "-skipping training candidate: rna54071.p1, not unique enough\n",
      "-skipping training candidate: rna9385.p1, not unique enough\n",
      "-skipping training candidate: rna32226.p1, not unique enough\n",
      "-skipping training candidate: rna62758.p1, not unique enough\n",
      "-skipping training candidate: rna22306.p1, not unique enough\n",
      "-skipping training candidate: rna25558.p1, not unique enough\n",
      "-skipping training candidate: rna33549.p1, not unique enough\n",
      "-skipping training candidate: rna33550.p1, not unique enough\n",
      "-skipping training candidate: rna33548.p1, not unique enough\n",
      "-skipping training candidate: rna34289.p1, not unique enough\n",
      "-skipping training candidate: rna44771.p1, not unique enough\n",
      "-skipping training candidate: rna42805.p1, not unique enough\n",
      "-skipping training candidate: rna46638.p1, not unique enough\n",
      "-skipping training candidate: rna52545.p1, not unique enough\n",
      "-skipping training candidate: rna59360.p1, not unique enough\n",
      "-skipping training candidate: rna29474.p1, not unique enough\n",
      "-skipping training candidate: rna38918.p1, not unique enough\n",
      "-skipping training candidate: rna45818.p1, not unique enough\n",
      "-skipping training candidate: rna52430.p1, not unique enough\n",
      "-skipping training candidate: rna31517.p1, not unique enough\n",
      "-skipping training candidate: rna38317.p1, not unique enough\n",
      "-skipping training candidate: rna44766.p1, not unique enough\n",
      "-skipping training candidate: rna45954.p1, not unique enough\n",
      "-skipping training candidate: rna45956.p1, not unique enough\n",
      "-skipping training candidate: rna53757.p1, not unique enough\n",
      "-skipping training candidate: rna59362.p1, not unique enough\n",
      "-skipping training candidate: rna18833.p1, not unique enough\n",
      "-skipping training candidate: rna61134.p1, not unique enough\n",
      "-skipping training candidate: rna31000.p1, not unique enough\n",
      "-skipping training candidate: rna63290.p1, not unique enough\n",
      "-skipping training candidate: rna1783.p1, not unique enough\n",
      "-skipping training candidate: rna1781.p1, not unique enough\n",
      "-skipping training candidate: rna46635.p1, not unique enough\n",
      "-skipping training candidate: rna46002.p1, not unique enough\n",
      "-skipping training candidate: rna4780.p1, not unique enough\n",
      "-skipping training candidate: rna30999.p1, not unique enough\n",
      "-skipping training candidate: rna36703.p1, not unique enough\n",
      "-skipping training candidate: rna38320.p1, not unique enough\n",
      "-skipping training candidate: rna8454.p1, not unique enough\n",
      "-skipping training candidate: rna9384.p1, not unique enough\n",
      "-skipping training candidate: rna22555.p1, not unique enough\n",
      "-skipping training candidate: rna22504.p1, not unique enough\n",
      "-skipping training candidate: rna52543.p1, not unique enough\n",
      "-skipping training candidate: rna9388.p1, not unique enough\n",
      "-skipping training candidate: rna32592.p1, not unique enough\n",
      "-skipping training candidate: rna49053.p1, not unique enough\n",
      "-skipping training candidate: rna30993.p1, not unique enough\n",
      "-skipping training candidate: rna62076.p1, not unique enough\n",
      "-skipping training candidate: rna20463.p1, not unique enough\n",
      "-skipping training candidate: rna59019.p1, not unique enough\n",
      "-skipping training candidate: rna9391.p1, not unique enough\n",
      "-skipping training candidate: rna1102.p1, not unique enough\n",
      "-skipping training candidate: rna31061.p1, not unique enough\n",
      "-skipping training candidate: rna32335.p1, not unique enough\n",
      "-skipping training candidate: rna42944.p1, not unique enough\n",
      "-skipping training candidate: rna42942.p1, not unique enough\n",
      "-skipping training candidate: rna50965.p1, not unique enough\n",
      "-skipping training candidate: rna50964.p1, not unique enough\n",
      "-skipping training candidate: rna52541.p1, not unique enough\n",
      "-skipping training candidate: rna30998.p1, not unique enough\n",
      "-skipping training candidate: rna46636.p1, not unique enough\n",
      "-skipping training candidate: rna28765.p1, not unique enough\n",
      "-skipping training candidate: rna63116.p1, not unique enough\n",
      "-skipping training candidate: rna9389.p1, not unique enough\n",
      "-skipping training candidate: rna21183.p1, not unique enough\n",
      "-skipping training candidate: rna21186.p1, not unique enough\n",
      "-skipping training candidate: rna9390.p1, not unique enough\n",
      "-skipping training candidate: rna21184.p1, not unique enough\n",
      "-skipping training candidate: rna42794.p1, not unique enough\n",
      "-skipping training candidate: rna50976.p1, not unique enough\n",
      "-skipping training candidate: rna50975.p1, not unique enough\n",
      "-skipping training candidate: rna58020.p1, not unique enough\n",
      "-skipping training candidate: rna50966.p1, not unique enough\n",
      "-skipping training candidate: rna22783.p1, not unique enough\n",
      "-skipping training candidate: rna44768.p1, not unique enough\n",
      "-skipping training candidate: rna50967.p1, not unique enough\n",
      "-skipping training candidate: rna50974.p1, not unique enough\n",
      "-skipping training candidate: rna50973.p1, not unique enough\n",
      "-skipping training candidate: rna63288.p1, not unique enough\n",
      "-skipping training candidate: rna39774.p1, not unique enough\n",
      "-skipping training candidate: rna50832.p1, not unique enough\n",
      "-skipping training candidate: rna50835.p1, not unique enough\n",
      "-skipping training candidate: rna4782.p1, not unique enough\n",
      "-skipping training candidate: rna22797.p1, not unique enough\n",
      "-skipping training candidate: rna25295.p1, not unique enough\n",
      "-skipping training candidate: rna7137.p1, not unique enough\n",
      "-skipping training candidate: rna5953.p1, not unique enough\n",
      "-skipping training candidate: rna9033.p1, not unique enough\n",
      "-skipping training candidate: rna25567.p1, not unique enough\n",
      "-skipping training candidate: rna5952.p1, not unique enough\n",
      "-skipping training candidate: rna22782.p1, not unique enough\n",
      "-skipping training candidate: rna21682.p1, not unique enough\n",
      "-skipping training candidate: rna38319.p1, not unique enough\n",
      "-skipping training candidate: rna63118.p1, not unique enough\n",
      "-skipping training candidate: rna20607.p1, not unique enough\n",
      "-skipping training candidate: rna21681.p1, not unique enough\n",
      "-skipping training candidate: rna38323.p1, not unique enough\n",
      "-skipping training candidate: rna43820.p1, not unique enough\n",
      "-skipping training candidate: rna44769.p1, not unique enough\n",
      "-skipping training candidate: rna49052.p1, not unique enough\n",
      "-skipping training candidate: rna49055.p1, not unique enough\n",
      "-skipping training candidate: rna50971.p1, not unique enough\n",
      "-skipping training candidate: rna22780.p1, not unique enough\n",
      "-skipping training candidate: rna42796.p1, not unique enough\n",
      "-skipping training candidate: rna41753.p1, not unique enough\n",
      "-skipping training candidate: rna44772.p1, not unique enough\n",
      "-skipping training candidate: rna7141.p1, not unique enough\n",
      "-skipping training candidate: rna31756.p1, not unique enough\n",
      "-skipping training candidate: rna931.p1, not unique enough\n",
      "-skipping training candidate: rna932.p1, not unique enough\n",
      "-skipping training candidate: rna35895.p1, not unique enough\n",
      "-skipping training candidate: rna50983.p1, not unique enough\n",
      "-skipping training candidate: rna36376.p1, not unique enough\n",
      "-skipping training candidate: rna42799.p1, not unique enough\n",
      "-skipping training candidate: rna43654.p1, not unique enough\n",
      "-skipping training candidate: rna50977.p1, not unique enough\n",
      "-skipping training candidate: rna31755.p1, not unique enough\n",
      "-skipping training candidate: rna50978.p1, not unique enough\n",
      "-skipping training candidate: rna50968.p1, not unique enough\n",
      "-skipping training candidate: rna1104.p1, not unique enough\n",
      "-skipping training candidate: rna1465.p1, not unique enough\n",
      "-skipping training candidate: rna35355.p1, not unique enough\n",
      "-skipping training candidate: rna829.p1, not unique enough\n",
      "-skipping training candidate: rna7144.p1, not unique enough\n",
      "-skipping training candidate: rna23989.p1, not unique enough\n",
      "-skipping training candidate: rna35354.p1, not unique enough\n",
      "-skipping training candidate: rna35357.p1, not unique enough\n",
      "-skipping training candidate: rna38760.p1, not unique enough\n",
      "-skipping training candidate: rna53758.p1, not unique enough\n",
      "-skipping training candidate: rna4362.p1, not unique enough\n",
      "-skipping training candidate: rna40867.p1, not unique enough\n",
      "-skipping training candidate: rna50992.p1, not unique enough\n",
      "-skipping training candidate: rna52544.p1, not unique enough\n",
      "-skipping training candidate: rna6929.p1, not unique enough\n",
      "-skipping training candidate: rna21237.p1, not unique enough\n",
      "-skipping training candidate: rna22789.p1, not unique enough\n",
      "-skipping training candidate: rna37785.p1, not unique enough\n",
      "-skipping training candidate: rna41752.p1, not unique enough\n",
      "-skipping training candidate: rna42800.p1, not unique enough\n",
      "-skipping training candidate: rna1105.p1, not unique enough\n",
      "-skipping training candidate: rna23864.p1, not unique enough\n",
      "-skipping training candidate: rna31003.p1, not unique enough\n",
      "-skipping training candidate: rna41748.p1, not unique enough\n",
      "-skipping training candidate: rna50979.p1, not unique enough\n",
      "-skipping training candidate: rna828.p1, not unique enough\n",
      "-skipping training candidate: rna21236.p1, not unique enough\n",
      "-skipping training candidate: rna22503.p1, not unique enough\n",
      "-skipping training candidate: rna27963.p1, not unique enough\n",
      "-skipping training candidate: rna29121.p1, not unique enough\n",
      "-skipping training candidate: rna29120.p1, not unique enough\n",
      "-skipping training candidate: rna29473.p1, not unique enough\n",
      "-skipping training candidate: rna31011.p1, not unique enough\n",
      "-skipping training candidate: rna50980.p1, not unique enough\n",
      "-skipping training candidate: rna62759.p1, not unique enough\n",
      "-skipping training candidate: rna3384.p1, not unique enough\n",
      "-skipping training candidate: rna3381.p1, not unique enough\n",
      "-skipping training candidate: rna3383.p1, not unique enough\n",
      "-skipping training candidate: rna3380.p1, not unique enough\n",
      "-skipping training candidate: rna38325.p1, not unique enough\n",
      "-skipping training candidate: rna49501.p1, not unique enough\n",
      "-skipping training candidate: rna52338.p1, not unique enough\n",
      "-skipping training candidate: rna3382.p1, not unique enough\n",
      "-skipping training candidate: rna7495.p1, not unique enough\n",
      "-skipping training candidate: rna18461.p1, not unique enough\n",
      "-skipping training candidate: rna25556.p1, not unique enough\n",
      "-skipping training candidate: rna7494.p1, not unique enough\n",
      "-skipping training candidate: rna31754.p1, not unique enough\n",
      "-skipping training candidate: rna59135.p1, not unique enough\n",
      "-skipping training candidate: rna41469.p1, not unique enough\n",
      "-skipping training candidate: rna48943.p1, not unique enough\n",
      "-skipping training candidate: rna48941.p1, not unique enough\n",
      "-skipping training candidate: rna29359.p1, not unique enough\n",
      "-skipping training candidate: rna38901.p1, not unique enough\n",
      "-skipping training candidate: rna22784.p1, not unique enough\n",
      "-skipping training candidate: rna27964.p1, not unique enough\n",
      "-skipping training candidate: rna30558.p1, not unique enough\n",
      "-skipping training candidate: rna38327.p1, not unique enough\n",
      "-skipping training candidate: rna46640.p1, not unique enough\n",
      "-skipping training candidate: rna47995.p1, not unique enough\n",
      "-skipping training candidate: rna50969.p1, not unique enough\n",
      "-skipping training candidate: rna21680.p1, not unique enough\n",
      "-skipping training candidate: rna31752.p1, not unique enough\n",
      "-skipping training candidate: rna38324.p1, not unique enough\n",
      "-skipping training candidate: rna46264.p1, not unique enough\n",
      "-skipping training candidate: rna43821.p1, not unique enough\n",
      "-skipping training candidate: rna60130.p1, not unique enough\n",
      "-skipping training candidate: rna23515.p1, not unique enough\n",
      "-skipping training candidate: rna59136.p1, not unique enough\n",
      "-skipping training candidate: rna28366.p1, not unique enough\n",
      "-skipping training candidate: rna57332.p1, not unique enough\n",
      "-skipping training candidate: rna28365.p1, not unique enough\n",
      "-skipping training candidate: rna32338.p1, not unique enough\n",
      "-skipping training candidate: rna42804.p1, not unique enough\n",
      "-skipping training candidate: rna59138.p1, not unique enough\n",
      "-skipping training candidate: rna7142.p1, not unique enough\n",
      "-skipping training candidate: rna28761.p1, not unique enough\n",
      "-skipping training candidate: rna30559.p1, not unique enough\n",
      "-skipping training candidate: rna37790.p1, not unique enough\n",
      "-skipping training candidate: rna50989.p1, not unique enough\n",
      "-skipping training candidate: rna60496.p1, not unique enough\n",
      "-skipping training candidate: rna60494.p1, not unique enough\n",
      "-skipping training candidate: rna60495.p1, not unique enough\n",
      "-skipping training candidate: rna61768.p1, not unique enough\n",
      "-skipping training candidate: rna7143.p1, not unique enough\n",
      "-skipping training candidate: rna31001.p1, not unique enough\n",
      "-skipping training candidate: rna40877.p1, not unique enough\n",
      "-skipping training candidate: rna41298.p1, not unique enough\n",
      "-skipping training candidate: rna52542.p1, not unique enough\n",
      "-skipping training candidate: rna53756.p1, not unique enough\n",
      "-skipping training candidate: rna19935.p1, not unique enough\n",
      "-skipping training candidate: rna38210.p1, not unique enough\n",
      "-skipping training candidate: rna52546.p1, not unique enough\n",
      "-skipping training candidate: rna34292.p1, not unique enough\n",
      "-skipping training candidate: rna40711.p1, not unique enough\n",
      "-skipping training candidate: rna40709.p1, not unique enough\n",
      "-skipping training candidate: rna56315.p1, not unique enough\n",
      "-skipping training candidate: rna61769.p1, not unique enough\n",
      "-skipping training candidate: rna32337.p1, not unique enough\n",
      "-skipping training candidate: rna40713.p1, not unique enough\n",
      "-skipping training candidate: rna42803.p1, not unique enough\n",
      "-skipping training candidate: rna56316.p1, not unique enough\n",
      "-skipping training candidate: rna56314.p1, not unique enough\n",
      "-skipping training candidate: rna61767.p1, not unique enough\n",
      "-skipping training candidate: rna691.p1, not unique enough\n",
      "-skipping training candidate: rna38322.p1, not unique enough\n",
      "-skipping training candidate: rna56313.p1, not unique enough\n",
      "-skipping training candidate: rna60627.p1, not unique enough\n",
      "-skipping training candidate: rna1106.p1, not unique enough\n",
      "-skipping training candidate: rna7140.p1, not unique enough\n",
      "-skipping training candidate: rna49054.p1, not unique enough\n",
      "-skipping training candidate: rna50984.p1, not unique enough\n",
      "-skipping training candidate: rna40712.p1, not unique enough\n",
      "-skipping training candidate: rna47993.p1, not unique enough\n",
      "-skipping training candidate: rna694.p1, not unique enough\n",
      "-skipping training candidate: rna26090.p1, not unique enough\n",
      "-skipping training candidate: rna42816.p1, not unique enough\n",
      "-skipping training candidate: rna59957.p1, not unique enough\n",
      "-skipping training candidate: rna25010.p1, not unique enough\n",
      "-skipping training candidate: rna43822.p1, not unique enough\n",
      "-skipping training candidate: rna56312.p1, not unique enough\n",
      "-skipping training candidate: rna50982.p1, not unique enough\n",
      "-skipping training candidate: rna22785.p1, not unique enough\n",
      "-skipping training candidate: rna31007.p1, not unique enough\n",
      "-skipping training candidate: rna38761.p1, not unique enough\n",
      "-skipping training candidate: rna50970.p1, not unique enough\n",
      "-skipping training candidate: rna49782.p1, not unique enough\n",
      "-skipping training candidate: rna49780.p1, not unique enough\n",
      "-skipping training candidate: rna56311.p1, not unique enough\n",
      "-skipping training candidate: rna693.p1, not unique enough\n",
      "-skipping training candidate: rna45117.p1, not unique enough\n",
      "-skipping training candidate: rna60615.p1, not unique enough\n",
      "-skipping training candidate: rna63771.p1, not unique enough\n",
      "-skipping training candidate: rna41750.p1, not unique enough\n",
      "-skipping training candidate: rna46718.p1, not unique enough\n",
      "-skipping training candidate: rna56310.p1, not unique enough\n",
      "-skipping training candidate: rna60616.p1, not unique enough\n",
      "-skipping training candidate: rna50985.p1, not unique enough\n",
      "-skipping training candidate: rna63773.p1, not unique enough\n",
      "-skipping training candidate: rna7211.p1, not unique enough\n",
      "-skipping training candidate: rna37655.p1, not unique enough\n",
      "-skipping training candidate: rna46641.p1, not unique enough\n",
      "-skipping training candidate: rna56319.p1, not unique enough\n",
      "-skipping training candidate: rna5954.p1, not unique enough\n",
      "-skipping training candidate: rna62428.p1, not unique enough\n",
      "-skipping training candidate: rna62252.p1, not unique enough\n",
      "-skipping training candidate: rna7991.p1, not unique enough\n",
      "-skipping training candidate: rna22781.p1, not unique enough\n",
      "-skipping training candidate: rna27894.p1, not unique enough\n",
      "-skipping training candidate: rna27895.p1, not unique enough\n",
      "-skipping training candidate: rna38073.p1, not unique enough\n",
      "-skipping training candidate: rna38076.p1, not unique enough\n",
      "-skipping training candidate: rna22788.p1, not unique enough\n",
      "-skipping training candidate: rna22787.p1, not unique enough\n",
      "-skipping training candidate: rna26253.p1, not unique enough\n",
      "-skipping training candidate: rna26254.p1, not unique enough\n",
      "-skipping training candidate: rna28161.p1, not unique enough\n",
      "-skipping training candidate: rna1147.p1, not unique enough\n",
      "-skipping training candidate: rna37654.p1, not unique enough\n",
      "-skipping training candidate: rna56318.p1, not unique enough\n",
      "-skipping training candidate: rna38069.p1, not unique enough\n",
      "-skipping training candidate: rna50981.p1, not unique enough\n",
      "-skipping training candidate: rna59931.p1, not unique enough\n",
      "-skipping training candidate: rna23649.p1, not unique enough\n",
      "-skipping training candidate: rna33890.p1, not unique enough\n",
      "-skipping training candidate: rna33878.p1, not unique enough\n",
      "-skipping training candidate: rna36227.p1, not unique enough\n",
      "-skipping training candidate: rna45120.p1, not unique enough\n",
      "-skipping training candidate: rna9870.p1, not unique enough\n",
      "-skipping training candidate: rna23651.p1, not unique enough\n",
      "-skipping training candidate: rna29122.p1, not unique enough\n",
      "-skipping training candidate: rna38161.p1, not unique enough\n",
      "-skipping training candidate: rna9873.p1, not unique enough\n",
      "-skipping training candidate: rna9872.p1, not unique enough\n",
      "-skipping training candidate: rna23646.p1, not unique enough\n",
      "-skipping training candidate: rna23650.p1, not unique enough\n",
      "-skipping training candidate: rna36216.p1, not unique enough\n",
      "-skipping training candidate: rna18814.p1, not unique enough\n",
      "-skipping training candidate: rna29123.p1, not unique enough\n",
      "-skipping training candidate: rna36226.p1, not unique enough\n",
      "-skipping training candidate: rna36220.p1, not unique enough\n",
      "-skipping training candidate: rna38899.p1, not unique enough\n",
      "-skipping training candidate: rna41295.p1, not unique enough\n",
      "-skipping training candidate: rna61766.p1, not unique enough\n",
      "-skipping training candidate: rna827.p1, not unique enough\n",
      "-skipping training candidate: rna27854.p1, not unique enough\n",
      "-skipping training candidate: rna36228.p1, not unique enough\n",
      "-skipping training candidate: rna41297.p1, not unique enough\n",
      "-skipping training candidate: rna50972.p1, not unique enough\n",
      "-skipping training candidate: rna55143.p1, not unique enough\n",
      "-skipping training candidate: rna27852.p1, not unique enough\n",
      "-skipping training candidate: rna33881.p1, not unique enough\n",
      "-skipping training candidate: rna36229.p1, not unique enough\n",
      "-skipping training candidate: rna36214.p1, not unique enough\n",
      "-skipping training candidate: rna36230.p1, not unique enough\n",
      "-skipping training candidate: rna38902.p1, not unique enough\n",
      "-skipping training candidate: rna62339.p1, not unique enough\n",
      "-skipping training candidate: rna23865.p1, not unique enough\n",
      "-skipping training candidate: rna33879.p1, not unique enough\n",
      "-skipping training candidate: rna43823.p1, not unique enough\n",
      "-skipping training candidate: rna50986.p1, not unique enough\n",
      "-skipping training candidate: rna36225.p1, not unique enough\n",
      "-skipping training candidate: rna37653.p1, not unique enough\n",
      "-skipping training candidate: rna50987.p1, not unique enough\n",
      "-skipping training candidate: rna883.p1, not unique enough\n",
      "-skipping training candidate: rna19018.p1, not unique enough\n",
      "-skipping training candidate: rna36224.p1, not unique enough\n",
      "-skipping training candidate: rna36219.p1, not unique enough\n",
      "-skipping training candidate: rna60133.p1, not unique enough\n",
      "-skipping training candidate: rna23648.p1, not unique enough\n",
      "-skipping training candidate: rna27853.p1, not unique enough\n",
      "-skipping training candidate: rna1107.p1, not unique enough\n",
      "-skipping training candidate: rna23866.p1, not unique enough\n",
      "-skipping training candidate: rna31004.p1, not unique enough\n",
      "-skipping training candidate: rna45118.p1, not unique enough\n",
      "-skipping training candidate: rna53348.p1, not unique enough\n",
      "-skipping training candidate: rna55976.p1, not unique enough\n",
      "-skipping training candidate: rna26044.p1, not unique enough\n",
      "-skipping training candidate: rna58449.p1, not unique enough\n",
      "-skipping training candidate: rna9571.p1, not unique enough\n",
      "-skipping training candidate: rna50990.p1, not unique enough\n",
      "-skipping training candidate: rna3889.p1, not unique enough\n",
      "-skipping training candidate: rna9576.p1, not unique enough\n",
      "-skipping training candidate: rna28654.p1, not unique enough\n",
      "-skipping training candidate: rna33888.p1, not unique enough\n",
      "-skipping training candidate: rna33880.p1, not unique enough\n",
      "-skipping training candidate: rna58997.p1, not unique enough\n",
      "-skipping training candidate: rna1149.p1, not unique enough\n",
      "-skipping training candidate: rna7493.p1, not unique enough\n",
      "-skipping training candidate: rna28364.p1, not unique enough\n",
      "-skipping training candidate: rna29899.p1, not unique enough\n",
      "-skipping training candidate: rna36704.p1, not unique enough\n",
      "-skipping training candidate: rna37122.p1, not unique enough\n",
      "-skipping training candidate: rna37652.p1, not unique enough\n",
      "-skipping training candidate: rna40029.p1, not unique enough\n",
      "-skipping training candidate: rna47996.p1, not unique enough\n",
      "-skipping training candidate: rna50988.p1, not unique enough\n",
      "-skipping training candidate: rna1784.p1, not unique enough\n",
      "-skipping training candidate: rna768.p1, not unique enough\n",
      "-skipping training candidate: rna23647.p1, not unique enough\n",
      "-skipping training candidate: rna36221.p1, not unique enough\n",
      "-skipping training candidate: rna59225.p1, not unique enough\n",
      "-skipping training candidate: rna9093.p1, not unique enough\n",
      "-skipping training candidate: rna32868.p1, not unique enough\n",
      "-skipping training candidate: rna59224.p1, not unique enough\n",
      "-skipping training candidate: rna19019.p1, not unique enough\n",
      "-skipping training candidate: rna37651.p1, not unique enough\n",
      "-skipping training candidate: rna60483.p1, not unique enough\n",
      "-skipping training candidate: rna765.p1, not unique enough\n",
      "-skipping training candidate: rna221.p1, not unique enough\n",
      "-skipping training candidate: rna23645.p1, not unique enough\n",
      "-skipping training candidate: rna31008.p1, not unique enough\n",
      "-skipping training candidate: rna33887.p1, not unique enough\n",
      "-skipping training candidate: rna60136.p1, not unique enough\n",
      "-skipping training candidate: rna60134.p1, not unique enough\n",
      "-skipping training candidate: rna19015.p1, not unique enough\n",
      "-skipping training candidate: rna27896.p1, not unique enough\n",
      "-skipping training candidate: rna31005.p1, not unique enough\n",
      "-skipping training candidate: rna34516.p1, not unique enough\n",
      "-skipping training candidate: rna34515.p1, not unique enough\n",
      "-skipping training candidate: rna33882.p1, not unique enough\n",
      "-skipping training candidate: rna45115.p1, not unique enough\n",
      "-skipping training candidate: rna58451.p1, not unique enough\n",
      "-skipping training candidate: rna7146.p1, not unique enough\n",
      "-skipping training candidate: rna46705.p1, not unique enough\n",
      "-skipping training candidate: rna882.p1, not unique enough\n",
      "-skipping training candidate: rna7309.p1, not unique enough\n",
      "-skipping training candidate: rna23644.p1, not unique enough\n",
      "-skipping training candidate: rna50033.p1, not unique enough\n",
      "-skipping training candidate: rna64757.p1, not unique enough\n",
      "-skipping training candidate: rna19016.p1, not unique enough\n",
      "-skipping training candidate: rna19020.p1, not unique enough\n",
      "-skipping training candidate: rna21684.p1, not unique enough\n",
      "-skipping training candidate: rna31408.p1, not unique enough\n",
      "-skipping training candidate: rna31409.p1, not unique enough\n",
      "-skipping training candidate: rna767.p1, not unique enough\n",
      "-skipping training candidate: rna28655.p1, not unique enough\n",
      "-skipping training candidate: rna40878.p1, not unique enough\n",
      "-skipping training candidate: rna51467.p1, not unique enough\n",
      "-skipping training candidate: rna21679.p1, not unique enough\n",
      "-skipping training candidate: rna33885.p1, not unique enough\n",
      "-skipping training candidate: rna50034.p1, not unique enough\n",
      "-skipping training candidate: rna52591.p1, not unique enough\n",
      "-skipping training candidate: rna46265.p1, not unique enough\n",
      "-skipping training candidate: rna52592.p1, not unique enough\n",
      "-skipping training candidate: rna22345.p1, not unique enough\n",
      "-skipping training candidate: rna7139.p1, not unique enough\n",
      "-skipping training candidate: rna28961.p1, not unique enough\n",
      "-skipping training candidate: rna31006.p1, not unique enough\n",
      "-skipping training candidate: rna31009.p1, not unique enough\n",
      "-skipping training candidate: rna41810.p1, not unique enough\n",
      "-skipping training candidate: rna45790.p1, not unique enough\n",
      "-skipping training candidate: rna50833.p1, not unique enough\n",
      "-skipping training candidate: rna50836.p1, not unique enough\n",
      "-skipping training candidate: rna60135.p1, not unique enough\n",
      "-skipping training candidate: rna23643.p1, not unique enough\n",
      "-skipping training candidate: rna21678.p1, not unique enough\n",
      "-skipping training candidate: rna29476.p1, not unique enough\n",
      "-skipping training candidate: rna32865.p1, not unique enough\n",
      "-skipping training candidate: rna52648.p1, not unique enough\n",
      "-skipping training candidate: rna36217.p1, not unique enough\n",
      "-skipping training candidate: rna47065.p1, not unique enough\n",
      "-skipping training candidate: rna30395.p1, not unique enough\n",
      "-skipping training candidate: rna32867.p1, not unique enough\n",
      "-skipping training candidate: rna37301.p1, not unique enough\n",
      "-skipping training candidate: rna51199.p1, not unique enough\n",
      "-skipping training candidate: rna51368.p1, not unique enough\n",
      "-skipping training candidate: rna52649.p1, not unique enough\n",
      "-skipping training candidate: rna23642.p1, not unique enough\n",
      "-skipping training candidate: rna26736.p1, not unique enough\n",
      "-skipping training candidate: rna34271.p1, not unique enough\n",
      "-skipping training candidate: rna33883.p1, not unique enough\n",
      "-skipping training candidate: rna36218.p1, not unique enough\n",
      "-skipping training candidate: rna53416.p1, not unique enough\n",
      "-skipping training candidate: rna53415.p1, not unique enough\n",
      "-skipping training candidate: rna37255.p1, not unique enough\n",
      "-skipping training candidate: rna22344.p1, not unique enough\n",
      "-skipping training candidate: rna31010.p1, not unique enough\n",
      "-skipping training candidate: rna53804.p1, not unique enough\n",
      "-skipping training candidate: rna5962.p1, not unique enough\n",
      "-skipping training candidate: rna37120.p1, not unique enough\n",
      "-skipping training candidate: rna38071.p1, not unique enough\n",
      "-skipping training candidate: rna50032.p1, not unique enough\n",
      "-skipping training candidate: rna3888.p1, not unique enough\n",
      "-skipping training candidate: rna35870.p1, not unique enough\n",
      "-skipping training candidate: rna41958.p1, not unique enough\n",
      "-skipping training candidate: rna5401.p1, not unique enough\n",
      "-skipping training candidate: rna27012.p1, not unique enough\n",
      "-skipping training candidate: rna53802.p1, not unique enough\n",
      "-skipping training candidate: rna18813.p1, not unique enough\n",
      "-skipping training candidate: rna28653.p1, not unique enough\n",
      "-skipping training candidate: rna37303.p1, not unique enough\n",
      "-skipping training candidate: rna28652.p1, not unique enough\n",
      "-skipping training candidate: rna27899.p1, not unique enough\n",
      "-skipping training candidate: rna35117.p1, not unique enough\n",
      "-skipping training candidate: rna50991.p1, not unique enough\n",
      "-skipping training candidate: rna59054.p1, not unique enough\n",
      "-skipping training candidate: rna64393.p1, not unique enough\n",
      "-skipping training candidate: rna64389.p1, not unique enough\n",
      "-skipping training candidate: rna9040.p1, not unique enough\n",
      "-skipping training candidate: rna9039.p1, not unique enough\n",
      "-skipping training candidate: rna26042.p1, not unique enough\n",
      "-skipping training candidate: rna34091.p1, not unique enough\n",
      "-skipping training candidate: rna45949.p1, not unique enough\n",
      "-skipping training candidate: rna45950.p1, not unique enough\n",
      "-skipping training candidate: rna64394.p1, not unique enough\n",
      "-skipping training candidate: rna9575.p1, not unique enough\n",
      "-skipping training candidate: rna35132.p1, not unique enough\n",
      "-skipping training candidate: rna36222.p1, not unique enough\n",
      "-skipping training candidate: rna50035.p1, not unique enough\n",
      "-skipping training candidate: rna64407.p1, not unique enough\n",
      "-skipping training candidate: rna5402.p1, not unique enough\n",
      "-skipping training candidate: rna64403.p1, not unique enough\n",
      "-skipping training candidate: rna41751.p1, not unique enough\n",
      "-skipping training candidate: rna45122.p1, not unique enough\n",
      "-skipping training candidate: rna54238.p1, not unique enough\n",
      "-skipping training candidate: rna54239.p1, not unique enough\n",
      "-skipping training candidate: rna54242.p1, not unique enough\n",
      "-skipping training candidate: rna54237.p1, not unique enough\n",
      "-skipping training candidate: rna54241.p1, not unique enough\n",
      "-skipping training candidate: rna53805.p1, not unique enough\n",
      "-skipping training candidate: rna57161.p1, not unique enough\n",
      "-skipping training candidate: rna64391.p1, not unique enough\n",
      "-skipping training candidate: rna5955.p1, not unique enough\n",
      "-skipping training candidate: rna19134.p1, not unique enough\n",
      "-skipping training candidate: rna32869.p1, not unique enough\n",
      "-skipping training candidate: rna36223.p1, not unique enough\n",
      "-skipping training candidate: rna38588.p1, not unique enough\n",
      "-skipping training candidate: rna50036.p1, not unique enough\n",
      "-skipping training candidate: rna54236.p1, not unique enough\n",
      "-skipping training candidate: rna59958.p1, not unique enough\n",
      "-skipping training candidate: rna64404.p1, not unique enough\n",
      "-skipping training candidate: rna35120.p1, not unique enough\n",
      "-skipping training candidate: rna51466.p1, not unique enough\n",
      "-skipping training candidate: rna5964.p1, not unique enough\n",
      "-skipping training candidate: rna26292.p1, not unique enough\n",
      "-skipping training candidate: rna34090.p1, not unique enough\n",
      "-skipping training candidate: rna41969.p1, not unique enough\n",
      "-skipping training candidate: rna44632.p1, not unique enough\n",
      "-skipping training candidate: rna44633.p1, not unique enough\n",
      "-skipping training candidate: rna1163.p1, not unique enough\n",
      "-skipping training candidate: rna35426.p1, not unique enough\n",
      "-skipping training candidate: rna49138.p1, not unique enough\n",
      "-skipping training candidate: rna49886.p1, not unique enough\n",
      "-skipping training candidate: rna64409.p1, not unique enough\n",
      "-skipping training candidate: rna1157.p1, not unique enough\n",
      "-skipping training candidate: rna9572.p1, not unique enough\n",
      "-skipping training candidate: rna44635.p1, not unique enough\n",
      "-skipping training candidate: rna32735.p1, not unique enough\n",
      "-skipping training candidate: rna42175.p1, not unique enough\n",
      "-skipping training candidate: rna41811.p1, not unique enough\n",
      "-skipping training candidate: rna62160.p1, not unique enough\n",
      "-skipping training candidate: rna8239.p1, not unique enough\n",
      "-skipping training candidate: rna23363.p1, not unique enough\n",
      "-skipping training candidate: rna28657.p1, not unique enough\n",
      "-skipping training candidate: rna61499.p1, not unique enough\n",
      "-skipping training candidate: rna1165.p1, not unique enough\n",
      "-skipping training candidate: rna44634.p1, not unique enough\n",
      "-skipping training candidate: rna49199.p1, not unique enough\n",
      "-skipping training candidate: rna60484.p1, not unique enough\n",
      "-skipping training candidate: rna64411.p1, not unique enough\n",
      "-skipping training candidate: rna1156.p1, not unique enough\n",
      "-skipping training candidate: rna5398.p1, not unique enough\n",
      "-skipping training candidate: rna21238.p1, not unique enough\n",
      "-skipping training candidate: rna22307.p1, not unique enough\n",
      "-skipping training candidate: rna44631.p1, not unique enough\n",
      "-skipping training candidate: rna64388.p1, not unique enough\n",
      "-skipping training candidate: rna45980.p1, not unique enough\n",
      "-skipping training candidate: rna63133.p1, not unique enough\n",
      "-skipping training candidate: rna64390.p1, not unique enough\n",
      "-skipping training candidate: rna30775.p1, not unique enough\n",
      "-skipping training candidate: rna34094.p1, not unique enough\n",
      "-skipping training candidate: rna53803.p1, not unique enough\n",
      "-skipping training candidate: rna64395.p1, not unique enough\n",
      "-skipping training candidate: rna37302.p1, not unique enough\n",
      "-skipping training candidate: rna40547.p1, not unique enough\n",
      "-skipping training candidate: rna44630.p1, not unique enough\n",
      "-skipping training candidate: rna20449.p1, not unique enough\n",
      "-skipping training candidate: rna20448.p1, not unique enough\n",
      "-skipping training candidate: rna26290.p1, not unique enough\n",
      "-skipping training candidate: rna44629.p1, not unique enough\n",
      "-skipping training candidate: rna48385.p1, not unique enough\n",
      "-skipping training candidate: rna55521.p1, not unique enough\n",
      "-skipping training candidate: rna61101.p1, not unique enough\n",
      "-skipping training candidate: rna40246.p1, not unique enough\n",
      "-skipping training candidate: rna45123.p1, not unique enough\n",
      "-skipping training candidate: rna51369.p1, not unique enough\n",
      "-skipping training candidate: rna54924.p1, not unique enough\n",
      "-skipping training candidate: rna64392.p1, not unique enough\n",
      "-skipping training candidate: rna30856.p1, not unique enough\n",
      "-skipping training candidate: rna31753.p1, not unique enough\n",
      "-skipping training candidate: rna41960.p1, not unique enough\n",
      "-skipping training candidate: rna56836.p1, not unique enough\n",
      "-skipping training candidate: rna56837.p1, not unique enough\n",
      "-skipping training candidate: rna64396.p1, not unique enough\n",
      "-skipping training candidate: rna41961.p1, not unique enough\n",
      "-skipping training candidate: rna58996.p1, not unique enough\n",
      "-skipping training candidate: rna1158.p1, not unique enough\n",
      "-skipping training candidate: rna1160.p1, not unique enough\n",
      "-skipping training candidate: rna7308.p1, not unique enough\n",
      "-skipping training candidate: rna7307.p1, not unique enough\n",
      "-skipping training candidate: rna23086.p1, not unique enough\n",
      "-skipping training candidate: rna28764.p1, not unique enough\n",
      "-skipping training candidate: rna40549.p1, not unique enough\n",
      "-skipping training candidate: rna51100.p1, not unique enough\n",
      "-skipping training candidate: rna51098.p1, not unique enough\n",
      "-skipping training candidate: rna52593.p1, not unique enough\n",
      "-skipping training candidate: rna54927.p1, not unique enough\n",
      "-skipping training candidate: rna60617.p1, not unique enough\n",
      "-skipping training candidate: rna1159.p1, not unique enough\n",
      "-skipping training candidate: rna7138.p1, not unique enough\n",
      "-skipping training candidate: rna31386.p1, not unique enough\n",
      "-skipping training candidate: rna34093.p1, not unique enough\n",
      "-skipping training candidate: rna33877.p1, not unique enough\n",
      "-skipping training candidate: rna44637.p1, not unique enough\n",
      "-skipping training candidate: rna54926.p1, not unique enough\n",
      "-skipping training candidate: rna18879.p1, not unique enough\n",
      "-skipping training candidate: rna26544.p1, not unique enough\n",
      "-skipping training candidate: rna27897.p1, not unique enough\n",
      "-skipping training candidate: rna44628.p1, not unique enough\n",
      "-skipping training candidate: rna21743.p1, not unique enough\n",
      "-skipping training candidate: rna32734.p1, not unique enough\n",
      "-skipping training candidate: rna21326.p1, not unique enough\n",
      "-skipping training candidate: rna18640.p1, not unique enough\n",
      "-skipping training candidate: rna26291.p1, not unique enough\n",
      "-skipping training candidate: rna37304.p1, not unique enough\n",
      "-skipping training candidate: rna46639.p1, not unique enough\n",
      "-skipping training candidate: rna4120.p1, not unique enough\n",
      "-skipping training candidate: rna26043.p1, not unique enough\n",
      "-skipping training candidate: rna35221.p1, not unique enough\n",
      "-skipping training candidate: rna45116.p1, not unique enough\n",
      "-skipping training candidate: rna44627.p1, not unique enough\n",
      "-skipping training candidate: rna6108.p1, not unique enough\n",
      "-skipping training candidate: rna4595.p1, not unique enough\n",
      "-skipping training candidate: rna9875.p1, not unique enough\n",
      "-skipping training candidate: rna19133.p1, not unique enough\n",
      "-skipping training candidate: rna26285.p1, not unique enough\n",
      "-skipping training candidate: rna51526.p1, not unique enough\n",
      "-skipping training candidate: rna64412.p1, not unique enough\n",
      "-skipping training candidate: rna4293.p1, not unique enough\n",
      "-skipping training candidate: rna9874.p1, not unique enough\n",
      "-skipping training candidate: rna59261.p1, not unique enough\n",
      "-skipping training candidate: rna63135.p1, not unique enough\n",
      "-skipping training candidate: rna4294.p1, not unique enough\n",
      "-skipping training candidate: rna49015.p1, not unique enough\n",
      "-skipping training candidate: rna52651.p1, not unique enough\n",
      "-skipping training candidate: rna7492.p1, not unique enough\n",
      "-skipping training candidate: rna35220.p1, not unique enough\n",
      "-skipping training candidate: rna45303.p1, not unique enough\n",
      "-skipping training candidate: rna45953.p1, not unique enough\n",
      "-skipping training candidate: rna21175.p1, not unique enough\n",
      "-skipping training candidate: rna53760.p1, not unique enough\n",
      "-skipping training candidate: rna64413.p1, not unique enough\n",
      "-skipping training candidate: rna5075.p1, not unique enough\n",
      "-skipping training candidate: rna32823.p1, not unique enough\n",
      "-skipping training candidate: rna51099.p1, not unique enough\n",
      "-skipping training candidate: rna20947.p1, not unique enough\n",
      "-skipping training candidate: rna20951.p1, not unique enough\n",
      "-skipping training candidate: rna27783.p1, not unique enough\n",
      "-skipping training candidate: rna30914.p1, not unique enough\n",
      "-skipping training candidate: rna33884.p1, not unique enough\n",
      "-skipping training candidate: rna32822.p1, not unique enough\n",
      "-skipping training candidate: rna48387.p1, not unique enough\n",
      "-skipping training candidate: rna7306.p1, not unique enough\n",
      "-skipping training candidate: rna7305.p1, not unique enough\n",
      "-skipping training candidate: rna7963.p1, not unique enough\n",
      "-skipping training candidate: rna7961.p1, not unique enough\n",
      "-skipping training candidate: rna7788.p1, not unique enough\n",
      "-skipping training candidate: rna7787.p1, not unique enough\n",
      "-skipping training candidate: rna9636.p1, not unique enough\n",
      "-skipping training candidate: rna32820.p1, not unique enough\n",
      "-skipping training candidate: rna1167.p1, not unique enough\n",
      "-skipping training candidate: rna9574.p1, not unique enough\n",
      "-skipping training candidate: rna31065.p1, not unique enough\n",
      "-skipping training candidate: rna61099.p1, not unique enough\n",
      "-skipping training candidate: rna4560.p1, not unique enough\n",
      "-skipping training candidate: rna19524.p1, not unique enough\n",
      "-skipping training candidate: rna27802.p1, not unique enough\n",
      "-skipping training candidate: rna27803.p1, not unique enough\n",
      "-skipping training candidate: rna34322.p1, not unique enough\n",
      "-skipping training candidate: rna582.p1, not unique enough\n",
      "-skipping training candidate: rna1164.p1, not unique enough\n",
      "-skipping training candidate: rna9639.p1, not unique enough\n",
      "-skipping training candidate: rna9637.p1, not unique enough\n",
      "-skipping training candidate: rna33886.p1, not unique enough\n",
      "-skipping training candidate: rna20948.p1, not unique enough\n",
      "-skipping training candidate: rna34325.p1, not unique enough\n",
      "-skipping training candidate: rna57556.p1, not unique enough\n",
      "-skipping training candidate: rna1302.p1, not unique enough\n",
      "-skipping training candidate: rna1830.p1, not unique enough\n",
      "-skipping training candidate: rna39755.p1, not unique enough\n",
      "-skipping training candidate: rna1828.p1, not unique enough\n",
      "-skipping training candidate: rna3665.p1, not unique enough\n",
      "-skipping training candidate: rna45121.p1, not unique enough\n",
      "-skipping training candidate: rna62408.p1, not unique enough\n",
      "-skipping training candidate: rna61805.p1, not unique enough\n",
      "-skipping training candidate: rna61804.p1, not unique enough\n",
      "-skipping training candidate: rna1121.p1, not unique enough\n",
      "-skipping training candidate: rna18880.p1, not unique enough\n",
      "-skipping training candidate: rna34323.p1, not unique enough\n",
      "-skipping training candidate: rna36553.p1, not unique enough\n",
      "-skipping training candidate: rna52024.p1, not unique enough\n",
      "-skipping training candidate: rna52025.p1, not unique enough\n",
      "-skipping training candidate: rna57555.p1, not unique enough\n",
      "-skipping training candidate: rna1301.p1, not unique enough\n",
      "-skipping training candidate: rna35998.p1, not unique enough\n",
      "-skipping training candidate: rna38326.p1, not unique enough\n",
      "-skipping training candidate: rna42115.p1, not unique enough\n",
      "-skipping training candidate: rna26286.p1, not unique enough\n",
      "-skipping training candidate: rna1829.p1, not unique enough\n",
      "-skipping training candidate: rna32733.p1, not unique enough\n",
      "-skipping training candidate: rna34326.p1, not unique enough\n",
      "-skipping training candidate: rna58797.p1, not unique enough\n",
      "-skipping training candidate: rna61802.p1, not unique enough\n",
      "-skipping training candidate: rna4596.p1, not unique enough\n",
      "-skipping training candidate: rna30855.p1, not unique enough\n",
      "-skipping training candidate: rna33001.p1, not unique enough\n",
      "-skipping training candidate: rna40548.p1, not unique enough\n",
      "-skipping training candidate: rna57558.p1, not unique enough\n",
      "-skipping training candidate: rna48270.p1, not unique enough\n",
      "-skipping training candidate: rna56811.p1, not unique enough\n",
      "-skipping training candidate: rna48225.p1, not unique enough\n",
      "-skipping training candidate: rna57554.p1, not unique enough\n",
      "-skipping training candidate: rna9635.p1, not unique enough\n",
      "-skipping training candidate: rna23530.p1, not unique enough\n",
      "-skipping training candidate: rna45125.p1, not unique enough\n",
      "-skipping training candidate: rna52646.p1, not unique enough\n",
      "-skipping training candidate: rna52653.p1, not unique enough\n",
      "-skipping training candidate: rna41608.p1, not unique enough\n",
      "-skipping training candidate: rna50560.p1, not unique enough\n",
      "-skipping training candidate: rna4712.p1, not unique enough\n",
      "-skipping training candidate: rna45124.p1, not unique enough\n",
      "-skipping training candidate: rna61742.p1, not unique enough\n",
      "-skipping training candidate: rna4710.p1, not unique enough\n",
      "-skipping training candidate: rna8528.p1, not unique enough\n",
      "-skipping training candidate: rna61102.p1, not unique enough\n",
      "-skipping training candidate: rna26288.p1, not unique enough\n",
      "-skipping training candidate: rna34327.p1, not unique enough\n",
      "-skipping training candidate: rna881.p1, not unique enough\n",
      "-skipping training candidate: rna51030.p1, not unique enough\n",
      "-skipping training candidate: rna28960.p1, not unique enough\n",
      "-skipping training candidate: rna28959.p1, not unique enough\n",
      "-skipping training candidate: rna35427.p1, not unique enough\n",
      "-skipping training candidate: rna50559.p1, not unique enough\n",
      "-skipping training candidate: rna61806.p1, not unique enough\n",
      "-skipping training candidate: rna8531.p1, not unique enough\n",
      "-skipping training candidate: rna32825.p1, not unique enough\n",
      "-skipping training candidate: rna40044.p1, not unique enough\n",
      "-skipping training candidate: rna41754.p1, not unique enough\n",
      "-skipping training candidate: rna47556.p1, not unique enough\n",
      "-skipping training candidate: rna8529.p1, not unique enough\n",
      "-skipping training candidate: rna52028.p1, not unique enough\n",
      "-skipping training candidate: rna3668.p1, not unique enough\n",
      "-skipping training candidate: rna21264.p1, not unique enough\n",
      "-skipping training candidate: rna19267.p1, not unique enough\n",
      "-skipping training candidate: rna31407.p1, not unique enough\n",
      "-skipping training candidate: rna38539.p1, not unique enough\n",
      "-skipping training candidate: rna18881.p1, not unique enough\n",
      "-skipping training candidate: rna32821.p1, not unique enough\n",
      "-skipping training candidate: rna33003.p1, not unique enough\n",
      "-skipping training candidate: rna33002.p1, not unique enough\n",
      "-skipping training candidate: rna41607.p1, not unique enough\n",
      "-skipping training candidate: rna1161.p1, not unique enough\n",
      "-skipping training candidate: rna1162.p1, not unique enough\n",
      "-skipping training candidate: rna1467.p1, not unique enough\n",
      "-skipping training candidate: rna20949.p1, not unique enough\n",
      "-skipping training candidate: rna47997.p1, not unique enough\n",
      "-skipping training candidate: rna53154.p1, not unique enough\n",
      "-skipping training candidate: rna51334.p1, not unique enough\n",
      "-skipping training candidate: rna53713.p1, not unique enough\n",
      "-skipping training candidate: rna53716.p1, not unique enough\n",
      "-skipping training candidate: rna32435.p1, not unique enough\n",
      "-skipping training candidate: rna32434.p1, not unique enough\n",
      "-skipping training candidate: rna34328.p1, not unique enough\n",
      "-skipping training candidate: rna51333.p1, not unique enough\n",
      "-skipping training candidate: rna19525.p1, not unique enough\n",
      "-skipping training candidate: rna34445.p1, not unique enough\n",
      "-skipping training candidate: rna30212.p1, not unique enough\n",
      "-skipping training candidate: rna48268.p1, not unique enough\n",
      "-skipping training candidate: rna52026.p1, not unique enough\n",
      "-skipping training candidate: rna61594.p1, not unique enough\n",
      "-skipping training candidate: rna28368.p1, not unique enough\n",
      "-skipping training candidate: rna30381.p1, not unique enough\n",
      "-skipping training candidate: rna35995.p1, not unique enough\n",
      "-skipping training candidate: rna52027.p1, not unique enough\n",
      "-skipping training candidate: rna61800.p1, not unique enough\n",
      "-skipping training candidate: rna32144.p1, not unique enough\n",
      "-skipping training candidate: rna62528.p1, not unique enough\n",
      "-skipping training candidate: rna31679.p1, not unique enough\n",
      "-skipping training candidate: rna37942.p1, not unique enough\n",
      "-skipping training candidate: rna44020.p1, not unique enough\n",
      "-skipping training candidate: rna4564.p1, not unique enough\n",
      "-skipping training candidate: rna23324.p1, not unique enough\n",
      "-skipping training candidate: rna37940.p1, not unique enough\n",
      "-skipping training candidate: rna43409.p1, not unique enough\n",
      "-skipping training candidate: rna51332.p1, not unique enough\n",
      "-skipping training candidate: rna57163.p1, not unique enough\n",
      "-skipping training candidate: rna18151.p1, not unique enough\n",
      "-skipping training candidate: rna23328.p1, not unique enough\n",
      "-skipping training candidate: rna30916.p1, not unique enough\n",
      "-skipping training candidate: rna34450.p1, not unique enough\n",
      "-skipping training candidate: rna38839.p1, not unique enough\n",
      "-skipping training candidate: rna19268.p1, not unique enough\n",
      "-skipping training candidate: rna18152.p1, not unique enough\n",
      "-skipping training candidate: rna23088.p1, not unique enough\n",
      "-skipping training candidate: rna28957.p1, not unique enough\n",
      "-skipping training candidate: rna41823.p1, not unique enough\n",
      "-skipping training candidate: rna41818.p1, not unique enough\n",
      "-skipping training candidate: rna41822.p1, not unique enough\n",
      "-skipping training candidate: rna41819.p1, not unique enough\n",
      "-skipping training candidate: rna50834.p1, not unique enough\n",
      "-skipping training candidate: rna49313.p1, not unique enough\n",
      "-skipping training candidate: rna49309.p1, not unique enough\n",
      "-skipping training candidate: rna49312.p1, not unique enough\n",
      "-skipping training candidate: rna49310.p1, not unique enough\n",
      "-skipping training candidate: rna49314.p1, not unique enough\n",
      "-skipping training candidate: rna4318.p1, not unique enough\n",
      "-skipping training candidate: rna40092.p1, not unique enough\n",
      "-skipping training candidate: rna41817.p1, not unique enough\n",
      "-skipping training candidate: rna41821.p1, not unique enough\n",
      "-skipping training candidate: rna46439.p1, not unique enough\n",
      "-skipping training candidate: rna51307.p1, not unique enough\n",
      "-skipping training candidate: rna8527.p1, not unique enough\n",
      "-skipping training candidate: rna57331.p1, not unique enough\n",
      "-skipping training candidate: rna8526.p1, not unique enough\n",
      "-skipping training candidate: rna23326.p1, not unique enough\n",
      "-skipping training candidate: rna31519.p1, not unique enough\n",
      "-skipping training candidate: rna34447.p1, not unique enough\n",
      "-skipping training candidate: rna23325.p1, not unique enough\n",
      "-skipping training candidate: rna24687.p1, not unique enough\n",
      "-skipping training candidate: rna24688.p1, not unique enough\n",
      "-skipping training candidate: rna31614.p1, not unique enough\n",
      "-skipping training candidate: rna33004.p1, not unique enough\n",
      "-skipping training candidate: rna50837.p1, not unique enough\n",
      "-skipping training candidate: rna34330.p1, not unique enough\n",
      "-skipping training candidate: rna34453.p1, not unique enough\n",
      "-skipping training candidate: rna43411.p1, not unique enough\n",
      "-skipping training candidate: rna18882.p1, not unique enough\n",
      "-skipping training candidate: rna25695.p1, not unique enough\n",
      "-skipping training candidate: rna60129.p1, not unique enough\n",
      "-skipping training candidate: rna44776.p1, not unique enough\n",
      "-skipping training candidate: rna51305.p1, not unique enough\n",
      "-skipping training candidate: rna26289.p1, not unique enough\n",
      "-skipping training candidate: rna43412.p1, not unique enough\n",
      "-skipping training candidate: rna48269.p1, not unique enough\n",
      "-skipping training candidate: rna48085.p1, not unique enough\n",
      "-skipping training candidate: rna39860.p1, not unique enough\n",
      "-skipping training candidate: rna52595.p1, not unique enough\n",
      "-skipping training candidate: rna22261.p1, not unique enough\n",
      "-skipping training candidate: rna24527.p1, not unique enough\n",
      "-skipping training candidate: rna32597.p1, not unique enough\n",
      "-skipping training candidate: rna37941.p1, not unique enough\n",
      "-skipping training candidate: rna37939.p1, not unique enough\n",
      "-skipping training candidate: rna60046.p1, not unique enough\n",
      "-skipping training candidate: rna61002.p1, not unique enough\n",
      "-skipping training candidate: rna5957.p1, not unique enough\n",
      "-skipping training candidate: rna32598.p1, not unique enough\n",
      "-skipping training candidate: rna57637.p1, not unique enough\n",
      "-skipping training candidate: rna28955.p1, not unique enough\n",
      "-skipping training candidate: rna34648.p1, not unique enough\n",
      "-skipping training candidate: rna37709.p1, not unique enough\n",
      "-skipping training candidate: rna51306.p1, not unique enough\n",
      "-skipping training candidate: rna61003.p1, not unique enough\n",
      "-skipping training candidate: rna31012.p1, not unique enough\n",
      "-skipping training candidate: rna50838.p1, not unique enough\n",
      "-skipping training candidate: rna769.p1, not unique enough\n",
      "-skipping training candidate: rna20799.p1, not unique enough\n",
      "-skipping training candidate: rna20798.p1, not unique enough\n",
      "-skipping training candidate: rna31411.p1, not unique enough\n",
      "-skipping training candidate: rna32146.p1, not unique enough\n",
      "-skipping training candidate: rna23327.p1, not unique enough\n",
      "-skipping training candidate: rna25195.p1, not unique enough\n",
      "-skipping training candidate: rna60044.p1, not unique enough\n",
      "-skipping training candidate: rna55803.p1, not unique enough\n",
      "-skipping training candidate: rna57634.p1, not unique enough\n",
      "-skipping training candidate: rna23369.p1, not unique enough\n",
      "-skipping training candidate: rna51308.p1, not unique enough\n",
      "-skipping training candidate: rna34449.p1, not unique enough\n",
      "-skipping training candidate: rna43413.p1, not unique enough\n",
      "-skipping training candidate: rna64397.p1, not unique enough\n",
      "-skipping training candidate: rna28953.p1, not unique enough\n",
      "-skipping training candidate: rna43410.p1, not unique enough\n",
      "-skipping training candidate: rna60119.p1, not unique enough\n",
      "-skipping training candidate: rna4222.p1, not unique enough\n",
      "-skipping training candidate: rna20647.p1, not unique enough\n",
      "-skipping training candidate: rna43782.p1, not unique enough\n",
      "-skipping training candidate: rna37291.p1, not unique enough\n",
      "-skipping training candidate: rna48084.p1, not unique enough\n",
      "-skipping training candidate: rna61801.p1, not unique enough\n",
      "-skipping training candidate: rna4608.p1, not unique enough\n",
      "-skipping training candidate: rna51304.p1, not unique enough\n",
      "-skipping training candidate: rna23370.p1, not unique enough\n",
      "-skipping training candidate: rna22259.p1, not unique enough\n",
      "-skipping training candidate: rna42176.p1, not unique enough\n",
      "-skipping training candidate: rna51742.p1, not unique enough\n",
      "-skipping training candidate: rna21687.p1, not unique enough\n",
      "-skipping training candidate: rna25196.p1, not unique enough\n",
      "-skipping training candidate: rna42148.p1, not unique enough\n",
      "-skipping training candidate: rna55804.p1, not unique enough\n",
      "-skipping training candidate: rna23330.p1, not unique enough\n",
      "-skipping training candidate: rna41959.p1, not unique enough\n",
      "-skipping training candidate: rna51372.p1, not unique enough\n",
      "-skipping training candidate: rna59094.p1, not unique enough\n",
      "-skipping training candidate: rna7235.p1, not unique enough\n",
      "-skipping training candidate: rna51371.p1, not unique enough\n",
      "-skipping training candidate: rna59314.p1, not unique enough\n",
      "-skipping training candidate: rna4807.p1, not unique enough\n",
      "-skipping training candidate: rna4607.p1, not unique enough\n",
      "-skipping training candidate: rna29928.p1, not unique enough\n",
      "-skipping training candidate: rna36587.p1, not unique enough\n",
      "-skipping training candidate: rna780.p1, not unique enough\n",
      "-skipping training candidate: rna18157.p1, not unique enough\n",
      "-skipping training candidate: rna42057.p1, not unique enough\n",
      "-skipping training candidate: rna46472.p1, not unique enough\n",
      "-skipping training candidate: rna49256.p1, not unique enough\n",
      "-skipping training candidate: rna23332.p1, not unique enough\n",
      "-skipping training candidate: rna51741.p1, not unique enough\n",
      "-skipping training candidate: rna1677.p1, not unique enough\n",
      "-skipping training candidate: rna4118.p1, not unique enough\n",
      "-skipping training candidate: rna18156.p1, not unique enough\n",
      "-skipping training candidate: rna8538.p1, not unique enough\n",
      "-skipping training candidate: rna38908.p1, not unique enough\n",
      "-skipping training candidate: rna61179.p1, not unique enough\n",
      "-skipping training candidate: rna1675.p1, not unique enough\n",
      "-skipping training candidate: rna34703.p1, not unique enough\n",
      "-skipping training candidate: rna60762.p1, not unique enough\n",
      "-skipping training candidate: rna64414.p1, not unique enough\n",
      "-skipping training candidate: rna25194.p1, not unique enough\n",
      "-skipping training candidate: rna47066.p1, not unique enough\n",
      "-skipping training candidate: rna19230.p1, not unique enough\n",
      "-skipping training candidate: rna42102.p1, not unique enough\n",
      "-skipping training candidate: rna44746.p1, not unique enough\n",
      "-skipping training candidate: rna50562.p1, not unique enough\n",
      "-skipping training candidate: rna54928.p1, not unique enough\n",
      "-skipping training candidate: rna25200.p1, not unique enough\n",
      "-skipping training candidate: rna60425.p1, not unique enough\n",
      "-skipping training candidate: rna25110.p1, not unique enough\n",
      "-skipping training candidate: rna25109.p1, not unique enough\n",
      "-skipping training candidate: rna34332.p1, not unique enough\n",
      "-skipping training candidate: rna38911.p1, not unique enough\n",
      "-skipping training candidate: rna38914.p1, not unique enough\n",
      "-skipping training candidate: rna42058.p1, not unique enough\n",
      "-skipping training candidate: rna56974.p1, not unique enough\n",
      "-skipping training candidate: rna60424.p1, not unique enough\n",
      "-skipping training candidate: rna3221.p1, not unique enough\n",
      "-skipping training candidate: rna38912.p1, not unique enough\n",
      "-skipping training candidate: rna60426.p1, not unique enough\n",
      "-skipping training candidate: rna23331.p1, not unique enough\n",
      "-skipping training candidate: rna31458.p1, not unique enough\n",
      "-skipping training candidate: rna60384.p1, not unique enough\n",
      "-skipping training candidate: rna50996.p1, not unique enough\n",
      "-skipping training candidate: rna62527.p1, not unique enough\n",
      "-skipping training candidate: rna31792.p1, not unique enough\n",
      "-skipping training candidate: rna51265.p1, not unique enough\n",
      "-skipping training candidate: rna61054.p1, not unique enough\n",
      "-skipping training candidate: rna34331.p1, not unique enough\n",
      "-skipping training candidate: rna38718.p1, not unique enough\n",
      "-skipping training candidate: rna54254.p1, not unique enough\n",
      "-skipping training candidate: rna59436.p1, not unique enough\n",
      "-skipping training candidate: rna63577.p1, not unique enough\n",
      "-skipping training candidate: rna3220.p1, not unique enough\n",
      "-skipping training candidate: rna3666.p1, not unique enough\n",
      "-skipping training candidate: rna45856.p1, not unique enough\n",
      "-skipping training candidate: rna50995.p1, not unique enough\n",
      "-skipping training candidate: rna61053.p1, not unique enough\n",
      "-skipping training candidate: rna23333.p1, not unique enough\n",
      "-skipping training candidate: rna24543.p1, not unique enough\n",
      "-skipping training candidate: rna29583.p1, not unique enough\n",
      "-skipping training candidate: rna50993.p1, not unique enough\n",
      "-skipping training candidate: rna51266.p1, not unique enough\n",
      "-skipping training candidate: rna59092.p1, not unique enough\n",
      "-skipping training candidate: rna2483.p1, not unique enough\n",
      "-skipping training candidate: rna32143.p1, not unique enough\n",
      "-skipping training candidate: rna23321.p1, not unique enough\n",
      "-skipping training candidate: rna50830.p1, not unique enough\n",
      "-skipping training candidate: rna61104.p1, not unique enough\n",
      "-skipping training candidate: rna4703.p1, not unique enough\n",
      "-skipping training candidate: rna19279.p1, not unique enough\n",
      "-skipping training candidate: rna43408.p1, not unique enough\n",
      "-skipping training candidate: rna3667.p1, not unique enough\n",
      "-skipping training candidate: rna25191.p1, not unique enough\n",
      "-skipping training candidate: rna31680.p1, not unique enough\n",
      "-skipping training candidate: rna34329.p1, not unique enough\n",
      "-skipping training candidate: rna48990.p1, not unique enough\n",
      "-skipping training candidate: rna49315.p1, not unique enough\n",
      "-skipping training candidate: rna50563.p1, not unique enough\n",
      "-skipping training candidate: rna4702.p1, not unique enough\n",
      "-skipping training candidate: rna9349.p1, not unique enough\n",
      "-skipping training candidate: rna9338.p1, not unique enough\n",
      "-skipping training candidate: rna26242.p1, not unique enough\n",
      "-skipping training candidate: rna39861.p1, not unique enough\n",
      "-skipping training candidate: rna4701.p1, not unique enough\n",
      "-skipping training candidate: rna29141.p1, not unique enough\n",
      "-skipping training candidate: rna36289.p1, not unique enough\n",
      "-skipping training candidate: rna40091.p1, not unique enough\n",
      "-skipping training candidate: rna57541.p1, not unique enough\n",
      "-skipping training candidate: rna64415.p1, not unique enough\n",
      "-skipping training candidate: rna842.p1, not unique enough\n",
      "-skipping training candidate: rna8449.p1, not unique enough\n",
      "-skipping training candidate: rna24459.p1, not unique enough\n",
      "-skipping training candidate: rna29611.p1, not unique enough\n",
      "-skipping training candidate: rna29610.p1, not unique enough\n",
      "-skipping training candidate: rna29926.p1, not unique enough\n",
      "-skipping training candidate: rna36288.p1, not unique enough\n",
      "-skipping training candidate: rna41606.p1, not unique enough\n",
      "-skipping training candidate: rna48993.p1, not unique enough\n",
      "-skipping training candidate: rna51097.p1, not unique enough\n",
      "-skipping training candidate: rna57537.p1, not unique enough\n",
      "-skipping training candidate: rna57544.p1, not unique enough\n",
      "-skipping training candidate: rna61100.p1, not unique enough\n",
      "-skipping training candidate: rna8451.p1, not unique enough\n",
      "-skipping training candidate: rna8450.p1, not unique enough\n",
      "-skipping training candidate: rna8541.p1, not unique enough\n",
      "-skipping training candidate: rna57539.p1, not unique enough\n",
      "-skipping training candidate: rna6176.p1, not unique enough\n",
      "-skipping training candidate: rna25192.p1, not unique enough\n",
      "-skipping training candidate: rna29143.p1, not unique enough\n",
      "-skipping training candidate: rna50866.p1, not unique enough\n",
      "-skipping training candidate: rna1169.p1, not unique enough\n",
      "-skipping training candidate: rna3119.p1, not unique enough\n",
      "-skipping training candidate: rna9346.p1, not unique enough\n",
      "-skipping training candidate: rna38909.p1, not unique enough\n",
      "-skipping training candidate: rna59090.p1, not unique enough\n",
      "-skipping training candidate: rna59089.p1, not unique enough\n",
      "-skipping training candidate: rna61005.p1, not unique enough\n",
      "-skipping training candidate: rna62370.p1, not unique enough\n",
      "-skipping training candidate: rna63134.p1, not unique enough\n",
      "-skipping training candidate: rna25197.p1, not unique enough\n",
      "-skipping training candidate: rna29142.p1, not unique enough\n",
      "-skipping training candidate: rna29581.p1, not unique enough\n",
      "-skipping training candidate: rna38910.p1, not unique enough\n",
      "-skipping training candidate: rna49317.p1, not unique enough\n",
      "-skipping training candidate: rna51469.p1, not unique enough\n",
      "-skipping training candidate: rna57561.p1, not unique enough\n",
      "-skipping training candidate: rna6172.p1, not unique enough\n",
      "-skipping training candidate: rna23322.p1, not unique enough\n",
      "-skipping training candidate: rna60715.p1, not unique enough\n",
      "-skipping training candidate: rna60716.p1, not unique enough\n",
      "-skipping training candidate: rna1168.p1, not unique enough\n",
      "-skipping training candidate: rna2482.p1, not unique enough\n",
      "-skipping training candidate: rna6169.p1, not unique enough\n",
      "-skipping training candidate: rna8536.p1, not unique enough\n",
      "-skipping training candidate: rna8539.p1, not unique enough\n",
      "-skipping training candidate: rna25804.p1, not unique enough\n",
      "-skipping training candidate: rna45119.p1, not unique enough\n",
      "-skipping training candidate: rna16440.p1, not unique enough\n",
      "-skipping training candidate: rna9343.p1, not unique enough\n",
      "-skipping training candidate: rna8535.p1, not unique enough\n",
      "-skipping training candidate: rna23334.p1, not unique enough\n",
      "-skipping training candidate: rna34704.p1, not unique enough\n",
      "-skipping training candidate: rna46440.p1, not unique enough\n",
      "-skipping training candidate: rna48991.p1, not unique enough\n",
      "-skipping training candidate: rna51740.p1, not unique enough\n",
      "-skipping training candidate: rna4809.p1, not unique enough\n",
      "-skipping training candidate: rna8534.p1, not unique enough\n",
      "-skipping training candidate: rna29144.p1, not unique enough\n",
      "-skipping training candidate: rna34451.p1, not unique enough\n",
      "-skipping training candidate: rna41605.p1, not unique enough\n",
      "-skipping training candidate: rna48992.p1, not unique enough\n",
      "-skipping training candidate: rna57543.p1, not unique enough\n",
      "-skipping training candidate: rna3943.p1, not unique enough\n",
      "-skipping training candidate: rna9340.p1, not unique enough\n",
      "-skipping training candidate: rna9877.p1, not unique enough\n",
      "-skipping training candidate: rna28796.p1, not unique enough\n",
      "-skipping training candidate: rna29145.p1, not unique enough\n",
      "-skipping training candidate: rna36554.p1, not unique enough\n",
      "-skipping training candidate: rna52846.p1, not unique enough\n",
      "-skipping training candidate: rna5951.p1, not unique enough\n",
      "-skipping training candidate: rna9516.p1, not unique enough\n",
      "-skipping training candidate: rna9876.p1, not unique enough\n",
      "-skipping training candidate: rna21894.p1, not unique enough\n",
      "-skipping training candidate: rna23928.p1, not unique enough\n",
      "-skipping training candidate: rna38913.p1, not unique enough\n",
      "-skipping training candidate: rna49227.p1, not unique enough\n",
      "-skipping training candidate: rna57540.p1, not unique enough\n",
      "-skipping training candidate: rna6173.p1, not unique enough\n",
      "-skipping training candidate: rna25202.p1, not unique enough\n",
      "-skipping training candidate: rna6175.p1, not unique enough\n",
      "-skipping training candidate: rna5965.p1, not unique enough\n",
      "-skipping training candidate: rna8537.p1, not unique enough\n",
      "-skipping training candidate: rna38719.p1, not unique enough\n",
      "-skipping training candidate: rna40245.p1, not unique enough\n",
      "-skipping training candidate: rna63769.p1, not unique enough\n",
      "-skipping training candidate: rna63768.p1, not unique enough\n",
      "-skipping training candidate: rna1368.p1, not unique enough\n",
      "-skipping training candidate: rna6174.p1, not unique enough\n",
      "-skipping training candidate: rna20455.p1, not unique enough\n",
      "-skipping training candidate: rna32599.p1, not unique enough\n",
      "-skipping training candidate: rna46605.p1, not unique enough\n",
      "-skipping training candidate: rna1366.p1, not unique enough\n",
      "-skipping training candidate: rna3947.p1, not unique enough\n",
      "-skipping training candidate: rna25198.p1, not unique enough\n",
      "-skipping training candidate: rna29613.p1, not unique enough\n",
      "-skipping training candidate: rna62371.p1, not unique enough\n",
      "-skipping training candidate: rna63770.p1, not unique enough\n",
      "-skipping training candidate: rna6171.p1, not unique enough\n",
      "-skipping training candidate: rna25199.p1, not unique enough\n",
      "-skipping training candidate: rna59139.p1, not unique enough\n",
      "-skipping training candidate: rna4097.p1, not unique enough\n",
      "-skipping training candidate: rna9514.p1, not unique enough\n",
      "-skipping training candidate: rna21744.p1, not unique enough\n",
      "-skipping training candidate: rna34334.p1, not unique enough\n",
      "-skipping training candidate: rna9345.p1, not unique enough\n",
      "-skipping training candidate: rna3574.p1, not unique enough\n",
      "-skipping training candidate: rna3670.p1, not unique enough\n",
      "-skipping training candidate: rna31842.p1, not unique enough\n",
      "-skipping training candidate: rna32507.p1, not unique enough\n",
      "-skipping training candidate: rna50397.p1, not unique enough\n",
      "-skipping training candidate: rna60492.p1, not unique enough\n",
      "-skipping training candidate: rna61004.p1, not unique enough\n",
      "-skipping training candidate: rna1365.p1, not unique enough\n",
      "-skipping training candidate: rna18540.p1, not unique enough\n",
      "-skipping training candidate: rna32508.p1, not unique enough\n",
      "-skipping training candidate: rna37878.p1, not unique enough\n",
      "-skipping training candidate: rna48988.p1, not unique enough\n",
      "-skipping training candidate: rna57538.p1, not unique enough\n",
      "-skipping training candidate: rna3575.p1, not unique enough\n",
      "-skipping training candidate: rna31640.p1, not unique enough\n",
      "-skipping training candidate: rna37880.p1, not unique enough\n",
      "-skipping training candidate: rna1367.p1, not unique enough\n",
      "-skipping training candidate: rna19980.p1, not unique enough\n",
      "-skipping training candidate: rna40440.p1, not unique enough\n",
      "-skipping training candidate: rna44927.p1, not unique enough\n",
      "-skipping training candidate: rna53034.p1, not unique enough\n",
      "-skipping training candidate: rna54256.p1, not unique enough\n",
      "-skipping training candidate: rna61098.p1, not unique enough\n",
      "-skipping training candidate: rna18539.p1, not unique enough\n",
      "-skipping training candidate: rna26555.p1, not unique enough\n",
      "-skipping training candidate: rna58005.p1, not unique enough\n",
      "-skipping training candidate: rna62047.p1, not unique enough\n",
      "-skipping training candidate: rna8673.p1, not unique enough\n",
      "-skipping training candidate: rna34802.p1, not unique enough\n",
      "-skipping training candidate: rna53035.p1, not unique enough\n",
      "-skipping training candidate: rna58613.p1, not unique enough\n",
      "-skipping training candidate: rna58617.p1, not unique enough\n",
      "-skipping training candidate: rna62046.p1, not unique enough\n",
      "-skipping training candidate: rna3669.p1, not unique enough\n",
      "-skipping training candidate: rna9613.p1, not unique enough\n",
      "-skipping training candidate: rna58793.p1, not unique enough\n",
      "-skipping training candidate: rna58794.p1, not unique enough\n",
      "-skipping training candidate: rna62670.p1, not unique enough\n",
      "-skipping training candidate: rna1593.p1, not unique enough\n",
      "-skipping training candidate: rna49226.p1, not unique enough\n",
      "-skipping training candidate: rna62669.p1, not unique enough\n",
      "-skipping training candidate: rna19528.p1, not unique enough\n",
      "-skipping training candidate: rna19984.p1, not unique enough\n",
      "-skipping training candidate: rna23451.p1, not unique enough\n",
      "-skipping training candidate: rna36116.p1, not unique enough\n",
      "-skipping training candidate: rna37885.p1, not unique enough\n",
      "-skipping training candidate: rna44890.p1, not unique enough\n",
      "-skipping training candidate: rna54040.p1, not unique enough\n",
      "-skipping training candidate: rna54035.p1, not unique enough\n",
      "-skipping training candidate: rna54033.p1, not unique enough\n",
      "-skipping training candidate: rna54039.p1, not unique enough\n",
      "-skipping training candidate: rna54037.p1, not unique enough\n",
      "-skipping training candidate: rna54036.p1, not unique enough\n",
      "-skipping training candidate: rna54034.p1, not unique enough\n",
      "-skipping training candidate: rna1594.p1, not unique enough\n",
      "-skipping training candidate: rna6170.p1, not unique enough\n",
      "-skipping training candidate: rna1363.p1, not unique enough\n",
      "-skipping training candidate: rna1369.p1, not unique enough\n",
      "-skipping training candidate: rna35808.p1, not unique enough\n",
      "-skipping training candidate: rna61097.p1, not unique enough\n",
      "-skipping training candidate: rna9513.p1, not unique enough\n",
      "-skipping training candidate: rna35806.p1, not unique enough\n",
      "-skipping training candidate: rna36673.p1, not unique enough\n",
      "-skipping training candidate: rna7766.p1, not unique enough\n",
      "-skipping training candidate: rna19529.p1, not unique enough\n",
      "-skipping training candidate: rna28445.p1, not unique enough\n",
      "-skipping training candidate: rna54261.p1, not unique enough\n",
      "-skipping training candidate: rna54958.p1, not unique enough\n",
      "-skipping training candidate: rna3946.p1, not unique enough\n",
      "-skipping training candidate: rna9341.p1, not unique enough\n",
      "-skipping training candidate: rna31841.p1, not unique enough\n",
      "-skipping training candidate: rna9348.p1, not unique enough\n",
      "-skipping training candidate: rna23458.p1, not unique enough\n",
      "-skipping training candidate: rna26556.p1, not unique enough\n",
      "-skipping training candidate: rna48989.p1, not unique enough\n",
      "-skipping training candidate: rna50994.p1, not unique enough\n",
      "-skipping training candidate: rna25425.p1, not unique enough\n",
      "-skipping training candidate: rna25201.p1, not unique enough\n",
      "-skipping training candidate: rna54258.p1, not unique enough\n",
      "-skipping training candidate: rna60801.p1, not unique enough\n",
      "-skipping training candidate: rna9352.p1, not unique enough\n",
      "-skipping training candidate: rna26757.p1, not unique enough\n",
      "-skipping training candidate: rna26754.p1, not unique enough\n",
      "-skipping training candidate: rna43332.p1, not unique enough\n",
      "-skipping training candidate: rna59020.p1, not unique enough\n",
      "-skipping training candidate: rna30028.p1, not unique enough\n",
      "-skipping training candidate: rna36117.p1, not unique enough\n",
      "-skipping training candidate: rna45744.p1, not unique enough\n",
      "-skipping training candidate: rna45746.p1, not unique enough\n",
      "-skipping training candidate: rna1370.p1, not unique enough\n",
      "-skipping training candidate: rna9879.p1, not unique enough\n",
      "-skipping training candidate: rna9878.p1, not unique enough\n",
      "-skipping training candidate: rna29625.p1, not unique enough\n",
      "-skipping training candidate: rna36290.p1, not unique enough\n",
      "-skipping training candidate: rna36118.p1, not unique enough\n",
      "-skipping training candidate: rna50997.p1, not unique enough\n",
      "-skipping training candidate: rna9612.p1, not unique enough\n",
      "-skipping training candidate: rna18541.p1, not unique enough\n",
      "-skipping training candidate: rna31840.p1, not unique enough\n",
      "-skipping training candidate: rna34336.p1, not unique enough\n",
      "-skipping training candidate: rna34290.p1, not unique enough\n",
      "-skipping training candidate: rna49228.p1, not unique enough\n",
      "-skipping training candidate: rna32738.p1, not unique enough\n",
      "-skipping training candidate: rna51258.p1, not unique enough\n",
      "-skipping training candidate: rna9634.p1, not unique enough\n",
      "-skipping training candidate: rna23457.p1, not unique enough\n",
      "-skipping training candidate: rna25426.p1, not unique enough\n",
      "-skipping training candidate: rna44321.p1, not unique enough\n",
      "-skipping training candidate: rna18566.p1, not unique enough\n",
      "-skipping training candidate: rna29582.p1, not unique enough\n",
      "-skipping training candidate: rna51261.p1, not unique enough\n",
      "-skipping training candidate: rna54257.p1, not unique enough\n",
      "-skipping training candidate: rna3120.p1, not unique enough\n",
      "-skipping training candidate: rna3121.p1, not unique enough\n",
      "-skipping training candidate: rna23329.p1, not unique enough\n",
      "-skipping training candidate: rna27405.p1, not unique enough\n",
      "-skipping training candidate: rna32658.p1, not unique enough\n",
      "-skipping training candidate: rna32655.p1, not unique enough\n",
      "-skipping training candidate: rna32656.p1, not unique enough\n",
      "-skipping training candidate: rna36672.p1, not unique enough\n",
      "-skipping training candidate: rna37884.p1, not unique enough\n",
      "-skipping training candidate: rna2731.p1, not unique enough\n",
      "-skipping training candidate: rna40211.p1, not unique enough\n",
      "-skipping training candidate: rna61985.p1, not unique enough\n",
      "-skipping training candidate: rna26758.p1, not unique enough\n",
      "-skipping training candidate: rna31874.p1, not unique enough\n",
      "-skipping training candidate: rna34333.p1, not unique enough\n",
      "-skipping training candidate: rna9830.p1, not unique enough\n",
      "-skipping training candidate: rna9831.p1, not unique enough\n",
      "-skipping training candidate: rna19978.p1, not unique enough\n",
      "-skipping training candidate: rna23459.p1, not unique enough\n",
      "-skipping training candidate: rna38074.p1, not unique enough\n",
      "-skipping training candidate: rna46368.p1, not unique enough\n",
      "-skipping training candidate: rna3203.p1, not unique enough\n",
      "-skipping training candidate: rna9829.p1, not unique enough\n",
      "-skipping training candidate: rna9633.p1, not unique enough\n",
      "-skipping training candidate: rna22615.p1, not unique enough\n",
      "-skipping training candidate: rna22614.p1, not unique enough\n",
      "-skipping training candidate: rna30730.p1, not unique enough\n",
      "-skipping training candidate: rna34799.p1, not unique enough\n",
      "-skipping training candidate: rna34798.p1, not unique enough\n",
      "-skipping training candidate: rna51267.p1, not unique enough\n",
      "-skipping training candidate: rna51268.p1, not unique enough\n",
      "-skipping training candidate: rna4125.p1, not unique enough\n",
      "-skipping training candidate: rna2732.p1, not unique enough\n",
      "-skipping training candidate: rna19979.p1, not unique enough\n",
      "-skipping training candidate: rna32657.p1, not unique enough\n",
      "-skipping training candidate: rna36671.p1, not unique enough\n",
      "-skipping training candidate: rna44295.p1, not unique enough\n",
      "-skipping training candidate: rna48431.p1, not unique enough\n",
      "-skipping training candidate: rna51936.p1, not unique enough\n",
      "-skipping training candidate: rna27406.p1, not unique enough\n",
      "-skipping training candidate: rna31839.p1, not unique enough\n",
      "-skipping training candidate: rna38072.p1, not unique enough\n",
      "-skipping training candidate: rna38077.p1, not unique enough\n",
      "-skipping training candidate: rna46474.p1, not unique enough\n",
      "-skipping training candidate: rna49302.p1, not unique enough\n",
      "-skipping training candidate: rna4098.p1, not unique enough\n",
      "-skipping training candidate: rna28421.p1, not unique enough\n",
      "-skipping training candidate: rna28420.p1, not unique enough\n",
      "-skipping training candidate: rna40209.p1, not unique enough\n",
      "-skipping training candidate: rna9616.p1, not unique enough\n",
      "-skipping training candidate: rna29470.p1, not unique enough\n",
      "-skipping training candidate: rna51259.p1, not unique enough\n",
      "-skipping training candidate: rna9834.p1, not unique enough\n",
      "-skipping training candidate: rna19983.p1, not unique enough\n",
      "-skipping training candidate: rna37896.p1, not unique enough\n",
      "-skipping training candidate: rna1596.p1, not unique enough\n",
      "-skipping training candidate: rna19981.p1, not unique enough\n",
      "-skipping training candidate: rna25427.p1, not unique enough\n",
      "-skipping training candidate: rna36292.p1, not unique enough\n",
      "-skipping training candidate: rna37988.p1, not unique enough\n",
      "-skipping training candidate: rna48894.p1, not unique enough\n",
      "-skipping training candidate: rna54262.p1, not unique enough\n",
      "-skipping training candidate: rna54263.p1, not unique enough\n",
      "-skipping training candidate: rna9992.p1, not unique enough\n",
      "-skipping training candidate: rna9991.p1, not unique enough\n",
      "-skipping training candidate: rna8540.p1, not unique enough\n",
      "-skipping training candidate: rna18567.p1, not unique enough\n",
      "-skipping training candidate: rna26554.p1, not unique enough\n",
      "-skipping training candidate: rna27343.p1, not unique enough\n",
      "-skipping training candidate: rna9990.p1, not unique enough\n",
      "-skipping training candidate: rna30382.p1, not unique enough\n",
      "-skipping training candidate: rna31838.p1, not unique enough\n",
      "-skipping training candidate: rna37897.p1, not unique enough\n",
      "-skipping training candidate: rna54670.p1, not unique enough\n",
      "-skipping training candidate: rna57545.p1, not unique enough\n",
      "-skipping training candidate: rna46.p1, not unique enough\n",
      "-skipping training candidate: rna20625.p1, not unique enough\n",
      "-skipping training candidate: rna26077.p1, not unique enough\n",
      "-skipping training candidate: rna30196.p1, not unique enough\n",
      "-skipping training candidate: rna50998.p1, not unique enough\n",
      "-skipping training candidate: rna50010.p1, not unique enough\n",
      "-skipping training candidate: rna1108.p1, not unique enough\n",
      "-skipping training candidate: rna4102.p1, not unique enough\n",
      "-skipping training candidate: rna9624.p1, not unique enough\n",
      "-skipping training candidate: rna9625.p1, not unique enough\n",
      "-skipping training candidate: rna25794.p1, not unique enough\n",
      "-skipping training candidate: rna26756.p1, not unique enough\n",
      "-skipping training candidate: rna30195.p1, not unique enough\n",
      "-skipping training candidate: rna30197.p1, not unique enough\n",
      "-skipping training candidate: rna47105.p1, not unique enough\n",
      "-skipping training candidate: rna48429.p1, not unique enough\n",
      "-skipping training candidate: rna50999.p1, not unique enough\n",
      "-skipping training candidate: rna9993.p1, not unique enough\n",
      "-skipping training candidate: rna20627.p1, not unique enough\n",
      "-skipping training candidate: rna23125.p1, not unique enough\n",
      "-skipping training candidate: rna33342.p1, not unique enough\n",
      "-skipping training candidate: rna37669.p1, not unique enough\n",
      "-skipping training candidate: rna3572.p1, not unique enough\n",
      "-skipping training candidate: rna23455.p1, not unique enough\n",
      "-skipping training candidate: rna37895.p1, not unique enough\n",
      "-skipping training candidate: rna49316.p1, not unique enough\n",
      "-skipping training candidate: rna58979.p1, not unique enough\n",
      "-skipping training candidate: rna9828.p1, not unique enough\n",
      "-skipping training candidate: rna9350.p1, not unique enough\n",
      "-skipping training candidate: rna19908.p1, not unique enough\n",
      "-skipping training candidate: rna21176.p1, not unique enough\n",
      "-skipping training candidate: rna26863.p1, not unique enough\n",
      "-skipping training candidate: rna34706.p1, not unique enough\n",
      "-skipping training candidate: rna36291.p1, not unique enough\n",
      "-skipping training candidate: rna37668.p1, not unique enough\n",
      "-skipping training candidate: rna8532.p1, not unique enough\n",
      "-skipping training candidate: rna26623.p1, not unique enough\n",
      "-skipping training candidate: rna26622.p1, not unique enough\n",
      "-skipping training candidate: rna30031.p1, not unique enough\n",
      "-skipping training candidate: rna33938.p1, not unique enough\n",
      "-skipping training candidate: rna52529.p1, not unique enough\n",
      "-skipping training candidate: rna34335.p1, not unique enough\n",
      "-skipping training candidate: rna48383.p1, not unique enough\n",
      "-skipping training candidate: rna51260.p1, not unique enough\n",
      "-skipping training candidate: rna54052.p1, not unique enough\n",
      "-skipping training candidate: rna54054.p1, not unique enough\n",
      "-skipping training candidate: rna54057.p1, not unique enough\n",
      "-skipping training candidate: rna9836.p1, not unique enough\n",
      "-skipping training candidate: rna27314.p1, not unique enough\n",
      "-skipping training candidate: rna33869.p1, not unique enough\n",
      "-skipping training candidate: rna33870.p1, not unique enough\n",
      "-skipping training candidate: rna46878.p1, not unique enough\n",
      "-skipping training candidate: rna54056.p1, not unique enough\n",
      "-skipping training candidate: rna57160.p1, not unique enough\n",
      "-skipping training candidate: rna25203.p1, not unique enough\n",
      "-skipping training candidate: rna9832.p1, not unique enough\n",
      "-skipping training candidate: rna40000.p1, not unique enough\n",
      "-skipping training candidate: rna40107.p1, not unique enough\n",
      "-skipping training candidate: rna51262.p1, not unique enough\n",
      "-skipping training candidate: rna1904.p1, not unique enough\n",
      "-skipping training candidate: rna20572.p1, not unique enough\n",
      "-skipping training candidate: rna23986.p1, not unique enough\n",
      "-skipping training candidate: rna31835.p1, not unique enough\n",
      "-skipping training candidate: rna36294.p1, not unique enough\n",
      "-skipping training candidate: rna37876.p1, not unique enough\n",
      "-skipping training candidate: rna37875.p1, not unique enough\n",
      "-skipping training candidate: rna40517.p1, not unique enough\n",
      "-skipping training candidate: rna40106.p1, not unique enough\n",
      "-skipping training candidate: rna46879.p1, not unique enough\n",
      "-skipping training candidate: rna54260.p1, not unique enough\n",
      "-skipping training candidate: rna8533.p1, not unique enough\n",
      "-skipping training candidate: rna20571.p1, not unique enough\n",
      "-skipping training candidate: rna30030.p1, not unique enough\n",
      "-skipping training candidate: rna34337.p1, not unique enough\n",
      "-skipping training candidate: rna37667.p1, not unique enough\n",
      "-skipping training candidate: rna40516.p1, not unique enough\n",
      "-skipping training candidate: rna61006.p1, not unique enough\n",
      "-skipping training candidate: rna22179.p1, not unique enough\n",
      "-skipping training candidate: rna23984.p1, not unique enough\n",
      "-skipping training candidate: rna30991.p1, not unique enough\n",
      "-skipping training candidate: rna31758.p1, not unique enough\n",
      "-skipping training candidate: rna47.p1, not unique enough\n",
      "-skipping training candidate: rna10032.p1, not unique enough\n",
      "-skipping training candidate: rna30483.p1, not unique enough\n",
      "-skipping training candidate: rna31837.p1, not unique enough\n",
      "-skipping training candidate: rna33871.p1, not unique enough\n",
      "-skipping training candidate: rna33868.p1, not unique enough\n",
      "-skipping training candidate: rna33940.p1, not unique enough\n",
      "-skipping training candidate: rna44417.p1, not unique enough\n",
      "-skipping training candidate: rna49499.p1, not unique enough\n",
      "-skipping training candidate: rna9833.p1, not unique enough\n",
      "-skipping training candidate: rna19907.p1, not unique enough\n",
      "-skipping training candidate: rna27408.p1, not unique enough\n",
      "-skipping training candidate: rna30857.p1, not unique enough\n",
      "-skipping training candidate: rna36670.p1, not unique enough\n",
      "-skipping training candidate: rna3991.p1, not unique enough\n",
      "-skipping training candidate: rna3992.p1, not unique enough\n",
      "-skipping training candidate: rna21029.p1, not unique enough\n",
      "-skipping training candidate: rna25793.p1, not unique enough\n",
      "-skipping training candidate: rna34291.p1, not unique enough\n",
      "-skipping training candidate: rna3948.p1, not unique enough\n",
      "-skipping training candidate: rna19095.p1, not unique enough\n",
      "-skipping training candidate: rna25795.p1, not unique enough\n",
      "-skipping training candidate: rna29008.p1, not unique enough\n",
      "-skipping training candidate: rna30670.p1, not unique enough\n",
      "-skipping training candidate: rna30667.p1, not unique enough\n",
      "-skipping training candidate: rna35965.p1, not unique enough\n",
      "-skipping training candidate: rna50011.p1, not unique enough\n",
      "-skipping training candidate: rna3993.p1, not unique enough\n",
      "-skipping training candidate: rna3573.p1, not unique enough\n",
      "-skipping training candidate: rna19096.p1, not unique enough\n",
      "-skipping training candidate: rna26488.p1, not unique enough\n",
      "-skipping training candidate: rna37666.p1, not unique enough\n",
      "-skipping training candidate: rna58978.p1, not unique enough\n",
      "-skipping training candidate: rna8525.p1, not unique enough\n",
      "-skipping training candidate: rna20623.p1, not unique enough\n",
      "-skipping training candidate: rna30480.p1, not unique enough\n",
      "-skipping training candidate: rna19521.p1, not unique enough\n",
      "-skipping training candidate: rna25792.p1, not unique enough\n",
      "-skipping training candidate: rna31690.p1, not unique enough\n",
      "-skipping training candidate: rna46369.p1, not unique enough\n",
      "-skipping training candidate: rna51087.p1, not unique enough\n",
      "-skipping training candidate: rna51089.p1, not unique enough\n",
      "-skipping training candidate: rna53809.p1, not unique enough\n",
      "-skipping training candidate: rna1903.p1, not unique enough\n",
      "-skipping training candidate: rna5152.p1, not unique enough\n",
      "-skipping training candidate: rna31836.p1, not unique enough\n",
      "-skipping training candidate: rna36338.p1, not unique enough\n",
      "-skipping training candidate: rna35450.p1, not unique enough\n",
      "-skipping training candidate: rna42713.p1, not unique enough\n",
      "-skipping training candidate: rna58615.p1, not unique enough\n",
      "-skipping training candidate: rna23452.p1, not unique enough\n",
      "-skipping training candidate: rna49192.p1, not unique enough\n",
      "-skipping training candidate: rna11939.p1, not unique enough\n",
      "-skipping training candidate: rna44523.p1, not unique enough\n",
      "-skipping training candidate: rna43928.p1, not unique enough\n",
      "-skipping training candidate: rna61587.p1, not unique enough\n",
      "-skipping training candidate: rna62602.p1, not unique enough\n",
      "-skipping training candidate: rna62601.p1, not unique enough\n",
      "-skipping training candidate: rna1597.p1, not unique enough\n",
      "-skipping training candidate: rna30199.p1, not unique enough\n",
      "-skipping training candidate: rna39197.p1, not unique enough\n",
      "-skipping training candidate: rna53705.p1, not unique enough\n",
      "-skipping training candidate: rna25796.p1, not unique enough\n",
      "-skipping training candidate: rna23985.p1, not unique enough\n",
      "-skipping training candidate: rna37954.p1, not unique enough\n",
      "-skipping training candidate: rna25797.p1, not unique enough\n",
      "-skipping training candidate: rna33936.p1, not unique enough\n",
      "-skipping training candidate: rna54053.p1, not unique enough\n",
      "-skipping training candidate: rna658.p1, not unique enough\n",
      "-skipping training candidate: rna5956.p1, not unique enough\n",
      "-skipping training candidate: rna9838.p1, not unique enough\n",
      "-skipping training candidate: rna26178.p1, not unique enough\n",
      "-skipping training candidate: rna33876.p1, not unique enough\n",
      "-skipping training candidate: rna37886.p1, not unique enough\n",
      "-skipping training candidate: rna52527.p1, not unique enough\n",
      "-skipping training candidate: rna3042.p1, not unique enough\n",
      "-skipping training candidate: rna22573.p1, not unique enough\n",
      "-skipping training candidate: rna35784.p1, not unique enough\n",
      "-skipping training candidate: rna62316.p1, not unique enough\n",
      "-skipping training candidate: rna29974.p1, not unique enough\n",
      "-skipping training candidate: rna30484.p1, not unique enough\n",
      "-skipping training candidate: rna47102.p1, not unique enough\n",
      "-skipping training candidate: rna55741.p1, not unique enough\n",
      "-skipping training candidate: rna20626.p1, not unique enough\n",
      "-skipping training candidate: rna30335.p1, not unique enough\n",
      "-skipping training candidate: rna37887.p1, not unique enough\n",
      "-skipping training candidate: rna47103.p1, not unique enough\n",
      "-skipping training candidate: rna50420.p1, not unique enough\n",
      "-skipping training candidate: rna51310.p1, not unique enough\n",
      "-skipping training candidate: rna29978.p1, not unique enough\n",
      "-skipping training candidate: rna30383.p1, not unique enough\n",
      "-skipping training candidate: rna37879.p1, not unique enough\n",
      "-skipping training candidate: rna51513.p1, not unique enough\n",
      "-skipping training candidate: rna59038.p1, not unique enough\n",
      "-skipping training candidate: rna9339.p1, not unique enough\n",
      "-skipping training candidate: rna19522.p1, not unique enough\n",
      "-skipping training candidate: rna18883.p1, not unique enough\n",
      "-skipping training candidate: rna43173.p1, not unique enough\n",
      "-skipping training candidate: rna49311.p1, not unique enough\n",
      "-skipping training candidate: rna1680.p1, not unique enough\n",
      "-skipping training candidate: rna22181.p1, not unique enough\n",
      "-skipping training candidate: rna39716.p1, not unique enough\n",
      "-skipping training candidate: rna46266.p1, not unique enough\n",
      "-skipping training candidate: rna51198.p1, not unique enough\n",
      "-skipping training candidate: rna53808.p1, not unique enough\n",
      "-skipping training candidate: rna4553.p1, not unique enough\n",
      "-skipping training candidate: rna31240.p1, not unique enough\n",
      "-skipping training candidate: rna31237.p1, not unique enough\n",
      "-skipping training candidate: rna31238.p1, not unique enough\n",
      "-skipping training candidate: rna31239.p1, not unique enough\n",
      "-skipping training candidate: rna31242.p1, not unique enough\n",
      "-skipping training candidate: rna32437.p1, not unique enough\n",
      "-skipping training candidate: rna33937.p1, not unique enough\n",
      "-skipping training candidate: rna37881.p1, not unique enough\n",
      "-skipping training candidate: rna52528.p1, not unique enough\n",
      "-skipping training candidate: rna62368.p1, not unique enough\n",
      "-skipping training candidate: rna3995.p1, not unique enough\n",
      "-skipping training candidate: rna3994.p1, not unique enough\n",
      "-skipping training candidate: rna9494.p1, not unique enough\n",
      "-skipping training candidate: rna54347.p1, not unique enough\n",
      "-skipping training candidate: rna47857.p1, not unique enough\n",
      "-skipping training candidate: rna59036.p1, not unique enough\n",
      "-skipping training candidate: rna19527.p1, not unique enough\n",
      "-skipping training candidate: rna43927.p1, not unique enough\n",
      "-skipping training candidate: rna29146.p1, not unique enough\n",
      "-skipping training candidate: rna32160.p1, not unique enough\n",
      "-skipping training candidate: rna35496.p1, not unique enough\n",
      "-skipping training candidate: rna38591.p1, not unique enough\n",
      "-skipping training candidate: rna38212.p1, not unique enough\n",
      "-skipping training candidate: rna48896.p1, not unique enough\n",
      "-skipping training candidate: rna48895.p1, not unique enough\n",
      "-skipping training candidate: rna59093.p1, not unique enough\n",
      "-skipping training candidate: rna59905.p1, not unique enough\n",
      "-skipping training candidate: rna37882.p1, not unique enough\n",
      "-skipping training candidate: rna40188.p1, not unique enough\n",
      "-skipping training candidate: rna1019.p1, not unique enough\n",
      "-skipping training candidate: rna30386.p1, not unique enough\n",
      "-skipping training candidate: rna60870.p1, not unique enough\n",
      "-skipping training candidate: rna3859.p1, not unique enough\n",
      "-skipping training candidate: rna25798.p1, not unique enough\n",
      "-skipping training candidate: rna32737.p1, not unique enough\n",
      "-skipping training candidate: rna49500.p1, not unique enough\n",
      "-skipping training candidate: rna54041.p1, not unique enough\n",
      "-skipping training candidate: rna26180.p1, not unique enough\n",
      "-skipping training candidate: rna32159.p1, not unique enough\n",
      "-skipping training candidate: rna41604.p1, not unique enough\n",
      "-skipping training candidate: rna36336.p1, not unique enough\n",
      "-skipping training candidate: rna58516.p1, not unique enough\n",
      "-skipping training candidate: rna9605.p1, not unique enough\n",
      "-skipping training candidate: rna26759.p1, not unique enough\n",
      "-skipping training candidate: rna26190.p1, not unique enough\n",
      "-skipping training candidate: rna1648.p1, not unique enough\n",
      "-skipping training candidate: rna23981.p1, not unique enough\n",
      "-skipping training candidate: rna25495.p1, not unique enough\n",
      "-skipping training candidate: rna23988.p1, not unique enough\n",
      "-skipping training candidate: rna30447.p1, not unique enough\n",
      "-skipping training candidate: rna50829.p1, not unique enough\n",
      "-skipping training candidate: rna1299.p1, not unique enough\n",
      "-skipping training candidate: rna27316.p1, not unique enough\n",
      "-skipping training candidate: rna61635.p1, not unique enough\n",
      "-skipping training candidate: rna34633.p1, not unique enough\n",
      "-skipping training candidate: rna37955.p1, not unique enough\n",
      "-skipping training candidate: rna61636.p1, not unique enough\n",
      "-skipping training candidate: rna35779.p1, not unique enough\n",
      "-skipping training candidate: rna45803.p1, not unique enough\n",
      "-skipping training candidate: rna50144.p1, not unique enough\n",
      "-skipping training candidate: rna61634.p1, not unique enough\n",
      "-skipping training candidate: rna30482.p1, not unique enough\n",
      "-skipping training candidate: rna34590.p1, not unique enough\n",
      "-skipping training candidate: rna43930.p1, not unique enough\n",
      "-skipping training candidate: rna43926.p1, not unique enough\n",
      "-skipping training candidate: rna46367.p1, not unique enough\n",
      "-skipping training candidate: rna59903.p1, not unique enough\n",
      "-skipping training candidate: rna60121.p1, not unique enough\n",
      "-skipping training candidate: rna60726.p1, not unique enough\n",
      "-skipping training candidate: rna9837.p1, not unique enough\n",
      "-skipping training candidate: rna35119.p1, not unique enough\n",
      "-skipping training candidate: rna37892.p1, not unique enough\n",
      "-skipping training candidate: rna40903.p1, not unique enough\n",
      "-skipping training candidate: rna60732.p1, not unique enough\n",
      "-skipping training candidate: rna3945.p1, not unique enough\n",
      "-skipping training candidate: rna37619.p1, not unique enough\n",
      "-skipping training candidate: rna49657.p1, not unique enough\n",
      "-skipping training candidate: rna61631.p1, not unique enough\n",
      "-skipping training candidate: rna34404.p1, not unique enough\n",
      "-skipping training candidate: rna35785.p1, not unique enough\n",
      "-skipping training candidate: rna40632.p1, not unique enough\n",
      "-skipping training candidate: rna49654.p1, not unique enough\n",
      "-skipping training candidate: rna3858.p1, not unique enough\n",
      "-skipping training candidate: rna38329.p1, not unique enough\n",
      "-skipping training candidate: rna37893.p1, not unique enough\n",
      "-skipping training candidate: rna40994.p1, not unique enough\n",
      "-skipping training candidate: rna41001.p1, not unique enough\n",
      "-skipping training candidate: rna40997.p1, not unique enough\n",
      "-skipping training candidate: rna49229.p1, not unique enough\n",
      "-skipping training candidate: rna3339.p1, not unique enough\n",
      "-skipping training candidate: rna26181.p1, not unique enough\n",
      "-skipping training candidate: rna15058.p1, not unique enough\n",
      "-skipping training candidate: rna30336.p1, not unique enough\n",
      "-skipping training candidate: rna37888.p1, not unique enough\n",
      "-skipping training candidate: rna41006.p1, not unique enough\n",
      "-skipping training candidate: rna3862.p1, not unique enough\n",
      "-skipping training candidate: rna30385.p1, not unique enough\n",
      "-skipping training candidate: rna40993.p1, not unique enough\n",
      "-skipping training candidate: rna59037.p1, not unique enough\n",
      "-skipping training candidate: rna61180.p1, not unique enough\n",
      "-skipping training candidate: rna34402.p1, not unique enough\n",
      "-skipping training candidate: rna34860.p1, not unique enough\n",
      "-skipping training candidate: rna37899.p1, not unique enough\n",
      "-skipping training candidate: rna60731.p1, not unique enough\n",
      "-skipping training candidate: rna60730.p1, not unique enough\n",
      "-skipping training candidate: rna19906.p1, not unique enough\n",
      "-skipping training candidate: rna18830.p1, not unique enough\n",
      "-skipping training candidate: rna27717.p1, not unique enough\n",
      "-skipping training candidate: rna29982.p1, not unique enough\n",
      "-skipping training candidate: rna29983.p1, not unique enough\n",
      "-skipping training candidate: rna37889.p1, not unique enough\n",
      "-skipping training candidate: rna47104.p1, not unique enough\n",
      "-skipping training candidate: rna1697.p1, not unique enough\n",
      "-skipping training candidate: rna23454.p1, not unique enough\n",
      "-skipping training candidate: rna37891.p1, not unique enough\n",
      "-skipping training candidate: rna41005.p1, not unique enough\n",
      "-skipping training candidate: rna48475.p1, not unique enough\n",
      "-skipping training candidate: rna51340.p1, not unique enough\n",
      "-skipping training candidate: rna659.p1, not unique enough\n",
      "-skipping training candidate: rna3861.p1, not unique enough\n",
      "-skipping training candidate: rna29977.p1, not unique enough\n",
      "-skipping training candidate: rna34432.p1, not unique enough\n",
      "-skipping training candidate: rna42245.p1, not unique enough\n",
      "-skipping training candidate: rna48476.p1, not unique enough\n",
      "-skipping training candidate: rna60844.p1, not unique enough\n",
      "-skipping training candidate: rna61630.p1, not unique enough\n",
      "-skipping training candidate: rna1708.p1, not unique enough\n",
      "-skipping training candidate: rna3660.p1, not unique enough\n",
      "-skipping training candidate: rna34434.p1, not unique enough\n",
      "-skipping training candidate: rna51338.p1, not unique enough\n",
      "-skipping training candidate: rna59612.p1, not unique enough\n",
      "-skipping training candidate: rna1241.p1, not unique enough\n",
      "-skipping training candidate: rna23245.p1, not unique enough\n",
      "-skipping training candidate: rna26189.p1, not unique enough\n",
      "-skipping training candidate: rna30450.p1, not unique enough\n",
      "-skipping training candidate: rna34274.p1, not unique enough\n",
      "-skipping training candidate: rna34401.p1, not unique enough\n",
      "-skipping training candidate: rna38211.p1, not unique enough\n",
      "-skipping training candidate: rna39388.p1, not unique enough\n",
      "-skipping training candidate: rna41603.p1, not unique enough\n",
      "-skipping training candidate: rna42630.p1, not unique enough\n",
      "-skipping training candidate: rna36005.p1, not unique enough\n",
      "-skipping training candidate: rna37956.p1, not unique enough\n",
      "-skipping training candidate: rna27411.p1, not unique enough\n",
      "-skipping training candidate: rna32157.p1, not unique enough\n",
      "-skipping training candidate: rna35782.p1, not unique enough\n",
      "-skipping training candidate: rna41007.p1, not unique enough\n",
      "-skipping training candidate: rna58514.p1, not unique enough\n",
      "-skipping training candidate: rna59095.p1, not unique enough\n",
      "-skipping training candidate: rna58984.p1, not unique enough\n",
      "-skipping training candidate: rna21663.p1, not unique enough\n",
      "-skipping training candidate: rna25428.p1, not unique enough\n",
      "-skipping training candidate: rna34403.p1, not unique enough\n",
      "-skipping training candidate: rna41002.p1, not unique enough\n",
      "-skipping training candidate: rna47856.p1, not unique enough\n",
      "-skipping training candidate: rna49653.p1, not unique enough\n",
      "-skipping training candidate: rna60734.p1, not unique enough\n",
      "-skipping training candidate: rna60727.p1, not unique enough\n",
      "-skipping training candidate: rna1705.p1, not unique enough\n",
      "-skipping training candidate: rna15057.p1, not unique enough\n",
      "-skipping training candidate: rna9342.p1, not unique enough\n",
      "-skipping training candidate: rna30668.p1, not unique enough\n",
      "-skipping training candidate: rna33591.p1, not unique enough\n",
      "-skipping training candidate: rna33942.p1, not unique enough\n",
      "-skipping training candidate: rna45858.p1, not unique enough\n",
      "-skipping training candidate: rna47855.p1, not unique enough\n",
      "-skipping training candidate: rna49655.p1, not unique enough\n",
      "-skipping training candidate: rna27409.p1, not unique enough\n",
      "-skipping training candidate: rna30669.p1, not unique enough\n",
      "-skipping training candidate: rna37957.p1, not unique enough\n",
      "-skipping training candidate: rna28134.p1, not unique enough\n",
      "-skipping training candidate: rna33872.p1, not unique enough\n",
      "-skipping training candidate: rna33873.p1, not unique enough\n",
      "-skipping training candidate: rna40992.p1, not unique enough\n",
      "-skipping training candidate: rna58515.p1, not unique enough\n",
      "-skipping training candidate: rna30448.p1, not unique enough\n",
      "-skipping training candidate: rna30387.p1, not unique enough\n",
      "-skipping training candidate: rna36337.p1, not unique enough\n",
      "-skipping training candidate: rna35799.p1, not unique enough\n",
      "-skipping training candidate: rna38716.p1, not unique enough\n",
      "-skipping training candidate: rna50537.p1, not unique enough\n",
      "-skipping training candidate: rna58518.p1, not unique enough\n",
      "-skipping training candidate: rna3865.p1, not unique enough\n",
      "-skipping training candidate: rna30479.p1, not unique enough\n",
      "-skipping training candidate: rna33875.p1, not unique enough\n",
      "-skipping training candidate: rna33874.p1, not unique enough\n",
      "-skipping training candidate: rna50541.p1, not unique enough\n",
      "-skipping training candidate: rna9611.p1, not unique enough\n",
      "-skipping training candidate: rna29313.p1, not unique enough\n",
      "-skipping training candidate: rna59486.p1, not unique enough\n",
      "-skipping training candidate: rna59858.p1, not unique enough\n",
      "-skipping training candidate: rna60729.p1, not unique enough\n",
      "-skipping training candidate: rna25429.p1, not unique enough\n",
      "-skipping training candidate: rna32214.p1, not unique enough\n",
      "-skipping training candidate: rna60842.p1, not unique enough\n",
      "-skipping training candidate: rna9615.p1, not unique enough\n",
      "-skipping training candidate: rna34446.p1, not unique enough\n",
      "-skipping training candidate: rna1710.p1, not unique enough\n",
      "-skipping training candidate: rna27075.p1, not unique enough\n",
      "-skipping training candidate: rna47349.p1, not unique enough\n",
      "-skipping training candidate: rna47352.p1, not unique enough\n",
      "-skipping training candidate: rna53764.p1, not unique enough\n",
      "-skipping training candidate: rna9344.p1, not unique enough\n",
      "-skipping training candidate: rna34430.p1, not unique enough\n",
      "-skipping training candidate: rna38330.p1, not unique enough\n",
      "-skipping training candidate: rna58614.p1, not unique enough\n",
      "-skipping training candidate: rna160.p1, not unique enough\n",
      "-skipping training candidate: rna6844.p1, not unique enough\n",
      "-skipping training candidate: rna21293.p1, not unique enough\n",
      "-skipping training candidate: rna34431.p1, not unique enough\n",
      "-skipping training candidate: rna26624.p1, not unique enough\n",
      "-skipping training candidate: rna31834.p1, not unique enough\n",
      "-skipping training candidate: rna31567.p1, not unique enough\n",
      "-skipping training candidate: rna31562.p1, not unique enough\n",
      "-skipping training candidate: rna31566.p1, not unique enough\n",
      "-skipping training candidate: rna31565.p1, not unique enough\n",
      "-skipping training candidate: rna37901.p1, not unique enough\n",
      "-skipping training candidate: rna1466.p1, not unique enough\n",
      "-skipping training candidate: rna46371.p1, not unique enough\n",
      "-skipping training candidate: rna53328.p1, not unique enough\n",
      "-skipping training candidate: rna4122.p1, not unique enough\n",
      "-skipping training candidate: rna28424.p1, not unique enough\n",
      "-skipping training candidate: rna34275.p1, not unique enough\n",
      "-skipping training candidate: rna43978.p1, not unique enough\n",
      "-skipping training candidate: rna60728.p1, not unique enough\n",
      "-skipping training candidate: rna27466.p1, not unique enough\n",
      "-skipping training candidate: rna30481.p1, not unique enough\n",
      "-skipping training candidate: rna36492.p1, not unique enough\n",
      "-skipping training candidate: rna57742.p1, not unique enough\n",
      "-skipping training candidate: rna61632.p1, not unique enough\n",
      "-skipping training candidate: rna9347.p1, not unique enough\n",
      "-skipping training candidate: rna21292.p1, not unique enough\n",
      "-skipping training candidate: rna25878.p1, not unique enough\n",
      "-skipping training candidate: rna25876.p1, not unique enough\n",
      "-skipping training candidate: rna26637.p1, not unique enough\n",
      "-skipping training candidate: rna27062.p1, not unique enough\n",
      "-skipping training candidate: rna31564.p1, not unique enough\n",
      "-skipping training candidate: rna38919.p1, not unique enough\n",
      "-skipping training candidate: rna37883.p1, not unique enough\n",
      "-skipping training candidate: rna53326.p1, not unique enough\n",
      "-skipping training candidate: rna58982.p1, not unique enough\n",
      "-skipping training candidate: rna3918.p1, not unique enough\n",
      "-skipping training candidate: rna9610.p1, not unique enough\n",
      "-skipping training candidate: rna22866.p1, not unique enough\n",
      "-skipping training candidate: rna22864.p1, not unique enough\n",
      "-skipping training candidate: rna22863.p1, not unique enough\n",
      "-skipping training candidate: rna27061.p1, not unique enough\n",
      "-skipping training candidate: rna27078.p1, not unique enough\n",
      "-skipping training candidate: rna34276.p1, not unique enough\n",
      "-skipping training candidate: rna39199.p1, not unique enough\n",
      "-skipping training candidate: rna37890.p1, not unique enough\n",
      "-skipping training candidate: rna60845.p1, not unique enough\n",
      "-skipping training candidate: rna2570.p1, not unique enough\n",
      "-skipping training candidate: rna9301.p1, not unique enough\n",
      "-skipping training candidate: rna19726.p1, not unique enough\n",
      "-skipping training candidate: rna27079.p1, not unique enough\n",
      "-skipping training candidate: rna59930.p1, not unique enough\n",
      "-skipping training candidate: rna62579.p1, not unique enough\n",
      "-skipping training candidate: rna1695.p1, not unique enough\n",
      "-skipping training candidate: rna1696.p1, not unique enough\n",
      "-skipping training candidate: rna11647.p1, not unique enough\n",
      "-skipping training candidate: rna25877.p1, not unique enough\n",
      "-skipping training candidate: rna34066.p1, not unique enough\n",
      "-skipping training candidate: rna34436.p1, not unique enough\n",
      "-skipping training candidate: rna60841.p1, not unique enough\n",
      "-skipping training candidate: rna4508.p1, not unique enough\n",
      "-skipping training candidate: rna9302.p1, not unique enough\n",
      "-skipping training candidate: rna26640.p1, not unique enough\n",
      "-skipping training candidate: rna24544.p1, not unique enough\n",
      "-skipping training candidate: rna27412.p1, not unique enough\n",
      "-skipping training candidate: rna29980.p1, not unique enough\n",
      "-skipping training candidate: rna29981.p1, not unique enough\n",
      "-skipping training candidate: rna34437.p1, not unique enough\n",
      "-skipping training candidate: rna34433.p1, not unique enough\n",
      "-skipping training candidate: rna36461.p1, not unique enough\n",
      "-skipping training candidate: rna48951.p1, not unique enough\n",
      "-skipping training candidate: rna56817.p1, not unique enough\n",
      "-skipping training candidate: rna62582.p1, not unique enough\n",
      "-skipping training candidate: rna38752.p1, not unique enough\n",
      "-skipping training candidate: rna45577.p1, not unique enough\n",
      "-skipping training candidate: rna45576.p1, not unique enough\n",
      "-skipping training candidate: rna54146.p1, not unique enough\n",
      "-skipping training candidate: rna62023.p1, not unique enough\n",
      "-skipping training candidate: rna65121.p1, not unique enough\n",
      "-skipping training candidate: rna4121.p1, not unique enough\n",
      "-skipping training candidate: rna27077.p1, not unique enough\n",
      "-skipping training candidate: rna29976.p1, not unique enough\n",
      "-skipping training candidate: rna30921.p1, not unique enough\n",
      "-skipping training candidate: rna34448.p1, not unique enough\n",
      "-skipping training candidate: rna37167.p1, not unique enough\n",
      "-skipping training candidate: rna40936.p1, not unique enough\n",
      "-skipping training candidate: rna54148.p1, not unique enough\n",
      "-skipping training candidate: rna58981.p1, not unique enough\n",
      "-skipping training candidate: rna62103.p1, not unique enough\n",
      "-skipping training candidate: rna3919.p1, not unique enough\n",
      "-skipping training candidate: rna22657.p1, not unique enough\n",
      "-skipping training candidate: rna27410.p1, not unique enough\n",
      "-skipping training candidate: rna36008.p1, not unique enough\n",
      "-skipping training candidate: rna39892.p1, not unique enough\n",
      "-skipping training candidate: rna1021.p1, not unique enough\n",
      "-skipping training candidate: rna9299.p1, not unique enough\n",
      "-skipping training candidate: rna25417.p1, not unique enough\n",
      "-skipping training candidate: rna29951.p1, not unique enough\n",
      "-skipping training candidate: rna37166.p1, not unique enough\n",
      "-skipping training candidate: rna50540.p1, not unique enough\n",
      "-skipping training candidate: rna50539.p1, not unique enough\n",
      "-skipping training candidate: rna3577.p1, not unique enough\n",
      "-skipping training candidate: rna27064.p1, not unique enough\n",
      "-skipping training candidate: rna31243.p1, not unique enough\n",
      "-skipping training candidate: rna43667.p1, not unique enough\n",
      "-skipping training candidate: rna51040.p1, not unique enough\n",
      "-skipping training candidate: rna52150.p1, not unique enough\n",
      "-skipping training candidate: rna52151.p1, not unique enough\n",
      "-skipping training candidate: rna52472.p1, not unique enough\n",
      "-skipping training candidate: rna58513.p1, not unique enough\n",
      "-skipping training candidate: rna61741.p1, not unique enough\n",
      "-skipping training candidate: rna23456.p1, not unique enough\n",
      "-skipping training candidate: rna32392.p1, not unique enough\n",
      "-skipping training candidate: rna40112.p1, not unique enough\n",
      "-skipping training candidate: rna41003.p1, not unique enough\n",
      "-skipping training candidate: rna43933.p1, not unique enough\n",
      "-skipping training candidate: rna54151.p1, not unique enough\n",
      "-skipping training candidate: rna9607.p1, not unique enough\n",
      "-skipping training candidate: rna30378.p1, not unique enough\n",
      "-skipping training candidate: rna42281.p1, not unique enough\n",
      "-skipping training candidate: rna45538.p1, not unique enough\n",
      "-skipping training candidate: rna54147.p1, not unique enough\n",
      "-skipping training candidate: rna1706.p1, not unique enough\n",
      "-skipping training candidate: rna3864.p1, not unique enough\n",
      "-skipping training candidate: rna53765.p1, not unique enough\n",
      "-skipping training candidate: rna54149.p1, not unique enough\n",
      "-skipping training candidate: rna60846.p1, not unique enough\n",
      "-skipping training candidate: rna62102.p1, not unique enough\n",
      "-skipping training candidate: rna19567.p1, not unique enough\n",
      "-skipping training candidate: rna21918.p1, not unique enough\n",
      "-skipping training candidate: rna54150.p1, not unique enough\n",
      "-skipping training candidate: rna58983.p1, not unique enough\n",
      "-skipping training candidate: rna7810.p1, not unique enough\n",
      "-skipping training candidate: rna9305.p1, not unique enough\n",
      "-skipping training candidate: rna9293.p1, not unique enough\n",
      "-skipping training candidate: rna21917.p1, not unique enough\n",
      "-skipping training candidate: rna33168.p1, not unique enough\n",
      "-skipping training candidate: rna33169.p1, not unique enough\n",
      "-skipping training candidate: rna55674.p1, not unique enough\n",
      "-skipping training candidate: rna7811.p1, not unique enough\n",
      "-skipping training candidate: rna35780.p1, not unique enough\n",
      "-skipping training candidate: rna36006.p1, not unique enough\n",
      "-skipping training candidate: rna39685.p1, not unique enough\n",
      "-skipping training candidate: rna41009.p1, not unique enough\n",
      "-skipping training candidate: rna52149.p1, not unique enough\n",
      "-skipping training candidate: rna52153.p1, not unique enough\n",
      "-skipping training candidate: rna9300.p1, not unique enough\n",
      "-skipping training candidate: rna33171.p1, not unique enough\n",
      "-skipping training candidate: rna47350.p1, not unique enough\n",
      "-skipping training candidate: rna51926.p1, not unique enough\n",
      "-skipping training candidate: rna58520.p1, not unique enough\n",
      "-skipping training candidate: rna58519.p1, not unique enough\n",
      "-skipping training candidate: rna3222.p1, not unique enough\n",
      "-skipping training candidate: rna9609.p1, not unique enough\n",
      "-skipping training candidate: rna9307.p1, not unique enough\n",
      "-skipping training candidate: rna19721.p1, not unique enough\n",
      "-skipping training candidate: rna27242.p1, not unique enough\n",
      "-skipping training candidate: rna33170.p1, not unique enough\n",
      "-skipping training candidate: rna40999.p1, not unique enough\n",
      "-skipping training candidate: rna49643.p1, not unique enough\n",
      "-skipping training candidate: rna55849.p1, not unique enough\n",
      "-skipping training candidate: rna21177.p1, not unique enough\n",
      "-skipping training candidate: rna24343.p1, not unique enough\n",
      "-skipping training candidate: rna30732.p1, not unique enough\n",
      "-skipping training candidate: rna40730.p1, not unique enough\n",
      "-skipping training candidate: rna43962.p1, not unique enough\n",
      "-skipping training candidate: rna43964.p1, not unique enough\n",
      "-skipping training candidate: rna49644.p1, not unique enough\n",
      "-skipping training candidate: rna55850.p1, not unique enough\n",
      "-skipping training candidate: rna59485.p1, not unique enough\n",
      "-skipping training candidate: rna1709.p1, not unique enough\n",
      "-skipping training candidate: rna20457.p1, not unique enough\n",
      "-skipping training candidate: rna26085.p1, not unique enough\n",
      "-skipping training candidate: rna39891.p1, not unique enough\n",
      "-skipping training candidate: rna49652.p1, not unique enough\n",
      "-skipping training candidate: rna50542.p1, not unique enough\n",
      "-skipping training candidate: rna62576.p1, not unique enough\n",
      "-skipping training candidate: rna27262.p1, not unique enough\n",
      "-skipping training candidate: rna30486.p1, not unique enough\n",
      "-skipping training candidate: rna30389.p1, not unique enough\n",
      "-skipping training candidate: rna9298.p1, not unique enough\n",
      "-skipping training candidate: rna21919.p1, not unique enough\n",
      "-skipping training candidate: rna30478.p1, not unique enough\n",
      "-skipping training candidate: rna47370.p1, not unique enough\n",
      "-skipping training candidate: rna45692.p1, not unique enough\n",
      "-skipping training candidate: rna45698.p1, not unique enough\n",
      "-skipping training candidate: rna62672.p1, not unique enough\n",
      "-skipping training candidate: rna3863.p1, not unique enough\n",
      "-skipping training candidate: rna62580.p1, not unique enough\n",
      "-skipping training candidate: rna27317.p1, not unique enough\n",
      "-skipping training candidate: rna45684.p1, not unique enough\n",
      "-skipping training candidate: rna52647.p1, not unique enough\n",
      "-skipping training candidate: rna60120.p1, not unique enough\n",
      "-skipping training candidate: rna782.p1, not unique enough\n",
      "-skipping training candidate: rna27042.p1, not unique enough\n",
      "-skipping training candidate: rna34636.p1, not unique enough\n",
      "-skipping training candidate: rna40113.p1, not unique enough\n",
      "-skipping training candidate: rna3977.p1, not unique enough\n",
      "-skipping training candidate: rna57315.p1, not unique enough\n",
      "-skipping training candidate: rna21752.p1, not unique enough\n",
      "-skipping training candidate: rna29221.p1, not unique enough\n",
      "-skipping training candidate: rna50757.p1, not unique enough\n",
      "-skipping training candidate: rna58616.p1, not unique enough\n",
      "-skipping training candidate: rna58618.p1, not unique enough\n",
      "-skipping training candidate: rna61567.p1, not unique enough\n",
      "-skipping training candidate: rna29147.p1, not unique enough\n",
      "-skipping training candidate: rna30384.p1, not unique enough\n",
      "-skipping training candidate: rna37894.p1, not unique enough\n",
      "-skipping training candidate: rna60848.p1, not unique enough\n",
      "-skipping training candidate: rna60847.p1, not unique enough\n",
      "-skipping training candidate: rna60849.p1, not unique enough\n",
      "-skipping training candidate: rna3976.p1, not unique enough\n",
      "-skipping training candidate: rna19006.p1, not unique enough\n",
      "-skipping training candidate: rna22526.p1, not unique enough\n",
      "-skipping training candidate: rna38067.p1, not unique enough\n",
      "-skipping training candidate: rna37898.p1, not unique enough\n",
      "-skipping training candidate: rna47380.p1, not unique enough\n",
      "-skipping training candidate: rna45694.p1, not unique enough\n",
      "-skipping training candidate: rna48474.p1, not unique enough\n",
      "-skipping training candidate: rna62101.p1, not unique enough\n",
      "-skipping training candidate: rna1305.p1, not unique enough\n",
      "-skipping training candidate: rna1304.p1, not unique enough\n",
      "-skipping training candidate: rna20831.p1, not unique enough\n",
      "-skipping training candidate: rna24359.p1, not unique enough\n",
      "-skipping training candidate: rna48958.p1, not unique enough\n",
      "-skipping training candidate: rna55866.p1, not unique enough\n",
      "-skipping training candidate: rna63306.p1, not unique enough\n",
      "-skipping training candidate: rna1701.p1, not unique enough\n",
      "-skipping training candidate: rna41004.p1, not unique enough\n",
      "-skipping training candidate: rna2533.p1, not unique enough\n",
      "-skipping training candidate: rna5464.p1, not unique enough\n",
      "-skipping training candidate: rna37900.p1, not unique enough\n",
      "-skipping training candidate: rna40674.p1, not unique enough\n",
      "-skipping training candidate: rna22803.p1, not unique enough\n",
      "-skipping training candidate: rna22802.p1, not unique enough\n",
      "-skipping training candidate: rna26046.p1, not unique enough\n",
      "-skipping training candidate: rna29979.p1, not unique enough\n",
      "-skipping training candidate: rna29703.p1, not unique enough\n",
      "-skipping training candidate: rna36231.p1, not unique enough\n",
      "-skipping training candidate: rna39374.p1, not unique enough\n",
      "-skipping training candidate: rna59484.p1, not unique enough\n",
      "-skipping training candidate: rna11665.p1, not unique enough\n",
      "-skipping training candidate: rna52111.p1, not unique enough\n",
      "-skipping training candidate: rna1700.p1, not unique enough\n",
      "-skipping training candidate: rna5002.p1, not unique enough\n",
      "-skipping training candidate: rna5003.p1, not unique enough\n",
      "-skipping training candidate: rna29975.p1, not unique enough\n",
      "-skipping training candidate: rna62577.p1, not unique enough\n",
      "-skipping training candidate: rna29800.p1, not unique enough\n",
      "-skipping training candidate: rna49656.p1, not unique enough\n",
      "-skipping training candidate: rna49658.p1, not unique enough\n",
      "-skipping training candidate: rna59219.p1, not unique enough\n",
      "-skipping training candidate: rna60890.p1, not unique enough\n",
      "-skipping training candidate: rna1470.p1, not unique enough\n",
      "-skipping training candidate: rna9308.p1, not unique enough\n",
      "-skipping training candidate: rna9297.p1, not unique enough\n",
      "-skipping training candidate: rna10019.p1, not unique enough\n",
      "-skipping training candidate: rna32217.p1, not unique enough\n",
      "-skipping training candidate: rna1707.p1, not unique enough\n",
      "-skipping training candidate: rna21753.p1, not unique enough\n",
      "-skipping training candidate: rna34435.p1, not unique enough\n",
      "-skipping training candidate: rna35999.p1, not unique enough\n",
      "-skipping training candidate: rna10034.p1, not unique enough\n",
      "-skipping training candidate: rna27310.p1, not unique enough\n",
      "-skipping training candidate: rna27849.p1, not unique enough\n",
      "-skipping training candidate: rna37657.p1, not unique enough\n",
      "-skipping training candidate: rna45051.p1, not unique enough\n",
      "-skipping training candidate: rna55343.p1, not unique enough\n",
      "-skipping training candidate: rna9608.p1, not unique enough\n",
      "-skipping training candidate: rna26638.p1, not unique enough\n",
      "-skipping training candidate: rna25853.p1, not unique enough\n",
      "-skipping training candidate: rna27855.p1, not unique enough\n",
      "-skipping training candidate: rna30165.p1, not unique enough\n",
      "-skipping training candidate: rna35353.p1, not unique enough\n",
      "-skipping training candidate: rna60800.p1, not unique enough\n",
      "-skipping training candidate: rna22557.p1, not unique enough\n",
      "-skipping training candidate: rna41972.p1, not unique enough\n",
      "-skipping training candidate: rna48398.p1, not unique enough\n",
      "-skipping training candidate: rna47651.p1, not unique enough\n",
      "-skipping training candidate: rna50471.p1, not unique enough\n",
      "-skipping training candidate: rna50231.p1, not unique enough\n",
      "-skipping training candidate: rna50232.p1, not unique enough\n",
      "-skipping training candidate: rna681.p1, not unique enough\n",
      "-skipping training candidate: rna6614.p1, not unique enough\n",
      "-skipping training candidate: rna6298.p1, not unique enough\n",
      "-skipping training candidate: rna6726.p1, not unique enough\n",
      "-skipping training candidate: rna9304.p1, not unique enough\n",
      "-skipping training candidate: rna19109.p1, not unique enough\n",
      "-skipping training candidate: rna36000.p1, not unique enough\n",
      "-skipping training candidate: rna40672.p1, not unique enough\n",
      "-skipping training candidate: rna49374.p1, not unique enough\n",
      "-skipping training candidate: rna54385.p1, not unique enough\n",
      "-skipping training candidate: rna1699.p1, not unique enough\n",
      "-skipping training candidate: rna19963.p1, not unique enough\n",
      "-skipping training candidate: rna19964.p1, not unique enough\n",
      "-skipping training candidate: rna19474.p1, not unique enough\n",
      "-skipping training candidate: rna24344.p1, not unique enough\n",
      "-skipping training candidate: rna24349.p1, not unique enough\n",
      "-skipping training candidate: rna24345.p1, not unique enough\n",
      "-skipping training candidate: rna24355.p1, not unique enough\n",
      "-skipping training candidate: rna34839.p1, not unique enough\n",
      "-skipping training candidate: rna55851.p1, not unique enough\n",
      "-skipping training candidate: rna55852.p1, not unique enough\n",
      "-skipping training candidate: rna55862.p1, not unique enough\n",
      "-skipping training candidate: rna55856.p1, not unique enough\n",
      "-skipping training candidate: rna60529.p1, not unique enough\n",
      "-skipping training candidate: rna24351.p1, not unique enough\n",
      "-skipping training candidate: rna35471.p1, not unique enough\n",
      "-skipping training candidate: rna35472.p1, not unique enough\n",
      "-skipping training candidate: rna35474.p1, not unique enough\n",
      "-skipping training candidate: rna41923.p1, not unique enough\n",
      "-skipping training candidate: rna45693.p1, not unique enough\n",
      "-skipping training candidate: rna55858.p1, not unique enough\n",
      "-skipping training candidate: rna60528.p1, not unique enough\n",
      "-skipping training candidate: rna60690.p1, not unique enough\n",
      "-skipping training candidate: rna27469.p1, not unique enough\n",
      "-skipping training candidate: rna34490.p1, not unique enough\n",
      "-skipping training candidate: rna33941.p1, not unique enough\n",
      "-skipping training candidate: rna39893.p1, not unique enough\n",
      "-skipping training candidate: rna48040.p1, not unique enough\n",
      "-skipping training candidate: rna49470.p1, not unique enough\n",
      "-skipping training candidate: rna57158.p1, not unique enough\n",
      "-skipping training candidate: rna57155.p1, not unique enough\n",
      "-skipping training candidate: rna60737.p1, not unique enough\n",
      "-skipping training candidate: rna445.p1, not unique enough\n",
      "-skipping training candidate: rna21992.p1, not unique enough\n",
      "-skipping training candidate: rna27468.p1, not unique enough\n",
      "-skipping training candidate: rna33960.p1, not unique enough\n",
      "-skipping training candidate: rna34635.p1, not unique enough\n",
      "-skipping training candidate: rna34218.p1, not unique enough\n",
      "-skipping training candidate: rna40406.p1, not unique enough\n",
      "-skipping training candidate: rna48396.p1, not unique enough\n",
      "-skipping training candidate: rna48041.p1, not unique enough\n",
      "-skipping training candidate: rna53762.p1, not unique enough\n",
      "-skipping training candidate: rna57156.p1, not unique enough\n",
      "-skipping training candidate: rna60740.p1, not unique enough\n",
      "-skipping training candidate: rna25836.p1, not unique enough\n",
      "-skipping training candidate: rna59212.p1, not unique enough\n",
      "-skipping training candidate: rna61615.p1, not unique enough\n",
      "-skipping training candidate: rna60693.p1, not unique enough\n",
      "-skipping training candidate: rna163.p1, not unique enough\n",
      "-skipping training candidate: rna28447.p1, not unique enough\n",
      "-skipping training candidate: rna31200.p1, not unique enough\n",
      "-skipping training candidate: rna36624.p1, not unique enough\n",
      "-skipping training candidate: rna39577.p1, not unique enough\n",
      "-skipping training candidate: rna29468.p1, not unique enough\n",
      "-skipping training candidate: rna58565.p1, not unique enough\n",
      "-skipping training candidate: rna660.p1, not unique enough\n",
      "-skipping training candidate: rna4131.p1, not unique enough\n",
      "-skipping training candidate: rna21908.p1, not unique enough\n",
      "-skipping training candidate: rna29642.p1, not unique enough\n",
      "-skipping training candidate: rna38229.p1, not unique enough\n",
      "-skipping training candidate: rna48286.p1, not unique enough\n",
      "-skipping training candidate: rna48287.p1, not unique enough\n",
      "-skipping training candidate: rna57640.p1, not unique enough\n",
      "-skipping training candidate: rna19111.p1, not unique enough\n",
      "-skipping training candidate: rna24353.p1, not unique enough\n",
      "-skipping training candidate: rna24347.p1, not unique enough\n",
      "-skipping training candidate: rna31560.p1, not unique enough\n",
      "-skipping training candidate: rna37621.p1, not unique enough\n",
      "-skipping training candidate: rna47653.p1, not unique enough\n",
      "-skipping training candidate: rna47652.p1, not unique enough\n",
      "-skipping training candidate: rna55860.p1, not unique enough\n",
      "-skipping training candidate: rna55854.p1, not unique enough\n",
      "-skipping training candidate: rna58566.p1, not unique enough\n",
      "-skipping training candidate: rna60692.p1, not unique enough\n",
      "-skipping training candidate: rna60807.p1, not unique enough\n",
      "-skipping training candidate: rna19476.p1, not unique enough\n",
      "-skipping training candidate: rna31199.p1, not unique enough\n",
      "-skipping training candidate: rna32659.p1, not unique enough\n",
      "-skipping training candidate: rna34638.p1, not unique enough\n",
      "-skipping training candidate: rna37926.p1, not unique enough\n",
      "-skipping training candidate: rna52590.p1, not unique enough\n",
      "-skipping training candidate: rna6364.p1, not unique enough\n",
      "-skipping training candidate: rna35015.p1, not unique enough\n",
      "-skipping training candidate: rna40404.p1, not unique enough\n",
      "-skipping training candidate: rna50249.p1, not unique enough\n",
      "-skipping training candidate: rna57154.p1, not unique enough\n",
      "-skipping training candidate: rna31201.p1, not unique enough\n",
      "-skipping training candidate: rna33462.p1, not unique enough\n",
      "-skipping training candidate: rna52588.p1, not unique enough\n",
      "-skipping training candidate: rna3944.p1, not unique enough\n",
      "-skipping training candidate: rna7599.p1, not unique enough\n",
      "-skipping training candidate: rna6365.p1, not unique enough\n",
      "-skipping training candidate: rna19110.p1, not unique enough\n",
      "-skipping training candidate: rna25240.p1, not unique enough\n",
      "-skipping training candidate: rna27467.p1, not unique enough\n",
      "-skipping training candidate: rna29702.p1, not unique enough\n",
      "-skipping training candidate: rna33697.p1, not unique enough\n",
      "-skipping training candidate: rna36626.p1, not unique enough\n",
      "-skipping training candidate: rna48946.p1, not unique enough\n",
      "-skipping training candidate: rna444.p1, not unique enough\n",
      "-skipping training candidate: rna18054.p1, not unique enough\n",
      "-skipping training candidate: rna9763.p1, not unique enough\n",
      "-skipping training candidate: rna30779.p1, not unique enough\n",
      "-skipping training candidate: rna30777.p1, not unique enough\n",
      "-skipping training candidate: rna30626.p1, not unique enough\n",
      "-skipping training candidate: rna30624.p1, not unique enough\n",
      "-skipping training candidate: rna31202.p1, not unique enough\n",
      "-skipping training candidate: rna29902.p1, not unique enough\n",
      "-skipping training candidate: rna62578.p1, not unique enough\n",
      "-skipping training candidate: rna447.p1, not unique enough\n",
      "-skipping training candidate: rna8521.p1, not unique enough\n",
      "-skipping training candidate: rna9296.p1, not unique enough\n",
      "-skipping training candidate: rna25205.p1, not unique enough\n",
      "-skipping training candidate: rna30032.p1, not unique enough\n",
      "-skipping training candidate: rna30623.p1, not unique enough\n",
      "-skipping training candidate: rna30622.p1, not unique enough\n",
      "-skipping training candidate: rna32944.p1, not unique enough\n",
      "-skipping training candidate: rna31650.p1, not unique enough\n",
      "-skipping training candidate: rna42571.p1, not unique enough\n",
      "-skipping training candidate: rna45691.p1, not unique enough\n",
      "-skipping training candidate: rna59809.p1, not unique enough\n",
      "-skipping training candidate: rna7601.p1, not unique enough\n",
      "-skipping training candidate: rna22528.p1, not unique enough\n",
      "-skipping training candidate: rna36517.p1, not unique enough\n",
      "-skipping training candidate: rna37049.p1, not unique enough\n",
      "-skipping training candidate: rna53329.p1, not unique enough\n",
      "-skipping training candidate: rna30302.p1, not unique enough\n",
      "-skipping training candidate: rna35448.p1, not unique enough\n",
      "-skipping training candidate: rna40403.p1, not unique enough\n",
      "-skipping training candidate: rna40673.p1, not unique enough\n",
      "-skipping training candidate: rna51927.p1, not unique enough\n",
      "-skipping training candidate: rna62581.p1, not unique enough\n",
      "-skipping training candidate: rna6751.p1, not unique enough\n",
      "-skipping training candidate: rna30301.p1, not unique enough\n",
      "-skipping training candidate: rna38258.p1, not unique enough\n",
      "-skipping training candidate: rna46567.p1, not unique enough\n",
      "-skipping training candidate: rna45579.p1, not unique enough\n",
      "-skipping training candidate: rna54203.p1, not unique enough\n",
      "-skipping training candidate: rna1793.p1, not unique enough\n",
      "-skipping training candidate: rna6630.p1, not unique enough\n",
      "-skipping training candidate: rna40640.p1, not unique enough\n",
      "-skipping training candidate: rna41000.p1, not unique enough\n",
      "-skipping training candidate: rna40998.p1, not unique enough\n",
      "-skipping training candidate: rna40996.p1, not unique enough\n",
      "-skipping training candidate: rna40991.p1, not unique enough\n",
      "-skipping training candidate: rna51923.p1, not unique enough\n",
      "-skipping training candidate: rna61617.p1, not unique enough\n",
      "-skipping training candidate: rna35014.p1, not unique enough\n",
      "-skipping training candidate: rna43208.p1, not unique enough\n",
      "-skipping training candidate: rna8520.p1, not unique enough\n",
      "-skipping training candidate: rna10020.p1, not unique enough\n",
      "-skipping training candidate: rna18289.p1, not unique enough\n",
      "-skipping training candidate: rna28782.p1, not unique enough\n",
      "-skipping training candidate: rna37366.p1, not unique enough\n",
      "-skipping training candidate: rna50840.p1, not unique enough\n",
      "-skipping training candidate: rna60533.p1, not unique enough\n",
      "-skipping training candidate: rna10417.p1, not unique enough\n",
      "-skipping training candidate: rna23825.p1, not unique enough\n",
      "-skipping training candidate: rna26182.p1, not unique enough\n",
      "-skipping training candidate: rna25379.p1, not unique enough\n",
      "-skipping training candidate: rna32054.p1, not unique enough\n",
      "-skipping training candidate: rna34634.p1, not unique enough\n",
      "-skipping training candidate: rna40902.p1, not unique enough\n",
      "-skipping training candidate: rna44456.p1, not unique enough\n",
      "-skipping training candidate: rna44449.p1, not unique enough\n",
      "-skipping training candidate: rna60530.p1, not unique enough\n",
      "-skipping training candidate: rna661.p1, not unique enough\n",
      "-skipping training candidate: rna3803.p1, not unique enough\n",
      "-skipping training candidate: rna6071.p1, not unique enough\n",
      "-skipping training candidate: rna10416.p1, not unique enough\n",
      "-skipping training candidate: rna18177.p1, not unique enough\n",
      "-skipping training candidate: rna9814.p1, not unique enough\n",
      "-skipping training candidate: rna19722.p1, not unique enough\n",
      "-skipping training candidate: rna24467.p1, not unique enough\n",
      "-skipping training candidate: rna30621.p1, not unique enough\n",
      "-skipping training candidate: rna38448.p1, not unique enough\n",
      "-skipping training candidate: rna40901.p1, not unique enough\n",
      "-skipping training candidate: rna42873.p1, not unique enough\n",
      "-skipping training candidate: rna42874.p1, not unique enough\n",
      "-skipping training candidate: rna44448.p1, not unique enough\n",
      "-skipping training candidate: rna44455.p1, not unique enough\n",
      "-skipping training candidate: rna44454.p1, not unique enough\n",
      "-skipping training candidate: rna49373.p1, not unique enough\n",
      "-skipping training candidate: rna3942.p1, not unique enough\n",
      "-skipping training candidate: rna20476.p1, not unique enough\n",
      "-skipping training candidate: rna20477.p1, not unique enough\n",
      "-skipping training candidate: rna26606.p1, not unique enough\n",
      "-skipping training candidate: rna26605.p1, not unique enough\n",
      "-skipping training candidate: rna40602.p1, not unique enough\n",
      "-skipping training candidate: rna44452.p1, not unique enough\n",
      "-skipping training candidate: rna8723.p1, not unique enough\n",
      "-skipping training candidate: rna35783.p1, not unique enough\n",
      "-skipping training candidate: rna47655.p1, not unique enough\n",
      "-skipping training candidate: rna60527.p1, not unique enough\n",
      "-skipping training candidate: rna3580.p1, not unique enough\n",
      "-skipping training candidate: rna24356.p1, not unique enough\n",
      "-skipping training candidate: rna24350.p1, not unique enough\n",
      "-skipping training candidate: rna32229.p1, not unique enough\n",
      "-skipping training candidate: rna40407.p1, not unique enough\n",
      "-skipping training candidate: rna40256.p1, not unique enough\n",
      "-skipping training candidate: rna45969.p1, not unique enough\n",
      "-skipping training candidate: rna45669.p1, not unique enough\n",
      "-skipping training candidate: rna55857.p1, not unique enough\n",
      "-skipping training candidate: rna55863.p1, not unique enough\n",
      "-skipping training candidate: rna8725.p1, not unique enough\n",
      "-skipping training candidate: rna19475.p1, not unique enough\n",
      "-skipping training candidate: rna40257.p1, not unique enough\n",
      "-skipping training candidate: rna58568.p1, not unique enough\n",
      "-skipping training candidate: rna58569.p1, not unique enough\n",
      "-skipping training candidate: rna58562.p1, not unique enough\n",
      "-skipping training candidate: rna4123.p1, not unique enough\n",
      "-skipping training candidate: rna6626.p1, not unique enough\n",
      "-skipping training candidate: rna9295.p1, not unique enough\n",
      "-skipping training candidate: rna31563.p1, not unique enough\n",
      "-skipping training candidate: rna36708.p1, not unique enough\n",
      "-skipping training candidate: rna40398.p1, not unique enough\n",
      "-skipping training candidate: rna42570.p1, not unique enough\n",
      "-skipping training candidate: rna45942.p1, not unique enough\n",
      "-skipping training candidate: rna46505.p1, not unique enough\n",
      "-skipping training candidate: rna59800.p1, not unique enough\n",
      "-skipping training candidate: rna19113.p1, not unique enough\n",
      "-skipping training candidate: rna20125.p1, not unique enough\n",
      "-skipping training candidate: rna29595.p1, not unique enough\n",
      "-skipping training candidate: rna4399.p1, not unique enough\n",
      "-skipping training candidate: rna27708.p1, not unique enough\n",
      "-skipping training candidate: rna41964.p1, not unique enough\n",
      "-skipping training candidate: rna41963.p1, not unique enough\n",
      "-skipping training candidate: rna45690.p1, not unique enough\n",
      "-skipping training candidate: rna50250.p1, not unique enough\n",
      "-skipping training candidate: rna1792.p1, not unique enough\n",
      "-skipping training candidate: rna6628.p1, not unique enough\n",
      "-skipping training candidate: rna18523.p1, not unique enough\n",
      "-skipping training candidate: rna20123.p1, not unique enough\n",
      "-skipping training candidate: rna26170.p1, not unique enough\n",
      "-skipping training candidate: rna35352.p1, not unique enough\n",
      "-skipping training candidate: rna42347.p1, not unique enough\n",
      "-skipping training candidate: rna42349.p1, not unique enough\n",
      "-skipping training candidate: rna42350.p1, not unique enough\n",
      "-skipping training candidate: rna42348.p1, not unique enough\n",
      "-skipping training candidate: rna57447.p1, not unique enough\n",
      "-skipping training candidate: rna60601.p1, not unique enough\n",
      "-skipping training candidate: rna60694.p1, not unique enough\n",
      "-skipping training candidate: rna8522.p1, not unique enough\n",
      "-skipping training candidate: rna25235.p1, not unique enough\n",
      "-skipping training candidate: rna25378.p1, not unique enough\n",
      "-skipping training candidate: rna30388.p1, not unique enough\n",
      "-skipping training candidate: rna42884.p1, not unique enough\n",
      "-skipping training candidate: rna62573.p1, not unique enough\n",
      "-skipping training candidate: rna874.p1, not unique enough\n",
      "-skipping training candidate: rna7332.p1, not unique enough\n",
      "-skipping training candidate: rna20479.p1, not unique enough\n",
      "-skipping training candidate: rna34428.p1, not unique enough\n",
      "-skipping training candidate: rna35996.p1, not unique enough\n",
      "-skipping training candidate: rna36687.p1, not unique enough\n",
      "-skipping training candidate: rna47654.p1, not unique enough\n",
      "-skipping training candidate: rna53861.p1, not unique enough\n",
      "-skipping training candidate: rna58564.p1, not unique enough\n",
      "-skipping training candidate: rna61546.p1, not unique enough\n",
      "-skipping training candidate: rna3480.p1, not unique enough\n",
      "-skipping training candidate: rna20119.p1, not unique enough\n",
      "-skipping training candidate: rna22081.p1, not unique enough\n",
      "-skipping training candidate: rna24348.p1, not unique enough\n",
      "-skipping training candidate: rna25687.p1, not unique enough\n",
      "-skipping training candidate: rna33948.p1, not unique enough\n",
      "-skipping training candidate: rna55855.p1, not unique enough\n",
      "-skipping training candidate: gene1377.p1, not unique enough\n",
      "-skipping training candidate: rna6627.p1, not unique enough\n",
      "-skipping training candidate: rna20122.p1, not unique enough\n",
      "-skipping training candidate: rna23740.p1, not unique enough\n",
      "-skipping training candidate: rna22078.p1, not unique enough\n",
      "-skipping training candidate: rna25686.p1, not unique enough\n",
      "-skipping training candidate: rna25683.p1, not unique enough\n",
      "-skipping training candidate: rna42353.p1, not unique enough\n",
      "-skipping training candidate: rna42352.p1, not unique enough\n",
      "-skipping training candidate: rna49375.p1, not unique enough\n",
      "-skipping training candidate: rna58567.p1, not unique enough\n",
      "-skipping training candidate: rna2809.p1, not unique enough\n",
      "-skipping training candidate: rna20118.p1, not unique enough\n",
      "-skipping training candidate: rna60132.p1, not unique enough\n",
      "-skipping training candidate: rna20124.p1, not unique enough\n",
      "-skipping training candidate: rna20121.p1, not unique enough\n",
      "-skipping training candidate: rna28762.p1, not unique enough\n",
      "-skipping training candidate: rna36710.p1, not unique enough\n",
      "-skipping training candidate: rna37365.p1, not unique enough\n",
      "-skipping training candidate: rna44414.p1, not unique enough\n",
      "-skipping training candidate: rna61614.p1, not unique enough\n",
      "-skipping training candidate: rna3974.p1, not unique enough\n",
      "-skipping training candidate: rna34379.p1, not unique enough\n",
      "-skipping training candidate: rna44983.p1, not unique enough\n",
      "-skipping training candidate: rna46020.p1, not unique enough\n",
      "-skipping training candidate: rna2898.p1, not unique enough\n",
      "-skipping training candidate: rna2900.p1, not unique enough\n",
      "-skipping training candidate: rna4985.p1, not unique enough\n",
      "-skipping training candidate: rna9762.p1, not unique enough\n",
      "-skipping training candidate: rna18455.p1, not unique enough\n",
      "-skipping training candidate: rna57446.p1, not unique enough\n",
      "-skipping training candidate: rna2899.p1, not unique enough\n",
      "-skipping training candidate: rna20117.p1, not unique enough\n",
      "-skipping training candidate: rna21644.p1, not unique enough\n",
      "-skipping training candidate: rna25931.p1, not unique enough\n",
      "-skipping training candidate: rna32941.p1, not unique enough\n",
      "-skipping training candidate: rna44412.p1, not unique enough\n",
      "-skipping training candidate: rna44410.p1, not unique enough\n",
      "-skipping training candidate: rna3975.p1, not unique enough\n",
      "-skipping training candidate: rna23320.p1, not unique enough\n",
      "-skipping training candidate: rna22079.p1, not unique enough\n",
      "-skipping training candidate: rna39811.p1, not unique enough\n",
      "-skipping training candidate: rna44413.p1, not unique enough\n",
      "-skipping training candidate: rna48290.p1, not unique enough\n",
      "-skipping training candidate: rna20126.p1, not unique enough\n",
      "-skipping training candidate: rna41053.p1, not unique enough\n",
      "-skipping training candidate: rna55895.p1, not unique enough\n",
      "-skipping training candidate: rna61613.p1, not unique enough\n",
      "-skipping training candidate: rna205.p1, not unique enough\n",
      "-skipping training candidate: rna9294.p1, not unique enough\n",
      "-skipping training candidate: rna18373.p1, not unique enough\n",
      "-skipping training candidate: rna26293.p1, not unique enough\n",
      "-skipping training candidate: rna28763.p1, not unique enough\n",
      "-skipping training candidate: rna29640.p1, not unique enough\n",
      "-skipping training candidate: rna46908.p1, not unique enough\n",
      "-skipping training candidate: rna52587.p1, not unique enough\n",
      "-skipping training candidate: rna53626.p1, not unique enough\n",
      "-skipping training candidate: rna60696.p1, not unique enough\n",
      "-skipping training candidate: rna58241.p1, not unique enough\n",
      "-skipping training candidate: rna61618.p1, not unique enough\n",
      "-skipping training candidate: rna2348.p1, not unique enough\n",
      "-skipping training candidate: rna22568.p1, not unique enough\n",
      "-skipping training candidate: rna30574.p1, not unique enough\n",
      "-skipping training candidate: rna30303.p1, not unique enough\n",
      "-skipping training candidate: rna32818.p1, not unique enough\n",
      "-skipping training candidate: rna34378.p1, not unique enough\n",
      "-skipping training candidate: rna40633.p1, not unique enough\n",
      "-skipping training candidate: rna9718.p1, not unique enough\n",
      "-skipping training candidate: rna33949.p1, not unique enough\n",
      "-skipping training candidate: rna2896.p1, not unique enough\n",
      "-skipping training candidate: rna4158.p1, not unique enough\n",
      "-skipping training candidate: rna17939.p1, not unique enough\n",
      "-skipping training candidate: rna9716.p1, not unique enough\n",
      "-skipping training candidate: rna34377.p1, not unique enough\n",
      "-skipping training candidate: rna35825.p1, not unique enough\n",
      "-skipping training candidate: rna44411.p1, not unique enough\n",
      "-skipping training candidate: rna59893.p1, not unique enough\n",
      "-skipping training candidate: rna58570.p1, not unique enough\n",
      "-skipping training candidate: rna2765.p1, not unique enough\n",
      "-skipping training candidate: rna2897.p1, not unique enough\n",
      "-skipping training candidate: rna2865.p1, not unique enough\n",
      "-skipping training candidate: rna32307.p1, not unique enough\n",
      "-skipping training candidate: rna47348.p1, not unique enough\n",
      "-skipping training candidate: rna53862.p1, not unique enough\n",
      "-skipping training candidate: rna31354.p1, not unique enough\n",
      "-skipping training candidate: rna56724.p1, not unique enough\n",
      "-skipping training candidate: rna60534.p1, not unique enough\n",
      "-skipping training candidate: rna3582.p1, not unique enough\n",
      "-skipping training candidate: rna20597.p1, not unique enough\n",
      "-skipping training candidate: rna30577.p1, not unique enough\n",
      "-skipping training candidate: rna31373.p1, not unique enough\n",
      "-skipping training candidate: rna29639.p1, not unique enough\n",
      "-skipping training candidate: rna35946.p1, not unique enough\n",
      "-skipping training candidate: rna37066.p1, not unique enough\n",
      "-skipping training candidate: rna40682.p1, not unique enough\n",
      "-skipping training candidate: rna44416.p1, not unique enough\n",
      "-skipping training candidate: rna48998.p1, not unique enough\n",
      "-skipping training candidate: rna25787.p1, not unique enough\n",
      "-skipping training candidate: rna24346.p1, not unique enough\n",
      "-skipping training candidate: rna55853.p1, not unique enough\n",
      "-skipping training candidate: rna25234.p1, not unique enough\n",
      "-skipping training candidate: rna25232.p1, not unique enough\n",
      "-skipping training candidate: rna25233.p1, not unique enough\n",
      "-skipping training candidate: rna24352.p1, not unique enough\n",
      "-skipping training candidate: rna41924.p1, not unique enough\n",
      "-skipping training candidate: rna55859.p1, not unique enough\n",
      "-skipping training candidate: rna6362.p1, not unique enough\n",
      "-skipping training candidate: rna8025.p1, not unique enough\n",
      "-skipping training candidate: rna9717.p1, not unique enough\n",
      "-skipping training candidate: rna20480.p1, not unique enough\n",
      "-skipping training candidate: rna20618.p1, not unique enough\n",
      "-skipping training candidate: rna23204.p1, not unique enough\n",
      "-skipping training candidate: rna24989.p1, not unique enough\n",
      "-skipping training candidate: rna43189.p1, not unique enough\n",
      "-skipping training candidate: rna53763.p1, not unique enough\n",
      "-skipping training candidate: rna56859.p1, not unique enough\n",
      "-skipping training candidate: rna59042.p1, not unique enough\n",
      "-skipping training candidate: rna59812.p1, not unique enough\n",
      "-skipping training candidate: rna20620.p1, not unique enough\n",
      "-skipping training candidate: rna25933.p1, not unique enough\n",
      "-skipping training candidate: rna25685.p1, not unique enough\n",
      "-skipping training candidate: rna25682.p1, not unique enough\n",
      "-skipping training candidate: rna29329.p1, not unique enough\n",
      "-skipping training candidate: rna29645.p1, not unique enough\n",
      "-skipping training candidate: rna41925.p1, not unique enough\n",
      "-skipping training candidate: rna44453.p1, not unique enough\n",
      "-skipping training candidate: rna50235.p1, not unique enough\n",
      "-skipping training candidate: rna50234.p1, not unique enough\n",
      "-skipping training candidate: rna59811.p1, not unique enough\n",
      "-skipping training candidate: rna59810.p1, not unique enough\n",
      "-skipping training candidate: rna59853.p1, not unique enough\n",
      "-skipping training candidate: rna2424.p1, not unique enough\n",
      "-skipping training candidate: rna1876.p1, not unique enough\n",
      "-skipping training candidate: rna9816.p1, not unique enough\n",
      "-skipping training candidate: rna9715.p1, not unique enough\n",
      "-skipping training candidate: rna20622.p1, not unique enough\n",
      "-skipping training candidate: rna25684.p1, not unique enough\n",
      "-skipping training candidate: rna34847.p1, not unique enough\n",
      "-skipping training candidate: rna39894.p1, not unique enough\n",
      "-skipping training candidate: rna39659.p1, not unique enough\n",
      "-skipping training candidate: rna48787.p1, not unique enough\n",
      "-skipping training candidate: rna55783.p1, not unique enough\n",
      "-skipping training candidate: rna58318.p1, not unique enough\n",
      "-skipping training candidate: rna2427.p1, not unique enough\n",
      "-skipping training candidate: rna2426.p1, not unique enough\n",
      "-skipping training candidate: rna1877.p1, not unique enough\n",
      "-skipping training candidate: rna34976.p1, not unique enough\n",
      "-skipping training candidate: rna40400.p1, not unique enough\n",
      "-skipping training candidate: rna42029.p1, not unique enough\n",
      "-skipping training candidate: rna41156.p1, not unique enough\n",
      "-skipping training candidate: rna44451.p1, not unique enough\n",
      "-skipping training candidate: rna45939.p1, not unique enough\n",
      "-skipping training candidate: rna47154.p1, not unique enough\n",
      "-skipping training candidate: rna48291.p1, not unique enough\n",
      "-skipping training candidate: rna48288.p1, not unique enough\n",
      "-skipping training candidate: rna52654.p1, not unique enough\n",
      "-skipping training candidate: rna59799.p1, not unique enough\n",
      "-skipping training candidate: rna18155.p1, not unique enough\n",
      "-skipping training candidate: rna18154.p1, not unique enough\n",
      "-skipping training candidate: rna18468.p1, not unique enough\n",
      "-skipping training candidate: rna25846.p1, not unique enough\n",
      "-skipping training candidate: rna27163.p1, not unique enough\n",
      "-skipping training candidate: rna33698.p1, not unique enough\n",
      "-skipping training candidate: rna38447.p1, not unique enough\n",
      "-skipping training candidate: rna44450.p1, not unique enough\n",
      "-skipping training candidate: rna49301.p1, not unique enough\n",
      "-skipping training candidate: rna3579.p1, not unique enough\n",
      "-skipping training candidate: rna20032.p1, not unique enough\n",
      "-skipping training candidate: rna18466.p1, not unique enough\n",
      "-skipping training candidate: rna26639.p1, not unique enough\n",
      "-skipping training candidate: rna33699.p1, not unique enough\n",
      "-skipping training candidate: rna34754.p1, not unique enough\n",
      "-skipping training candidate: rna34757.p1, not unique enough\n",
      "-skipping training candidate: rna41351.p1, not unique enough\n",
      "-skipping training candidate: rna42377.p1, not unique enough\n",
      "-skipping training candidate: rna42376.p1, not unique enough\n",
      "-skipping training candidate: rna63032.p1, not unique enough\n",
      "-skipping training candidate: rna13055.p1, not unique enough\n",
      "-skipping training candidate: rna20420.p1, not unique enough\n",
      "-skipping training candidate: rna20419.p1, not unique enough\n",
      "-skipping training candidate: rna24354.p1, not unique enough\n",
      "-skipping training candidate: rna24357.p1, not unique enough\n",
      "-skipping training candidate: rna48999.p1, not unique enough\n",
      "-skipping training candidate: rna48167.p1, not unique enough\n",
      "-skipping training candidate: rna48037.p1, not unique enough\n",
      "-skipping training candidate: rna51940.p1, not unique enough\n",
      "-skipping training candidate: rna51943.p1, not unique enough\n",
      "-skipping training candidate: rna55861.p1, not unique enough\n",
      "-skipping training candidate: rna55864.p1, not unique enough\n",
      "-skipping training candidate: rna2767.p1, not unique enough\n",
      "-skipping training candidate: rna30166.p1, not unique enough\n",
      "-skipping training candidate: rna29901.p1, not unique enough\n",
      "-skipping training candidate: rna38514.p1, not unique enough\n",
      "-skipping training candidate: rna55784.p1, not unique enough\n",
      "-skipping training candidate: rna57051.p1, not unique enough\n",
      "-skipping training candidate: rna58629.p1, not unique enough\n",
      "-skipping training candidate: rna58628.p1, not unique enough\n",
      "-skipping training candidate: rna60661.p1, not unique enough\n",
      "-skipping training candidate: rna121.p1, not unique enough\n",
      "-skipping training candidate: rna3979.p1, not unique enough\n",
      "-skipping training candidate: rna29644.p1, not unique enough\n",
      "-skipping training candidate: rna34978.p1, not unique enough\n",
      "-skipping training candidate: rna42399.p1, not unique enough\n",
      "-skipping training candidate: rna42573.p1, not unique enough\n",
      "-skipping training candidate: rna45940.p1, not unique enough\n",
      "-skipping training candidate: rna49943.p1, not unique enough\n",
      "-skipping training candidate: rna49937.p1, not unique enough\n",
      "-skipping training candidate: rna49936.p1, not unique enough\n",
      "-skipping training candidate: rna52389.p1, not unique enough\n",
      "-skipping training candidate: rna62574.p1, not unique enough\n",
      "-skipping training candidate: rna27857.p1, not unique enough\n",
      "-skipping training candidate: rna33165.p1, not unique enough\n",
      "-skipping training candidate: rna34975.p1, not unique enough\n",
      "-skipping training candidate: rna46504.p1, not unique enough\n",
      "-skipping training candidate: rna49944.p1, not unique enough\n",
      "-skipping training candidate: rna50876.p1, not unique enough\n",
      "-skipping training candidate: rna60741.p1, not unique enough\n",
      "-skipping training candidate: rna3578.p1, not unique enough\n",
      "-skipping training candidate: rna19798.p1, not unique enough\n",
      "-skipping training candidate: rna32751.p1, not unique enough\n",
      "-skipping training candidate: rna32759.p1, not unique enough\n",
      "-skipping training candidate: rna42070.p1, not unique enough\n",
      "-skipping training candidate: rna41970.p1, not unique enough\n",
      "-skipping training candidate: rna56725.p1, not unique enough\n",
      "-skipping training candidate: rna58316.p1, not unique enough\n",
      "-skipping training candidate: rna60695.p1, not unique enough\n",
      "-skipping training candidate: rna61058.p1, not unique enough\n",
      "-skipping training candidate: rna1879.p1, not unique enough\n",
      "-skipping training candidate: rna9813.p1, not unique enough\n",
      "-skipping training candidate: rna24938.p1, not unique enough\n",
      "-skipping training candidate: rna24937.p1, not unique enough\n",
      "-skipping training candidate: rna24939.p1, not unique enough\n",
      "-skipping training candidate: rna48258.p1, not unique enough\n",
      "-skipping training candidate: rna9631.p1, not unique enough\n",
      "-skipping training candidate: rna20475.p1, not unique enough\n",
      "-skipping training candidate: rna25204.p1, not unique enough\n",
      "-skipping training candidate: rna33164.p1, not unique enough\n",
      "-skipping training candidate: rna42572.p1, not unique enough\n",
      "-skipping training candidate: rna47869.p1, not unique enough\n",
      "-skipping training candidate: rna53863.p1, not unique enough\n",
      "-skipping training candidate: rna59855.p1, not unique enough\n",
      "-skipping training candidate: rna188.p1, not unique enough\n",
      "-skipping training candidate: rna24663.p1, not unique enough\n",
      "-skipping training candidate: rna30576.p1, not unique enough\n",
      "-skipping training candidate: rna33167.p1, not unique enough\n",
      "-skipping training candidate: rna47656.p1, not unique enough\n",
      "-skipping training candidate: rna22672.p1, not unique enough\n",
      "-skipping training candidate: rna39895.p1, not unique enough\n",
      "-skipping training candidate: rna43210.p1, not unique enough\n",
      "-skipping training candidate: rna46363.p1, not unique enough\n",
      "-skipping training candidate: rna53253.p1, not unique enough\n",
      "-skipping training candidate: rna59854.p1, not unique enough\n",
      "-skipping training candidate: rna60662.p1, not unique enough\n",
      "-skipping training candidate: rna8038.p1, not unique enough\n",
      "-skipping training candidate: rna20977.p1, not unique enough\n",
      "-skipping training candidate: rna23351.p1, not unique enough\n",
      "-skipping training candidate: rna22311.p1, not unique enough\n",
      "-skipping training candidate: rna29944.p1, not unique enough\n",
      "-skipping training candidate: rna31356.p1, not unique enough\n",
      "-skipping training candidate: rna48853.p1, not unique enough\n",
      "-skipping training candidate: rna60077.p1, not unique enough\n",
      "-skipping training candidate: rna9720.p1, not unique enough\n",
      "-skipping training candidate: rna18850.p1, not unique enough\n",
      "-skipping training candidate: rna18851.p1, not unique enough\n",
      "-skipping training candidate: rna27318.p1, not unique enough\n",
      "-skipping training candidate: rna27319.p1, not unique enough\n",
      "-skipping training candidate: rna27311.p1, not unique enough\n",
      "-skipping training candidate: rna30954.p1, not unique enough\n",
      "-skipping training candidate: rna31066.p1, not unique enough\n",
      "-skipping training candidate: rna33012.p1, not unique enough\n",
      "-skipping training candidate: rna34880.p1, not unique enough\n",
      "-skipping training candidate: rna34882.p1, not unique enough\n",
      "-skipping training candidate: rna34879.p1, not unique enough\n",
      "-skipping training candidate: rna39235.p1, not unique enough\n",
      "-skipping training candidate: rna45936.p1, not unique enough\n",
      "-skipping training candidate: rna49959.p1, not unique enough\n",
      "-skipping training candidate: rna51924.p1, not unique enough\n",
      "-skipping training candidate: rna51941.p1, not unique enough\n",
      "-skipping training candidate: rna1794.p1, not unique enough\n",
      "-skipping training candidate: rna1880.p1, not unique enough\n",
      "-skipping training candidate: rna8036.p1, not unique enough\n",
      "-skipping training candidate: rna8039.p1, not unique enough\n",
      "-skipping training candidate: rna32754.p1, not unique enough\n",
      "-skipping training candidate: rna38517.p1, not unique enough\n",
      "-skipping training candidate: rna42267.p1, not unique enough\n",
      "-skipping training candidate: rna42956.p1, not unique enough\n",
      "-skipping training candidate: rna50877.p1, not unique enough\n",
      "-skipping training candidate: rna49376.p1, not unique enough\n",
      "-skipping training candidate: rna51287.p1, not unique enough\n",
      "-skipping training candidate: rna2764.p1, not unique enough\n",
      "-skipping training candidate: rna43525.p1, not unique enough\n",
      "-skipping training candidate: rna43526.p1, not unique enough\n",
      "-skipping training candidate: rna47451.p1, not unique enough\n",
      "-skipping training candidate: rna50404.p1, not unique enough\n",
      "-skipping training candidate: rna3586.p1, not unique enough\n",
      "-skipping training candidate: rna6974.p1, not unique enough\n",
      "-skipping training candidate: rna5441.p1, not unique enough\n",
      "-skipping training candidate: rna38516.p1, not unique enough\n",
      "-skipping training candidate: rna38228.p1, not unique enough\n",
      "-skipping training candidate: rna44408.p1, not unique enough\n",
      "-skipping training candidate: rna48849.p1, not unique enough\n",
      "-skipping training candidate: rna47452.p1, not unique enough\n",
      "-skipping training candidate: rna60532.p1, not unique enough\n",
      "-skipping training candidate: rna25844.p1, not unique enough\n",
      "-skipping training candidate: rna32752.p1, not unique enough\n",
      "-skipping training candidate: rna39662.p1, not unique enough\n",
      "-skipping training candidate: rna57448.p1, not unique enough\n",
      "-skipping training candidate: rna58254.p1, not unique enough\n",
      "-skipping training candidate: rna60697.p1, not unique enough\n",
      "-skipping training candidate: rna19341.p1, not unique enough\n",
      "-skipping training candidate: rna21454.p1, not unique enough\n",
      "-skipping training candidate: rna21488.p1, not unique enough\n",
      "-skipping training candidate: rna25842.p1, not unique enough\n",
      "-skipping training candidate: rna44389.p1, not unique enough\n",
      "-skipping training candidate: rna47765.p1, not unique enough\n",
      "-skipping training candidate: rna47766.p1, not unique enough\n",
      "-skipping training candidate: rna61059.p1, not unique enough\n",
      "-skipping training candidate: rna736.p1, not unique enough\n",
      "-skipping training candidate: rna25788.p1, not unique enough\n",
      "-skipping training candidate: rna31352.p1, not unique enough\n",
      "-skipping training candidate: rna33344.p1, not unique enough\n",
      "-skipping training candidate: rna32764.p1, not unique enough\n",
      "-skipping training candidate: rna45580.p1, not unique enough\n",
      "-skipping training candidate: rna49535.p1, not unique enough\n",
      "-skipping training candidate: rna4126.p1, not unique enough\n",
      "-skipping training candidate: rna29027.p1, not unique enough\n",
      "-skipping training candidate: rna27691.p1, not unique enough\n",
      "-skipping training candidate: rna35781.p1, not unique enough\n",
      "-skipping training candidate: rna57839.p1, not unique enough\n",
      "-skipping training candidate: rna60526.p1, not unique enough\n",
      "-skipping training candidate: rna33867.p1, not unique enough\n",
      "-skipping training candidate: rna35339.p1, not unique enough\n",
      "-skipping training candidate: rna49000.p1, not unique enough\n",
      "-skipping training candidate: rna47648.p1, not unique enough\n",
      "-skipping training candidate: rna50154.p1, not unique enough\n",
      "-skipping training candidate: rna57118.p1, not unique enough\n",
      "-skipping training candidate: rna22479.p1, not unique enough\n",
      "-skipping training candidate: rna25845.p1, not unique enough\n",
      "-skipping training candidate: rna32943.p1, not unique enough\n",
      "-skipping training candidate: rna32755.p1, not unique enough\n",
      "-skipping training candidate: rna34954.p1, not unique enough\n",
      "-skipping training candidate: rna35349.p1, not unique enough\n",
      "-skipping training candidate: rna35351.p1, not unique enough\n",
      "-skipping training candidate: rna37788.p1, not unique enough\n",
      "-skipping training candidate: rna37294.p1, not unique enough\n",
      "-skipping training candidate: rna42681.p1, not unique enough\n",
      "-skipping training candidate: rna45935.p1, not unique enough\n",
      "-skipping training candidate: rna47469.p1, not unique enough\n",
      "-skipping training candidate: rna55883.p1, not unique enough\n",
      "-skipping training candidate: rna57840.p1, not unique enough\n",
      "-skipping training candidate: rna57085.p1, not unique enough\n",
      "-skipping training candidate: rna57084.p1, not unique enough\n",
      "-skipping training candidate: rna32770.p1, not unique enough\n",
      "-skipping training candidate: rna34657.p1, not unique enough\n",
      "-skipping training candidate: rna38128.p1, not unique enough\n",
      "-skipping training candidate: rna49001.p1, not unique enough\n",
      "-skipping training candidate: rna48168.p1, not unique enough\n",
      "-skipping training candidate: rna50776.p1, not unique enough\n",
      "-skipping training candidate: rna56860.p1, not unique enough\n",
      "-skipping training candidate: rna2393.p1, not unique enough\n",
      "-skipping training candidate: rna2391.p1, not unique enough\n",
      "-skipping training candidate: rna2390.p1, not unique enough\n",
      "-skipping training candidate: rna2381.p1, not unique enough\n",
      "-skipping training candidate: rna2366.p1, not unique enough\n",
      "-skipping training candidate: rna2364.p1, not unique enough\n",
      "-skipping training candidate: rna19462.p1, not unique enough\n",
      "-skipping training candidate: rna19460.p1, not unique enough\n",
      "-skipping training candidate: rna27149.p1, not unique enough\n",
      "-skipping training candidate: rna42883.p1, not unique enough\n",
      "-skipping training candidate: rna57083.p1, not unique enough\n",
      "-skipping training candidate: rna59029.p1, not unique enough\n",
      "-skipping training candidate: rna59028.p1, not unique enough\n",
      "-skipping training candidate: rna62044.p1, not unique enough\n",
      "-skipping training candidate: rna682.p1, not unique enough\n",
      "-skipping training candidate: rna8127.p1, not unique enough\n",
      "-skipping training candidate: rna19725.p1, not unique enough\n",
      "-skipping training candidate: rna19724.p1, not unique enough\n",
      "-skipping training candidate: rna25543.p1, not unique enough\n",
      "-skipping training candidate: rna28486.p1, not unique enough\n",
      "-skipping training candidate: rna32753.p1, not unique enough\n",
      "-skipping training candidate: rna34492.p1, not unique enough\n",
      "-skipping training candidate: rna57116.p1, not unique enough\n",
      "-skipping training candidate: rna60738.p1, not unique enough\n",
      "-skipping training candidate: rna62111.p1, not unique enough\n",
      "-skipping training candidate: rna62108.p1, not unique enough\n",
      "-skipping training candidate: rna62109.p1, not unique enough\n",
      "-skipping training candidate: rna62112.p1, not unique enough\n",
      "-skipping training candidate: rna1579.p1, not unique enough\n",
      "-skipping training candidate: rna6727.p1, not unique enough\n",
      "-skipping training candidate: rna18349.p1, not unique enough\n",
      "-skipping training candidate: rna39233.p1, not unique enough\n",
      "-skipping training candidate: rna52109.p1, not unique enough\n",
      "-skipping training candidate: rna4211.p1, not unique enough\n",
      "-skipping training candidate: rna22675.p1, not unique enough\n",
      "-skipping training candidate: rna32731.p1, not unique enough\n",
      "-skipping training candidate: rna33308.p1, not unique enough\n",
      "-skipping training candidate: rna33312.p1, not unique enough\n",
      "-skipping training candidate: rna34639.p1, not unique enough\n",
      "-skipping training candidate: rna37295.p1, not unique enough\n",
      "-skipping training candidate: rna49933.p1, not unique enough\n",
      "-skipping training candidate: rna60251.p1, not unique enough\n",
      "-skipping training candidate: rna60253.p1, not unique enough\n",
      "-skipping training candidate: rna4319.p1, not unique enough\n",
      "-skipping training candidate: rna19688.p1, not unique enough\n",
      "-skipping training candidate: rna27306.p1, not unique enough\n",
      "-skipping training candidate: rna32732.p1, not unique enough\n",
      "-skipping training candidate: rna33311.p1, not unique enough\n",
      "-skipping training candidate: rna34659.p1, not unique enough\n",
      "-skipping training candidate: rna42643.p1, not unique enough\n",
      "-skipping training candidate: rna41783.p1, not unique enough\n",
      "-skipping training candidate: rna49938.p1, not unique enough\n",
      "-skipping training candidate: rna49954.p1, not unique enough\n",
      "-skipping training candidate: rna6724.p1, not unique enough\n",
      "-skipping training candidate: rna6723.p1, not unique enough\n",
      "-skipping training candidate: rna7108.p1, not unique enough\n",
      "-skipping training candidate: rna7107.p1, not unique enough\n",
      "-skipping training candidate: rna34758.p1, not unique enough\n",
      "-skipping training candidate: rna49945.p1, not unique enough\n",
      "-skipping training candidate: rna49940.p1, not unique enough\n",
      "-skipping training candidate: rna51068.p1, not unique enough\n",
      "-skipping training candidate: rna51066.p1, not unique enough\n",
      "-skipping training candidate: rna56726.p1, not unique enough\n",
      "-skipping training candidate: rna1444.p1, not unique enough\n",
      "-skipping training candidate: rna10258.p1, not unique enough\n",
      "-skipping training candidate: rna23004.p1, not unique enough\n",
      "-skipping training candidate: rna25789.p1, not unique enough\n",
      "-skipping training candidate: rna29929.p1, not unique enough\n",
      "-skipping training candidate: rna32772.p1, not unique enough\n",
      "-skipping training candidate: rna34591.p1, not unique enough\n",
      "-skipping training candidate: rna39815.p1, not unique enough\n",
      "-skipping training candidate: rna42876.p1, not unique enough\n",
      "-skipping training candidate: rna43524.p1, not unique enough\n",
      "-skipping training candidate: rna49946.p1, not unique enough\n",
      "-skipping training candidate: rna1145.p1, not unique enough\n",
      "-skipping training candidate: rna8643.p1, not unique enough\n",
      "-skipping training candidate: rna19339.p1, not unique enough\n",
      "-skipping training candidate: rna23008.p1, not unique enough\n",
      "-skipping training candidate: rna29175.p1, not unique enough\n",
      "-skipping training candidate: rna48850.p1, not unique enough\n",
      "-skipping training candidate: rna47770.p1, not unique enough\n",
      "-skipping training candidate: rna53040.p1, not unique enough\n",
      "-skipping training candidate: rna51928.p1, not unique enough\n",
      "-skipping training candidate: rna54288.p1, not unique enough\n",
      "-skipping training candidate: rna19800.p1, not unique enough\n",
      "-skipping training candidate: rna31228.p1, not unique enough\n",
      "-skipping training candidate: rna30727.p1, not unique enough\n",
      "-skipping training candidate: rna41780.p1, not unique enough\n",
      "-skipping training candidate: rna44409.p1, not unique enough\n",
      "-skipping training candidate: rna46881.p1, not unique enough\n",
      "-skipping training candidate: rna47769.p1, not unique enough\n",
      "-skipping training candidate: rna6728.p1, not unique enough\n",
      "-skipping training candidate: rna41779.p1, not unique enough\n",
      "-skipping training candidate: rna47764.p1, not unique enough\n",
      "-skipping training candidate: rna55881.p1, not unique enough\n",
      "-skipping training candidate: rna18504.p1, not unique enough\n",
      "-skipping training candidate: rna39883.p1, not unique enough\n",
      "-skipping training candidate: rna42511.p1, not unique enough\n",
      "-skipping training candidate: rna47763.p1, not unique enough\n",
      "-skipping training candidate: rna49533.p1, not unique enough\n",
      "-skipping training candidate: rna60664.p1, not unique enough\n",
      "-skipping training candidate: rna29025.p1, not unique enough\n",
      "-skipping training candidate: rna36188.p1, not unique enough\n",
      "-skipping training candidate: rna41776.p1, not unique enough\n",
      "-skipping training candidate: rna55754.p1, not unique enough\n",
      "-skipping training candidate: rna62721.p1, not unique enough\n",
      "-skipping training candidate: rna1881.p1, not unique enough\n",
      "-skipping training candidate: rna18778.p1, not unique enough\n",
      "-skipping training candidate: rna29026.p1, not unique enough\n",
      "-skipping training candidate: rna34756.p1, not unique enough\n",
      "-skipping training candidate: rna39816.p1, not unique enough\n",
      "-skipping training candidate: rna41702.p1, not unique enough\n",
      "-skipping training candidate: rna47657.p1, not unique enough\n",
      "-skipping training candidate: rna49953.p1, not unique enough\n",
      "-skipping training candidate: rna59603.p1, not unique enough\n",
      "-skipping training candidate: rna59602.p1, not unique enough\n",
      "-skipping training candidate: rna61649.p1, not unique enough\n",
      "-skipping training candidate: rna61650.p1, not unique enough\n",
      "-skipping training candidate: rna64398.p1, not unique enough\n",
      "-skipping training candidate: rna3481.p1, not unique enough\n",
      "-skipping training candidate: rna41777.p1, not unique enough\n",
      "-skipping training candidate: rna64399.p1, not unique enough\n",
      "-skipping training candidate: rna6616.p1, not unique enough\n",
      "-skipping training candidate: rna6610.p1, not unique enough\n",
      "-skipping training candidate: rna19709.p1, not unique enough\n",
      "-skipping training candidate: rna22621.p1, not unique enough\n",
      "-skipping training candidate: rna41782.p1, not unique enough\n",
      "-skipping training candidate: rna58139.p1, not unique enough\n",
      "-skipping training candidate: rna58730.p1, not unique enough\n",
      "-skipping training candidate: rna58731.p1, not unique enough\n",
      "-skipping training candidate: rna266.p1, not unique enough\n",
      "-skipping training candidate: rna3512.p1, not unique enough\n",
      "-skipping training candidate: rna19928.p1, not unique enough\n",
      "-skipping training candidate: rna22619.p1, not unique enough\n",
      "-skipping training candidate: rna26534.p1, not unique enough\n",
      "-skipping training candidate: rna29124.p1, not unique enough\n",
      "-skipping training candidate: rna42269.p1, not unique enough\n",
      "-skipping training candidate: rna41781.p1, not unique enough\n",
      "-skipping training candidate: rna59220.p1, not unique enough\n",
      "-skipping training candidate: rna59267.p1, not unique enough\n",
      "-skipping training candidate: rna5604.p1, not unique enough\n",
      "-skipping training candidate: rna9099.p1, not unique enough\n",
      "-skipping training candidate: rna27300.p1, not unique enough\n",
      "-skipping training candidate: rna30618.p1, not unique enough\n",
      "-skipping training candidate: rna32756.p1, not unique enough\n",
      "-skipping training candidate: rna34517.p1, not unique enough\n",
      "-skipping training candidate: rna40675.p1, not unique enough\n",
      "-skipping training candidate: rna41775.p1, not unique enough\n",
      "-skipping training candidate: rna50001.p1, not unique enough\n",
      "-skipping training candidate: rna57583.p1, not unique enough\n",
      "-skipping training candidate: rna61108.p1, not unique enough\n",
      "-skipping training candidate: rna334.p1, not unique enough\n",
      "-skipping training candidate: rna16022.p1, not unique enough\n",
      "-skipping training candidate: rna35553.p1, not unique enough\n",
      "-skipping training candidate: rna44755.p1, not unique enough\n",
      "-skipping training candidate: rna54342.p1, not unique enough\n",
      "-skipping training candidate: rna2741.p1, not unique enough\n",
      "-skipping training candidate: rna7537.p1, not unique enough\n",
      "-skipping training candidate: rna7539.p1, not unique enough\n",
      "-skipping training candidate: rna32765.p1, not unique enough\n",
      "-skipping training candidate: rna47762.p1, not unique enough\n",
      "-skipping training candidate: rna50237.p1, not unique enough\n",
      "-skipping training candidate: rna60061.p1, not unique enough\n",
      "-skipping training candidate: rna22674.p1, not unique enough\n",
      "-skipping training candidate: rna34755.p1, not unique enough\n",
      "-skipping training candidate: rna39234.p1, not unique enough\n",
      "-skipping training candidate: rna41778.p1, not unique enough\n",
      "-skipping training candidate: rna44387.p1, not unique enough\n",
      "-skipping training candidate: rna46875.p1, not unique enough\n",
      "-skipping training candidate: rna19322.p1, not unique enough\n",
      "-skipping training candidate: rna41790.p1, not unique enough\n",
      "-skipping training candidate: rna59269.p1, not unique enough\n",
      "-skipping training candidate: rna162.p1, not unique enough\n",
      "-skipping training candidate: rna6331.p1, not unique enough\n",
      "-skipping training candidate: rna18064.p1, not unique enough\n",
      "-skipping training candidate: rna20120.p1, not unique enough\n",
      "-skipping training candidate: rna29048.p1, not unique enough\n",
      "-skipping training candidate: rna32613.p1, not unique enough\n",
      "-skipping training candidate: rna41774.p1, not unique enough\n",
      "-skipping training candidate: rna44754.p1, not unique enough\n",
      "-skipping training candidate: rna60491.p1, not unique enough\n",
      "-skipping training candidate: rna2359.p1, not unique enough\n",
      "-skipping training candidate: rna8128.p1, not unique enough\n",
      "-skipping training candidate: rna23006.p1, not unique enough\n",
      "-skipping training candidate: rna22622.p1, not unique enough\n",
      "-skipping training candidate: rna33013.p1, not unique enough\n",
      "-skipping training candidate: rna39814.p1, not unique enough\n",
      "-skipping training candidate: rna40189.p1, not unique enough\n",
      "-skipping training candidate: rna58457.p1, not unique enough\n",
      "-skipping training candidate: rna9009.p1, not unique enough\n",
      "-skipping training candidate: rna47649.p1, not unique enough\n",
      "-skipping training candidate: rna61866.p1, not unique enough\n",
      "-skipping training candidate: rna61865.p1, not unique enough\n",
      "-skipping training candidate: rna8992.p1, not unique enough\n",
      "-skipping training candidate: rna23003.p1, not unique enough\n",
      "-skipping training candidate: rna46549.p1, not unique enough\n",
      "-skipping training candidate: rna48994.p1, not unique enough\n",
      "-skipping training candidate: rna50591.p1, not unique enough\n",
      "-skipping training candidate: rna61883.p1, not unique enough\n",
      "-skipping training candidate: rna61881.p1, not unique enough\n",
      "-skipping training candidate: rna19713.p1, not unique enough\n",
      "-skipping training candidate: rna19767.p1, not unique enough\n",
      "-skipping training candidate: rna47760.p1, not unique enough\n",
      "-skipping training candidate: rna19321.p1, not unique enough\n",
      "-skipping training candidate: rna25372.p1, not unique enough\n",
      "-skipping training candidate: rna25375.p1, not unique enough\n",
      "-skipping training candidate: rna33419.p1, not unique enough\n",
      "-skipping training candidate: rna38085.p1, not unique enough\n",
      "-skipping training candidate: rna42506.p1, not unique enough\n",
      "-skipping training candidate: rna49935.p1, not unique enough\n",
      "-skipping training candidate: rna59372.p1, not unique enough\n",
      "-skipping training candidate: rna62813.p1, not unique enough\n",
      "-skipping training candidate: rna18350.p1, not unique enough\n",
      "-skipping training candidate: rna32771.p1, not unique enough\n",
      "-skipping training candidate: rna41791.p1, not unique enough\n",
      "-skipping training candidate: rna55503.p1, not unique enough\n",
      "-skipping training candidate: rna61040.p1, not unique enough\n",
      "-skipping training candidate: rna61041.p1, not unique enough\n",
      "-skipping training candidate: rna64400.p1, not unique enough\n",
      "-skipping training candidate: rna64406.p1, not unique enough\n",
      "\n",
      "\t-redundancy-minimized set includes 2058 / 5000 = 41.16%\n",
      "\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr 500 > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest\n",
      "PCT_GC: 46.2\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/seq_n_baseprobs_to_loglikelihood_vals.pl ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//base_freqs.dat > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//hexamer.scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/score_CDS_likelihood_all_6_frames.pl ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//hexamer.scores > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/select_best_ORFs_per_transcript.pl --gff3_file ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.gff3 --cds_scores ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.scores  --min_length_auto_accept 749  > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 \n",
      "Selecting best orfs\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/train_start_PWM.pl --transcripts ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna --selected_orfs ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest --out_prefix ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement\n",
      "Training start codon pattern recognition* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/build_atgPWM_+-.pl  --transcripts ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna  --selected_orfs ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.top_500_longest  --out_prefix ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement --pwm_left 20 --pwm_right 10 \n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scoring.+-.pl  --features_plus ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.+.features  --features_minus ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.-.features  --atg_position 20  > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.feature.scores\n",
      "-round: 1\n",
      "-round: 2\n",
      "-round: 3\n",
      "-round: 4\n",
      "-round: 5\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scores_to_ROC.pl ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.feature.scores > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.feature.scores.roc\n",
      "-parsing scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/plot_ROC.Rscript ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.feature.scores.roc || :\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/compute_AUC.pl ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.feature.scores.roc\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.+.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.-.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/deplete_feature_noise.pl  --features_plus ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.+.features  --pwm_minus ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.-.pwm  --out_prefix ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.enhanced\n",
      "num features: 397\tnum_incorporate: 119\n",
      "-feature swap of score: -2.21692784225968 instead of -2.82480127800276\n",
      "-feature swap of score: -0.696828751663083 instead of -2.21669338229439\n",
      "-feature swap of score: 0.763599817269247 instead of -2.08002913214012\n",
      "-feature swap of score: -1.11519786089795 instead of -2.13198058788708\n",
      "-feature swap of score: 3.53408341556213 instead of -2.26417417643694\n",
      "-feature swap of score: 1.86101588908586 instead of -2.22369485081157\n",
      "-feature swap of score: -0.218665863983171 instead of -2.28581831991281\n",
      "-feature swap of score: 2.02886036026552 instead of -1.9859531223204\n",
      "-feature swap of score: 1.10702231475703 instead of -2.17554202724522\n",
      "-feature swap of score: 2.5383655822014 instead of -2.20399741332803\n",
      "-feature swap of score: 2.31679310064595 instead of -1.82450535806742\n",
      "-feature swap of score: -0.659719692413243 instead of -1.78167925724792\n",
      "-feature swap of score: 1.94987485463704 instead of -1.7706989701408\n",
      "-feature swap of score: 2.83377523996852 instead of -1.76507044820695\n",
      "-feature swap of score: 2.04112913248544 instead of -1.6645444711762\n",
      "-feature swap of score: 2.23140986944903 instead of -1.43593801295698\n",
      "-feature swap of score: -0.468853215613425 instead of -1.3662057797844\n",
      "-feature swap of score: 2.8501343048918 instead of -1.33918648637577\n",
      "-feature swap of score: -0.82297065064561 instead of -1.26654294850041\n",
      "-feature swap of score: 1.74847096833361 instead of -1.23636035760393\n",
      "-feature swap of score: 2.35820960262574 instead of -1.33451927243332\n",
      "-feature swap of score: 0.174525520245121 instead of -1.33831867169981\n",
      "-feature swap of score: -0.589301270618693 instead of -1.32496614856519\n",
      "-feature swap of score: 0.253617627828586 instead of -1.21867443551391\n",
      "-feature swap of score: 1.52673011593545 instead of -1.0299387224734\n",
      "-feature swap of score: -0.213045721889497 instead of -0.85861641458173\n",
      "-feature swap of score: 0.958395940861797 instead of -0.820298163482403\n",
      "-feature swap of score: -0.17652808834708 instead of -0.927096043151601\n",
      "-feature swap of score: -0.713001792305057 instead of -0.943313963296813\n",
      "-feature swap of score: 0.569552634718937 instead of -1.13043289339032\n",
      "-feature swap of score: 0.786660533945287 instead of -1.06490212259205\n",
      "-feature swap of score: -0.864984377267684 instead of -1.03253596898551\n",
      "-feature swap of score: 0.526490575345434 instead of -0.914478403205129\n",
      "-feature swap of score: -0.784147873225043 instead of -1.00113771518839\n",
      "-feature swap of score: 1.3138356220685 instead of -0.64912022078675\n",
      "-feature swap of score: 2.79690373635375 instead of -0.601493959624492\n",
      "-feature swap of score: 1.76065980912418 instead of -0.365952128306905\n",
      "-feature swap of score: 2.26891484656023 instead of -0.378676603838264\n",
      "-feature swap of score: -0.210905168398584 instead of -0.446565302579578\n",
      "-feature swap of score: 1.89889203183362 instead of -0.358936596302073\n",
      "-feature swap of score: 1.00688750298899 instead of -0.353916354013543\n",
      "-feature swap of score: 1.79484698552421 instead of -0.65989441137551\n",
      "-feature swap of score: 1.19827463110793 instead of -0.514872410851988\n",
      "-feature swap of score: 5.42964733568679 instead of -0.500999227608865\n",
      "-feature swap of score: 0.79055074424589 instead of -0.305912642936535\n",
      "-feature swap of score: 0.504275415107053 instead of -0.0829589300527132\n",
      "-feature swap of score: 1.6071457004952 instead of 0.0798266020192396\n",
      "-feature swap of score: 2.55300576811539 instead of 0.00962164936263132\n",
      "-feature swap of score: 1.6198141794378 instead of -0.275832437734224\n",
      "-feature swap of score: 0.923907367442004 instead of 0.0616025828884617\n",
      "-feature swap of score: 3.5634443444382 instead of 0.0244157032017905\n",
      "-feature swap of score: 4.46845866730229 instead of 0.0398402559058269\n",
      "-feature swap of score: 1.60001406794808 instead of -0.0230489817681658\n",
      "-feature swap of score: 1.49782608245693 instead of 0.0873432674626391\n",
      "-feature swap of score: 2.33907393406013 instead of -0.0916375721415191\n",
      "-feature swap of score: 1.86427405105594 instead of -0.0766140301656006\n",
      "-feature swap of score: 4.33559769833044 instead of -0.204162854577705\n",
      "-feature swap of score: 4.89324968832146 instead of -0.0290853636308332\n",
      "-feature swap of score: 0.630765880399525 instead of -0.0687196902425364\n",
      "-feature swap of score: 1.95903724520605 instead of 0.126086386012773\n",
      "-feature swap of score: 2.60481753120288 instead of 0.191334999026586\n",
      "-feature swap of score: 4.3608068902744 instead of -0.156283516370599\n",
      "-feature swap of score: 1.92944996439441 instead of 0.290193244748163\n",
      "-feature swap of score: 1.86578603701568 instead of 0.225229275180886\n",
      "-feature swap of score: 1.94018317547513 instead of 0.116538243846511\n",
      "-feature swap of score: 2.38816702354996 instead of 0.28047302730733\n",
      "-feature swap of score: 1.81725122339717 instead of 0.386253265777538\n",
      "-feature swap of score: 1.92520130938505 instead of 0.503716109145222\n",
      "-feature swap of score: 1.5452210972483 instead of 0.53356735390578\n",
      "-feature swap of score: 2.88801770441359 instead of 0.492389607438148\n",
      "-feature swap of score: 1.27405643003176 instead of 0.432658714686395\n",
      "-feature swap of score: 6.05437493391006 instead of 0.194328395891367\n",
      "-feature swap of score: 0.616082456763985 instead of 0.413620079700086\n",
      "-feature swap of score: 4.78033553337365 instead of 0.384021327172111\n",
      "-feature swap of score: 1.86738731571186 instead of 0.313622161722505\n",
      "-feature swap of score: 4.88405640273726 instead of 0.532053576165131\n",
      "-feature swap of score: 1.75822800393298 instead of 0.311359625224178\n",
      "-feature swap of score: 1.19262966750173 instead of 0.578854442017519\n",
      "-feature swap of score: 3.24795027257256 instead of 0.711014404123423\n",
      "-feature swap of score: 4.23546031605856 instead of 0.799980961414935\n",
      "-feature swap of score: 1.69783338063326 instead of 0.790876916862567\n",
      "-feature swap of score: 1.81965502104673 instead of 0.665324393576442\n",
      "-feature swap of score: 2.69788577770424 instead of 0.644831256446135\n",
      "-feature swap of score: 3.40645518604735 instead of 0.664482830522306\n",
      "-feature swap of score: 2.96028090277577 instead of 0.610638586788851\n",
      "-feature swap of score: 4.82604103608634 instead of 1.03340711309669\n",
      "-feature swap of score: 2.50275871642906 instead of 1.07193423555004\n",
      "-feature swap of score: 3.60957312686455 instead of 0.956754186985914\n",
      "-feature swap of score: 4.4433301270039 instead of 1.09402741546021\n",
      "-feature swap of score: 3.796273651055 instead of 1.02467993649294\n",
      "-feature swap of score: 3.89046488250719 instead of 1.05325815845505\n",
      "-feature swap of score: 2.0867183861946 instead of 1.08090722177749\n",
      "-feature swap of score: 1.57796425823252 instead of 0.922201284791306\n",
      "-feature swap of score: 5.33850009445087 instead of 0.928992085852854\n",
      "-feature swap of score: 2.32343864603735 instead of 0.898484973715832\n",
      "-feature swap of score: 5.91450664599095 instead of 1.22525033588609\n",
      "-feature swap of score: 2.11672909445256 instead of 1.11491412545647\n",
      "-feature swap of score: 1.9099999944661 instead of 1.33961086227587\n",
      "-feature swap of score: 1.32614707249276 instead of 1.10819707516468\n",
      "-feature swap of score: 2.18649035304408 instead of 1.08677612853226\n",
      "-feature swap of score: 3.19770920810262 instead of 1.18217498300053\n",
      "-feature swap of score: 3.87207599357921 instead of 1.38330442423035\n",
      "-feature swap of score: 3.34411985762808 instead of 1.35902255088362\n",
      "-feature swap of score: 1.64762094990613 instead of 1.41633056149429\n",
      "-feature swap of score: 6.85681092414695 instead of 1.49804798673973\n",
      "-feature swap of score: 5.33997426279726 instead of 1.48377281057682\n",
      "-feature swap of score: 5.42090313439803 instead of 1.53651511977846\n",
      "-feature swap of score: 2.93580997720439 instead of 1.4588028225823\n",
      "-feature swap of score: 2.28270524102015 instead of 1.44457420528128\n",
      "-feature swap of score: 3.81157422004727 instead of 1.62346847737815\n",
      "-feature swap of score: 5.54701787577771 instead of 1.4646282098063\n",
      "-feature swap of score: 4.10280722740544 instead of 1.70740657298957\n",
      "-feature swap of score: 2.52849153508133 instead of 1.62755282396467\n",
      "-feature swap of score: 4.93042683261043 instead of 1.57956790241985\n",
      "-feature swap of score: 4.4909098849646 instead of 1.54536669803315\n",
      "-feature swap of score: 7.14916040206657 instead of 1.57809830201934\n",
      "-feature swap of score: 1.60342145364952 instead of 1.45793981560608\n",
      "-num feature swaps: 117\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scoring.+-.pl  --features_plus ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.enhanced.+.features  --features_minus ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.-.features  --atg_position 20  > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores\n",
      "-round: 1\n",
      "-round: 2\n",
      "-round: 3\n",
      "-round: 4\n",
      "-round: 5\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scores_to_ROC.pl ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc\n",
      "-parsing scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/plot_ROC.Rscript ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc || :\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/compute_AUC.pl ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//start_refinement.enhanced.+.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/start_codon_refinement.pl --transcripts ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna --gff3_file ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 --workdir ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir/ > ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3\n",
      "Refining start codon selections.\n",
      "-number of revised start positions: 2462\n",
      "* Running CMD: cp ../../output/Drer_adultbrain/Danio_rerio_adultbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3 GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3\n",
      "copying output to final output file: GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_bed.pl GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 > GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed\n",
      "Making bed file: GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 --fasta ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna  --genetic_code Universal > GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep\n",
      "Making pep file: GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 --fasta ../../output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna --seqType CDS  --genetic_code Universal > GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds\n",
      "Making cds file: GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds\n"
     ]
    }
   ],
   "source": [
    "cdna = sample_Docket.genome_fasta.get_transdecoder_cdna_gtf(models_asgtf_updated, TRANSDECODER_LOC)\n",
    "sample_Docket.add_keyfile(cdna, 'cdna')\n",
    "\n",
    "transdecoder_files = sample_Docket.cdna.to_pep_files(TDLONGORF_LOC, TDPREDICT_LOC)\n",
    "sample_Docket.add_keyfiles(transdecoder_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "iEug6TQ6Cc9D"
   },
   "source": [
    "# 12. Push files to AWS S3\n",
    "\n",
    "Iteratively moves through the file_set and file_dict variables and populates files into the right place in AWS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 13. Pickle the `sample_Docket` variable for use by the next script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_Docket.pickle()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "Xtro_bro.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
