{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "L56fjA27DmhA"
   },
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "t0BTCsz7Y48y",
    "outputId": "f6510678-f977-4ec8-ce30-1f2acd4fe666"
   },
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess, os, dill, sys\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/glial-origins/output/Mmus_adolescentbrain/ already exists\n",
      "Files will be saved into /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/\n"
     ]
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify the name of the species folder in Amazon S3\n",
    "species = 'Mus_musculus'\n",
    "\n",
    "# Specify any particular identifying conditions, eg tissue type:\n",
    "conditions = 'adolescentbrain'\n",
    "\n",
    "################\n",
    "################\n",
    "\n",
    "sample_BFD = BioFileDocket(species, conditions)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Download and describe data\n",
    "\n",
    "This notebook collects data from the [Han et al. 2018](https://www.sciencedirect.com/science/article/pii/S0092867418301168#sec4) mouse cell atlas.\n",
    "\n",
    "## Dataset description\n",
    "\n",
    "- Data can be found at the [Linnarsson Lab Website](http://mousebrain.org/adolescent/).\n",
    "- This notebook collects data from all samples in LOOM format for the purposes of data analysis and exploration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "inferring file name as GCF_000001635.23_GRCm38.p3_genomic.fna.gz\n",
      "file GCF_000001635.23_GRCm38.p3_genomic.fna.gz already exists at /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/GCF_000001635.23_GRCm38.p3_genomic.fna.gz\n",
      "file GCF_000001635.23_GRCm38.p3_genomic.fna.gz unzipped and object renamed to GCF_000001635.23_GRCm38.p3_genomic.fna\n",
      "inferring file name as GCF_000001635.23_GRCm38.p3_genomic.gff.gz\n",
      "file GCF_000001635.23_GRCm38.p3_genomic.gff.gz already exists at /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/GCF_000001635.23_GRCm38.p3_genomic.gff.gz\n",
      "file GCF_000001635.23_GRCm38.p3_genomic.gff.gz unzipped and object renamed to GCF_000001635.23_GRCm38.p3_genomic.gff\n",
      "inferring file name as 10X05_1.loom\n",
      "file 10X05_1.loom already exists at /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/10X05_1.loom\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "gzip: /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/GCF_000001635.23_GRCm38.p3_genomic.fna: unknown suffix -- ignored\n",
      "gzip: /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/GCF_000001635.23_GRCm38.p3_genomic.gff: unknown suffix -- ignored\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GxcFile already exists at /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/10X05_1.loom.asgxc.tsv\n",
      "IdmmFile already exists at /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/10X05_1.loom.idmm.tsv\n",
      "CellAnnotFile already exists at /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/10X05_1.loom.cellannot.tsv\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'species': 'Mus_musculus',\n",
       " 'conditions': 'adolescentbrain',\n",
       " 'directory': '/home/ec2-user/glial-origins/output/Mmus_adolescentbrain/',\n",
       " 'files': {},\n",
       " 'metadata': <biofile_handling.metadata_object at 0x7fa67bdafee0>,\n",
       " 'annot': <biofile_handling.GenomeGffFile at 0x7fa5ca9378b0>,\n",
       " 'genome_fasta': <biofile_handling.GenomeFastaFile at 0x7fa5ca936440>,\n",
       " 'gxc': <biofile_handling.GxcFile at 0x7fa5ca9348e0>,\n",
       " 'loom': <biofile_handling.LoomFile at 0x7fa5ca936470>,\n",
       " 'loom_idmm': <biofile_handling.IdmmFile at 0x7fa67bdaf970>,\n",
       " 'loom_cellannot': <biofile_handling.CellAnnotFile at 0x7fa5c26633d0>}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Specify url and other variables\n",
    "genome_fasta_url = 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/635/GCF_000001635.23_GRCm38.p3/GCF_000001635.23_GRCm38.p3_genomic.fna.gz'\n",
    "genome_version = 'GRCm38.p3'\n",
    "\n",
    "annot_url = 'https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/635/GCF_000001635.23_GRCm38.p3/GCF_000001635.23_GRCm38.p3_genomic.gff.gz'\n",
    "loom_url = 'https://storage.googleapis.com/linnarsson-lab-chromium/10X05_1/10X05_1.loom'\n",
    "\n",
    "###########\n",
    "# runtime #\n",
    "###########\n",
    "\n",
    "protocol = 'curl'\n",
    "\n",
    "genome_fasta = GenomeFastaFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    version = genome_version,\n",
    "    url = genome_fasta_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "annot = GenomeGffFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    reference_genome = genome_fasta,\n",
    "    url = annot_url,\n",
    "    protocol = protocol\n",
    ")\n",
    "\n",
    "loom = LoomFile(\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    url = loom_url,\n",
    "    protocol = protocol,\n",
    "    reference_genome = genome_fasta,\n",
    "    reference_annot = annot\n",
    ")\n",
    "\n",
    "gxc = loom.to_gxc()\n",
    "loom_idmm = loom.get_idmm(id_type = 'ensembl_id')\n",
    "loom_cellannot = loom.get_cellannot()\n",
    "\n",
    "keyfiles = {\n",
    "    'annot': annot,\n",
    "    'genome_fasta': genome_fasta,\n",
    "    'gxc': gxc,\n",
    "    'loom': loom,\n",
    "    'loom_idmm': loom_idmm,\n",
    "    'loom_cellannot': loom_cellannot\n",
    "}\n",
    "\n",
    "sample_BFD.add_keyfiles(keyfiles)\n",
    "display(vars(sample_BFD))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Load in the gxc matrix and get gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>10X05_1:AAACATACTTTACC-1</th>\n",
       "      <th>10X05_1:AAACATTGCCTCCA-1</th>\n",
       "      <th>10X05_1:AAACATTGCTTACT-1</th>\n",
       "      <th>10X05_1:AAACATTGGTACGT-1</th>\n",
       "      <th>10X05_1:AAACCGTGACTCAG-1</th>\n",
       "      <th>10X05_1:AAACCGTGCGTAGT-1</th>\n",
       "      <th>10X05_1:AAACGCACACACGT-1</th>\n",
       "      <th>10X05_1:AAACGCACGTCGAT-1</th>\n",
       "      <th>10X05_1:AAACGGCTGTGCTA-1</th>\n",
       "      <th>...</th>\n",
       "      <th>10X05_1:TTTCTACTCGCTAA-1</th>\n",
       "      <th>10X05_1:TTTCTACTCTGACA-1</th>\n",
       "      <th>10X05_1:TTTCTACTTGTGCA-1</th>\n",
       "      <th>10X05_1:TTTCTACTTTCTCA-1</th>\n",
       "      <th>10X05_1:TTTCTACTTTGGCA-1</th>\n",
       "      <th>10X05_1:TTTGACTGTTCAGG-1</th>\n",
       "      <th>10X05_1:TTTGCATGCAGAGG-1</th>\n",
       "      <th>10X05_1:TTTGCATGGCGATT-1</th>\n",
       "      <th>10X05_1:TTTGCATGTACTGG-1</th>\n",
       "      <th>10X05_1:TTTGCATGTTCCGC-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gm1992</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gm37381</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rp1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rp1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Sox17</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Gm37323</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Lypla1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Gm37988</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 4291 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name  10X05_1:AAACATACTTTACC-1  10X05_1:AAACATTGCCTCCA-1  \\\n",
       "0      Xkr4                       1.0                       0.0   \n",
       "1    Gm1992                       0.0                       0.0   \n",
       "2   Gm37381                       0.0                       0.0   \n",
       "3       Rp1                       0.0                       0.0   \n",
       "4       Rp1                       0.0                       0.0   \n",
       "5     Sox17                       0.0                       0.0   \n",
       "6   Gm37323                       0.0                       0.0   \n",
       "7    Mrpl15                       0.0                       0.0   \n",
       "8    Lypla1                       0.0                       0.0   \n",
       "9   Gm37988                       0.0                       0.0   \n",
       "\n",
       "   10X05_1:AAACATTGCTTACT-1  10X05_1:AAACATTGGTACGT-1  \\\n",
       "0                       0.0                       0.0   \n",
       "1                       0.0                       0.0   \n",
       "2                       0.0                       0.0   \n",
       "3                       0.0                       0.0   \n",
       "4                       0.0                       0.0   \n",
       "5                       0.0                       0.0   \n",
       "6                       0.0                       0.0   \n",
       "7                       0.0                       0.0   \n",
       "8                       0.0                       0.0   \n",
       "9                       0.0                       0.0   \n",
       "\n",
       "   10X05_1:AAACCGTGACTCAG-1  10X05_1:AAACCGTGCGTAGT-1  \\\n",
       "0                       0.0                       0.0   \n",
       "1                       0.0                       0.0   \n",
       "2                       0.0                       0.0   \n",
       "3                       0.0                       0.0   \n",
       "4                       0.0                       0.0   \n",
       "5                       0.0                       0.0   \n",
       "6                       0.0                       0.0   \n",
       "7                       0.0                       0.0   \n",
       "8                       0.0                       0.0   \n",
       "9                       0.0                       0.0   \n",
       "\n",
       "   10X05_1:AAACGCACACACGT-1  10X05_1:AAACGCACGTCGAT-1  \\\n",
       "0                       0.0                       0.0   \n",
       "1                       0.0                       0.0   \n",
       "2                       0.0                       0.0   \n",
       "3                       0.0                       0.0   \n",
       "4                       0.0                       0.0   \n",
       "5                       5.0                       0.0   \n",
       "6                       0.0                       0.0   \n",
       "7                       0.0                       0.0   \n",
       "8                       0.0                       0.0   \n",
       "9                       0.0                       0.0   \n",
       "\n",
       "   10X05_1:AAACGGCTGTGCTA-1  ...  10X05_1:TTTCTACTCGCTAA-1  \\\n",
       "0                       0.0  ...                       0.0   \n",
       "1                       0.0  ...                       0.0   \n",
       "2                       0.0  ...                       0.0   \n",
       "3                       0.0  ...                       0.0   \n",
       "4                       0.0  ...                       0.0   \n",
       "5                       0.0  ...                       0.0   \n",
       "6                       0.0  ...                       0.0   \n",
       "7                       0.0  ...                       1.0   \n",
       "8                       0.0  ...                       0.0   \n",
       "9                       0.0  ...                       0.0   \n",
       "\n",
       "   10X05_1:TTTCTACTCTGACA-1  10X05_1:TTTCTACTTGTGCA-1  \\\n",
       "0                       0.0                       0.0   \n",
       "1                       0.0                       0.0   \n",
       "2                       0.0                       0.0   \n",
       "3                       0.0                       0.0   \n",
       "4                       0.0                       0.0   \n",
       "5                       0.0                       0.0   \n",
       "6                       0.0                       0.0   \n",
       "7                       0.0                       0.0   \n",
       "8                       0.0                       0.0   \n",
       "9                       0.0                       0.0   \n",
       "\n",
       "   10X05_1:TTTCTACTTTCTCA-1  10X05_1:TTTCTACTTTGGCA-1  \\\n",
       "0                       0.0                       0.0   \n",
       "1                       0.0                       0.0   \n",
       "2                       0.0                       0.0   \n",
       "3                       0.0                       0.0   \n",
       "4                       0.0                       0.0   \n",
       "5                       0.0                       0.0   \n",
       "6                       0.0                       0.0   \n",
       "7                       0.0                       1.0   \n",
       "8                       0.0                       0.0   \n",
       "9                       0.0                       0.0   \n",
       "\n",
       "   10X05_1:TTTGACTGTTCAGG-1  10X05_1:TTTGCATGCAGAGG-1  \\\n",
       "0                       0.0                       0.0   \n",
       "1                       0.0                       0.0   \n",
       "2                       0.0                       0.0   \n",
       "3                       0.0                       0.0   \n",
       "4                       0.0                       0.0   \n",
       "5                       0.0                       0.0   \n",
       "6                       0.0                       0.0   \n",
       "7                       0.0                       0.0   \n",
       "8                       0.0                       0.0   \n",
       "9                       0.0                       0.0   \n",
       "\n",
       "   10X05_1:TTTGCATGGCGATT-1  10X05_1:TTTGCATGTACTGG-1  \\\n",
       "0                       0.0                       0.0   \n",
       "1                       0.0                       0.0   \n",
       "2                       0.0                       0.0   \n",
       "3                       0.0                       0.0   \n",
       "4                       0.0                       0.0   \n",
       "5                       0.0                       0.0   \n",
       "6                       0.0                       0.0   \n",
       "7                       0.0                       0.0   \n",
       "8                       0.0                       0.0   \n",
       "9                       0.0                       0.0   \n",
       "\n",
       "   10X05_1:TTTGCATGTTCCGC-1  \n",
       "0                       0.0  \n",
       "1                       0.0  \n",
       "2                       0.0  \n",
       "3                       0.0  \n",
       "4                       0.0  \n",
       "5                       0.0  \n",
       "6                       0.0  \n",
       "7                       1.0  \n",
       "8                       2.0  \n",
       "9                       0.0  \n",
       "\n",
       "[10 rows x 4291 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Xkr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gm1992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gm37381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27993</th>\n",
       "      <td>AC168977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27994</th>\n",
       "      <td>PISD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27995</th>\n",
       "      <td>DHRSX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27996</th>\n",
       "      <td>Vmn2r122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27997</th>\n",
       "      <td>CAAA01147332.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27998 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_name\n",
       "0                Xkr4\n",
       "1              Gm1992\n",
       "2             Gm37381\n",
       "3                 Rp1\n",
       "4                 Rp1\n",
       "...               ...\n",
       "27993      AC168977.1\n",
       "27994            PISD\n",
       "27995           DHRSX\n",
       "27996        Vmn2r122\n",
       "27997  CAAA01147332.1\n",
       "\n",
       "[27998 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes_matrix = pd.read_csv(sample_BFD.gxc.path, sep = '\\t', nrows = 10)\n",
    "display(genes_matrix)\n",
    "\n",
    "gxc_genes_list = pd.read_csv(sample_BFD.gxc.path, sep = '\\t', usecols=[0])\n",
    "display(gxc_genes_list)\n",
    "\n",
    "sample_BFD.metadata.add('num_cells', len(genes_matrix.columns) - 1)\n",
    "sample_BFD.metadata.add('num_genes', len(genes_matrix))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "jvfcFNT2FuaZ"
   },
   "source": [
    "# 3. Get mapping identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>195471971</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id0;Dbxref=taxon:10090;Name=1;chromosome=1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>BestRefSeq%2CGnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>3199731</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Dbxref=GeneID:497097,MGI:MGI:3528744;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3199731</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Dbxref=GeneID:497097,Genb...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Dbxref=GeneID:497097,Genban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Dbxref=GeneID:497097,Genban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284235</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>15451</td>\n",
       "      <td>15509</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169986;Note=ETAS1%3B extended terminatio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284236</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>15515</td>\n",
       "      <td>15558</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169987;Note=ETAS2%3B extended terminatio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284237</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>16035</td>\n",
       "      <td>16058</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169988;Note=CSB1%3B conserved sequencing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284238</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>16089</td>\n",
       "      <td>16104</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169989;Note=CSB2%3B conserved sequencing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284239</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>16114</td>\n",
       "      <td>16131</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169990;Note=CSB3%3B conserved sequencing...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2284240 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0                    1       2        3          4  5  6  \\\n",
       "0        NC_000067.6               RefSeq  region        1  195471971  .  +   \n",
       "1        NC_000067.6  BestRefSeq%2CGnomon    gene  3199731    3671742  .  -   \n",
       "2        NC_000067.6               Gnomon    mRNA  3199731    3671742  .  -   \n",
       "3        NC_000067.6               Gnomon    exon  3670552    3671742  .  -   \n",
       "4        NC_000067.6               Gnomon    exon  3421702    3421901  .  -   \n",
       "...              ...                  ...     ...      ...        ... .. ..   \n",
       "2284235  NC_005089.1               RefSeq  region    15451      15509  .  +   \n",
       "2284236  NC_005089.1               RefSeq  region    15515      15558  .  +   \n",
       "2284237  NC_005089.1               RefSeq  region    16035      16058  .  +   \n",
       "2284238  NC_005089.1               RefSeq  region    16089      16104  .  +   \n",
       "2284239  NC_005089.1               RefSeq  region    16114      16131  .  +   \n",
       "\n",
       "         7                                                  8  \n",
       "0        .  ID=id0;Dbxref=taxon:10090;Name=1;chromosome=1;...  \n",
       "1        .  ID=gene0;Dbxref=GeneID:497097,MGI:MGI:3528744;...  \n",
       "2        .  ID=rna0;Parent=gene0;Dbxref=GeneID:497097,Genb...  \n",
       "3        .  ID=id1;Parent=rna0;Dbxref=GeneID:497097,Genban...  \n",
       "4        .  ID=id2;Parent=rna0;Dbxref=GeneID:497097,Genban...  \n",
       "...     ..                                                ...  \n",
       "2284235  .  ID=id1169986;Note=ETAS1%3B extended terminatio...  \n",
       "2284236  .  ID=id1169987;Note=ETAS2%3B extended terminatio...  \n",
       "2284237  .  ID=id1169988;Note=CSB1%3B conserved sequencing...  \n",
       "2284238  .  ID=id1169989;Note=CSB2%3B conserved sequencing...  \n",
       "2284239  .  ID=id1169990;Note=CSB3%3B conserved sequencing...  \n",
       "\n",
       "[2284240 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'ID=id1;Parent=rna0;Dbxref=GeneID:497097,Genbank:XM_006495550.2,MGI:MGI:3528744;gbkey=mRNA;gene=Xkr4;product=X Kell blood group precursor related family member 4%2C transcript variant X1;transcript_id=XM_006495550.2'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load in the original GFF-based annotation\n",
    "models = pd.read_csv(sample_BFD.annot.path, comment = '#', header = None, sep = '\\t')\n",
    "display(models)\n",
    "\n",
    "attributes_column = 8\n",
    "\n",
    "# Check the structure of fields in the GFF additional fields section\n",
    "display(models[attributes_column][3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>Dbxref</th>\n",
       "      <th>dbxref_dict</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>195471971</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id0;Dbxref=taxon:10090;Name=1;chromosome=1;...</td>\n",
       "      <td>{'ID': 'id0', 'Dbxref': 'taxon:10090', 'Name':...</td>\n",
       "      <td>None</td>\n",
       "      <td>taxon:10090</td>\n",
       "      <td>{'taxon': '10090'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>BestRefSeq%2CGnomon</td>\n",
       "      <td>gene</td>\n",
       "      <td>3199731</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene0;Dbxref=GeneID:497097,MGI:MGI:3528744;...</td>\n",
       "      <td>{'ID': 'gene0', 'Dbxref': 'GeneID:497097,MGI:M...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>GeneID:497097,MGI:MGI:3528744</td>\n",
       "      <td>{'GeneID': '497097', 'MGI': '3528744'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3199731</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=rna0;Parent=gene0;Dbxref=GeneID:497097,Genb...</td>\n",
       "      <td>{'ID': 'rna0', 'Parent': 'gene0', 'Dbxref': 'G...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>GeneID:497097,Genbank:XM_006495550.2,MGI:MGI:3...</td>\n",
       "      <td>{'GeneID': '497097', 'Genbank': 'XM_006495550....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1;Parent=rna0;Dbxref=GeneID:497097,Genban...</td>\n",
       "      <td>{'ID': 'id1', 'Parent': 'rna0', 'Dbxref': 'Gen...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>GeneID:497097,Genbank:XM_006495550.2,MGI:MGI:3...</td>\n",
       "      <td>{'GeneID': '497097', 'Genbank': 'XM_006495550....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id2;Parent=rna0;Dbxref=GeneID:497097,Genban...</td>\n",
       "      <td>{'ID': 'id2', 'Parent': 'rna0', 'Dbxref': 'Gen...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>GeneID:497097,Genbank:XM_006495550.2,MGI:MGI:3...</td>\n",
       "      <td>{'GeneID': '497097', 'Genbank': 'XM_006495550....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284235</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>15451</td>\n",
       "      <td>15509</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169986;Note=ETAS1%3B extended terminatio...</td>\n",
       "      <td>{'ID': 'id1169986', 'Note': 'ETAS1%3B extended...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284236</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>15515</td>\n",
       "      <td>15558</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169987;Note=ETAS2%3B extended terminatio...</td>\n",
       "      <td>{'ID': 'id1169987', 'Note': 'ETAS2%3B extended...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284237</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>16035</td>\n",
       "      <td>16058</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169988;Note=CSB1%3B conserved sequencing...</td>\n",
       "      <td>{'ID': 'id1169988', 'Note': 'CSB1%3B conserved...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284238</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>16089</td>\n",
       "      <td>16104</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169989;Note=CSB2%3B conserved sequencing...</td>\n",
       "      <td>{'ID': 'id1169989', 'Note': 'CSB2%3B conserved...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2284239</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>region</td>\n",
       "      <td>16114</td>\n",
       "      <td>16131</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=id1169990;Note=CSB3%3B conserved sequencing...</td>\n",
       "      <td>{'ID': 'id1169990', 'Note': 'CSB3%3B conserved...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2284240 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0                    1       2        3          4  5  6  \\\n",
       "0        NC_000067.6               RefSeq  region        1  195471971  .  +   \n",
       "1        NC_000067.6  BestRefSeq%2CGnomon    gene  3199731    3671742  .  -   \n",
       "2        NC_000067.6               Gnomon    mRNA  3199731    3671742  .  -   \n",
       "3        NC_000067.6               Gnomon    exon  3670552    3671742  .  -   \n",
       "4        NC_000067.6               Gnomon    exon  3421702    3421901  .  -   \n",
       "...              ...                  ...     ...      ...        ... .. ..   \n",
       "2284235  NC_005089.1               RefSeq  region    15451      15509  .  +   \n",
       "2284236  NC_005089.1               RefSeq  region    15515      15558  .  +   \n",
       "2284237  NC_005089.1               RefSeq  region    16035      16058  .  +   \n",
       "2284238  NC_005089.1               RefSeq  region    16089      16104  .  +   \n",
       "2284239  NC_005089.1               RefSeq  region    16114      16131  .  +   \n",
       "\n",
       "         7                                                  8  \\\n",
       "0        .  ID=id0;Dbxref=taxon:10090;Name=1;chromosome=1;...   \n",
       "1        .  ID=gene0;Dbxref=GeneID:497097,MGI:MGI:3528744;...   \n",
       "2        .  ID=rna0;Parent=gene0;Dbxref=GeneID:497097,Genb...   \n",
       "3        .  ID=id1;Parent=rna0;Dbxref=GeneID:497097,Genban...   \n",
       "4        .  ID=id2;Parent=rna0;Dbxref=GeneID:497097,Genban...   \n",
       "...     ..                                                ...   \n",
       "2284235  .  ID=id1169986;Note=ETAS1%3B extended terminatio...   \n",
       "2284236  .  ID=id1169987;Note=ETAS2%3B extended terminatio...   \n",
       "2284237  .  ID=id1169988;Note=CSB1%3B conserved sequencing...   \n",
       "2284238  .  ID=id1169989;Note=CSB2%3B conserved sequencing...   \n",
       "2284239  .  ID=id1169990;Note=CSB3%3B conserved sequencing...   \n",
       "\n",
       "                                          field_dictionary gene_name  \\\n",
       "0        {'ID': 'id0', 'Dbxref': 'taxon:10090', 'Name':...      None   \n",
       "1        {'ID': 'gene0', 'Dbxref': 'GeneID:497097,MGI:M...      Xkr4   \n",
       "2        {'ID': 'rna0', 'Parent': 'gene0', 'Dbxref': 'G...      Xkr4   \n",
       "3        {'ID': 'id1', 'Parent': 'rna0', 'Dbxref': 'Gen...      Xkr4   \n",
       "4        {'ID': 'id2', 'Parent': 'rna0', 'Dbxref': 'Gen...      Xkr4   \n",
       "...                                                    ...       ...   \n",
       "2284235  {'ID': 'id1169986', 'Note': 'ETAS1%3B extended...      None   \n",
       "2284236  {'ID': 'id1169987', 'Note': 'ETAS2%3B extended...      None   \n",
       "2284237  {'ID': 'id1169988', 'Note': 'CSB1%3B conserved...      None   \n",
       "2284238  {'ID': 'id1169989', 'Note': 'CSB2%3B conserved...      None   \n",
       "2284239  {'ID': 'id1169990', 'Note': 'CSB3%3B conserved...      None   \n",
       "\n",
       "                                                    Dbxref  \\\n",
       "0                                              taxon:10090   \n",
       "1                            GeneID:497097,MGI:MGI:3528744   \n",
       "2        GeneID:497097,Genbank:XM_006495550.2,MGI:MGI:3...   \n",
       "3        GeneID:497097,Genbank:XM_006495550.2,MGI:MGI:3...   \n",
       "4        GeneID:497097,Genbank:XM_006495550.2,MGI:MGI:3...   \n",
       "...                                                    ...   \n",
       "2284235                                               None   \n",
       "2284236                                               None   \n",
       "2284237                                               None   \n",
       "2284238                                               None   \n",
       "2284239                                               None   \n",
       "\n",
       "                                               dbxref_dict  \n",
       "0                                       {'taxon': '10090'}  \n",
       "1                   {'GeneID': '497097', 'MGI': '3528744'}  \n",
       "2        {'GeneID': '497097', 'Genbank': 'XM_006495550....  \n",
       "3        {'GeneID': '497097', 'Genbank': 'XM_006495550....  \n",
       "4        {'GeneID': '497097', 'Genbank': 'XM_006495550....  \n",
       "...                                                    ...  \n",
       "2284235                                               None  \n",
       "2284236                                               None  \n",
       "2284237                                               None  \n",
       "2284238                                               None  \n",
       "2284239                                               None  \n",
       "\n",
       "[2284240 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Remove any rows with NaNs\n",
    "models.dropna(inplace = True)\n",
    "\n",
    "# Extract field and database cross-ref (dbxref) information into columns\n",
    "models['field_dictionary'] = models[attributes_column].apply(convert_fields_to_dict_gff)\n",
    "models['gene_name'] = [d.get('gene') for d in models['field_dictionary']]\n",
    "models['Dbxref'] = [d.get('Dbxref') for d in models['field_dictionary']]\n",
    "models['dbxref_dict'] = models['Dbxref'].apply(convert_dbxref_to_dict)\n",
    "\n",
    "display(models)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Extract gene IDs for mapping to UniprotKB\n",
    "Specify which set of identifiers will be use to query the [Uniprot ID Mapping Tool](https://www.uniprot.org/id-mapping) via API.\n",
    "\n",
    "If using an identifier from the `dbxref_dict`, specify the name via string in the `dbxref_datafield` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ensembl_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Rp1</td>\n",
       "      <td>ENSMUSG00000025900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Rp1</td>\n",
       "      <td>ENSMUSG00000109048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Sox17</td>\n",
       "      <td>ENSMUSG00000025902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Mrpl15</td>\n",
       "      <td>ENSMUSG00000033845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46282</th>\n",
       "      <td>Gm21294</td>\n",
       "      <td>ENSMUSG00000102045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46283</th>\n",
       "      <td>Gm21996</td>\n",
       "      <td>ENSMUSG00000100608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46286</th>\n",
       "      <td>Gm29504</td>\n",
       "      <td>ENSMUSG00000100533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46288</th>\n",
       "      <td>Gm20837</td>\n",
       "      <td>ENSMUSG00000096178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46291</th>\n",
       "      <td>Gm28301</td>\n",
       "      <td>ENSMUSG00000099399</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23090 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name          ensembl_id\n",
       "0          Xkr4  ENSMUSG00000051951\n",
       "7           Rp1  ENSMUSG00000025900\n",
       "8           Rp1  ENSMUSG00000109048\n",
       "9         Sox17  ENSMUSG00000025902\n",
       "19       Mrpl15  ENSMUSG00000033845\n",
       "...         ...                 ...\n",
       "46282   Gm21294  ENSMUSG00000102045\n",
       "46283   Gm21996  ENSMUSG00000100608\n",
       "46286   Gm29504  ENSMUSG00000100533\n",
       "46288   Gm20837  ENSMUSG00000096178\n",
       "46291   Gm28301  ENSMUSG00000099399\n",
       "\n",
       "[23090 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "loom_idmm_df = pd.read_csv(loom_idmm.path, sep = '\\t')\n",
    "annot_gene_names = models.dropna(axis = 0, subset = ['gene_name'])[['gene_name']].drop_duplicates()\n",
    "\n",
    "models_subset = annot_gene_names.merge(loom_idmm_df, on = 'gene_name', how = 'left').dropna()\n",
    "\n",
    "dbxref_datafield = 'ensembl_id'\n",
    "display(models_subset)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Generate gene list file to query Uniprot ID Mapping API\n",
    "Generate a text file ending in `_ids.txt` for sending to the ID mapping API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 23090 gene ids to /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mmus_adolescentbrain_ensembl_id_ids.txt\n"
     ]
    }
   ],
   "source": [
    "gene_list = models_subset[dbxref_datafield].unique()\n",
    "\n",
    "genelist_object = GeneListFile(\n",
    "    filename = '', # filename is automatically generated from sampledict\n",
    "    sampledict = sample_BFD.sampledict,\n",
    "    sources = [sample_BFD.annot],\n",
    "    genes = gene_list,\n",
    "    identifier = dbxref_datafield\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Query Uniprot ID Mapping API\n",
    "Specify the `from_type` variable based on the Uniprot name of the identifier.  \n",
    "The table below lists some databases and the `from_type` string that the API accepts for that datatype.  \n",
    "\n",
    "| datatype | `from_type` string | description |\n",
    "| ---: | :--- | :--- |\n",
    "| Mouse Genome Informatics | `MGI` | ID starts with `MGI:` |\n",
    "| Zebrafish Information Network | `ZFIN` | ID starts with `ZDB-GENE-` |\n",
    "| Xenbase | `Xenbase` | ID starts with `XB-GENE-` |\n",
    "\n",
    "__NOTE:__ You may have to run the cell below twice - UniProt sometimes throws an \"Resource not found\" message on the first query to the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100  428k    0    52  100  428k     23   191k  0:00:02  0:00:02 --:--:--  191k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100    24    0    24    0     0     31      0 --:--:-- --:--:-- --:--:--    31\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>From</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "      <td>Q5GH67</td>\n",
       "      <td>XKR4_MOUSE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>XK-related protein 4 [Cleaved into: XK-related...</td>\n",
       "      <td>Xkr4 Xrg4</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000025900</td>\n",
       "      <td>P56716</td>\n",
       "      <td>RP1_MOUSE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Oxygen-regulated protein 1 (Retinitis pigmento...</td>\n",
       "      <td>Rp1 Orp1 Rp1h</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>2095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000025900</td>\n",
       "      <td>A0A140LHJ6</td>\n",
       "      <td>A0A140LHJ6_MOUSE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Oxygen-regulated protein 1</td>\n",
       "      <td>Rp1</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>1371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000025902</td>\n",
       "      <td>Q61473</td>\n",
       "      <td>SOX17_MOUSE</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Transcription factor SOX-17</td>\n",
       "      <td>Sox17 Sox-17</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000025902</td>\n",
       "      <td>A0A0A6YWS4</td>\n",
       "      <td>A0A0A6YWS4_MOUSE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Transcription factor SOX-17</td>\n",
       "      <td>Sox17</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51510</th>\n",
       "      <td>ENSMUSG00000099856</td>\n",
       "      <td>A0A087WRK1</td>\n",
       "      <td>A0A087WRK1_MOUSE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Predicted gene, 20814 (Predicted gene, 20855) ...</td>\n",
       "      <td>Gm20905 Gm20814 Gm20835 Gm20850 Gm20855 Gm2086...</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51511</th>\n",
       "      <td>ENSMUSG00000102045</td>\n",
       "      <td>A0A087WS79</td>\n",
       "      <td>A0A087WS79_MOUSE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Predicted gene, 21294</td>\n",
       "      <td>Gm21294</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51512</th>\n",
       "      <td>ENSMUSG00000100608</td>\n",
       "      <td>A0A087WSS9</td>\n",
       "      <td>A0A087WSS9_MOUSE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Predicted gene 21996</td>\n",
       "      <td>Gm21996</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51513</th>\n",
       "      <td>ENSMUSG00000096178</td>\n",
       "      <td>A0A087WP97</td>\n",
       "      <td>A0A087WP97_MOUSE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Predicted gene, 20837</td>\n",
       "      <td>Gm20837</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51514</th>\n",
       "      <td>ENSMUSG00000096178</td>\n",
       "      <td>J3QP42</td>\n",
       "      <td>J3QP42_MOUSE</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Predicted gene 28891 (Predicted gene, 20837) (...</td>\n",
       "      <td>Gm20838 Gm20837 Gm21094 Gm21256 Gm21739 Gm2186...</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "      <td>166</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>51515 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     From       Entry        Entry Name    Reviewed  \\\n",
       "0      ENSMUSG00000051951      Q5GH67        XKR4_MOUSE    reviewed   \n",
       "1      ENSMUSG00000025900      P56716         RP1_MOUSE    reviewed   \n",
       "2      ENSMUSG00000025900  A0A140LHJ6  A0A140LHJ6_MOUSE  unreviewed   \n",
       "3      ENSMUSG00000025902      Q61473       SOX17_MOUSE    reviewed   \n",
       "4      ENSMUSG00000025902  A0A0A6YWS4  A0A0A6YWS4_MOUSE  unreviewed   \n",
       "...                   ...         ...               ...         ...   \n",
       "51510  ENSMUSG00000099856  A0A087WRK1  A0A087WRK1_MOUSE  unreviewed   \n",
       "51511  ENSMUSG00000102045  A0A087WS79  A0A087WS79_MOUSE  unreviewed   \n",
       "51512  ENSMUSG00000100608  A0A087WSS9  A0A087WSS9_MOUSE  unreviewed   \n",
       "51513  ENSMUSG00000096178  A0A087WP97  A0A087WP97_MOUSE  unreviewed   \n",
       "51514  ENSMUSG00000096178      J3QP42      J3QP42_MOUSE  unreviewed   \n",
       "\n",
       "                                           Protein names  \\\n",
       "0      XK-related protein 4 [Cleaved into: XK-related...   \n",
       "1      Oxygen-regulated protein 1 (Retinitis pigmento...   \n",
       "2                             Oxygen-regulated protein 1   \n",
       "3                            Transcription factor SOX-17   \n",
       "4                            Transcription factor SOX-17   \n",
       "...                                                  ...   \n",
       "51510  Predicted gene, 20814 (Predicted gene, 20855) ...   \n",
       "51511                              Predicted gene, 21294   \n",
       "51512                               Predicted gene 21996   \n",
       "51513                              Predicted gene, 20837   \n",
       "51514  Predicted gene 28891 (Predicted gene, 20837) (...   \n",
       "\n",
       "                                              Gene Names  \\\n",
       "0                                              Xkr4 Xrg4   \n",
       "1                                          Rp1 Orp1 Rp1h   \n",
       "2                                                    Rp1   \n",
       "3                                           Sox17 Sox-17   \n",
       "4                                                  Sox17   \n",
       "...                                                  ...   \n",
       "51510  Gm20905 Gm20814 Gm20835 Gm20850 Gm20855 Gm2086...   \n",
       "51511                                            Gm21294   \n",
       "51512                                            Gm21996   \n",
       "51513                                            Gm20837   \n",
       "51514  Gm20838 Gm20837 Gm21094 Gm21256 Gm21739 Gm2186...   \n",
       "\n",
       "                   Organism  Length  \n",
       "0      Mus musculus (Mouse)     647  \n",
       "1      Mus musculus (Mouse)    2095  \n",
       "2      Mus musculus (Mouse)    1371  \n",
       "3      Mus musculus (Mouse)     419  \n",
       "4      Mus musculus (Mouse)     169  \n",
       "...                     ...     ...  \n",
       "51510  Mus musculus (Mouse)     222  \n",
       "51511  Mus musculus (Mouse)     222  \n",
       "51512  Mus musculus (Mouse)     222  \n",
       "51513  Mus musculus (Mouse)     188  \n",
       "51514  Mus musculus (Mouse)     166  \n",
       "\n",
       "[51515 rows x 8 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from_type = 'Ensembl'\n",
    "to_type = 'UniProtKB'\n",
    "\n",
    "uniprot_idmm_object = genelist_object.get_uniprot_ids(ID_MAPPER_LOC, from_type, to_type)\n",
    "sample_BFD.add_keyfiles({'uniprot_output_idmm': uniprot_idmm_object})\n",
    "\n",
    "uniprot_idmm = pd.read_csv(sample_BFD.uniprot_output_idmm.path, sep = '\\t')\n",
    "display(uniprot_idmm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 7. Extract results and generate Uniprot IDMM\n",
    "Generates an idmm that links `gene_name`, the `dbxref_datafield` seleted above, and `uniprot_id` returned by API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_93813/3311337425.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  uniprot_idpairs.rename(columns = {'From': dbxref_datafield, 'Entry': 'uniprot_id'}, inplace = True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "      <td>Q5GH67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000025900</td>\n",
       "      <td>P56716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000025900</td>\n",
       "      <td>A0A140LHJ6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000025902</td>\n",
       "      <td>Q61473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000025902</td>\n",
       "      <td>A0A0A6YWS4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51510</th>\n",
       "      <td>ENSMUSG00000099856</td>\n",
       "      <td>A0A087WRK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51511</th>\n",
       "      <td>ENSMUSG00000102045</td>\n",
       "      <td>A0A087WS79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51512</th>\n",
       "      <td>ENSMUSG00000100608</td>\n",
       "      <td>A0A087WSS9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51513</th>\n",
       "      <td>ENSMUSG00000096178</td>\n",
       "      <td>A0A087WP97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51514</th>\n",
       "      <td>ENSMUSG00000096178</td>\n",
       "      <td>J3QP42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>51515 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               ensembl_id  uniprot_id\n",
       "0      ENSMUSG00000051951      Q5GH67\n",
       "1      ENSMUSG00000025900      P56716\n",
       "2      ENSMUSG00000025900  A0A140LHJ6\n",
       "3      ENSMUSG00000025902      Q61473\n",
       "4      ENSMUSG00000025902  A0A0A6YWS4\n",
       "...                   ...         ...\n",
       "51510  ENSMUSG00000099856  A0A087WRK1\n",
       "51511  ENSMUSG00000102045  A0A087WS79\n",
       "51512  ENSMUSG00000100608  A0A087WSS9\n",
       "51513  ENSMUSG00000096178  A0A087WP97\n",
       "51514  ENSMUSG00000096178      J3QP42\n",
       "\n",
       "[51515 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>uniprot_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "      <td>Q5GH67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Rp1</td>\n",
       "      <td>ENSMUSG00000025900</td>\n",
       "      <td>P56716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Rp1</td>\n",
       "      <td>ENSMUSG00000025900</td>\n",
       "      <td>A0A140LHJ6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sox17</td>\n",
       "      <td>ENSMUSG00000025902</td>\n",
       "      <td>Q61473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Sox17</td>\n",
       "      <td>ENSMUSG00000025902</td>\n",
       "      <td>A0A0A6YWS4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51510</th>\n",
       "      <td>Gm20906</td>\n",
       "      <td>ENSMUSG00000099856</td>\n",
       "      <td>A0A087WRK1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51511</th>\n",
       "      <td>Gm21294</td>\n",
       "      <td>ENSMUSG00000102045</td>\n",
       "      <td>A0A087WS79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51512</th>\n",
       "      <td>Gm21996</td>\n",
       "      <td>ENSMUSG00000100608</td>\n",
       "      <td>A0A087WSS9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51513</th>\n",
       "      <td>Gm20837</td>\n",
       "      <td>ENSMUSG00000096178</td>\n",
       "      <td>A0A087WP97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51514</th>\n",
       "      <td>Gm20837</td>\n",
       "      <td>ENSMUSG00000096178</td>\n",
       "      <td>J3QP42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>51515 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_name          ensembl_id  uniprot_id\n",
       "0          Xkr4  ENSMUSG00000051951      Q5GH67\n",
       "1           Rp1  ENSMUSG00000025900      P56716\n",
       "2           Rp1  ENSMUSG00000025900  A0A140LHJ6\n",
       "3         Sox17  ENSMUSG00000025902      Q61473\n",
       "4         Sox17  ENSMUSG00000025902  A0A0A6YWS4\n",
       "...         ...                 ...         ...\n",
       "51510   Gm20906  ENSMUSG00000099856  A0A087WRK1\n",
       "51511   Gm21294  ENSMUSG00000102045  A0A087WS79\n",
       "51512   Gm21996  ENSMUSG00000100608  A0A087WSS9\n",
       "51513   Gm20837  ENSMUSG00000096178  A0A087WP97\n",
       "51514   Gm20837  ENSMUSG00000096178      J3QP42\n",
       "\n",
       "[51515 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "uniprot_idpairs = uniprot_idmm[['From', 'Entry']]\n",
    "uniprot_idpairs.rename(columns = {'From': dbxref_datafield, 'Entry': 'uniprot_id'}, inplace = True)\n",
    "display(uniprot_idpairs)\n",
    "\n",
    "uniprot_output_idmm = models_subset.merge(uniprot_idpairs, on = dbxref_datafield)\n",
    "display(uniprot_output_idmm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate a filename and file for the idmm\n",
    "uniprot_output_idmm_filename = '_'.join([sample_BFD.species_prefix, conditions, 'uniprot-idmm.tsv'])\n",
    "uniprot_output_idmm_object = IdmmFile(\n",
    "    filename = uniprot_output_idmm_filename, \n",
    "    sampledict = sample_BFD.sampledict, \n",
    "    kind = 'uniprot_idmm', \n",
    "    sources = [sample_BFD.annot]\n",
    ")\n",
    "\n",
    "# save to file and add to the BioFileDocket\n",
    "uniprot_output_idmm.to_csv(uniprot_output_idmm_object.path, sep = '\\t')\n",
    "sample_BFD.add_keyfile('uniprot_idmm', uniprot_output_idmm_object)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 8. Convert GFF to GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted file GCF_000001635.23_GRCm38.p3_genomic.gtf already exists at:\n",
      " /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/GCF_000001635.23_GRCm38.p3_genomic.gtf\n"
     ]
    }
   ],
   "source": [
    "# convert the GFF file to GTF using gffread\n",
    "models_asgtf = sample_BFD.annot.to_gtf(GFFREAD_LOC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3199731</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3199731</td>\n",
       "      <td>3207317</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3213439</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224794</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>14145</td>\n",
       "      <td>15288</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"gene48832\"; gene_name \"CYTB\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224795</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111579\"; gene_id \"gene48833\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224796</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111579\"; gene_id \"gene48833\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224797</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111580\"; gene_id \"gene48834\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224798</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111580\"; gene_id \"gene48834\"...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2224799 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0       1           2        3        4  5  6  7  \\\n",
       "0        NC_000067.6  Gnomon  transcript  3199731  3671742  .  -  .   \n",
       "1        NC_000067.6  Gnomon        exon  3199731  3207317  .  -  .   \n",
       "2        NC_000067.6  Gnomon        exon  3213439  3216968  .  -  .   \n",
       "3        NC_000067.6  Gnomon        exon  3421702  3421901  .  -  .   \n",
       "4        NC_000067.6  Gnomon        exon  3670552  3671742  .  -  .   \n",
       "...              ...     ...         ...      ...      ... .. .. ..   \n",
       "2224794  NC_005089.1  RefSeq         CDS    14145    15288  .  +  0   \n",
       "2224795  NC_005089.1  RefSeq  transcript    15289    15355  .  +  .   \n",
       "2224796  NC_005089.1  RefSeq        exon    15289    15355  .  +  .   \n",
       "2224797  NC_005089.1  RefSeq  transcript    15356    15422  .  -  .   \n",
       "2224798  NC_005089.1  RefSeq        exon    15356    15422  .  -  .   \n",
       "\n",
       "                                                         8  \n",
       "0        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "1        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "2        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "3        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "4        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...  \n",
       "...                                                    ...  \n",
       "2224794       transcript_id \"gene48832\"; gene_name \"CYTB\";  \n",
       "2224795  transcript_id \"rna111579\"; gene_id \"gene48833\"...  \n",
       "2224796  transcript_id \"rna111579\"; gene_id \"gene48833\"...  \n",
       "2224797  transcript_id \"rna111580\"; gene_id \"gene48834\"...  \n",
       "2224798  transcript_id \"rna111580\"; gene_id \"gene48834\"...  \n",
       "\n",
       "[2224799 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "'transcript_id \"rna0\"; gene_id \"gene0\"; gene_name \"Xkr4\";'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# load the newly-generated GTF file as a dataframe\n",
    "models_asgtf_df = pd.read_csv(models_asgtf.path, skiprows = 0, header = None, sep = '\\t')\n",
    "\n",
    "display(models_asgtf_df)\n",
    "display(models_asgtf_df[attributes_column][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>field_dictionary</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3199731</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3199731</td>\n",
       "      <td>3207317</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3213439</td>\n",
       "      <td>3216968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3421702</td>\n",
       "      <td>3421901</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_000067.6</td>\n",
       "      <td>Gnomon</td>\n",
       "      <td>exon</td>\n",
       "      <td>3670552</td>\n",
       "      <td>3671742</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...</td>\n",
       "      <td>{'transcript_id': 'rna0', 'gene_id': 'gene0', ...</td>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224793</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14145</td>\n",
       "      <td>15288</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"gene48832\"; gene_id \"gene48832\"...</td>\n",
       "      <td>{'transcript_id': 'gene48832', 'gene_id': 'gen...</td>\n",
       "      <td>CYTB</td>\n",
       "      <td>gene48832</td>\n",
       "      <td>gene48832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224795</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111579\"; gene_id \"gene48833\"...</td>\n",
       "      <td>{'transcript_id': 'rna111579', 'gene_id': 'gen...</td>\n",
       "      <td>TrnT</td>\n",
       "      <td>gene48833</td>\n",
       "      <td>rna111579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224796</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15289</td>\n",
       "      <td>15355</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111579\"; gene_id \"gene48833\"...</td>\n",
       "      <td>{'transcript_id': 'rna111579', 'gene_id': 'gen...</td>\n",
       "      <td>TrnT</td>\n",
       "      <td>gene48833</td>\n",
       "      <td>rna111579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224797</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111580\"; gene_id \"gene48834\"...</td>\n",
       "      <td>{'transcript_id': 'rna111580', 'gene_id': 'gen...</td>\n",
       "      <td>TrnP</td>\n",
       "      <td>gene48834</td>\n",
       "      <td>rna111580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224798</th>\n",
       "      <td>NC_005089.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>exon</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"rna111580\"; gene_id \"gene48834\"...</td>\n",
       "      <td>{'transcript_id': 'rna111580', 'gene_id': 'gen...</td>\n",
       "      <td>TrnP</td>\n",
       "      <td>gene48834</td>\n",
       "      <td>rna111580</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1287511 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0       1           2        3        4  5  6  7  \\\n",
       "0        NC_000067.6  Gnomon  transcript  3199731  3671742  .  -  .   \n",
       "1        NC_000067.6  Gnomon        exon  3199731  3207317  .  -  .   \n",
       "2        NC_000067.6  Gnomon        exon  3213439  3216968  .  -  .   \n",
       "3        NC_000067.6  Gnomon        exon  3421702  3421901  .  -  .   \n",
       "4        NC_000067.6  Gnomon        exon  3670552  3671742  .  -  .   \n",
       "...              ...     ...         ...      ...      ... .. .. ..   \n",
       "2224793  NC_005089.1  RefSeq  transcript    14145    15288  .  +  .   \n",
       "2224795  NC_005089.1  RefSeq  transcript    15289    15355  .  +  .   \n",
       "2224796  NC_005089.1  RefSeq        exon    15289    15355  .  +  .   \n",
       "2224797  NC_005089.1  RefSeq  transcript    15356    15422  .  -  .   \n",
       "2224798  NC_005089.1  RefSeq        exon    15356    15422  .  -  .   \n",
       "\n",
       "                                                         8  \\\n",
       "0        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "1        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "2        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "3        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "4        transcript_id \"rna0\"; gene_id \"gene0\"; gene_na...   \n",
       "...                                                    ...   \n",
       "2224793  transcript_id \"gene48832\"; gene_id \"gene48832\"...   \n",
       "2224795  transcript_id \"rna111579\"; gene_id \"gene48833\"...   \n",
       "2224796  transcript_id \"rna111579\"; gene_id \"gene48833\"...   \n",
       "2224797  transcript_id \"rna111580\"; gene_id \"gene48834\"...   \n",
       "2224798  transcript_id \"rna111580\"; gene_id \"gene48834\"...   \n",
       "\n",
       "                                          field_dictionary gene_name  \\\n",
       "0        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...      Xkr4   \n",
       "1        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...      Xkr4   \n",
       "2        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...      Xkr4   \n",
       "3        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...      Xkr4   \n",
       "4        {'transcript_id': 'rna0', 'gene_id': 'gene0', ...      Xkr4   \n",
       "...                                                    ...       ...   \n",
       "2224793  {'transcript_id': 'gene48832', 'gene_id': 'gen...      CYTB   \n",
       "2224795  {'transcript_id': 'rna111579', 'gene_id': 'gen...      TrnT   \n",
       "2224796  {'transcript_id': 'rna111579', 'gene_id': 'gen...      TrnT   \n",
       "2224797  {'transcript_id': 'rna111580', 'gene_id': 'gen...      TrnP   \n",
       "2224798  {'transcript_id': 'rna111580', 'gene_id': 'gen...      TrnP   \n",
       "\n",
       "           gene_id transcript_id  \n",
       "0            gene0          rna0  \n",
       "1            gene0          rna0  \n",
       "2            gene0          rna0  \n",
       "3            gene0          rna0  \n",
       "4            gene0          rna0  \n",
       "...            ...           ...  \n",
       "2224793  gene48832     gene48832  \n",
       "2224795  gene48833     rna111579  \n",
       "2224796  gene48833     rna111579  \n",
       "2224797  gene48834     rna111580  \n",
       "2224798  gene48834     rna111580  \n",
       "\n",
       "[1287511 rows x 13 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Use a custom function to extract useful fields from the additional fields section (column 8)\n",
    "# Pull from that dict to fill in additional useful columns\n",
    "models_asgtf_df['field_dictionary'] = models_asgtf_df[attributes_column].apply(convert_fields_to_dict_gtf)\n",
    "models_asgtf_df['gene_name'] = [d.get('gene_name') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['gene_id'] = [d.get('gene_id') for d in models_asgtf_df['field_dictionary']]\n",
    "models_asgtf_df['transcript_id'] = [d.get('transcript_id') for d in models_asgtf_df['field_dictionary']]\n",
    "\n",
    "# Remove CDS annotations because they interfere with TransDecoder cDNA generation\n",
    "models_asgtf_df = models_asgtf_df[models_asgtf_df[2] != 'CDS']\n",
    "display(models_asgtf_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9. Generate gtf-idmm\n",
    "This file maps the `gene_name` to `gene_id` and `transcript_id` fields generated by the conversion from GFF to GTF, which will be needed for downstream processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Xkr4</td>\n",
       "      <td>gene0</td>\n",
       "      <td>rna3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>LOC105243853</td>\n",
       "      <td>gene2</td>\n",
       "      <td>rna4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224789</th>\n",
       "      <td>ND6</td>\n",
       "      <td>gene48830</td>\n",
       "      <td>gene48830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224791</th>\n",
       "      <td>TrnE</td>\n",
       "      <td>gene48831</td>\n",
       "      <td>rna111578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224793</th>\n",
       "      <td>CYTB</td>\n",
       "      <td>gene48832</td>\n",
       "      <td>gene48832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224795</th>\n",
       "      <td>TrnT</td>\n",
       "      <td>gene48833</td>\n",
       "      <td>rna111579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224797</th>\n",
       "      <td>TrnP</td>\n",
       "      <td>gene48834</td>\n",
       "      <td>rna111580</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>116982 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_name    gene_id transcript_id\n",
       "0                Xkr4      gene0          rna0\n",
       "8                Xkr4      gene0          rna1\n",
       "15               Xkr4      gene0          rna2\n",
       "22               Xkr4      gene0          rna3\n",
       "29       LOC105243853      gene2          rna4\n",
       "...               ...        ...           ...\n",
       "2224789           ND6  gene48830     gene48830\n",
       "2224791          TrnE  gene48831     rna111578\n",
       "2224793          CYTB  gene48832     gene48832\n",
       "2224795          TrnT  gene48833     rna111579\n",
       "2224797          TrnP  gene48834     rna111580\n",
       "\n",
       "[116982 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract gene_name, gene_id, and transcript_id fields to generate an ID mapping matrix (idmm)\n",
    "idmm_df = models_asgtf_df[['gene_name', 'gene_id', 'transcript_id']].drop_duplicates()\n",
    "idmm_df.dropna(inplace = True)\n",
    "display(idmm_df)\n",
    "\n",
    "# generate a filename and file for the idmm\n",
    "idmm_filename = '_'.join([sample_BFD.species_prefix, conditions, 'gtf-idmm.tsv'])\n",
    "idmm = IdmmFile(\n",
    "    filename = idmm_filename, \n",
    "    sampledict = sample_BFD.sampledict, \n",
    "    kind = 'gtf_idmm', \n",
    "    sources = [sample_BFD.annot]\n",
    ")\n",
    "\n",
    "# save to file and add to the BioFileDocket\n",
    "idmm_df.to_csv(idmm.path, sep = '\\t')\n",
    "sample_BFD.add_keyfile('gtf_idmm', idmm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 10. Generate updated gtf\n",
    "Generated an updated GTF file using transcript_id as the key. For some datasets, transcripts do not consistently get gene names and gene IDs added, which causes Transdecoder to throw errors. This resolves that problem."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "models_asgtf_updated_df = models_asgtf_df.merge(idmm_df, on = 'transcript_id')\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_name': x['gene_name_y']}), axis = 1)\n",
    "models_asgtf_updated_df.apply(lambda x: x['field_dictionary'].update({'gene_id': x['gene_id_y']}), axis = 1)\n",
    "models_asgtf_updated_df['new_fields'] = models_asgtf_updated_df['field_dictionary'].apply(convert_dict_to_fields_gtf)\n",
    "models_asgtf_updated_df = models_asgtf_updated_df[[0, 1, 2, 3, 4, 5, 6, 7, 'new_fields']]\n",
    "\n",
    "models_asgtf_updated_filename = models_asgtf.filename.replace('.gtf', '_updated.gtf')\n",
    "models_asgtf_updated = GenomeGtfFile(\n",
    "    filename = models_asgtf_updated_filename, \n",
    "    sampledict = sample_BFD.sampledict, \n",
    "    reference_genome = sample_BFD.genome_fasta\n",
    ")\n",
    "\n",
    "models_asgtf_updated_df.to_csv(models_asgtf_updated.path, header = None, index = None, sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 11. Generate cDNA and peptide files\n",
    "Using the updated gtf file and genome file, generate cDNA sequence.\n",
    "\n",
    "Then, using the cDNA sequence, generate peptide sequences using transdecoder.\n",
    "\n",
    "Expect this step to take some time, probably ~20-30min."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/compute_base_probs.pl /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna 0 > /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//base_freqs.dat, checkpoint [/home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir/.__checkpoints_longorfs/base_freqs_file.ok] exists.\n",
      "-skipping long orf extraction, already completed earlier as per checkpoint: /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir/.__checkpoints_longorfs/TD.longorfs.ok\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.cds 5000 5000 > /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000, checkpoint [/home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir/.__checkpoints/get_longest_orfs.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/exclude_similar_proteins.pl /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000 > /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr, checkpoint [/home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir/.__checkpoints/nr.ok] exists.\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/get_top_longest_fasta_entries.pl /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.cds.top_longest_5000.nr 500 > /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.cds.top_500_longest, checkpoint [/home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir/.__checkpoints/top_train_select.ok] exists.\n",
      "PCT_GC: 49\n",
      "-- Skipping CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/seq_n_baseprobs_to_loglikelihood_vals.pl /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.cds.top_500_longest /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//base_freqs.dat > /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//hexamer.scores, checkpoint [/home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir/.__checkpoints/hexamer_scores.ok] exists.\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/score_CDS_likelihood_all_6_frames.pl /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.cds /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//hexamer.scores > /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.cds.scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/select_best_ORFs_per_transcript.pl --gff3_file /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.gff3 --cds_scores /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.cds.scores  --min_length_auto_accept 749  > /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 \n",
      "Selecting best orfs\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/train_start_PWM.pl --transcripts /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna --selected_orfs /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.cds.top_500_longest --out_prefix /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement\n",
      "Training start codon pattern recognition* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/build_atgPWM_+-.pl  --transcripts /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna  --selected_orfs /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.cds.top_500_longest  --out_prefix /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement --pwm_left 20 --pwm_right 10 \n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scoring.+-.pl  --features_plus /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.+.features  --features_minus /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.-.features  --atg_position 20  > /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.feature.scores\n",
      "-round: 1\n",
      "-round: 2\n",
      "-round: 3\n",
      "-round: 4\n",
      "-round: 5\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scores_to_ROC.pl /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.feature.scores > /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.feature.scores.roc\n",
      "-parsing scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/plot_ROC.Rscript /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.feature.scores.roc || :\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/compute_AUC.pl /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.feature.scores.roc\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.+.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.-.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/deplete_feature_noise.pl  --features_plus /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.+.features  --pwm_minus /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.-.pwm  --out_prefix /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.enhanced\n",
      "num features: 372\tnum_incorporate: 111\n",
      "-feature swap of score: -3.07961583728188 instead of -5.54882077937692\n",
      "-feature swap of score: -3.30776323967247 instead of -3.72508370886427\n",
      "-feature swap of score: -0.0566651310729297 instead of -3.48745360339976\n",
      "-feature swap of score: -0.1066035017871 instead of -3.35751991183622\n",
      "-feature swap of score: -2.32688324568908 instead of -3.51672790760349\n",
      "-feature swap of score: 1.8114796564168 instead of -2.94059170574088\n",
      "-feature swap of score: 4.83802230792251 instead of -2.28287526955426\n",
      "-feature swap of score: 7.2633675769597 instead of -2.22193955042837\n",
      "-feature swap of score: 2.64429947764902 instead of -2.26881819278539\n",
      "-feature swap of score: -0.633073690423108 instead of -2.28933764829901\n",
      "-feature swap of score: 0.807602350694878 instead of -2.14869304172604\n",
      "-feature swap of score: 1.87849758385338 instead of -2.04409251007407\n",
      "-feature swap of score: 2.53508845401154 instead of -1.92154403311267\n",
      "-feature swap of score: 1.4827443183465 instead of -2.23457874458801\n",
      "-feature swap of score: -1.90263860146324 instead of -1.95656737532091\n",
      "-feature swap of score: 5.30053833348517 instead of -1.81579476032879\n",
      "-feature swap of score: 2.57898389281976 instead of -1.95514639540647\n",
      "-feature swap of score: 3.85965337184732 instead of -1.8335029489198\n",
      "-feature swap of score: -1.19241575565706 instead of -2.04545075914044\n",
      "-feature swap of score: 2.40154644030053 instead of -1.71522649701155\n",
      "-feature swap of score: -1.9937470054408 instead of -2.13408354386525\n",
      "-feature swap of score: -0.515168612958254 instead of -1.61899135589908\n",
      "-feature swap of score: 4.31453632600487 instead of -1.46969779936152\n",
      "-feature swap of score: -1.01860791073695 instead of -1.19753773431071\n",
      "-feature swap of score: 6.41435467912981 instead of -0.954438968886407\n",
      "-feature swap of score: 0.217574663928277 instead of -1.18472503318556\n",
      "-feature swap of score: 4.23538083729836 instead of -1.21672531036628\n",
      "-feature swap of score: -0.760978331870938 instead of -1.14188563293433\n",
      "-feature swap of score: -0.854186750053127 instead of -1.13473699052445\n",
      "-feature swap of score: 0.0120390683483282 instead of -1.34665761789958\n",
      "-feature swap of score: 2.55509736758517 instead of -1.18709587017204\n",
      "-feature swap of score: 1.33663230455492 instead of -1.25457651957661\n",
      "-feature swap of score: 0.798616039694534 instead of -1.2409138289344\n",
      "-feature swap of score: 7.29447784238645 instead of -0.960638422446972\n",
      "-feature swap of score: 1.34693539231036 instead of -1.20951809428308\n",
      "-feature swap of score: 1.42258513632555 instead of -0.984013896475219\n",
      "-feature swap of score: 5.74367384048423 instead of -0.795168345234279\n",
      "-feature swap of score: 2.41873604984524 instead of -0.545796957788624\n",
      "-feature swap of score: 2.75232358566149 instead of -0.509794672621013\n",
      "-feature swap of score: 0.657365576225011 instead of -0.521509140907423\n",
      "-feature swap of score: 1.9901835456408 instead of -0.512171191198489\n",
      "-feature swap of score: 4.62328806293469 instead of -0.642419692918477\n",
      "-feature swap of score: 0.567205789631084 instead of -0.952989167407225\n",
      "-feature swap of score: 0.695385539421008 instead of -0.845108737093088\n",
      "-feature swap of score: 3.75711985634357 instead of -0.814618152482949\n",
      "-feature swap of score: 1.71148630501813 instead of -1.00344950475216\n",
      "-feature swap of score: 0.306022872551889 instead of -0.669123887530112\n",
      "-feature swap of score: 0.326132468387415 instead of -0.429437425138595\n",
      "-feature swap of score: 0.79240355718279 instead of -0.343762288130981\n",
      "-feature swap of score: 7.65965837081165 instead of -0.483541227455327\n",
      "-feature swap of score: -0.162159106088733 instead of -0.518700627007177\n",
      "-feature swap of score: 8.0476308141254 instead of -0.734206033537888\n",
      "-feature swap of score: -0.397641012760179 instead of -0.683454203075261\n",
      "-feature swap of score: -0.722595975422018 instead of -0.73241965022018\n",
      "-feature swap of score: 0.782800178294548 instead of -0.384750547137005\n",
      "-feature swap of score: -0.150680796028597 instead of -0.191588582542292\n",
      "-feature swap of score: 0.0655944712063897 instead of -0.222955940456792\n",
      "-feature swap of score: 3.22244270057735 instead of -0.271188996517164\n",
      "-feature swap of score: 6.8554938406139 instead of -0.26956239060823\n",
      "-feature swap of score: 7.61694439089122 instead of -0.181865994490802\n",
      "-feature swap of score: 6.07358934118585 instead of 0.184900218191189\n",
      "-feature swap of score: 3.60167698213654 instead of 0.0912999728967031\n",
      "-feature swap of score: 4.67632131402229 instead of 0.00738384002417303\n",
      "-feature swap of score: 1.09594113050263 instead of 0.0216855118623636\n",
      "-feature swap of score: 4.96606951816637 instead of -0.0531526549668888\n",
      "-feature swap of score: 1.50873675635421 instead of -0.0537301405745084\n",
      "-feature swap of score: 0.0982920630932862 instead of -0.207428492241072\n",
      "-feature swap of score: 4.04999261725813 instead of 0.0893332263972173\n",
      "-feature swap of score: 0.29947565857975 instead of 0.124841587813719\n",
      "-feature swap of score: 1.64098297822191 instead of 0.257913417748148\n",
      "-feature swap of score: 1.70707542372114 instead of 0.168516997593578\n",
      "-feature swap of score: 6.01008163467429 instead of 0.168439364183266\n",
      "-feature swap of score: 1.29865694346257 instead of 0.114417077630474\n",
      "-feature swap of score: 6.65179528171786 instead of 0.0850952342549961\n",
      "-feature swap of score: 5.33194515890524 instead of 0.116080483215178\n",
      "-feature swap of score: 11.834744428935 instead of 0.143562338729346\n",
      "-feature swap of score: 2.99864674855024 instead of 0.084371877847171\n",
      "-feature swap of score: 0.900074989311438 instead of -0.249270048125017\n",
      "-feature swap of score: 2.09038662023652 instead of -0.000791353306212828\n",
      "-feature swap of score: 2.02651137732016 instead of 0.143004270951102\n",
      "-feature swap of score: 6.41201594952697 instead of 0.158735813342829\n",
      "-feature swap of score: 4.27062108015547 instead of 0.318303055944072\n",
      "-feature swap of score: 2.7791210440446 instead of 0.33238484355153\n",
      "-feature swap of score: 5.27972273736135 instead of 0.189744106355082\n",
      "-feature swap of score: 4.7092729970992 instead of 0.192614734839833\n",
      "-feature swap of score: 5.97662558282433 instead of 0.526943616177558\n",
      "-feature swap of score: 3.00315861384792 instead of 0.249457289995526\n",
      "-feature swap of score: 10.1296442880705 instead of 0.682588532868068\n",
      "-feature swap of score: 8.78774463359561 instead of 0.618495741513837\n",
      "-feature swap of score: 2.37201194624862 instead of 0.531176531335652\n",
      "-feature swap of score: 5.80850539794257 instead of 0.665206674836657\n",
      "-feature swap of score: 1.54010964335504 instead of 0.850949971706014\n",
      "-feature swap of score: 3.62876869147803 instead of 1.01998231836701\n",
      "-feature swap of score: 5.38947098180129 instead of 1.00302332558127\n",
      "-feature swap of score: 1.79773066447291 instead of 1.21446563395174\n",
      "-feature swap of score: 4.42909149489641 instead of 0.974314502676825\n",
      "-feature swap of score: 1.96932596327398 instead of 0.810685469521272\n",
      "-feature swap of score: 5.4568442861618 instead of 0.909178832882726\n",
      "-feature swap of score: 3.62938596943756 instead of 0.857599261887494\n",
      "-feature swap of score: 7.05389482473966 instead of -0.267863285375342\n",
      "-feature swap of score: 1.38165908531913 instead of 1.18023429410558\n",
      "-feature swap of score: 3.91473902499198 instead of 1.15154078317504\n",
      "-feature swap of score: 1.83031848191679 instead of 1.28215172047157\n",
      "-feature swap of score: 8.86213235385091 instead of 1.2417456082789\n",
      "-feature swap of score: 1.35002865912594 instead of 1.3342118173477\n",
      "-feature swap of score: 2.01253195931627 instead of 1.7504867644111\n",
      "-feature swap of score: 9.61342473815962 instead of 1.77130004315643\n",
      "-feature swap of score: 2.71604825386934 instead of 1.49935143371896\n",
      "-feature swap of score: 3.10767284907619 instead of 1.75352645364397\n",
      "-feature swap of score: 11.8220561028179 instead of 1.75441353620049\n",
      "-feature swap of score: 4.57170559326327 instead of 1.74457413011621\n",
      "-num feature swaps: 111\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scoring.+-.pl  --features_plus /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.enhanced.+.features  --features_minus /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.-.features  --atg_position 20  > /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.enhanced.feature.scores\n",
      "-round: 1\n",
      "-round: 2\n",
      "-round: 3\n",
      "-round: 4\n",
      "-round: 5\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/feature_scores_to_ROC.pl /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.enhanced.feature.scores > /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc\n",
      "-parsing scores\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/plot_ROC.Rscript /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc || :\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/compute_AUC.pl /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.enhanced.feature.scores.roc\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null device \n",
      "          1 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/PWM/make_seqLogo.Rscript /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//start_refinement.enhanced.+.pwm || :\n",
      "Error in library(seqLogo) : there is no package called ‘seqLogo’\n",
      "Execution halted\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/start_codon_refinement.pl --transcripts /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna --gff3_file /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3 --workdir /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir/ > /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3\n",
      "Refining start codon selections.\n",
      "-number of revised start positions: 8692\n",
      "* Running CMD: cp /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/Mus_musculus_adolescentbrain.transdecoder_dir//longest_orfs.cds.best_candidates.gff3.revised_starts.gff3 GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3\n",
      "copying output to final output file: GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_bed.pl GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3 > GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.bed\n",
      "Making bed file: GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.bed\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3 --fasta /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna  --genetic_code Universal > GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep\n",
      "Making pep file: GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep\n",
      "* Running CMD: /home/ec2-user/miniconda3/pkgs/transdecoder-5.5.0-pl526_1/opt/transdecoder/util/gff3_file_to_proteins.pl --gff3 GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3 --fasta /home/ec2-user/glial-origins/output/Mmus_adolescentbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna --seqType CDS  --genetic_code Universal > GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.cds\n",
      "Making cds file: GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.cds\n",
      "transdecoder is finished.  See output files GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.*\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cdna = sample_BFD.genome_fasta.get_transdecoder_cdna_gtf(models_asgtf_updated, TRANSDECODER_LOC)\n",
    "sample_BFD.add_keyfile('cdna', cdna)\n",
    "\n",
    "transdecoder_files = sample_BFD.cdna.to_pep_files(TDLONGORF_LOC, TDPREDICT_LOC)\n",
    "sample_BFD.add_keyfiles(transdecoder_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "iEug6TQ6Cc9D"
   },
   "source": [
    "# 12. Push files to AWS S3\n",
    "\n",
    "Iteratively moves through the file_set and file_dict variables and populates files into the right place in AWS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_000001635.23_GRCm38.p3_genomic.gff already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "upload: ../../output/Mmus_adolescentbrain/10X05_1.loom.asgxc.tsv to s3://arcadia-reference-datasets/organisms/Mus_musculus/functional_sequencing/scRNA-Seq/10X05_1.loom.asgxc.tsv\n",
      "upload: ../../output/Mmus_adolescentbrain/10X05_1.loom to s3://arcadia-reference-datasets/organisms/Mus_musculus/functional_sequencing/scRNA-Seq/10X05_1.loom\n",
      "upload: ../../output/Mmus_adolescentbrain/10X05_1.loom.idmm.tsv to s3://arcadia-reference-datasets/organisms/Mus_musculus/genomics_reference/mapping_file/10X05_1.loom.idmm.tsv\n",
      "upload: ../../output/Mmus_adolescentbrain/10X05_1.loom.cellannot.tsv to s3://arcadia-reference-datasets/organisms/Mus_musculus/functional_sequencing/scRNA-Seq/10X05_1.loom.cellannot.tsv\n",
      "upload: ../../output/Mmus_adolescentbrain/Mmus_adolescentbrain_ensembl_id_UniProtIDs.txt to s3://arcadia-reference-datasets/organisms/Mus_musculus/genomics_reference/mapping_file/Mmus_adolescentbrain_ensembl_id_UniProtIDs.txt\n",
      "upload: ../../output/Mmus_adolescentbrain/Mmus_adolescentbrain_uniprot-idmm.tsv to s3://arcadia-reference-datasets/organisms/Mus_musculus/genomics_reference/mapping_file/Mmus_adolescentbrain_uniprot-idmm.tsv\n",
      "upload: ../../output/Mmus_adolescentbrain/Mmus_adolescentbrain_gtf-idmm.tsv to s3://arcadia-reference-datasets/organisms/Mus_musculus/genomics_reference/mapping_file/Mmus_adolescentbrain_gtf-idmm.tsv\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.bed already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.cds already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3 already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n"
     ]
    }
   ],
   "source": [
    "sample_BFD.local_to_s3()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 13. Pickle the `sample_BFD` variable for use by the next script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upload: ../../output/Mmus_adolescentbrain/Mmus_adolescentbrain_BioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/Mmus_adolescentbrain_BioFileDocket.pkl\n"
     ]
    }
   ],
   "source": [
    "# Generate a .pkl file for the Docket\n",
    "sample_BFD.pickle()\n",
    "\n",
    "# Push to S3, optionally overwriting existing file\n",
    "sample_BFD.push_to_s3(overwrite = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "Xtro_bro.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
