{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "fe72b412-070e-4c6a-8626-595386403737",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import os\n",
    "import dill\n",
    "import sys\n",
    "import datetime\n",
    "import numpy as np\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c4d9e051-6242-4387-afe6-f25b86dac7be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/ already exists\n",
      "file GCF_000002035.5_GRCz10_genomic.gff already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.gff\n",
      "file GCF_000002035.5_GRCz10_genomic.fna already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.fna\n",
      "file GSM3768152_Brain_8_dge.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GSM3768152_Brain_8_dge.txt\n",
      "file Drer_adultbrain_ZFIN_UniProtIDs.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_ZFIN_UniProtIDs.txt\n",
      "file Drer_adultbrain_uniprot-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_uniprot-idmm.tsv\n",
      "file Drer_adultbrain_gtf-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_gtf-idmm.tsv\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep\n",
      "file Drer_taxid_genes.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_taxid_genes.txt\n",
      "file Drer_adultbrain_sc_idmm.tsv already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_sc_idmm.tsv\n",
      "file DrerMmusXlae_adultbrain_struclusters_file.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_struclusters_file.tsv\n",
      "file GSM3768152_Brain_8_dge_asStruCluster.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GSM3768152_Brain_8_dge_asStruCluster.txt\n",
      "file DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv\n",
      "file Drer_adultbrain_og_idmm.tsv already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_og_idmm.tsv\n",
      "file OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv\n",
      "file GSM3768152_Brain_8_dge_asOrthogroup.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GSM3768152_Brain_8_dge_asOrthogroup.txt\n",
      "file DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv\n",
      "file Drer_adultbrain_Brain_8_cellannot.tsv already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_Brain_8_cellannot.tsv\n",
      "file Drer_FoldSeek_top_100_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_FoldSeek_top_100_StruCluster_ids.txt\n",
      "file Drer_FoldSeek_top_150_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_FoldSeek_top_150_StruCluster_ids.txt\n",
      "file Drer_FoldSeek_top_200_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_FoldSeek_top_200_StruCluster_ids.txt\n",
      "file Drer_Exc_top_250DE_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_Exc_top_250DE_StruCluster_ids.txt\n",
      "file Drer_Exc_top_250DE_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_Exc_top_250DE_StruCluster_ids.txt\n",
      "file Drer_FoldSeek_top_300_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_FoldSeek_top_300_StruCluster_ids.txt\n",
      "file Drer_OrthoFinder_top_100_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_OrthoFinder_top_100_Orthogroup_ids.txt\n",
      "file Drer_OrthoFinder_top_150_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_OrthoFinder_top_150_Orthogroup_ids.txt\n",
      "file Drer_Exc_top_200DE_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_Exc_top_200DE_Orthogroup_ids.txt\n",
      "file Drer_OrthoFinder_top_250_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_OrthoFinder_top_250_Orthogroup_ids.txt\n",
      "file Drer_Exc_top_200DE_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_Exc_top_200DE_Orthogroup_ids.txt\n",
      "file Drer_OrthoFinder_top_300_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_OrthoFinder_top_300_Orthogroup_ids.txt\n",
      "file GCF_000001635.23_GRCm38.p3_genomic.gff already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic.gff\n",
      "file GCF_000001635.23_GRCm38.p3_genomic.fna already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic.fna\n",
      "file GSM2906405_Brain1_dge_coerced.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GSM2906405_Brain1_dge_coerced.txt\n",
      "file Mmus_adultbrain_MGI_UniProtIDs.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_adultbrain_MGI_UniProtIDs.txt\n",
      "file Mmus_adultbrain_uniprot-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_adultbrain_uniprot-idmm.tsv\n",
      "file Mmus_adultbrain_gtf-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_adultbrain_gtf-idmm.tsv\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.bed already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.bed\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.cds already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.cds\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3 already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep\n",
      "file Mmus_adultbrain_Brain_1_cellannot.tsv already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_adultbrain_Brain_1_cellannot.tsv\n",
      "file GSM2906405_Brain1_dge_coerced_asOrthogroup.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GSM2906405_Brain1_dge_coerced_asOrthogroup.txt\n",
      "file DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv\n",
      "file Mmus_taxid_genes.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_taxid_genes.txt\n",
      "file GSM2906405_Brain1_dge_coerced_asStruCluster.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GSM2906405_Brain1_dge_coerced_asStruCluster.txt\n",
      "file DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv\n",
      "file Mmus_adultbrain_sc_idmm.tsv already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_adultbrain_sc_idmm.tsv\n",
      "file DrerMmusXlae_adultbrain_struclusters_file.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_struclusters_file.tsv\n",
      "file Mmus_adultbrain_og_idmm.tsv already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_adultbrain_og_idmm.tsv\n",
      "file OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv\n",
      "file Mmus_FoldSeek_top_100_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_FoldSeek_top_100_StruCluster_ids.txt\n",
      "file Mmus_FoldSeek_top_150_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_FoldSeek_top_150_StruCluster_ids.txt\n",
      "file Mmus_FoldSeek_top_200_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_FoldSeek_top_200_StruCluster_ids.txt\n",
      "file Mmus_Exc_top_250DE_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_Exc_top_250DE_StruCluster_ids.txt\n",
      "file Mmus_Exc_top_250DE_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_Exc_top_250DE_StruCluster_ids.txt\n",
      "file Mmus_FoldSeek_top_300_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_FoldSeek_top_300_StruCluster_ids.txt\n",
      "file Mmus_OrthoFinder_top_100_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_OrthoFinder_top_100_Orthogroup_ids.txt\n",
      "file Mmus_OrthoFinder_top_150_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_OrthoFinder_top_150_Orthogroup_ids.txt\n",
      "file Mmus_Exc_top_200DE_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_Exc_top_200DE_Orthogroup_ids.txt\n",
      "file Mmus_adultbrain_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_adultbrain_Orthogroup_ids.txt\n",
      "file Mmus_OrthoFinder_top_250_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_OrthoFinder_top_250_Orthogroup_ids.txt\n",
      "file Mmus_OrthoFinder_top_300_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_OrthoFinder_top_300_Orthogroup_ids.txt\n",
      "file Mmus_Exc_top_200DE_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_Exc_top_200DE_Orthogroup_ids.txt\n",
      "file XENLA_9.2_Xenbase.gtf already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_Xenbase.gtf\n",
      "file XENLA_9.2_GCA.gff already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_GCA.gff\n",
      "file XENLA_9.2_genome.fa already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome.fa\n",
      "file GSM6214268_Xenopus_brain_COL65_dge.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge.txt\n",
      "file Xlae_adultbrain_Xenbase_UniProtIDs.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_Xenbase_UniProtIDs.txt\n",
      "file Xlae_adultbrain_uniprot-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_uniprot-idmm.tsv\n",
      "file Xlae_adultbrain_gtf-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_gtf-idmm.tsv\n",
      "file XENLA_9.2_genome_cDNA.fa already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.bed already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.bed\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.cds already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.cds\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.gff3\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.pep already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.pep\n",
      "file Xlae_adultbrain_Brain3_4_cellannot.tsv already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_Brain3_4_cellannot.tsv\n",
      "file GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup.txt\n",
      "file DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv\n",
      "file Xlae_taxid_genes.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_taxid_genes.txt\n",
      "file GSM6214268_Xenopus_brain_COL65_dge_asStruCluster.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge_asStruCluster.txt\n",
      "file DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv\n",
      "file GSM6214268_Xenopus_brain_COL65_dge_subsampled.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge_subsampled.txt\n",
      "file GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup_subsampled.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup_subsampled.txt\n",
      "file Xlae_adultbrain_sc_idmm.tsv already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_sc_idmm.tsv\n",
      "file DrerMmusXlae_adultbrain_struclusters_file.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_struclusters_file.tsv\n",
      "file Xlae_adultbrain_og_idmm.tsv already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_og_idmm.tsv\n",
      "file OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv\n",
      "file GSM6214268_Xenopus_brain_COL65_dge_asStruCluster_subsampled.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge_asStruCluster_subsampled.txt\n",
      "file Xlae_FoldSeek_top_100_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_FoldSeek_top_100_StruCluster_ids.txt\n",
      "file Xlae_FoldSeek_top_150_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_FoldSeek_top_150_StruCluster_ids.txt\n",
      "file Xlae_FoldSeek_top_200_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_FoldSeek_top_200_StruCluster_ids.txt\n",
      "file Xlae_Exc_top_250DE_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_Exc_top_250DE_StruCluster_ids.txt\n",
      "file Xlae_Exc_top_250DE_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_Exc_top_250DE_StruCluster_ids.txt\n",
      "file Xlae_FoldSeek_top_300_StruCluster_ids.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_FoldSeek_top_300_StruCluster_ids.txt\n",
      "file Xlae_OrthoFinder_top_100_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_OrthoFinder_top_100_Orthogroup_ids.txt\n",
      "file Xlae_OrthoFinder_top_150_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_OrthoFinder_top_150_Orthogroup_ids.txt\n",
      "file Xlae_Exc_top_200DE_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_Exc_top_200DE_Orthogroup_ids.txt\n",
      "file Xlae_adultbrain_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_Orthogroup_ids.txt\n",
      "file Xlae_OrthoFinder_top_250_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_OrthoFinder_top_250_Orthogroup_ids.txt\n",
      "file Xlae_OrthoFinder_top_300_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_OrthoFinder_top_300_Orthogroup_ids.txt\n",
      "file Xlae_Exc_top_200DE_Orthogroup_ids.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_Exc_top_200DE_Orthogroup_ids.txt\n"
     ]
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "species_dict = {\n",
    "    'Danio_rerio': 'adultbrain',\n",
    "    'Mus_musculus': 'adultbrain',\n",
    "    'Xenopus_laevis': 'adultbrain'\n",
    "}\n",
    "\n",
    "global_conditions = 'adultbrain'\n",
    "\n",
    "analysis_type = 'OrthoFinder'\n",
    "\n",
    "################\n",
    "\n",
    "sample_MSD = MultiSpeciesBioFileDocket(\n",
    "    species_dict = species_dict,\n",
    "    global_conditions = global_conditions,\n",
    "    analysis_type = analysis_type\n",
    "    ).unpickle()\n",
    "\n",
    "sample_MSD.get_BioFileDockets()\n",
    "sample_MSD.s3_to_local()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "e0407af4-0fc7-4f02-b0e1-6cf55414e5de",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder</th>\n",
       "      <th>GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder</th>\n",
       "      <th>XENLA_9.2_genome_cDNA.fa.transdecoder</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OG0000000</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene10395_t.p1, gene12148_t.p1, gene1307_t.p1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OG0000001</td>\n",
       "      <td>[rna75711.p2]</td>\n",
       "      <td>[gene10036.p1, gene10201.p2, gene10535.p1, gen...</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OG0000002</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene2928_t.p3, gene38029_t.p2, gene41218_t.p1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OG0000003</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene264_t.p1, rna1157.p2, rna12598.p1, rna151...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OG0000004</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[gene50425_t.p1, rna27145.p2, rna34455.p1, rna...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32348</th>\n",
       "      <td>OG0032348</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[rna97220.p1, rna97221.p1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32349</th>\n",
       "      <td>OG0032349</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[rna97260.p1, rna97260.p2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32350</th>\n",
       "      <td>OG0032350</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[rna97447.p1, rna97448.p1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32351</th>\n",
       "      <td>OG0032351</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[rna97780.p1, rna97781.p2]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32352</th>\n",
       "      <td>OG0032352</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[rna9904.p1, rna9906.p1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32353 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Orthogroup GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder  \\\n",
       "0      OG0000000                                                 []         \n",
       "1      OG0000001                                      [rna75711.p2]         \n",
       "2      OG0000002                                                 []         \n",
       "3      OG0000003                                                 []         \n",
       "4      OG0000004                                                 []         \n",
       "...          ...                                                ...         \n",
       "32348  OG0032348                                                 []         \n",
       "32349  OG0032349                                                 []         \n",
       "32350  OG0032350                                                 []         \n",
       "32351  OG0032351                                                 []         \n",
       "32352  OG0032352                                                 []         \n",
       "\n",
       "      GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder  \\\n",
       "0                                                     []     \n",
       "1      [gene10036.p1, gene10201.p2, gene10535.p1, gen...     \n",
       "2                                                     []     \n",
       "3                                                     []     \n",
       "4                                                     []     \n",
       "...                                                  ...     \n",
       "32348                                                 []     \n",
       "32349                                                 []     \n",
       "32350                                                 []     \n",
       "32351                                                 []     \n",
       "32352                                                 []     \n",
       "\n",
       "                   XENLA_9.2_genome_cDNA.fa.transdecoder  \n",
       "0      [gene10395_t.p1, gene12148_t.p1, gene1307_t.p1...  \n",
       "1                                                     []  \n",
       "2      [gene2928_t.p3, gene38029_t.p2, gene41218_t.p1...  \n",
       "3      [gene264_t.p1, rna1157.p2, rna12598.p1, rna151...  \n",
       "4      [gene50425_t.p1, rna27145.p2, rna34455.p1, rna...  \n",
       "...                                                  ...  \n",
       "32348                         [rna97220.p1, rna97221.p1]  \n",
       "32349                         [rna97260.p1, rna97260.p2]  \n",
       "32350                         [rna97447.p1, rna97448.p1]  \n",
       "32351                         [rna97780.p1, rna97781.p2]  \n",
       "32352                           [rna9904.p1, rna9906.p1]  \n",
       "\n",
       "[32353 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "orthogroups_file = pd.read_csv(sample_MSD.orthogroups_file.path, sep = '\\t')\n",
    "\n",
    "sp_cols = [i for i in orthogroups_file.columns if i != 'Orthogroup']\n",
    "\n",
    "col_unpacker = lambda x: [i for i in x.split(', ')] if x is not np.nan else []\n",
    "\n",
    "for col in sp_cols:\n",
    "    orthogroups_file[col] = orthogroups_file[col].apply(col_unpacker)\n",
    "    \n",
    "display(orthogroups_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "e3ad5bea-3eb8-4da8-9615-3a0a16af45bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "og_idmm_dict = {}\n",
    "\n",
    "for sp in sample_MSD.species_BioFileDockets:\n",
    "    og_idmm_dict[sp] = pd.read_csv(sample_MSD.species_BioFileDockets[sp].og_idmm.path, sep = '\\t')\n",
    "\n",
    "og_mapper_dict = {}\n",
    "\n",
    "for sp in og_idmm_dict:\n",
    "    og_mapper_dict[sp] = dict(zip(og_idmm_dict[sp]['protein_id'], og_idmm_dict[sp]['gene_name']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "6f56eeea-173a-48e3-bab3-5fcec6c8a156",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder</th>\n",
       "      <th>GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder</th>\n",
       "      <th>XENLA_9.2_genome_cDNA.fa.transdecoder</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OG0000000</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{LOC108700417 [provisional:trim25], LOC1086994...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OG0000001</td>\n",
       "      <td>{Zfp389}</td>\n",
       "      <td>{zgc:174696, LOC108183688, LOC108183294, LOC10...</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OG0000002</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Xelaev18028992m, Xelaev18041830m, Xelaev18024...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OG0000003</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{Xelaev18000208m, LOC108709124, Xelaev18022432...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OG0000004</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{LOC108708057 [provisional:prkcq], LOC10871647...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32348</th>\n",
       "      <td>OG0032348</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{trim8.L}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32349</th>\n",
       "      <td>OG0032349</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{LOC108700451 [provisional:kcnj6]}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32350</th>\n",
       "      <td>OG0032350</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{gls.L}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32351</th>\n",
       "      <td>OG0032351</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{vstm5.L}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32352</th>\n",
       "      <td>OG0032352</td>\n",
       "      <td>{}</td>\n",
       "      <td>{}</td>\n",
       "      <td>{egln1.S}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32353 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Orthogroup GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder  \\\n",
       "0      OG0000000                                                 {}         \n",
       "1      OG0000001                                           {Zfp389}         \n",
       "2      OG0000002                                                 {}         \n",
       "3      OG0000003                                                 {}         \n",
       "4      OG0000004                                                 {}         \n",
       "...          ...                                                ...         \n",
       "32348  OG0032348                                                 {}         \n",
       "32349  OG0032349                                                 {}         \n",
       "32350  OG0032350                                                 {}         \n",
       "32351  OG0032351                                                 {}         \n",
       "32352  OG0032352                                                 {}         \n",
       "\n",
       "      GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder  \\\n",
       "0                                                     {}     \n",
       "1      {zgc:174696, LOC108183688, LOC108183294, LOC10...     \n",
       "2                                                     {}     \n",
       "3                                                     {}     \n",
       "4                                                     {}     \n",
       "...                                                  ...     \n",
       "32348                                                 {}     \n",
       "32349                                                 {}     \n",
       "32350                                                 {}     \n",
       "32351                                                 {}     \n",
       "32352                                                 {}     \n",
       "\n",
       "                   XENLA_9.2_genome_cDNA.fa.transdecoder  \n",
       "0      {LOC108700417 [provisional:trim25], LOC1086994...  \n",
       "1                                                     {}  \n",
       "2      {Xelaev18028992m, Xelaev18041830m, Xelaev18024...  \n",
       "3      {Xelaev18000208m, LOC108709124, Xelaev18022432...  \n",
       "4      {LOC108708057 [provisional:prkcq], LOC10871647...  \n",
       "...                                                  ...  \n",
       "32348                                          {trim8.L}  \n",
       "32349                 {LOC108700451 [provisional:kcnj6]}  \n",
       "32350                                            {gls.L}  \n",
       "32351                                          {vstm5.L}  \n",
       "32352                                          {egln1.S}  \n",
       "\n",
       "[32353 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sp_col_mapper = {'GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder': 'Mmus',\n",
    "                'GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder': 'Drer',\n",
    "                'XENLA_9.2_genome_cDNA.fa.transdecoder': 'Xlae'}\n",
    "\n",
    "def dict_mapper (x, col, col_mapper, mapper_dict):\n",
    "    key = col_mapper[col]\n",
    "    return set([mapper_dict[key][i] for i in x])\n",
    "\n",
    "for col in sp_cols:\n",
    "    orthogroups_file[col] = orthogroups_file[col].apply(lambda x: dict_mapper(x, col, sp_col_mapper, og_mapper_dict))\n",
    "\n",
    "display(orthogroups_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "a82abd0f-8c40-4b83-9ef8-902626f3f9dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orthogroup</th>\n",
       "      <th>GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder</th>\n",
       "      <th>GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder</th>\n",
       "      <th>XENLA_9.2_genome_cDNA.fa.transdecoder</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OG0000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC108700417 [provisional:trim25], LOC10869946...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OG0000001</td>\n",
       "      <td>Zfp389</td>\n",
       "      <td>zgc:174696, LOC108183688, LOC108183294, LOC100...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OG0000002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Xelaev18028992m, Xelaev18041830m, Xelaev180246...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OG0000003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Xelaev18000208m, LOC108709124, Xelaev18022432m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OG0000004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC108708057 [provisional:prkcq], LOC108716471...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32348</th>\n",
       "      <td>OG0032348</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>trim8.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32349</th>\n",
       "      <td>OG0032349</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LOC108700451 [provisional:kcnj6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32350</th>\n",
       "      <td>OG0032350</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gls.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32351</th>\n",
       "      <td>OG0032351</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>vstm5.L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32352</th>\n",
       "      <td>OG0032352</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>egln1.S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32353 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Orthogroup GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder  \\\n",
       "0      OG0000000                                                NaN         \n",
       "1      OG0000001                                             Zfp389         \n",
       "2      OG0000002                                                NaN         \n",
       "3      OG0000003                                                NaN         \n",
       "4      OG0000004                                                NaN         \n",
       "...          ...                                                ...         \n",
       "32348  OG0032348                                                NaN         \n",
       "32349  OG0032349                                                NaN         \n",
       "32350  OG0032350                                                NaN         \n",
       "32351  OG0032351                                                NaN         \n",
       "32352  OG0032352                                                NaN         \n",
       "\n",
       "      GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder  \\\n",
       "0                                                    NaN     \n",
       "1      zgc:174696, LOC108183688, LOC108183294, LOC100...     \n",
       "2                                                    NaN     \n",
       "3                                                    NaN     \n",
       "4                                                    NaN     \n",
       "...                                                  ...     \n",
       "32348                                                NaN     \n",
       "32349                                                NaN     \n",
       "32350                                                NaN     \n",
       "32351                                                NaN     \n",
       "32352                                                NaN     \n",
       "\n",
       "                   XENLA_9.2_genome_cDNA.fa.transdecoder  \n",
       "0      LOC108700417 [provisional:trim25], LOC10869946...  \n",
       "1                                                    NaN  \n",
       "2      Xelaev18028992m, Xelaev18041830m, Xelaev180246...  \n",
       "3      Xelaev18000208m, LOC108709124, Xelaev18022432m...  \n",
       "4      LOC108708057 [provisional:prkcq], LOC108716471...  \n",
       "...                                                  ...  \n",
       "32348                                            trim8.L  \n",
       "32349                   LOC108700451 [provisional:kcnj6]  \n",
       "32350                                              gls.L  \n",
       "32351                                            vstm5.L  \n",
       "32352                                            egln1.S  \n",
       "\n",
       "[32353 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "col_repacker = lambda x: ', '.join([i for i in x]) if len(x) > 0 else np.nan\n",
    "\n",
    "for col in sp_cols:\n",
    "    orthogroups_file[col] = orthogroups_file[col].apply(col_repacker)\n",
    "    \n",
    "display(orthogroups_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "2430b43a-9035-44f6-97bb-52690aaf6e71",
   "metadata": {},
   "outputs": [],
   "source": [
    "orthogroups_file.to_csv('DrerMmusXlae_Orthogroups_bygenename.tsv', sep = '\\t', index = None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
