{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3df6c5ba-21b7-4f9a-aedb-9db92d7882cf",
   "metadata": {},
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "aac310d7-4800-4b49-aa99-48d455aa03d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import os\n",
    "import dill\n",
    "import sys\n",
    "import datetime\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84395a54-f8ad-4342-bbd1-6e959c6470da",
   "metadata": {},
   "source": [
    "# 1. Collect BioFileDockets for each starting dataset\n",
    "Collects BioFileDockets for each of the species datasets in the `species_dict`.\n",
    "\n",
    "This `dict` expects key:value pairs in the form of `species`:`conditions` and is passed to the MultiSpeciesBioFileDocket class, which creates a folder in `output/` for the analysis.\n",
    "\n",
    "The folder name takes the following format:\n",
    "\n",
    "- `species_concat` + `_` + `global_conditions` + `analysis_type`\n",
    "\n",
    "Where `species_concat` is the string concatenation of the alphabetize list of species prefixes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e1ea15e7-b9de-4247-a0c8-c641879a58c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_FoldSeek/ already exists\n",
      "file GCF_000002035.5_GRCz10_genomic.gff already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.gff\n",
      "file GCF_000002035.5_GRCz10_genomic.fna already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.fna\n",
      "file GSM3768152_Brain_8_dge.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GSM3768152_Brain_8_dge.txt\n",
      "file Drer_adultbrain_ZFIN_UniProtIDs.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_ZFIN_UniProtIDs.txt\n",
      "file Drer_adultbrain_gtf-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_gtf-idmm.tsv\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep\n",
      "file Drer_adultbrain_og_idmm.tsv already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_og_idmm.tsv\n",
      "file OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv\n",
      "file GSM3768152_Brain_8_dge_asOrthogroup.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GSM3768152_Brain_8_dge_asOrthogroup.txt\n",
      "file DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv\n",
      "file GCF_000001635.23_GRCm38.p3_genomic.gff already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic.gff\n",
      "file GCF_000001635.23_GRCm38.p3_genomic.fna already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic.fna\n",
      "file GSM2906405_Brain1_dge_coerced.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GSM2906405_Brain1_dge_coerced.txt\n",
      "file Mmus_adultbrain_MGI_UniProtIDs.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_adultbrain_MGI_UniProtIDs.txt\n",
      "file Mmus_adultbrain_gtf-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_adultbrain_gtf-idmm.tsv\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.bed already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.bed\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.cds already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.cds\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3 already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3\n",
      "file GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep\n",
      "file Mmus_adultbrain_og_idmm.tsv already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/Mmus_adultbrain_og_idmm.tsv\n",
      "file OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv\n",
      "file GSM2906405_Brain1_dge_coerced_asOrthogroup.txt already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GSM2906405_Brain1_dge_coerced_asOrthogroup.txt\n",
      "file DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv\n",
      "file XENLA_9.2_Xenbase.gtf already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_Xenbase.gtf\n",
      "file XENLA_9.2_GCA.gff already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_GCA.gff\n",
      "file XENLA_9.2_genome.fa already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome.fa\n",
      "file GSM6214268_Xenopus_brain_COL65_dge.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge.txt\n",
      "file Xlae_adultbrain_Xenbase_UniProtIDs.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_Xenbase_UniProtIDs.txt\n",
      "file Xlae_adultbrain_uniprot-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_uniprot-idmm.tsv\n",
      "file Xlae_adultbrain_gtf-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_gtf-idmm.tsv\n",
      "file XENLA_9.2_genome_cDNA.fa already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.bed already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.bed\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.cds already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.cds\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.gff3\n",
      "file XENLA_9.2_genome_cDNA.fa.transdecoder.pep already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/XENLA_9.2_genome_cDNA.fa.transdecoder.pep\n",
      "file Xlae_adultbrain_og_idmm.tsv already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/Xlae_adultbrain_og_idmm.tsv\n",
      "file OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv\n",
      "file GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup.txt already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup.txt\n",
      "file DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv\n"
     ]
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "species_dict = {\n",
    "    'Danio_rerio': 'adultbrain',\n",
    "    'Mus_musculus': 'adultbrain',\n",
    "    'Xenopus_laevis': 'adultbrain'\n",
    "}\n",
    "\n",
    "global_conditions = 'adultbrain'\n",
    "\n",
    "analysis_type = 'FoldSeek'\n",
    "\n",
    "################\n",
    "\n",
    "sample_MSD = MultiSpeciesBioFileDocket(\n",
    "    species_dict = species_dict,\n",
    "    global_conditions = global_conditions,\n",
    "    analysis_type = analysis_type\n",
    "    )\n",
    "\n",
    "sample_MSD.get_BioFileDockets()\n",
    "sample_MSD.s3_to_local()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e3bf493-da0f-42bc-904d-fc4c6dd6ca2b",
   "metadata": {},
   "source": [
    "# 2. Get gene list from UniProt using Taxid\n",
    "Also, set the taxid for each species BioFileDocket."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "14a07a0a-e0fe-4a87-af94-7af04036216f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "overwriting taxid\n",
      "key \"taxid_genes\" already exists, ignoring\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Organism (ID)</th>\n",
       "      <th>Organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0D5W690</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0G2KQY6</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0G2KTI4</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0G2KYH9</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0G2L7I0</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52322</th>\n",
       "      <td>Z4YHP6</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52323</th>\n",
       "      <td>Z4YHQ6</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52324</th>\n",
       "      <td>Z4YHZ1</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52325</th>\n",
       "      <td>Z4YJ74</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52326</th>\n",
       "      <td>Z4YJA1</td>\n",
       "      <td>7955</td>\n",
       "      <td>Danio rerio (Zebrafish) (Brachydanio rerio)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>52327 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Entry  Organism (ID)                                     Organism\n",
       "0      A0A0D5W690           7955  Danio rerio (Zebrafish) (Brachydanio rerio)\n",
       "1      A0A0G2KQY6           7955  Danio rerio (Zebrafish) (Brachydanio rerio)\n",
       "2      A0A0G2KTI4           7955  Danio rerio (Zebrafish) (Brachydanio rerio)\n",
       "3      A0A0G2KYH9           7955  Danio rerio (Zebrafish) (Brachydanio rerio)\n",
       "4      A0A0G2L7I0           7955  Danio rerio (Zebrafish) (Brachydanio rerio)\n",
       "...           ...            ...                                          ...\n",
       "52322      Z4YHP6           7955  Danio rerio (Zebrafish) (Brachydanio rerio)\n",
       "52323      Z4YHQ6           7955  Danio rerio (Zebrafish) (Brachydanio rerio)\n",
       "52324      Z4YHZ1           7955  Danio rerio (Zebrafish) (Brachydanio rerio)\n",
       "52325      Z4YJ74           7955  Danio rerio (Zebrafish) (Brachydanio rerio)\n",
       "52326      Z4YJA1           7955  Danio rerio (Zebrafish) (Brachydanio rerio)\n",
       "\n",
       "[52327 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "overwriting taxid\n",
      "key \"taxid_genes\" already exists, ignoring\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Organism (ID)</th>\n",
       "      <th>Organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A075F5C6</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A087WPF7</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A087WPU4</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A087WRK1</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A087WRT4</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88529</th>\n",
       "      <td>Z4YN82</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88530</th>\n",
       "      <td>Z4YN86</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88531</th>\n",
       "      <td>Z4YN92</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88532</th>\n",
       "      <td>Z4YN97</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88533</th>\n",
       "      <td>Z4YNA9</td>\n",
       "      <td>10090</td>\n",
       "      <td>Mus musculus (Mouse)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88534 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Entry  Organism (ID)              Organism\n",
       "0      A0A075F5C6          10090  Mus musculus (Mouse)\n",
       "1      A0A087WPF7          10090  Mus musculus (Mouse)\n",
       "2      A0A087WPU4          10090  Mus musculus (Mouse)\n",
       "3      A0A087WRK1          10090  Mus musculus (Mouse)\n",
       "4      A0A087WRT4          10090  Mus musculus (Mouse)\n",
       "...           ...            ...                   ...\n",
       "88529      Z4YN82          10090  Mus musculus (Mouse)\n",
       "88530      Z4YN86          10090  Mus musculus (Mouse)\n",
       "88531      Z4YN92          10090  Mus musculus (Mouse)\n",
       "88532      Z4YN97          10090  Mus musculus (Mouse)\n",
       "88533      Z4YNA9          10090  Mus musculus (Mouse)\n",
       "\n",
       "[88534 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "overwriting taxid\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 5678k    0 5678k    0     0  43673      0 --:--:--  0:02:13 --:--:--  100k"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "key \"taxid_genes\" already exists, ignoring\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100 5724k    0 5724k    0     0  43743      0 --:--:--  0:02:14 --:--:-- 69191\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Organism (ID)</th>\n",
       "      <th>Organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A191ZDL1</td>\n",
       "      <td>8355</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A1L8ETL3</td>\n",
       "      <td>8355</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A1L8F1P8</td>\n",
       "      <td>8355</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A1L8F5J9</td>\n",
       "      <td>8355</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A1L8FDW4</td>\n",
       "      <td>8355</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111802</th>\n",
       "      <td>W8W3R8</td>\n",
       "      <td>8355</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111803</th>\n",
       "      <td>X2FP34</td>\n",
       "      <td>8355</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111804</th>\n",
       "      <td>X5FKY5</td>\n",
       "      <td>8355</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111805</th>\n",
       "      <td>X5FX98</td>\n",
       "      <td>8355</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111806</th>\n",
       "      <td>X5J6M9</td>\n",
       "      <td>8355</td>\n",
       "      <td>Xenopus laevis (African clawed frog)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>111807 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Entry  Organism (ID)                              Organism\n",
       "0       A0A191ZDL1           8355  Xenopus laevis (African clawed frog)\n",
       "1       A0A1L8ETL3           8355  Xenopus laevis (African clawed frog)\n",
       "2       A0A1L8F1P8           8355  Xenopus laevis (African clawed frog)\n",
       "3       A0A1L8F5J9           8355  Xenopus laevis (African clawed frog)\n",
       "4       A0A1L8FDW4           8355  Xenopus laevis (African clawed frog)\n",
       "...            ...            ...                                   ...\n",
       "111802      W8W3R8           8355  Xenopus laevis (African clawed frog)\n",
       "111803      X2FP34           8355  Xenopus laevis (African clawed frog)\n",
       "111804      X5FKY5           8355  Xenopus laevis (African clawed frog)\n",
       "111805      X5FX98           8355  Xenopus laevis (African clawed frog)\n",
       "111806      X5J6M9           8355  Xenopus laevis (African clawed frog)\n",
       "\n",
       "[111807 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "taxid_dict = {\n",
    "    'Drer': 7955,\n",
    "    'Mmus': 10090,\n",
    "    'Xlae': 8355\n",
    "}\n",
    "\n",
    "for pre in sample_MSD.species_BioFileDockets:\n",
    "    taxid = taxid_dict[pre]\n",
    "    sample_MSD.species_BioFileDockets[pre].set_taxid(taxid)\n",
    "    uniprot_genes = UniProtTaxidListFile(\n",
    "        filename = pre + '_taxid_genes.txt',\n",
    "        sampledict = sample_MSD.species_BioFileDockets[pre].sampledict,\n",
    "        taxid = taxid_dict[pre]\n",
    "    )\n",
    "    sample_MSD.species_BioFileDockets[pre].add_keyfile('taxid_genes', uniprot_genes)\n",
    "    \n",
    "    display(pd.read_csv(sample_MSD.species_BioFileDockets[pre].taxid_genes.path, sep = '\\t'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4963be4b-109a-416b-bc59-4459f8d3dd3b",
   "metadata": {},
   "source": [
    "# 3. Download files from all species into a single directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "03e128ac-c038-49ed-bf45-2c19b59630ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Download and untar for taxid 7955 has already been processed at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_FoldSeek/\n",
      "Download and untar for taxid 10090 has already been processed at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_FoldSeek/\n",
      "Download and untar for taxid 8355 has already been processed at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_FoldSeek/\n"
     ]
    }
   ],
   "source": [
    "output_folder = sample_MSD.directory\n",
    "alphafold_folder = output_folder + 'alphafold/'\n",
    "if not os.path.exists(alphafold_folder):\n",
    "    os.mkdir(alphafold_folder)\n",
    "\n",
    "for pre in sample_MSD.species_BioFileDockets:\n",
    "    taxid = sample_MSD.species_BioFileDockets[pre].metadata.taxid\n",
    "    target_files = 'gs://public-datasets-deepmind-alphafold/proteomes/proteome-tax_id-' + taxid + '-*'\n",
    "    \n",
    "    checkpoint_filepath = output_folder + taxid + '.checkpoint'\n",
    "    \n",
    "    if not os.path.exists(checkpoint_filepath):\n",
    "        subprocess.run(['gsutil', '-m', 'cp', target_files, output_folder])\n",
    "        print('Download and untar for taxid', taxid, 'completed;\\ncheckpoint file created at', checkpoint_filepath)\n",
    "        \n",
    "        filepaths = [output_folder + file for file in os.listdir(output_folder) if '.tar' in file]\n",
    "        for file in filepaths:\n",
    "            # Important to put all of the untar'd files into a separate folder\n",
    "            # Otherwise filesystem stutters b/c too many files in a single directory\n",
    "            subprocess.run(['tar', '-xf', file, '--directory', alphafold_folder])\n",
    "        \n",
    "        subprocess.run(['touch', checkpoint_filepath])\n",
    "    else:\n",
    "        print('Download and untar for taxid', taxid, 'has already been processed at', output_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7594805-097c-46b5-ba21-76b2d279ae0d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5b48b3dd-a027-451f-a3ec-de4d0308d804",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "final cluster tsv file already found at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_FoldSeek/clu_greedy.tsv\n"
     ]
    }
   ],
   "source": [
    "db_folder = output_folder + 'all_foldomesDB'\n",
    "foldseek_clustertsv = output_folder + 'clu_greedy.tsv'\n",
    "\n",
    "if not os.path.exists(foldseek_clustertsv):\n",
    "\n",
    "    subprocess.run(['foldseek', 'createdb', alphafold_folder, db_folder])\n",
    "\n",
    "    foldseek_out = output_folder + 'all_by_all'\n",
    "    foldseek_tmp = output_folder + 'tmp'\n",
    "    subprocess.run(['foldseek', 'search', db_folder, db_folder, foldseek_out, foldseek_tmp, '-a'])\n",
    "\n",
    "    foldseek_tmscore = output_folder + 'all_by_all_tmscore'\n",
    "    subprocess.run(['foldseek', 'aln2tmscore', db_folder, db_folder, foldseek_out, foldseek_tmscore])\n",
    "\n",
    "    foldseek_tsv = foldseek_tmscore + '.tsv'\n",
    "    subprocess.run(['foldseek', 'createtsv', db_folder, db_folder, foldseek_tmscore, foldseek_tsv])\n",
    "    \n",
    "    foldseek_cluster = output_folder + 'clu'\n",
    "    subprocess.run(['foldseek', 'clust', db_folder, foldseek_out, foldseek_cluster, '--cluster-mode', '0', '--similarity-type', '2'])\n",
    "    \n",
    "    subprocess.run(['foldseek', 'createtsv', db_folder, db_folder, foldseek_cluster, foldseek_clustertsv])\n",
    "\n",
    "else:\n",
    "    print('final cluster tsv file already found at', foldseek_clustertsv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0434465c-c7f4-443e-a10d-c4046f13dd19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>StruCluster</th>\n",
       "      <th>Drer</th>\n",
       "      <th>Mmus</th>\n",
       "      <th>Xlae</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SC0</td>\n",
       "      <td></td>\n",
       "      <td>A0A023NE65,A0A4P2VRD2,A0A023NE78,B3Y998</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SC1</td>\n",
       "      <td>Q4KMH7,E7FGK1,A0A1D5NSW6,E7EXY5,A0A2R8RMY7</td>\n",
       "      <td>D3Z4J9,A0A024A3D6,Q7TNG5,A0A0R4J1G7,A0A024A2Q0...</td>\n",
       "      <td>Q5U577,A0A1L8G1A2,A0A1L8GJX2,Q2TAF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SC2</td>\n",
       "      <td></td>\n",
       "      <td>A0A024CD92</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SC3</td>\n",
       "      <td></td>\n",
       "      <td>A0A024CDU6,A0A024CD39,A0A024CEZ3</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SC4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>A0A060CJT7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36646</th>\n",
       "      <td>SC36646</td>\n",
       "      <td></td>\n",
       "      <td>Z4YN28</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36647</th>\n",
       "      <td>SC36647</td>\n",
       "      <td></td>\n",
       "      <td>Z4YN31</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36648</th>\n",
       "      <td>SC36648</td>\n",
       "      <td>A5D8S0,A0A2R8PWB6</td>\n",
       "      <td>Q3UXL4,Z4YN37</td>\n",
       "      <td>A0A1L8G1N3,A1L2H3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36649</th>\n",
       "      <td>SC36649</td>\n",
       "      <td></td>\n",
       "      <td>Z4YN77</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36650</th>\n",
       "      <td>SC36650</td>\n",
       "      <td></td>\n",
       "      <td>Z4YN82</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36651 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      StruCluster                                        Drer  \\\n",
       "0             SC0                                               \n",
       "1             SC1  Q4KMH7,E7FGK1,A0A1D5NSW6,E7EXY5,A0A2R8RMY7   \n",
       "2             SC2                                               \n",
       "3             SC3                                               \n",
       "4             SC4                                               \n",
       "...           ...                                         ...   \n",
       "36646     SC36646                                               \n",
       "36647     SC36647                                               \n",
       "36648     SC36648                           A5D8S0,A0A2R8PWB6   \n",
       "36649     SC36649                                               \n",
       "36650     SC36650                                               \n",
       "\n",
       "                                                    Mmus  \\\n",
       "0                A0A023NE65,A0A4P2VRD2,A0A023NE78,B3Y998   \n",
       "1      D3Z4J9,A0A024A3D6,Q7TNG5,A0A0R4J1G7,A0A024A2Q0...   \n",
       "2                                             A0A024CD92   \n",
       "3                       A0A024CDU6,A0A024CD39,A0A024CEZ3   \n",
       "4                                                          \n",
       "...                                                  ...   \n",
       "36646                                             Z4YN28   \n",
       "36647                                             Z4YN31   \n",
       "36648                                      Q3UXL4,Z4YN37   \n",
       "36649                                             Z4YN77   \n",
       "36650                                             Z4YN82   \n",
       "\n",
       "                                      Xlae  \n",
       "0                                           \n",
       "1      Q5U577,A0A1L8G1A2,A0A1L8GJX2,Q2TAF3  \n",
       "2                                           \n",
       "3                                           \n",
       "4                               A0A060CJT7  \n",
       "...                                    ...  \n",
       "36646                                       \n",
       "36647                                       \n",
       "36648                    A0A1L8G1N3,A1L2H3  \n",
       "36649                                       \n",
       "36650                                       \n",
       "\n",
       "[36651 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df = pd.read_csv(foldseek_clustertsv, sep = '\\t', names = ['ClusterRep', 'uniprot_id'])\n",
    "df['ClusterRep'] = df['ClusterRep'].str.split('-', expand = True)[1]\n",
    "df['uniprot_id'] = df['uniprot_id'].str.split('-', expand = True)[1]\n",
    "\n",
    "df_merged = df.groupby('ClusterRep').agg({i: ('first' if i == 'ClusterRep' else lambda x: [i for i in x]) for i in df.columns}).reset_index(drop = True)\n",
    "df_merged.drop(columns = ['ClusterRep'], inplace = True)\n",
    "df_merged.insert(0, 'StruCluster', 'SC' + df_merged.index.astype('str'))\n",
    "\n",
    "for pre in taxid_dict:\n",
    "    species_genes = pd.read_csv(sample_MSD.species_BioFileDockets[pre].taxid_genes.path, sep = '\\t')\n",
    "    species_genelist = set(species_genes['Entry'])\n",
    "    df_merged[pre] = df_merged['uniprot_id'].apply(lambda x: ','.join(set(x).intersection(species_genelist)))\n",
    "\n",
    "struclusters_df = df_merged.drop(columns = ['uniprot_id'])\n",
    "struclusters_filename = '_'.join([sample_MSD.species_concat, sample_MSD.global_conditions, 'struclusters_file.tsv'])\n",
    "\n",
    "struclusters_file = FoldSeekOutputFile(\n",
    "    filename = struclusters_filename,\n",
    "    species_dict = sample_MSD.species_dict,\n",
    "    sampledict = sample_MSD.sampledict\n",
    ")\n",
    "\n",
    "struclusters_df.to_csv(struclusters_file.path, sep = '\\t', index = None)\n",
    "\n",
    "display(struclusters_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61be3acd-f098-44bb-bbf4-9c9bea6a1150",
   "metadata": {},
   "source": [
    "# 4. Generate Orthofinder_exc file from gxc file of each starting dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "11515d31-b501-4b47-90b2-aea14ca5db69",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using StruCluster embeddings as expected from FoldSeek\n",
      "StruCluster_excfile already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GSM3768152_Brain_8_dge_asStruCluster.txt skipping\n",
      "StruCluster_excfile already exists at /home/ec2-user/glial-origins/output/Mmus_adultbrain/GSM2906405_Brain1_dge_coerced_asStruCluster.txt skipping\n",
      "StruCluster_excfile already exists at /home/ec2-user/glial-origins/output/Xlae_adultbrain/GSM6214268_Xenopus_brain_COL65_dge_asStruCluster.txt skipping\n"
     ]
    }
   ],
   "source": [
    "gxc_to_exc(sample_MSD = sample_MSD,\n",
    "           embedding_df = struclusters_df, \n",
    "           exc_file = struclusters_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c135c9ee-8f04-4bf5-9916-7b3eeb0f42b0",
   "metadata": {},
   "source": [
    "# 5. Generate merged exc file for all samples in dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c30778eb-5480-4fda-84a2-462932b09b6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "collector_df = pd.DataFrame()\n",
    "\n",
    "export_filename = '_'.join([sample_MSD.species_concat, sample_MSD.global_conditions, 'asStruCluster', 'combined_exc.tsv'])\n",
    "\n",
    "jointexc = JointExcFile(\n",
    "    species_dict = sample_MSD.species_dict,\n",
    "    sampledict = sample_MSD.sampledict,\n",
    "    filename = export_filename,\n",
    "    embedding = 'StruCluster',\n",
    "    sources = [sample_MSD.species_BioFileDockets[pre].StruCluster_exc for pre in sample_MSD.species_BioFileDockets]\n",
    ")\n",
    "\n",
    "if not os.path.exists(jointexc.path):\n",
    "\n",
    "    # Iterates through all of the species in the Species BioFileDocket\n",
    "    for i, pre in enumerate(sample_MSD.species_BioFileDockets.keys()):\n",
    "    \n",
    "        df = pd.read_csv(sample_MSD.species_BioFileDockets[pre].StruCluster_exc.path, sep = '\\t')\n",
    "        df = df.add_prefix(pre + '_')\n",
    "        df.rename(columns = {pre + '_' + 'StruCluster': 'StruCluster'}, inplace = True)\n",
    "    \n",
    "        if i == 0:\n",
    "            collector_df = df\n",
    "        else:\n",
    "            collector_df = collector_df.merge(df, on = 'StruCluster', how = 'inner')\n",
    "\n",
    "    collector_df.to_csv(jointexc.path, sep = '\\t', index = None)\n",
    "\n",
    "# Iterates through all of the species in the Species BioFileDocket\n",
    "for pre in sample_MSD.species_BioFileDockets.keys():\n",
    "    sample_MSD.species_BioFileDockets[pre].add_keyfile('StruCluster_jointexc', jointexc)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "984a991a-f659-464f-aaf0-62637fb3bc88",
   "metadata": {},
   "source": [
    "# 6. Pickle the species_BioFileDocket files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f873483e-f652-4d15-bc4c-fe2ff9892456",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_000002035.5_GRCz10_genomic.gff already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM3768152_Brain_8_dge.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_ZFIN_UniProtIDs.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_gtf-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_og_idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM3768152_Brain_8_dge_asOrthogroup.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "upload: ../../output/Drer_adultbrain/Drer_taxid_genes.txt to s3://arcadia-reference-datasets/organisms/Danio_rerio/genomics_reference/mapping_file/Drer_taxid_genes.txt\n",
      "GSM3768152_Brain_8_dge_asStruCluster.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_BioFileDocket.pkl already exists in S3 bucket; overwriting.\n",
      "upload: ../../output/Drer_adultbrain/Drer_adultbrain_BioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/Drer_adultbrain_BioFileDocket.pkl\n",
      "GCF_000001635.23_GRCm38.p3_genomic.gff already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM2906405_Brain1_dge_coerced.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Mmus_adultbrain_MGI_UniProtIDs.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Mmus_adultbrain_gtf-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.bed already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.cds already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.gff3 already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000001635.23_GRCm38.p3_genomic_cDNA.fna.transdecoder.pep already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Mmus_adultbrain_og_idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM2906405_Brain1_dge_coerced_asOrthogroup.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "upload: ../../output/Mmus_adultbrain/Mmus_taxid_genes.txt to s3://arcadia-reference-datasets/organisms/Mus_musculus/genomics_reference/mapping_file/Mmus_taxid_genes.txt\n",
      "GSM2906405_Brain1_dge_coerced_asStruCluster.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Mmus_adultbrain_BioFileDocket.pkl already exists in S3 bucket; overwriting.\n",
      "upload: ../../output/Mmus_adultbrain/Mmus_adultbrain_BioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/Mmus_adultbrain_BioFileDocket.pkl\n",
      "XENLA_9.2_Xenbase.gtf already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_GCA.gff already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome.fa already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM6214268_Xenopus_brain_COL65_dge.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_Xenbase_UniProtIDs.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_uniprot-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_gtf-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.bed already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.cds already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.gff3 already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "XENLA_9.2_genome_cDNA.fa.transdecoder.pep already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_og_idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM6214268_Xenopus_brain_COL65_dge_asOrthogroup.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "upload: ../../output/Xlae_adultbrain/Xlae_taxid_genes.txt to s3://arcadia-reference-datasets/organisms/Xenopus_laevis/genomics_reference/mapping_file/Xlae_taxid_genes.txt\n",
      "GSM6214268_Xenopus_brain_COL65_dge_asStruCluster.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Xlae_adultbrain_BioFileDocket.pkl already exists in S3 bucket; overwriting.\n",
      "upload: ../../output/Xlae_adultbrain/Xlae_adultbrain_BioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/Xlae_adultbrain_BioFileDocket.pkl\n"
     ]
    }
   ],
   "source": [
    "for pre in sample_MSD.species_BioFileDockets.keys():\n",
    "    sample_MSD.species_BioFileDockets[pre].local_to_s3()\n",
    "    sample_MSD.species_BioFileDockets[pre].pickle()\n",
    "    sample_MSD.species_BioFileDockets[pre].push_to_s3(overwrite = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b1e678ce-1ec4-4a8d-9ebf-c77ed8d236b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upload: ../../output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_FoldSeek_MultiSpeciesBioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/DrerMmusXlae_adultbrain_FoldSeek_MultiSpeciesBioFileDocket.pkl\n"
     ]
    }
   ],
   "source": [
    "sample_MSD.pickle()\n",
    "sample_MSD.push_to_s3()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d38322d-c4a4-49b3-9f64-8d64594feec7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
