<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Scanpy Wrappers Usage - Glial Origins</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
        <link href="../assets/_mkdocstrings.css" rel="stylesheet" />
        <link href="../css/extra.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Scanpy Wrappers Usage";
        var mkdocs_page_input_path = "scanpy_wrappers_doc.md";
        var mkdocs_page_url = null;
      </script>
    
    <script src="../js/jquery-3.6.0.min.js" defer></script>
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
      <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> Glial Origins
        </a>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../BioFile_README/">About BioFile Handling</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../BioFile_tutorial/">BioFile Tutorial</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../biofile_doc/">BioFile Usage</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../string_functions_doc/">String Functions Usage</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="./">Scanpy Wrappers Usage</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#utils.scanpy_wrappers">utils.scanpy_wrappers</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#utils.scanpy_wrappers.ScanpyMetaObject">ScanpyMetaObject</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#utils.scanpy_wrappers.ScanpyMetaObject.cellgene_filter">cellgene_filter()</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.scanpy_wrappers.ScanpyMetaObject.export_top_genes">export_top_genes()</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.scanpy_wrappers.ScanpyMetaObject.get_top_genes">get_top_genes()</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.scanpy_wrappers.ScanpyMetaObject.map_cellannots">map_cellannots()</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.scanpy_wrappers.ScanpyMetaObject.map_cellannots_multispecies">map_cellannots_multispecies()</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.scanpy_wrappers.ScanpyMetaObject.map_gene_to_id">map_gene_to_id()</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.scanpy_wrappers.ScanpyMetaObject.normalize">normalize()</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.scanpy_wrappers.ScanpyMetaObject.pca_basic">pca_basic()</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.scanpy_wrappers.ScanpyMetaObject.rank_genes">rank_genes()</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.scanpy_wrappers.ScanpyMetaObject.read">read()</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.scanpy_wrappers.ScanpyMetaObject.regress_scale">regress_scale()</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.scanpy_wrappers.ScanpyMetaObject.umap_leiden">umap_leiden()</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.scanpy_wrappers.ScanpyMetaObject.variable_filter">variable_filter()</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#utils.scanpy_wrappers.ScanpyMetaObject.violin">violin()</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#utils.scanpy_wrappers.diagonalize_df">diagonalize_df()</a>
    </li>
    </ul>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Glial Origins</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" alt="Docs"></a> &raquo;</li>
      <li>Scanpy Wrappers Usage</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="string-functions-docmuentation">String Functions Docmuentation</h1>
<p>This page describes the functions in <code>scanpy_wrappers.py</code> and their usage.<br />
Functions here default to showing their source code for greater transparency and interoperability with base scanpy functions.</p>


<div class="doc doc-object doc-module">


<a id="utils.scanpy_wrappers"></a>
  <div class="doc doc-contents first">

  

  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="utils.scanpy_wrappers.ScanpyMetaObject" class="doc doc-heading">
        <code>ScanpyMetaObject</code>


</h2>


  <div class="doc doc-contents ">

  
      <p>Object that collects standardized collections of scanpy functions.</p>
<p>You can access the specific scanpy data in this object using <code>ScanpyMetaObject.adata</code>.<br />
Use this variable like you would use <code>adata</code> in any scanpy operation.  </p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>matrix</code></td>
          <td>
                <code>GxcFile | ExcFile</code>
          </td>
          <td><p>BioFile object of the matrix file.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>sampledict</code></td>
          <td>
                <code><span title="biofile_handling.SampleDict">SampleDict</span></code>
          </td>
          <td><p>a SampleDict object from the BioFileDocket.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>


        <details class="quote">
          <summary>Source code in <code>utils/scanpy_wrappers.py</code></summary>
          <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  8</span>
<span class="normal">  9</span>
<span class="normal"> 10</span>
<span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">ScanpyMetaObject</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;Object that collects standardized collections of scanpy functions.</span>

<span class="sd">    You can access the specific scanpy data in this object using `ScanpyMetaObject.adata`.  </span>
<span class="sd">    Use this variable like you would use `adata` in any scanpy operation.  </span>

<span class="sd">    Args:</span>
<span class="sd">        matrix (GxcFile | ExcFile): BioFile object of the matrix file.</span>
<span class="sd">        sampledict (SampleDict): a SampleDict object from the BioFileDocket.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">matrix</span><span class="p">,</span> <span class="n">sampledict</span><span class="p">:</span> <span class="n">SampleDict</span><span class="p">):</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">matrix</span> <span class="o">=</span> <span class="n">matrix</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sampledict</span> <span class="o">=</span> <span class="n">sampledict</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">matrixname</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">matrix</span><span class="o">.</span><span class="n">filename</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">species</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sampledict</span><span class="o">.</span><span class="n">species</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">directory</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sampledict</span><span class="o">.</span><span class="n">directory</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">conditions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sampledict</span><span class="o">.</span><span class="n">conditions</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">species_prefix</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">if</span> <span class="s1">&#39;_&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">species</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">species</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">prefixify</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">species</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">matrixpath</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">directory</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">matrixname</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">datatype</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">matrix</span><span class="p">))</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;File&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">resultsfile</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">matrixname</span> <span class="o">+</span> <span class="s1">&#39;_scresults.h5ad&#39;</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">resultspath</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">directory</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">resultsfile</span>

    <span class="k">def</span> <span class="nf">read</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">delimiter</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">cache</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">transpose</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">filter_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()):</span>
        <span class="sd">&quot;&quot;&quot;Reads the contained matrix into a scanpy.adata object, transposing if needed.</span>

<span class="sd">        Args:</span>
<span class="sd">            delimiter (str): a delimeter for the data, such as &#39;,&#39; or &#39;\t&#39;. Defaults to &#39;\t&#39;.</span>
<span class="sd">            cache (bool): whether or not to create a cache file for the matrix. Defaults to True.</span>
<span class="sd">            transpose (bool): whether or not to transpose the data matrix after loading. Defaults to True.</span>
<span class="sd">            filter_set (set): a set of gene ids to keep from the data, removing those that don&#39;t match.  </span>
<span class="sd">                Transposes matrix after filtering if transpose == True.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">matrixpath</span><span class="p">,</span> <span class="n">cache</span> <span class="o">=</span> <span class="n">cache</span><span class="p">,</span> <span class="n">delimiter</span> <span class="o">=</span> <span class="n">delimiter</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">filter_set</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">filter_set</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">transpose</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">violin</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span> <span class="o">=</span> <span class="s1">&#39;total_counts&#39;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">,</span> <span class="n">plot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Runs sc.pp.calculate_qc_metrics, sc.pl.violin, and sc.pl.scatter.</span>

<span class="sd">        Args:</span>
<span class="sd">            x (str): label of the feature to be plotted on the x axis of the violin &amp; scatter plot.  </span>
<span class="sd">                Defaults to &#39;total_counts&#39;.</span>
<span class="sd">            y (str): label of the feature to be plotted on the y axis of the violin &amp; scatter plot.</span>
<span class="sd">                Defaults to &#39;n_genes_by_counts&#39;.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">percent_top</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">log1p</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">plot</span><span class="p">:</span>
            <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="p">[</span><span class="n">y</span><span class="p">,</span> <span class="n">x</span><span class="p">],</span>
             <span class="n">jitter</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">cellgene_filter</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">20</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Filters data by minimum number of genes expressed per cell and minimum expressing cells per gene.</span>

<span class="sd">        Args:</span>
<span class="sd">            min_genes (int): minimum number of genes expressed per cell as a cutoff.</span>
<span class="sd">            min_cells (int): minimum number of cells expressed per gene as a cutoff.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="n">min_genes</span><span class="p">)</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="n">min_cells</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">normalize</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">max_n_genes_by_counts</span> <span class="o">=</span> <span class="mi">7000</span><span class="p">,</span> <span class="n">target_sum</span> <span class="o">=</span> <span class="mf">1e4</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Normalizes data using a max_n_genes_by_counts cutoff and a target sum.</span>

<span class="sd">        Modifies underlying `adata` using cutoff and normalization.</span>

<span class="sd">        Args:</span>
<span class="sd">            max_n_genes_by_counts (int): removes cells that exceed this number of counds.</span>
<span class="sd">            target_sum (float): target number of counts per cell, using scientific notation.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">n_genes_by_counts</span> <span class="o">&lt;</span> <span class="n">max_n_genes_by_counts</span><span class="p">,</span> <span class="p">:]</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">target_sum</span> <span class="o">=</span> <span class="n">target_sum</span><span class="p">)</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">variable_filter</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">min_mean</span> <span class="o">=</span> <span class="mf">0.0125</span><span class="p">,</span> <span class="n">max_mean</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">min_disp</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">max_disp</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">plot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Filters genes to use for dimensionality reduction using max/min of mean and max/min of dispersion, plotting optionally.</span>

<span class="sd">        Modifies underlying `adata` and saves original data under `adata.raw`.</span>

<span class="sd">        Args:</span>
<span class="sd">            min_mean (float | int): minimum mean expression of genes.</span>
<span class="sd">            max_mean (float | int): maximum mean expression of genes.</span>
<span class="sd">            min_disp (float | int): minimum dispersion of gene expression.</span>
<span class="sd">            max_disp (float | int): maximum dispersion of gene expression.</span>
<span class="sd">            plot (bool): whether or not to generate a scatter plot showing cutoffs.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_mean</span> <span class="o">=</span> <span class="n">min_mean</span><span class="p">,</span> <span class="n">max_mean</span><span class="o">=</span><span class="n">max_mean</span><span class="p">,</span> <span class="n">min_disp</span><span class="o">=</span><span class="n">min_disp</span><span class="p">,</span> <span class="n">max_disp</span><span class="o">=</span><span class="n">max_disp</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">plot</span><span class="p">:</span>
            <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">raw</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">[:,</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">highly_variable</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">regress_scale</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;total_counts&#39;</span><span class="p">],</span> <span class="n">max_value</span> <span class="o">=</span> <span class="mi">10</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Runs regression of specific features and scales data to a maximum value.</span>

<span class="sd">        Args:</span>
<span class="sd">            how (list): feature to regress out. Defaults to &#39;total_counts&#39;. </span>
<span class="sd">            max_value (float | int): maximum value to scale data.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">regress_out</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">how</span><span class="p">)</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">max_value</span> <span class="o">=</span> <span class="n">max_value</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">map_cellannots</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cellannot</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Adds an additional .obs feature for cell annotation.</span>

<span class="sd">        Imports from a file that has two columns: `cell_barcode` and `cell_type`.  </span>
<span class="sd">        The `cell_barcode` field should be an exact match for cells in the data.  </span>
<span class="sd">        Cells without a matching barcode are given the `celltype` label `&#39;Unlabeled&#39;`.</span>

<span class="sd">        Args:</span>
<span class="sd">            cellannot (CellAnnotFile): CellAnnotFile object, usually in the same BioFileDocket.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">cell_ids</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;cell_barcode&#39;</span><span class="p">:</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">})</span>
        <span class="n">cell_annots</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">cellannot</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">cell_ids</span> <span class="o">=</span> <span class="n">cell_ids</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">cell_annots</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="s1">&#39;cell_barcode&#39;</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>
        <span class="n">cell_ids</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;Unlabeled&#39;</span><span class="p">,</span> <span class="n">inplace</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cell_ids</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

    <span class="k">def</span> <span class="nf">map_cellannots_multispecies</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">msd</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Adds an additional .obs feature for cell annotation from a MultiSpeciesBioFileDocket.</span>

<span class="sd">        Imports from a file that has two columns: `cell_barcode` and `cell_type`.  </span>
<span class="sd">        The `cell_barcode` field should be an exact match for cells in the data.  </span>
<span class="sd">        Cells without a matching barcode are given the `celltype` label `&#39;Unlabeled&#39;`.</span>

<span class="sd">        Args:</span>
<span class="sd">            msd (MultiSpeciesBioFileDocket): the docket containing information about files from all species in the dataset.  </span>
<span class="sd">                The function looks for the BioFile object at the key `cellannot` for each species.  </span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">cell_ids</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;cell_barcode&#39;</span><span class="p">:</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">})</span>

        <span class="n">cell_annots</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">pre</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">msd</span><span class="o">.</span><span class="n">species_BioFileDockets</span><span class="p">):</span>
            <span class="n">dummy</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">msd</span><span class="o">.</span><span class="n">species_BioFileDockets</span><span class="p">[</span><span class="n">pre</span><span class="p">]</span><span class="o">.</span><span class="n">cellannot</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">dummy</span><span class="p">[</span><span class="s1">&#39;cell_barcode&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pre</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">dummy</span><span class="p">[</span><span class="s1">&#39;cell_barcode&#39;</span><span class="p">]</span>
            <span class="n">dummy</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pre</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">dummy</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span>

            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">cell_annots</span> <span class="o">=</span> <span class="n">dummy</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">cell_annots</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">cell_annots</span><span class="p">,</span> <span class="n">dummy</span><span class="p">])</span>

        <span class="n">cell_ids</span> <span class="o">=</span> <span class="n">cell_ids</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">cell_annots</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="s1">&#39;cell_barcode&#39;</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>
        <span class="n">cell_ids</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;Unlabeled&#39;</span><span class="p">,</span> <span class="n">inplace</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cell_ids</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>

    <span class="k">def</span> <span class="nf">pca_basic</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s1">&#39;arpack&#39;</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">plot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Runs a simple PCA, displaying the first two components and variance plot.</span>

<span class="sd">        Args:</span>
<span class="sd">            svd_solver (str): SVD sovling function, passed to `sc.tl.pca()`.</span>
<span class="sd">            color (list): list of coloring schemes for points in the data; creates one plot per color scheme.</span>
<span class="sd">            plot (bool): whether or not to make a plot.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="n">svd_solver</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">plot</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="n">color</span> <span class="o">!=</span> <span class="p">[]:</span>
            <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">color</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_prefix</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">datatype</span> <span class="o">+</span> <span class="s1">&#39;_pca.pdf&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_prefix</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">datatype</span> <span class="o">+</span> <span class="s1">&#39;_pca.pdf&#39;</span><span class="p">)</span>
            <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca_variance_ratio</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">umap_leiden</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">n_pcs</span> <span class="o">=</span> <span class="mi">40</span><span class="p">,</span> <span class="n">legend_loc</span><span class="o">=</span><span class="s1">&#39;on data&#39;</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">plot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Runs Leiden clustering, followed by UMAP, on the data.</span>

<span class="sd">        Args:</span>
<span class="sd">            n_neighbors (int): number of neighbors to pass to `sc.pp.neighbors`.</span>
<span class="sd">            n_pcs (int): number of Principal Components to use from PCA, passed to `sc.pp.neighbors`.</span>
<span class="sd">            legend_loc (str): where the legend should go (e.g. `&#39;on data&#39;`.)</span>
<span class="sd">            save (bool): whether to save the plot.</span>
<span class="sd">            plot (bool): whether to make the plot.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="n">n_neighbors</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="n">n_pcs</span><span class="p">)</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">)</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">plot</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="n">save</span><span class="p">:</span>
            <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">],</span> <span class="n">legend_loc</span><span class="o">=</span><span class="n">legend_loc</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">species_prefix</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">datatype</span><span class="p">,</span> <span class="s1">&#39;leiden.pdf&#39;</span><span class="p">]))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">],</span> <span class="n">legend_loc</span><span class="o">=</span><span class="n">legend_loc</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">rank_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;t-test&#39;</span><span class="p">,</span> <span class="n">plot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">n_genes</span> <span class="o">=</span> <span class="mi">25</span><span class="p">,</span> <span class="n">sharey</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Runs `sc.tl.rank_genes_groups` based on passed parameters.</span>

<span class="sd">        Parameters can be changed to alter the ranking scheme.</span>

<span class="sd">        Args:</span>
<span class="sd">            on (str): clustering feature to use. Defaults to &#39;leiden&#39;.</span>
<span class="sd">            method (str): how to compare groups for gene ranking. Defaults to &#39;t-test&#39;.</span>
<span class="sd">            plot (bool): whether or not to plot the data.</span>
<span class="sd">            n_genes (int): number of genes to plot.</span>
<span class="sd">            sharey (bool): pass to `sc.pl.rank_genes_groups`.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">on</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">)</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="n">n_genes</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="n">sharey</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">resultsfile</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">get_top_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">datatype</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">top_number</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span> <span class="n">tofile</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get the top_number genes per cluster across all clusters without repetition.  </span>

<span class="sd">        Also get the top marker gene per cluster.  </span>
<span class="sd">        Save these values as object attributes with a special key to the parent object.</span>

<span class="sd">        Args:</span>
<span class="sd">            datatype (str): descriptor for the data type, to be included in the outfile name.  </span>
<span class="sd">                Note: This is not checked for corretness!</span>
<span class="sd">            top_number (int): number of top genes to pull out from each cluster.</span>
<span class="sd">            tofile (bool): whether to save a file. (Not implemented yet.)</span>

<span class="sd">        Returns:</span>
<span class="sd">            (tuple[str, str]): keys of the top gene list files.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">biofile_handling</span> <span class="kn">import</span> <span class="n">GeneListFile</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">marker_genes_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;rank_genes_groups&#39;</span><span class="p">][</span><span class="s1">&#39;names&#39;</span><span class="p">])</span>
        <span class="n">top_genelist</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">marker_genes_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_number</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()])))</span>
        <span class="n">top_marker_genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">marker_genes_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>

        <span class="n">genelisttype</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">top_number</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;DE&#39;</span><span class="p">,</span> <span class="n">datatype</span><span class="p">,</span> <span class="s1">&#39;list&#39;</span><span class="p">])</span>
        <span class="n">markergenetype</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="s1">&#39;marker&#39;</span><span class="p">,</span> <span class="n">datatype</span><span class="p">,</span> <span class="s1">&#39;list&#39;</span><span class="p">])</span>
        <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genelisttype</span><span class="p">,</span> <span class="n">top_genelist</span><span class="p">)</span>
        <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">markergenetype</span><span class="p">,</span> <span class="n">top_marker_genes</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">genelisttype</span><span class="p">,</span> <span class="n">markergenetype</span>

    <span class="k">def</span> <span class="nf">export_top_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Exports the top genes list to a text file, assigning the object as an attribute of the parent object.</span>

<span class="sd">        Args:</span>
<span class="sd">            key (str): a key for the gene list, ending in `&#39;_list&#39;`.</span>

<span class="sd">        Raises:</span>
<span class="sd">            Exception: when the key does not end in `&#39;_list&#39;`.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="s1">&#39;_list&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">key</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s1">&#39;key must be an attribute ending in &quot;_list&quot;&#39;</span><span class="p">)</span>

        <span class="n">top_type</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">datatype</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span>

        <span class="n">top_genelist_filename</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">species_prefix</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">datatype</span><span class="p">,</span> <span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">top_type</span><span class="p">,</span> <span class="n">datatype</span><span class="p">,</span> <span class="s1">&#39;ids.txt&#39;</span><span class="p">])</span>
        <span class="n">top_genelist_file</span> <span class="o">=</span> <span class="n">GeneListFile</span><span class="p">(</span><span class="n">filename</span> <span class="o">=</span> <span class="n">top_genelist_filename</span><span class="p">,</span> 
                                         <span class="n">sampledict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sampledict</span><span class="p">,</span> 
                                         <span class="n">sources</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">matrix</span><span class="p">,</span> 
                                         <span class="n">genes</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">),</span> 
                                         <span class="n">identifier</span> <span class="o">=</span> <span class="n">datatype</span><span class="p">)</span>
        <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span> <span class="o">+</span> <span class="s1">&#39;_file&#39;</span><span class="p">,</span> <span class="n">top_genelist_file</span><span class="p">)</span>

    <span class="c1"># Using an IdmmFile object, convert IDs from one type to another</span>
    <span class="c1"># Usually you&#39;ll be converting from gene_name to the embedding (e.g. Orthogroup)</span>
    <span class="c1"># Prints a table of the mappings</span>
    <span class="c1"># Returns a list of ids for the ones actually present in the scanpy object</span>
    <span class="k">def</span> <span class="nf">map_gene_to_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">idmm</span><span class="p">,</span> <span class="n">gene_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">from_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">to_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">check_ids</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Maps IDs from one type into another using an IdmmFile object.</span>

<span class="sd">        Args:</span>
<span class="sd">            idmm (IdmmFile): the IdmmFile object you&#39;re converting IDs between.</span>
<span class="sd">            gene_list (list of str): list of genes in the dataset you want to convert.</span>
<span class="sd">            from_id (str): starting feature column in the idmm.</span>
<span class="sd">            to_id (str): ending feature in the idmm.</span>
<span class="sd">            check_ids (bool): whether to check if the ids in gene_list are present in the parent object.  </span>
<span class="sd">                If true, only returns mappings that are actually found in the parent data.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
        <span class="n">idmm</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">idmm</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

        <span class="n">id_table</span> <span class="o">=</span> <span class="n">idmm</span><span class="p">[</span><span class="n">idmm</span><span class="p">[</span><span class="n">from_id</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)][[</span><span class="n">from_id</span><span class="p">,</span> <span class="n">to_id</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
        <span class="n">display</span><span class="p">(</span><span class="n">id_table</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">check_ids</span><span class="p">:</span>
            <span class="n">ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">id_table</span><span class="p">[</span><span class="n">to_id</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">id_table</span><span class="p">[</span><span class="n">to_id</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">ids</span>
</code></pre></div></td></tr></table></div>
        </details>

  

  <div class="doc doc-children">









<div class="doc doc-object doc-function">



<h3 id="utils.scanpy_wrappers.ScanpyMetaObject.cellgene_filter" class="doc doc-heading">
<code class="highlight language-python"><span class="n">cellgene_filter</span><span class="p">(</span><span class="n">min_genes</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Filters data by minimum number of genes expressed per cell and minimum expressing cells per gene.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>min_genes</code></td>
          <td>
                <code>int</code>
          </td>
          <td><p>minimum number of genes expressed per cell as a cutoff.</p></td>
          <td>
                <code>100</code>
          </td>
        </tr>
        <tr>
          <td><code>min_cells</code></td>
          <td>
                <code>int</code>
          </td>
          <td><p>minimum number of cells expressed per gene as a cutoff.</p></td>
          <td>
                <code>20</code>
          </td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>utils/scanpy_wrappers.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">cellgene_filter</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">20</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Filters data by minimum number of genes expressed per cell and minimum expressing cells per gene.</span>

<span class="sd">    Args:</span>
<span class="sd">        min_genes (int): minimum number of genes expressed per cell as a cutoff.</span>
<span class="sd">        min_cells (int): minimum number of cells expressed per gene as a cutoff.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_cells</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="n">min_genes</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="n">min_cells</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="utils.scanpy_wrappers.ScanpyMetaObject.export_top_genes" class="doc doc-heading">
<code class="highlight language-python"><span class="n">export_top_genes</span><span class="p">(</span><span class="n">key</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Exports the top genes list to a text file, assigning the object as an attribute of the parent object.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>key</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>a key for the gene list, ending in <code>'_list'</code>.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Raises:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code>Exception</code>
          </td>
          <td><p>when the key does not end in <code>'_list'</code>.</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>utils/scanpy_wrappers.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">export_top_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Exports the top genes list to a text file, assigning the object as an attribute of the parent object.</span>

<span class="sd">    Args:</span>
<span class="sd">        key (str): a key for the gene list, ending in `&#39;_list&#39;`.</span>

<span class="sd">    Raises:</span>
<span class="sd">        Exception: when the key does not end in `&#39;_list&#39;`.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s1">&#39;_list&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">key</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s1">&#39;key must be an attribute ending in &quot;_list&quot;&#39;</span><span class="p">)</span>

    <span class="n">top_type</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">datatype</span> <span class="o">=</span> <span class="n">key</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span>

    <span class="n">top_genelist_filename</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">species_prefix</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">datatype</span><span class="p">,</span> <span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="n">top_type</span><span class="p">,</span> <span class="n">datatype</span><span class="p">,</span> <span class="s1">&#39;ids.txt&#39;</span><span class="p">])</span>
    <span class="n">top_genelist_file</span> <span class="o">=</span> <span class="n">GeneListFile</span><span class="p">(</span><span class="n">filename</span> <span class="o">=</span> <span class="n">top_genelist_filename</span><span class="p">,</span> 
                                     <span class="n">sampledict</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sampledict</span><span class="p">,</span> 
                                     <span class="n">sources</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">matrix</span><span class="p">,</span> 
                                     <span class="n">genes</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">),</span> 
                                     <span class="n">identifier</span> <span class="o">=</span> <span class="n">datatype</span><span class="p">)</span>
    <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span> <span class="o">+</span> <span class="s1">&#39;_file&#39;</span><span class="p">,</span> <span class="n">top_genelist_file</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="utils.scanpy_wrappers.ScanpyMetaObject.get_top_genes" class="doc doc-heading">
<code class="highlight language-python"><span class="n">get_top_genes</span><span class="p">(</span><span class="n">datatype</span><span class="p">,</span> <span class="n">top_number</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">tofile</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Get the top_number genes per cluster across all clusters without repetition.  </p>
<p>Also get the top marker gene per cluster.<br />
Save these values as object attributes with a special key to the parent object.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>datatype</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>descriptor for the data type, to be included in the outfile name.<br />
Note: This is not checked for corretness!</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>top_number</code></td>
          <td>
                <code>int</code>
          </td>
          <td><p>number of top genes to pull out from each cluster.</p></td>
          <td>
                <code>200</code>
          </td>
        </tr>
        <tr>
          <td><code>tofile</code></td>
          <td>
                <code>bool</code>
          </td>
          <td><p>whether to save a file. (Not implemented yet.)</p></td>
          <td>
                <code>True</code>
          </td>
        </tr>
    </tbody>
  </table>

  <p><strong>Returns:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Type</th>
        <th>Description</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td>
                <code>tuple[str, str]</code>
          </td>
          <td><p>keys of the top gene list files.</p></td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>utils/scanpy_wrappers.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">get_top_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">datatype</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">top_number</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span> <span class="n">tofile</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Get the top_number genes per cluster across all clusters without repetition.  </span>

<span class="sd">    Also get the top marker gene per cluster.  </span>
<span class="sd">    Save these values as object attributes with a special key to the parent object.</span>

<span class="sd">    Args:</span>
<span class="sd">        datatype (str): descriptor for the data type, to be included in the outfile name.  </span>
<span class="sd">            Note: This is not checked for corretness!</span>
<span class="sd">        top_number (int): number of top genes to pull out from each cluster.</span>
<span class="sd">        tofile (bool): whether to save a file. (Not implemented yet.)</span>

<span class="sd">    Returns:</span>
<span class="sd">        (tuple[str, str]): keys of the top gene list files.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span> <span class="nn">biofile_handling</span> <span class="kn">import</span> <span class="n">GeneListFile</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">marker_genes_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;rank_genes_groups&#39;</span><span class="p">][</span><span class="s1">&#39;names&#39;</span><span class="p">])</span>
    <span class="n">top_genelist</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">chain</span><span class="o">.</span><span class="n">from_iterable</span><span class="p">([</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">marker_genes_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">top_number</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()])))</span>
    <span class="n">top_marker_genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">marker_genes_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>

    <span class="n">genelisttype</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">top_number</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;DE&#39;</span><span class="p">,</span> <span class="n">datatype</span><span class="p">,</span> <span class="s1">&#39;list&#39;</span><span class="p">])</span>
    <span class="n">markergenetype</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="s1">&#39;top&#39;</span><span class="p">,</span> <span class="s1">&#39;marker&#39;</span><span class="p">,</span> <span class="n">datatype</span><span class="p">,</span> <span class="s1">&#39;list&#39;</span><span class="p">])</span>
    <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genelisttype</span><span class="p">,</span> <span class="n">top_genelist</span><span class="p">)</span>
    <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">markergenetype</span><span class="p">,</span> <span class="n">top_marker_genes</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">genelisttype</span><span class="p">,</span> <span class="n">markergenetype</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="utils.scanpy_wrappers.ScanpyMetaObject.map_cellannots" class="doc doc-heading">
<code class="highlight language-python"><span class="n">map_cellannots</span><span class="p">(</span><span class="n">cellannot</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Adds an additional .obs feature for cell annotation.</p>
<p>Imports from a file that has two columns: <code>cell_barcode</code> and <code>cell_type</code>.<br />
The <code>cell_barcode</code> field should be an exact match for cells in the data.<br />
Cells without a matching barcode are given the <code>celltype</code> label <code>'Unlabeled'</code>.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>cellannot</code></td>
          <td>
                <code><span title="biofile_handling.CellAnnotFile">CellAnnotFile</span></code>
          </td>
          <td><p>CellAnnotFile object, usually in the same BioFileDocket.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>utils/scanpy_wrappers.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">map_cellannots</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cellannot</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Adds an additional .obs feature for cell annotation.</span>

<span class="sd">    Imports from a file that has two columns: `cell_barcode` and `cell_type`.  </span>
<span class="sd">    The `cell_barcode` field should be an exact match for cells in the data.  </span>
<span class="sd">    Cells without a matching barcode are given the `celltype` label `&#39;Unlabeled&#39;`.</span>

<span class="sd">    Args:</span>
<span class="sd">        cellannot (CellAnnotFile): CellAnnotFile object, usually in the same BioFileDocket.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">cell_ids</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;cell_barcode&#39;</span><span class="p">:</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">})</span>
    <span class="n">cell_annots</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">cellannot</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">cell_ids</span> <span class="o">=</span> <span class="n">cell_ids</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">cell_annots</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="s1">&#39;cell_barcode&#39;</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>
    <span class="n">cell_ids</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;Unlabeled&#39;</span><span class="p">,</span> <span class="n">inplace</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cell_ids</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="utils.scanpy_wrappers.ScanpyMetaObject.map_cellannots_multispecies" class="doc doc-heading">
<code class="highlight language-python"><span class="n">map_cellannots_multispecies</span><span class="p">(</span><span class="n">msd</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Adds an additional .obs feature for cell annotation from a MultiSpeciesBioFileDocket.</p>
<p>Imports from a file that has two columns: <code>cell_barcode</code> and <code>cell_type</code>.<br />
The <code>cell_barcode</code> field should be an exact match for cells in the data.<br />
Cells without a matching barcode are given the <code>celltype</code> label <code>'Unlabeled'</code>.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>msd</code></td>
          <td>
                <code>MultiSpeciesBioFileDocket</code>
          </td>
          <td><p>the docket containing information about files from all species in the dataset.<br />
The function looks for the BioFile object at the key <code>cellannot</code> for each species.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>utils/scanpy_wrappers.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">map_cellannots_multispecies</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">msd</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Adds an additional .obs feature for cell annotation from a MultiSpeciesBioFileDocket.</span>

<span class="sd">    Imports from a file that has two columns: `cell_barcode` and `cell_type`.  </span>
<span class="sd">    The `cell_barcode` field should be an exact match for cells in the data.  </span>
<span class="sd">    Cells without a matching barcode are given the `celltype` label `&#39;Unlabeled&#39;`.</span>

<span class="sd">    Args:</span>
<span class="sd">        msd (MultiSpeciesBioFileDocket): the docket containing information about files from all species in the dataset.  </span>
<span class="sd">            The function looks for the BioFile object at the key `cellannot` for each species.  </span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">cell_ids</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">&#39;cell_barcode&#39;</span><span class="p">:</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">})</span>

    <span class="n">cell_annots</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">pre</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">msd</span><span class="o">.</span><span class="n">species_BioFileDockets</span><span class="p">):</span>
        <span class="n">dummy</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">msd</span><span class="o">.</span><span class="n">species_BioFileDockets</span><span class="p">[</span><span class="n">pre</span><span class="p">]</span><span class="o">.</span><span class="n">cellannot</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="n">dummy</span><span class="p">[</span><span class="s1">&#39;cell_barcode&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pre</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">dummy</span><span class="p">[</span><span class="s1">&#39;cell_barcode&#39;</span><span class="p">]</span>
        <span class="n">dummy</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pre</span> <span class="o">+</span> <span class="s1">&#39;_&#39;</span> <span class="o">+</span> <span class="n">dummy</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">cell_annots</span> <span class="o">=</span> <span class="n">dummy</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">cell_annots</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">cell_annots</span><span class="p">,</span> <span class="n">dummy</span><span class="p">])</span>

    <span class="n">cell_ids</span> <span class="o">=</span> <span class="n">cell_ids</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">cell_annots</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="s1">&#39;cell_barcode&#39;</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;left&#39;</span><span class="p">)</span>
    <span class="n">cell_ids</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="s1">&#39;Unlabeled&#39;</span><span class="p">,</span> <span class="n">inplace</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cell_ids</span><span class="p">[</span><span class="s1">&#39;celltype&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="utils.scanpy_wrappers.ScanpyMetaObject.map_gene_to_id" class="doc doc-heading">
<code class="highlight language-python"><span class="n">map_gene_to_id</span><span class="p">(</span><span class="n">idmm</span><span class="p">,</span> <span class="n">gene_list</span><span class="p">,</span> <span class="n">from_id</span><span class="p">,</span> <span class="n">to_id</span><span class="p">,</span> <span class="n">check_ids</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Maps IDs from one type into another using an IdmmFile object.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>idmm</code></td>
          <td>
                <code>IdmmFile</code>
          </td>
          <td><p>the IdmmFile object you're converting IDs between.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>gene_list</code></td>
          <td>
                <code>list of str</code>
          </td>
          <td><p>list of genes in the dataset you want to convert.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>from_id</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>starting feature column in the idmm.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>to_id</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>ending feature in the idmm.</p></td>
          <td>
              <em>required</em>
          </td>
        </tr>
        <tr>
          <td><code>check_ids</code></td>
          <td>
                <code>bool</code>
          </td>
          <td><p>whether to check if the ids in gene_list are present in the parent object.<br />
If true, only returns mappings that are actually found in the parent data.</p></td>
          <td>
                <code>True</code>
          </td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>utils/scanpy_wrappers.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">map_gene_to_id</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">idmm</span><span class="p">,</span> <span class="n">gene_list</span><span class="p">:</span> <span class="nb">list</span><span class="p">,</span> <span class="n">from_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">to_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">check_ids</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Maps IDs from one type into another using an IdmmFile object.</span>

<span class="sd">    Args:</span>
<span class="sd">        idmm (IdmmFile): the IdmmFile object you&#39;re converting IDs between.</span>
<span class="sd">        gene_list (list of str): list of genes in the dataset you want to convert.</span>
<span class="sd">        from_id (str): starting feature column in the idmm.</span>
<span class="sd">        to_id (str): ending feature in the idmm.</span>
<span class="sd">        check_ids (bool): whether to check if the ids in gene_list are present in the parent object.  </span>
<span class="sd">            If true, only returns mappings that are actually found in the parent data.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
    <span class="n">idmm</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">idmm</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="n">id_table</span> <span class="o">=</span> <span class="n">idmm</span><span class="p">[</span><span class="n">idmm</span><span class="p">[</span><span class="n">from_id</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)][[</span><span class="n">from_id</span><span class="p">,</span> <span class="n">to_id</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="n">display</span><span class="p">(</span><span class="n">id_table</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">check_ids</span><span class="p">:</span>
        <span class="n">ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">id_table</span><span class="p">[</span><span class="n">to_id</span><span class="p">]</span><span class="o">.</span><span class="n">values</span> <span class="k">if</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">id_table</span><span class="p">[</span><span class="n">to_id</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">ids</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="utils.scanpy_wrappers.ScanpyMetaObject.normalize" class="doc doc-heading">
<code class="highlight language-python"><span class="n">normalize</span><span class="p">(</span><span class="n">max_n_genes_by_counts</span><span class="o">=</span><span class="mi">7000</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">10000.0</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Normalizes data using a max_n_genes_by_counts cutoff and a target sum.</p>
<p>Modifies underlying <code>adata</code> using cutoff and normalization.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>max_n_genes_by_counts</code></td>
          <td>
                <code>int</code>
          </td>
          <td><p>removes cells that exceed this number of counds.</p></td>
          <td>
                <code>7000</code>
          </td>
        </tr>
        <tr>
          <td><code>target_sum</code></td>
          <td>
                <code>float</code>
          </td>
          <td><p>target number of counts per cell, using scientific notation.</p></td>
          <td>
                <code>10000.0</code>
          </td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>utils/scanpy_wrappers.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">normalize</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">max_n_genes_by_counts</span> <span class="o">=</span> <span class="mi">7000</span><span class="p">,</span> <span class="n">target_sum</span> <span class="o">=</span> <span class="mf">1e4</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Normalizes data using a max_n_genes_by_counts cutoff and a target sum.</span>

<span class="sd">    Modifies underlying `adata` using cutoff and normalization.</span>

<span class="sd">    Args:</span>
<span class="sd">        max_n_genes_by_counts (int): removes cells that exceed this number of counds.</span>
<span class="sd">        target_sum (float): target number of counts per cell, using scientific notation.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">n_genes_by_counts</span> <span class="o">&lt;</span> <span class="n">max_n_genes_by_counts</span><span class="p">,</span> <span class="p">:]</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">target_sum</span> <span class="o">=</span> <span class="n">target_sum</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="utils.scanpy_wrappers.ScanpyMetaObject.pca_basic" class="doc doc-heading">
<code class="highlight language-python"><span class="n">pca_basic</span><span class="p">(</span><span class="n">svd_solver</span><span class="o">=</span><span class="s1">&#39;arpack&#39;</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[],</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Runs a simple PCA, displaying the first two components and variance plot.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>svd_solver</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>SVD sovling function, passed to <code>sc.tl.pca()</code>.</p></td>
          <td>
                <code>&#39;arpack&#39;</code>
          </td>
        </tr>
        <tr>
          <td><code>color</code></td>
          <td>
                <code>list</code>
          </td>
          <td><p>list of coloring schemes for points in the data; creates one plot per color scheme.</p></td>
          <td>
                <code>[]</code>
          </td>
        </tr>
        <tr>
          <td><code>plot</code></td>
          <td>
                <code>bool</code>
          </td>
          <td><p>whether or not to make a plot.</p></td>
          <td>
                <code>True</code>
          </td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>utils/scanpy_wrappers.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">pca_basic</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="s1">&#39;arpack&#39;</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">plot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Runs a simple PCA, displaying the first two components and variance plot.</span>

<span class="sd">    Args:</span>
<span class="sd">        svd_solver (str): SVD sovling function, passed to `sc.tl.pca()`.</span>
<span class="sd">        color (list): list of coloring schemes for points in the data; creates one plot per color scheme.</span>
<span class="sd">        plot (bool): whether or not to make a plot.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">svd_solver</span><span class="o">=</span><span class="n">svd_solver</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">plot</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="n">color</span> <span class="o">!=</span> <span class="p">[]:</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="n">color</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_prefix</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">datatype</span> <span class="o">+</span> <span class="s1">&#39;_pca.pdf&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">species_prefix</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">datatype</span> <span class="o">+</span> <span class="s1">&#39;_pca.pdf&#39;</span><span class="p">)</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">pca_variance_ratio</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="utils.scanpy_wrappers.ScanpyMetaObject.rank_genes" class="doc doc-heading">
<code class="highlight language-python"><span class="n">rank_genes</span><span class="p">(</span><span class="n">on</span><span class="o">=</span><span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s1">&#39;t-test&#39;</span><span class="p">,</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Runs <code>sc.tl.rank_genes_groups</code> based on passed parameters.</p>
<p>Parameters can be changed to alter the ranking scheme.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>on</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>clustering feature to use. Defaults to 'leiden'.</p></td>
          <td>
                <code>&#39;leiden&#39;</code>
          </td>
        </tr>
        <tr>
          <td><code>method</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>how to compare groups for gene ranking. Defaults to 't-test'.</p></td>
          <td>
                <code>&#39;t-test&#39;</code>
          </td>
        </tr>
        <tr>
          <td><code>plot</code></td>
          <td>
                <code>bool</code>
          </td>
          <td><p>whether or not to plot the data.</p></td>
          <td>
                <code>True</code>
          </td>
        </tr>
        <tr>
          <td><code>n_genes</code></td>
          <td>
                <code>int</code>
          </td>
          <td><p>number of genes to plot.</p></td>
          <td>
                <code>25</code>
          </td>
        </tr>
        <tr>
          <td><code>sharey</code></td>
          <td>
                <code>bool</code>
          </td>
          <td><p>pass to <code>sc.pl.rank_genes_groups</code>.</p></td>
          <td>
                <code>False</code>
          </td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>utils/scanpy_wrappers.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">rank_genes</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">on</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;t-test&#39;</span><span class="p">,</span> <span class="n">plot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">n_genes</span> <span class="o">=</span> <span class="mi">25</span><span class="p">,</span> <span class="n">sharey</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Runs `sc.tl.rank_genes_groups` based on passed parameters.</span>

<span class="sd">    Parameters can be changed to alter the ranking scheme.</span>

<span class="sd">    Args:</span>
<span class="sd">        on (str): clustering feature to use. Defaults to &#39;leiden&#39;.</span>
<span class="sd">        method (str): how to compare groups for gene ranking. Defaults to &#39;t-test&#39;.</span>
<span class="sd">        plot (bool): whether or not to plot the data.</span>
<span class="sd">        n_genes (int): number of genes to plot.</span>
<span class="sd">        sharey (bool): pass to `sc.pl.rank_genes_groups`.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">on</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="n">method</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_genes</span><span class="o">=</span><span class="n">n_genes</span><span class="p">,</span> <span class="n">sharey</span><span class="o">=</span><span class="n">sharey</span><span class="p">)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">resultsfile</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="utils.scanpy_wrappers.ScanpyMetaObject.read" class="doc doc-heading">
<code class="highlight language-python"><span class="n">read</span><span class="p">(</span><span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">cache</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">transpose</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">filter_set</span><span class="o">=</span><span class="nb">set</span><span class="p">())</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Reads the contained matrix into a scanpy.adata object, transposing if needed.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>delimiter</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>a delimeter for the data, such as ',' or ' '. Defaults to '        '.</p></td>
          <td>
                <code>&#39;\t&#39;</code>
          </td>
        </tr>
        <tr>
          <td><code>cache</code></td>
          <td>
                <code>bool</code>
          </td>
          <td><p>whether or not to create a cache file for the matrix. Defaults to True.</p></td>
          <td>
                <code>True</code>
          </td>
        </tr>
        <tr>
          <td><code>transpose</code></td>
          <td>
                <code>bool</code>
          </td>
          <td><p>whether or not to transpose the data matrix after loading. Defaults to True.</p></td>
          <td>
                <code>True</code>
          </td>
        </tr>
        <tr>
          <td><code>filter_set</code></td>
          <td>
                <code>set</code>
          </td>
          <td><p>a set of gene ids to keep from the data, removing those that don't match.<br />
Transposes matrix after filtering if transpose == True.</p></td>
          <td>
                <code>set()</code>
          </td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>utils/scanpy_wrappers.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">read</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">delimiter</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">cache</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">transpose</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">filter_set</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()):</span>
    <span class="sd">&quot;&quot;&quot;Reads the contained matrix into a scanpy.adata object, transposing if needed.</span>

<span class="sd">    Args:</span>
<span class="sd">        delimiter (str): a delimeter for the data, such as &#39;,&#39; or &#39;\t&#39;. Defaults to &#39;\t&#39;.</span>
<span class="sd">        cache (bool): whether or not to create a cache file for the matrix. Defaults to True.</span>
<span class="sd">        transpose (bool): whether or not to transpose the data matrix after loading. Defaults to True.</span>
<span class="sd">        filter_set (set): a set of gene ids to keep from the data, removing those that don&#39;t match.  </span>
<span class="sd">            Transposes matrix after filtering if transpose == True.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">matrixpath</span><span class="p">,</span> <span class="n">cache</span> <span class="o">=</span> <span class="n">cache</span><span class="p">,</span> <span class="n">delimiter</span> <span class="o">=</span> <span class="n">delimiter</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">filter_set</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">filter_set</span><span class="p">)]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">transpose</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="utils.scanpy_wrappers.ScanpyMetaObject.regress_scale" class="doc doc-heading">
<code class="highlight language-python"><span class="n">regress_scale</span><span class="p">(</span><span class="n">how</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;total_counts&#39;</span><span class="p">],</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Runs regression of specific features and scales data to a maximum value.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>how</code></td>
          <td>
                <code>list</code>
          </td>
          <td><p>feature to regress out. Defaults to 'total_counts'. </p></td>
          <td>
                <code>[&#39;total_counts&#39;]</code>
          </td>
        </tr>
        <tr>
          <td><code>max_value</code></td>
          <td>
                <code>float | int</code>
          </td>
          <td><p>maximum value to scale data.</p></td>
          <td>
                <code>10</code>
          </td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>utils/scanpy_wrappers.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">regress_scale</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">how</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;total_counts&#39;</span><span class="p">],</span> <span class="n">max_value</span> <span class="o">=</span> <span class="mi">10</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Runs regression of specific features and scales data to a maximum value.</span>

<span class="sd">    Args:</span>
<span class="sd">        how (list): feature to regress out. Defaults to &#39;total_counts&#39;. </span>
<span class="sd">        max_value (float | int): maximum value to scale data.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">regress_out</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">how</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">max_value</span> <span class="o">=</span> <span class="n">max_value</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="utils.scanpy_wrappers.ScanpyMetaObject.umap_leiden" class="doc doc-heading">
<code class="highlight language-python"><span class="n">umap_leiden</span><span class="p">(</span><span class="n">n_neighbors</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span> <span class="n">legend_loc</span><span class="o">=</span><span class="s1">&#39;on data&#39;</span><span class="p">,</span> <span class="n">save</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Runs Leiden clustering, followed by UMAP, on the data.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>n_neighbors</code></td>
          <td>
                <code>int</code>
          </td>
          <td><p>number of neighbors to pass to <code>sc.pp.neighbors</code>.</p></td>
          <td>
                <code>50</code>
          </td>
        </tr>
        <tr>
          <td><code>n_pcs</code></td>
          <td>
                <code>int</code>
          </td>
          <td><p>number of Principal Components to use from PCA, passed to <code>sc.pp.neighbors</code>.</p></td>
          <td>
                <code>40</code>
          </td>
        </tr>
        <tr>
          <td><code>legend_loc</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>where the legend should go (e.g. <code>'on data'</code>.)</p></td>
          <td>
                <code>&#39;on data&#39;</code>
          </td>
        </tr>
        <tr>
          <td><code>save</code></td>
          <td>
                <code>bool</code>
          </td>
          <td><p>whether to save the plot.</p></td>
          <td>
                <code>True</code>
          </td>
        </tr>
        <tr>
          <td><code>plot</code></td>
          <td>
                <code>bool</code>
          </td>
          <td><p>whether to make the plot.</p></td>
          <td>
                <code>True</code>
          </td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>utils/scanpy_wrappers.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">umap_leiden</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">n_pcs</span> <span class="o">=</span> <span class="mi">40</span><span class="p">,</span> <span class="n">legend_loc</span><span class="o">=</span><span class="s1">&#39;on data&#39;</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">plot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Runs Leiden clustering, followed by UMAP, on the data.</span>

<span class="sd">    Args:</span>
<span class="sd">        n_neighbors (int): number of neighbors to pass to `sc.pp.neighbors`.</span>
<span class="sd">        n_pcs (int): number of Principal Components to use from PCA, passed to `sc.pp.neighbors`.</span>
<span class="sd">        legend_loc (str): where the legend should go (e.g. `&#39;on data&#39;`.)</span>
<span class="sd">        save (bool): whether to save the plot.</span>
<span class="sd">        plot (bool): whether to make the plot.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="n">n_neighbors</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="n">n_pcs</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">plot</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">if</span> <span class="n">save</span><span class="p">:</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">],</span> <span class="n">legend_loc</span><span class="o">=</span><span class="n">legend_loc</span><span class="p">,</span> <span class="n">save</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">species_prefix</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">datatype</span><span class="p">,</span> <span class="s1">&#39;leiden.pdf&#39;</span><span class="p">]))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;leiden&#39;</span><span class="p">],</span> <span class="n">legend_loc</span><span class="o">=</span><span class="n">legend_loc</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="utils.scanpy_wrappers.ScanpyMetaObject.variable_filter" class="doc doc-heading">
<code class="highlight language-python"><span class="n">variable_filter</span><span class="p">(</span><span class="n">min_mean</span><span class="o">=</span><span class="mf">0.0125</span><span class="p">,</span> <span class="n">max_mean</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">min_disp</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">max_disp</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Filters genes to use for dimensionality reduction using max/min of mean and max/min of dispersion, plotting optionally.</p>
<p>Modifies underlying <code>adata</code> and saves original data under <code>adata.raw</code>.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>min_mean</code></td>
          <td>
                <code>float | int</code>
          </td>
          <td><p>minimum mean expression of genes.</p></td>
          <td>
                <code>0.0125</code>
          </td>
        </tr>
        <tr>
          <td><code>max_mean</code></td>
          <td>
                <code>float | int</code>
          </td>
          <td><p>maximum mean expression of genes.</p></td>
          <td>
                <code>3</code>
          </td>
        </tr>
        <tr>
          <td><code>min_disp</code></td>
          <td>
                <code>float | int</code>
          </td>
          <td><p>minimum dispersion of gene expression.</p></td>
          <td>
                <code>0.1</code>
          </td>
        </tr>
        <tr>
          <td><code>max_disp</code></td>
          <td>
                <code>float | int</code>
          </td>
          <td><p>maximum dispersion of gene expression.</p></td>
          <td>
                <code>10</code>
          </td>
        </tr>
        <tr>
          <td><code>plot</code></td>
          <td>
                <code>bool</code>
          </td>
          <td><p>whether or not to generate a scatter plot showing cutoffs.</p></td>
          <td>
                <code>True</code>
          </td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>utils/scanpy_wrappers.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">variable_filter</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">min_mean</span> <span class="o">=</span> <span class="mf">0.0125</span><span class="p">,</span> <span class="n">max_mean</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">min_disp</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span> <span class="n">max_disp</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span> <span class="n">plot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Filters genes to use for dimensionality reduction using max/min of mean and max/min of dispersion, plotting optionally.</span>

<span class="sd">    Modifies underlying `adata` and saves original data under `adata.raw`.</span>

<span class="sd">    Args:</span>
<span class="sd">        min_mean (float | int): minimum mean expression of genes.</span>
<span class="sd">        max_mean (float | int): maximum mean expression of genes.</span>
<span class="sd">        min_disp (float | int): minimum dispersion of gene expression.</span>
<span class="sd">        max_disp (float | int): maximum dispersion of gene expression.</span>
<span class="sd">        plot (bool): whether or not to generate a scatter plot showing cutoffs.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_mean</span> <span class="o">=</span> <span class="n">min_mean</span><span class="p">,</span> <span class="n">max_mean</span><span class="o">=</span><span class="n">max_mean</span><span class="p">,</span> <span class="n">min_disp</span><span class="o">=</span><span class="n">min_disp</span><span class="p">,</span> <span class="n">max_disp</span><span class="o">=</span><span class="n">max_disp</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">plot</span><span class="p">:</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">raw</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">adata</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">[:,</span> <span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">highly_variable</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>

<div class="doc doc-object doc-function">



<h3 id="utils.scanpy_wrappers.ScanpyMetaObject.violin" class="doc doc-heading">
<code class="highlight language-python"><span class="n">violin</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;total_counts&#39;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">,</span> <span class="n">plot</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

</h3>


  <div class="doc doc-contents ">
  
      <p>Runs sc.pp.calculate_qc_metrics, sc.pl.violin, and sc.pl.scatter.</p>

  <p><strong>Parameters:</strong></p>
  <table>
    <thead>
      <tr>
        <th>Name</th>
        <th>Type</th>
        <th>Description</th>
        <th>Default</th>
      </tr>
    </thead>
    <tbody>
        <tr>
          <td><code>x</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>label of the feature to be plotted on the x axis of the violin &amp; scatter plot.<br />
Defaults to 'total_counts'.</p></td>
          <td>
                <code>&#39;total_counts&#39;</code>
          </td>
        </tr>
        <tr>
          <td><code>y</code></td>
          <td>
                <code>str</code>
          </td>
          <td><p>label of the feature to be plotted on the y axis of the violin &amp; scatter plot.
Defaults to 'n_genes_by_counts'.</p></td>
          <td>
                <code>&#39;n_genes_by_counts&#39;</code>
          </td>
        </tr>
    </tbody>
  </table>

      <details class="quote">
        <summary>Source code in <code>utils/scanpy_wrappers.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">violin</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span> <span class="o">=</span> <span class="s1">&#39;total_counts&#39;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s1">&#39;n_genes_by_counts&#39;</span><span class="p">,</span> <span class="n">plot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Runs sc.pp.calculate_qc_metrics, sc.pl.violin, and sc.pl.scatter.</span>

<span class="sd">    Args:</span>
<span class="sd">        x (str): label of the feature to be plotted on the x axis of the violin &amp; scatter plot.  </span>
<span class="sd">            Defaults to &#39;total_counts&#39;.</span>
<span class="sd">        y (str): label of the feature to be plotted on the y axis of the violin &amp; scatter plot.</span>
<span class="sd">            Defaults to &#39;n_genes_by_counts&#39;.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">calculate_qc_metrics</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">percent_top</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">log1p</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">plot</span><span class="p">:</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">violin</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="p">[</span><span class="n">y</span><span class="p">,</span> <span class="n">x</span><span class="p">],</span>
         <span class="n">jitter</span><span class="o">=</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">multi_panel</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">adata</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>



  </div>

  </div>

</div>


<div class="doc doc-object doc-function">



<h2 id="utils.scanpy_wrappers.diagonalize_df" class="doc doc-heading">
<code class="highlight language-python"><span class="n">diagonalize_df</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Sorts values of a 2d dataframe along their diagonal for prettier plotting.</p>

      <details class="quote">
        <summary>Source code in <code>utils/scanpy_wrappers.py</code></summary>
        <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">diagonalize_df</span><span class="p">(</span><span class="n">df</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Sorts values of a 2d dataframe along their diagonal for prettier plotting.&quot;&quot;&quot;</span>
    <span class="n">max_order</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">])</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">i</span><span class="p">]))</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">index</span><span class="p">]</span>
    <span class="n">reordered</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">copy</span><span class="p">(</span><span class="n">deep</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
    <span class="n">reordered</span><span class="p">[</span><span class="s1">&#39;max_col&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">max_order</span>
    <span class="n">reordered</span> <span class="o">=</span> <span class="n">reordered</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="s1">&#39;index&#39;</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s1">&#39;max_col&#39;</span><span class="p">)</span>
    <span class="n">reordered</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;max_col&#39;</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">reordered</span>
</code></pre></div></td></tr></table></div>
      </details>
  </div>

</div>



  </div>

  </div>

</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../string_functions_doc/" class="btn btn-neutral float-left" title="String Functions Usage"><span class="icon icon-circle-arrow-left"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../string_functions_doc/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
  </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme_extra.js" defer></script>
    <script src="../js/theme.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
