{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c8afc416-e911-44b6-bb5c-19a25e64c74f",
   "metadata": {},
   "source": [
    "# Overview\n",
    "\n",
    "This notebook downloads cell annotation from the [Jiang et al. 2021](https://www.frontiersin.org/articles/10.3389/fcell.2021.743421/full) zebrafish cell atlas for the sample `Brain_8`.\n",
    "\n",
    "\n",
    "## Guidelines\n",
    "\n",
    "Due to the wide variability in potential formats for these filetypes, it is difficult to generate a standardized analysis pipeline for these files.  \n",
    "For downstream analysis purposes, we'll coerce celltype annotation information into a standard format.\n",
    "\n",
    "The final output generated by this script should be a .tsv file with the following format:\n",
    "\n",
    "| cell_barcode | celltype |\n",
    "|--------------|----------|\n",
    "| ATGCAGATACAC | pyramidal cell |\n",
    "| CCATACAGACTA | neuroblast |\n",
    "| TTTCAAAGACAG | astrocyte |\n",
    "| ... | ... |\n",
    "\n",
    "The final datatype expects two columns: `cell_barcode` and `celltype`.  \n",
    "The `cell_barcode` names should be properly matched to the same format as the `gxc` file.  \n",
    "The `celltype` can have any format, but should include cell identity information.  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "953ac006-8f59-49f1-856b-0c9d18b7b602",
   "metadata": {},
   "source": [
    "# 0. Setup\n",
    "\n",
    "Import packages and specify any important functions here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "58f78eca-7ccb-4e60-b6ca-bdc3acb26d7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import standard python packages\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import os\n",
    "import dill\n",
    "import datetime\n",
    "\n",
    "# add the utils and env directories to the path\n",
    "import sys\n",
    "sys.path.append('../../utils/')\n",
    "sys.path.append('../../env/')\n",
    "\n",
    "# import functions from utils directory files\n",
    "from string_functions import *\n",
    "from biofile_handling import *\n",
    "\n",
    "# import paths to software installs from env\n",
    "from install_locs import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "883bb4a4-e467-4882-9586-f7fa488664ae",
   "metadata": {},
   "source": [
    "# 1. Load BioFileDocket\n",
    "\n",
    "Load in the BioFileDocket file for this dataset using `BioFileDocket.unpickle()`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8ddbd0a5-f4e0-4321-9e8d-5a12c5bb386b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/glial-origins/output/Drer_adultbrain/ already exists\n",
      "Files will be saved into /home/ec2-user/glial-origins/output/Drer_adultbrain/\n",
      "file Drer_adultbrain_BioFileDocket.pkl already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_BioFileDocket.pkl\n",
      "file GCF_000002035.5_GRCz10_genomic.gff already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.gff\n",
      "file GCF_000002035.5_GRCz10_genomic.fna already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic.fna\n",
      "file GSM3768152_Brain_8_dge.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GSM3768152_Brain_8_dge.txt\n",
      "file Drer_adultbrain_ZFIN_UniProtIDs.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_ZFIN_UniProtIDs.txt\n",
      "file Drer_adultbrain_uniprot-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_uniprot-idmm.tsv\n",
      "file Drer_adultbrain_gtf-idmm.tsv already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_gtf-idmm.tsv\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3\n",
      "file GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep\n",
      "file Drer_adultbrain_og_idmm.tsv already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_og_idmm.tsv\n",
      "file OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv\n",
      "file GSM3768152_Brain_8_dge_asOrthogroup.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GSM3768152_Brain_8_dge_asOrthogroup.txt\n",
      "file DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_OrthoFinder/DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv\n",
      "file Drer_taxid_genes.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_taxid_genes.txt\n",
      "file Drer_adultbrain_sc_idmm.tsv already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/Drer_adultbrain_sc_idmm.tsv\n",
      "file DrerMmusXlae_adultbrain_struclusters_file.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_struclusters_file.tsv\n",
      "file GSM3768152_Brain_8_dge_asStruCluster.txt already exists at /home/ec2-user/glial-origins/output/Drer_adultbrain/GSM3768152_Brain_8_dge_asStruCluster.txt\n",
      "file DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv already exists at /home/ec2-user/glial-origins/output/DrerMmusXlae_adultbrain_FoldSeek/DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv\n"
     ]
    }
   ],
   "source": [
    "################\n",
    "# general info #\n",
    "################\n",
    "\n",
    "# Specify the name of the species folder in Amazon S3\n",
    "species = 'Danio_rerio'\n",
    "\n",
    "# Specify any particular identifying conditions, eg tissue type:\n",
    "conditions = 'adultbrain'\n",
    "\n",
    "################\n",
    "\n",
    "species_BFD = BioFileDocket(species, conditions).get_from_s3().unpickle()\n",
    "species_BFD.s3_to_local()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "008f0b2e-1a6e-44ec-af36-25cb303ba0d1",
   "metadata": {},
   "source": [
    "# 2. Download cell annotation file and load in raw file\n",
    "\n",
    "Using `subprocess.run`, download the file using an appropriate protocol and load it as a pandas DataFrame.  \n",
    "For .csv and .tsv files, use `pd.read_csv`.  \n",
    "For .xlsx and .xls files, use `pd.read_excel`.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d604e913-34cd-49ba-bef0-287d0217ac9f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell barcode</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Batch</th>\n",
       "      <th>cell_barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Brain_1.TTCCGCAAGTACTGTGCG</td>\n",
       "      <td>Brain_cluster3</td>\n",
       "      <td>Brain.Microglia</td>\n",
       "      <td>Brain_1</td>\n",
       "      <td>TTCCGCAAGTACTGTGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Brain_1.TTCATATGTGCGTTTAGG</td>\n",
       "      <td>Brain_cluster3</td>\n",
       "      <td>Brain.Microglia</td>\n",
       "      <td>Brain_1</td>\n",
       "      <td>TTCATATGTGCGTTTAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Brain_1.TGCGGAACTTATGCTCAA</td>\n",
       "      <td>Brain_cluster3</td>\n",
       "      <td>Brain.Microglia</td>\n",
       "      <td>Brain_1</td>\n",
       "      <td>TGCGGAACTTATGCTCAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Brain_1.TGCGGAGTTGCCGCGTCC</td>\n",
       "      <td>Brain_cluster3</td>\n",
       "      <td>Brain.Microglia</td>\n",
       "      <td>Brain_1</td>\n",
       "      <td>TGCGGAGTTGCCGCGTCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Brain_1.GCAGGAGAATTATGATCA</td>\n",
       "      <td>Brain_cluster3</td>\n",
       "      <td>Brain.Microglia</td>\n",
       "      <td>Brain_1</td>\n",
       "      <td>GCAGGAGAATTATGATCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201089</th>\n",
       "      <td>Z72h2.CATCCCCAACAACGTATT</td>\n",
       "      <td>Z72h_cluster1</td>\n",
       "      <td>Z72h.Hatching gland</td>\n",
       "      <td>Z72h2</td>\n",
       "      <td>CATCCCCAACAACGTATT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201090</th>\n",
       "      <td>Z72h2.CCAGACCCGACGAGCGAG</td>\n",
       "      <td>Z72h_cluster1</td>\n",
       "      <td>Z72h.Hatching gland</td>\n",
       "      <td>Z72h2</td>\n",
       "      <td>CCAGACCCGACGAGCGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201091</th>\n",
       "      <td>Z72h2.CCGCTAACACCCAAAGTT</td>\n",
       "      <td>Z72h_cluster1</td>\n",
       "      <td>Z72h.Hatching gland</td>\n",
       "      <td>Z72h2</td>\n",
       "      <td>CCGCTAACACCCAAAGTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201092</th>\n",
       "      <td>Z72h2.CGGCAGGGACATCTGTGT</td>\n",
       "      <td>Z72h_cluster1</td>\n",
       "      <td>Z72h.Hatching gland</td>\n",
       "      <td>Z72h2</td>\n",
       "      <td>CGGCAGGGACATCTGTGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201093</th>\n",
       "      <td>Z72h2.CTCCATATCTCTTGCGGA</td>\n",
       "      <td>Z72h_cluster8</td>\n",
       "      <td>Z72h.Pancreas cell</td>\n",
       "      <td>Z72h2</td>\n",
       "      <td>CTCCATATCTCTTGCGGA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>201094 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      cell barcode         Cluster           Annotation  \\\n",
       "0       Brain_1.TTCCGCAAGTACTGTGCG  Brain_cluster3      Brain.Microglia   \n",
       "1       Brain_1.TTCATATGTGCGTTTAGG  Brain_cluster3      Brain.Microglia   \n",
       "2       Brain_1.TGCGGAACTTATGCTCAA  Brain_cluster3      Brain.Microglia   \n",
       "3       Brain_1.TGCGGAGTTGCCGCGTCC  Brain_cluster3      Brain.Microglia   \n",
       "4       Brain_1.GCAGGAGAATTATGATCA  Brain_cluster3      Brain.Microglia   \n",
       "...                            ...             ...                  ...   \n",
       "201089    Z72h2.CATCCCCAACAACGTATT   Z72h_cluster1  Z72h.Hatching gland   \n",
       "201090    Z72h2.CCAGACCCGACGAGCGAG   Z72h_cluster1  Z72h.Hatching gland   \n",
       "201091    Z72h2.CCGCTAACACCCAAAGTT   Z72h_cluster1  Z72h.Hatching gland   \n",
       "201092    Z72h2.CGGCAGGGACATCTGTGT   Z72h_cluster1  Z72h.Hatching gland   \n",
       "201093    Z72h2.CTCCATATCTCTTGCGGA   Z72h_cluster8   Z72h.Pancreas cell   \n",
       "\n",
       "          Batch        cell_barcode  \n",
       "0       Brain_1  TTCCGCAAGTACTGTGCG  \n",
       "1       Brain_1  TTCATATGTGCGTTTAGG  \n",
       "2       Brain_1  TGCGGAACTTATGCTCAA  \n",
       "3       Brain_1  TGCGGAGTTGCCGCGTCC  \n",
       "4       Brain_1  GCAGGAGAATTATGATCA  \n",
       "...         ...                 ...  \n",
       "201089    Z72h2  CATCCCCAACAACGTATT  \n",
       "201090    Z72h2  CCAGACCCGACGAGCGAG  \n",
       "201091    Z72h2  CCGCTAACACCCAAAGTT  \n",
       "201092    Z72h2  CGGCAGGGACATCTGTGT  \n",
       "201093    Z72h2  CTCCATATCTCTTGCGGA  \n",
       "\n",
       "[201094 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GENE</th>\n",
       "      <th>ACAATATATTGTACCTGA</th>\n",
       "      <th>ACGTTGATGGCGTAGAGA</th>\n",
       "      <th>AACCTAACCTGAATTTGC</th>\n",
       "      <th>CTCGCAGCCCTCTATGTA</th>\n",
       "      <th>ACGTTGCGTATTTAGTCG</th>\n",
       "      <th>AACCTATAGAGACCGACG</th>\n",
       "      <th>ACGAGCGCTGTGGCCTAG</th>\n",
       "      <th>GCGAATGGACATGGACAT</th>\n",
       "      <th>TCTACCGCTCAAGCTCAA</th>\n",
       "      <th>...</th>\n",
       "      <th>CGGCAGTCAAAGATCTCT</th>\n",
       "      <th>GACACTGCGAATCTGTGT</th>\n",
       "      <th>GCAGGAGGCTGCTAAGGG</th>\n",
       "      <th>TATGTATACTTCCGCACC</th>\n",
       "      <th>TGGATGTTCCGCACAATA</th>\n",
       "      <th>AACCTATGGATGGGGTTT</th>\n",
       "      <th>AAGCGGAGGACTCTCCAT</th>\n",
       "      <th>ACCTGACTCGCAAGCGAG</th>\n",
       "      <th>ACGTTGCAAAGTTTCATA</th>\n",
       "      <th>ATCAACTGCAATTTCCGC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCF3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACOT12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACSF3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACTC1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACVR1C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ADAM12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ADAMTSL4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ADGRL3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AKAP13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AL590134.1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 10001 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         GENE  ACAATATATTGTACCTGA  ACGTTGATGGCGTAGAGA  AACCTAACCTGAATTTGC  \\\n",
       "0       ABCF3                   0                   0                   0   \n",
       "1      ACOT12                   0                   0                   0   \n",
       "2       ACSF3                   0                   0                   0   \n",
       "3       ACTC1                   0                   0                   0   \n",
       "4      ACVR1C                   0                   0                   0   \n",
       "5      ADAM12                   0                   0                   0   \n",
       "6    ADAMTSL4                   0                   0                   0   \n",
       "7      ADGRL3                   0                   0                   0   \n",
       "8      AKAP13                   0                   0                   0   \n",
       "9  AL590134.1                   0                   0                   0   \n",
       "\n",
       "   CTCGCAGCCCTCTATGTA  ACGTTGCGTATTTAGTCG  AACCTATAGAGACCGACG  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   ACGAGCGCTGTGGCCTAG  GCGAATGGACATGGACAT  TCTACCGCTCAAGCTCAA  ...  \\\n",
       "0                   0                   0                   0  ...   \n",
       "1                   0                   0                   0  ...   \n",
       "2                   0                   0                   0  ...   \n",
       "3                   0                   0                   0  ...   \n",
       "4                   0                   0                   0  ...   \n",
       "5                   0                   0                   0  ...   \n",
       "6                   0                   0                   0  ...   \n",
       "7                   0                   0                   0  ...   \n",
       "8                   0                   0                   0  ...   \n",
       "9                   1                   0                   0  ...   \n",
       "\n",
       "   CGGCAGTCAAAGATCTCT  GACACTGCGAATCTGTGT  GCAGGAGGCTGCTAAGGG  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   TATGTATACTTCCGCACC  TGGATGTTCCGCACAATA  AACCTATGGATGGGGTTT  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   AAGCGGAGGACTCTCCAT  ACCTGACTCGCAAGCGAG  ACGTTGCAAAGTTTCATA  \\\n",
       "0                   0                   0                   0   \n",
       "1                   0                   0                   0   \n",
       "2                   0                   0                   0   \n",
       "3                   0                   0                   0   \n",
       "4                   0                   0                   0   \n",
       "5                   0                   0                   0   \n",
       "6                   0                   0                   0   \n",
       "7                   0                   0                   0   \n",
       "8                   0                   0                   0   \n",
       "9                   0                   0                   0   \n",
       "\n",
       "   ATCAACTGCAATTTCCGC  \n",
       "0                   0  \n",
       "1                   0  \n",
       "2                   0  \n",
       "3                   0  \n",
       "4                   0  \n",
       "5                   0  \n",
       "6                   0  \n",
       "7                   0  \n",
       "8                   0  \n",
       "9                   0  \n",
       "\n",
       "[10 rows x 10001 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Specify file url and output information\n",
    "file_url = 'https://ndownloader.figstatic.com/files/30949762'\n",
    "output_filename = 'Drer_cellannot.xlsx'\n",
    "output_filepath = species_BFD.sampledict.directory + output_filename\n",
    "\n",
    "subprocess.run(['wget', file_url, '-O', output_filepath],\n",
    "               stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)\n",
    "\n",
    "cell_annots = pd.read_excel(output_filepath, sheet_name= 'cell barcode of each cell')\n",
    "cell_annots['Batch'] = cell_annots['cell barcode'].str.split('.', expand = True)[0]\n",
    "cell_annots['cell_barcode'] = cell_annots['cell barcode'].str.split('.', expand = True)[1]\n",
    "display(cell_annots)\n",
    "\n",
    "# Load in gxc matrix to check proper cell names\n",
    "gxc = pd.read_csv(species_BFD.gxc.path, sep = '\\t', nrows = 10)\n",
    "display(gxc)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f892ee12-7439-49c4-8eb2-6296cb8261ef",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 3. Extract and coerce data into the correct format. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1e1acbb5-5c14-470b-b51f-6b1e368a2b66",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_75932/1596347762.py:13: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  sample_cell_annots['celltype'] = sample_cell_annots['celltype'].str.replace('Brain\\.', '')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>celltype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3400</th>\n",
       "      <td>ATCTCTGCTCAAAAAACG</td>\n",
       "      <td>Microglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3401</th>\n",
       "      <td>TAAGGGACGTTGATTCCA</td>\n",
       "      <td>Microglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3402</th>\n",
       "      <td>GCAGGAACAATACGGCAG</td>\n",
       "      <td>Microglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3403</th>\n",
       "      <td>GCGAATTGTGCGGCGTCC</td>\n",
       "      <td>Microglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3404</th>\n",
       "      <td>AACCTACTGTGTGAGATC</td>\n",
       "      <td>Microglia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20786</th>\n",
       "      <td>TGATCAGGGCGAATACAG</td>\n",
       "      <td>Innate_immune_cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20787</th>\n",
       "      <td>TTGGACGCTCAAGTCCCG</td>\n",
       "      <td>Innate_immune_cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20788</th>\n",
       "      <td>AAAACGTCTACCCTCGCA</td>\n",
       "      <td>Innate_immune_cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20789</th>\n",
       "      <td>TAAGGGAGGACTGCGTCC</td>\n",
       "      <td>Innate_immune_cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20790</th>\n",
       "      <td>GTCCCGGTCGGTCAAAGT</td>\n",
       "      <td>Innate_immune_cell</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17176 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             cell_barcode            celltype\n",
       "3400   ATCTCTGCTCAAAAAACG           Microglia\n",
       "3401   TAAGGGACGTTGATTCCA           Microglia\n",
       "3402   GCAGGAACAATACGGCAG           Microglia\n",
       "3403   GCGAATTGTGCGGCGTCC           Microglia\n",
       "3404   AACCTACTGTGTGAGATC           Microglia\n",
       "...                   ...                 ...\n",
       "20786  TGATCAGGGCGAATACAG  Innate_immune_cell\n",
       "20787  TTGGACGCTCAAGTCCCG  Innate_immune_cell\n",
       "20788  AAAACGTCTACCCTCGCA  Innate_immune_cell\n",
       "20789  TAAGGGAGGACTGCGTCC  Innate_immune_cell\n",
       "20790  GTCCCGGTCGGTCAAAGT  Innate_immune_cell\n",
       "\n",
       "[17176 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Specify sample name for filtering\n",
    "sample_name = 'Brain_8'\n",
    "\n",
    "# Get cells from the Batch matching the sample_name\n",
    "sample_cell_annots = cell_annots[cell_annots['Batch'] == sample_name]\n",
    "\n",
    "# Extract cell barcode and celltype info\n",
    "sample_cell_annots = sample_cell_annots[['cell_barcode', 'Annotation']]\n",
    "\n",
    "# Rename columns to have the correct header\n",
    "sample_cell_annots.rename(columns = {'Annotation': 'celltype'}, inplace = True)\n",
    "sample_cell_annots.drop_duplicates(subset = 'cell_barcode', inplace = True)\n",
    "sample_cell_annots['celltype'] = sample_cell_annots['celltype'].str.replace('Brain\\.', '')\n",
    "display(sample_cell_annots)\n",
    "\n",
    "# Generate filename\n",
    "output_filename = '_'.join([prefixify(species_BFD.species), species_BFD.conditions, sample_name, 'cellannot.tsv'])\n",
    "\n",
    "# Generate BioFile object\n",
    "output_CellAnnotFile = CellAnnotFile(\n",
    "    filename = output_filename,\n",
    "    sampledict = species_BFD.sampledict,\n",
    "    sources = [species_BFD.gxc]\n",
    ")\n",
    "\n",
    "# Export file and add to BioFileDocket\n",
    "sample_cell_annots.to_csv(output_CellAnnotFile.path, sep = '\\t', index = None)\n",
    "species_BFD.add_keyfile('cellannot', output_CellAnnotFile)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f853053c-6713-4671-b9c1-fec83d1557f5",
   "metadata": {},
   "source": [
    "# 4. Update BioFileDocket and push files to S3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "709a46f7-256e-4c8f-909f-edaea94620f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCF_000002035.5_GRCz10_genomic.gff already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM3768152_Brain_8_dge.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_ZFIN_UniProtIDs.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_uniprot-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_gtf-idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.bed already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.cds already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.gff3 already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GCF_000002035.5_GRCz10_genomic_cDNA.fna.transdecoder.pep already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_og_idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "OrthoFinder/Results_Sep28/Orthogroups/Orthogroups.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM3768152_Brain_8_dge_asOrthogroup.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_asOrthogroup_combined_exc.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_taxid_genes.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_sc_idmm.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_struclusters_file.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "GSM3768152_Brain_8_dge_asStruCluster.txt already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "DrerMmusXlae_adultbrain_asStruCluster_combined_exc.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_Brain_8_cellannot.tsv already exists in S3 bucket, skipping upload. set overwrite = True to overwrite the existing file.\n",
      "Drer_adultbrain_BioFileDocket.pkl already exists in S3 bucket; overwriting.\n",
      "upload: ../../output/Drer_adultbrain/Drer_adultbrain_BioFileDocket.pkl to s3://arcadia-reference-datasets/glial-origins-pkl/Drer_adultbrain_BioFileDocket.pkl\n"
     ]
    }
   ],
   "source": [
    "species_BFD.local_to_s3()\n",
    "species_BFD.pickle()\n",
    "species_BFD.push_to_s3(overwrite = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5533452-cf4e-41de-9e0d-6d9901d9f14d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a237ba26-2fa7-4bb6-b33a-85afd4e016bc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
